<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言 我们已经通过ExpressionSet类和BamFileList文件（含有BAM比对后的文件）来了解了一些关于数据管理的基本知识。  S4类有助于控制某个数据结构中不同组件之间的复杂程度和一致性。 R包能够进一步用于协调和发布数据、软件程度和文档。在这个课程的后面部分里我们会学习如何构建R包。  在这个一部分的结尾处，我们来解决一个明显没有最佳解决方案的问题。我们考虑打包GRan">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS007-管理与处理大量BED文件">
<meta property="og:url" content="http://rvdsd.top/2019/09/07/Genomics Data Analysis Series/GDAS007-管理与处理大量BED文件/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 我们已经通过ExpressionSet类和BamFileList文件（含有BAM比对后的文件）来了解了一些关于数据管理的基本知识。  S4类有助于控制某个数据结构中不同组件之间的复杂程度和一致性。 R包能够进一步用于协调和发布数据、软件程度和文档。在这个课程的后面部分里我们会学习如何构建R包。  在这个一部分的结尾处，我们来解决一个明显没有最佳解决方案的问题。我们考虑打包GRan">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkxy-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkbr-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkbr2-1.png">
<meta property="og:updated_time" content="2019-11-07T11:29:44.124Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS007-管理与处理大量BED文件">
<meta name="twitter:description" content="前言 我们已经通过ExpressionSet类和BamFileList文件（含有BAM比对后的文件）来了解了一些关于数据管理的基本知识。  S4类有助于控制某个数据结构中不同组件之间的复杂程度和一致性。 R包能够进一步用于协调和发布数据、软件程度和文档。在这个课程的后面部分里我们会学习如何构建R包。  在这个一部分的结尾处，我们来解决一个明显没有最佳解决方案的问题。我们考虑打包GRan">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkxy-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/07/Genomics Data Analysis Series/GDAS007-管理与处理大量BED文件/"/>





  <title>GDAS007-管理与处理大量BED文件 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/07/Genomics Data Analysis Series/GDAS007-管理与处理大量BED文件/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS007-管理与处理大量BED文件</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-07T12:00:00+08:00">
                2019-09-07
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,782
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="前言">前言</h1>
<p>我们已经通过<code>ExpressionSet</code>类和<code>BamFileList</code>文件（含有BAM比对后的文件）来了解了一些关于数据管理的基本知识。</p>
<ul>
<li>S4类有助于控制某个数据结构中不同组件之间的复杂程度和一致性。</li>
<li>R包能够进一步用于协调和发布数据、软件程度和文档。在这个课程的后面部分里我们会学习如何构建R包。</li>
</ul>
<p>在这个一部分的结尾处，我们来解决一个明显没有最佳解决方案的问题。我们考虑打包GRanges的序列化后的实例以及已经建立了索引的压缩文本文件，这些文件都可以使用GenomicFiles进行并行计算处理。</p>
<p>为什么这是一个比较难解决的问题呢？因为在某些情况下，含有大量功能的GRanges实例将会非常大，加载它们可能会花费大量的时间，并且会占用大量的内存。这可能导致在RAM有限时，在群集节点上使用这些表示方法出现困难。如果数据以允许随机访问的索引格式保存在磁盘上，则根据需要将文件内容的目标提取，并转换为GRanges，就可能为某些应用程度带来更好的性能。创建<code>erma</code>包的一个原因就是获取关于这些不同方法之间平衡后的数据。</p>
<h1 id="erma表观遗传路线图简介">erma：表观遗传路线图简介</h1>
<p>为了促进对DNA突变体的理解，我们对细胞类型之间的表观基因组变异有所兴趣。NIH表观基因组路线图收集了很多关于ChIP-seq实验分析的数据，在这个公众数据库里记录了数百个不同细胞类型和标志特异性文件供其他研究者使用。</p>
<p>管理这类数据的有效方法是什么呢？其中<a href="http://www.bioconductor.org/packages/erma" target="_blank" rel="external">erma</a> 包就提供了这样一种方法。</p>
<h2 id="bed文件集合">BED文件集合</h2>
<p>如同Ernst和Kellis使用大量细胞系的数据所报道的那样，他们已经进行了染色质状态的统计“估算”，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(erma)</div><div class="line">ef = dir(system.file(<span class="string">"bed_tabix"</span>, package=<span class="string">"erma"</span>), patt=<span class="string">"bed.gz$"</span>)</div><div class="line">length(ef)</div><div class="line"><span class="comment">## [1] 31</span></div><div class="line">head(ef)</div><div class="line"><span class="comment">## [1] "E002_25_imputed12marks_mnemonics.bed.gz"</span></div><div class="line"><span class="comment">## [2] "E003_25_imputed12marks_mnemonics.bed.gz"</span></div><div class="line"><span class="comment">## [3] "E021_25_imputed12marks_mnemonics.bed.gz"</span></div><div class="line"><span class="comment">## [4] "E032_25_imputed12marks_mnemonics.bed.gz"</span></div><div class="line"><span class="comment">## [5] "E033_25_imputed12marks_mnemonics.bed.gz"</span></div><div class="line"><span class="comment">## [6] "E034_25_imputed12marks_mnemonics.bed.gz"</span></div></pre></td></tr></table></figure>
<h2 id="使用genomicfiles进行管理">使用GenomicFiles进行管理</h2>
<p>我们认为文件名是不变的，但是储存BED文件路径的向量则可以被命名得更详细。通过<code>makeErmaSet</code>函数可以获取元数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div></pre></td><td class="code"><pre><div class="line">mm = makeErmaSet()</div><div class="line"><span class="comment">## <span class="doctag">NOTE:</span> input data had non-ASCII characters replaced by ' '.</span></div><div class="line">mm</div><div class="line"><span class="comment">## ErmaSet object with 0 ranges and 31 files: </span></div><div class="line"><span class="comment">## files: E002_25_imputed12marks_mnemonics.bed.gz, E003_25_imputed12marks_mnemonics.bed.gz, ..., E088_25_imputed12marks_mnemonics.bed.gz, E096_25_imputed12marks_mnemonics.bed.gz </span></div><div class="line"><span class="comment">## detail: use files(), rowRanges(), colData(), ...</span></div><div class="line">head(colData(mm))</div><div class="line"><span class="comment">## (showing narrow slice of  6 x 95  DataFrame)   narrDF with 6 rows and 6 columns</span></div><div class="line"><span class="comment">##      Epigenome.ID..EID.          GROUP Epigenome.Mnemonic     ANATOMY</span></div><div class="line"><span class="comment">##             &lt;character&gt;    &lt;character&gt;        &lt;character&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">## E002               E002            ESC            ESC.WA7         ESC</span></div><div class="line"><span class="comment">## E003               E003            ESC             ESC.H1         ESC</span></div><div class="line"><span class="comment">## E021               E021           iPSC        IPSC.DF.6.9        IPSC</span></div><div class="line"><span class="comment">## E032               E032   HSC &amp; B-cell       BLD.CD19.PPC       BLOOD</span></div><div class="line"><span class="comment">## E033               E033 Blood &amp; T-cell        BLD.CD3.CPC       BLOOD</span></div><div class="line"><span class="comment">## E034               E034 Blood &amp; T-cell        BLD.CD3.PPC       BLOOD</span></div><div class="line"><span class="comment">##                TYPE         SEX</span></div><div class="line"><span class="comment">##         &lt;character&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">## E002 PrimaryCulture      Female</span></div><div class="line"><span class="comment">## E003 PrimaryCulture        Male</span></div><div class="line"><span class="comment">## E021 PrimaryCulture        Male</span></div><div class="line"><span class="comment">## E032    PrimaryCell        Male</span></div><div class="line"><span class="comment">## E033    PrimaryCell     Unknown</span></div><div class="line"><span class="comment">## E034    PrimaryCell        Male</span></div><div class="line"><span class="comment">## use colnames() for full set of metadata attributes.</span></div><div class="line">names(files(mm)) = colData(mm)$Epigenome.Mnemonic</div><div class="line">head(files(mm))</div><div class="line"><span class="comment">##                                                                                                                 ESC.WA7 </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E002_25_imputed12marks_mnemonics.bed.gz" </span></div><div class="line"><span class="comment">##                                                                                                                  ESC.H1 </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E003_25_imputed12marks_mnemonics.bed.gz" </span></div><div class="line"><span class="comment">##                                                                                                             IPSC.DF.6.9 </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E021_25_imputed12marks_mnemonics.bed.gz" </span></div><div class="line"><span class="comment">##                                                                                                            BLD.CD19.PPC </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E032_25_imputed12marks_mnemonics.bed.gz" </span></div><div class="line"><span class="comment">##                                                                                                             BLD.CD3.CPC </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E033_25_imputed12marks_mnemonics.bed.gz" </span></div><div class="line"><span class="comment">##                                                                                                             BLD.CD3.PPC </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.3/Resources/library/erma/bed_tabix/E034_25_imputed12marks_mnemonics.bed.gz"</span></div></pre></td></tr></table></figure>
<h1 id="使用表格描述染色质状态信息">使用表格描述染色质状态信息</h1>
<h2 id="定义染色体状态与标记概述">定义染色体状态与标记概述</h2>
<p>下表显示了不同类型的染色质标记是如何组合并用于判断染色体质片段的状态的。那些缩写的意义为：活跃TSS(TssA)，两种类型的上下流启动子(PromU, PromD1, PromD2)，转录为5’或3‘优先或略强或弱(Tx5’, Tx3’, Tx, TxWk)，联合转录或调控(TxReg)，优先转录，增强子活性和可能弱(TxEnh5‘，TxEnh3‘，TxEnhW)，两种类型的活跃增强子，可能是侧翼或弱的(EnhA1，EnhA2，EnhAF，EnhW1，EnhW2)或由H3K27ac(EnhAc)，初级DNase(DNase)，ZNF基因和重复序列(ZNF/rpts)，异染色质(Het)，二价或平衡型启动子(Propp，PromBiv)，抑制多梳(ReprPC)或静默(Quies)。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(png)</div><div class="line">im = readPNG(system.file(<span class="string">"pngs/emparms.png"</span>, package=<span class="string">"erma"</span>))</div><div class="line">grid.raster(im)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkxy-1.png" alt="plot of chunk lkxy">
<p class="caption">plot of chunk lkxy</p>
</div>
<p><code>erma</code>包中含有的文件包括31个不同细胞系和基因组的信息（原文里面是tilings，这个我的理解就是表示了一些信息），并根据上面显示了状态图将这些信息分配给染色体状态。</p>
<p>下面展示了两个基因BRCA2和Eome启动子区域的状态标记，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">stateProfile(mm, <span class="string">"BRCA2"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">## Warning: executing %dopar% sequentially: no parallel backend registered</span></div><div class="line"><span class="comment">## Scale for 'y' is already present. Adding another scale for 'y', which</span></div><div class="line"><span class="comment">## will replace the existing scale.</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkbr-1.png" alt="plot of chunk lkbr">
<p class="caption">plot of chunk lkbr</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">stateProfile(mm, <span class="string">"EOMES"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">## Scale for 'y' is already present. Adding another scale for 'y', which</span></div><div class="line"><span class="comment">## will replace the existing scale.</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_multibed-lkbr2-1.png" alt="plot of chunk lkbr2">
<p class="caption">plot of chunk lkbr2</p>
</div>
<p>有两个基因在启动子区域的状态与常规的状态不同，以及与细胞类型的一致性方面也不同。</p>
<h2 id="并行目标查询">并行目标查询</h2>
<p>虽然可视化很直观，但是我们也希望能够进行文本编辑，从而提取一些信息供下游使用。在这里我们简单地将每个细胞类型的状态制成表格。使用<code>readuceByFile</code>函数就能将文件分配给可用的内核或主机进行并行处理，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line">gm = promoters(range(genemodel(<span class="string">"BRCA2"</span>))) <span class="comment"># 2000 upstream, 200 down by default</span></div><div class="line"><span class="keyword">library</span>(BiocParallel)</div><div class="line">register(MulticoreParam(workers=<span class="number">2</span>))</div><div class="line"><span class="keyword">library</span>(GenomicFiles)</div><div class="line">ans = reduceByFile( gm, files(mm), MAP=<span class="keyword">function</span>(range,file) &#123;</div><div class="line">   table( import(file, genome=<span class="string">"hg19"</span>, which=range)$name ) &#125; )</div><div class="line">ans= unlist(ans, recursive=<span class="literal">FALSE</span>)</div><div class="line">names(ans) = colData(mm)$Epigenome.Mnemonic</div><div class="line">ans[<span class="number">1</span>:<span class="number">4</span>]</div><div class="line"><span class="comment">## $$ESC.WA7</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   1_TssA 16_EnhW1  2_PromU 25_Quies </span></div><div class="line"><span class="comment">##        1        1        1        1 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$ESC.H1</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   1_TssA 16_EnhW1 19_DNase  2_PromU 25_Quies 3_PromD1 </span></div><div class="line"><span class="comment">##        1        1        1        1        1        1 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$IPSC.DF.6.9</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   1_TssA 16_EnhW1 19_DNase  2_PromU 25_Quies </span></div><div class="line"><span class="comment">##        1        1        1        1        1 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$BLD.CD19.PPC</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   1_TssA 16_EnhW1 19_DNase  2_PromU 25_Quies 3_PromD1 </span></div><div class="line"><span class="comment">##        1        1        1        1        1        1</span></div></pre></td></tr></table></figure>
<p>因此，我们看到，在某个特定的基因的启动子区域的基因序列方面，不同的细胞类型存在着数量和状态类型的变化。在后面的练习中，我们将会考虑如何以这种方式开发其它潜在感兴趣的信息。</p>
<h1 id="总结">总结</h1>
<p>在这一部分中我们看到：</p>
<ul>
<li>在磁盘中，使用了大量添加了索引的文本来表示基因组注释或实验结果（原文是：“leave on disk” large numbers of indexed textual representations of genomic annotation or experimental result）；</li>
<li>使用R包的方式来简化标记这些数据；</li>
<li>使用<code>GenomicFiles</code>来管理有关文件的元数据；</li>
<li>使用<code>reduceByFile</code>来并行处理文件内容。</li>
</ul>
<p>这些技术避免了创建，加载大量的R对象，从而提供了某种程度的便捷性。在内存资源有限的情况下，对于那些需要大量集群(clusters)的应用来说，这种策略就显得性价比较高。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_multibed.html" target="_blank" rel="external">Managing and processing large numbers of BED files</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/07/Experiment/HIPC介绍/" rel="next" title="HIPC介绍">
                <i class="fa fa-chevron-left"></i> HIPC介绍
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/" rel="prev" title="GDAS010-生物学变异与技术变异">
                GDAS010-生物学变异与技术变异 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">243</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">117</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#erma表观遗传路线图简介"><span class="nav-number">2.</span> <span class="nav-text">erma：表观遗传路线图简介</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#bed文件集合"><span class="nav-number">2.1.</span> <span class="nav-text">BED文件集合</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#使用genomicfiles进行管理"><span class="nav-number">2.2.</span> <span class="nav-text">使用GenomicFiles进行管理</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#使用表格描述染色质状态信息"><span class="nav-number">3.</span> <span class="nav-text">使用表格描述染色质状态信息</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#定义染色体状态与标记概述"><span class="nav-number">3.1.</span> <span class="nav-text">定义染色体状态与标记概述</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#并行目标查询"><span class="nav-number">3.2.</span> <span class="nav-text">并行目标查询</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#总结"><span class="nav-number">4.</span> <span class="nav-text">总结</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">4.1.</span> <span class="nav-text">参考资料</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共963.7k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
