<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="RNA-seq,Bioconductor," />










<meta name="description" content="前言 在后面的章节中，我们将介绍基因组学实验中的统计推断。我们会使用前面几节中介绍的一些概念，其中包括t检验和多重比较。在这一节中，我们先介绍一个在基因组学数据分析中一个极为重要的概念：生物学变异与技术变异(biological and technical variability)以及它们之间的区别。 通常而言，我们在一个种群里所观察到的生物学单位，例如某个个体的变异称为生物学变异(biolo">
<meta name="keywords" content="RNA-seq,Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS010-生物学变异与技术变异">
<meta property="og:url" content="http://rvdsd.top/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 在后面的章节中，我们将介绍基因组学实验中的统计推断。我们会使用前面几节中介绍的一些概念，其中包括t检验和多重比较。在这一节中，我们先介绍一个在基因组学数据分析中一个极为重要的概念：生物学变异与技术变异(biological and technical variability)以及它们之间的区别。 通常而言，我们在一个种群里所观察到的生物学单位，例如某个个体的变异称为生物学变异(biolo">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-3-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-6-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-10-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-11-1.png">
<meta property="og:updated_time" content="2019-11-13T08:51:22.892Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS010-生物学变异与技术变异">
<meta name="twitter:description" content="前言 在后面的章节中，我们将介绍基因组学实验中的统计推断。我们会使用前面几节中介绍的一些概念，其中包括t检验和多重比较。在这一节中，我们先介绍一个在基因组学数据分析中一个极为重要的概念：生物学变异与技术变异(biological and technical variability)以及它们之间的区别。 通常而言，我们在一个种群里所观察到的生物学单位，例如某个个体的变异称为生物学变异(biolo">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-3-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/"/>





  <title>GDAS010-生物学变异与技术变异 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS010-生物学变异与技术变异</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-10T12:00:00+08:00">
                2019-09-10
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,677
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="前言">前言</h1>
<p>在后面的章节中，我们将介绍基因组学实验中的统计推断。我们会使用前面几节中介绍的一些概念，其中包括t检验和多重比较。在这一节中，我们先介绍一个在基因组学数据分析中一个极为重要的概念：生物学变异与技术变异(biological and technical variability)以及它们之间的区别。</p>
<p>通常而言，我们在一个种群里所观察到的生物学单位，例如某个个体的变异称为生物学变异(biological variability)。我们将那些对同一个生物学单位进行测量时所观察到的变异称为技术性变异(technical variability，例如同一个样本测3次，这3次就是技术重复，3次数据的变异称为技术变异)。由于在基因组学研究中经常会用到新技术，因此技术重复(technical replicates)在很多时候都会用于评估实验数据。理想情况下，同一样本的检测结果应该是相同的，但在实际检测中同一样本的多次检测会出现一定的偏差，因此使用技术重复我们就能我们就能够评估新技术的技术变异。我还经常使用生物学重复(biological replicates)与技术重复(technical replicates)来分别代指生物学变异和技术变异。</p>
<p>我们在进行统计推断来解释有差异时，需要注意不要混淆生物学变异和技术变异，因为它们所代表的意义是不同的。例如，当我们分析技术重复时，样本仅仅只有一个，而不是代表某个总体，例如人类或小鼠。这一节中，我们通过设计一个实验来说明一下技术重复和生物重复。</p>
<h1 id="合并实验数据">合并实验数据</h1>
<p>我们将要研究的数据集来源于基因表达微阵列。在这个实验中，我们从两个品系的小鼠中分别随机选择了12只小鼠，并提取了它们的RNA。我们将得到的24个样本使用微阵列的手段进行检测，引外，我们还构成了两个库，这两个库中的样本则是分别来源于这两个品系的12只小鼠的混合样本。</p>
<p>现在我们安装以下包，这个包中含有这些数据，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(devtools)</div><div class="line">install_github(&quot;genomicsclass/maPooling&quot;)</div></pre></td></tr></table></figure>
<p>使用<code>pData</code>函数我们就能查看实验设计。每行表示一个样本，每列表示1只小鼠。例如<code>i,j</code>单元格，中如果是1，则表示RNA数据来源于<code>j</code>只小鼠的样本<code>i</code>。通过行名我们可以知道品系信息（不推荐这种方法来进行分组，你可以向phenoData中添加一个变量，用于指定小鼠品系信息），数据如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line"><span class="keyword">library</span>(maPooling)</div><div class="line">data(maPooling)</div><div class="line">head(pData(maPooling))</div><div class="line"><span class="comment">##          a2 a3 a4 a5 a6 a7 a8 a9 a10 a11 a12 a14 b2 b3 b5 b6 b8 b9 b10 b11</span></div><div class="line"><span class="comment">## a10       0  0  0  0  0  0  0  0   1   0   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a10a11    0  0  0  0  0  0  0  0   1   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a10a11a4  0  0  1  0  0  0  0  0   1   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a11       0  0  0  0  0  0  0  0   0   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a12       0  0  0  0  0  0  0  0   0   0   1   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a12a14    0  0  0  0  0  0  0  0   0   0   1   1  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">##          b12 b13 b14 b15</span></div><div class="line"><span class="comment">## a10        0   0   0   0</span></div><div class="line"><span class="comment">## a10a11     0   0   0   0</span></div><div class="line"><span class="comment">## a10a11a4   0   0   0   0</span></div><div class="line"><span class="comment">## a11        0   0   0   0</span></div><div class="line"><span class="comment">## a12        0   0   0   0</span></div><div class="line"><span class="comment">## a12a14     0   0   0   0</span></div></pre></td></tr></table></figure>
<p>现在我们创建一个函数，用于说明小鼠的样本信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar()</div><div class="line">flipt &lt;- <span class="keyword">function</span>(m) t(m[nrow(m):<span class="number">1</span>,])</div><div class="line">myimage &lt;- <span class="keyword">function</span>(m,<span class="keyword">...</span>) &#123;</div><div class="line">  image(flipt(m),xaxt=<span class="string">"n"</span>,yaxt=<span class="string">"n"</span>,<span class="keyword">...</span>)</div><div class="line">  &#125;</div><div class="line"></div><div class="line">myimage(as.matrix(pData(maPooling)),col=c(<span class="string">"white"</span>,<span class="string">"black"</span>),</div><div class="line">        xlab=<span class="string">"experiments"</span>,</div><div class="line">        ylab=<span class="string">"individuals"</span>,</div><div class="line">        main=<span class="string">"phenoData"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-3-1.png" alt="plot of chunk unnamed-chunk-3">
<p class="caption">plot of chunk unnamed-chunk-3</p>
</div>
<p>需要注意的是，最终我们只对两个品系小鼠的差异基因感兴趣，也就是数据中标为0的品系和1的品系。我们可以对合并样本的技术重复进行检验或来源于12只小鼠的数据进行检验。我们可以提取出这些合并的样本，因为来自于每个品系的所有小鼠都都可以用0或1来表示，因此那些来源于某品系（例如用1表示的这个品系）代表了这些样本，因此我们提取出矩阵行之和为12的基因，这段话不太理解，原文如下：</p>
<blockquote>
<p>Note that ultimately we are interested in detecting genes that are differentially expressed between the two strains of mice which we will refer to as strain 0 and 1. We can apply tests to the technical replicates of pooled samples or the data from 12 individual mice. We can identify these pooled samples because all mice from each strain were represented in these samples and thus the sum of the rows of experimental design matrix add up to 12:</p>
</blockquote>
<p>如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">data(maPooling)</div><div class="line">pd=pData(maPooling)</div><div class="line">pooled=which(rowSums(pd)==12)</div></pre></td></tr></table></figure>
<p>我们可以从列名确定品系，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">factor(as.numeric(grepl(<span class="string">"b"</span>,names(pooled))))</div><div class="line"><span class="comment">## [1] 0 0 0 0 1 1 1 1</span></div><div class="line"><span class="comment">## Levels: 0 1</span></div></pre></td></tr></table></figure>
<p>如果我们要比较每个基因在不同品系之间的差异，我们就会发现有几个基因出现了比较一致的差异，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="comment">###look at 2 pre-selected genes for illustration</span></div><div class="line">i=<span class="number">11425</span>;j=<span class="number">11878</span></div><div class="line">pooled_y=exprs(maPooling[,pooled])</div><div class="line">pooled_g=factor(as.numeric(grepl(<span class="string">"b"</span>,names(pooled))))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">2</span>)</div><div class="line">stripchart(split(pooled_y[i,],pooled_g),vertical=<span class="literal">TRUE</span>,method=<span class="string">"jitter"</span>,col=c(<span class="number">1</span>,<span class="number">2</span>),</div><div class="line">           main=<span class="string">"Gene 1"</span>,xlab=<span class="string">"Group"</span>,pch=<span class="number">15</span>)</div><div class="line">stripchart(split(pooled_y[j,],pooled_g),vertical=<span class="literal">TRUE</span>,method=<span class="string">"jitter"</span>,col=c(<span class="number">1</span>,<span class="number">2</span>),</div><div class="line">           main=<span class="string">"Gene 2"</span>,xlab=<span class="string">"Group"</span>,pch=<span class="number">15</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<p>这里要注意一下，如果我们根据这些值来进行t检验，那么我们就会得到非常显著的结果，也就是说p值极低，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(genefilter)</div><div class="line">pooled_tt=rowttests(pooled_y,pooled_g)</div><div class="line">pooled_tt$p.value[i]</div><div class="line"><span class="comment">## [1] 2.075617e-07</span></div><div class="line">pooled_tt$p.value[j]</div><div class="line"><span class="comment">## [1] 3.400476e-07</span></div></pre></td></tr></table></figure>
<p>但是，如果我们再次从两个品系的小鼠中各选择12个小鼠的话，这样的结论会成立吗？请注意，t检验的定义包括要比较的总体的标准差。我们这里做的计算结果是否能够反映出总体差异？（注：这里要回顾一下t检验的定义，样本与方差的差异等概念）。</p>
<p>需要注意的是，我们现在重复计算的过程就是在重复原来的实验流程。我们为每个合并的样本创建了4个技术重复。或许Gene 1在小鼠品系内存在着很高的变异，而Gene 2的变异则比较低，但是我们不能看到这一点，因为小鼠之间的变异被合并后的数据所掩盖了（注：这里作者只是假设了Gene 1可能存在着比较高的组内变异）。</p>
<p>我们还有每只小鼠的微阵列数据。对于每个品系的小鼠来说，我们有12个生物学重复。我们可以通过查询1来找到这个品系，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">individuals=which(rowSums(pd)==1)</div></pre></td></tr></table></figure>
<p>事实证明，某些单独的小鼠本身就含有技术重复，因此现在我们将这种情况剔除掉，从而用于说明仅用生物学重复的分析结果，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="comment">##remove replicates</span></div><div class="line">individuals=individuals[-grep(<span class="string">"tr"</span>,names(individuals))]</div><div class="line">y=exprs(maPooling)[,individuals]</div><div class="line">g=factor(as.numeric(grepl(<span class="string">"b"</span>,names(individuals))))</div></pre></td></tr></table></figure>
<p>我们可以计算每个基因的在某个品系中的方差，并与通过技术重复获得的标准差进行比较，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">technicalsd &lt;- rowSds(pooled_y[,pooled_g==<span class="number">0</span>])</div><div class="line">biologicalsd &lt;- rowSds(y[,g==<span class="number">0</span>])</div><div class="line">LIM=range(c(technicalsd,biologicalsd))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">boxplot(technicalsd,biologicalsd,names=c(<span class="string">"technical"</span>,<span class="string">"biological"</span>),ylab=<span class="string">"standard deviation"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>从上图上我们可以看到，生物学方差(variance)远大于技术方差。并且对于方差的变异而言，生物学方差更大。这只是我们前面提到的2个基因的情况，现在我们展示每只小鼠上测得的这2个基因的表达值，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">mypar(1,2)</div><div class="line">stripchart(split(y[i,],g),vertical=TRUE,method=&quot;jitter&quot;,col=c(1,2),xlab=&quot;Gene 1&quot;,pch=15)</div><div class="line">points(c(1,2),tapply(y[i,],g,mean),pch=4,cex=1.5)</div><div class="line">stripchart(split(y[j,],g),vertical=TRUE,method=&quot;jitter&quot;,col=c(1,2),xlab=&quot;Gene 2&quot;,pch=15)</div><div class="line">points(c(1,2),tapply(y[j,],g,mean),pch=4,cex=1.5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11">
<p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<p>现在p值就发生了一点变化，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(genefilter)</div><div class="line">tt=rowttests(y,g)</div><div class="line">tt$p.value[i]</div><div class="line">## [1] 0.0898726</div><div class="line">tt$p.value[j]</div><div class="line">## [1] 1.979172e-07</div></pre></td></tr></table></figure>
<p>当我们在不同品系的小鼠之间比较这2个基因的差异时，我们对哪个基因的差异更有信心呢（Gene 1还是Gene 2）？如果另一位研究人员另外一批随机样本来做实验，你认识他会发现哪个基因有差异（Gene 1还是Gene 2）？如果我们希望我们的结论是关于小鼠品系的通用结论（通用结论这里我的理解就是，无论放哪个实验室做，都能做出这个结果，得到这个结论），而不仅仅是我们这次实验的结论（也就是说，有可能就是只有自己实验室能做出这个结果，那么这个结果就不太可信，有可能是假阳性），那么检测生物学的变异则非常重要。</p>
<p>在这一节中，我们使用了两种分析方法。第一种分析方法是将这两个品系的生物学重复做为总体总体来进行分析析（也就是说把某个品系的12只小鼠数据汇总起来分析）。第二种则是将合并后的数据来作为总体进行分析（我的理解就是把12只小鼠的RNA混合在一起，作为一个样本，然后使用4个技术重复来进行分析，这种分析方法其实检测的就是技术变异）。在科学研究中，我们通常关注的是总体（也就是第一种分析方法）。作为一个非常实际的例子，这里需要注意的是，如果另外一个实验室重复该实验，他们会有另外一组12个小鼠的样本，因此关于我们实验中对总体的推断也会在另外的实验室中得到重复。</p>
<blockquote>
<p>An analysis with biological replicates has as a population these two strains of mice. An analysis with technical replicates has as a population the twelve mice we selected and the variability is related to the measurement technology. In science we typically are concerned with populations. As a very practical example, note that if another lab performs this experiment they will have another set of twelve mice and thus inferences about <em>populations</em> are more likely to be reproducible.</p>
</blockquote>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_btvari.html" target="_blank" rel="external">Biological versus technical variability</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/RNA-seq/" <i class="fa fa-tag"></i> RNA-seq</a>
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/10/BioTools/fgsea/" rel="next" title="fgsea包做GSEA分析">
                <i class="fa fa-chevron-left"></i> fgsea包做GSEA分析
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/11/Genomics Data Analysis Series/GDAS011-基因组级数据中的t检验/" rel="prev" title="GDAS011-基因组级数据中的t检验与多重比较">
                GDAS011-基因组级数据中的t检验与多重比较 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">246</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">117</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#合并实验数据"><span class="nav-number">2.</span> <span class="nav-text">合并实验数据</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">2.1.</span> <span class="nav-text">参考资料</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共980.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
