<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="RNA-Seq," />










<meta name="description" content="前言 利用DEseq2包进行分析不同组之间的差异。都是参考徐洲更.(伪)从零开始学转录组(7):差异基因表达分析的。 导入数据，构建 DESeq2 所需的 DESeqDataSet 对象，如下所示： 1234library(DESeq2)countData &amp;lt;- raw_count_filt[,2:5]condition &amp;lt;- factor(c(&amp;quot;control&amp;quo">
<meta name="keywords" content="RNA-Seq">
<meta property="og:type" content="article">
<meta property="og:title" content="转录组学习笔记07-差异基因分析">
<meta property="og:url" content="http://rvdsd.top/2017/10/07/BioItem/生信-转录组学习笔记07差异基因分析/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 利用DEseq2包进行分析不同组之间的差异。都是参考徐洲更.(伪)从零开始学转录组(7):差异基因表达分析的。 导入数据，构建 DESeq2 所需的 DESeqDataSet 对象，如下所示： 1234library(DESeq2)countData &amp;lt;- raw_count_filt[,2:5]condition &amp;lt;- factor(c(&amp;quot;control&amp;quo">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/rnaseq112.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/rnaseq113.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/rnqseq114.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/rnaseq115.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/rnaseq116.png">
<meta property="og:updated_time" content="2018-08-24T08:39:41.046Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="转录组学习笔记07-差异基因分析">
<meta name="twitter:description" content="前言 利用DEseq2包进行分析不同组之间的差异。都是参考徐洲更.(伪)从零开始学转录组(7):差异基因表达分析的。 导入数据，构建 DESeq2 所需的 DESeqDataSet 对象，如下所示： 1234library(DESeq2)countData &amp;lt;- raw_count_filt[,2:5]condition &amp;lt;- factor(c(&amp;quot;control&amp;quo">
<meta name="twitter:image" content="http://ortxpw68f.bkt.clouddn.com/rnaseq112.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2017/10/07/BioItem/生信-转录组学习笔记07差异基因分析/"/>





  <title>转录组学习笔记07-差异基因分析 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">学习过程中的输出</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2017/10/07/BioItem/生信-转录组学习笔记07差异基因分析/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">转录组学习笔记07-差异基因分析</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2017-10-07T12:00:00+08:00">
                2017-10-07
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信实战项目/" itemprop="url" rel="index">
                    <span itemprop="name">生信实战项目</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,050
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  9
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>利用DEseq2包进行分析不同组之间的差异。都是参考<a href="http://www.360doc.com/content/17/1023/22/41791033_697550480.shtml" target="_blank" rel="external">徐洲更.(伪)从零开始学转录组(7):差异基因表达分析</a>的。</p>
<p>导入数据，构建 DESeq2 所需的 DESeqDataSet 对象，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">library(DESeq2)</div><div class="line">countData &lt;- raw_count_filt[,2:5]</div><div class="line">condition &lt;- factor(c(&quot;control&quot;,&quot;KD&quot;,&quot;KD&quot;,&quot;control&quot;))</div><div class="line">dds &lt;- DESeqDataSetFromMatrix(countData, DataFrame(condition), design= ~ condition )</div></pre></td></tr></table></figure>
<h2 id="过滤掉低low-count">过滤掉低low count</h2>
<p>接着， 过滤掉一些low count数据（通常是count为0的read），节省内存，提高运行速度，如下所示： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">nrow(dds)</div><div class="line">dds &lt;- dds[ rowSums(counts(dds)) &gt; 1, ]</div><div class="line">nrow(dds)</div></pre></td></tr></table></figure></p>
<h2 id="使用deseq进行差异表达分析">使用DESeq进行差异表达分析：</h2>
<p>DESeq包含三步，estimation of size factors（estimateSizeFactors)， estimation of dispersion（estimateDispersons)， Negative Binomial GLM fitting and Wald statistics（nbinomWaldTest），可以分布运行，也可用一步到位，最后返回 results可用的DESeqDataSet对象，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">dds &lt;- DESeq(dds)</div></pre></td></tr></table></figure>
<blockquote>
<p>estimating size factors estimating dispersions gene-wise dispersion estimates mean-dispersion relationship final dispersion estimates fitting model and testing</p>
</blockquote>
<p>出现上述结果表示运行结束，过程大概是十几秒。 可以通过<code>result</code>来查看运行的结果，如下所示： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">res &lt;- results(dds)</div></pre></td></tr></table></figure></p>
<p>可用mcols查看每一项结果的具体含义，比如说log2FoldChange 表示倍数变化取log2结果，还能画个火山图。padj 就是用BH对多重试验进行矫正。如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">mcols(res, use.names = TRUE)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/rnaseq112.png">

</div>
<p>用summary可以看描述性的结果，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/rnaseq113.png"></p>
<h3 id="绘制ma图">绘制MA图</h3>
<blockquote>
<p>An MA plot is an application of a Bland–Altman plot for visual representation of genomic data. The plot visualises the differences between measurements taken in two samples, by transforming the data onto M (log ratio) and A (mean average) scales, then plotting these values. Though originally applied in the context of two channel DNA microarray gene expression data, MA plots are also used to visualise high-throughput sequencing analysis —From wikipeida</p>
</blockquote>
<p>M表示log fold change，衡量基因表达量变化，上调还是下调。A表示每个基因的count的均值。根据summary可知，low count的比率很高，所以大部分基因表达量不高，也就是集中在0的附近（log2(1)=0，也就是变化1倍）,提供了模型预测系数的分布总览。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">plotMA(res, ylim = c(-5,5))</div><div class="line">topGene &lt;- rownames(res)[which.min(res$padj)]</div><div class="line">with(res[topGene, ], &#123;</div><div class="line">  points(baseMean, log2FoldChange, col=&quot;dodgerblue&quot;, cex=2, lwd=2)</div><div class="line">  text(baseMean, log2FoldChange, topGene, pos=2, col=&quot;dodgerblue&quot;)</div><div class="line">&#125;)</div></pre></td></tr></table></figure>
<p>如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/rnqseq114.png"></p>
<p>如果经过lfcShrink 收缩log2 fold change， 结果会好看很多 <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">res.shrink &lt;- lfcShrink(dds, contrast = c(&quot;condition&quot;,&quot;KD&quot;,&quot;control&quot;), res=res)</div><div class="line">plotMA(res.shrink, ylim = c(-5,5))</div><div class="line">topGene &lt;- rownames(res)[which.min(res$padj)]</div><div class="line">with(res[topGene, ], &#123;</div><div class="line">  points(baseMean, log2FoldChange, col=&quot;dodgerblue&quot;, cex=2, lwd=2)</div><div class="line">  text(baseMean, log2FoldChange, topGene, pos=2, col=&quot;dodgerblue&quot;)</div><div class="line">&#125;)</div></pre></td></tr></table></figure></p>
<p>如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/rnaseq115.png"></p>
<h3 id="寻找差异基因">寻找差异基因</h3>
<p>提取差异基因</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">res.deseq2 &lt;- subset(res, padj &lt; 0.05)</div></pre></td></tr></table></figure>
<h3 id="使用edger进行差异基因分析">使用edgeR进行差异基因分析</h3>
<p>edgeR使用DGEList函数读取count matrix数据，也就说你需要提供一个现成的matrix数据，而不是指望它能读取单独的文件，然后进行合并，其实可以用 tximport 或 DESeqDataSetFromHTSeq 读取单独的文件，然后传递给DGEList。</p>
<h4 id="第一步-构建dgelist对象">第一步： 构建DGEList对象</h4>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(edgeR)</div><div class="line">group &lt;- factor(c(&quot;control&quot;,&quot;KD&quot;,&quot;KD&quot;,&quot;control&quot;))</div><div class="line">genelist &lt;- DGEList(counts=raw_count_filt[,2:5], group = group)</div></pre></td></tr></table></figure>
<h4 id="第二步-过滤-low-counts数据">第二步： 过滤 low counts数据。</h4>
<p>与DESeq2的预过滤不同，DESeq2的预过滤只是为了改善后续运算性能，在运行过程中依旧会自动处理low count数据，edgeR需要在分析前就要排除那些low count数据，而且非常严格。从生物学角度，有生物学意义的基因的表达量必须高于某一个阈值。从统计学角度上， low count的数据不太可能有显著性差异，而且在多重试验矫正阶段还会拖后腿。 综上所诉，放心大胆的过滤吧。</p>
<p>根据经验，基因至少在某一些文库的count超过10 ~ 15 才被认为是表达。这一步全靠尝试， 剔除太多就缓缓，剔除太少就严格点。 我们可以简单的对每个基因的raw count进行比较，但是建议用CPM（count-per-million)标准化 后再比较，避免了文库大小的影响。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"># 简单粗暴的方法</div><div class="line">keep &lt;- rowSums(genelist$count) &gt; 50</div><div class="line"># 利用CPM标准化</div><div class="line">keep &lt;- rowSums(cpm(genelist) &gt; 0.5 ) &gt;=2</div><div class="line">table(keep)</div><div class="line">genelist.filted &lt;- genelist[keep,,keep.lib.sizes=FALSE]</div></pre></td></tr></table></figure>
<p>这里的0.5(即阈值）等于 10/(最小的文库的 read count数 /1000000)，keep.lib.size=FALSE表示重新计算文库大小。</p>
<p>第三步： 根据组成偏好(composition bias)标准化。edgeR的calcNormFactors函数使用TMM算法对DGEList标准化.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">genelist.norm &lt;- calcNormFactors(genelist.filted)</div></pre></td></tr></table></figure>
<p>注 大部分的mRNA-Seq数据分析用TMM标准化就行了，但是也有例外，比如说single-cell RNA-Seq(Lun, Bach, and Marioni 2016), 还有就是global differential expression， 基因组一半以上的基因都是差异表达的，请尽力避免，(D. Wu et al. 2013)， 不然就需要用到内参进行标准化了(Risso et al. 2014).</p>
<p>第四步： 实验设计矩阵(Design matrix)， 类似于DESeq2中的design参数。 edgeR的线性模型和差异表达分析需要定义一个实验设计矩阵。很直白的就能发现是1vs0</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">design &lt;- model.matrix(~0+group)</div><div class="line">colnames(design) &lt;- levels(group)</div></pre></td></tr></table></figure>
<p>第五步： 估计离散值（Dispersion）。前面已经提到负二项分布（negative binomial，NB)需要均值和离散值两个参数。edgeR对每个基因都估测一个经验贝叶斯稳健离散值（mpirical Bayes moderated dispersion），还有一个公共离散值（common dispersion，所有基因的经验贝叶斯稳健离散值的均值）以及一个趋势离散值</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">genelist.Disp &lt;- estimateDisp(genelist.norm, design, robust = TRUE)</div><div class="line">plotBCV(genelist.Disp)</div></pre></td></tr></table></figure>
<p>还可以进一步通过quasi-likelihood (QL)拟合NB模型，用于解释生物学和技术性导致的基因特异性变异 (Lund et al. 2012; Lun, Chen, and Smyth 2016).</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">fit &lt;- glmQLFit(genelist.Disp, design, robust=TRUE)</div><div class="line">head(fit$coefficients)</div></pre></td></tr></table></figure>
<p>第六步: 差异表达检验（1）。这一步主要构建比较矩阵，类似于DESeq2中的results函数的 contrast 参数。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">cntr.vs.KD &lt;- makeContrasts(control-KD, levels=design)</div><div class="line">res &lt;- glmQLFTest(fit, contrast=cntr.vs.KD)</div><div class="line">ig.edger &lt;- res$table[p.adjust(res$table$PValue, method = &quot;BH&quot;) &lt; 0.01, ]</div></pre></td></tr></table></figure>
<p>这里用的是glmQLFTest而不是glmLRT是因为前面用了glmQLTFit进行拟合，所以需要用QL F-test进行检验。如果前面用的是glmFit，那么对应的就是glmLRT. 作者称QL F-test更加严格。多重试验矫正用的也是BH方法。</p>
<p>后续就是提取显著性差异的基因用作下游分析，做一些图看看</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">topTags(res,n=10)</div><div class="line">is.de &lt;- decideTestsDGE(res)</div><div class="line">summary(is.de)</div><div class="line">plotMD(res, status=is.de, values=c(1,-1), col=c(&quot;red&quot;,&quot;blue&quot;),legend=&quot;topright&quot;)</div></pre></td></tr></table></figure>
<p>如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/rnaseq116.png"></p>
<p>第六步：差异表达检验（2）。上面找到的显著性差异的基因，没有考虑效应值，也就是具体变化了多少倍。我们也可用找表达量变化比较大的基因，对应的函数是 glmTreat。</p>
<h3 id="使用limma进行差异分析">使用limma进行差异分析</h3>
<p>导入read count， 保存为专门的对象用于后续分析 原始数据过滤，根据标准化read count 或者 raw count 作为筛选标准 raw read count 标准化 通过各种算法（如经验贝叶斯，EM）预测dispersion离散值 广义线性模型拟合数据 差异分析，也就是统计检验部分</p>
<p>Limma原先用于处理基因表达芯片数据。如果你仔细看edgeR导入界面，你就会发现，edgeR有一部分功能依赖于limma包。Limma采用经验贝叶斯模型（ Empirical Bayesian model）让结果更稳健。</p>
<p>在处理RNA-Seq数据时，raw read count先被转成log2-counts-per-million (logCPM)，然后对mean-variance关系建模。建模有两种方法：</p>
<p>精确权重法（precision weights）也就是“voom” 经验贝叶斯先验趋势（empirical Bayes prior trend），也就是”limma-trend“</p>
<p>数据预处理： Limma使用edgeR的DGEList对象，并且过滤方法都是一致的，对应edgeR的第一步,第二步， 第三步</p>
<p><figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">library(edgeR)library(limma)</div><div class="line">group &lt;- factor(c(&quot;control&quot;,&quot;KD&quot;,&quot;KD&quot;,&quot;control&quot;))</div><div class="line">genelist &lt;- DGEList(counts=raw_count_filt[,2:5], group = group)</div><div class="line">### filter base  use CPM</div><div class="line">keep &lt;- rowSums(cpm(genelist) &gt; 0.5 ) &gt;=2table(keep)</div><div class="line">genelist.filted &lt;- genelist[keep, ,keep.lib.sizes=FALSE]</div><div class="line">### normalizaition</div><div class="line">x &lt;- calcNormFactors(x, method = &quot;TMM&quot;)</div></pre></td></tr></table></figure></p>
<p>差异表达分析: 使用“limma-trend”</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">design &lt;- model.matrix(~0+group)</div><div class="line">colnames(design) &lt;- levels(group)</div><div class="line">logCPM &lt;- cpm(genelist.norm, log=TRUE, prior.count=3)</div><div class="line">fit &lt;- lmFit(logCPM, design)</div><div class="line">fit &lt;- eBayes(fit, trend=TRUE)</div><div class="line">topTable(fit, coef=ncol(design))</div></pre></td></tr></table></figure>
<p>差异表达分析: 使用”limma-voom“</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">### DGE with voom</div><div class="line">v &lt;- voom(genelist.norm, design, plot=TRUE)</div><div class="line">#v &lt;- voom(counts, design, plot=TRUE)</div><div class="line">fit &lt;- lmFit(v, design)</div><div class="line">fit &lt;- eBayes(fit)</div><div class="line">all &lt;- topTable(fit, coef=ncol(design), number=10000)</div><div class="line">sig.limma &lt;- all[all$adj.P.Val &lt; 0.01, ]</div><div class="line">fit &lt;- treat(fit, lfc=log2(1.2))</div><div class="line">topTreat(fit, coef=ncol(design))</div></pre></td></tr></table></figure>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/RNA-Seq/" <i class="fa fa-tag"></i> RNA-Seq</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2017/10/06/BioItem/生信-转录组学习笔记06reads计数/" rel="next" title="转录组学习笔记06-reads计数">
                <i class="fa fa-chevron-left"></i> 转录组学习笔记06-reads计数
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2017/10/08/BioItem/生信-转录组学习笔记08差异基因结果注释/" rel="prev" title="转录组学习笔记08-差异基因结果注释">
                转录组学习笔记08-差异基因结果注释 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">157</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">10</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">45</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#过滤掉低low-count"><span class="nav-number">2.</span> <span class="nav-text">过滤掉低low count</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#使用deseq进行差异表达分析"><span class="nav-number">3.</span> <span class="nav-text">使用DESeq进行差异表达分析：</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#绘制ma图"><span class="nav-number">3.1.</span> <span class="nav-text">绘制MA图</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#寻找差异基因"><span class="nav-number">3.2.</span> <span class="nav-text">寻找差异基因</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#使用edger进行差异基因分析"><span class="nav-number">3.3.</span> <span class="nav-text">使用edgeR进行差异基因分析</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#第一步-构建dgelist对象"><span class="nav-number">3.3.1.</span> <span class="nav-text">第一步： 构建DGEList对象</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#第二步-过滤-low-counts数据"><span class="nav-number">3.3.2.</span> <span class="nav-text">第二步： 过滤 low counts数据。</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#使用limma进行差异分析"><span class="nav-number">3.4.</span> <span class="nav-text">使用limma进行差异分析</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共589.7k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
