<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言在生命科学研究领域的相关人员，经常使用电子表格来记录和管理实验数据。针对这个问题，我们将会提出一个使用R和Bioconductor管理基因组级实验数据的原则。使用基于R的软件来管理数据似乎并不太常见。通常来说，R仅用于编写程序，构建统计模型。但是，针对复杂生物实验数据的多方面需求这种情况，Bioconductor已经非常着手以统一的方式来详细地注释这些数据。Bioconductor的这种做法可">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS004-Bioconductor与基因组级数据分析简介">
<meta property="og:url" content="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言在生命科学研究领域的相关人员，经常使用电子表格来记录和管理实验数据。针对这个问题，我们将会提出一个使用R和Bioconductor管理基因组级实验数据的原则。使用基于R的软件来管理数据似乎并不太常见。通常来说，R仅用于编写程序，构建统计模型。但是，针对复杂生物实验数据的多方面需求这种情况，Bioconductor已经非常着手以统一的方式来详细地注释这些数据。Bioconductor的这种做法可">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-dochks-1.png">
<meta property="og:updated_time" content="2019-11-13T08:50:05.330Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS004-Bioconductor与基因组级数据分析简介">
<meta name="twitter:description" content="前言在生命科学研究领域的相关人员，经常使用电子表格来记录和管理实验数据。针对这个问题，我们将会提出一个使用R和Bioconductor管理基因组级实验数据的原则。使用基于R的软件来管理数据似乎并不太常见。通常来说，R仅用于编写程序，构建统计模型。但是，针对复杂生物实验数据的多方面需求这种情况，Bioconductor已经非常着手以统一的方式来详细地注释这些数据。Bioconductor的这种做法可">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/"/>





  <title>GDAS004-Bioconductor与基因组级数据分析简介 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS004-Bioconductor与基因组级数据分析简介</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-04T12:00:00+08:00">
                2019-09-04
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,649
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  15
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h1><p>在生命科学研究领域的相关人员，经常使用电子表格来记录和管理实验数据。针对这个问题，我们将会提出一个使用R和Bioconductor管理基因组级实验数据的原则。使用基于R的软件来管理数据似乎并不太常见。通常来说，R仅用于编写程序，构建统计模型。但是，针对复杂生物实验数据的多方面需求这种情况，Bioconductor已经非常着手以统一的方式来详细地注释这些数据。Bioconductor的这种做法可以让我们导入初步分析的结果（例如多个微阵列的扫描结果以及多个测序结果的比对信息），将这些初步的结果与有关实验证的元数据，以及检测的样本信息结合起来，从而将所有的信息都保存在一个R变量中。通过周全的设计该变量的数据结构，程序可以对这些数据进行各种操作，最终获得更加有效的管理和分析环境。</p>
<p>在这一节中，我们会通过一些实验数据来演示一下基本的操作方法，从而，逐步过滤到更加全面的表示方法（</p>
<p>我们将通过对Bioconductor中所表示的实验数据的一些非常基本的表示法，逐步走向更先进的综合表示法，从而理解这种数据操作方法所表现出来的高效率。</p>
<h2 id="案例说明-Youtube上有相应视频"><a href="#案例说明-Youtube上有相应视频" class="headerlink" title="案例说明 (Youtube上有相应视频)"></a>案例说明 (Youtube上有相应视频)</h2><p>1.0 Distinct components from an experiment.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSE5859Subset)</div><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line">.oldls = ls()</div><div class="line">data(GSE5859Subset)</div><div class="line">.newls = ls()</div></pre></td></tr></table></figure>
<p>我们导入了一个<code>GSE5859Subset</code>的实验数据，现在我们使用 <code>setdiff()</code>函数来比较一下导入前后文件的区别，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">newstuff = setdiff(.newls, .oldls)</div><div class="line">newstuff</div><div class="line"><span class="comment">## [1] "geneAnnotation" "geneExpression" "sampleInfo"</span></div></pre></td></tr></table></figure>
<p>从结果中我们可以发现，当导入了 <code>GSE5859Subset</code>实验数据中，这个数据中含有3个文件。它们3个分别属于不同的类，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">cls = lapply(newstuff, <span class="keyword">function</span>(x) class(get(x)))</div><div class="line">names(cls) = newstuff</div><div class="line">cls</div><div class="line"><span class="comment">## $$geneAnnotation</span></div><div class="line"><span class="comment">## [1] "data.frame"</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$geneExpression</span></div><div class="line"><span class="comment">## [1] "matrix"</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$sampleInfo</span></div><div class="line"><span class="comment">## [1] "data.frame"</span></div></pre></td></tr></table></figure>
<p>我们可以按照种族绘制绘制出它们不同时间的图形，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(date~factor(ethnicity), data=sampleInfo, ylab=&quot;chip date&quot;)</div></pre></td></tr></table></figure>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png" alt="plot of chunk lket"></p>
<p>假如我们想查看一下BRCA2在不同种族中的表达情况，我们可以按以下操作进行：</p>
<ul>
<li>我们需要知道<code>geneExpression</code> 这个矩阵中哪一行代表BRCA2的数据，并且我们要知道<code>geneAnnotation</code>哪一列提供了BRCA2的信息（探针名与基因名的对应关系）；</li>
<li>我们需要知道 <code>geneExpression</code> 这个矩阵中的顺序与<code>samkpleInfo</code>是一致的，要知道<code>sampleInfo</code>中的信息与<code>geneExpression</code>中的标签的对应关系。</li>
</ul>
<p>因此，我为了将基因表达矩阵与样本信息联系起来，我们需要对三个数据进行操作（分别为表达矩阵，样本信息，基因注释信息），从而确保这些信息能够匹配起来，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">all.equal(colnames(geneExpression), sampleInfo$filename)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">all.equal(rownames(geneExpression), geneAnnotation$PROBEID)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">ind = which(geneAnnotation$SYMBOL==<span class="string">"BRCA2"</span>)</div><div class="line">boxplot(geneExpression[ind,]~sampleInfo$ethnicity, ylab=<span class="string">"BRCA2 by hgfocus"</span>)</div></pre></td></tr></table></figure>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-dochks-1.png" alt="plot of chunk dochks"></p>
<p>虽然我们绘制出了图形，但是这略微复杂，为了降低操作难度，Bioconductor定义了一个名为<code>ExpressionSet</code>数据结构。</p>
<h3 id="ExpressionSet数据结构"><a href="#ExpressionSet数据结构" class="headerlink" title="ExpressionSet数据结构"></a>ExpressionSet数据结构</h3><p>我们使用简单的命令就能构建起ExpressionSet这种数据结构，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">es1 = ExpressionSet(geneExpression)</div><div class="line">pData(es1) = sampleInfo</div><div class="line">fData(es1) = geneAnnotation</div><div class="line">es1</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 24 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: 107 122 ... 159 (24 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename group</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 1 30 ... 12919 (8793 total)</span></div><div class="line"><span class="comment">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</span></div><div class="line"><span class="comment">##   fvarMetadata: labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation:</span></div></pre></td></tr></table></figure>
<p>上面的<code>es</code> 这个单一变量就浓缩了整个实验的数据，将原来的三个数据变为了一个。现在我们使用<code>es1</code>就能直接绘图，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(es1$date~es1$ethnicity)</div></pre></td></tr></table></figure>
<p>现在我们已经了解了<code>ExpressionSet</code>容器的设计，你自己也能构建一个ExpressionSet实例。我们将会在下面的实验中，使用一个更加广泛的数据表示方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSE5859)</div><div class="line">data(GSE5859)</div><div class="line">e</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 208 samples </span></div><div class="line"><span class="comment">##   element names: exprs, se.exprs </span></div><div class="line"><span class="comment">## protocolData</span></div><div class="line"><span class="comment">##   sampleNames: GSM25349.CEL.gz GSM25350.CEL.gz ...</span></div><div class="line"><span class="comment">##     GSM136729.CEL.gz (208 total)</span></div><div class="line"><span class="comment">##   varLabels: ScanDate</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: 1 2 ... 208 (208 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData: none</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation: hgfocus</span></div></pre></td></tr></table></figure>
<h3 id="元数据管理-Metadata-management"><a href="#元数据管理-Metadata-management" class="headerlink" title="元数据管理(Metadata management)"></a>元数据管理(Metadata management)</h3><p>及时地了解关于实验意图和方法的信息非常有用。GSE5859中使用的一些微阵列数据就是类来源于一篇发表于2004年（PubmedID为15269782）的论文。在联网的情况下，我们可以将论文的信息导入R，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(annotate)</div><div class="line">p = pmid2MIAME(<span class="string">"15269782"</span>)</div><div class="line">p</div><div class="line"><span class="comment">## Experiment data</span></div><div class="line"><span class="comment">##   Experimenter name: Morley M </span></div><div class="line"><span class="comment">##   Laboratory: NA </span></div><div class="line"><span class="comment">##   Contact information:  </span></div><div class="line"><span class="comment">##   Title: Genetic analysis of genome-wide variation in human gene expression. </span></div><div class="line"><span class="comment">##   URL:  </span></div><div class="line"><span class="comment">##   PMIDs: 15269782 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   Abstract: A 167 word abstract is available. Use 'abstract' method.</span></div></pre></td></tr></table></figure>
<p>现在查看你R环境中的<code>e</code>变量，使用<code>experimentData(e) &lt;- p</code>后，就能够将PubmedID信息添加到<code>e</code>中。使用<code>experimentData(e)</code>也能发现<code>e</code>中的信息改变了，也含有Pubmed ID信息。</p>
<h1 id="关于ExpressionSet的Why与How"><a href="#关于ExpressionSet的Why与How" class="headerlink" title="关于ExpressionSet的Why与How"></a>关于ExpressionSet的Why与How</h1><p>现在我们通过创建以及使用<code>ExpressionSet</code> 实例已经快速了解了这种数据结构的特点。我们现在将要了解一下，为什么要这样设计这种数据结构，以及如何使用这些数据结构。</p>
<p>设计<code>ExpressionSet</code> 这种数据结构的初衷有很多，以下几点最为关键：</p>
<ul>
<li>降低复杂分析，注释以及样本信息编程的难度；</li>
<li>支持特征值（其实就是表达值）和样本信息的简单的过滤；</li>
<li>在将分析数据传递给分析函数时控制内存占用。</li>
</ul>
<p>以上就是为什么要设计ExpressionSet这种数据结构。关于如何使用这种结构，这些内容就要涉及一些设计模式问题，在R中，与面向对象编程相关类被称为S4类。</p>
<h2 id="面向对象编程与S4类"><a href="#面向对象编程与S4类" class="headerlink" title="面向对象编程与S4类"></a>面向对象编程与S4类</h2><p>面向对象编程(OOP, Object-oriented programming)是计算机编程一个学科，它有助于降低程序的复杂性和维护要求，并能增强程度的互动性。OOP有着各种解决方式，即使在R中，也可以进行OOP，这就是所谓的S4方法。在R中，OOP的基本思想就是一个对象有些特征：</p>
<ul>
<li>将不同类型的数据集中放到一个统一的结构中，</li>
<li>它被标记为某个正式指定类的成员，并且可以正式地位于相关对象的系统中，</li>
<li>以类特异性方式对泛型方法进行响应。</li>
</ul>
<p>我们可以使用<code>getClass</code>函数查看一个<code>ExpresionSet</code>的类结构，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"ExpressionSet"</span>)</div><div class="line"><span class="comment">## Class "ExpressionSet" [package "Biobase"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                                </span></div><div class="line"><span class="comment">## Name:      experimentData          assayData          phenoData</span></div><div class="line"><span class="comment">## Class:              MIAME          AssayData AnnotatedDataFrame</span></div><div class="line"><span class="comment">##                                                                </span></div><div class="line"><span class="comment">## Name:         featureData         annotation       protocolData</span></div><div class="line"><span class="comment">## Class: AnnotatedDataFrame          character AnnotatedDataFrame</span></div><div class="line"><span class="comment">##                          </span></div><div class="line"><span class="comment">## Name:   .__classVersion__</span></div><div class="line"><span class="comment">## Class:           Versions</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "eSet", directly</span></div><div class="line"><span class="comment">## Class "VersionedBiobase", by class "eSet", distance 2</span></div><div class="line"><span class="comment">## Class "Versioned", by class "eSet", distance 3</span></div></pre></td></tr></table></figure>
<p>从上面我们可以看到，一共有7个组件定义了该类。只有<code>annoataion</code>这个组件占据了一个单独的R类型，也就是<code>character</code>类型。至于其它的组件，都是S4类的一个实例（或者说是更多的R数据类型）。</p>
<p>举例说明一下，<code>MIAME</code>类的定义如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"MIAME"</span>)</div><div class="line"><span class="comment">## Class "MIAME" [package "Biobase"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:               name               lab           contact</span></div><div class="line"><span class="comment">## Class:         character         character         character</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:              title          abstract               url</span></div><div class="line"><span class="comment">## Class:         character         character         character</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:          pubMedIds           samples    hybridizations</span></div><div class="line"><span class="comment">## Class:         character              list              list</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:       normControls     preprocessing             other</span></div><div class="line"><span class="comment">## Class:              list              list              list</span></div><div class="line"><span class="comment">##                         </span></div><div class="line"><span class="comment">## Name:  .__classVersion__</span></div><div class="line"><span class="comment">## Class:          Versions</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "MIAxE", directly</span></div><div class="line"><span class="comment">## Class "characterORMIAME", directly</span></div><div class="line"><span class="comment">## Class "Versioned", by class "MIAxE", distance 2</span></div></pre></td></tr></table></figure>
<p>因此，我们可以发现，S4可以让我们定义包含各种信息的新数据结。我们虽然可以使用R中非常通用的列表结构来实现这一目的（就是说实现封装不同信息），但是S4允许我们能保证有新结构的属性，以便为这些结构定义的程序不必“检查”它们正在操作什么。这些检查过程内置在了类系统中。</p>
<h2 id="数据框-DataFrame-表格数据的增强型结构"><a href="#数据框-DataFrame-表格数据的增强型结构" class="headerlink" title="数据框(DataFrame): 表格数据的增强型结构"></a>数据框(DataFrame): 表格数据的增强型结构</h2><p>数据框是R中非常基础的数据结构。Bioconductor中引入了<code>DataFrame</code>类，它可以实现其它一些额外的功能，后面我们会提到。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(S4Vectors)</div><div class="line">sa2 = DataFrame(sampleInfo)</div><div class="line">sa2</div><div class="line"><span class="comment">## DataFrame with 24 rows and 4 columns</span></div><div class="line"><span class="comment">##     ethnicity       date         filename     group</span></div><div class="line"><span class="comment">##      &lt;factor&gt;     &lt;Date&gt;      &lt;character&gt; &lt;numeric&gt;</span></div><div class="line"><span class="comment">## 1         ASN 2005-06-23 GSM136508.CEL.gz         1</span></div><div class="line"><span class="comment">## 2         ASN 2005-06-27 GSM136530.CEL.gz         1</span></div><div class="line"><span class="comment">## 3         ASN 2005-06-27 GSM136517.CEL.gz         1</span></div><div class="line"><span class="comment">## 4         ASN 2005-10-28 GSM136576.CEL.gz         1</span></div><div class="line"><span class="comment">## 5         ASN 2005-10-07 GSM136566.CEL.gz         1</span></div><div class="line"><span class="comment">## ...       ...        ...              ...       ...</span></div><div class="line"><span class="comment">## 20        ASN 2005-06-27 GSM136524.CEL.gz         0</span></div><div class="line"><span class="comment">## 21        ASN 2005-06-23 GSM136514.CEL.gz         0</span></div><div class="line"><span class="comment">## 22        ASN 2005-10-07 GSM136563.CEL.gz         0</span></div><div class="line"><span class="comment">## 23        ASN 2005-10-07 GSM136564.CEL.gz         0</span></div><div class="line"><span class="comment">## 24        ASN 2005-10-07 GSM136572.CEL.gz         0</span></div></pre></td></tr></table></figure>
<p><code>show</code>方法描述顶部和底部的行数据，提供有关列数据类型的信息，并且能够比<code>data.frame</code>数据结构更有效地管理列中的复杂数据类所，下面是类定义：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"DataFrame"</span>)</div><div class="line"><span class="comment">## Class "DataFrame" [package "S4Vectors"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:           rownames             nrows          listData</span></div><div class="line"><span class="comment">## Class: character_OR_NULL           integer              list</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:        elementType   elementMetadata          metadata</span></div><div class="line"><span class="comment">## Class:         character DataTable_OR_NULL              list</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "DataTable", directly</span></div><div class="line"><span class="comment">## Class "SimpleList", directly</span></div><div class="line"><span class="comment">## Class "DataTable_OR_NULL", by class "DataTable", distance 2</span></div><div class="line"><span class="comment">## Class "List", by class "SimpleList", distance 2</span></div><div class="line"><span class="comment">## Class "Vector", by class "SimpleList", distance 3</span></div><div class="line"><span class="comment">## Class "Annotated", by class "SimpleList", distance 4</span></div></pre></td></tr></table></figure>
<h2 id="源于多样本NGS数据的成熟数据表示方法"><a href="#源于多样本NGS数据的成熟数据表示方法" class="headerlink" title="源于多样本NGS数据的成熟数据表示方法"></a>源于多样本NGS数据的成熟数据表示方法</h2><p>短读长测序产生了远比微阵列多得多的数据。测序实验也可以通过更加多样化的手段进行分析，因此测序实验能够在不同粒度级别(at various levels of granularity)提供有效的输入。</p>
<p>处理RNA-seq数据的典型流程包括：</p>
<ol>
<li>展示每个周期中碱基所调用的色谱（2008年后就不再提供了）；</li>
<li>碱基的读取质量值，它们通常以FASTQ格式进行储存；</li>
<li>通过基因组或转录本比对后，对各个基因统计的结果；</li>
<li>基因组极的多个样本的数据储存，用于推断感兴趣的生物学问题。通常一个样本对应两个FASTQ文件。</li>
</ol>
<h3 id="Bioconductor中的BAM文件"><a href="#Bioconductor中的BAM文件" class="headerlink" title="Bioconductor中的BAM文件"></a>Bioconductor中的BAM文件</h3><p><code>RNAseqData.HNRNPC.bam.chr14</code> 中含有一些BAM文件，这些BAM文件是8个Hela细胞系样本的双端测序数据，其中4个样本通过RNAi的方式敲低了异质核核糖核蛋白C1和C2（hnRNPC,heterogeneous nuclear ribonucleoproteins C1 and C2)。这个家庭的核蛋白参与pre-mRNA的加工。现在我们演示一下，如何从BAM文件中来统计与HNRNPC有关的读长数目(reads count)。</p>
<p>BAM文件储存在<code>RNAseqData.HNRNPC.bam.chr14</code> 包中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">bfp = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line">length(bfp)</div><div class="line"><span class="comment">## [1] 8</span></div><div class="line">bfp[<span class="number">1</span>]</div><div class="line"><span class="comment">##                                                                                                                ERR127306 </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RNAseqData.HNRNPC.bam.chr14/extdata/ERR127306_chr14.bam"</span></div></pre></td></tr></table></figure>
<p>我们使用<code>Rsamtools</code>包以一个变量的形式管理这些信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Rsamtools)</div><div class="line">bfl = BamFileList(file=bfp)</div><div class="line">bfl</div><div class="line"><span class="comment">## BamFileList of length 8</span></div><div class="line"><span class="comment">## names(8): ERR127306 ERR127307 ERR127308 ... ERR127303 ERR127304 ERR127305</span></div></pre></td></tr></table></figure>
<p>BAM文件包括关于读长(reads)与基因组比对后的有限数量的元数据。这里我们以<code>BamFileList</code>实例来说明一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">seqinfo(bfl)</div><div class="line"><span class="comment">## Seqinfo object with 93 sequences from an unspecified genome:</span></div><div class="line"><span class="comment">##   seqnames              seqlengths isCircular genome</span></div><div class="line"><span class="comment">##   chr1                   249250621       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr10                  135534747       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr11                  135006516       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr11_gl000202_random      40103       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr12                  133851895       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ...                          ...        ...    ...</span></div><div class="line"><span class="comment">##   chrUn_gl000247             36422       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrUn_gl000248             39786       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrUn_gl000249             38502       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrX                   155270560       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrY                    59373566       &lt;NA&gt;   &lt;NA&gt;</span></div></pre></td></tr></table></figure>
<p>我们看到基因”没有指定“，但是我们碰巧知道参考基因组<code>hg19</code>用于进入基因序列的比对。使用<code>[genome(bfl) &lt;- “hg19”</code>可以纠正这点，但是还存在着一个bug。这个bug应该修正一下，以便可以标记使用不同的参考基因组来进行比对或纠正错误，注：这段不太理解，原文如下：</p>
<blockquote>
<p>We see that the genome is “unspecified”, but we happen to know that the reference build ‘hg19’ was used to define the genomic sequence for alignment.  [genome(bfl) &lt;- “hg19” should rectify this but at present there is a bug.  This should be fixed so that attempts to compare alignments using different<br>references can be flagged or throw errors.]</p>
</blockquote>
<p>为了验证是否成功敲除，我们将检查编码HNRNPC的基因组附近的序列的比对情况。稍后，我们将展示如何使用Bioconductor注释工具获取基因地址；不过这个地址仅用于hg19基因组，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">hnrnpcLoc = GRanges(<span class="string">"chr14"</span>, IRanges(<span class="number">21677296</span>, <span class="number">21737638</span>))</div><div class="line">hnrnpcLoc</div><div class="line"><span class="comment">## GRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames               ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]    chr14 [21677296, 21737638]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p>这标识了HNRNPC编码基因为hg19注释的染色体14的区域。评估表达式量化的一种简单方法是使用GenomicAlignmentspackage中的summary izeOverlaps。此计数方法将隐式使用多核计算。</p>
<p>上面的代码标识了14号染色体中编码HNRNPC附近的序列。评估基因表达量化的一种简单方法就是使用<code>GenomicAlignmentspackage</code>包中的<code>summarizeOverlaps</code> 函数，这个计算方法隐式地使用了多核计算，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line"><span class="keyword">library</span>(BiocParallel)</div><div class="line">register(SerialParam())</div><div class="line">hnse = summarizeOverlaps(hnrnpcLoc,bfl)</div></pre></td></tr></table></figure>
<h3 id="练习"><a href="#练习" class="headerlink" title="练习"></a>练习</h3><ol>
<li>检查<code>hnse</code>中的<code>count</code>组件内容。与HNRNPC敲除相对应的有关量化的指标是什么？</li>
<li>向<code>hnse</code>的<code>colData</code>组件中添加一个变量<code>condition</code>，用于指示敲除株与野生株。</li>
<li>具有后缀306和307的样本是单个对照的技术重复；308和309同样如此。因此使用了6个不同的生物学样本。现在向<code>colData</code>中添加一个变量，用于区别这6个样本。</li>
</ol>
<h2 id="ExpressionSet设计的一些复杂性-进阶"><a href="#ExpressionSet设计的一些复杂性-进阶" class="headerlink" title="ExpressionSet设计的一些复杂性(进阶)"></a>ExpressionSet设计的一些复杂性(进阶)</h2><p>生物信息学家是在一个不断变化的环境中工作。自Bioconductor建立以来，计算机的容量和通量已经大幅度增加。R语言也性现了很大的变量。在早期的时候，大家都比较注意内存的使用，因为当时在R中，向一个函数传递参数时，会复制内存中的对象， 这种操作比较低效。而“环境”类中的R对象则不会发生复制，这就导致了使用环境来储存大量的数据用于分析。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">class(slot(es1, <span class="string">"assayData"</span>))</div><div class="line"><span class="comment">## [1] "environment"</span></div><div class="line">ae = slot(es1, <span class="string">"assayData"</span>)</div><div class="line">ae</div><div class="line"><span class="comment">## &lt;environment: 0x7f813b4dc698&gt;</span></div></pre></td></tr></table></figure>
<p>“环境”是R中一种非常特殊的对象类型，可以将它视为对表达值矩阵引用的容器。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">ls(ae)</div><div class="line"><span class="comment">## [1] "exprs"</span></div><div class="line">dim(ae$exprs)</div><div class="line"><span class="comment">## [1] 8793   24</span></div><div class="line">dim(exprs(es1))</div><div class="line"><span class="comment">## [1] 8793   24</span></div><div class="line">object.size(ae)</div><div class="line"><span class="comment">## 56 bytes</span></div><div class="line">object.size(ae$exprs)</div><div class="line"><span class="comment">## 2253824 bytes</span></div></pre></td></tr></table></figure>
<h3 id="练习（可选）"><a href="#练习（可选）" class="headerlink" title="练习（可选）"></a>练习（可选）</h3><p>请解释以下运行结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">e1 = new.env()</div><div class="line">l1 = list()</div><div class="line">e1$x = <span class="number">5</span></div><div class="line">l1$x = <span class="number">5</span></div><div class="line">all.equal(e1$x, l1$x)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">f = <span class="keyword">function</span>(z) &#123;z$x = <span class="number">4</span>; z$x&#125;</div><div class="line">all.equal(f(e1), f(l1))</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">all.equal(e1$x, l1$x)  <span class="comment"># different from before</span></div><div class="line"><span class="comment">## [1] "Mean relative difference: 0.25"</span></div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">&gt; e1 = new.env()</div><div class="line">&gt; l1 = list()</div><div class="line">&gt; e1$x = <span class="number">5</span></div><div class="line">&gt; l1$x = <span class="number">5</span></div><div class="line">&gt; all.equal(e1$x, l1$x)</div><div class="line">[<span class="number">1</span>] <span class="literal">TRUE</span></div><div class="line">&gt; <span class="comment">## [1] TRUE</span></div><div class="line">&gt; f = <span class="keyword">function</span>(z) &#123;z$x = <span class="number">4</span>; z$x&#125;</div><div class="line">&gt; all.equal(f(e1), f(l1))</div><div class="line">[<span class="number">1</span>] <span class="literal">TRUE</span></div><div class="line">&gt; <span class="comment">## [1] TRUE</span></div><div class="line">&gt; all.equal(e1$x, l1$x)  <span class="comment"># different from before</span></div><div class="line">[<span class="number">1</span>] <span class="string">"Mean relative difference: 0.25"</span></div><div class="line">&gt; <span class="comment">## [1] "Mean relative difference: 0.25"</span></div></pre></td></tr></table></figure>
<h2 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h2><ol>
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_mgt_gsd.html" target="_blank" rel="external">Management of genome-scale data</a> </li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/" rel="next" title="GDAS003-Bioconductor与基因组级数据分析简介">
                <i class="fa fa-chevron-left"></i> GDAS003-Bioconductor与基因组级数据分析简介
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/05/Genomics Data Analysis Series/GDAS005-基因组实验中的数据结构与管理/" rel="prev" title="GDAS005-基因组实验中的数据结构与管理">
                GDAS005-基因组实验中的数据结构与管理 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">249</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">118</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#案例说明-Youtube上有相应视频"><span class="nav-number">1.1.</span> <span class="nav-text">案例说明 (Youtube上有相应视频)</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#ExpressionSet数据结构"><span class="nav-number">1.1.1.</span> <span class="nav-text">ExpressionSet数据结构</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#元数据管理-Metadata-management"><span class="nav-number">1.1.2.</span> <span class="nav-text">元数据管理(Metadata management)</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#关于ExpressionSet的Why与How"><span class="nav-number">2.</span> <span class="nav-text">关于ExpressionSet的Why与How</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#面向对象编程与S4类"><span class="nav-number">2.1.</span> <span class="nav-text">面向对象编程与S4类</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#数据框-DataFrame-表格数据的增强型结构"><span class="nav-number">2.2.</span> <span class="nav-text">数据框(DataFrame): 表格数据的增强型结构</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#源于多样本NGS数据的成熟数据表示方法"><span class="nav-number">2.3.</span> <span class="nav-text">源于多样本NGS数据的成熟数据表示方法</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Bioconductor中的BAM文件"><span class="nav-number">2.3.1.</span> <span class="nav-text">Bioconductor中的BAM文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#练习"><span class="nav-number">2.3.2.</span> <span class="nav-text">练习</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#ExpressionSet设计的一些复杂性-进阶"><span class="nav-number">2.4.</span> <span class="nav-text">ExpressionSet设计的一些复杂性(进阶)</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#练习（可选）"><span class="nav-number">2.4.1.</span> <span class="nav-text">练习（可选）</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">2.5.</span> <span class="nav-text">参考资料</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共981.6k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
