<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Hexo, NexT" />










<meta property="og:type" content="website">
<meta property="og:title" content="RVDSD的个人笔记本">
<meta property="og:url" content="http://rvdsd.top/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RVDSD的个人笔记本">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/"/>





  <title>RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/10/16/BioStatistics/标准差为什么除以n-1/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/10/16/BioStatistics/标准差为什么除以n-1/" itemprop="url">标准差为什么除以n-1</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-10-16T12:10:54+08:00">
                2019-10-16
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,749
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  7
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>在学习统计学的时候，我遇到过这么一个问题，也就是说，样本的标准差公式，如下所示： <span class="math display">\[
s=\sqrt{s^{2}}=\sqrt{\frac{S S}{n-1}}
\]</span> 很多统计学书上都提到，在样本标准差的计算公式中，平方根中的分子是 <span class="math inline">\(n-1\)</span>，而总体标准差则是 <span class="math inline">\(n\)</span> 。其理由是为了校正样本变异性而做出的调整，这是对总体标准差的无偏估计。</p>
<p>但是，为什么说这是一种无偏估计，很多书中并没有提及，或者说是只用了很粗略的语言简单地说了一下，其实也没必过于纠结这个问题，记住就行。但是，如果实在是想弄明白这个问题，网上也有人给出了证明过程，但是证明过程对于没有数学基础的人来讲，还是有点难的，这个完整的证明过程的可以参考知乎上的这个帖子<a href="https://www.zhihu.com/question/20099757?rf=21126585" target="_blank" rel="external">《为什么样本方差(sample variance)的分子是n-1》</a>。</p>
<p>最近我看到了一本统计学的书《行为科学统计》（第七版）作者：[美]FrederickJ Gravetter，这本书中对这个问题的描述很清楚，通过用举例子的方式说明了一下（并非严格证明），为什么在样本标准差中，使用 <span class="math inline">\(n-1\)</span> 是对总体方差的无偏估计。</p>
<p>另外说明一下，《行为科学统计》这本书原本就是给社会学的学生学习统计学准备的，里面的语言浅显易懂，没有复杂的公式，对于数学功底差的学生来说，非常友好，最新一版已经到了第9版。</p>
<h2 id="背景知识">背景知识</h2>
<ol style="list-style-type: decimal">
<li><strong>离差</strong>：数据到平均数的距离，例如对于一个 <span class="math inline">\(\mu = 50\)</span> 的分布来说，如果你的一个数据是 <span class="math inline">\(X=53\)</span> ，那么离差就是 <span class="math inline">\(X-\mu = 53 - 50 = 3\)</span>。如果数据是45，那么离差就是 <span class="math inline">\(45-50=-5\)</span>。</li>
<li><strong>离均差平方和（SS，sum of squares of deviation from mean）</strong>：由于离差有正有负，最终所有离均差的和即 <span class="math inline">\((X-\mu)\)</span> 为0，因此离均差的和无法描述一组数据的变异大小。因此将离均差平方后相加得到平方和<span class="math inline">\(Var(X)=E(X-\mu)^2\)</span>，这就是离均差平方和（sum of squares of deviations from mean, SS）。</li>
<li><strong>方差</strong>：方差定义为离均差平方和的平均数，如下所示：</li>
</ol>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191016130630.png">

</div>
<ol start="4" style="list-style-type: decimal">
<li><strong>标准差</strong>：方差的平方根。</li>
</ol>
<h2 id="计算过程">计算过程</h2>
<p>先来看一组数据，即<code>1, 9, 5, 8, 7</code>，我们把这个数据当作是总体，现在我们计算它的离差，离差的平方等，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">a &lt;- c(<span class="number">1</span>,<span class="number">9</span>, <span class="number">5</span>, <span class="number">8</span>, <span class="number">7</span>) <span class="comment"># 原始数据</span></div><div class="line">a - mean(a) <span class="comment"># 离差</span></div><div class="line">(a- mean(a))^<span class="number">2</span> <span class="comment"># 离差的平方</span></div><div class="line">sum((a- mean(a))^<span class="number">2</span>) <span class="comment"># 离均差平方和</span></div><div class="line">sum((a- mean(a))^<span class="number">2</span>)/length(a) <span class="comment"># 方差</span></div><div class="line">sqrt(sum((a- mean(a))^<span class="number">2</span>)/length(a)) <span class="comment">#标准差</span></div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">&gt; a &lt;- c(<span class="number">1</span>,<span class="number">9</span>, <span class="number">5</span>, <span class="number">8</span>, <span class="number">7</span>) <span class="comment"># 原始数据</span></div><div class="line">&gt; a - mean(a) <span class="comment"># 离差</span></div><div class="line">[<span class="number">1</span>] -<span class="number">5</span>  <span class="number">3</span> -<span class="number">1</span>  <span class="number">2</span>  <span class="number">1</span></div><div class="line">&gt; (a- mean(a))^<span class="number">2</span> <span class="comment"># 离差的平方</span></div><div class="line">[<span class="number">1</span>] <span class="number">25</span>  <span class="number">9</span>  <span class="number">1</span>  <span class="number">4</span>  <span class="number">1</span></div><div class="line">&gt; sum((a- mean(a))^<span class="number">2</span>) <span class="comment"># 离差的平方和</span></div><div class="line">[<span class="number">1</span>] <span class="number">40</span></div><div class="line">&gt; sum((a- mean(a))^<span class="number">2</span>)/length(a) <span class="comment"># 方差</span></div><div class="line">[<span class="number">1</span>] <span class="number">8</span></div><div class="line">&gt; sqrt(sum((a- mean(a))^<span class="number">2</span>)/length(a)) <span class="comment">#标准差</span></div><div class="line">[<span class="number">1</span>] <span class="number">2.828427</span></div></pre></td></tr></table></figure>
<h2 id="总体方差与样本方差">总体方差与样本方差</h2>
<p>总体方差的计算公式就是方差的定义：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191016130134.png">

</div>
<p>总体标准差就是总体方差的平方根，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191016130243.png">

</div>
<p>样本方差与总体方差有所不同，为了校正样本变异性，我们需要对方差和标准有效期的公式做出调整，此时样本方差公式中的分母是 <span class="math inline">\(n-1\)</span>， 如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191016123038.png">

</div>
<p>样本标准差的公式如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191016123229.png">

</div>
<p>这里要注意的是，公式使用了 <span class="math inline">\(n-1\)</span> 来代替 <span class="math inline">\(n\)</span> ，这是为了校正样本变异性的偏误做出的调整，调整的结果使所得的结果变大，从而使样本方差成为对总体方差精确的无偏估计（如果是n，则是有偏估计）。</p>
<p>下面我们用不太严谨的一个案例来说明一下为什么样本方差中的分母是 <span class="math inline">\(n-1\)</span> 。</p>
<h2 id="举例说明为什么是n-1">举例说明为什么是n-1</h2>
<p>现在我们设计一个N=6的总体，它的元素为<code>0, 0, 3, 3, 9, 9</code>，现在我们计算可知它的总体均数为 <span class="math inline">\(\mu=4\)</span>， 方差 <span class="math inline">\(\sigma^2 = 14\)</span>。</p>
<p>现在我们从这个总体中选择一个 <span class="math inline">\(n=2\)</span> 的样本，我们选出所有可能的组合，并计算出其平均数，有偏方差，无偏方差，如下所示：</p>
<table>
<thead>
<tr class="header">
<th>样本编号</th>
<th>第1个数</th>
<th>第2个数</th>
<th>平均数</th>
<th>有偏的方差(n)</th>
<th>无偏的方差(n-1)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>2</td>
<td>0</td>
<td>3</td>
<td>1.5</td>
<td>2.25</td>
<td>4.5</td>
</tr>
<tr class="odd">
<td>3</td>
<td>0</td>
<td>9</td>
<td>4.5</td>
<td>20.25</td>
<td>40.5</td>
</tr>
<tr class="even">
<td>4</td>
<td>3</td>
<td>0</td>
<td>1.5</td>
<td>2.25</td>
<td>4.5</td>
</tr>
<tr class="odd">
<td>5</td>
<td>3</td>
<td>3</td>
<td>3</td>
<td>.</td>
<td>0</td>
</tr>
<tr class="even">
<td>6</td>
<td>3</td>
<td>9</td>
<td>6</td>
<td>9</td>
<td>18</td>
</tr>
<tr class="odd">
<td>7</td>
<td>9</td>
<td>0</td>
<td>4.5</td>
<td>20.25</td>
<td>40.5</td>
</tr>
<tr class="even">
<td>8</td>
<td>9</td>
<td>3</td>
<td>6</td>
<td>9</td>
<td>18</td>
</tr>
<tr class="odd">
<td>9</td>
<td>9</td>
<td>9</td>
<td>9</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>总和</td>
<td>36</td>
<td>63</td>
<td>126</td>
</tr>
</tbody>
</table>
<p>现在我们观察平均数这一列，原始的总体均数为 <span class="math inline">\(\mu = 4\)</span>。虽然没有一个样本的均数恰好为4，但是如果考虑整组样本，将会发现，9个样本的平均数总和为36，因此样本均数数的平均数为 36/9=4，此时样本平均数恰好等于总体平均数。根据定义，这是一个无偏的统计量，也就是说，样本精确地代表了总体。</p>
<p>现在我们考虑用除以n得到的存在偏误的样本方差这一列。原始的总体方差是 <span class="math inline">\(\sigma^2 = 14\)</span>。 然而，9个样本方差的总和为63， 这使得63/9=7。注意，这些样本方差的平均值不等于总体方差，也就是说，如果用除以n得到的样本方差，得出的结果不能精确估计总体方差， 也就是说，这些样本方差低估了总体方差，因此是存在偏误的统计量。</p>
<p>现在我们再考虑除了n-1得到的样本详这一列，虽然总体方差为<span class="math inline">\(\sigma^2=14\)</span>，然而没有一个样本的方差恰好等于14。但是，如果考虑整组样本方差，将会发现这9个值总和为126，因此方差的平均值为126/9=14。因此，样本方差的平均值恰好等于总体方差。也就是说，样本方差（此时是使用了n-1来代替n）是对总体方差的一个精确的、无偏的估计。</p>
<p>结论就是，样本平均数和样本方差（使用n-1）都是无偏估计的例子。这个事实使样本平均数和样本方差在推论统计方面变得非常重要。虽然没有单个样本恰好具有与总体一样的平均数和方差，但是，样本平均数和样本方差的平均值确实提供了对相应总体参数的精确估计。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>行为科学统计 作者: [美] F. J. Gravetter / [美] L. B. Wallnau，出版社: 中国轻工业出版社，原作名: Statistics for the behavioral sciences，译者: 王爱民 / 李悦，出版年: 2008-7</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/10/15/R/R语言多线程/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/10/15/R/R语言多线程/" itemprop="url">R语言的多线程</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-10-15T12:00:00+08:00">
                2019-10-15
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index">
                    <span itemprop="name">R</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  410
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  2
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这是一棵树（网名）同学在群里分享的教程，有关R语言多线程的内容。</p>
<p>关于多线程是什么，可以搜一下，网上很多资料。</p>
<p>R的一个比较大的缺点就是它只能单线程操作，例如如果你的电脑有200个处理器，R一次只能使用一个，这个时候工作效率会降低。为了解决这个问题，可以调用R的多线程。</p>
<p>R中与多线程有关的包是<code>parallel</code>。</p>
<h2 id="parallel包"><code>parallel</code>包</h2>
<p><code>paralle</code>包是R中的基础包，不用额外安装，直接调用即可，它的使用如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># Load parallel package</span></div><div class="line"><span class="keyword">library</span>(parallel)</div><div class="line"></div><div class="line"><span class="comment"># Check how many threads and cores in your computer</span></div><div class="line">detectCores()</div><div class="line">detectCores(logical = <span class="literal">F</span>)</div><div class="line"></div><div class="line"><span class="comment"># Open multi-threads</span></div><div class="line"><span class="comment"># You will see 4 R programs run in your taks managers simultaneously</span></div><div class="line">cl &lt;- makeCluster(<span class="number">3</span>)</div><div class="line"></div><div class="line"><span class="comment"># Close cluster</span></div><div class="line"><span class="comment"># Run following command will close 4 R program</span></div><div class="line">stopCluster(cl)</div></pre></td></tr></table></figure>
<p><code>detectCores()</code>这个命令用来查看你的电脑有几个核心，如果添加上<code>logical=F</code>参数，则表示查看电脑有几个物理核心，如果是<code>logical=T</code>则表示有几个逻辑核心。</p>
<p><code>makecluster()</code>函数用于创建几个并行运算。</p>
<p>我们先来看一下运行这个函数之前的任务管理器，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191015200553.png">

</div>
<p>只有一个R，现在我们使用并行计算，即运行这个<code>cl &lt;- makeCluster(3)</code>命令，再看一下任务管理器，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191015200655.png">

</div>
<p>现在再关闭这个并行计算，即<code>stopCluster(cl)</code>，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191015200741.png">

</div>
<p>使用并行计算的思路就是：打开并行计算，计算，关闭并行计算。</p>
<p>并行计算的分配</p>
<p>当我们使用<code>makeCluster(4)</code>打开了4个R后，如何分配计算任务呢？</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/10/12/R/R语言的爬虫/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/10/12/R/R语言的爬虫/" itemprop="url">R语言的爬虫</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-10-12T12:00:00+08:00">
                2019-10-12
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index">
                    <span itemprop="name">R</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,838
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  18
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这个教程是一棵树（网名）演示的爬虫笔记，他的Github为：https://github.com/yikeshu0611</p>
<p>爬取科学网，网址为：http://fund.sciencenet.cn/search/smallSubject?subject=H0101&amp;yearStart=2018&amp;yearEnd=2018&amp;submit=list</p>
<p>内容如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012175107.png">

</div>
<p>爬虫的思路就是：</p>
<ol style="list-style-type: decimal">
<li>读取网页；</li>
<li>提取数据。</li>
</ol>
<h2 id="r包">R包</h2>
<p>使用<code>rvest</code>包中的<code>read_html()</code>函数提取网页中的内容。</p>
<h2 id="读取国自然操作">读取国自然操作</h2>
<h3 id="读取网页">1. 读取网页</h3>
<ol style="list-style-type: decimal">
<li>安装并加载<code>rvest</code>包；</li>
<li>将网址赋值给url；</li>
<li>使用<code>read_html()</code>函数读取，如下所示：</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">install.packages(<span class="string">"rvest"</span>)</div><div class="line"><span class="keyword">library</span>(rvest)</div><div class="line">url=<span class="string">'http://fund.sciencenet.cn/search/smallSubject?subject=H0101&amp;yearStart=2018&amp;yearEnd=2018&amp;submit=list'</span></div><div class="line">read_html(url)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; read_html(url)</div><div class="line">&#123;html_document&#125;</div><div class="line">&lt;html xmlns=<span class="string">"http://www.w3.org/1999/xhtml"</span>&gt;</div><div class="line">[<span class="number">1</span>] &lt;head&gt;\n&lt;meta http-equiv=<span class="string">"Content-Type"</span> content=<span class="string">"text/html; charset=UTF-8"</span> <span class="keyword">...</span></div><div class="line">[<span class="number">2</span>] &lt;body&gt;\n        &lt;div id=<span class="string">"hd"</span>&gt;\n    &lt;div class=<span class="string">"wp"</span>&gt;\n                &lt;ifra <span class="keyword">...</span></div></pre></td></tr></table></figure>
<p>上面的结果就是网页结构，其中<code>&lt;head&gt;</code>是网页的头部，<code>&lt;body&gt;</code>是网页的主体。</p>
<h3 id="读取数据">2. 读取数据</h3>
<p>读取数据则是要定位从哪里开始读取，还是先回到网页部分，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012175839.png">

</div>
<p>把鼠标放到上面的题目上，然后单击右键，选择审查元素（chrome浏览器），如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012180023.png">

</div>
<p>在上面网址那一行单击右键，<code>复制</code>-&gt;<code>Xpath</code>，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012180536.png">

</div>
<p>复制的内容是<code>//*[@id=&quot;resultLst&quot;]/div[1]/p/a</code>。现在复制另外一个题目的xpath，内容为<code>//*[@id=&quot;resultLst&quot;]/div[2]/p/a</code>。</p>
<p>从这两个内容上我们可以大概知道，<code>id=&quot;resultLst&quot;</code>对应了<code>&lt;div id=&quot;resultLst&quot;--------&lt;/div&gt;</code>，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012181120.png">

</div>
<p>在上面的html代码中，我们哦可以发现，这一部分有2个<code>div</code>，第一个题目是<code>div[1]</code>，第2个题目是<code>div[2]</code>，依次类推，最后一个就是<code>div[last()]</code>（不要过于纠结这种写法）。</p>
<p>现在我们看第1个div，div下面是p节点，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012181427.png">

</div>
<p>p节点下面又有2个节点，b和a，b节点那里是1，就是项目前面的标号，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012181527.png">

</div>
<p>a节点下面是<code>href=&quot;...&quot;</code>，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012181613.png">

</div>
<p>我们可以看到，在a节点现在有2个内容，第1个是链接，第2个是文本，也就是标题，我们的目标就是这个项目标题，现在我们从div那个节点开始，来写这个标题的地址，这个网址的结果如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012181740.png">

</div>
<p>在<code>rvest</code>包中，网页的定位是使用<code>html_nodes()</code>函数，现在我们定位第1个标题的位置，现在将读取的网页赋值给<code>content</code>，来定位网页中的某个东西，例如标题1，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">content &lt;- read_html(url)</div><div class="line">html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[1]/p/a'</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; content &lt;- read_html(url)</div><div class="line">&gt; </div><div class="line">&gt; html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[1]/p/a'</span>)</div><div class="line">&#123;xml_nodeset (<span class="number">1</span>)&#125;</div><div class="line">[<span class="number">1</span>] &lt;a href=<span class="string">"http://fund.sciencenet.cn/project/50 ...</span></div></pre></td></tr></table></figure>
<p>标题的xpath地址赋值给xpath，上面的结果就是相应的内容，里面就是一个文本，我们使用<code>html_text()</code>函数来提取这些内容，并将定位的内容赋值给<code>location</code>，然后再提取，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[1]/p/a'</span>)</div><div class="line">html_text(location)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[1]/p/a'</span>)</div><div class="line">&gt; html_text(location)</div><div class="line">[<span class="number">1</span>] <span class="string">"CFTR/EGFR反馈环路调控肺液清除功能在支气管肺发育不良发病中的作用和分子机制"</span></div></pre></td></tr></table></figure>
<p>现在我们提取第2个标题,如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[2]/p/a'</span>)</div><div class="line">html_text(location)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[2]/p/a'</span>)</div><div class="line">&gt; html_text(location)</div><div class="line">[<span class="number">1</span>] <span class="string">"II型肺泡上皮细胞（AT2）在重症流感肺泡损伤修复过程中的参与作用及调控机制"</span></div></pre></td></tr></table></figure>
<p>其中改变的就是<code>div[2]</code>这个参数。如果是要提取最后1个题目，就是<code>div[last()]</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]/div[last()]/p/a'</span>)</div><div class="line">&gt; html_text(location)</div><div class="line">[<span class="number">1</span>] <span class="string">"II型肺泡上皮细胞（AT2）在重症流感肺泡损伤修复过程中的参与作用及调控机制"</span></div></pre></td></tr></table></figure>
<p>如果是100个题目，不能这么干，我们需要把<code>div[1]</code>这类字符串删掉，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]//p/a'</span>)</div><div class="line">&gt; html_text(location)</div><div class="line">[<span class="number">1</span>] <span class="string">"CFTR/EGFR反馈环路调控肺液清除功能在支气管肺发育不良发病中的作用和分子机制"</span></div><div class="line">[<span class="number">2</span>] <span class="string">"II型肺泡上皮细胞（AT2）在重症流感肺泡损伤修复过程中的参与作用及调控机制"</span></div></pre></td></tr></table></figure>
<p>现在就提取了所有的题目。</p>
<p>第二个任务：提取姓名。</p>
<p>现在我们再来提取作者的姓名，例如<code>赵冬莹</code>，前面的xpath操作和前面的相同，即<code>//*[@id=&quot;resultLst&quot;]/div[1]/div/p[1]/span[1]/i</code>，现在再来复制第2个名字，即<code>//*[@id=&quot;resultLst&quot;]/div[2]/div/p[1]/span[1]/i</code>，复制2个名字主要是为了找到规律。现在我们把<code>div[1]</code>删掉，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]//div/p[1]/span[1]/i'</span>)</div><div class="line">html_text(location)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]//div/p[1]/span[1]/i'</span>)</div><div class="line">&gt; html_text(location)</div><div class="line">[<span class="number">1</span>] <span class="string">"赵冬莹"</span> <span class="string">"李辉"</span></div></pre></td></tr></table></figure>
<p>第三个任务：提取标题部分的网址，这个网址，就是标题后面链接的网址，有时候，我们需要爬取二级页面，就地需要获得二级页面的网址，我们看到这个网址不是文本，它虽然和标题在同一个位置，都是a节点下面，但是我们使用<code>html_text()</code>命令并没有将其提取出现，因为这个函数认为它不是文本，而是链接，对应的是<code>herf=&quot;----------------&quot;</code>这种格式，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012190255.png">

</div>
<p>现在我们要提取某一个具体的网页(html)属性(attribute)内容，此时我们使用<code>html_attr()</code>命令，例如我们要提取超链接，就写成<code>html_attr(&quot;href&quot;)</code>，所以，如果我们要提取标题处的链接，就需要先定位到标题那里，然后使用<code>html_attr()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]//p/a'</span>)</div><div class="line">html_attr(location, name = <span class="string">'href'</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; location &lt;- html_nodes(x = content, xpath = <span class="string">'//*[@id="resultLst"]//p/a'</span>)</div><div class="line">&gt; html_attr(location, name = <span class="string">'href'</span>)</div><div class="line">[<span class="number">1</span>] <span class="string">"http://fund.sciencenet.cn/project/509194"</span></div><div class="line">[<span class="number">2</span>] <span class="string">"http://fund.sciencenet.cn/project/509195"</span></div></pre></td></tr></table></figure>
<h2 id="读取pubmed">读取Pubmed</h2>
<p>现在来讲一下大致思路：第一，找到网址；第二，定位，也就是说从哪个地方开始抓取数据；第三步，抓取数据。</p>
<ol style="list-style-type: decimal">
<li>打开pubmed，https://www.ncbi.nlm.nih.gov/pubmed，输入circulation，点击搜索如下所示：</li>
</ol>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012191518.png">

</div>
<ol start="2" style="list-style-type: decimal">
<li>加载rvest包，输入目标网址，如下所示：</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## Crawl pubmed</span></div><div class="line"><span class="comment">## Setup 1</span></div><div class="line">page_content &lt;- read_html(x = <span class="string">'https://www.ncbi.nlm.nih.gov/pubmed/?term=circulation'</span>)</div></pre></td></tr></table></figure>
<ol start="3" style="list-style-type: decimal">
<li>像前面一样，右键xpath，如下所示：</li>
</ol>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012193009.png">

</div>
<p>其中，一个rprt对应的就是左侧的蓝色阴影部分，一共有20个这样的结构（其实就是一页中的20个结果），如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012193151.png">

</div>
<p>我们再回到第1个题目，右键，copy xpath，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## Crawl pubmed</span></div><div class="line"><span class="comment">## Setup 1</span></div><div class="line">page_content &lt;- read_html(x = <span class="string">'https://www.ncbi.nlm.nih.gov/pubmed/?term=circulation'</span>)</div><div class="line"></div><div class="line"><span class="comment">## Step 2: Crawl content</span></div><div class="line"><span class="comment">#xpath</span></div><div class="line">node = <span class="string">'//*[@id="maincontent"]/div/div[5]/div[1]/div[2]/p/a'</span></div><div class="line">html_nodes(x = page_content, xpath = node)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; html_nodes(x = page_content, xpath = node)</div><div class="line">&#123;xml_nodeset (<span class="number">1</span>)&#125;</div><div class="line">[<span class="number">1</span>] &lt;a href=<span class="string">"/pubmed/31603578"</span> ref=<span class="string">"ordinalpos=1&amp;amp;ncbi_uid=31603578&amp;amp;link_ ...</span></div></pre></td></tr></table></figure>
<p>读取上面内容中的文本部分，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## Step 2: Crawl content</span></div><div class="line"><span class="comment">#xpath</span></div><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//*[@id="maincontent"]/div/div[5]/div[1]/div[2]/p/a'</span></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"></div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_text(x = nodes_content, trim = <span class="literal">T</span>)</div></pre></td></tr></table></figure>
<p>现在我们提取第2个题目的xpath，再将它与第1个比较，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">//*[@id=<span class="string">"maincontent"</span>]/div/div[<span class="number">5</span>]/div[<span class="number">1</span>]/div[<span class="number">2</span>]/p/a</div><div class="line">//*[@id=<span class="string">"maincontent"</span>]/div/div[<span class="number">5</span>]/div[<span class="number">2</span>]/div[<span class="number">2</span>]/p/a</div></pre></td></tr></table></figure>
<p>我们可以找到它们的共同规律，即<code>//*[@id=&quot;maincontent&quot;]/div/div[5]//div[2]/p/a</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//*[@id="maincontent"]/div/div[5]//div[2]/p/a'</span></div><div class="line"><span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line"></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"></div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_text(x = nodes_content, trim = <span class="literal">T</span>)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment">### 2.1 Location</span></div><div class="line">&gt; node = <span class="string">'//*[@id="maincontent"]/div/div[5]//div[2]/p/a'</span></div><div class="line">&gt; nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment">### 2.2 Crawl data</span></div><div class="line">&gt; html_text(x = nodes_content, trim = <span class="literal">T</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"Variation of organ position in snakes."</span>                                                                                                                                   </div><div class="line"> [<span class="number">2</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line"> [<span class="number">3</span>] <span class="string">"Pathogenesis of Cluster Headache: From Episodic to Chronic Form, the Role of Neurotransmitters and Neuromodulators."</span>                                                      </div><div class="line"> [<span class="number">4</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line"> [<span class="number">5</span>] <span class="string">"Haemodynamic evaluation of the new pulsatile-flow generation method in vitro."</span>                                                                                            </div><div class="line"> [<span class="number">6</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line"> [<span class="number">7</span>] <span class="string">"Stacking as a Key Property for Creating Nanoparticles with Tunable Shape: The Case of Squalenoyl-Doxorubicin."</span>                                                            </div><div class="line"> [<span class="number">8</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line"> [<span class="number">9</span>] <span class="string">"[Study on absorbed components of Aconitum kusnezoffii under Yunnan Baiyao compatibility in effect of activating blood circulation and removing blood stasis]."</span>            </div><div class="line">[<span class="number">10</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">11</span>] <span class="string">"[Application of quality constant method in grade evaluation of Aurantii Fructus Immaturus pieces]."</span>                                                                       </div><div class="line">[<span class="number">12</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">13</span>] <span class="string">"[Study on regularity of Tibetan medicine in treatment of gZav-Grib disease (apoplexy sequelae) based on HIS clinical medical records]."</span>                                   </div><div class="line">[<span class="number">14</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">15</span>] <span class="string">"Exercise responses in children and adults with a Fontan circulation at simulated altitude."</span>                                                                               </div><div class="line">[<span class="number">16</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">17</span>] <span class="string">"The Course of Immune Stimulation by Photodynamic Therapy: Bridging fundamentals of photochemically-induced Immunogenic Cell Death to the Enrichment of T Cell Repertoire."</span></div><div class="line">[<span class="number">18</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">19</span>] <span class="string">"Design of Dense Brush Conformation Bearing Gold Nanoparticles as Theranostic Agent for Cancer."</span>                                                                           </div><div class="line">[<span class="number">20</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">21</span>] <span class="string">"Liquid biopsy: one cell at a time."</span>                                                                                                                                       </div><div class="line">[<span class="number">22</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">23</span>] <span class="string">"The Evolving Role of Neutrophils in Liver Transplant Ischemia-Reperfusion Injury."</span>                                                                                        </div><div class="line">[<span class="number">24</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">25</span>] <span class="string">"Predictors of Balloon Guide Catheter Assistance Success in Stent-retrieval Thrombectomy for an Anterior Circulation Acute Ischemic Stroke."</span>                               </div><div class="line">[<span class="number">26</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">27</span>] <span class="string">"High mobility group box protein 1 neutralization therapy in ovine bacteremia: Lessons learned from an ovine septic shock model incorporating intensive care support."</span>     </div><div class="line">[<span class="number">28</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">29</span>] <span class="string">"Low human and murine Mcl-1 expression leads to a pro-apoptotic plaque phenotype enriched in giant-cells."</span>                                                                 </div><div class="line">[<span class="number">30</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">31</span>] <span class="string">"Upper-tropospheric bridging of wintertime surface climate variability in the Euro-Atlantic region and northern Asia."</span>                                                     </div><div class="line">[<span class="number">32</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">33</span>] <span class="string">"Discordances between predicted and actual risk in obese patients with suspected cardiac ischaemia."</span>                                                                       </div><div class="line">[<span class="number">34</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">35</span>] <span class="string">"Tumor-Associated Release of Prostatic Cells into the Blood after Transrectal Ultrasound-Guided Biopsy in Patients with Histologically Confirmed Prostate Cancer."</span>         </div><div class="line">[<span class="number">36</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">37</span>] <span class="string">"Circulating tumor cells exhibit metastatic tropism and reveal brain metastasis drivers."</span>                                                                                  </div><div class="line">[<span class="number">38</span>] <span class="string">"Similar articles"</span>                                                                                                                                                         </div><div class="line">[<span class="number">39</span>] <span class="string">"Permanent pacing post-Fontan is not associated with reduced long-term survival."</span>                                                                                          </div><div class="line">[<span class="number">40</span>] <span class="string">"Similar articles"</span></div></pre></td></tr></table></figure>
<p>一共有40个结果，里面有相同的字符串，即<code>&quot;Similar articles&quot;</code>，这个不是我们需要的，这因为定位出现了问题，也就是说<code>node = '//*[@id=&quot;maincontent&quot;]/div/div[5]//div[2]/p/a'</code>这段代码有问题，现在我们查看原题目与Similar articles的元素，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012201418.png">

</div>
<p>其中，红框是我们要爬取的题目，而蓝框则similar articles的内容，因此我们需要把蓝框的内容给剔掉，只爬取到<code>class=&quot;title&quot;</code>这个字段就行，也就是说添加上<code>p[@class=&quot;title&quot;]</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//*[@id="maincontent"]/div/div[5]//div[2]/p[@class="title"]/a'</span></div><div class="line"><span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line"></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"></div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_text(x = nodes_content, trim = <span class="literal">T</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment">### 2.1 Location</span></div><div class="line">&gt; node = <span class="string">'//*[@id="maincontent"]/div/div[5]//div[2]/p[@class="title"]/a'</span></div><div class="line">&gt; <span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line">&gt; </div><div class="line">&gt; nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment">### 2.2 Crawl data</span></div><div class="line">&gt; html_text(x = nodes_content, trim = <span class="literal">T</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"Variation of organ position in snakes."</span>                                                                                                                                   </div><div class="line"> [<span class="number">2</span>] <span class="string">"Pathogenesis of Cluster Headache: From Episodic to Chronic Form, the Role of Neurotransmitters and Neuromodulators."</span>                                                      </div><div class="line"> [<span class="number">3</span>] <span class="string">"Haemodynamic evaluation of the new pulsatile-flow generation method in vitro."</span>                                                                                            </div><div class="line"> [<span class="number">4</span>] <span class="string">"Stacking as a Key Property for Creating Nanoparticles with Tunable Shape: The Case of Squalenoyl-Doxorubicin."</span>                                                            </div><div class="line"> [<span class="number">5</span>] <span class="string">"[Study on absorbed components of Aconitum kusnezoffii under Yunnan Baiyao compatibility in effect of activating blood circulation and removing blood stasis]."</span>            </div><div class="line"> [<span class="number">6</span>] <span class="string">"[Application of quality constant method in grade evaluation of Aurantii Fructus Immaturus pieces]."</span>                                                                       </div><div class="line"> [<span class="number">7</span>] <span class="string">"[Study on regularity of Tibetan medicine in treatment of gZav-Grib disease (apoplexy sequelae) based on HIS clinical medical records]."</span>                                   </div><div class="line"> [<span class="number">8</span>] <span class="string">"Exercise responses in children and adults with a Fontan circulation at simulated altitude."</span>                                                                               </div><div class="line"> [<span class="number">9</span>] <span class="string">"The Course of Immune Stimulation by Photodynamic Therapy: Bridging fundamentals of photochemically-induced Immunogenic Cell Death to the Enrichment of T Cell Repertoire."</span></div><div class="line">[<span class="number">10</span>] <span class="string">"Design of Dense Brush Conformation Bearing Gold Nanoparticles as Theranostic Agent for Cancer."</span>                                                                           </div><div class="line">[<span class="number">11</span>] <span class="string">"Liquid biopsy: one cell at a time."</span>                                                                                                                                       </div><div class="line">[<span class="number">12</span>] <span class="string">"The Evolving Role of Neutrophils in Liver Transplant Ischemia-Reperfusion Injury."</span>                                                                                        </div><div class="line">[<span class="number">13</span>] <span class="string">"Predictors of Balloon Guide Catheter Assistance Success in Stent-retrieval Thrombectomy for an Anterior Circulation Acute Ischemic Stroke."</span>                               </div><div class="line">[<span class="number">14</span>] <span class="string">"High mobility group box protein 1 neutralization therapy in ovine bacteremia: Lessons learned from an ovine septic shock model incorporating intensive care support."</span>     </div><div class="line">[<span class="number">15</span>] <span class="string">"Low human and murine Mcl-1 expression leads to a pro-apoptotic plaque phenotype enriched in giant-cells."</span>                                                                 </div><div class="line">[<span class="number">16</span>] <span class="string">"Upper-tropospheric bridging of wintertime surface climate variability in the Euro-Atlantic region and northern Asia."</span>                                                     </div><div class="line">[<span class="number">17</span>] <span class="string">"Discordances between predicted and actual risk in obese patients with suspected cardiac ischaemia."</span>                                                                       </div><div class="line">[<span class="number">18</span>] <span class="string">"Tumor-Associated Release of Prostatic Cells into the Blood after Transrectal Ultrasound-Guided Biopsy in Patients with Histologically Confirmed Prostate Cancer."</span>         </div><div class="line">[<span class="number">19</span>] <span class="string">"Circulating tumor cells exhibit metastatic tropism and reveal brain metastasis drivers."</span>                                                                                  </div><div class="line">[<span class="number">20</span>] <span class="string">"Permanent pacing post-Fontan is not associated with reduced long-term survival."</span></div></pre></td></tr></table></figure>
<p>此时，如果想读取链接，则如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//*[@id="maincontent"]/div/div[5]//div[2]/p/a'</span></div><div class="line"><span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line"></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"></div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">&gt; html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"/pubmed/31603578"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31603578"</span></div><div class="line"> [<span class="number">3</span>] <span class="string">"/pubmed/31603552"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31603552"</span></div><div class="line"> [<span class="number">5</span>] <span class="string">"/pubmed/31603372"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31603372"</span></div><div class="line"> [<span class="number">7</span>] <span class="string">"/pubmed/31603305"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31603305"</span></div><div class="line"> [<span class="number">9</span>] <span class="string">"/pubmed/31602894"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602894"</span></div><div class="line">[<span class="number">11</span>] <span class="string">"/pubmed/31602882"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602882"</span></div><div class="line">[<span class="number">13</span>] <span class="string">"/pubmed/31602864"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602864"</span></div><div class="line">[<span class="number">15</span>] <span class="string">"/pubmed/31602790"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602790"</span></div><div class="line">[<span class="number">17</span>] <span class="string">"/pubmed/31602649"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602649"</span></div><div class="line">[<span class="number">19</span>] <span class="string">"/pubmed/31602558"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602558"</span></div><div class="line">[<span class="number">21</span>] <span class="string">"/pubmed/31602399"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602399"</span></div><div class="line">[<span class="number">23</span>] <span class="string">"/pubmed/31602356"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602356"</span></div><div class="line">[<span class="number">25</span>] <span class="string">"/pubmed/31602354"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602354"</span></div><div class="line">[<span class="number">27</span>] <span class="string">"/pubmed/31602200"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31602200"</span></div><div class="line">[<span class="number">29</span>] <span class="string">"/pubmed/31601924"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601924"</span></div><div class="line">[<span class="number">31</span>] <span class="string">"/pubmed/31601851"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601851"</span></div><div class="line">[<span class="number">33</span>] <span class="string">"/pubmed/31601728"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601728"</span></div><div class="line">[<span class="number">35</span>] <span class="string">"/pubmed/31601564"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601564"</span></div><div class="line">[<span class="number">37</span>] <span class="string">"/pubmed/31601552"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601552"</span></div><div class="line">[<span class="number">39</span>] <span class="string">"/pubmed/31601284"</span>                                 <span class="string">"/pubmed?linkname=pubmed_pubmed&amp;from_uid=31601284"</span></div></pre></td></tr></table></figure>
<p>此时读取了40项内容，又多了一倍，还是跟前面的问题一样，此时需要过滤一下（核心内容就是正则表达式），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//div[@class="rprt"]/div[@class="rslt"]/p[@class="title"]/a'</span></div><div class="line"></div><div class="line"></div><div class="line"><span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line"></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"></div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"/pubmed/31603578"</span> <span class="string">"/pubmed/31603552"</span> <span class="string">"/pubmed/31603372"</span> <span class="string">"/pubmed/31603305"</span> <span class="string">"/pubmed/31602894"</span> <span class="string">"/pubmed/31602882"</span></div><div class="line"> [<span class="number">7</span>] <span class="string">"/pubmed/31602864"</span> <span class="string">"/pubmed/31602790"</span> <span class="string">"/pubmed/31602649"</span> <span class="string">"/pubmed/31602558"</span> <span class="string">"/pubmed/31602399"</span> <span class="string">"/pubmed/31602356"</span></div><div class="line">[<span class="number">13</span>] <span class="string">"/pubmed/31602354"</span> <span class="string">"/pubmed/31602200"</span> <span class="string">"/pubmed/31601924"</span> <span class="string">"/pubmed/31601851"</span> <span class="string">"/pubmed/31601728"</span> <span class="string">"/pubmed/31601564"</span></div><div class="line">[<span class="number">19</span>] <span class="string">"/pubmed/31601552"</span> <span class="string">"/pubmed/31601284"</span></div></pre></td></tr></table></figure>
<p>其实我们可以发现，<code>node = '//div[@class=&quot;rprt&quot;]/div[@class=&quot;rslt&quot;]/p[@class=&quot;title&quot;]/a'</code>这一句中的最后一部分<code>node = '//p[@class=&quot;title&quot;]/a'</code>其实是唯一标记的，也就是说在这个路径中没有重复，因此我们还可以改一下代码，把这个字符前面的都删掉，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="comment">### 2.1 Location</span></div><div class="line">node = <span class="string">'//p[@class="title"]/a'</span></div><div class="line"><span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line">nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line"><span class="comment">### 2.2 Crawl data</span></div><div class="line">html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div></pre></td></tr></table></figure>
<p>运行一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment">### 2.1 Location</span></div><div class="line">&gt; node = <span class="string">'//p[@class="title"]/a'</span></div><div class="line">&gt; <span class="comment"># 这里要注意//与/的区别，/表示绝对路径，//表示相对路径</span></div><div class="line">&gt; nodes_content &lt;- html_nodes(x = page_content, xpath = node)</div><div class="line">&gt; <span class="comment">### 2.2 Crawl data</span></div><div class="line">&gt; html_attr(x = nodes_content, name = <span class="string">"href"</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"/pubmed/31603578"</span> <span class="string">"/pubmed/31603552"</span> <span class="string">"/pubmed/31603372"</span> <span class="string">"/pubmed/31603305"</span> <span class="string">"/pubmed/31602894"</span> <span class="string">"/pubmed/31602882"</span></div><div class="line"> [<span class="number">7</span>] <span class="string">"/pubmed/31602864"</span> <span class="string">"/pubmed/31602790"</span> <span class="string">"/pubmed/31602649"</span> <span class="string">"/pubmed/31602558"</span> <span class="string">"/pubmed/31602399"</span> <span class="string">"/pubmed/31602356"</span></div><div class="line">[<span class="number">13</span>] <span class="string">"/pubmed/31602354"</span> <span class="string">"/pubmed/31602200"</span> <span class="string">"/pubmed/31601924"</span> <span class="string">"/pubmed/31601851"</span> <span class="string">"/pubmed/31601728"</span> <span class="string">"/pubmed/31601564"</span></div><div class="line">[<span class="number">19</span>] <span class="string">"/pubmed/31601552"</span> <span class="string">"/pubmed/31601284"</span></div></pre></td></tr></table></figure>
<p>结果是一样的，这里要学习的就是唯一标记符，使用这种方法非常高效（核心就是找到唯一的节点）。</p>
<h3 id="简化操作之管道">简化操作之管道</h3>
<p>上面介绍的这个爬虫过程，都需要找到网址，输入节点，比较麻烦，因此可以采取管道（<code>%&gt;%</code>）来简化操作，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## pipeline operation</span></div><div class="line">page_content &lt;- read_html(x = <span class="string">'https://www.ncbi.nlm.nih.gov/pubmed/?term=circulation'</span>)</div><div class="line"></div><div class="line"><span class="comment"># title</span></div><div class="line">node &lt;-  <span class="string">'//p[@class="title"]/a'</span></div><div class="line">page_content %&gt;%</div><div class="line">  html_nodes(xpath = node) %&gt;%</div><div class="line">  html_text(trim = <span class="literal">T</span>)</div><div class="line"></div><div class="line"><span class="comment"># Link</span></div><div class="line">page_content %&gt;%</div><div class="line">  html_nodes(xpath = node) %&gt;%</div><div class="line">  html_attr(name = <span class="string">'href'</span>)</div></pre></td></tr></table></figure>
<h3 id="简化操作之函数">简化操作之函数</h3>
<p>上面的操作还能继续简化，也就是写成一个函数，R中构建函数的内容可以找本书看看，这里直接放代码，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## Contruction function for web site</span></div><div class="line">read_html.my &lt;- <span class="keyword">function</span>(url)&#123;</div><div class="line">  <span class="keyword">library</span>(rvest)</div><div class="line">  page_concent &lt;&lt;- read_html(x = url)</div><div class="line">  <span class="comment"># note symbol "&lt;&lt;-" is transform local variable into global variable</span></div><div class="line">&#125;</div><div class="line"></div><div class="line">read_html.my(<span class="string">'https://www.ncbi.nlm.nih.gov/pubmed/?term=circulation'</span>)</div><div class="line"></div><div class="line"><span class="comment"># Read text from web site</span></div><div class="line">html_text.my &lt;- <span class="keyword">function</span>(xpath)&#123;</div><div class="line">  page_concent %&gt;%</div><div class="line">    html_nodes(xpath = xpath) %&gt;%</div><div class="line">    html_text(trim = <span class="literal">T</span>)</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="comment"># Read link from web site</span></div><div class="line">html_href.my &lt;- <span class="keyword">function</span>(xpath)&#123;</div><div class="line">  page_concent %&gt;%</div><div class="line">    html_nodes(xpath = xpath) %&gt;%</div><div class="line">    html_attr(name = <span class="string">'href'</span>)</div><div class="line">&#125;</div><div class="line"></div><div class="line">html_text.my(<span class="string">'//p[@class="title"]/a'</span>)</div><div class="line">html_href.my(<span class="string">'//p[@class="title"]/a'</span>)</div></pre></td></tr></table></figure>
<h2 id="总结">总结</h2>
<p>涉及到的知识点大概如下所示：</p>
<ol style="list-style-type: decimal">
<li>网页的构成（xpath，html，css，绝对路径(<code>/</code>)与相对路径(<code>//</code>，节点，内容）；</li>
<li>正则表达式；</li>
<li>R中函数的构建（局部变量，变局变量，局部变量切换为全局变量<code>&lt;&lt;-</code>）；</li>
<li>管道操作(<code>%&gt;%</code>)。</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/10/01/BioTools/NormqPCR笔记对应文献之一/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/10/01/BioTools/NormqPCR笔记对应文献之一/" itemprop="url">NormqPCR笔记对应文献之一</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-10-01T12:00:00+08:00">
                2019-10-01
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,653
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  20
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>最近在处理qPCR数据过程中，设计的qPCR阵列中采用了多个内参，我对于多内参的处理不太理解。后来找到了一篇文献，就是讲qPCR中的多内参问题的。文献信息如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">Vandesompele, J., et al. (2002). &quot;Accurate normalization of real-time quantitative RT-PCR data by geometric averaging of multiple internal control genes.&quot; Genome Biology 3(7): research0034.0031.</div></pre></td></tr></table></figure>
<p>此篇文献引用量很高（到2019年10月为止，引用量约为16000），虽然文献比较古老，但是了解一下文献中的思想还是比较有用的，以下是文献的阅读笔记（基本上就是翻译原文外加自己检索）。</p>
<p>文献中涉及一定的线性代数知识，我不太懂，只能先暂时先记录下来，等看懂的时候再补充一篇笔记。</p>
<h2 id="研究背景">研究背景</h2>
<p>​ 在生物学研究中，基因表达的分析有着重要的作用。对于基因表达模式的理解可以为复杂的调控网络研究提供思路。最近开发的两种检测转录本丰度的方法得到了非常广泛的应用（该文章是2002年写的，这里的最近已经是非常久远了）。微阵列的方法（微阵列其实就是芯片法）可以同时检测数以千计的基因表达情况，而RT-PCR的方法则是用于检测许多不同的样本中数量有限基因的表达情况（在细胞数量比较少的情况下用这种方法更好，说服力更强）。与传统的方法例如northern-blot，核糖核酸酶保护，或者是竞争RT-PCR（competitive rt-PCR）相比，这两种方法（微阵列与RT-qPCR）都有快速，高通量的优点。但是这两种方法都需要与传统方法一样，需要对目的mRNA进行均一化（normalization）。</p>
<p>以下是补充一些与核酸保护实验相关的背景知识。</p>
<h3 id="核酸酶保护实验">核酸酶保护实验</h3>
<p>​ 糖核酸酶保护实验（Ribonuclease protection assay，RPA）是一种mRNA定量分析方法。其基本原理是将标记的特异RNA探针（32P或生物素）与待测的RNA样品液相杂交，标记的特异RNA探针按碱基互补的原则与目的基因特异性结合，形成双链RNA；未结合的单链RNA经RNA酶A或RNA酶T1消化形成寡核糖核酸，而待测目的基因与特异RNA探针结合后形成双链RNA，免受RNA酶的消化，故该方法命名为RNA酶保护实验。</p>
<p>​ 对于32P标记的探针，杂交双链进行变性聚丙酰胺凝胶电泳，用放射自显影或磷屏成像系统检测被保护的探针的信号; 对于生物素标记的探针，杂交双链经过变性聚丙酰胺凝胶电泳后电转移至尼龙膜，采用链霉亲和素－辣根过氧化物酶（Streptavidin-HRP）和化学发光底物与膜上biotin标记的探针结合，X射线胶片或化学发光图像分析仪检测杂交信号。</p>
<p>​ 在利用RT-PCR对基因进行定量时，为了缩小误差，我们需要对一些因素进行控制，例如原始样品的数量（细胞或者是组织），酶的效率，组织或细胞的整个转录活动差异等因素。在控制条件的理想情况下，提取的高质量RNA的转录本数量的可重复结果与细胞数量是呈正比的，但是精确的细胞数目很难进行定量（尤其是当原始材料是组织时更是如此）。</p>
<p>​ 另外一种常用的定量手段就是RNA的量（这里指的是数量，原文写的是mass quantity），尤其是northern blot分析方法。但是这种方法也存在着一些争议。因为这种方法没有考虑到RNA的质量（这里侧重的是RNA的完整性，原文写的是quality）与相关酶的反应效率。此外，在某些情况下，这些因素也无法定量，例如从提取的微切割的组织（microdissected tissue）中得到的RNA数量极少，这个时候就无法对RNA进行定量。</p>
<p>​ 其中最大的争议还在于，用于均一化时所考虑的总RNA量，因为总的RNA主要是由rRNA组成的，它并不代表mRNA。这里大概介绍一下使用总RNA量来作为RNA质控的方法。</p>
<p>​ 当我们提取了RNA后，在进行质控时，通常是使用1%的琼脂糖凝胶电泳来跑200ng到400ng的总RNA，然后观察RNA的条带。良好的RNA（这里指的是动物组织或细胞的样本）中可以看到3个条带，从上到下依次是28S、18S、5S（这是核糖体RNA的主要成分），其中28S条带与18S条带非常清楚，5S条件非常弱，28S和18S比值可以判断RNA样本的完整性，通常这个比值是1.5（肉眼不太好判断，通常来说就是28S比18S亮就行了），如果这两个条带变弱，5S条件变亮，则说明RNA降解，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012111600.png">

</div>
<p>​ 不过电泳法是比较粗糙的方法，毕竟这篇文献是2002年的，现在如果要进行测序建库，要求就比较高，会采用Qubit检测RNA的浓度，用安捷伦2100来检测RNA的完整性，具体的检测指标可以检索相关资料。</p>
<p>但是在总的RNA中，mRNA的含量仅占7.5%。此外，还有报告指出，rRNA的转录还受到生物因素与药物的影响。利用18S或28S作为RNA定量的标准缺陷还在于，在纯的mRNA分子中，并不含有18S或28S，与靶mRNA转录本相比，18S或28S的含量太高。在利用RT-PCR进行分析时，很难排除基线(baseline)。</p>
<p>现在补充一些竞争性PCR的背景知识。</p>
<h3 id="竞争性pcr">竞争性PCR</h3>
<p>​ 在竞争性PCR中，逆转录之前加入一个外源的RNA转录本（内部标准RNA）作为样品间差异的对照。内部标准RNA被逆转录并同目的模板一起扩增，作为cDNA转化和扩增效率差异的对照。通过将特异性的目的序列同已知浓度的内部标准RNA一起扩增进行定量。通过比较由内部标准获得的信号和目的模板所获得的信号可以确定目的模板的丰度。</p>
<p>​ 到目前为止（也是2018年为止），内参基因是最常用的对mRNA进行均一化的方法。内参基因（internal control）通常指的是管家基因（housekeeping gene），这类基因并不会随着组织或细胞的不同而变化， 或者是不会随着实验条件而变化。但是在许多的研究里，研究者使用的这些持续表达的内参基因中，并没有对这些基因的稳定表达进行验证。文献中使用的这些管家基因偶尔会出现变化极大的现象。随着RT-PCR灵敏度，可靠性以及使用范围的增加，对于合适内参基因的需求也日益迫切。些外，通过利用利用公共微阵列数量，这种均一化因素（normalization factor）也在广泛的应用的微阵列定标系数（scaling factor）中得到了验证。</p>
<h2 id="需要了解的几个公式">需要了解的几个公式</h2>
<p>文献末尾处提到了几个公式，现在汇总如下所示：</p>
<h3 id="公式1单一内参均一化错误值single-control-normalization-error">公式1:单一内参均一化错误值（Single control normalization error）</h3>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924164713.png">

</div>
<p>公式上面的英文注释如下所示：</p>
<p>E值： 对于任何m个样本，用real-time RT-PCR技术可以测量其n个内参基因的表达水平，其表达水平为aij。对于每2个样本（分别记为p和q）的组合来说，它们的两个内参基因j和k的每一种组合来说，可以计算出它们的单一内参均一化错误值E（single control normalization error，公式1）。</p>
<p>当样本p和q分别针对其内参基因j或k进行均一化的时候，其样本p和样本q的表达总倍数差异就是错误值E。 n表示内参的数目（注：既然是多内参，这个n必然是大于等于2）。</p>
<p>aij表示的是内参的表达水平（这是一组数据）。</p>
<p>j与k的范围为1到n，p与q的值为1到m（都是闭区间），并且j不等于k，p不等于q。<strong>Rjkpq的倒数就是E值</strong>，公式1中，aqj是说在样本q中，以内参基因j为标准进行均一化后的数据，aqk则是说，以内参基因k为标准进行均一化后的数据，aqj/aqk则是这二者之比，apk/apj为例，aqj/aqk与apk/apj的比值则是指p与q样本中，a基因表达水平的误差。</p>
<p>个人理解：这个E值其实就是每个样本，每个内参均一化后误差的乘积。</p>
<h3 id="公式2内参基因稳定检测值m">公式2:内参基因稳定检测值M</h3>
<p>原文如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924172122.png">

</div>
<p>公式解释：</p>
<p>M值：对于任意两个内参基因j与k，可以计算出由它们构成的一个Ajk数列（array），这个数列包含一组经log2转化来的比值（aij/aik），如果是m个样本，那么Ajk中一共是m个值（公式2）。作者定义Vjk是内参基因j和k的<strong>成对变异（pairwise variation）</strong>它是Ajk的标准差（公式3）。其中内参基因j的稳定表达值Mj就是所有成对基因变异Vjk的算术平均数（公式4）。</p>
<p>画个图，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924174227.png">

</div>
<p>以上计算的<strong>只是内参基因j的稳定值</strong>。</p>
<h2 id="研究思路">研究思路</h2>
<p>提取了不同组织或细胞的RNA进行检测，然后检测这些RNA中的几个选定的内参，再通过一定的算法进行比较。</p>
<h2 id="实验过程">实验过程</h2>
<h3 id="一管家基因表达谱">（一）管家基因表达谱</h3>
<p>​ 本文研究了10个常用管家基因在人类13个不同组织中的表达，这10个管家基因如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924162706.png">

</div>
<p>管家基因列表。</p>
<p>注：选这个10个管家基因的思路是要尽量使这10个基因属于不同的功能分类，这样就能极大地了降低它们之间存在的共调控现象。</p>
<p>作者检测了80个组织中的这10个管家基因的表达情况，这80个组织分别为：在34个神经母细胞瘤（来源不同的实验室以及不同的病人，NB1-34）；20个短期培养的常规成纤维细胞（FIB1-20）；13个常规的淋巴细胞样本（LEU1-13）；9个骨髓样本（BM1-9）；9个其他的人类正常组织（分别为heart，brain，fetal brain，lung，trachea，kidney，mammary gland，small intestine和uterus）。</p>
<h3 id="二单一内参均一化错误值single-control-normalization-error">（二）单一内参均一化错误值（Single control normalization error）</h3>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924162802.png">

</div>
<p>Figure 1 -对于两个理想的内参基因来说，E值等于1，这个很好理解，我们还看一下前面的E公式，如下所示： <span class="math display">\[
{R_{j k p q}=\frac{a_{q j}}{a_{q k}} / \frac{a_{p j}}{a_{p k}} \quad\left(\text { if } R&lt;1, \text { then } E=R^{-1}, \text { else } E=R\right)}
\]</span> 也就是说，理想的两个内参基因表达水平必然是相同的，因此E等于1。</p>
<p>事实上观察到的E值通常大于1，这就构成了两个样本之间的E倍差异，具体的差异值取决用于均一化的特定管家基因。对所有的10个内参基因两两配对（一共45种组合），以及所有样本的两两配对（一共是865个组合）绘制E值，如figure 1所示。此外，通过分析重复运行的相同内参基因绘制出了系统错误的分位数。E值的75百分位数和90百分位分别是3.0（范围是2.1-3.9）和6.4（范围是3.0-10.9）。</p>
<p>注：所有样本的两两配对是指，同一个组织内的两两配对，因此根据前面的内容：</p>
<blockquote>
<p>34个神经母细胞瘤（来源不同的实验室以及不同的病人，NB1-34）；20个短期培养的常规成纤维细胞（FIB1-20）；13个常规的淋巴细胞样本（LEU1-13）；9个骨髓样本（BM1-9）；9个其他的人类正常组织（分别为heart，brain，fetal brain，lung，trachea，kidney，mammary gland，small intestine和uterus）</p>
</blockquote>
<p>排列组合数目就是865。</p>
<h3 id="三基因表达稳定性检测以及所选基因的秩">（三）基因表达稳定性检测以及所选基因的秩</h3>
<p>​ 在研究中，通常会认为基因的表达水平应该精选一个稳定表达的基因进行均一化。但是为了验证选定的这个内参基因的稳定表达，我们需要一种可靠的检测手段手段来确定这个内参基因确实是稳定表达的，从而剔除掉非特异性的变异。为了解决这个循环问题（circular problem，这里的循环问题就是说，我要选择稳定的内参，就需要“内参”进行均一化，但是选择均一化的这个“内参”又得是稳定的，因此是一个不断循环的问题），作者以非均一化表达水平为基础，提出了一种基因稳定性检测算法。</p>
<p>​ 这种算法的原理是，无论实验条件，无论细胞类型，两种理想的内参基因在所有样本中的表达必然是相同的。而在实际条件下，两种真实内参基因的表达比值（ratio）的变异就会反映出一种（或者是两种）内参基因的表达并非是恒定不变的，这种比值的差异与基因表达的稳定程度呈反比。对于每一个内参基因来说，作者都研究了它与其他所有内参基因两两的差异，这种差异用表达比值的对数转换的标准差来表示，并且定义了一个内参基因表达稳定检测值M作为一个特定基因与其他所有内参基因的平均两两变异。</p>
<p>​ 其中，最小M值的基因是表达最稳定的内参基因。假如选定的这些内参基因不存在共调控，那么分步排除掉的最高M值的基因后，最终剩下的内参基因就会产生一对组合，构成这种组合的两个基因是稳定表达的管家基因，即它们是样品中最稳定的内参基因。</p>
<p>​ 为了计算最佳的内参基因组合，作者利用VBA开发了一个Excel插件，命名为geNorm，它能自动计算所有内参基因的的M值（现在这个插件已经整合到了qbase+中，网址为：https://www.qbaseplus.com/）。这个程序可以剔除那些不好的内参基因（即最高M值的基因），并且重新计算剩余内参基因的M值。利用geNorm，作者计算了5类组织中的10个内参基因的M值，并按从大到小的顺序排列（参见figure2与table 3）。此外，作者还计算了系统变异，其方法对相同的基因进行重复的RT-PCR实验，这种变异反映了仪器自身，加样，酶的固有变异。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924162839.png">

</div>
<p>Figure 2 横坐标表示的是剩余的内参数目。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190924181905.png">

</div>
<p>注：qbase+这个软件我并不熟悉，这是没有重现数据的分析过程。但是在前一篇笔记中，我使用了NormqPCR这个R包来重现了数据分析过程。</p>
<h3 id="四基于多内参基因的几何均数进行均一化因子计算">（四）基于多内参基因的几何均数进行均一化因子计算</h3>
<p>这一部分就是说明要选择几个内参来进行均一化。</p>
<p>​ 为了精确地评估基因的表达水平，作者推荐使用多个内参基因进行均一化。将多个内参取几何均数。至于使用的内参基因数目，需要在实际情况与精确性方面进行取舍。但是就实际情况来看，选择太多的内参基因浪费（例如10个太浪费了），如果太少，不为精确，作者推荐最少使用3个表达稳定的内参基因作为均一化因子（normalizaiton factor, NFn, n=3）。</p>
<p>​ 作者还研究了一些情况，即那些利用了超过3种内参基因用于均一化的情况，作者计算了相同组织类型内的所有样本，按照两次均一化因子（NFn和NFn+1）的两两变异Vn/n+1（即aij=NFn,i和aik=NFn+1,n代表用于均一化的基因数目，范围是3到9，i表示样本，具体参照公式2和3）。我的理解是，在选定了几个内参，例如5个，对所有的样本进行均一化，计算其变异，即V5，然后再排除一个内参，剩了4个内参，再对所有的样本均一化，计算其变异，即V4，最后计算V4/V5）。这种算法研究的是，如果存在着一个很大的变异，就意味着，这个加入的内参基因（我的理解是，排除的那个内参）有着很强的效应，它应该优先选择用于均一化因子的计算（不太理解这种算法）。</p>
<p>​ 对于所有的组织类型，用于计算的均一化因子是3个最稳定的内参基因，通过逐步添加剩余的7个内参基因。逐步计算两两配对变异，对于连续的NFn和NFn+1均一化因子，这就反应了加入的第n+1基因的效应。 （如figure 3a所示）。从图中可以明显看出来，对于leukicytes, fibroblast和bone marrow来说，添加的第4个基因没有特别明显的影响（就是说低V3/4值），这也说明，在NF3和NF4值之间存在着很强的关联，如figure 3b所示。以这些数据为基础，作者采用了0.15作为阈值，低于0.15的内参基因不需要添加进来。对于neuroblastoma和正常组织来说，分别需要加入1个和2个内参基因（参见figure 3b）。最高的V8/9和V9/10值对于normal pool，neuroblastoma和leukocyte来说就确认了figure 2中逐步排除的最差内参基因。这种分析表明了平均M值的在开始时的强烈下降，这就指出了，leukocyte的两个内参基因的表达异常，neuroblastoma和normal tissues一个内参基因的不稳定。此外，对于后两种组织来说，需要包含另外的内参基因，从而与内参基因的高度变异保持一致。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012105801.jpg">

</div>
<h3 id="验证可能的rt-pcr均一化因子">验证可能的RT-PCR均一化因子</h3>
<p>​ 为了验证这种基因检测算法的可靠性，即最低M值是最稳定的表达的内参基因，作者检测了每个内参基因的特异性变化，作为均一化之后表达的变异系数（variation coefficient）。对于合适的内参基因来说，这个系数应该是最小的。作者以三个基因的几何平均数为基础，分别计算了三个不同的均一化因子，最低的M值（NF3（1-3）），最高的M值（NF3（8-10）和中等的M值（NF3（6-8））。接着，作者计算用最低变异系数，在每种组织类型内，计算了这三个基因的平均基因特异性变异（figure 4a）。从计算结果可以看出，当数据用NF3（1-3）进行均一化时，这些基因特异性变异最小。</p>
<p>​ 这就表明，基因稳定性检测能够确定地确定表达最稳定的内参基因。为了确认高M值是一类表达不稳定，或者是表达有差异的基因，作者分析了MYCN的表达水平，MYCN这个基因是一种在neuroblastoma中表达有差异的基因。在neuroblastoma中，MYN的M值是6.02，而B2M（表达最稳定的内参基因）的M值是2.17。进一步的观察发现，用单一的内参基因进行均一化会导致其它内参基因更高的的基因特异性变异，这就进一步说明了利用多个管家基因进行均一化的优势。</p>
<p>​ 为了说明最佳的个内参基因不受细胞增殖的影响，作者分析了PCNA的表达水平（PCNA是增殖的标志），研究了四个最佳管家基因和标志基因PCNA之间的Spearman秩相关系数。从分析结果来看，管家基因之间存在着关联（p值小于0.001，相关系数在0.6到0.76之间）。相比之下，PCNA和三个管家基因之间不存在联系，而与HPRT1则存在着弱相关（p值为0.024，相关系数为0.43）。这些数据清楚地说明了，最稳定的管家基因与细胞的增殖状态无关。</p>
<p>​ 为了进一步研究选定的内参基因几何均数的精确性，作者研究了公共数据库中的微阵列数据，利用这些数据中的内参基因的几何均数进行了均一化，具体的选择过程参见文献，作者选定了5个内参基因，其标准是M值小于0.7，其计算结果与微阵列的计算结果类似。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012105826.png">

</div>
<h3 id="组织特异性管家基因的表达">组织特异性管家基因的表达</h3>
<p>​ 为了比较13所测组织中的管家基因表达水平的异质性。作者剔除了差异最大的4个管家基因（这4个管家基因分别为B2M，RPL13A，ACTB与HMBS，参见Table 2）后，计算了6个内参基因的几何均数作为“伪内参”，然后将这10个内参基因再根据这个“伪内参”进行均一化。Figure 5显示了所有的10个基因的不同丰度类型，表达最高的ACTB基因是最低的HMBS的400倍。虽然内参基因的丰度在不同的组织中类似，但还是观察到了组织特异性表达的差异，例如，B2M的表达水平在leukocytes中比Fetal Brain中高出112倍，而Fibroiblast和Heart tisssue中，ACTB的差异有22倍。与这两个基因相比，有2个基因的表达比较稳定，例如UBC和HPRT1。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191012105923.jpg">

</div>
<p>讨论部分略。</p>
<p>参考资料：</p>
<ol style="list-style-type: decimal">
<li><a href="https://www.biomart.cn/experiment/430/457/458/2284875.htm" target="_blank" rel="external">PCR和RT-PCR基础（强烈推荐）</a></li>
<li><a href="http://www.bioon.com/experiment/mob/275439.shtml" target="_blank" rel="external">核糖核酸酶保护实验的基本原理</a></li>
<li><a href="http://www.sohu.com/a/197299181_307557" target="_blank" rel="external">大家好，给大家介绍一下，这是RNA质检结果解读指南</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/30/BioTools/NormqPCR包学习笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/30/BioTools/NormqPCR包学习笔记/" itemprop="url">NormqPCR包笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-30T12:00:00+08:00">
                2019-09-30
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  8,176
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  43
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p><code>NormqPcR</code>这个包用于对qPCR数据进行均一化(normalization，这里需要注意的是，这里的normalization只是使用内参来校正目标基因的表达的意思，与统计学中的归一化(Normalization)，中心化(center)，标准化(scales)之类的术语没有关系)。这篇笔记主要是翻译了这个包的英文文档，本文档以一个小型案例(vignette)来演示了这个包的主要功能，笔记主要内容如下所示：</p>
<ul>
<li>合并技术重复(technical replicates)（了解即可）；</li>
<li>处理低Cq值(undetermined values)（了解即可）；</li>
<li>计算用于均一化的最佳管家基因(housekeeping genes)（这个最重要，其余的很多函数其实了解即可，这一部分要了解geNorm和NormFinder的算法原理）；</li>
<li>使用管家基因来对数据进行均一化(normalization)（了解）。</li>
</ul>
<p>注：在这篇笔记中，管家基因(housekeeping genes)与内参(reference)指的是同一个意思。</p>
<h2 id="合并技术重复">合并技术重复</h2>
<p>我们先使用<code>ReadqPCR</code>包来导入原始数据，这个原始数据中含有技术重复，其后缀都是<code>_TechRep.n</code>，其中n表示总的重复数。读取数据过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># BiocManager::install("ReadqPCR")</span></div><div class="line"><span class="comment"># BiocManager::install("NormqPCR")</span></div><div class="line"><span class="comment"># input raw data for qPCR</span></div><div class="line"><span class="keyword">library</span>(NormqPCR)</div><div class="line"><span class="keyword">library</span>(ReadqPCR)</div><div class="line">path &lt;- system.file(<span class="string">"exData"</span>, package = <span class="string">"ReadqPCR"</span>)</div><div class="line">qPCR.example.techReps &lt;- file.path(path, <span class="string">"qPCR.techReps.txt"</span>)</div><div class="line">qPCRBatch.qPCR.techReps &lt;- read.qPCR(qPCR.example.techReps)</div><div class="line">rownames(exprs(qPCRBatch.qPCR.techReps))[<span class="number">1</span>:<span class="number">8</span>]</div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; rownames(exprs(qPCRBatch.qPCR.techReps))[<span class="number">1</span>:<span class="number">8</span>]</div><div class="line">[<span class="number">1</span>] <span class="string">"gene_aj_TechReps.1"</span> <span class="string">"gene_aj_TechReps.2"</span> <span class="string">"gene_al_TechReps.1"</span></div><div class="line">[<span class="number">4</span>] <span class="string">"gene_al_TechReps.2"</span> <span class="string">"gene_ax_TechReps.1"</span> <span class="string">"gene_ax_TechReps.2"</span></div><div class="line">[<span class="number">7</span>] <span class="string">"gene_bo_TechReps.1"</span> <span class="string">"gene_bo_TechReps.2"</span></div></pre></td></tr></table></figure>
<p>上面是读取数据的一些信息，从中我们可以发现，里面有技术重复，例如<code>gene_aj_TechReps.1</code>与<code>gene_aj_TechReps.2</code>就是做了2个技术重复。</p>
<p>在我们计算之前，我们还要计算技术重复Cq的算术平均值(arithmetic mean)，这一步我们使用<code>combineTechReps()</code>函数来计算一下，这个函数会生成一个qPCRBatch对象，这样就计算出了技术重复的算术平均值，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># Cal arithmetic mean of the raw qPCR Cq value</span></div><div class="line"><span class="comment"># combineTechReps function will return a new qPCRBatch object</span></div><div class="line">combinedTechReps &lt;- combineTechReps(qPCRBatch.qPCR.techReps)</div><div class="line">combinedTechReps</div><div class="line">exprs(qPCRBatch.qPCR.techReps)</div><div class="line">exprs(combinedTechReps)</div><div class="line">apply(exprs(qPCRBatch.qPCR.techReps)[<span class="number">1</span>:<span class="number">2</span>,],<span class="number">2</span>,mean)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div></pre></td><td class="code"><pre><div class="line">&gt; combinedTechReps &lt;- combineTechReps(qPCRBatch.qPCR.techReps)</div><div class="line">&gt; combinedTechReps</div><div class="line">qPCRBatch (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">8</span> features, <span class="number">4</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: four one three two</div><div class="line">  varLabels: sample</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData: none</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation:  </div><div class="line">&gt; exprs(qPCRBatch.qPCR.techReps)</div><div class="line">                       four      one    three      two</div><div class="line">gene_aj_TechReps.1 <span class="number">18.02512</span> <span class="number">20.72960</span> <span class="number">21.87109</span> <span class="number">17.51878</span></div><div class="line">gene_aj_TechReps.2 <span class="number">17.59243</span> <span class="number">19.58196</span> <span class="number">21.71061</span> <span class="number">18.40029</span></div><div class="line">gene_al_TechReps.1 <span class="number">20.30233</span> <span class="number">22.23137</span> <span class="number">19.30596</span> <span class="number">21.32223</span></div><div class="line">gene_al_TechReps.2 <span class="number">20.56011</span> <span class="number">21.66608</span> <span class="number">20.07169</span> <span class="number">19.53765</span></div><div class="line">gene_ax_TechReps.1 <span class="number">20.67044</span> <span class="number">19.09129</span> <span class="number">21.90034</span> <span class="number">21.15897</span></div><div class="line">gene_ax_TechReps.2 <span class="number">20.44186</span> <span class="number">18.22294</span> <span class="number">21.28660</span> <span class="number">21.31048</span></div><div class="line">gene_bo_TechReps.1 <span class="number">20.62939</span> <span class="number">18.77342</span> <span class="number">19.70848</span> <span class="number">21.08782</span></div><div class="line">gene_bo_TechReps.2 <span class="number">20.70826</span> <span class="number">17.95030</span> <span class="number">19.53260</span> <span class="number">21.53267</span></div><div class="line">gene_cp_TechReps.1 <span class="number">23.50768</span> <span class="number">20.39530</span> <span class="number">22.06632</span> <span class="number">22.99578</span></div><div class="line">gene_cp_TechReps.2 <span class="number">22.73511</span> <span class="number">19.64446</span> <span class="number">20.74793</span> <span class="number">22.34119</span></div><div class="line">gene_dx_TechReps.1 <span class="number">21.16744</span> <span class="number">19.57131</span> <span class="number">21.62546</span> <span class="number">21.65467</span></div><div class="line">gene_dx_TechReps.2 <span class="number">21.36217</span> <span class="number">20.21180</span> <span class="number">21.41619</span> <span class="number">20.73930</span></div><div class="line">gene_ei_TechReps.1 <span class="number">19.90058</span> <span class="number">19.78343</span> <span class="number">20.26527</span> <span class="number">19.83862</span></div><div class="line">gene_ei_TechReps.2 <span class="number">19.67431</span> <span class="number">20.97861</span> <span class="number">19.82184</span> <span class="number">19.16033</span></div><div class="line">gene_jy_TechReps.1 <span class="number">21.60261</span> <span class="number">19.23546</span> <span class="number">18.22053</span> <span class="number">22.21470</span></div><div class="line">gene_jy_TechReps.2 <span class="number">21.04445</span> <span class="number">18.75998</span> <span class="number">19.31650</span> <span class="number">20.94050</span></div><div class="line">&gt; exprs(combinedTechReps)</div><div class="line">            four      one    three      two</div><div class="line">gene_aj <span class="number">17.80877</span> <span class="number">20.15578</span> <span class="number">21.79085</span> <span class="number">17.95954</span></div><div class="line">gene_al <span class="number">20.43122</span> <span class="number">21.94873</span> <span class="number">19.68883</span> <span class="number">20.42994</span></div><div class="line">gene_ax <span class="number">20.55615</span> <span class="number">18.65711</span> <span class="number">21.59347</span> <span class="number">21.23473</span></div><div class="line">gene_bo <span class="number">20.66882</span> <span class="number">18.36186</span> <span class="number">19.62054</span> <span class="number">21.31024</span></div><div class="line">gene_cp <span class="number">23.12139</span> <span class="number">20.01988</span> <span class="number">21.40713</span> <span class="number">22.66848</span></div><div class="line">gene_dx <span class="number">21.26481</span> <span class="number">19.89156</span> <span class="number">21.52082</span> <span class="number">21.19699</span></div><div class="line">gene_ei <span class="number">19.78745</span> <span class="number">20.38102</span> <span class="number">20.04355</span> <span class="number">19.49948</span></div><div class="line">gene_jy <span class="number">21.32353</span> <span class="number">18.99772</span> <span class="number">18.76851</span> <span class="number">21.57760</span></div><div class="line">&gt; apply(exprs(qPCRBatch.qPCR.techReps)[<span class="number">1</span>:<span class="number">2</span>,],<span class="number">2</span>,mean)</div><div class="line">    four      one    three      two </div><div class="line"><span class="number">17.80877</span> <span class="number">20.15578</span> <span class="number">21.79085</span> <span class="number">17.95954</span></div></pre></td></tr></table></figure>
<p>从上面我们可以看到：</p>
<ol style="list-style-type: decimal">
<li>做了2个技术重复；</li>
<li>使用<code>combineTechReps()</code>函数计算了技术重复的算术平均值；</li>
<li>又用<code>apply()</code>函数核对了一下，没问题。</li>
</ol>
<h2 id="处理未检测值">处理未检测值</h2>
<p>在qPCR实验中，当Cq值超过一定阈值后，qPCR仪有可能会将这个读数标记为Undetermined，也就是数值太小（Cq值越大，表示信号越弱），在<code>qPCRBatch</code>对象中，这个值就会被标记为<code>NA</code>。不同的人有不同的处理策略，常见的处理策略有两种。</p>
<p>第一种就是将那些大于38的Cq值认为是无法检测值，并将其标记为40，这是Taqman软件中的常见做法。而在我们的这个案例中，我们不会将其替换为40，而是替换为NA值，现在我们读入一个Taqman的qPCR数据，这批数据是用96孔板测的，其中，实验组(mia)4个重复，对照组(no-mia)4个重复，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">path &lt;- system.file(<span class="string">"exData"</span>, package = <span class="string">"NormqPCR"</span>)</div><div class="line">taqman.example &lt;- file.path(path, <span class="string">"/example.txt"</span>)</div><div class="line">qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div></pre></td></tr></table></figure>
<p>现在我们只看某个样本的检测值，即<code>Ccl20.Rn00570287_m1</code>这个数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">exprs(qPCRBatch.taqman)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; exprs(qPCRBatch.taqman)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">    fp1.day3.v     fp2.day3.v   fp5.day3.mia   fp6.day3.mia   fp.3.day.3.v </div><div class="line">            <span class="literal">NA</span>             <span class="literal">NA</span>       <span class="number">35.74190</span>       <span class="number">34.05922</span>       <span class="number">35.02052</span> </div><div class="line">  fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia </div><div class="line">            <span class="literal">NA</span>       <span class="number">35.93689</span>       <span class="number">36.57921</span></div></pre></td></tr></table></figure>
<p>现在我们使用<code>replaceAboveCutOff()</code>函数将大于35的数值替换为<code>NA</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># Use replaceAboveCutOff method in order to replace anything above 35 with NA</span></div><div class="line">qPCRBatch.taqman.replaced &lt;- replaceAboveCutOff(qPCRBatch.taqman,newVal = <span class="literal">NA</span>, cutOff = <span class="number">35</span>)</div><div class="line">exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">    fp1.day3.v     fp2.day3.v   fp5.day3.mia   fp6.day3.mia   fp.3.day.3.v </div><div class="line">            <span class="literal">NA</span>             <span class="literal">NA</span>             <span class="literal">NA</span>       <span class="number">34.05922</span>             <span class="literal">NA</span> </div><div class="line">  fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia </div><div class="line">            <span class="literal">NA</span>             <span class="literal">NA</span>             <span class="literal">NA</span></div></pre></td></tr></table></figure>
<p>但是对于有些同学来说，我不想把大于35的数值替换为NA，而是想把所有的NA都替换为40，那么可以使用<code>replaceNAs()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># Use replaesNAs methods to replace all NAs with 40</span></div><div class="line">qPCRBatch.taqman.replaced &lt;- replaceNAs(qPCRBatch.taqman, newNA = <span class="number">40</span>)</div><div class="line">exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">    fp1.day3.v     fp2.day3.v   fp5.day3.mia   fp6.day3.mia   fp.3.day.3.v </div><div class="line">      <span class="number">40.00000</span>       <span class="number">40.00000</span>       <span class="number">35.74190</span>       <span class="number">34.05922</span>       <span class="number">35.02052</span> </div><div class="line">  fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia </div><div class="line">      <span class="number">40.00000</span>       <span class="number">35.93689</span>       <span class="number">36.57921</span></div></pre></td></tr></table></figure>
<p>有的时候我们会遇到这样的一样种情况，在一次实验中，例如我们做了96孔板的qPCR实验（做了好几块96孔板），某个孔的检测器中检测的数值大于设定的阈值，而有的孔的检测器检测到的数值则没有大于设定的阈值。有的同学可能就会认为，即然一些孔的检测器检测到的值设为NAs，那么这个孔的检测器在读取其它数值时，也应该设为NAs。因为否则一个检测器的计数出现了异学，那么会导致某个基因表达水平的计算出现错误估计（这一段不太理解，原话如下）：</p>
<blockquote>
<p>In addition, the situation sometimes arises where some readings for a given detector are above a given cycle threshold, but some others are not. The user may decide for example that if a given number of readings are NAs, then all of the readings for this detector should be NAs. This is important because otherwise an unusual reading for one detector might lead to an inaccurate estimate for the expression of a given gene.</p>
</blockquote>
<p>对于不同的样本类型这个过程应该分开处理，因为你也许想比较不同样本之间的某个特定的基因。因此有必要使用比较矩阵来指定每个样本类型的重复数。因此，在上面的案例中，我们一共有两个类型，分别是实验组与对照组，每组4个重复，我们可以使用<code>contrastMatrix()</code>函数和<code>sampleMaxMatrix()</code>函数来创建比较矩阵，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># Use contrastMatrix() and sampleMaxMatrix() to generate contrast matrix</span></div><div class="line">sampleNames(qPCRBatch.taqman) <span class="comment"># Check group information</span></div><div class="line">a &lt;- c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>) <span class="comment"># one for each sample type, with 1 representing</span></div><div class="line">b &lt;- c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>) <span class="comment"># position of sample type in samplenames vector</span></div><div class="line">contM &lt;- cbind(a,b)</div><div class="line">colnames(contM) &lt;- c(<span class="string">"case"</span>,<span class="string">"control"</span>) <span class="comment"># set the names of each sample type</span></div><div class="line">rownames(contM) &lt;- sampleNames(qPCRBatch.taqman) <span class="comment"># set row names</span></div><div class="line">contM</div><div class="line">sMaxM &lt;- t(as.matrix(c(<span class="number">3</span>,<span class="number">3</span>))) <span class="comment"># now make the contrast matrix</span></div><div class="line">colnames(sMaxM) &lt;- c(<span class="string">"case"</span>,<span class="string">"control"</span>) <span class="comment"># make sure these line up with samples</span></div><div class="line">sMaxM</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment"># Use contrastMatrix() and sampleMaxMatrix() to generate contrast matrix</span></div><div class="line">&gt; sampleNames(qPCRBatch.taqman) <span class="comment"># Check group information</span></div><div class="line">[<span class="number">1</span>] <span class="string">"fp1.day3.v"</span>     <span class="string">"fp2.day3.v"</span>     <span class="string">"fp5.day3.mia"</span>   <span class="string">"fp6.day3.mia"</span>  </div><div class="line">[<span class="number">5</span>] <span class="string">"fp.3.day.3.v"</span>   <span class="string">"fp.4.day.3.v"</span>   <span class="string">"fp.7.day.3.mia"</span> <span class="string">"fp.8.day.3.mia"</span></div><div class="line">&gt; a &lt;- c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>) <span class="comment"># one for each sample type, with 1 representing</span></div><div class="line">&gt; b &lt;- c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>) <span class="comment"># position of sample type in samplenames vector</span></div><div class="line">&gt; contM &lt;- cbind(a,b)</div><div class="line">&gt; colnames(contM) &lt;- c(<span class="string">"case"</span>,<span class="string">"control"</span>) <span class="comment"># set the names of each sample type</span></div><div class="line">&gt; rownames(contM) &lt;- sampleNames(qPCRBatch.taqman) <span class="comment"># set row names</span></div><div class="line">&gt; contM</div><div class="line">               case control</div><div class="line">fp1.day3.v        <span class="number">0</span>       <span class="number">1</span></div><div class="line">fp2.day3.v        <span class="number">0</span>       <span class="number">1</span></div><div class="line">fp5.day3.mia      <span class="number">1</span>       <span class="number">0</span></div><div class="line">fp6.day3.mia      <span class="number">1</span>       <span class="number">0</span></div><div class="line">fp.3.day.3.v      <span class="number">0</span>       <span class="number">1</span></div><div class="line">fp.4.day.3.v      <span class="number">0</span>       <span class="number">1</span></div><div class="line">fp.7.day.3.mia    <span class="number">1</span>       <span class="number">0</span></div><div class="line">fp.8.day.3.mia    <span class="number">1</span>       <span class="number">0</span></div><div class="line">&gt; sMaxM &lt;- t(as.matrix(c(<span class="number">3</span>,<span class="number">3</span>))) <span class="comment"># now make the contrast matrix</span></div><div class="line">&gt; colnames(sMaxM) &lt;- c(<span class="string">"case"</span>,<span class="string">"control"</span>) <span class="comment"># make sure these line up with samples</span></div><div class="line">&gt; sMaxM</div><div class="line">     case control</div><div class="line">[<span class="number">1</span>,]    <span class="number">3</span>       <span class="number">3</span></div></pre></td></tr></table></figure>
<p>有关更多的算法细节可以参考<code>limmma</code>包中的内容，那个包中创建差异基因比较的矩阵也是类似方法。</p>
<p>例如，如果用户决定将那些4个重复中，有3个是NA的值都设为NA的话，可以使用<code>makeAllNewVal()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">qPCRBatch.taqman.replaced &lt;- makeAllNewVal(qPCRBatch.taqman, contM,sMaxM, newVal=<span class="literal">NA</span>)</div></pre></td></tr></table></figure>
<p>现在我们来看一下<code>Ccl20.Rn00570287 m1</code>这个样本的检测值，对照组的值都为NA，因为原始的4个技术重复中，有3个是NA，有一个是35，因此全设为NA。不过实验组的数据全部进行了保留，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">exprs(qPCRBatch.taqman)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">&gt; exprs(qPCRBatch.taqman)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">    fp1.day3.v     fp2.day3.v   fp5.day3.mia   fp6.day3.mia   fp.3.day.3.v </div><div class="line">            <span class="literal">NA</span>             <span class="literal">NA</span>       <span class="number">35.74190</span>       <span class="number">34.05922</span>       <span class="number">35.02052</span> </div><div class="line">  fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia </div><div class="line">            <span class="literal">NA</span>       <span class="number">35.93689</span>       <span class="number">36.57921</span> </div><div class="line">&gt; exprs(qPCRBatch.taqman.replaced)[<span class="string">"Ccl20.Rn00570287_m1"</span>,]</div><div class="line">    fp1.day3.v     fp2.day3.v   fp5.day3.mia   fp6.day3.mia   fp.3.day.3.v </div><div class="line">            <span class="literal">NA</span>             <span class="literal">NA</span>       <span class="number">35.74190</span>       <span class="number">34.05922</span>             <span class="literal">NA</span> </div><div class="line">  fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia </div><div class="line">            <span class="literal">NA</span>       <span class="number">35.93689</span>       <span class="number">36.57921</span></div></pre></td></tr></table></figure>
<p>前面这两部分内容（合并技术重复与处理未检测值）了解一下就行，平时也用不到，最重要的是下面的内容。</p>
<h2 id="选择最稳定内参基因">选择最稳定内参基因</h2>
<p>这一部分涉及选择最稳定的内参基因，其中用的是geNorm算法，这个算法的相关文献信息如下所示：</p>
<blockquote>
<p>Jo Vandesompele, Katleen De Preter, Filip Pattyn et al. (2002). Accurate normalization of real-time quantitative RT-PCR data by geometric averaging of multiple internal control genes. Genome Biology 2002. 3(7):research0034.1-0034.11. http://genomebiology.com/2002/3/7/research/0034/</p>
</blockquote>
<h3 id="genorm">geNorm</h3>
<p>先加载数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">options(width = <span class="number">68</span>)</div><div class="line">data(geNorm)</div><div class="line">str(exprs(geNorm.qPCRBatch))</div></pre></td></tr></table></figure>
<p>数据结构如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; str(exprs(geNorm.qPCRBatch))</div><div class="line"> num [<span class="number">1</span>:<span class="number">10</span>, <span class="number">1</span>:<span class="number">85</span>] <span class="number">0.0425</span> <span class="number">0.0576</span> <span class="number">0.1547</span> <span class="number">0.1096</span> <span class="number">0.118</span> <span class="keyword">...</span></div><div class="line"> - attr(*, <span class="string">"dimnames"</span>)=List of <span class="number">2</span></div><div class="line">  ..$ : chr [<span class="number">1</span>:<span class="number">10</span>] <span class="string">"ACTB"</span> <span class="string">"B2M"</span> <span class="string">"GAPD"</span> <span class="string">"HMBS"</span> <span class="keyword">...</span></div><div class="line">  ..$ : chr [<span class="number">1</span>:<span class="number">85</span>] <span class="string">"BM1"</span> <span class="string">"BM2"</span> <span class="string">"BM3"</span> <span class="string">"BM4"</span> <span class="keyword">...</span></div></pre></td></tr></table></figure>
<p>现在我们对选择的内参基因进行排序。我们在这里使用<code>selectHKs()</code>函数来实现geNorm算法中的逐步处理过程，这里的数据都是文献中材料与方法中提到的数据。</p>
<p>第一步，我们计算所有内参基因稳定检测值M后，排除那些M值最高的内参基因；</p>
<p>第二步，再次计算M值，再排除掉最高M值的内参基因，直到剩下两个内参基因（也就是参数<code>minNrHK=2</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">tissue &lt;- as.factor(c(rep(<span class="string">"BM"</span>, <span class="number">9</span>), rep(<span class="string">"FIB"</span>, <span class="number">20</span>), rep(<span class="string">"LEU"</span>, <span class="number">13</span>),rep(<span class="string">"NB"</span>, <span class="number">34</span>), rep(<span class="string">"POOL"</span>, <span class="number">9</span>)))</div><div class="line">res.BM &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"BM"</span>], method = <span class="string">"geNorm"</span>,Symbols = featureNames(geNorm.qPCRBatch),minNrHK = <span class="number">2</span>, log = <span class="literal">FALSE</span>)</div></pre></td></tr></table></figure>
<p>计算过程中，这个函数会给出不断计算的M值，结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div><div class="line">58</div><div class="line">59</div><div class="line">60</div><div class="line">61</div><div class="line">62</div><div class="line">63</div><div class="line">64</div><div class="line">65</div><div class="line">66</div><div class="line">67</div><div class="line">68</div><div class="line">69</div><div class="line">70</div><div class="line">71</div><div class="line">72</div><div class="line">73</div><div class="line">74</div><div class="line">75</div><div class="line">76</div><div class="line">77</div><div class="line">78</div><div class="line">79</div><div class="line">80</div><div class="line">81</div><div class="line">82</div><div class="line">83</div><div class="line">84</div><div class="line">85</div><div class="line">86</div><div class="line">87</div><div class="line">88</div><div class="line">89</div><div class="line">90</div><div class="line">91</div><div class="line">92</div><div class="line">93</div><div class="line">94</div><div class="line">95</div><div class="line">96</div><div class="line">97</div><div class="line">98</div><div class="line">99</div><div class="line">100</div><div class="line">101</div><div class="line">102</div><div class="line">103</div><div class="line">104</div><div class="line">105</div><div class="line">106</div><div class="line">107</div><div class="line">108</div><div class="line">109</div><div class="line">110</div><div class="line">111</div><div class="line">112</div><div class="line">113</div><div class="line">114</div><div class="line">115</div><div class="line">116</div><div class="line">117</div><div class="line">118</div><div class="line">119</div><div class="line">120</div><div class="line">121</div><div class="line">122</div><div class="line">123</div><div class="line">124</div><div class="line">125</div><div class="line">126</div><div class="line">127</div><div class="line">128</div><div class="line">129</div><div class="line">130</div><div class="line">131</div><div class="line">132</div><div class="line">133</div><div class="line">134</div><div class="line">135</div><div class="line">136</div><div class="line">137</div><div class="line">138</div><div class="line">139</div><div class="line">140</div><div class="line">141</div><div class="line">142</div><div class="line">143</div><div class="line">144</div><div class="line">145</div><div class="line">146</div><div class="line">147</div></pre></td><td class="code"><pre><div class="line">res.BM &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"BM"</span>], method = <span class="string">"geNorm"</span>,Symbols = featureNames(geNorm.qPCRBatch),minNrHK = <span class="number">2</span>, log = <span class="literal">FALSE</span>)</div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">1</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">    HPRT1     YWHAZ    RPL13A       UBC      GAPD      SDHA       TBP </div><div class="line"><span class="number">0.5160313</span> <span class="number">0.5314564</span> <span class="number">0.5335963</span> <span class="number">0.5700961</span> <span class="number">0.6064919</span> <span class="number">0.6201470</span> <span class="number">0.6397969</span> </div><div class="line">     HMBS       B2M      ACTB </div><div class="line"><span class="number">0.7206013</span> <span class="number">0.7747634</span> <span class="number">0.8498739</span> </div><div class="line">average stability M:	<span class="number">0.63628545246682</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	ACTB</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">9</span>/<span class="number">10</span>):	<span class="number">0.0764690052563778</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">2</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">    HPRT1    RPL13A     YWHAZ       UBC      GAPD      SDHA       TBP </div><div class="line"><span class="number">0.4705664</span> <span class="number">0.5141375</span> <span class="number">0.5271169</span> <span class="number">0.5554718</span> <span class="number">0.5575295</span> <span class="number">0.5738460</span> <span class="number">0.6042110</span> </div><div class="line">     HMBS       B2M </div><div class="line"><span class="number">0.6759176</span> <span class="number">0.7671985</span> </div><div class="line">average stability M:	<span class="number">0.582888329316757</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	B2M</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">8</span>/<span class="number">9</span>):	<span class="number">0.0776534266912183</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">3</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">    HPRT1    RPL13A      SDHA     YWHAZ       UBC      GAPD       TBP </div><div class="line"><span class="number">0.4391222</span> <span class="number">0.4733732</span> <span class="number">0.5243665</span> <span class="number">0.5253471</span> <span class="number">0.5403137</span> <span class="number">0.5560120</span> <span class="number">0.5622094</span> </div><div class="line">     HMBS </div><div class="line"><span class="number">0.6210820</span> </div><div class="line">average stability M:	<span class="number">0.530228279613623</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	HMBS</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">7</span>/<span class="number">8</span>):	<span class="number">0.0671119963410967</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">4</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">    HPRT1    RPL13A     YWHAZ       UBC      SDHA      GAPD       TBP </div><div class="line"><span class="number">0.4389069</span> <span class="number">0.4696398</span> <span class="number">0.4879728</span> <span class="number">0.5043292</span> <span class="number">0.5178634</span> <span class="number">0.5245346</span> <span class="number">0.5563591</span> </div><div class="line">average stability M:	<span class="number">0.499943693933222</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	TBP</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">6</span>/<span class="number">7</span>):	<span class="number">0.0681320232188603</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">5</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">    HPRT1    RPL13A       UBC     YWHAZ      GAPD      SDHA </div><div class="line"><span class="number">0.4292808</span> <span class="number">0.4447874</span> <span class="number">0.4594181</span> <span class="number">0.4728920</span> <span class="number">0.5012107</span> <span class="number">0.5566762</span> </div><div class="line">average stability M:	<span class="number">0.477377523800525</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	SDHA</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">5</span>/<span class="number">6</span>):	<span class="number">0.0806194432580746</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">6</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">      UBC    RPL13A     HPRT1     YWHAZ      GAPD </div><div class="line"><span class="number">0.4195958</span> <span class="number">0.4204997</span> <span class="number">0.4219179</span> <span class="number">0.4424631</span> <span class="number">0.4841646</span> </div><div class="line">average stability M:	<span class="number">0.437728198765878</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	GAPD</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">4</span>/<span class="number">5</span>):	<span class="number">0.0841653121631615</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">7</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">   RPL13A       UBC     YWHAZ     HPRT1 </div><div class="line"><span class="number">0.3699163</span> <span class="number">0.3978736</span> <span class="number">0.4173706</span> <span class="number">0.4419220</span> </div><div class="line">average stability M:	<span class="number">0.406770625156432</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	HPRT1</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">3</span>/<span class="number">4</span>):	<span class="number">0.097678269387021</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">8</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">      UBC    RPL13A     YWHAZ </div><div class="line"><span class="number">0.3559286</span> <span class="number">0.3761358</span> <span class="number">0.3827933</span> </div><div class="line">average stability M:	<span class="number">0.371619241507029</span></div><div class="line"></div><div class="line">variable with lowest stability (largest M value):	YWHAZ</div><div class="line"></div><div class="line">Pairwise variation, (<span class="number">2</span>/<span class="number">3</span>):	<span class="number">0.113745049966055</span></div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">9</span>:</div><div class="line"></div><div class="line">stability values M:</div><div class="line"></div><div class="line">   RPL13A       UBC </div><div class="line"><span class="number">0.3492712</span> <span class="number">0.3492712</span> </div><div class="line">average stability M:	<span class="number">0.349271187472188</span></div><div class="line"></div><div class="line">&gt; res.BM</div><div class="line">$ranking</div><div class="line">       <span class="number">1</span>        <span class="number">1</span>        <span class="number">3</span>        <span class="number">4</span>        <span class="number">5</span>        <span class="number">6</span>        <span class="number">7</span> </div><div class="line"><span class="string">"RPL13A"</span>    <span class="string">"UBC"</span>  <span class="string">"YWHAZ"</span>  <span class="string">"HPRT1"</span>   <span class="string">"GAPD"</span>   <span class="string">"SDHA"</span>    <span class="string">"TBP"</span> </div><div class="line">       <span class="number">8</span>        <span class="number">9</span>       <span class="number">10</span> </div><div class="line">  <span class="string">"HMBS"</span>    <span class="string">"B2M"</span>   <span class="string">"ACTB"</span> </div><div class="line"></div><div class="line">$variation</div><div class="line">      <span class="number">9</span>/<span class="number">10</span>        <span class="number">8</span>/<span class="number">9</span>        <span class="number">7</span>/<span class="number">8</span>        <span class="number">6</span>/<span class="number">7</span>        <span class="number">5</span>/<span class="number">6</span>        <span class="number">4</span>/<span class="number">5</span> </div><div class="line"><span class="number">0.07646901</span> <span class="number">0.07765343</span> <span class="number">0.06711200</span> <span class="number">0.06813202</span> <span class="number">0.08061944</span> <span class="number">0.08416531</span> </div><div class="line">       <span class="number">3</span>/<span class="number">4</span>        <span class="number">2</span>/<span class="number">3</span> </div><div class="line"><span class="number">0.09767827</span> <span class="number">0.11374505</span> </div><div class="line"></div><div class="line">$meanM</div><div class="line">       <span class="number">10</span>         <span class="number">9</span>         <span class="number">8</span>         <span class="number">7</span>         <span class="number">6</span>         <span class="number">5</span> </div><div class="line"><span class="number">0.6362855</span> <span class="number">0.5828883</span> <span class="number">0.5302283</span> <span class="number">0.4999437</span> <span class="number">0.4773775</span> <span class="number">0.4377282</span> </div><div class="line">        <span class="number">4</span>         <span class="number">3</span>         <span class="number">2</span> </div><div class="line"><span class="number">0.4067706</span> <span class="number">0.3716192</span> <span class="number">0.3492712</span></div></pre></td></tr></table></figure>
<p>以上只是BM这类组织的计算结果。现在我们把其它的组织都计算出来，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">res.POOL &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"POOL"</span>],</div><div class="line">	method = <span class="string">"geNorm"</span>,</div><div class="line">	Symbols = featureNames(geNorm.qPCRBatch),</div><div class="line">	minNrHK = <span class="number">2</span>, trace = <span class="literal">FALSE</span>, log = <span class="literal">FALSE</span>)</div><div class="line">res.FIB &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"FIB"</span>],</div><div class="line">	method = <span class="string">"geNorm"</span>,</div><div class="line">	Symbols = featureNames(geNorm.qPCRBatch),</div><div class="line">	minNrHK = <span class="number">2</span>, trace = <span class="literal">FALSE</span>, log = <span class="literal">FALSE</span>)</div><div class="line">res.LEU &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"LEU"</span>],</div><div class="line">	method = <span class="string">"geNorm"</span>,</div><div class="line">	Symbols = featureNames(geNorm.qPCRBatch),</div><div class="line">	minNrHK = <span class="number">2</span>, trace = <span class="literal">FALSE</span>, log = <span class="literal">FALSE</span>)</div><div class="line">res.NB &lt;- selectHKs(geNorm.qPCRBatch[,tissue == <span class="string">"NB"</span>],</div><div class="line">	method = <span class="string">"geNorm"</span>,</div><div class="line">	Symbols = featureNames(geNorm.qPCRBatch),</div><div class="line">	minNrHK = <span class="number">2</span>, trace = <span class="literal">FALSE</span>, log = <span class="literal">FALSE</span>)</div></pre></td></tr></table></figure>
<p>我们就获得了基因的下面排序信息，这个就是文献中的Table 3，文献中的信息如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925163029.png">

</div>
<p>我们自己的计算结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">ranks &lt;- data.frame(c(<span class="number">1</span>, <span class="number">1</span>:<span class="number">9</span>),</div><div class="line">	res.BM$ranking, res.POOL$ranking,</div><div class="line">	res.FIB$ranking, res.LEU$ranking,</div><div class="line">	res.NB$ranking)</div><div class="line">names(ranks) &lt;- c(<span class="string">"rank"</span>, <span class="string">"BM"</span>, <span class="string">"POOL"</span>, <span class="string">"FIB"</span>, <span class="string">"LEU"</span>, <span class="string">"NB"</span>)</div><div class="line">ranks</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">ranks</div><div class="line">   rank     BM   POOL    FIB    LEU     NB</div><div class="line"><span class="number">1</span>     <span class="number">1</span> RPL13A   GAPD   GAPD    UBC   GAPD</div><div class="line"><span class="number">2</span>     <span class="number">1</span>    UBC   SDHA  HPRT1  YWHAZ  HPRT1</div><div class="line"><span class="number">3</span>     <span class="number">2</span>  YWHAZ   HMBS  YWHAZ    B2M   SDHA</div><div class="line"><span class="number">4</span>     <span class="number">3</span>  HPRT1  HPRT1    UBC   GAPD    UBC</div><div class="line"><span class="number">5</span>     <span class="number">4</span>   GAPD    TBP   ACTB RPL13A   HMBS</div><div class="line"><span class="number">6</span>     <span class="number">5</span>   SDHA    UBC    TBP    TBP  YWHAZ</div><div class="line"><span class="number">7</span>     <span class="number">6</span>    TBP RPL13A   SDHA   SDHA    TBP</div><div class="line"><span class="number">8</span>     <span class="number">7</span>   HMBS  YWHAZ RPL13A  HPRT1   ACTB</div><div class="line"><span class="number">9</span>     <span class="number">8</span>    B2M   ACTB    B2M   HMBS RPL13A</div><div class="line"><span class="number">10</span>    <span class="number">9</span>   ACTB    B2M   HMBS   ACTB    B2M</div></pre></td></tr></table></figure>
<p>现在绘制出每个细胞类型的平均基因表达稳定值M的折线图，这个对应于文献中的Figure 2，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925163314.png">

</div>
<p>代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RColorBrewer)</div><div class="line">mypalette &lt;- brewer.pal(<span class="number">5</span>, <span class="string">"Set1"</span>)</div><div class="line">matplot(cbind(res.BM$meanM, res.POOL$meanM, res.FIB$meanM,</div><div class="line">res.LEU$meanM, res.NB$meanM), type = <span class="string">"b"</span>,</div><div class="line">ylab = <span class="string">"Average expression stability M"</span>,</div><div class="line">xlab = <span class="string">"Number of remaining control genes"</span>,</div><div class="line">axes = <span class="literal">FALSE</span>, pch = <span class="number">19</span>, col = mypalette,</div><div class="line">ylim = c(<span class="number">0.2</span>, <span class="number">1.22</span>), lty = <span class="number">1</span>, lwd = <span class="number">2</span>,</div><div class="line">main = <span class="string">"Figure 2 in Vandesompele et al. (2002)"</span>)</div><div class="line">axis(<span class="number">1</span>, at = <span class="number">1</span>:<span class="number">9</span>, labels = as.character(<span class="number">10</span>:<span class="number">2</span>))</div><div class="line">axis(<span class="number">2</span>, at = seq(<span class="number">0.2</span>, <span class="number">1.2</span>, by = <span class="number">0.2</span>), labels = seq(<span class="number">0.2</span>, <span class="number">1.2</span>, by = <span class="number">0.2</span>))</div><div class="line">box()</div><div class="line">abline(h = seq(<span class="number">0.2</span>, <span class="number">1.2</span>, by = <span class="number">0.2</span>), lty = <span class="number">2</span>, lwd = <span class="number">1</span>, col = <span class="string">"grey"</span>)</div><div class="line">legend(<span class="string">"topright"</span>, legend = c(<span class="string">"BM"</span>, <span class="string">"POOL"</span>, <span class="string">"FIB"</span>, <span class="string">"LEU"</span>, <span class="string">"NB"</span>),</div><div class="line">fill = mypalette)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925181159.jpeg">

</div>
<p>绘制每个细胞类型的成对方差(pairwise variation)图，这个对应文献中的Figure 3，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925163523.png">

</div>
<p>代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">mypalette &lt;- brewer.pal(<span class="number">8</span>, <span class="string">"YlGnBu"</span>)</div><div class="line">barplot(cbind(res.POOL$variation, res.LEU$variation, res.NB$variation,</div><div class="line">res.FIB$variation, res.BM$variation), beside = <span class="literal">TRUE</span>,</div><div class="line">col = mypalette, space = c(<span class="number">0</span>, <span class="number">2</span>),</div><div class="line">names.arg = c(<span class="string">"POOL"</span>, <span class="string">"LEU"</span>, <span class="string">"NB"</span>, <span class="string">"FIB"</span>, <span class="string">"BM"</span>),</div><div class="line">ylab = <span class="string">"Pairwise variation V"</span>,</div><div class="line">main = <span class="string">"Figure 3(a) in Vandesompele et al. (2002)"</span>)</div><div class="line">legend(<span class="string">"topright"</span>, legend = c(<span class="string">"V9/10"</span>, <span class="string">"V8/9"</span>, <span class="string">"V7/8"</span>, <span class="string">"V6/7"</span>,</div><div class="line"><span class="string">"V5/6"</span>, <span class="string">"V4/5"</span>, <span class="string">"V3/4"</span>, <span class="string">"V2/3"</span>),</div><div class="line">fill = mypalette, ncol = <span class="number">2</span>)</div><div class="line">abline(h = seq(<span class="number">0.05</span>, <span class="number">0.25</span>, by = <span class="number">0.05</span>), lty = <span class="number">2</span>, col = <span class="string">"grey"</span>)</div><div class="line">abline(h = <span class="number">0.15</span>, lty = <span class="number">1</span>, col = <span class="string">"black"</span>)</div></pre></td></tr></table></figure>
<p>注：文献中推荐的成对方差值的截止值为0.15，因此低于此值后，就不需要使用其它的内参基因了。</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925181222.jpeg">

</div>
<h3 id="normfinder">NormFinder</h3>
<p>第二种选择内参的基因的方法是<code>NormFinder</code>，这种方法来源于以下文献：</p>
<blockquote>
<p>Andersen, C. L., et al. (2004). “Normalization of real-time quantitative reverse transcription-PCR data: a model-based variance estimation approach to identify genes suited for normalization, applied to bladder and colon cancer data sets.” Cancer Res 64(15): 5245-5250.</p>
</blockquote>
<p>现在我们计算Table 3，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925181651.png">

</div>
<p>第一步，导入数据：：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="comment">## NormFinder</span></div><div class="line">data(Colon)</div><div class="line">Colon</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">&gt; data(Colon)</div><div class="line">&gt; Colon</div><div class="line">qPCRBatch (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">13</span> features, <span class="number">40</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: I459N <span class="number">90</span> <span class="keyword">...</span> I-C1056T (<span class="number">40</span> total)</div><div class="line">  varLabels: Sample.no. Classification</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData</div><div class="line">  featureNames: UBC UBB <span class="keyword">...</span> TUBA6 (<span class="number">13</span> total)</div><div class="line">  fvarLabels: Symbol Gene.name</div><div class="line">  fvarMetadata: labelDescription</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: Table <span class="number">1</span> <span class="keyword">in</span> Andersen et al. (<span class="number">2004</span>)</div></pre></td></tr></table></figure>
<p>继续计算，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">Class &lt;- pData(Colon)[,<span class="string">"Classification"</span>]</div><div class="line">res.Colon &lt;- stabMeasureRho(Colon, group = Class, log = <span class="literal">FALSE</span>)</div><div class="line">sort(res.Colon) <span class="comment"># see Table 3 in Andersen et al (2004)</span></div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; sort(res.Colon) <span class="comment"># see Table 3 in Andersen et al (2004)</span></div><div class="line">      UBC      GAPD      TPT1       UBB     TUBA6     RPS13      NACA </div><div class="line"><span class="number">0.1821707</span> <span class="number">0.2146061</span> <span class="number">0.2202956</span> <span class="number">0.2471573</span> <span class="number">0.2700641</span> <span class="number">0.2813039</span> <span class="number">0.2862397</span> </div><div class="line">     CFL1      SUI1      ACTB      CLTC     RPS23  FLJ20030 </div><div class="line"><span class="number">0.2870467</span> <span class="number">0.3139404</span> <span class="number">0.3235918</span> <span class="number">0.3692880</span> <span class="number">0.3784909</span> <span class="number">0.3935173</span></div></pre></td></tr></table></figure>
<p>再看另外的一个数据集<code>Bladder</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">data(Bladder)</div><div class="line">Bladder</div><div class="line">grade &lt;- pData(Bladder)[,<span class="string">"Grade"</span>]</div><div class="line">res.Bladder &lt;- stabMeasureRho(Bladder,</div><div class="line">                              group = grade,</div><div class="line">                              log = <span class="literal">FALSE</span>)</div><div class="line">sort(res.Bladder)</div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div></pre></td><td class="code"><pre><div class="line">&gt; data(Bladder)</div><div class="line">&gt; Bladder</div><div class="line">qPCRBatch (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">14</span> features, <span class="number">28</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: <span class="number">335</span>-<span class="number">6</span> <span class="number">1131</span>-<span class="number">1</span> <span class="keyword">...</span> <span class="number">1356</span>-<span class="number">1</span> (<span class="number">28</span> total)</div><div class="line">  varLabels: Sample.no. Grade</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData</div><div class="line">  featureNames: ATP5B HSPCB <span class="keyword">...</span> FLJ20030 (<span class="number">14</span> total)</div><div class="line">  fvarLabels: Symbol Gene.name</div><div class="line">  fvarMetadata: labelDescription</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: Table <span class="number">1</span> <span class="keyword">in</span> Andersen et al. (<span class="number">2004</span>) </div><div class="line">&gt; grade &lt;- pData(Bladder)[,<span class="string">"Grade"</span>]</div><div class="line">&gt; res.Bladder &lt;- stabMeasureRho(Bladder,</div><div class="line">+                               group = grade,</div><div class="line">+                               log = <span class="literal">FALSE</span>)</div><div class="line">Warning message:</div><div class="line">In .local(x, <span class="keyword">...</span>) : Argument <span class="string">'group'</span> is transformed to a factor vector.</div><div class="line">&gt; sort(res.Bladder)</div><div class="line">    HSPCB      TEGT     ATP5B       UBC     RPS23     RPS13      CFL1 </div><div class="line"><span class="number">0.1539598</span> <span class="number">0.1966556</span> <span class="number">0.1987227</span> <span class="number">0.2033477</span> <span class="number">0.2139626</span> <span class="number">0.2147852</span> <span class="number">0.2666129</span> </div><div class="line"> FLJ20030      TPT1       UBB     FLOT2      GAPD    S100A6      ACTB </div><div class="line"><span class="number">0.2672918</span> <span class="number">0.2691553</span> <span class="number">0.2826051</span> <span class="number">0.2960429</span> <span class="number">0.3408742</span> <span class="number">0.3453435</span> <span class="number">0.3497295</span></div></pre></td></tr></table></figure>
<p>我们也可以使用geNorm算法来计算，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">selectHKs(Colon, </div><div class="line">          log = <span class="literal">FALSE</span>, </div><div class="line">          trace = <span class="literal">FALSE</span>,</div><div class="line">          Symbols = featureNames(Colon))$ranking</div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; selectHKs(Colon, </div><div class="line">+           log = <span class="literal">FALSE</span>, </div><div class="line">+           trace = <span class="literal">FALSE</span>,</div><div class="line">+           Symbols = featureNames(Colon))$ranking</div><div class="line">         <span class="number">1</span>          <span class="number">1</span>          <span class="number">3</span>          <span class="number">4</span>          <span class="number">5</span>          <span class="number">6</span>          <span class="number">7</span> </div><div class="line">   <span class="string">"RPS23"</span>     <span class="string">"TPT1"</span>    <span class="string">"RPS13"</span>     <span class="string">"SUI1"</span>      <span class="string">"UBC"</span>     <span class="string">"GAPD"</span>    <span class="string">"TUBA6"</span> </div><div class="line">         <span class="number">8</span>          <span class="number">9</span>         <span class="number">10</span>         <span class="number">11</span>         <span class="number">12</span>         <span class="number">13</span> </div><div class="line">     <span class="string">"UBB"</span>     <span class="string">"NACA"</span>     <span class="string">"CFL1"</span>     <span class="string">"CLTC"</span>     <span class="string">"ACTB"</span> <span class="string">"FLJ20030"</span></div></pre></td></tr></table></figure>
<p>再来看Baldder的数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">selectHKs(Bladder,</div><div class="line">	log = <span class="literal">FALSE</span>,</div><div class="line">	trace = <span class="literal">FALSE</span>,</div><div class="line">	Symbols = featureNames(Bladder))$ranking</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; selectHKs(Bladder,</div><div class="line">+           log = <span class="literal">FALSE</span>,</div><div class="line">+           trace = <span class="literal">FALSE</span>,</div><div class="line">+           Symbols = featureNames(Bladder))$ranking</div><div class="line">         <span class="number">1</span>          <span class="number">1</span>          <span class="number">3</span>          <span class="number">4</span>          <span class="number">5</span>          <span class="number">6</span>          <span class="number">7</span> </div><div class="line">    <span class="string">"CFL1"</span>      <span class="string">"UBC"</span>    <span class="string">"ATP5B"</span>    <span class="string">"HSPCB"</span>     <span class="string">"GAPD"</span>     <span class="string">"TEGT"</span>    <span class="string">"RPS23"</span> </div><div class="line">         <span class="number">8</span>          <span class="number">9</span>         <span class="number">10</span>         <span class="number">11</span>         <span class="number">12</span>         <span class="number">13</span>         <span class="number">14</span> </div><div class="line">   <span class="string">"RPS13"</span>     <span class="string">"TPT1"</span> <span class="string">"FLJ20030"</span>    <span class="string">"FLOT2"</span>      <span class="string">"UBB"</span>     <span class="string">"ACTB"</span>   <span class="string">"S100A6"</span></div></pre></td></tr></table></figure>
<p>我们也可以使用NormFinder算法的分步计算方式工来看多个内参基因的情况，也就是文献中附录中的<code>Average control gene</code>部分，<code>selectHKs</code>函数中含有这种算法的参数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">Class &lt;- pData(Colon)[,<span class="string">"Classification"</span>]</div><div class="line">selectHKs(Colon, </div><div class="line">	group = Class, </div><div class="line">	log = <span class="literal">FALSE</span>, </div><div class="line">	trace = <span class="literal">TRUE</span>,</div><div class="line">	Symbols = featureNames(Colon), </div><div class="line">	minNrHKs = <span class="number">12</span>,</div><div class="line">	method = <span class="string">"NormFinder"</span>)$ranking</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div><div class="line">58</div><div class="line">59</div><div class="line">60</div><div class="line">61</div><div class="line">62</div><div class="line">63</div><div class="line">64</div><div class="line">65</div><div class="line">66</div><div class="line">67</div><div class="line">68</div><div class="line">69</div><div class="line">70</div><div class="line">71</div><div class="line">72</div><div class="line">73</div><div class="line">74</div><div class="line">75</div><div class="line">76</div><div class="line">77</div><div class="line">78</div><div class="line">79</div><div class="line">80</div><div class="line">81</div><div class="line">82</div><div class="line">83</div><div class="line">84</div><div class="line">85</div><div class="line">86</div><div class="line">87</div><div class="line">88</div><div class="line">89</div><div class="line">90</div><div class="line">91</div><div class="line">92</div><div class="line">93</div><div class="line">94</div><div class="line">95</div><div class="line">96</div><div class="line">97</div><div class="line">98</div><div class="line">99</div><div class="line">100</div><div class="line">101</div><div class="line">102</div><div class="line">103</div><div class="line">104</div><div class="line">105</div><div class="line">106</div><div class="line">107</div><div class="line">108</div><div class="line">109</div><div class="line">110</div><div class="line">111</div><div class="line">112</div><div class="line">113</div><div class="line">114</div><div class="line">115</div><div class="line">116</div><div class="line">117</div><div class="line">118</div><div class="line">119</div><div class="line">120</div><div class="line">121</div><div class="line">122</div><div class="line">123</div><div class="line">124</div><div class="line">125</div><div class="line">126</div><div class="line">127</div><div class="line">128</div><div class="line">129</div><div class="line">130</div><div class="line">131</div><div class="line">132</div><div class="line">133</div><div class="line">134</div><div class="line">135</div><div class="line">136</div><div class="line">137</div><div class="line">138</div><div class="line">139</div><div class="line">140</div><div class="line">141</div><div class="line">142</div><div class="line">143</div><div class="line">144</div><div class="line">145</div><div class="line">146</div></pre></td><td class="code"><pre><div class="line">&gt; Class &lt;- pData(Colon)[,<span class="string">"Classification"</span>]</div><div class="line">&gt; selectHKs(Colon, </div><div class="line">+           group = Class, </div><div class="line">+           log = <span class="literal">FALSE</span>, </div><div class="line">+           trace = <span class="literal">TRUE</span>,</div><div class="line">+           Symbols = featureNames(Colon), </div><div class="line">+           minNrHKs = <span class="number">12</span>,</div><div class="line">+           method = <span class="string">"NormFinder"</span>)$ranking</div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">1</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">      UBC      GAPD      TPT1       UBB     TUBA6     RPS13      NACA </div><div class="line"><span class="number">0.1821707</span> <span class="number">0.2146061</span> <span class="number">0.2202956</span> <span class="number">0.2471573</span> <span class="number">0.2700641</span> <span class="number">0.2813039</span> <span class="number">0.2862397</span> </div><div class="line">     CFL1      SUI1      ACTB      CLTC     RPS23  FLJ20030 </div><div class="line"><span class="number">0.2870467</span> <span class="number">0.3139404</span> <span class="number">0.3235918</span> <span class="number">0.3692880</span> <span class="number">0.3784909</span> <span class="number">0.3935173</span> </div><div class="line">variable with highest stability (smallest rho value):	UBC</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">2</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">     GAPD      TPT1       UBB      NACA      CFL1     RPS13     TUBA6 </div><div class="line"><span class="number">0.1375298</span> <span class="number">0.1424519</span> <span class="number">0.1578360</span> <span class="number">0.1657364</span> <span class="number">0.1729069</span> <span class="number">0.1837057</span> <span class="number">0.1849021</span> </div><div class="line">     SUI1      ACTB     RPS23  FLJ20030      CLTC </div><div class="line"><span class="number">0.2065531</span> <span class="number">0.2131651</span> <span class="number">0.2188277</span> <span class="number">0.2359623</span> <span class="number">0.2447588</span> </div><div class="line">variable with highest stability (smallest rho value):	GAPD</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">3</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">     TPT1      NACA       UBB     RPS13      CFL1     TUBA6  FLJ20030 </div><div class="line"><span class="number">0.1108474</span> <span class="number">0.1299802</span> <span class="number">0.1356690</span> <span class="number">0.1411173</span> <span class="number">0.1474242</span> <span class="number">0.1532953</span> <span class="number">0.1583031</span> </div><div class="line">     SUI1      ACTB     RPS23      CLTC </div><div class="line"><span class="number">0.1586250</span> <span class="number">0.1682972</span> <span class="number">0.1686139</span> <span class="number">0.1926907</span> </div><div class="line">variable with highest stability (smallest rho value):	TPT1</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">4</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">       UBB      TUBA6       ACTB       CFL1      RPS13       SUI1       CLTC </div><div class="line"><span class="number">0.09656546</span> <span class="number">0.09674897</span> <span class="number">0.10753445</span> <span class="number">0.10830099</span> <span class="number">0.11801680</span> <span class="number">0.12612399</span> <span class="number">0.12773131</span> </div><div class="line">      NACA   FLJ20030      RPS23 </div><div class="line"><span class="number">0.13422958</span> <span class="number">0.14609897</span> <span class="number">0.16530522</span> </div><div class="line">variable with highest stability (smallest rho value):	UBB</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">5</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">     RPS13       SUI1      TUBA6       NACA   FLJ20030       CFL1       ACTB </div><div class="line"><span class="number">0.09085973</span> <span class="number">0.09647829</span> <span class="number">0.09943424</span> <span class="number">0.10288912</span> <span class="number">0.11097074</span> <span class="number">0.11428399</span> <span class="number">0.11495336</span> </div><div class="line">     RPS23       CLTC </div><div class="line"><span class="number">0.12635109</span> <span class="number">0.13286210</span> </div><div class="line">variable with highest stability (smallest rho value):	RPS13</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">6</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">      ACTB      TUBA6       CFL1   FLJ20030       NACA       CLTC       SUI1 </div><div class="line"><span class="number">0.09215478</span> <span class="number">0.09499893</span> <span class="number">0.09674032</span> <span class="number">0.10528784</span> <span class="number">0.10718604</span> <span class="number">0.10879846</span> <span class="number">0.11368091</span> </div><div class="line">     RPS23 </div><div class="line"><span class="number">0.13134766</span> </div><div class="line">variable with highest stability (smallest rho value):	ACTB</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">7</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">      SUI1       NACA   FLJ20030      RPS23      TUBA6       CFL1       CLTC </div><div class="line"><span class="number">0.08281504</span> <span class="number">0.08444905</span> <span class="number">0.08922236</span> <span class="number">0.09072667</span> <span class="number">0.10559279</span> <span class="number">0.10993755</span> <span class="number">0.13142181</span> </div><div class="line">variable with highest stability (smallest rho value):	SUI1</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">8</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">      NACA       CFL1      TUBA6   FLJ20030       CLTC      RPS23 </div><div class="line"><span class="number">0.08336046</span> <span class="number">0.08410148</span> <span class="number">0.09315528</span> <span class="number">0.09775742</span> <span class="number">0.10499056</span> <span class="number">0.10554332</span> </div><div class="line">variable with highest stability (smallest rho value):	NACA</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">9</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">      CFL1      TUBA6       CLTC   FLJ20030      RPS23 </div><div class="line"><span class="number">0.07222968</span> <span class="number">0.07722737</span> <span class="number">0.08440691</span> <span class="number">0.09831958</span> <span class="number">0.12735605</span> </div><div class="line">variable with highest stability (smallest rho value):	CFL1</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">10</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">  FLJ20030      TUBA6       CLTC      RPS23 </div><div class="line"><span class="number">0.08162006</span> <span class="number">0.08189011</span> <span class="number">0.10705192</span> <span class="number">0.11430674</span> </div><div class="line">variable with highest stability (smallest rho value):	FLJ20030</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">11</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">     TUBA6       CLTC      RPS23 </div><div class="line"><span class="number">0.06978897</span> <span class="number">0.08069582</span> <span class="number">0.13702726</span> </div><div class="line">variable with highest stability (smallest rho value):	TUBA6</div><div class="line"></div><div class="line"><span class="comment">###############################################################</span></div><div class="line"></div><div class="line">Step <span class="number">12</span>:</div><div class="line"></div><div class="line">stability values rho:</div><div class="line"></div><div class="line">     CLTC     RPS23 </div><div class="line"><span class="number">0.1199009</span> <span class="number">0.1245241</span> </div><div class="line">variable with highest stability (smallest rho value):	CLTC</div><div class="line"></div><div class="line">         <span class="number">1</span>          <span class="number">2</span>          <span class="number">3</span>          <span class="number">4</span>          <span class="number">5</span>          <span class="number">6</span>          <span class="number">7</span> </div><div class="line">     <span class="string">"UBC"</span>     <span class="string">"GAPD"</span>     <span class="string">"TPT1"</span>      <span class="string">"UBB"</span>    <span class="string">"RPS13"</span>     <span class="string">"ACTB"</span>     <span class="string">"SUI1"</span> </div><div class="line">         <span class="number">8</span>          <span class="number">9</span>         <span class="number">10</span>         <span class="number">11</span>         <span class="number">12</span> </div><div class="line">    <span class="string">"NACA"</span>     <span class="string">"CFL1"</span> <span class="string">"FLJ20030"</span>    <span class="string">"TUBA6"</span>     <span class="string">"CLTC"</span> </div><div class="line">Warning message:</div><div class="line">In .local(x, <span class="keyword">...</span>) : Argument <span class="string">'group'</span> is transformed to a factor vector.</div></pre></td></tr></table></figure>
<p>在Bladder数据集中，我们发现，排列前面的2个内参基因是HSPCB与RPS13，也就是Figure 1中的内容，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190925182918.png">

</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">grade &lt;- pData(Bladder)[,<span class="string">"Grade"</span>]</div><div class="line">selectHKs(Bladder, </div><div class="line">	group = grade, </div><div class="line">	log = <span class="literal">FALSE</span>, </div><div class="line">	trace = <span class="literal">FALSE</span>,</div><div class="line">	Symbols = featureNames(Bladder), </div><div class="line">	minNrHKs = <span class="number">13</span>,</div><div class="line">	method = <span class="string">"NormFinder"</span>)$ranking</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; grade &lt;- pData(Bladder)[,<span class="string">"Grade"</span>]</div><div class="line">&gt; selectHKs(Bladder, </div><div class="line">+           group = grade, </div><div class="line">+           log = <span class="literal">FALSE</span>, </div><div class="line">+           trace = <span class="literal">FALSE</span>,</div><div class="line">+           Symbols = featureNames(Bladder), </div><div class="line">+           minNrHKs = <span class="number">13</span>,</div><div class="line">+           method = <span class="string">"NormFinder"</span>)$ranking</div><div class="line">         <span class="number">1</span>          <span class="number">2</span>          <span class="number">3</span>          <span class="number">4</span>          <span class="number">5</span>          <span class="number">6</span>          <span class="number">7</span> </div><div class="line">   <span class="string">"HSPCB"</span>    <span class="string">"RPS13"</span>      <span class="string">"UBC"</span>    <span class="string">"RPS23"</span>    <span class="string">"ATP5B"</span>     <span class="string">"TEGT"</span>      <span class="string">"UBB"</span> </div><div class="line">         <span class="number">8</span>          <span class="number">9</span>         <span class="number">10</span>         <span class="number">11</span>         <span class="number">12</span>         <span class="number">13</span> </div><div class="line"><span class="string">"FLJ20030"</span>     <span class="string">"CFL1"</span>   <span class="string">"S100A6"</span>    <span class="string">"FLOT2"</span>     <span class="string">"ACTB"</span>     <span class="string">"TPT1"</span></div></pre></td></tr></table></figure>
<h2 id="使用多内参进行均一化">使用多内参进行均一化</h2>
<h3 id="单内参的δcq方法">单内参的ΔCq方法</h3>
<p>单内参均一化的方法就是，使用某个基因的Cq值减去内参的Cq值。<code>NormqPCR</code>中就有相关的计算函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">path &lt;- system.file(<span class="string">"exData"</span>, package = <span class="string">"NormqPCR"</span>)</div><div class="line">taqman.example &lt;- file.path(path, <span class="string">"example.txt"</span>)</div><div class="line">qPCR.example &lt;- file.path(path, <span class="string">"qPCR.example.txt"</span>)</div><div class="line">qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div><div class="line">hkgs&lt;-<span class="string">"Actb-Rn00667869_m1"</span></div><div class="line">qPCRBatch.norm &lt;- deltaCq(qPCRBatch = qPCRBatch.taqman, hkgs = hkgs, calc=<span class="string">"arith"</span>)</div><div class="line">head(exprs(qPCRBatch.taqman)) <span class="comment"># raw data</span></div><div class="line">head(exprs(qPCRBatch.norm)) <span class="comment"># Normlazation data</span></div></pre></td></tr></table></figure>
<p>这里我们使用了Actb作为内参，计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div></pre></td><td class="code"><pre><div class="line">&gt; head(exprs(qPCRBatch.taqman)) <span class="comment"># raw data</span></div><div class="line">                     fp1.day3.v fp2.day3.v fp5.day3.mia fp6.day3.mia</div><div class="line">Actb.Rn00667869_m1     <span class="number">20.92301</span>   <span class="number">21.00262</span>     <span class="number">20.88438</span>     <span class="number">20.38190</span></div><div class="line">Adipoq.Rn00595250_m1   <span class="number">20.93906</span>   <span class="number">20.88610</span>     <span class="number">23.81790</span>     <span class="number">22.92289</span></div><div class="line">Adrbk1.Rn00562822_m1         <span class="literal">NA</span>         <span class="literal">NA</span>     <span class="number">27.45101</span>     <span class="number">27.02446</span></div><div class="line">Agtrl1.Rn00580252_s1   <span class="number">25.82239</span>   <span class="number">26.03846</span>     <span class="number">27.28175</span>     <span class="number">26.06274</span></div><div class="line">Alpl.Rn00564931_m1     <span class="number">33.45495</span>   <span class="number">32.81128</span>     <span class="number">33.91955</span>     <span class="number">32.62145</span></div><div class="line">B2m.Rn00560865_m1      <span class="number">21.66457</span>   <span class="number">21.89334</span>     <span class="number">22.92485</span>     <span class="number">22.61651</span></div><div class="line">                     fp.3.day.3.v fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia</div><div class="line">Actb.Rn00667869_m1       <span class="number">20.94854</span>     <span class="number">21.48838</span>       <span class="number">20.97863</span>       <span class="number">21.10281</span></div><div class="line">Adipoq.Rn00595250_m1     <span class="number">20.76957</span>     <span class="number">20.92512</span>       <span class="number">23.43714</span>       <span class="number">23.83928</span></div><div class="line">Adrbk1.Rn00562822_m1           <span class="literal">NA</span>           <span class="literal">NA</span>       <span class="number">24.71573</span>       <span class="number">27.97638</span></div><div class="line">Agtrl1.Rn00580252_s1     <span class="number">26.16934</span>     <span class="number">25.91374</span>       <span class="number">25.77341</span>       <span class="number">26.44801</span></div><div class="line">Alpl.Rn00564931_m1       <span class="number">33.34335</span>     <span class="number">33.26128</span>       <span class="number">33.08863</span>       <span class="number">33.35799</span></div><div class="line">B2m.Rn00560865_m1        <span class="number">21.45406</span>     <span class="number">22.36598</span>       <span class="number">22.90620</span>       <span class="number">23.00608</span></div><div class="line">&gt; head(exprs(qPCRBatch.norm)) <span class="comment"># Normlazation data</span></div><div class="line">                     fp1.day3.v fp2.day3.v fp5.day3.mia fp6.day3.mia</div><div class="line">Actb.Rn00667869_m1     <span class="number">0.000000</span>   <span class="number">0.000000</span>     <span class="number">0.000000</span>     <span class="number">0.000000</span></div><div class="line">Adipoq.Rn00595250_m1   <span class="number">0.016052</span>  -<span class="number">0.116520</span>     <span class="number">2.933523</span>     <span class="number">2.540987</span></div><div class="line">Adrbk1.Rn00562822_m1         <span class="literal">NA</span>         <span class="literal">NA</span>     <span class="number">6.566628</span>     <span class="number">6.642561</span></div><div class="line">Agtrl1.Rn00580252_s1   <span class="number">4.899380</span>   <span class="number">5.035841</span>     <span class="number">6.397364</span>     <span class="number">5.680837</span></div><div class="line">Alpl.Rn00564931_m1    <span class="number">12.531942</span>  <span class="number">11.808657</span>    <span class="number">13.035166</span>    <span class="number">12.239549</span></div><div class="line">B2m.Rn00560865_m1      <span class="number">0.741558</span>   <span class="number">0.890717</span>     <span class="number">2.040470</span>     <span class="number">2.234605</span></div><div class="line">                     fp.3.day.3.v fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia</div><div class="line">Actb.Rn00667869_m1       <span class="number">0.000000</span>     <span class="number">0.000000</span>       <span class="number">0.000000</span>       <span class="number">0.000000</span></div><div class="line">Adipoq.Rn00595250_m1    -<span class="number">0.178971</span>    -<span class="number">0.563263</span>       <span class="number">2.458509</span>       <span class="number">2.736475</span></div><div class="line">Adrbk1.Rn00562822_m1           <span class="literal">NA</span>           <span class="literal">NA</span>       <span class="number">3.737100</span>       <span class="number">6.873568</span></div><div class="line">Agtrl1.Rn00580252_s1     <span class="number">5.220796</span>     <span class="number">4.425364</span>       <span class="number">4.794776</span>       <span class="number">5.345202</span></div><div class="line">Alpl.Rn00564931_m1      <span class="number">12.394802</span>    <span class="number">11.772896</span>      <span class="number">12.110000</span>      <span class="number">12.255186</span></div><div class="line">B2m.Rn00560865_m1        <span class="number">0.505516</span>     <span class="number">0.877598</span>       <span class="number">1.927563</span>       <span class="number">1.903269</span></div></pre></td></tr></table></figure>
<p>生成的结果是一个qPCRBatch对象，很多R包例如heatmap就兼容这类对象。</p>
<h3 id="多内参的δcq方法">多内参的ΔCq方法</h3>
<p>如果使用前面提到的geNorm或NormFinder算法找到了多个内参，此时想要用这多个内参进行均一化，那么就需要计算这几个内参的均值，构成一个“伪内参“(pseudo-housekeeper)进行均一化。现在我们使用三个内参，分别为GAPDH，beta-2-microglobulin和Beta-actin来进行均一化，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">hkgs&lt;-c(<span class="string">"Actb-Rn00667869_m1"</span>, <span class="string">"B2m-Rn00560865_m1"</span>, <span class="string">"Gapdh-Rn99999916_s1"</span>)</div><div class="line">qPCRBatch.norm &lt;- deltaCq(qPCRBatch = qPCRBatch.taqman, hkgs = hkgs, calc=<span class="string">"arith"</span>)</div><div class="line">head(exprs(qPCRBatch.norm))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">&gt; head(exprs(qPCRBatch.norm))</div><div class="line">                     fp1.day3.v fp2.day3.v fp5.day3.mia fp6.day3.mia</div><div class="line">Actb.Rn00667869_m1   -<span class="number">1.2998917</span> -<span class="number">1.2816963</span>    -<span class="number">1.380296</span>   -<span class="number">1.5106197</span></div><div class="line">Adipoq.Rn00595250_m1 -<span class="number">1.2838397</span> -<span class="number">1.3982163</span>     <span class="number">1.553227</span>    <span class="number">1.0303673</span></div><div class="line">Adrbk1.Rn00562822_m1         <span class="literal">NA</span>         <span class="literal">NA</span>     <span class="number">5.186332</span>    <span class="number">5.1319413</span></div><div class="line">Agtrl1.Rn00580252_s1  <span class="number">3.5994883</span>  <span class="number">3.7541447</span>     <span class="number">5.017068</span>    <span class="number">4.1702173</span></div><div class="line">Alpl.Rn00564931_m1   <span class="number">11.2320503</span> <span class="number">10.5269607</span>    <span class="number">11.654870</span>   <span class="number">10.7289293</span></div><div class="line">B2m.Rn00560865_m1    -<span class="number">0.5583337</span> -<span class="number">0.3909793</span>     <span class="number">0.660174</span>    <span class="number">0.7239853</span></div><div class="line">                     fp.3.day.3.v fp.4.day.3.v fp.7.day.3.mia fp.8.day.3.mia</div><div class="line">Actb.Rn00667869_m1     -<span class="number">1.1644617</span>   -<span class="number">1.1714227</span>      -<span class="number">1.323712</span>      -<span class="number">1.286277</span></div><div class="line">Adipoq.Rn00595250_m1   -<span class="number">1.3434327</span>   -<span class="number">1.7346857</span>       <span class="number">1.134797</span>       <span class="number">1.450198</span></div><div class="line">Adrbk1.Rn00562822_m1           <span class="literal">NA</span>           <span class="literal">NA</span>       <span class="number">2.413388</span>       <span class="number">5.587291</span></div><div class="line">Agtrl1.Rn00580252_s1    <span class="number">4.0563343</span>    <span class="number">3.2539413</span>       <span class="number">3.471064</span>       <span class="number">4.058925</span></div><div class="line">Alpl.Rn00564931_m1     <span class="number">11.2303403</span>   <span class="number">10.6014733</span>      <span class="number">10.786288</span>      <span class="number">10.968909</span></div><div class="line">B2m.Rn00560865_m1      -<span class="number">0.6589457</span>   -<span class="number">0.2938247</span>       <span class="number">0.603851</span>       <span class="number">0.616992</span></div></pre></td></tr></table></figure>
<h3 id="单内参2-δδcq方法">单内参2^-ΔΔCq方法</h3>
<p>使用2^-ΔΔCq方法哦可以计算两个样本类型之间的相对表达值。默认情况下，同一个样本的目标基因与其内参基因配对进行计算，此时的标准差就体现在不同基因与不同内参上。但是，如果参数中设定了<code>paired=FALSE</code>，那么标准差的计算公式就是<code>s=sqrt(s1^2+s2^2)</code>，这里的s1是目标基因的标准差，s2是内参基因的标准差，但是，如果当你要比较的基因与内参基因是来源于同一个样本时，不推荐这种方法，例如使用Taqman板的情况就是如此，但是，如果要考虑到完整性(completeness)以及要使用不同的生物学重复中获得的内参基因时，可以考虑使用这种计算方法，例如我们最初设计的内参数据并不好时，当采用事后检验(post hoc)时，就采用这种方法，或者是当NormqPCR用于处理那些单独板孔中的扩增数据时，也是采用这种方法。</p>
<p>现在我们读取原始数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">path &lt;- system.file(<span class="string">"exData"</span>, package = <span class="string">"NormqPCR"</span>)</div><div class="line">taqman.example &lt;- file.path(path, <span class="string">"example.txt"</span>)</div><div class="line">qPCR.example &lt;- file.path(path, <span class="string">"qPCR.example.txt"</span>)</div><div class="line">qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div></pre></td></tr></table></figure>
<p><code>deltaDeltaCq()</code>函数在进行计算时需要一个比较矩阵。在这个矩阵中，列是样本名（例如case或control），这与<code>limma</code>包中的数据类似。这个比较矩阵的列中只包括1或0这两个数据，用于指定分类变量，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">contM &lt;- cbind(c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>),c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>))</div><div class="line">colnames(contM) &lt;- c(<span class="string">"interestingPhenotype"</span>,<span class="string">"wildTypePhenotype"</span>)</div><div class="line">rownames(contM) &lt;- sampleNames(qPCRBatch.taqman)</div><div class="line">contM</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">&gt; contM</div><div class="line">               interestingPhenotype wildTypePhenotype</div><div class="line">fp1.day3.v                        <span class="number">0</span>                 <span class="number">1</span></div><div class="line">fp2.day3.v                        <span class="number">0</span>                 <span class="number">1</span></div><div class="line">fp5.day3.mia                      <span class="number">1</span>                 <span class="number">0</span></div><div class="line">fp6.day3.mia                      <span class="number">1</span>                 <span class="number">0</span></div><div class="line">fp.3.day.3.v                      <span class="number">0</span>                 <span class="number">1</span></div><div class="line">fp.4.day.3.v                      <span class="number">0</span>                 <span class="number">1</span></div><div class="line">fp.7.day.3.mia                    <span class="number">1</span>                 <span class="number">0</span></div><div class="line">fp.8.day.3.mia                    <span class="number">1</span>                 <span class="number">0</span></div></pre></td></tr></table></figure>
<p>此时，我们可以指定一个内参基因来进行均一化，然后再看一个某个基因在实验组(case)和对照组(control)的比值，计算结果包括以下信息（按列显示）：</p>
<ol style="list-style-type: decimal">
<li>每个基因的表达值；</li>
<li>实验组的ΔCq值：</li>
<li>Cq值的sd；</li>
<li>对照组的ΔCq值；</li>
<li>对照组的ΔCq值的sd；</li>
<li>2^-ΔΔCq值，这是对照组与实验组的差异；</li>
<li>和8. 分别是1和0时的sd值，如下所示：</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">hkg &lt;- <span class="string">"Actb-Rn00667869_m1"</span></div><div class="line">ddCq.taqman &lt;- deltaDeltaCq(qPCRBatch = qPCRBatch.taqman, </div><div class="line">                            maxNACase=<span class="number">1</span>, maxNAControl=<span class="number">1</span>,</div><div class="line">                            hkg=hkg, </div><div class="line">                            contrastM=contM, </div><div class="line">                            case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">                            control=<span class="string">"wildTypePhenotype"</span>, </div><div class="line">                            statCalc=<span class="string">"geom"</span>,</div><div class="line">                            hkgCalc=<span class="string">"arith"</span>)</div><div class="line">head(ddCq.taqman)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">&gt; head(ddCq.taqman)</div><div class="line">                    ID <span class="number">2</span>^-dCt.interestingPhenotype interestingPhenotype.sd</div><div class="line"><span class="number">1</span>   Actb.Rn00667869_m1                   <span class="number">1.000e+00</span>               <span class="number">0.000e+00</span></div><div class="line"><span class="number">2</span> Adipoq.Rn00595250_m1                   <span class="number">1.587e-01</span>               <span class="number">2.280e-02</span></div><div class="line"><span class="number">3</span> Adrbk1.Rn00562822_m1                   <span class="number">2.602e-02</span>               <span class="number">3.266e-02</span></div><div class="line"><span class="number">4</span> Agtrl1.Rn00580252_s1                   <span class="number">2.300e-02</span>               <span class="number">1.014e-02</span></div><div class="line"><span class="number">5</span>   Alpl.Rn00564931_m1                   <span class="number">1.892e-04</span>               <span class="number">4.770e-05</span></div><div class="line"><span class="number">6</span>    B2m.Rn00560865_m1                   <span class="number">2.464e-01</span>               <span class="number">2.498e-02</span></div><div class="line">  <span class="number">2</span>^-dCt.wildTypePhenotype wildTypePhenotype.sd              <span class="number">2</span>^-ddCt <span class="number">2</span>^-ddCt.min</div><div class="line"><span class="number">1</span>                <span class="number">1.000e+00</span>            <span class="number">0.000e+00</span> <span class="number">1</span>                             <span class="literal">NA</span></div><div class="line"><span class="number">2</span>                <span class="number">1.171e+00</span>            <span class="number">2.131e-01</span> <span class="number">0.135541545192243</span>             <span class="literal">NA</span></div><div class="line"><span class="number">3</span>                       <span class="literal">NA</span>                   <span class="literal">NA</span> +                             <span class="literal">NA</span></div><div class="line"><span class="number">4</span>                <span class="number">3.434e-02</span>            <span class="number">8.584e-03</span> <span class="number">0.669721905042939</span>             <span class="literal">NA</span></div><div class="line"><span class="number">5</span>                <span class="number">2.298e-04</span>            <span class="number">6.107e-05</span> <span class="number">0.823327272466571</span>             <span class="literal">NA</span></div><div class="line"><span class="number">6</span>                <span class="number">5.965e-01</span>            <span class="number">7.668e-02</span> <span class="number">0.413128242070071</span>             <span class="literal">NA</span></div><div class="line">  <span class="number">2</span>^-ddCt.max</div><div class="line"><span class="number">1</span>          <span class="literal">NA</span></div><div class="line"><span class="number">2</span>          <span class="literal">NA</span></div><div class="line"><span class="number">3</span>          <span class="literal">NA</span></div><div class="line"><span class="number">4</span>          <span class="literal">NA</span></div><div class="line"><span class="number">5</span>          <span class="literal">NA</span></div><div class="line"><span class="number">6</span>          <span class="literal">NA</span></div></pre></td></tr></table></figure>
<p>也可以使用每个单独的样本/孔(samples/wells)方法来计算平均值。这里我们单独计算sd，然后将它们汇总起来。因此，同一个样本的内参与检测值的本领信息氷是人干活的，这同时会增加变异，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">hkg &lt;- <span class="string">"Actb-Rn00667869_m1"</span></div><div class="line">ddCqAvg.taqman &lt;- deltaDeltaCq(qPCRBatch = qPCRBatch.taqman, </div><div class="line">                               maxNACase=<span class="number">1</span>, </div><div class="line">                               maxNAControl=<span class="number">1</span>,</div><div class="line">                               hkg=hkg, </div><div class="line">                               contrastM=contM, </div><div class="line">                               case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">                               control=<span class="string">"wildTypePhenotype"</span>, </div><div class="line">                               paired=<span class="literal">FALSE</span>, statCalc=<span class="string">"geom"</span>,</div><div class="line">                               hkgCalc=<span class="string">"arith"</span>)</div><div class="line"></div><div class="line">head(ddCqAvg.taqman)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div></pre></td><td class="code"><pre><div class="line">&gt; hkg &lt;- <span class="string">"Actb-Rn00667869_m1"</span></div><div class="line">&gt; ddCqAvg.taqman &lt;- deltaDeltaCq(qPCRBatch = qPCRBatch.taqman, </div><div class="line">+                                maxNACase=<span class="number">1</span>, </div><div class="line">+                                maxNAControl=<span class="number">1</span>,</div><div class="line">+                                hkg=hkg, </div><div class="line">+                                contrastM=contM, </div><div class="line">+                                case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">+                                control=<span class="string">"wildTypePhenotype"</span>, </div><div class="line">+                                paired=<span class="literal">FALSE</span>, statCalc=<span class="string">"geom"</span>,</div><div class="line">+                                hkgCalc=<span class="string">"arith"</span>)</div><div class="line">There were <span class="number">12</span> warnings (use warnings() to see them)</div><div class="line">&gt; </div><div class="line">&gt; head(ddCqAvg.taqman)</div><div class="line">                    ID <span class="number">2</span>^-dCt.interestingPhenotype interestingPhenotype.sd</div><div class="line"><span class="number">1</span>   Actb.Rn00667869_m1                   <span class="number">1.000e+00</span>               <span class="number">0.000e+00</span></div><div class="line"><span class="number">2</span> Adipoq.Rn00595250_m1                   <span class="number">1.587e-01</span>               <span class="number">2.280e-02</span></div><div class="line"><span class="number">3</span> Adrbk1.Rn00562822_m1                   <span class="number">2.602e-02</span>               <span class="number">3.266e-02</span></div><div class="line"><span class="number">4</span> Agtrl1.Rn00580252_s1                   <span class="number">2.300e-02</span>               <span class="number">1.014e-02</span></div><div class="line"><span class="number">5</span>   Alpl.Rn00564931_m1                   <span class="number">1.892e-04</span>               <span class="number">4.770e-05</span></div><div class="line"><span class="number">6</span>    B2m.Rn00560865_m1                   <span class="number">2.464e-01</span>               <span class="number">2.498e-02</span></div><div class="line">  <span class="number">2</span>^-dCt.wildTypePhenotype wildTypePhenotype.sd              <span class="number">2</span>^-ddCt <span class="number">2</span>^-ddCt.min</div><div class="line"><span class="number">1</span>                <span class="number">1.000e+00</span>            <span class="number">0.000e+00</span> <span class="number">1</span>                             <span class="literal">NA</span></div><div class="line"><span class="number">2</span>                <span class="number">1.171e+00</span>            <span class="number">2.131e-01</span> <span class="number">0.135541545192243</span>             <span class="literal">NA</span></div><div class="line"><span class="number">3</span>                       <span class="literal">NA</span>                   <span class="literal">NA</span> +                             <span class="literal">NA</span></div><div class="line"><span class="number">4</span>                <span class="number">3.434e-02</span>            <span class="number">8.584e-03</span> <span class="number">0.669721905042939</span>             <span class="literal">NA</span></div><div class="line"><span class="number">5</span>                <span class="number">2.298e-04</span>            <span class="number">6.107e-05</span> <span class="number">0.823327272466571</span>             <span class="literal">NA</span></div><div class="line"><span class="number">6</span>                <span class="number">5.965e-01</span>            <span class="number">7.668e-02</span> <span class="number">0.413128242070071</span>             <span class="literal">NA</span></div><div class="line">  <span class="number">2</span>^-ddCt.max</div><div class="line"><span class="number">1</span>          <span class="literal">NA</span></div><div class="line"><span class="number">2</span>          <span class="literal">NA</span></div><div class="line"><span class="number">3</span>          <span class="literal">NA</span></div><div class="line"><span class="number">4</span>          <span class="literal">NA</span></div><div class="line"><span class="number">5</span>          <span class="literal">NA</span></div><div class="line"><span class="number">6</span>          <span class="literal">NA</span></div></pre></td></tr></table></figure>
<h3 id="多内参2-δδcq方法">多内参2^-ΔΔCq方法</h3>
<p>如果想使用大于1个内参进行均一化，例如，我们想使用NormFinder/geNorm算法来进行均一化。我们就可以使用几何平均数(geometric mean)来构建一个<code>伪内参(pseudo-housekeeper)</code>来实现这一目的。现在我们使用GAPDH，beta-2-microglobulin和Beta-actin这三个内参来进行均一化，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div><div class="line">contM &lt;- cbind(c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>),c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>))</div><div class="line">colnames(contM) &lt;- c(<span class="string">"interestingPhenotype"</span>,<span class="string">"wildTypePhenotype"</span>)</div><div class="line">rownames(contM) &lt;- sampleNames(qPCRBatch.taqman)</div><div class="line">hkgs&lt;-c(<span class="string">"Actb-Rn00667869_m1"</span>, <span class="string">"B2m-Rn00560865_m1"</span>, <span class="string">"Gapdh-Rn99999916_s1"</span>)</div><div class="line">ddCq.gM.taqman &lt;- deltaDeltaCq(qPCRBatch = qPCRBatch.taqman,</div><div class="line">                               maxNACase=<span class="number">1</span>,</div><div class="line">                               maxNAControl=<span class="number">1</span>,</div><div class="line">                               hkgs=hkgs, </div><div class="line">                               contrastM=contM,</div><div class="line">                               case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">                               control=<span class="string">"wildTypePhenotype"</span>, </div><div class="line">                               statCalc=<span class="string">"arith"</span>,</div><div class="line">                               hkgCalc=<span class="string">"arith"</span>)</div><div class="line">head(ddCq.gM.taqman)</div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div></pre></td><td class="code"><pre><div class="line">&gt; qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div><div class="line">Warning message:</div><div class="line">In read.taqman(taqman.example) :</div><div class="line">  Incompatible phenoData object. Created a new one using sample name data derived from raw data.</div><div class="line"></div><div class="line">&gt; contM &lt;- cbind(c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>),c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>))</div><div class="line">&gt; colnames(contM) &lt;- c(<span class="string">"interestingPhenotype"</span>,<span class="string">"wildTypePhenotype"</span>)</div><div class="line">&gt; rownames(contM) &lt;- sampleNames(qPCRBatch.taqman)</div><div class="line">&gt; hkgs&lt;-c(<span class="string">"Actb-Rn00667869_m1"</span>, <span class="string">"B2m-Rn00560865_m1"</span>, <span class="string">"Gapdh-Rn99999916_s1"</span>)</div><div class="line">&gt; ddCq.gM.taqman &lt;- deltaDeltaCq(qPCRBatch = qPCRBatch.taqman,</div><div class="line">+                                maxNACase=<span class="number">1</span>,</div><div class="line">+                                maxNAControl=<span class="number">1</span>,</div><div class="line">+                                hkgs=hkgs, </div><div class="line">+                                contrastM=contM,</div><div class="line">+                                case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">+                                control=<span class="string">"wildTypePhenotype"</span>, </div><div class="line">+                                statCalc=<span class="string">"arith"</span>,</div><div class="line">+                                hkgCalc=<span class="string">"arith"</span>)</div><div class="line">There were <span class="number">12</span> warnings (use warnings() to see them)</div><div class="line">&gt; head(ddCq.gM.taqman)</div><div class="line">                    ID <span class="number">2</span>^-dCt.interestingPhenotype interestingPhenotype.sd</div><div class="line"><span class="number">1</span>   Actb.Rn00667869_m1                   <span class="number">2.594e+00</span>                 <span class="number">0.09819</span></div><div class="line"><span class="number">2</span> Adipoq.Rn00595250_m1                   <span class="number">4.083e-01</span>                 <span class="number">0.24929</span></div><div class="line"><span class="number">3</span> Adrbk1.Rn00562822_m1                   <span class="number">4.182e-02</span>                 <span class="number">1.45844</span></div><div class="line"><span class="number">4</span> Agtrl1.Rn00580252_s1                   <span class="number">5.520e-02</span>                 <span class="number">0.63719</span></div><div class="line"><span class="number">5</span>   Alpl.Rn00564931_m1                   <span class="number">4.767e-04</span>                 <span class="number">0.42589</span></div><div class="line"><span class="number">6</span>    B2m.Rn00560865_m1                   <span class="number">6.367e-01</span>                 <span class="number">0.05413</span></div><div class="line">  <span class="number">2</span>^-dCt.wildTypePhenotype wildTypePhenotype.sd              <span class="number">2</span>^-ddCt <span class="number">2</span>^-ddCt.min</div><div class="line"><span class="number">1</span>                <span class="number">2.345e+00</span>             <span class="number">0.071373</span> <span class="number">1.10638851325547</span>       <span class="number">1.034e+00</span></div><div class="line"><span class="number">2</span>                <span class="number">2.713e+00</span>             <span class="number">0.201905</span> <span class="number">0.150497255530234</span>      <span class="number">1.266e-01</span></div><div class="line"><span class="number">3</span>                       <span class="literal">NA</span>                   <span class="literal">NA</span> +                             <span class="literal">NA</span></div><div class="line"><span class="number">4</span>                <span class="number">7.878e-02</span>             <span class="number">0.333840</span> <span class="number">0.700597907024805</span>      <span class="number">4.505e-01</span></div><div class="line"><span class="number">5</span>                <span class="number">5.242e-04</span>             <span class="number">0.386280</span> <span class="number">0.909381199520663</span>      <span class="number">6.769e-01</span></div><div class="line"><span class="number">6</span>                <span class="number">1.390e+00</span>             <span class="number">0.163975</span> <span class="number">0.457939394245865</span>      <span class="number">4.411e-01</span></div><div class="line">  <span class="number">2</span>^-ddCt.max</div><div class="line"><span class="number">1</span>    <span class="number">1.184310</span></div><div class="line"><span class="number">2</span>    <span class="number">0.178884</span></div><div class="line"><span class="number">3</span>          <span class="literal">NA</span></div><div class="line"><span class="number">4</span>    <span class="number">1.089636</span></div><div class="line"><span class="number">5</span>    <span class="number">1.221662</span></div><div class="line"><span class="number">6</span>    <span class="number">0.475448</span></div></pre></td></tr></table></figure>
<p>我们也可以使用那些在样本间方差相同的内参来进行均一化，这个类似于前面提到的第二个<code>deltaDeltaCq</code>方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div><div class="line">contM &lt;- cbind(c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>),c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>))</div><div class="line">colnames(contM) &lt;- c(<span class="string">"interestingPhenotype"</span>,<span class="string">"wildTypePhenotype"</span>)</div><div class="line">rownames(contM) &lt;- sampleNames(qPCRBatch.taqman)</div><div class="line">hkgs&lt;-c(<span class="string">"Actb-Rn00667869_m1"</span>, <span class="string">"B2m-Rn00560865_m1"</span>, <span class="string">"Gapdh-Rn99999916_s1"</span>)</div><div class="line">ddAvgCq.gM.taqman &lt;-deltaDeltaCq(qPCRBatch = qPCRBatch.taqman, </div><div class="line">                                 maxNACase=<span class="number">1</span>, maxNAControl=<span class="number">1</span>,</div><div class="line">                                 hkgs=hkgs, </div><div class="line">                                 contrastM=contM, </div><div class="line">                                 case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">                                 control=<span class="string">"wildTypePhenotype"</span>,</div><div class="line">                                 paired=<span class="literal">FALSE</span>, </div><div class="line">                                 statCalc=<span class="string">"arith"</span>,</div><div class="line">                                 hkgCalc=<span class="string">"arith"</span>)</div><div class="line">head(ddAvgCq.gM.taqman)</div></pre></td></tr></table></figure>
<p>计算结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div></pre></td><td class="code"><pre><div class="line">&gt; qPCRBatch.taqman &lt;- read.taqman(taqman.example)</div><div class="line">Warning message:</div><div class="line">In read.taqman(taqman.example) :</div><div class="line">  Incompatible phenoData object. Created a new one using sample name data derived from raw data.</div><div class="line"></div><div class="line">&gt; contM &lt;- cbind(c(<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>),c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>))</div><div class="line">&gt; colnames(contM) &lt;- c(<span class="string">"interestingPhenotype"</span>,<span class="string">"wildTypePhenotype"</span>)</div><div class="line">&gt; rownames(contM) &lt;- sampleNames(qPCRBatch.taqman)</div><div class="line">&gt; hkgs&lt;-c(<span class="string">"Actb-Rn00667869_m1"</span>, <span class="string">"B2m-Rn00560865_m1"</span>, <span class="string">"Gapdh-Rn99999916_s1"</span>)</div><div class="line">&gt; ddAvgCq.gM.taqman &lt;-deltaDeltaCq(qPCRBatch = qPCRBatch.taqman, </div><div class="line">+                                  maxNACase=<span class="number">1</span>, maxNAControl=<span class="number">1</span>,</div><div class="line">+                                  hkgs=hkgs, </div><div class="line">+                                  contrastM=contM, </div><div class="line">+                                  case=<span class="string">"interestingPhenotype"</span>,</div><div class="line">+                                  control=<span class="string">"wildTypePhenotype"</span>,</div><div class="line">+                                  paired=<span class="literal">FALSE</span>, </div><div class="line">+                                  statCalc=<span class="string">"arith"</span>,</div><div class="line">+                                  hkgCalc=<span class="string">"arith"</span>)</div><div class="line">There were <span class="number">12</span> warnings (use warnings() to see them)</div><div class="line">&gt; head(ddAvgCq.gM.taqman)</div><div class="line">                    ID <span class="number">2</span>^-dCt.interestingPhenotype interestingPhenotype.sd</div><div class="line"><span class="number">1</span>   Actb.Rn00667869_m1                   <span class="number">2.594e+00</span>                  <span class="number">0.3849</span></div><div class="line"><span class="number">2</span> Adipoq.Rn00595250_m1                   <span class="number">4.083e-01</span>                  <span class="number">0.4822</span></div><div class="line"><span class="number">3</span> Adrbk1.Rn00562822_m1                   <span class="number">4.182e-02</span>                  <span class="number">1.4545</span></div><div class="line"><span class="number">4</span> Agtrl1.Rn00580252_s1                   <span class="number">5.520e-02</span>                  <span class="number">0.6905</span></div><div class="line"><span class="number">5</span>   Alpl.Rn00564931_m1                   <span class="number">4.767e-04</span>                  <span class="number">0.5846</span></div><div class="line"><span class="number">6</span>    B2m.Rn00560865_m1                   <span class="number">6.367e-01</span>                  <span class="number">0.2777</span></div><div class="line">  <span class="number">2</span>^-dCt.wildTypePhenotype wildTypePhenotype.sd              <span class="number">2</span>^-ddCt <span class="number">2</span>^-ddCt.min</div><div class="line"><span class="number">1</span>                <span class="number">2.345e+00</span>               <span class="number">0.3574</span> <span class="number">1.10638851325547</span>       <span class="number">8.473e-01</span></div><div class="line"><span class="number">2</span>                <span class="number">2.713e+00</span>               <span class="number">0.2495</span> <span class="number">0.150497255530234</span>      <span class="number">1.077e-01</span></div><div class="line"><span class="number">3</span>                       <span class="literal">NA</span>                   <span class="literal">NA</span> +                             <span class="literal">NA</span></div><div class="line"><span class="number">4</span>                <span class="number">7.878e-02</span>               <span class="number">0.2813</span> <span class="number">0.700597907024805</span>      <span class="number">4.341e-01</span></div><div class="line"><span class="number">5</span>                <span class="number">5.242e-04</span>               <span class="number">0.3689</span> <span class="number">0.909381199520663</span>      <span class="number">6.064e-01</span></div><div class="line"><span class="number">6</span>                <span class="number">1.390e+00</span>               <span class="number">0.4576</span> <span class="number">0.457939394245865</span>      <span class="number">3.778e-01</span></div><div class="line">  <span class="number">2</span>^-ddCt.max</div><div class="line"><span class="number">1</span>    <span class="number">1.444684</span></div><div class="line"><span class="number">2</span>    <span class="number">0.210221</span></div><div class="line"><span class="number">3</span>          <span class="literal">NA</span></div><div class="line"><span class="number">4</span>    <span class="number">1.130625</span></div><div class="line"><span class="number">5</span>    <span class="number">1.363762</span></div><div class="line"><span class="number">6</span>    <span class="number">0.555126</span></div></pre></td></tr></table></figure>
<h3 id="nrq值计算">NRQ值计算</h3>
<p>导入数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">path &lt;- system.file(<span class="string">"exData"</span>, package = <span class="string">"ReadqPCR"</span>)</div><div class="line">qPCR.example &lt;- file.path(path, <span class="string">"qPCR.example.txt"</span>)</div><div class="line">Cq.data &lt;- read.qPCR(qPCR.example)</div></pre></td></tr></table></figure>
<p>合并技术重复，并计算SD，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">Cq.data1 &lt;- combineTechRepsWithSD(Cq.data)</div></pre></td></tr></table></figure>
<p>加载数据集的效应(efficiencies)，并将它们添加到数据集中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">Effs &lt;- file.path(path, <span class="string">"Efficiencies.txt"</span>)</div><div class="line">Cq.effs &lt;- read.table(file = Effs, row.names = <span class="number">1</span>, header = <span class="literal">TRUE</span>)</div><div class="line">rownames(Cq.effs) &lt;- featureNames(Cq.data1)</div><div class="line">effs(Cq.data1) &lt;- as.matrix(Cq.effs[,<span class="string">"efficiency"</span>,drop = <span class="literal">FALSE</span>])</div><div class="line">se.effs(Cq.data1) &lt;- as.matrix(Cq.effs[,<span class="string">"SD.efficiency"</span>,drop = <span class="literal">FALSE</span>])</div></pre></td></tr></table></figure>
<p>计算数据集的NRQ(normalized relative quantities)，我们只考虑两个特征值作为内参基因，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">res &lt;- ComputeNRQs(Cq.data1, hkgs = c(<span class="string">"gene_az"</span>, <span class="string">"gene_gx"</span>))</div><div class="line"><span class="comment">## NRQs</span></div><div class="line">exprs(res)</div><div class="line"></div><div class="line"><span class="comment">## SD of NRQs</span></div><div class="line">se.exprs(res)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div><div class="line">58</div><div class="line">59</div><div class="line">60</div></pre></td><td class="code"><pre><div class="line">&gt; res &lt;- ComputeNRQs(Cq.data1, hkgs = c(<span class="string">"gene_az"</span>, <span class="string">"gene_gx"</span>))</div><div class="line">Warning message:</div><div class="line">In .local(qPCRBatch, <span class="keyword">...</span>) : This <span class="keyword">function</span> is still experimental!</div><div class="line">&gt; <span class="comment">## NRQs</span></div><div class="line">&gt; exprs(res)</div><div class="line">            caseA     caseB  controlA  controlB</div><div class="line">gene_ai <span class="number">1.9253072</span> <span class="number">1.3586729</span> <span class="number">0.6479659</span> <span class="number">0.8749479</span></div><div class="line">gene_az <span class="number">1.0567118</span> <span class="number">1.1438982</span> <span class="number">1.0331980</span> <span class="number">0.9134997</span></div><div class="line">gene_bc <span class="number">1.1024935</span> <span class="number">0.7193500</span> <span class="number">0.7030487</span> <span class="number">1.2140836</span></div><div class="line">gene_by <span class="number">1.5102316</span> <span class="number">0.9573047</span> <span class="number">0.7527082</span> <span class="number">1.6008850</span></div><div class="line">gene_dh <span class="number">1.2982037</span> <span class="number">1.0722522</span> <span class="number">0.9623335</span> <span class="number">0.9392871</span></div><div class="line">gene_dm <span class="number">0.6590246</span> <span class="number">1.1690720</span> <span class="number">1.2475372</span> <span class="number">0.9366210</span></div><div class="line">gene_dq <span class="number">0.7541955</span> <span class="number">0.7036408</span> <span class="number">0.8327917</span> <span class="number">1.6165326</span></div><div class="line">gene_dr <span class="number">2.2192305</span> <span class="number">1.0581211</span> <span class="number">0.7026411</span> <span class="number">0.6900584</span></div><div class="line">gene_eg <span class="number">0.9366671</span> <span class="number">0.5800339</span> <span class="number">0.8313720</span> <span class="number">1.1848856</span></div><div class="line">gene_er <span class="number">0.5269062</span> <span class="number">0.9375427</span> <span class="number">0.6953326</span> <span class="number">2.3195978</span></div><div class="line">gene_ev <span class="number">1.4622280</span> <span class="number">2.3457021</span> <span class="number">0.9038912</span> <span class="number">1.1454535</span></div><div class="line">gene_fr <span class="number">1.4954763</span> <span class="number">1.6200792</span> <span class="number">0.9641192</span> <span class="number">0.7295680</span></div><div class="line">gene_fw <span class="number">0.6944248</span> <span class="number">0.8051075</span> <span class="number">1.5698382</span> <span class="number">0.7978611</span></div><div class="line">gene_gx <span class="number">0.9463318</span> <span class="number">0.8742037</span> <span class="number">0.9678687</span> <span class="number">1.0946911</span></div><div class="line">gene_hl <span class="number">1.0009372</span> <span class="number">1.4015267</span> <span class="number">0.7683665</span> <span class="number">0.7713712</span></div><div class="line">gene_il <span class="number">1.4632019</span> <span class="number">1.2595559</span> <span class="number">0.7216891</span> <span class="number">0.9318860</span></div><div class="line">gene_iv <span class="number">1.7263335</span> <span class="number">1.2275001</span> <span class="number">1.5464212</span> <span class="number">0.8881605</span></div><div class="line">gene_jr <span class="number">0.8984351</span> <span class="number">0.9834026</span> <span class="number">0.8754813</span> <span class="number">0.6637941</span></div><div class="line">gene_jw <span class="number">1.4655948</span> <span class="number">0.9340184</span> <span class="number">1.0505200</span> <span class="number">1.5504136</span></div><div class="line">gene_qs <span class="number">0.6730225</span> <span class="number">0.7610418</span> <span class="number">1.0665938</span> <span class="number">3.5329891</span></div><div class="line">gene_qy <span class="number">0.5287127</span> <span class="number">1.5722670</span> <span class="number">1.0615326</span> <span class="number">3.3252907</span></div><div class="line">gene_rz <span class="number">0.8690600</span> <span class="number">1.5588299</span> <span class="number">0.7287288</span> <span class="number">1.4812753</span></div><div class="line">gene_sw <span class="number">0.5975288</span> <span class="number">1.2406438</span> <span class="number">0.6982954</span> <span class="number">1.6007333</span></div><div class="line">gene_vx <span class="number">0.6942254</span> <span class="number">0.7168408</span> <span class="number">2.0253177</span> <span class="number">1.3190943</span></div><div class="line">gene_xz <span class="number">0.7668030</span> <span class="number">1.0218209</span> <span class="number">0.6136038</span> <span class="number">1.6729352</span></div><div class="line"></div><div class="line">&gt; <span class="comment">## SD of NRQs</span></div><div class="line">&gt; se.exprs(res)</div><div class="line">            caseA     caseB  controlA  controlB</div><div class="line">gene_ai <span class="number">1.3996554</span> <span class="number">0.8787290</span> <span class="number">0.4855882</span> <span class="number">1.0034912</span></div><div class="line">gene_az <span class="number">0.6832730</span> <span class="number">0.7601966</span> <span class="number">0.8971054</span> <span class="number">0.6031927</span></div><div class="line">gene_bc <span class="number">0.7225348</span> <span class="number">0.4746146</span> <span class="number">0.9626570</span> <span class="number">1.0478385</span></div><div class="line">gene_by <span class="number">1.1522746</span> <span class="number">0.6116269</span> <span class="number">0.6088836</span> <span class="number">2.0409211</span></div><div class="line">gene_dh <span class="number">1.2483072</span> <span class="number">0.7889984</span> <span class="number">0.6165041</span> <span class="number">0.9767947</span></div><div class="line">gene_dm <span class="number">0.4711409</span> <span class="number">0.7780238</span> <span class="number">0.8476053</span> <span class="number">0.7294405</span></div><div class="line">gene_dq <span class="number">0.7023561</span> <span class="number">0.4849899</span> <span class="number">0.5813310</span> <span class="number">1.4067670</span></div><div class="line">gene_dr <span class="number">1.4407662</span> <span class="number">1.0804211</span> <span class="number">0.4543153</span> <span class="number">0.5149367</span></div><div class="line">gene_eg <span class="number">0.7355269</span> <span class="number">0.5497433</span> <span class="number">0.5588801</span> <span class="number">1.0938601</span></div><div class="line">gene_er <span class="number">0.4301195</span> <span class="number">0.6119514</span> <span class="number">0.4471454</span> <span class="number">1.5115897</span></div><div class="line">gene_ev <span class="number">1.0094209</span> <span class="number">2.4267114</span> <span class="number">0.6337126</span> <span class="number">0.7782519</span></div><div class="line">gene_fr <span class="number">1.6760391</span> <span class="number">1.1119157</span> <span class="number">0.6226081</span> <span class="number">0.5040967</span></div><div class="line">gene_fw <span class="number">0.5041070</span> <span class="number">0.9131565</span> <span class="number">1.1153268</span> <span class="number">0.7234551</span></div><div class="line">gene_gx <span class="number">0.6046042</span> <span class="number">0.9027816</span> <span class="number">0.6713914</span> <span class="number">1.7394961</span></div><div class="line">gene_hl <span class="number">0.7633174</span> <span class="number">0.9123997</span> <span class="number">1.0000329</span> <span class="number">0.5005813</span></div><div class="line">gene_il <span class="number">1.4621406</span> <span class="number">0.9540445</span> <span class="number">0.5678634</span> <span class="number">0.6067147</span></div><div class="line">gene_iv <span class="number">1.2668346</span> <span class="number">0.8039841</span> <span class="number">0.9995225</span> <span class="number">0.8171996</span></div><div class="line">gene_jr <span class="number">0.5749672</span> <span class="number">0.6786989</span> <span class="number">0.5595295</span> <span class="number">0.4405919</span></div><div class="line">gene_jw <span class="number">0.9626606</span> <span class="number">0.7890401</span> <span class="number">0.7194378</span> <span class="number">1.1512512</span></div><div class="line">gene_qs <span class="number">0.5830335</span> <span class="number">0.5309990</span> <span class="number">0.6828952</span> <span class="number">2.8253799</span></div><div class="line">gene_qy <span class="number">0.5947918</span> <span class="number">1.1294199</span> <span class="number">0.6794829</span> <span class="number">2.1713340</span></div><div class="line">gene_rz <span class="number">0.5846751</span> <span class="number">1.7926435</span> <span class="number">0.4911506</span> <span class="number">2.1424580</span></div><div class="line">gene_sw <span class="number">0.6284440</span> <span class="number">0.8062083</span> <span class="number">0.9638307</span> <span class="number">1.8398593</span></div><div class="line">gene_vx <span class="number">0.5285361</span> <span class="number">1.0126959</span> <span class="number">1.3861226</span> <span class="number">0.8683886</span></div><div class="line">gene_xz <span class="number">0.5231477</span> <span class="number">0.9270275</span> <span class="number">0.3972901</span> <span class="number">1.3643840</span></div></pre></td></tr></table></figure>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>NormqPCR: Functions for normalisation of RT-qPCR data</li>
<li></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/28/Experiment/WB实验笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/28/Experiment/WB实验笔记/" itemprop="url">Western Blot笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-28T12:00:00+08:00">
                2019-09-28
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物实验笔记/" itemprop="url" rel="index">
                    <span itemprop="name">生物实验笔记</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  4,416
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  16
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="试剂配置">试剂配置</h2>
<h3 id="电泳液10x">电泳液(10X)</h3>
<table>
<thead>
<tr class="header">
<th></th>
<th>称量</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Tris</td>
<td>30.3 g</td>
</tr>
<tr class="even">
<td>Glycine</td>
<td>144 g</td>
</tr>
<tr class="odd">
<td>SDS</td>
<td>10g</td>
</tr>
</tbody>
</table>
<h3 id="转膜液10x">转膜液(10X)</h3>
<table>
<thead>
<tr class="header">
<th>成分</th>
<th>称量</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Tris</td>
<td>58g</td>
</tr>
<tr class="even">
<td>Glycine</td>
<td>29g</td>
</tr>
<tr class="odd">
<td>SDS（电泳级）</td>
<td>3.8g</td>
</tr>
</tbody>
</table>
<h3 id="tbs10x">TBS(10X)</h3>
<table>
<thead>
<tr class="header">
<th>成分</th>
<th>称量</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>NaCl</td>
<td>80g</td>
</tr>
<tr class="even">
<td>1M Tris-HCl((PH8.0)</td>
<td>24g</td>
</tr>
</tbody>
</table>
<p>使用时，取100mL TBS(10X)，加入900mL去离子水，再加入1mL 的Twen20（Sigma，货号：T2700-1L）。</p>
<h2 id="其它试剂">其它试剂</h2>
<h3 id="封闭液">封闭液</h3>
<p>5%脱脂奶（BD:232100），或索莱宝（货号D8340）</p>
<p>2.5g脱脂奶粉加入20ml的TBST Wash Buffer中，搅拌溶解，4℃或-20℃保存。</p>
<p>封闭液也可以直接使用商业公司的产品，例如碧云天的QuickBlock封闭液，货号：P0228</p>
<h3 id="一抗二抗稀释液">一抗二抗稀释液</h3>
<p>一抗稀释液：碧云天Western一抗稀释液（货号：P0023A）</p>
<p>二抗稀释液：碧云天Western二抗稀释液（货号：P0023D）</p>
<p>或用5%脱脂奶粉</p>
<h3 id="常用内参一抗与二抗">常用内参一抗与二抗</h3>
<p>一抗：鼠源小鼠actin抗体（碧云天，货号：AA128）</p>
<p>一抗：兔源小鼠Gapdh抗体（碧云天，货号：AF1186）</p>
<p>兔抗鼠Vinculin(CST，货号：13901S)</p>
<p>兔抗鼠beta-tubulin(CST，货号：15115S)</p>
<p>羊抗兔二抗（鼎国，货号：SA00001-2）</p>
<p>羊抗小鼠二抗（碧云天：货号：A0216）</p>
<p>羊抗兔二抗（CST，货号：7074S）。</p>
<p><strong>(4)</strong> 蛋白Marker</p>
<p>ThermoFisher protein Marker（货号：26619）</p>
<h3 id="其它试剂-1">其它试剂</h3>
<p>BCA蛋白浓度测定试剂盒（碧云天，货号：P0010） Twen20（Sigma，货号：T2700-1L）。</p>
<h2 id="实验步骤">实验步骤</h2>
<h3 id="一样本裂解">（一）样本裂解</h3>
<ol style="list-style-type: decimal">
<li>将2×106的细胞悬液离心后，除去上清，加入RIPA溶液40µL（根据细胞的数目而定），反复吹打，随后放冰上静置10分钟；（注：RIPA是一种传统的细胞组织快速裂解液。成分为 ①50 mMTris-HCl(pH 7.4)② 150 mM NaCl③1% NP-40）；</li>
<li>12000g，离心10min，4℃，取上清，检测蛋白含量。</li>
</ol>
<h3 id="二蛋白浓度测定bca法">（二）蛋白浓度测定(BCA法)</h3>
<p>试剂盒：碧云天（BCA蛋白浓度测定试剂盒，货号：P0010）或BCA蛋白浓度测定试剂盒(增强型，货号：P0012) ①BCA试剂A②BCA试剂B③蛋白标准品（5mg/mL BSA，BSA为牛血清白蛋白，bovine serum albumin）。BCA试剂A和B室温保存，蛋白标准于-20℃冻存（提前可以配置为梯度BSA溶液，于负20度保存）；④5XSDS变性剂（碧云天，货号：P0015）</p>
<p><strong>检测原理</strong> 碱性条件下，蛋白将Cu2+还原为Cu+， Cu+与BCA试剂形成紫颜色的络合物，测定其在562nm处的吸收值，并与标准曲线对比，即可计算待测蛋白的浓度。</p>
<p>检测流程：</p>
<ol style="list-style-type: decimal">
<li>根据样品数量，按试剂A：试剂B=50：1的比例进行稀释（检测的时候，此工作的体积为180µL，样品以及标准品的体积为20µL，因此要按照这个体积以及相当样品数来进行稀释），BCA工作液室温24小时内稳定（标准品的梯度为8个，样品是N个，则BCA工作液的体积一般是200x(8+N+1)µL）。注：根据试剂盒的说明进行操作，例如我使用的P0010试剂盒，里面是要求加200µL的反应液。</li>
<li>取提前配置好的蛋白标准品梯度溶液，即0µg/µL（编号1），0.025µg/µL（编号2），0.05µg/µL（编号3），0.1µg/µL（编号4），0.2µg/µL（编号5），0.3µg/µL（编号6），0.4µg/µL（编号7），0.5µg/µL（编号8）。</li>
<li>在96孔板相应的孔中加入200µL的BCA工作液，然后将编号为1到8的标准品加入到这些孔中，体积为20µL；</li>
<li>将样品稀释10或20倍，然后将20µL的样品加入到样品孔中（注：在这一步的操作中，通常是先在96孔的孔中提前加入9μL或19μL的双蒸水，然后再加入1μL的样本，最后加入200μL的反应液）；</li>
<li>在37℃条件下放置30分钟；</li>
<li>测定OD562nm处的吸光度，根据标准曲线计算出蛋白浓度，这一步注意，通常我们是使用<code>y = ax+b</code>来计算的，也就是x轴是吸光度，y轴是蛋白浓度，计算后再乘以稀释倍数即可。</li>
</ol>
<h3 id="三蛋白变性">（三）蛋白变性</h3>
<ol style="list-style-type: decimal">
<li>将准备上样的所有蛋白样品调至等浓度（一般上样量最多是20µg，体积是10µL，因此蛋白的终浓度控制在2µg/µL，但根据Biorad公司的说明，1.5mm的胶，大梳子每个孔最多上样66uL，但推荐最大上样是50μL，小梳子最多一个孔上样40μL），蛋白与5×SDS－PAGE Loading Buffer以4:1 混匀（体积比），1×loading buffer补充至等体积，SDS要在不超过37度的水浴中进行融化，用完放到负20度即可。</li>
</ol>
<p>建议：蛋白样品在加入SDS之前最好统一稀释至浓度为2.5µg/uL（因为最终上样体积为10uL，上样蛋白为20ug，其中在变性前，1次实验的蛋白样本为8µL，需要加入含有5XSDS的Loading Buffer为2uL，因此蛋白浓度最终就是2µg/µL）。每次制备的样品需要按3次实验进行分装，即每个试管中装24uL，每支的浓度为2.5µg/µL，使用的时候，直接加6uL的Loading Buffer，一次上样为10uL;如果细胞数量较多，蛋白浓度较大，可以将5XSDS稀释为1XSDS,并对蛋白样本进行稀释后再变性。</p>
<ol start="2" style="list-style-type: decimal">
<li>100℃干热或水浴（封口）10min充分变性，12000g离心30s后，置于冰上或4℃保存。</li>
</ol>
<h3 id="四sds-page胶配置">（四）SDS-PAGE胶配置</h3>
<ol style="list-style-type: decimal">
<li>清洗玻璃板：根据实验需要选择一定厚度的玻璃板 （1.0mm，1.5mm），注意一定要将玻璃板洗净（洗洁精），最后用无水乙醇冲洗，与胶接触的一面向上置于干净的纸巾晾干或用电吹风吹干。</li>
<li>装置玻璃板：将两块玻璃板装好，低的玻璃板朝内，注意玻璃底部平整，防止漏胶，四个螺丝钉一起拧紧，防止玻璃破裂，再固定于制胶板上，可同时做一块胶（对面有塑料挡板）或两块胶，在加入分离胶之前，可以加入无水乙醇，观察一下是否漏液；如使用Bio-Rad, 将两块玻璃板装好，低的玻璃板朝外，夹好玻璃。</li>
<li>配分离胶：分离胶的其它成分可事先配好（事先配置好的意思就是，先把除了10% AP（过硫酸铵）及TEMED外的其它成分加好，在实验之前再加入AP与TEMED），避光存放于4℃，可至少存放1个月，临用前取出室温平衡（否则凝胶过程产生的热量会使低温时溶解于储存液中的气体析出而导致气泡，有条件者可真空抽吸3min），加入10%AP及TEMED即可，也可临时按比例配制，注意胶用枪吹打混匀（鉴定的Rab32蛋白是按照10%SDS-PAGE的胶进行配置，体积为5mL，不同的蛋白分子量需要不同浓度的胶，通常是分子量越大的蛋白，需要浓度越低，同时胶也越脆弱）注：试剂盒的A盒（即带有TEMED的那个试剂盒）放4℃保存；B盒常温下保存。</li>
<li>灌胶：将凝胶溶液平缓地注入两层玻璃极中，灌入分离胶后（可预先插入梳子，梳子底下约0.5cm处标记预备配制的分离胶高度，拔出梳子）应立即封胶，可用在液面上小心注入一层无水乙醇，以阻止氧气进入凝胶溶液中，然后静置，封胶后切记勿动，分离胶的上层距离低的玻璃板的最上层距离约为1.5~2cm，配置分离胶的体积约为5mL，用掉约4.5mL。</li>
<li>待分离胶凝后（30min左右出现一条明显的分离线）将无水乙醇倒掉，用滤纸或纸巾吸干。</li>
<li>配浓缩胶： 同前按比例配制浓缩胶，一个板子约2mL，以连续平稳的液流注入凝胶溶液，然后小心插入梳子（不同厚度的玻璃板匹配不同厚度的梳子），注意不得在齿尖留有气泡，静置1h以保证完全聚合。</li>
</ol>
<p><strong>(5)</strong> <strong>电泳</strong></p>
<p>电泳：上样前将胶板下的气泡赶走，用注射器吸取电泳缓冲液冲孔，可根据实验需要设定阳性对照（即肯定可以表达你的目的蛋白的组织或者细胞，同时可以提示你一抗的质量）或阴性对照（即肯定不会表达你的目的蛋白的组织或者细胞），样品两侧的泳道用等体积的1×loading buffer上样，Marker（Ferments）也用1×loading buffer调整至与样品等体积，上样完毕后宜尽快电泳（注意电极正负）防止扩散，以初始电压为80V进行稳压电泳，当溴芬兰泳动至分离胶时，这个过程约为20min~30min，改电压为120 V继续稳压电泳，最终的蓝色缓冲液条件不要超过底部支架的绿色部分，电泳的过程约为60min~90min，整个过程要注意电泳的条件位置，不同分子量的蛋白电泳的时间与位置不同。</p>
<p><strong>(6)</strong> <strong>转膜</strong></p>
<ol style="list-style-type: decimal">
<li>取胶：在目的条带充分分离后（参考Marker的分离效果），或电泳至溴酚染料前沿下至凝胶末端处可结束电泳，将胶卸下，保留30-100KD或分子量范围更广些的胶（以便以后杂其他感兴趣的蛋白），在转移液（转移液即1X的transfer buffer，要现用现配）中浸泡。</li>
<li>转膜：转膜液是800mL的1X转膜液加上200mL的甲醇，根据胶的大小剪裁稍大的PVDF膜，在左上角剪角，甲醇中充分浸泡5min（活化PVDF膜）后转入转移液中，去掉PVDF膜的上下两层塑料封纸，将厚的白色滤纸及海绵也浸入转移液中。15min-30min后按如下顺序安装转移装置（如图1所示）：</li>
</ol>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191009124330.png">

</div>
<p>a.平放底部电极（负极），放一张海绵垫片。 b.在海绵垫片上放置1张用转移缓冲液浸泡过的滤纸，然后用一玻璃棒作滚筒以挤出所有气泡，必要时可滴加转膜液润湿。 c.取出浸在转膜液中的凝胶平放于滤纸上，排除所有气泡。 d.把PVDF膜放在聚丙烯酰胺凝胶上，在硝酸纤维素滤膜与聚丙烯酰胺凝胶之间不留有气泡。 e. 把最后1张滤纸放在PVDF膜上方，同样须确保不留气泡。 f.放上另一张海绵垫片，盖上阳极板，夹紧。保证对凝胶有一定的压力。加入预冷的转膜液及内置的冰盒，盖好盖子，连接电源，（注意正负极）电转移40min, 300mA（根据实验需要摸索最佳转膜条件）。电转过程中，尽量保持低温环境，可放在冰水里转膜。电转结束后，断开电源，拆卸转移装置，逐一掀去各层。将凝胶转移至盛有考马斯亮蓝染液的托盘中，进行染色，可检查蛋白质转移是否完全。要观察膜上蛋白的情况可将膜在丽春红中染色，然后蒸馏水冲洗。（电泳上槽液及转膜液可回收）。</p>
<ol start="3" style="list-style-type: decimal">
<li>如果不染色，可以跳过后面几步，直接进行封闭。染色和脱色 ：电泳完毕可通过染色和脱色评定其结果的优劣.对凝胶中分离成不同条带的蛋白进行检测.此外,根据不同的研究目的,也可将凝胶电印迹后的胶或膜染色。</li>
<li>考马斯亮蓝染色：将凝胶取出放入塑料盒中倒入少量的染色液,以浸没凝胶为宜,在摇床上震荡,直到凝胶上出现明显的条带为止,一般5-6小时或者4℃过夜;然后倾去染色液，用脱色液洗涤震荡,其间要不断换脱色液,直至脱色液颜色稍清亮.,凝胶背景清楚为止。（染色液可回收）</li>
<li>丽春红染色：详细方法见说明书。</li>
</ol>
<p><strong>(7)</strong> <strong>膜的封闭</strong></p>
<p>在进行抗体杂交之前，需要先对转印膜进行封闭，防止免疫试剂的非特异性吸附。封闭一般采用异源性蛋白质或去污剂，常用的有0.2%Tween-20，20%BSA，10%马血清以及5% Non-fat milk等，至于选择哪一类封闭液，首先应考虑与检测试剂相适应，如采用葡萄球蛋白A（SPA）作用检测试剂，就不能用全血清封闭，其次是尽可能使非特异着色背静浅，本实验室常用封闭液5%Non-fat milk。封闭过程如下所示（与胶接触的PVDF膜的一面是正面，要在正面的左上角剪下一小块，作为区分正反面的依据，前面剪过的则忽略此步）：</p>
<ol style="list-style-type: decimal">
<li>洗转印膜：用1xTBS-T室温漂洗3次×10min（漂洗1次也行），以尽量洗去转印膜上的SDS，防止影响后面的抗体结合。</li>
<li>取漂洗的转印膜，放入5% Non-fat milk的封闭液内，摇床震动，室温封闭2h，也可在4℃过夜， 或使用商用封闭液。</li>
<li>封闭后，使用1×TBS-T室温漂洗3次×10min。</li>
</ol>
<p><strong>(8)</strong> <strong>抗体杂交</strong></p>
<p>抗体表面主要采用间接法。即先加入未标记特异性抗体（Ab1，即一抗）与膜上抗原结合，再加入标记的抗抗体（Ab2，即二抗）进行杂交检测，标记Ab2 物质有放射性核素酶及生物素等。 （抗体稀释倍数需要摸索通常一抗浓度1：2000到1:1000，具体的稀释浓度可以参考一抗的说明书，二抗浓度1:1000）。杂交过程如下所示：</p>
<ol style="list-style-type: decimal">
<li>封闭后的杂交膜放入塑料盒或杂交袋中，加入抗体稀释液稀释的Ab1（宜设立内参照，beta-actin，beta-tubulin等），4℃孵育过夜或室温（22-25℃）摇动孵育2h。（一抗可回收3-5次）；</li>
<li>1×TBS-T洗膜3次×10min；</li>
<li>抗体稀释液稀释的Ab2（抗兔-HRP或抗鼠-HRP，根据一抗种属来源决定，注意避光）室温2h，随后洗膜3×10min。</li>
<li>ECL显色：将显色剂A、B按1：1 混匀，将杂交后PVDF膜放入显色盒中，加上混合好的显色液，约1-5min用纸巾吸去印迹膜边缘或者边角部分多余的显色液，在暗室中使胶片曝光1-5min。注：不同的显色仪参数不同，需要阅读一下说明书，找到最优曝光条件。</li>
</ol>
<h2 id="wb常见问题">WB常见问题</h2>
<ol style="list-style-type: decimal">
<li>如果我想观察2个蛋白，但这两个蛋白的分子量离得很近，怕剪膜剪坏，如何操作？</li>
</ol>
<p>答：通常这种情况，可以先孵A蛋白，然后将A蛋白的抗体洗掉，再进行封闭操作，随后孵B蛋白的抗体。</p>
<p>洗抗体通常会使用抗体去除剂来进行操作，这样通常可以重复利用膜3-5次。抗体去除剂通常有中性，酸性，碱性，要根据自己的目的来进行选择，官网通常就有，例如碧云天的官网<a href="https://www.beyotime.com/support/stripping-buffer.htm" target="_blank" rel="external">《碧云天生产的各种Western一抗二抗去除液的主要特点、差异和选择》</a>。</p>
<ol start="2" style="list-style-type: decimal">
<li>如果我想用5%的脱脂奶粉来配制一抗和二抗，但由于担心不好保存怎么办？</li>
</ol>
<p>答：如果要用奶粉来配制抗体，通常会加入叠氮化钠。实验室配制的叠氮钠是10%的储备液，常温放置备用，配制一抗时每1mL加入叠氮化钠10μL,能使一抗保存时间一个月左右，使用叠氮化钠时一定要小心，此试剂有毒。</p>
<ol start="3" style="list-style-type: decimal">
<li>转膜时间与蛋白分子量有没有关系？</li>
</ol>
<p>答：通常如果使用Bio-Rad的标准湿式转膜装置，可以设定转膜电流为300-400mA，转膜时间为30-60分钟。也可以在15-20mA转膜过夜。具体的转膜时间要根据目的蛋白的大小而定，目的蛋白的分子量越大， 需要的转膜时间越长，目的蛋白的分子量越小，需要的转膜时间越短。</p>
<ol start="4" style="list-style-type: decimal">
<li>转膜的效率与分子量大小有关系没？</li>
</ol>
<p>答：如果是大蛋白（大于100 kDa），需要注意以下几点：</p>
<ol style="list-style-type: decimal">
<li>对于大蛋白而言，其在凝胶电泳分离迁移较慢，而从凝胶转出也非常慢，因此对于这种大分子量蛋白应该用低浓度的凝胶，8% 或更低，但因低浓度的胶非常易碎，所以操作时需十分小心。</li>
<li>大蛋白易在凝胶里形成沉淀，从而影响转膜。转膜时在电转移缓冲液加入SDS 至终浓度0.1%，避免出现这种情况。甲醇易使SDS 从蛋白上脱失，因此相应降低甲醇的浓度至10% 或更低，以防止蛋白沉淀。</li>
<li>降低电转移缓冲液中甲醇的比例，这样可以促进凝胶的膨胀，更易于大蛋白的转出。</li>
<li>只有使用硝酸纤维素膜时，甲醇才是必需的。如果是PVDF 膜，甲醇可以不必加入电转移缓冲液中，但转膜前PVDF 需用甲醇活化。</li>
<li>选择湿式，4℃ 转膜过夜，以取代半干式转膜。</li>
</ol>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://mp.weixin.qq.com/s?__biz=MzI0OTE4NjA5Mg==&amp;mid=2652694527&amp;idx=1&amp;sn=5dd241334f139850551843a8727376b1&amp;chksm=f27d15bcc50a9caa9c2b905b6e876a4a3d4171fccfbc92691b64a428db2d66571b6925c6cb9d&amp;scene=21#wechat_redirect" target="_blank" rel="external">Western Blot实验——蛋白转膜和染色篇</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/24/ggplot2/R语言拼图/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/24/ggplot2/R语言拼图/" itemprop="url">R语言拼图</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-24T12:00:00+08:00">
                2019-09-24
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/ggplot2/" itemprop="url" rel="index">
                    <span itemprop="name">ggplot2</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  0
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/10/BioTools/fgsea/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/10/BioTools/fgsea/" itemprop="url">fgsea包做GSEA分析</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-10T12:00:00+08:00">
                2019-09-10
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,259
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="gsea图">GSEA图</h2>
<p>这篇教程翻译自生信博客DAVE TANGE’S BLOG(https://davetang.org/muse/)上的一篇教程，参考资料中已经列出了地址。</p>
<p><code>fgsea</code>这个包用于做GSEA分析，先来看一下使用这个包做的图，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911114456.png">

</div>
<p>现在简单解释一下这个图形：</p>
<p>x轴——排序后的基因列表<code>L</code>位置对应的坐标，也就是我们自己通过RNA-seq，芯片，qPCR等手段获得的基因表达值倍数变化，或p值排序，总之，这是一个有序列表。</p>
<p>垂直的黑色细胞——上图中类似条形码的图形，这是指的是某一个基因集<code>S</code>中基因对应于<code>L</code>基因中，的位置，在上图中，这个基因集是细胞周期(Cell Cycle)，明天看到<code>S</code>中的成员在<code>L</code>的左侧比较密集。</p>
<p>y轴——富集分布，从上面我们可以看到，细胞周期(Cell Cycle)这个基因集在左侧富集，也就是绿色曲线表示的位置。</p>
<h2 id="fgsea使用">fgsea使用</h2>
<h3 id="安装">安装</h3>
<p>先安装<code>fgsea</code>包，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">install.packages(<span class="string">"BiocManager"</span>)</div><div class="line">BiocManager::install(<span class="string">"fgsea"</span>)</div><div class="line"><span class="keyword">library</span>(fgsea)</div></pre></td></tr></table></figure>
<h3 id="数据">数据</h3>
<p><code>fgsea</code>包中内置的有数据集<code>examplePathways</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(fgsea)</div><div class="line"></div><div class="line">data(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># examplePathways是从'reactome.db'包中</span></div><div class="line"><span class="comment"># 提取的信息，并以列表的形式存了这些信息，</span></div><div class="line"><span class="comment"># 这些信息主要是小鼠的基因</span></div><div class="line">help(<span class="string">"examplePathways"</span>)</div><div class="line"></div><div class="line"><span class="comment"># 这个数据集是一个列表</span></div><div class="line">class(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># 一共有1,457 "通路"</span></div><div class="line">length(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># 第1列含有Meiotic_Synapsis pathway的EntrezID</span></div><div class="line">examplePathways[<span class="number">1</span>]</div><div class="line"></div><div class="line"><span class="comment"># exampleRanks储存的是排序信息</span></div><div class="line">data(exampleRanks)</div><div class="line"></div><div class="line"><span class="comment"># exampleRnak是一个数字型的向量</span></div><div class="line">class(exampleRanks)</div><div class="line"></div><div class="line"><span class="comment"># exampleRanks中的向量名称则是对称的Entrez ID</span></div><div class="line">head(exampleRanks)</div><div class="line">tail(unname(exampleRanks))</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(fgsea)</div><div class="line">&gt; </div><div class="line">&gt; data(examplePathways)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># examplePathways是从'reactome.db'包中</span></div><div class="line"><span class="comment"># 提取的信息，并以列表的形式</span></div><div class="line">&gt; <span class="comment"># 储存了这些信息，这些信息主要是小鼠的基因</span></div><div class="line">&gt; help(<span class="string">"examplePathways"</span>)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 这个数据集是一个列表</span></div><div class="line">&gt; class(examplePathways)</div><div class="line">[<span class="number">1</span>] <span class="string">"list"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 一共有1,457 "通路"</span></div><div class="line">&gt; length(examplePathways)</div><div class="line">[<span class="number">1</span>] <span class="number">1457</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 第1列含有Meiotic_Synapsis pathway的EntrezID</span></div><div class="line">&gt; examplePathways[<span class="number">1</span>]</div><div class="line">$`1221633_Meiotic_Synapsis`</div><div class="line"> [<span class="number">1</span>] <span class="string">"12189"</span>     <span class="string">"13006"</span>     <span class="string">"15077"</span>     <span class="string">"15078"</span>     <span class="string">"15270"</span>     <span class="string">"15512"</span>    </div><div class="line"> [<span class="number">7</span>] <span class="string">"16905"</span>     <span class="string">"16906"</span>     <span class="string">"19357"</span>     <span class="string">"20842"</span>     <span class="string">"20843"</span>     <span class="string">"20957"</span>    </div><div class="line">[<span class="number">13</span>] <span class="string">"20962"</span>     <span class="string">"21749"</span>     <span class="string">"21750"</span>     <span class="string">"22196"</span>     <span class="string">"23856"</span>     <span class="string">"24061"</span>    </div><div class="line">[<span class="number">19</span>] <span class="string">"28113"</span>     <span class="string">"50878"</span>     <span class="string">"56739"</span>     <span class="string">"57321"</span>     <span class="string">"64009"</span>     <span class="string">"66654"</span>    </div><div class="line">[<span class="number">25</span>] <span class="string">"69386"</span>     <span class="string">"71846"</span>     <span class="string">"74075"</span>     <span class="string">"77053"</span>     <span class="string">"94244"</span>     <span class="string">"97114"</span>    </div><div class="line">[<span class="number">31</span>] <span class="string">"97122"</span>     <span class="string">"97908"</span>     <span class="string">"101185"</span>    <span class="string">"140557"</span>    <span class="string">"223697"</span>    <span class="string">"260423"</span>   </div><div class="line">[<span class="number">37</span>] <span class="string">"319148"</span>    <span class="string">"319149"</span>    <span class="string">"319150"</span>    <span class="string">"319151"</span>    <span class="string">"319152"</span>    <span class="string">"319153"</span>   </div><div class="line">[<span class="number">43</span>] <span class="string">"319154"</span>    <span class="string">"319155"</span>    <span class="string">"319156"</span>    <span class="string">"319157"</span>    <span class="string">"319158"</span>    <span class="string">"319159"</span>   </div><div class="line">[<span class="number">49</span>] <span class="string">"319160"</span>    <span class="string">"319161"</span>    <span class="string">"319565"</span>    <span class="string">"320332"</span>    <span class="string">"320558"</span>    <span class="string">"326619"</span>   </div><div class="line">[<span class="number">55</span>] <span class="string">"326620"</span>    <span class="string">"360198"</span>    <span class="string">"497652"</span>    <span class="string">"544973"</span>    <span class="string">"625328"</span>    <span class="string">"667250"</span>   </div><div class="line">[<span class="number">61</span>] <span class="string">"100041230"</span> <span class="string">"102641229"</span> <span class="string">"102641751"</span> <span class="string">"102642045"</span></div><div class="line"></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRanks储存的是排序信息</span></div><div class="line">&gt; data(exampleRanks)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRnak是一个数字型的向量</span></div><div class="line">&gt; class(exampleRanks)</div><div class="line">[<span class="number">1</span>] <span class="string">"numeric"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRanks中的向量名称则是对称的Entrez ID</span></div><div class="line">&gt; head(exampleRanks)</div><div class="line">   <span class="number">170942</span>    <span class="number">109711</span>     <span class="number">18124</span>     <span class="number">12775</span>     <span class="number">72148</span>     <span class="number">16010</span> </div><div class="line">-<span class="number">63.33703</span> -<span class="number">49.74779</span> -<span class="number">43.63878</span> -<span class="number">41.51889</span> -<span class="number">33.26039</span> -<span class="number">32.77626</span> </div><div class="line">&gt; tail(unname(exampleRanks))</div><div class="line">[<span class="number">1</span>] <span class="number">47.58235</span> <span class="number">49.87543</span> <span class="number">50.25179</span> <span class="number">50.86532</span> <span class="number">51.16110</span> <span class="number">53.28400</span></div></pre></td></tr></table></figure>
<h3 id="分析">分析</h3>
<p>现在我们使用上面的数据进行GSEA分析，进行GSEA分析时，我们的通路文件（也就是GSEA官网中的GMT文件，对应的就是基因集<code>S</code>）储存在<code>pathways</code>参数中，用户自己的数据（排过序的数据）放在<code>stats</code>数据集中，剩下的参数不用管，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fgseaRes &lt;- fgsea(pathways = examplePathways, </div><div class="line">                  stats = exampleRanks,</div><div class="line">                  minSize=<span class="number">15</span>,</div><div class="line">                  maxSize=<span class="number">500</span>,</div><div class="line">                  nperm=<span class="number">100000</span>)</div></pre></td></tr></table></figure>
<p>分析的结果fgseaRes是一个<code>data.table</code>格式的文件，使用<code>plotEnrichment</code>函数可以绘制出GSEA分析的结果，如下所示： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">library(ggplot2)</div><div class="line"># 参考 https://github.com/Rdatatable/data.table/wiki</div><div class="line"># 来查看data.table相关的信息</div><div class="line">class(fgseaRes)</div><div class="line"></div><div class="line"># 查看p值小于0.01的通路的数目</div><div class="line">sum(fgseaRes[, padj &lt; 0.01])</div><div class="line"></div><div class="line"># 绘制出最显著的富集通路</div><div class="line">plotEnrichment(examplePathways[[head(fgseaRes[order(pval), ], 1)$pathway]],</div><div class="line">               exampleRanks) + </div><div class="line">               labs(title=head(fgseaRes[order(pval), ], 1)$pathway)</div></pre></td></tr></table></figure></p>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; class(fgseaRes)</div><div class="line">[<span class="number">1</span>] <span class="string">"data.table"</span> <span class="string">"data.frame"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 查看p值小于0.01的通路的数目</span></div><div class="line">&gt; sum(fgseaRes[, padj &lt; <span class="number">0.01</span>])</div><div class="line">[<span class="number">1</span>] <span class="number">78</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911122517.jpeg">

</div>
<p>可以画出具体的某条通路，如下所示；</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">plotEnrichment(examplePathways[[<span class="string">"5991130_Programmed_Cell_Death"</span>]],</div><div class="line">               exampleRanks) + </div><div class="line">  labs(title=<span class="string">"Programmed Cell Death"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911125331.jpeg">

</div>
<p>还可以在一张图中绘制出前10个富集通路，以及后10个富集通路，一共20个，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">topPathwaysUp &lt;- fgseaRes[ES &gt; 0][head(order(pval), n=10), pathway]</div><div class="line">topPathwaysDown &lt;- fgseaRes[ES &lt; 0][head(order(pval), n=10), pathway]</div><div class="line">topPathways &lt;- c(topPathwaysUp, rev(topPathwaysDown))</div><div class="line">plotGseaTable(examplePathways[topPathways], exampleRanks, fgseaRes, </div><div class="line">              gseaParam = 0.5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911122626.jpeg">

</div>
<h3 id="reactome">Reactome</h3>
<p>也可以使用<a href="https://reactome.org/" target="_blank" rel="external">Reactome</a>通路来进行GSEA分析，此时需要安装<code>reactome.db</code>包，这个包很大，600多M，下载的时间很长，安装过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># BiocManager::install("reactome.db")</span></div><div class="line"><span class="keyword">library</span>(reactome.db)</div><div class="line"></div><div class="line">my_pathways &lt;- reactomePathways(names(exampleRanks))</div><div class="line"> </div><div class="line"><span class="comment"># Reactome pathways have a median of 11 genes</span></div><div class="line">summary(sapply(my_pathways, length))</div><div class="line">   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </div><div class="line">    <span class="number">1.0</span>     <span class="number">4.0</span>    <span class="number">11.0</span>    <span class="number">30.3</span>    <span class="number">30.0</span>  <span class="number">1140.0</span></div><div class="line"> </div><div class="line">fgsea_reactome &lt;- fgsea(pathways = my_pathways, </div><div class="line">                        stats = exampleRanks,</div><div class="line">                        minSize=<span class="number">15</span>,</div><div class="line">                        maxSize=<span class="number">500</span>,</div><div class="line">                        nperm=<span class="number">100000</span>)</div><div class="line"> </div><div class="line">head(fgsea_reactome[order(pval), ])</div><div class="line">sum(fgsea_reactome[, padj &lt; <span class="number">0.01</span>])</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; head(fgsea_reactome[order(pval), ])</div><div class="line">                    pathway         pval         padj</div><div class="line"><span class="number">1</span>:               Cell Cycle <span class="number">1.212856e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">2</span>:      Cell Cycle, Mitotic <span class="number">1.236552e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">3</span>: Neutrophil degranulation <span class="number">1.274974e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">4</span>: Signaling by Rho GTPases <span class="number">1.291339e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">5</span>:                  M Phase <span class="number">1.303509e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">6</span>:   Cell Cycle Checkpoints <span class="number">1.350749e-05</span> <span class="number">0.0002571089</span></div><div class="line">          ES      NES nMoreExtreme size</div><div class="line"><span class="number">1</span>: <span class="number">0.5324037</span> <span class="number">2.674889</span>            <span class="number">0</span>  <span class="number">414</span></div><div class="line"><span class="number">2</span>: <span class="number">0.5475346</span> <span class="number">2.719737</span>            <span class="number">0</span>  <span class="number">363</span></div><div class="line"><span class="number">3</span>: <span class="number">0.4258088</span> <span class="number">2.074081</span>            <span class="number">0</span>  <span class="number">296</span></div><div class="line"><span class="number">4</span>: <span class="number">0.4073481</span> <span class="number">1.966039</span>            <span class="number">0</span>  <span class="number">271</span></div><div class="line"><span class="number">5</span>: <span class="number">0.5022189</span> <span class="number">2.408762</span>            <span class="number">0</span>  <span class="number">255</span></div><div class="line"><span class="number">6</span>: <span class="number">0.6085986</span> <span class="number">2.836881</span>            <span class="number">0</span>  <span class="number">200</span></div><div class="line">                                  leadingEdge</div><div class="line"><span class="number">1</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line"><span class="number">2</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line"><span class="number">3</span>:    <span class="number">11676</span>,<span class="number">14190</span>,<span class="number">53381</span>,<span class="number">12306</span>,<span class="number">20430</span>,<span class="number">12505</span>,<span class="keyword">...</span></div><div class="line"><span class="number">4</span>: <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">20430</span>,<span class="number">104215</span>,<span class="number">233406</span>,<span class="number">107995</span>,<span class="keyword">...</span></div><div class="line"><span class="number">5</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="number">52276</span>,<span class="keyword">...</span></div><div class="line"><span class="number">6</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="number">12428</span>,<span class="keyword">...</span></div><div class="line">&gt; sum(fgsea_reactome[, padj &lt; <span class="number">0.01</span>])</div><div class="line">[<span class="number">1</span>] <span class="number">103</span></div></pre></td></tr></table></figure>
<h3 id="leading-edge">Leading edge</h3>
<p>在GSEA分析中，我们通常会提取那些构成高得分的核心基因。我们对高得分的核心基因的定义就是，基因集<code>S</code>中位于排序基因列表<code>L</code>位置中的得分最大处之前或之后的那些基因集（也就是GSEA结果中绿色曲线最高点的前面或后面）。</p>
<p>前面我们注意到，GSEA分析的结果中有1列被命名为<code>leadingEdge</code>。这一列包含了高得分的基因。我们使用Reactome通路的富集结果来提取这些基因，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># the most significant pathway</span></div><div class="line">fgsea_reactome[order(pval),][<span class="number">1</span>,]</div><div class="line"></div><div class="line"><span class="comment"># list of Entrez gene IDs that contributed to the enrichment score</span></div><div class="line">fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge</div><div class="line"></div><div class="line"><span class="comment"># how many genes are in the leading edge?</span></div><div class="line">length(fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge[[<span class="number">1</span>]])</div><div class="line"></div><div class="line"><span class="comment"># how many genes are in the Cell Cycle pathway?</span></div><div class="line">length(my_pathways[[<span class="string">'Cell Cycle'</span>]])</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment"># the most significant pathway</span></div><div class="line">&gt; fgsea_reactome[order(pval),][<span class="number">1</span>,]</div><div class="line">      pathway         pval         padj        ES</div><div class="line"><span class="number">1</span>: Cell Cycle <span class="number">1.212856e-05</span> <span class="number">0.0002571089</span> <span class="number">0.5324037</span></div><div class="line">        NES nMoreExtreme size</div><div class="line"><span class="number">1</span>: <span class="number">2.674889</span>            <span class="number">0</span>  <span class="number">414</span></div><div class="line">                                leadingEdge</div><div class="line"><span class="number">1</span>: <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># list of Entrez gene IDs that contributed to the enrichment score</span></div><div class="line">&gt; fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge</div><div class="line">[[<span class="number">1</span>]]</div><div class="line">  [<span class="number">1</span>] <span class="string">"66336"</span>  <span class="string">"66977"</span>  <span class="string">"15366"</span>  <span class="string">"12442"</span>  <span class="string">"107995"</span></div><div class="line">  [<span class="number">6</span>] <span class="string">"66442"</span>  <span class="string">"12571"</span>  <span class="string">"12428"</span>  <span class="string">"52276"</span>  <span class="string">"54392"</span> </div><div class="line"> [<span class="number">11</span>] <span class="string">"66311"</span>  <span class="string">"215387"</span> <span class="string">"67629"</span>  <span class="string">"12649"</span>  <span class="string">"72415"</span> </div><div class="line"> [<span class="number">16</span>] <span class="string">"56150"</span>  <span class="string">"57441"</span>  <span class="string">"20877"</span>  <span class="string">"67121"</span>  <span class="string">"12615"</span> </div><div class="line"> [<span class="number">21</span>] <span class="string">"11799"</span>  <span class="string">"66468"</span>  <span class="string">"67849"</span>  <span class="string">"19053"</span>  <span class="string">"73804"</span> </div><div class="line"> [<span class="number">26</span>] <span class="string">"76044"</span>  <span class="string">"20878"</span>  <span class="string">"15270"</span>  <span class="string">"13555"</span>  <span class="string">"60411"</span> </div><div class="line"> [<span class="number">31</span>] <span class="string">"12580"</span>  <span class="string">"17219"</span>  <span class="string">"69270"</span>  <span class="string">"12575"</span>  <span class="string">"69263"</span> </div><div class="line"> [<span class="number">36</span>] <span class="string">"12448"</span>  <span class="string">"14211"</span>  <span class="string">"20873"</span>  <span class="string">"18005"</span>  <span class="string">"72119"</span> </div><div class="line"> [<span class="number">41</span>] <span class="string">"71988"</span>  <span class="string">"12189"</span>  <span class="string">"17215"</span>  <span class="string">"12534"</span>  <span class="string">"66156"</span> </div><div class="line"> [<span class="number">46</span>] <span class="string">"208628"</span> <span class="string">"237911"</span> <span class="string">"22390"</span>  <span class="string">"68240"</span>  <span class="string">"228421"</span></div><div class="line"> [<span class="number">51</span>] <span class="string">"68014"</span>  <span class="string">"269582"</span> <span class="string">"19348"</span>  <span class="string">"12236"</span>  <span class="string">"72151"</span> </div><div class="line"> [<span class="number">56</span>] <span class="string">"18817"</span>  <span class="string">"21781"</span>  <span class="string">"18968"</span>  <span class="string">"217653"</span> <span class="string">"66934"</span> </div><div class="line"> [<span class="number">61</span>] <span class="string">"272551"</span> <span class="string">"227613"</span> <span class="string">"67141"</span>  <span class="string">"67951"</span>  <span class="string">"68612"</span> </div><div class="line"> [<span class="number">66</span>] <span class="string">"68298"</span>  <span class="string">"108000"</span> <span class="string">"23834"</span>  <span class="string">"106344"</span> <span class="string">"56452"</span> </div><div class="line"> [<span class="number">71</span>] <span class="string">"242705"</span> <span class="string">"18141"</span>  <span class="string">"223921"</span> <span class="string">"26886"</span>  <span class="string">"13557"</span> </div><div class="line"> [<span class="number">76</span>] <span class="string">"26909"</span>  <span class="string">"72787"</span>  <span class="string">"268697"</span> <span class="string">"72155"</span>  <span class="string">"56371"</span> </div><div class="line"> [<span class="number">81</span>] <span class="string">"17535"</span>  <span class="string">"107823"</span> <span class="string">"12531"</span>  <span class="string">"327762"</span> <span class="string">"12567"</span> </div><div class="line"> [<span class="number">86</span>] <span class="string">"229841"</span> <span class="string">"67052"</span>  <span class="string">"16319"</span>  <span class="string">"66634"</span>  <span class="string">"171567"</span></div><div class="line"> [<span class="number">91</span>] <span class="string">"26931"</span>  <span class="string">"67203"</span>  <span class="string">"12235"</span>  <span class="string">"19891"</span>  <span class="string">"74470"</span> </div><div class="line"> [<span class="number">96</span>] <span class="string">"72083"</span>  <span class="string">"381318"</span> <span class="string">"66570"</span>  <span class="string">"17216"</span>  <span class="string">"76308"</span> </div><div class="line">[<span class="number">101</span>] <span class="string">"19687"</span>  <span class="string">"17218"</span>  <span class="string">"102920"</span> <span class="string">"29870"</span>  <span class="string">"18973"</span> </div><div class="line">[<span class="number">106</span>] <span class="string">"16881"</span>  <span class="string">"17463"</span>  <span class="string">"75786"</span>  <span class="string">"19645"</span>  <span class="string">"19075"</span> </div><div class="line">[<span class="number">111</span>] <span class="string">"26417"</span>  <span class="string">"69736"</span>  <span class="string">"19357"</span>  <span class="string">"76816"</span>  <span class="string">"70385"</span> </div><div class="line">[<span class="number">116</span>] <span class="string">"70645"</span>  <span class="string">"22628"</span>  <span class="string">"225182"</span> <span class="string">"22627"</span>  <span class="string">"52683"</span> </div><div class="line">[<span class="number">121</span>] <span class="string">"19076"</span>  <span class="string">"18972"</span>  <span class="string">"231863"</span> <span class="string">"26932"</span>  <span class="string">"12544"</span> </div><div class="line">[<span class="number">126</span>] <span class="string">"17997"</span>  <span class="string">"51788"</span>  <span class="string">"26440"</span>  <span class="string">"68549"</span>  <span class="string">"12445"</span> </div><div class="line">[<span class="number">131</span>] <span class="string">"19088"</span>  <span class="string">"269113"</span> <span class="string">"26444"</span>  <span class="string">"19324"</span>  <span class="string">"103733"</span></div><div class="line">[<span class="number">136</span>] <span class="string">"59001"</span>  <span class="string">"107976"</span> <span class="string">"19179"</span>  <span class="string">"12579"</span>  <span class="string">"232987"</span></div><div class="line">[<span class="number">141</span>] <span class="string">"17420"</span>  <span class="string">"228769"</span> <span class="string">"219072"</span> <span class="string">"26445"</span>  <span class="string">"105988"</span></div><div class="line">[<span class="number">146</span>] <span class="string">"69745"</span>  <span class="string">"18538"</span>  <span class="string">"69928"</span>  <span class="string">"11651"</span>  <span class="string">"235559"</span></div><div class="line">[<span class="number">151</span>] <span class="string">"68097"</span>  <span class="string">"57296"</span>  <span class="string">"63955"</span>  <span class="string">"14235"</span>  <span class="string">"19170"</span> </div><div class="line">[<span class="number">156</span>] <span class="string">"17246"</span>  <span class="string">"17220"</span>  <span class="string">"12144"</span>  <span class="string">"50793"</span>  <span class="string">"77605"</span> </div><div class="line">[<span class="number">161</span>] <span class="string">"18392"</span>  <span class="string">"236930"</span> <span class="string">"67151"</span>  <span class="string">"70024"</span>  <span class="string">"59126"</span> </div><div class="line">[<span class="number">166</span>] <span class="string">"66296"</span>  <span class="string">"16906"</span>  <span class="string">"109145"</span> <span class="string">"71819"</span>  <span class="string">"67733"</span> </div><div class="line">[<span class="number">171</span>] <span class="string">"50883"</span>  <span class="string">"12447"</span>  <span class="string">"12532"</span>  <span class="string">"14156"</span>  <span class="string">"26442"</span> </div><div class="line">[<span class="number">176</span>] <span class="string">"19177"</span>  <span class="string">"230376"</span> <span class="string">"245688"</span></div><div class="line"></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># how many genes are in the leading edge?</span></div><div class="line">&gt; length(fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge[[<span class="number">1</span>]])</div><div class="line">[<span class="number">1</span>] <span class="number">178</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># how many genes are in the Cell Cycle pathway?</span></div><div class="line">&gt; length(my_pathways[[<span class="string">'Cell Cycle'</span>]])</div><div class="line">[<span class="number">1</span>] <span class="number">414</span></div></pre></td></tr></table></figure>
<h2 id="总结">总结</h2>
<p>GSEA是于2005年首次提出来的，现在已经成了基因表达分析中的常规分析手段，它不同于GO分析，GO分析只关注差异基因，而GSEA分析则关注所有的基因。<code>fgsea</code>包可以使用预先排列好的基因一R中进行GSEA分析。p值的计算结果是基于置换检验(permutation test)，这种方法并不是十分精我，因为它忽略了基因之间的相关性，有可能会导致假阳性。但是，在这种方法在研究上调与上调基因方面还是很有用的。即使你计算出的GSEA结果中，p值大于0.05，但是是也可以参考leading edge基因集，为你的实验进行指导。</p>
<h2 id="案例分析">案例分析</h2>
<p>作者提供了用于生成类似于<strong>exampleRanks</strong>文件的R脚本，不过使用的GEO的数据，平时自己利用<code>fgsea</code>包进行GSEA分析时，生成就好，现在看一下如何将GEO的数据生成类似于****exampleRanks****文件的排序信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">source</span>(<span class="string">"https://raw.githubusercontent.com/assaron/r-utils/master/R/exprs.R"</span>)</div><div class="line"><span class="keyword">library</span>(GEOquery)</div><div class="line"><span class="keyword">library</span>(limma)</div><div class="line">gse14308 &lt;- getGEO(<span class="string">"GSE14308"</span>)[[<span class="number">1</span>]]</div><div class="line">pData(gse14308)$condition &lt;- sub(<span class="string">"-.*$"</span>, <span class="string">""</span>, gse14308$title)</div><div class="line">es &lt;- collapseBy(gse14308, fData(gse14308)$ENTREZ_GENE_ID, FUN=median)</div><div class="line">es &lt;- es[!grepl(<span class="string">"///"</span>, rownames(es)), ]</div><div class="line">es &lt;- es[rownames(es) != <span class="string">""</span>, ]</div><div class="line">exprs(es) &lt;- normalizeBetweenArrays(log2(exprs(es)+<span class="number">1</span>), method=<span class="string">"quantile"</span>)</div><div class="line">es.design &lt;- model.matrix(~<span class="number">0</span>+condition, data=pData(es))</div><div class="line">fit &lt;- lmFit(es, es.design)</div><div class="line">fit2 &lt;- contrasts.fit(fit, makeContrasts(conditionTh1-conditionNaive,</div><div class="line">                                         levels=es.design))</div><div class="line">fit2 &lt;- eBayes(fit2)</div><div class="line">de &lt;- data.table(topTable(fit2, adjust.method=<span class="string">"BH"</span>, number=<span class="number">12000</span>, sort.by = <span class="string">"B"</span>), keep.rownames = <span class="literal">T</span>)</div><div class="line">ranks &lt;- de[order(t), list(rn, t)]</div><div class="line">head(ranks)</div><div class="line">tail(ranks)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">&gt; head(ranks)</div><div class="line">       rn         t</div><div class="line"><span class="number">1</span>: <span class="number">109711</span> -<span class="number">50.99316</span></div><div class="line"><span class="number">2</span>:  <span class="number">18124</span> -<span class="number">44.40559</span></div><div class="line"><span class="number">3</span>:  <span class="number">12775</span> -<span class="number">43.22109</span></div><div class="line"><span class="number">4</span>:  <span class="number">72148</span> -<span class="number">33.74441</span></div><div class="line"><span class="number">5</span>:  <span class="number">16010</span> -<span class="number">33.34034</span></div><div class="line"><span class="number">6</span>:  <span class="number">16206</span> -<span class="number">30.67962</span></div><div class="line">&gt; tail(ranks)</div><div class="line">      rn        t</div><div class="line"><span class="number">1</span>: <span class="number">80901</span> <span class="number">47.55618</span></div><div class="line"><span class="number">2</span>: <span class="number">58801</span> <span class="number">48.53756</span></div><div class="line"><span class="number">3</span>: <span class="number">15937</span> <span class="number">49.88068</span></div><div class="line"><span class="number">4</span>: <span class="number">13730</span> <span class="number">50.17693</span></div><div class="line"><span class="number">5</span>: <span class="number">12772</span> <span class="number">50.32302</span></div><div class="line"><span class="number">6</span>: <span class="number">80876</span> <span class="number">51.72987</span></div></pre></td></tr></table></figure>
<p>现在绘制一下上述数据中6个上升与6个下降的基因热图：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(pheatmap)</div><div class="line"> </div><div class="line">my_group &lt;- data.frame(group = pData(es)$condition)</div><div class="line">row.names(my_group) &lt;- colnames(exprs(es))</div><div class="line"> </div><div class="line">pheatmap(</div><div class="line">  mat = es[c(head(de[order(t), <span class="number">1</span>])$rn, tail(de[order(t), <span class="number">1</span>])$rn),],</div><div class="line">  annotation_col = my_group,</div><div class="line">  cluster_rows = <span class="literal">FALSE</span>,</div><div class="line">  cellwidth=<span class="number">25</span>,</div><div class="line">  cellheight=<span class="number">15</span></div><div class="line">)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911133906.png">

</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html" target="_blank" rel="external">Using fgsea package</a></li>
<li><a href="https://www.r-bloggers.com/comparison-of-clusterprofiler-and-gsea-p/" target="_blank" rel="external">Comparison of clusterProfiler and GSEA-P</a></li>
<li><a href="https://davetang.org/muse/2018/01/10/using-fast-preranked-gene-set-enrichment-analysis-fgsea-package/" target="_blank" rel="external">Using the fast preranked gene set enrichment analysis (fgsea) package</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/07/Experiment/HIPC介绍/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/07/Experiment/HIPC介绍/" itemprop="url">HIPC介绍</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-07T12:00:00+08:00">
                2019-09-07
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物实验笔记/" itemprop="url" rel="index">
                    <span itemprop="name">生物实验笔记</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  803
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  3
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>最近在看GSEA有关的内容，GSEA官网里涉及到一个基因集，即<code>C7 collection: Immunologic signatures</code>，这是一个与免疫相关的基因集，根据官网的介绍，这个基因集是由BROAD实验室与丹娜法伯癌症研究院（Dana-Farber Cancer Institute）的Haining Lab和人类免疫学项目联合合会(HIPC)合作的一部分，当时还发了一篇Immunity，文献信息如下：</p>
<p>Godec J, Tan Y, Liberzon A, Tamayo P, Bhattacharya S, Butte A, Mesirov JP, Haining WN, Compendium of Immune Signatures Identifies Conserved and Species-Specific Biology in Response to Inflammation, 2016, Immunity 44(1), 194-206.</p>
<p>以下内容是HIPC官网的一些资料。</p>
<h2 id="hipc简介">HIPC简介</h2>
<p>HIPC英文全称是Human Immunology Project Consortium，即人类免疫学项目研究联合会，成立于2010年，后又于2015年重组，即由原来的NIAID的过敏所（注：NIAID的全称为National Institute of Allergy and Infectious Diseases，即<strong>国家过敏与感染性疾病研究院</strong>），免疫所和移植所进行合并，统一作为NIAID的构成部分，集中研究人类免疫学。</p>
<p>注：上面一段话是我按原文翻译后理解的，不一定准确，为了避免歧义我，这里贴出原文：</p>
<blockquote>
<p>The Human Immunology Project Consortium (HIPC) program was established in 2010, and renewed in 2015, by the NIAID Division of Allergy, Immunology, and Transplantation as part of the overall NIAID focus on human immunology.</p>
</blockquote>
<p>通过HIPC计划，研究者们使用各种现代分析工具分析人类多个免疫学方面的数据，其中包括多重转录(multiplex transcriptional)，细胞因子，蛋白质谱；白细胞亚型的多参数表型(multiparameter phenotyping of leukocyte subsets)；评估白细胞功能状态以及开发多种算法。</p>
<p>HIPC项目旨在建立一个集中的研究中心，以及一个广泛的，中心化的数据库，从而让更多的研究者们使用。HIPC项目产生的信息将会有助于研究者们广泛地理解人类免疫系统方方面面，包括免疫系统的调控，发现免疫系统组分与其它生物系统的新型关系，从而确定新的免疫介质与通路，以及构建不同种群疫苗案安全性的预测因子，从而确保对不同疫苗配方的快速评估，确立接种方案等。该知识库还将会构建起免疫系统介导的人类疾病的研究基础，这些人类免疫性疾病包括过敏，哮喘，移植排斥和自身免疫性疾病以及各种炎症性疾病。</p>
<h2 id="hipc的任务">HIPC的任务</h2>
<h3 id="目标">目标</h3>
<ol style="list-style-type: decimal">
<li>定义稳定和活化的人类免疫系统的表达谱/标签/足迹(Define profiles/signatures/fingerprints of steady-state and activated human immune system)</li>
<li>创建一个中心化的知识库或资源：</li>
</ol>
<ul>
<li>有利于研究人类免疫系统；</li>
<li>针对人类疾病开发新的疗法。</li>
</ul>
<h3 id="研究">研究</h3>
<ol style="list-style-type: decimal">
<li>感染导致的稳态异常(Perturbations of the steady-state by infection)；</li>
<li>疫苗；</li>
<li>佐剂研究(Adjuvant administration )</li>
</ol>
<h3 id="方法">方法</h3>
<ol style="list-style-type: decimal">
<li>定义人类转录组/蛋白质谱；</li>
<li>多重分析(multiplex assays)；</li>
<li>多参数表型，系统生物学；</li>
<li>质谱。</li>
</ol>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>https://www.immuneprofiling.org/hipc/page/showPage?pg=sci-about</li>
<li>http://software.broadinstitute.org/gsea/msigdb/collection_details.jsp#C7</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/04/PH525x/PH525xC11-04-基因组实验中的数据结构与管理/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/04/PH525x/PH525xC11-04-基因组实验中的数据结构与管理/" itemprop="url">PH525xC11-04-基因组实验中的数据结构与管理</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-04T12:00:00+08:00">
                2019-09-04
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/PH525x/" itemprop="url" rel="index">
                    <span itemprop="name">PH525x</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  9,924
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  45
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="前言">前言</h1>
<p>在科学研究中，大家都认为数据管理是一项专业且乏味的工作。因为数据管理一旦出现问题，会导致自身的资源与声誉付出高昂的代价，因此，数据管理必须要高度可靠且有效。在传统实验室中，研究人员通常使用电子表格来管理数据，但是这种方式存在着风险。因此，我们希望通常简化可靠的数据管理操作来增加数据的价值。</p>
<p>在Bioconductor中，指导软件开发的原则也可以应用于数据管理。模块化和可扩展的数据结构能够发挥更高的价值。打包和版本控制协议适用于定义数据类(data class)。我们可以将电子表格转化为语义丰富的对象，并联合使用NCBI GEO数据库，处理BAM和BED文件，以及外部存储来促进与大型多组学数据库（例如TCGA）衔接来说明一下上述的这些思路。</p>
<h1 id="汇总多个表格信息">汇总多个表格信息</h1>
<h2 id="演示包gse5859subset">演示包：GSE5859Subset</h2>
<p>GSE5859Subset是一个包，里面含有一个实验的表达数据。当加载了这个包后，我们可以看到这个包有三个组件：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">library(GSE5859Subset)</div><div class="line">data(GSE5859Subset)</div><div class="line">dim(geneExpression)</div><div class="line">## [1] 8793   24</div><div class="line">dim(geneAnnotation)</div><div class="line">## [1] 8793    4</div><div class="line">dim(sampleInfo)</div><div class="line">## [1] 24  4</div></pre></td></tr></table></figure>
<p>我们来查看一下这3个组件是否有关（一个矩阵，两个数据框），如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">all.equal(sampleInfo$filename, colnames(geneExpression))</div><div class="line">## [1] TRUE</div><div class="line">all.equal(rownames(geneExpression), geneAnnotation$PROBEID)</div><div class="line">## [1] TRUE</div></pre></td></tr></table></figure>
<p>简单来讲，我们可以把 <code>sampleInfo$filename</code> 视为一个按行连接的节点(key)，它连接了样本信息表格与基因表达表格。<code>colnames(geneExpression)</code>将该矩阵的列连接到<code>sampleInfo</code>的行中对应的样本。简单来说就是，geneExpression的列名是样本名，而sampleInfo中的某1列中也含有这些样本名，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191029171318.png">

</div>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191029171338.png">

</div>
<p>与之类似, <code>geneExpression</code> 的行名，对应了<code>geneAnnotation</code>中探针数据，即 <code>PROBEID</code> ，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">options(digits=2)</div><div class="line">cbind(sampleInfo[1:3,], colnames(geneExpression)[1:3], </div><div class="line">    t(geneExpression)[1:3,1:4])</div><div class="line">##     ethnicity       date         filename group</div><div class="line">## 107       ASN 2005-06-23 GSM136508.CEL.gz     1</div><div class="line">## 122       ASN 2005-06-27 GSM136530.CEL.gz     1</div><div class="line">## 113       ASN 2005-06-27 GSM136517.CEL.gz     1</div><div class="line">##     colnames(geneExpression)[1:3] 1007_s_at 1053_at 117_at 121_at</div><div class="line">## 107              GSM136508.CEL.gz       6.5     7.5    5.4    7.9</div><div class="line">## 122              GSM136530.CEL.gz       6.4     7.3    5.1    7.7</div><div class="line">## 113              GSM136517.CEL.gz       6.3     7.2    5.0    7.5</div></pre></td></tr></table></figure>
<h2 id="在expressionset汇总表格">在ExpressionSet汇总表格</h2>
<p><code>ExpressionSet</code> 容器用于管理一个项目中的所有信息。为了改善基因名与样本名命名的可视化效果，我们可以以下每个组件的注释信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">head(rownames(sampleInfo))</div><div class="line">head(rownames(geneAnnotation))</div><div class="line">rownames(sampleInfo) = sampleInfo$filename</div><div class="line">rownames(geneAnnotation) = geneAnnotation$PROBEID</div><div class="line">head(rownames(sampleInfo))</div><div class="line">head(rownames(geneAnnotation))</div></pre></td></tr></table></figure>
<p>结果如下所：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; head(rownames(sampleInfo))</div><div class="line">[<span class="number">1</span>] <span class="string">"107"</span> <span class="string">"122"</span> <span class="string">"113"</span> <span class="string">"163"</span> <span class="string">"153"</span> <span class="string">"161"</span></div><div class="line">&gt; head(rownames(geneAnnotation))</div><div class="line">[<span class="number">1</span>] <span class="string">"1"</span>  <span class="string">"30"</span> <span class="string">"31"</span> <span class="string">"32"</span> <span class="string">"33"</span> <span class="string">"34"</span></div><div class="line">&gt; rownames(sampleInfo) = sampleInfo$filename</div><div class="line">&gt; rownames(geneAnnotation) = geneAnnotation$PROBEID</div><div class="line">&gt; head(rownames(sampleInfo))</div><div class="line">[<span class="number">1</span>] <span class="string">"GSM136508.CEL.gz"</span> <span class="string">"GSM136530.CEL.gz"</span> <span class="string">"GSM136517.CEL.gz"</span></div><div class="line">[<span class="number">4</span>] <span class="string">"GSM136576.CEL.gz"</span> <span class="string">"GSM136566.CEL.gz"</span> <span class="string">"GSM136574.CEL.gz"</span></div><div class="line">&gt; head(rownames(geneAnnotation))</div><div class="line">[<span class="number">1</span>] <span class="string">"1007_s_at"</span> <span class="string">"1053_at"</span>   <span class="string">"117_at"</span>    <span class="string">"121_at"</span>    <span class="string">"1255_g_at"</span></div><div class="line">[<span class="number">6</span>] <span class="string">"1294_at"</span></div></pre></td></tr></table></figure>
<p>现在我们已经将样本的信息进行了更改，现在我们生成 <code>ExpressionSet</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line">es5859 = ExpressionSet(assayData=geneExpression)</div><div class="line">pData(es5859) = sampleInfo</div><div class="line">fData(es5859) = geneAnnotation</div><div class="line">es5859</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 24 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: GSM136508.CEL.gz GSM136530.CEL.gz ...</span></div><div class="line"><span class="comment">##     GSM136572.CEL.gz (24 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename group</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 1007_s_at 1053_at ... AFFX-r2-P1-cre-5_at (8793</span></div><div class="line"><span class="comment">##     total)</span></div><div class="line"><span class="comment">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</span></div><div class="line"><span class="comment">##   fvarMetadata: labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation:</span></div></pre></td></tr></table></figure>
<p>这样做的一个好处的就是，我们可以使用<code>fData</code>和<code>pData</code>组件中更高层次的注释来选择特征。例如，我们现在只提取Y染色体上的基因表达特征：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">es5859[which(fData(es5859)$CHR==<span class="string">"chrY"</span>),]</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 21 features, 24 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: GSM136508.CEL.gz GSM136530.CEL.gz ...</span></div><div class="line"><span class="comment">##     GSM136572.CEL.gz (24 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename group</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 201909_at 204409_s_at ... 211149_at (21 total)</span></div><div class="line"><span class="comment">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</span></div><div class="line"><span class="comment">##   fvarMetadata: labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation:</span></div></pre></td></tr></table></figure>
<p>我们可以看一下可<code>ExpressionSet</code>这种类所能使用的所有方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">methods(class=<span class="string">"ExpressionSet"</span>)</div><div class="line"><span class="comment">##  [1] [                [[               [[&lt;-             $$               </span></div><div class="line"><span class="comment">##  [5] $$&lt;-              abstract         annotation       annotation&lt;-    </span></div><div class="line"><span class="comment">##  [9] as.data.frame    as.matrix        assayData        assayData&lt;-     </span></div><div class="line"><span class="comment">## [13] boxplot          classVersion     classVersion&lt;-   coerce          </span></div><div class="line"><span class="comment">## [17] combine          description      description&lt;-    dim             </span></div><div class="line"><span class="comment">## [21] dimnames         dimnames&lt;-       dims             esApply         </span></div><div class="line"><span class="comment">## [25] experimentData   experimentData&lt;- exprs            exprs&lt;-         </span></div><div class="line"><span class="comment">## [29] fData            fData&lt;-          featureData      featureData&lt;-   </span></div><div class="line"><span class="comment">## [33] featureNames     featureNames&lt;-   fvarLabels       fvarLabels&lt;-    </span></div><div class="line"><span class="comment">## [37] fvarMetadata     fvarMetadata&lt;-   initialize       isCurrent       </span></div><div class="line"><span class="comment">## [41] isVersioned      KEGG2heatmap     KEGGmnplot       makeDataPackage </span></div><div class="line"><span class="comment">## [45] notes            notes&lt;-          pData            pData&lt;-         </span></div><div class="line"><span class="comment">## [49] phenoData        phenoData&lt;-      preproc          preproc&lt;-       </span></div><div class="line"><span class="comment">## [53] protocolData     protocolData&lt;-   pubMedIds        pubMedIds&lt;-     </span></div><div class="line"><span class="comment">## [57] rowMedians       rowQ             sampleNames      sampleNames&lt;-   </span></div><div class="line"><span class="comment">## [61] show             storageMode      storageMode&lt;-    updateObject    </span></div><div class="line"><span class="comment">## [65] updateObjectTo   varLabels        varLabels&lt;-      varMetadata     </span></div><div class="line"><span class="comment">## [69] varMetadata&lt;-    write.exprs     </span></div><div class="line"><span class="comment">## see '?methods' for accessing help and source code</span></div></pre></td></tr></table></figure>
<p>在这些方法中比较重要的有以下几个：</p>
<ul>
<li><code>exprs()</code>: 获取数值型表达值；</li>
<li><code>pData()</code>: 获取样本数据，就是样本名；</li>
<li><code>fData()</code>: 获取特征数据，就是基因表达值；</li>
<li><code>annotation()</code>: 获取特征名的标签；get a tag that identifies nomenclature for feature names</li>
<li><code>experimentData()</code>: 获取符合 <a href="https://www.ncbi.nlm.nih.gov/geo/info/MIAME.html" target="_blank" rel="external">MIAME</a>的元数据结构。</li>
</ul>
<p>许多方法都具有<code>setter</code>属性（注：这里的<code>setter</code>应该是R面向对象编程中的一个方法，我猜测可能是赋值一类的操作，具体的我还不清楚）。例如，</p>
<p><code>exprs&lt;-</code> 可以用于指定表达数值。此外，所有的组件都是可选的，因此我们的<code>es5859</code>对象中没有<code>annotation</code>或<code>experimentData</code>也不奇怪。我们随后会改善一下这个对象的自身的描述，首选我们设置注释信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">annotation(es5859)</div><div class="line">annotation(es5859) = &quot;hgfocus.db&quot; # need to look at GSE record in GEO, and know .db</div><div class="line">annotation(es5859)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; annotation(es5859)</div><div class="line">character(<span class="number">0</span>)</div><div class="line">&gt; annotation(es5859) = <span class="string">"hgfocus.db"</span> <span class="comment"># need to look at GSE record in GEO, and know .db</span></div><div class="line">&gt; </div><div class="line">&gt; annotation(es5859)</div><div class="line">[<span class="number">1</span>] <span class="string">"hgfocus.db"</span></div></pre></td></tr></table></figure>
<p>第二，获取符合MIAME文档的实验元数据信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(annotate)</div><div class="line">es5859</div><div class="line">mi = pmid2MIAME(<span class="string">"17206142"</span>)</div><div class="line">experimentData(es5859) = mi</div><div class="line">es5859</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 24 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: GSM136508.CEL.gz GSM136530.CEL.gz ...</span></div><div class="line"><span class="comment">##     GSM136572.CEL.gz (24 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename group</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 1007_s_at 1053_at ... AFFX-r2-P1-cre-5_at (8793</span></div><div class="line"><span class="comment">##     total)</span></div><div class="line"><span class="comment">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</span></div><div class="line"><span class="comment">##   fvarMetadata: labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">##   pubMedIds: 17206142 </span></div><div class="line"><span class="comment">## Annotation: hgfocus.db</span></div></pre></td></tr></table></figure>
<p>现在我们通过<code>abstract</code>方法就能获取一定信息：</p>
<p>Now, for example, the <code>abstract</code> method will function well:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">nchar(abstract(es5859))</div><div class="line"><span class="comment">## [1] 1057</span></div><div class="line">substr(abstract(es5859),<span class="number">1</span>,<span class="number">50</span>)</div><div class="line"><span class="comment">## [1] "Variation in DNA sequence contributes to individua"</span></div></pre></td></tr></table></figure>
<p>用于汇总这些表格的最新方法就是使用<code>SummarizedExperiment</code>类，我们后面会讲到。</p>
<h2 id="自同态环概念endomorphism">*自同态环概念(endomorphism)</h2>
<p>关于ExpressionSet容器设计的一个最后备注：假设 <code>x</code> 是一个ExpressionSet。方括号被定义为一操作符，当<code>G</code>和<code>S</code>分别是标识特征和合适样本的合适向量时，那么<code>X[G,S]</code>是具有仅限于<code>G</code>和<code>S</code>中标记的那些特征值和样本的ExrpessionSet。对<code>X</code>有限的所有操作也都对 <code>X[G, S]</code> 有效。此属性称为ExpressionSet相对于中括号进行子集的自同态。</p>
<p>上面一段不太懂，以下是原文：</p>
<blockquote>
<p>A final remark about the ExpressionSet container design: Suppose <code>X</code> is an ExpressionSet. The bracket operator has been defined so that whenever <code>G</code> and <code>S</code> are suitable vectors identifying features and samples respectively, <code>X[G, S]</code> is an ExpressionSet with features and samples restricted to those identified in <code>G</code> and <code>S</code>. All operations that are valid for <code>X</code> are valid for <code>X[G, S]</code>. This property is called the <em>endomorphism of ExpressionSet with respect to subsetting with bracket</em>.</p>
</blockquote>
<h1 id="geo-geoquery-arrayexpress与表达数据储存">GEO, GEOquery, ArrayExpress与表达数据储存</h1>
<p>基因表达数据库(GEO)储存了源于NIH资助的所有微阵列实验的数据。 Bioconductor的GEOquery软件包可以很方便地获取这些数据。而ArrayExpress软件包则可以很容易地获取欧洲分子生物学实验室赞助的ArrayExpress数据。</p>
<h2 id="geometadb">GEOmetadb</h2>
<p>GEO中有数以万计的实验结果。<em>GEOmetadb</em>包括了用于获取和检索SQLite数据库的工具，SQLite数据库含有对GEO内容的广泛注释信息。2017年10月检索到的信息超过了6G。因此，我们并不需要你使用这个包。如果你有兴趣的话，可以参考vignette中案例，介绍得非常详细。</p>
<h2 id="getgeo获取geo中的expressionset">getGEO获取GEO中的ExpressionSet</h2>
<p>例如我们对胶质母细胞瘤的基因组学特别感兴趣，并且已经性了一篇论文(Pmid 27746144)讨论了一种新陈代谢途径作可能会增强其治疗开发策略。这个实验使用了Affymetrix PrimeView阵列，并储存在GEO。现在我们使用getGEO获取这些数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GEOquery)</div><div class="line">glioMA = getGEO(<span class="string">"GSE78703"</span>)[[<span class="number">1</span>]]</div><div class="line"><span class="comment">## Found 1 file(s)</span></div><div class="line"><span class="comment">## GSE78703_series_matrix.txt.gz</span></div><div class="line"><span class="comment">## Parsed with column specification:</span></div><div class="line"><span class="comment">## cols(</span></div><div class="line"><span class="comment">##   ID_REF = col_character(),</span></div><div class="line"><span class="comment">##   GSM2072905 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072906 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072907 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072908 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072909 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072910 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072911 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072912 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072913 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072914 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072915 = col_double(),</span></div><div class="line"><span class="comment">##   GSM2072916 = col_double()</span></div><div class="line"><span class="comment">## )</span></div><div class="line"><span class="comment">## File stored at:</span></div><div class="line"><span class="comment">## /var/folders/5_/14ld0y7s0vbg_z0g2c9l8v300000gr/T//RtmpKRbO0i/GPL15207.soft</span></div><div class="line"><span class="comment">## Warning: 102 parsing failures.</span></div><div class="line"><span class="comment">##   row     col           expected         actual         file</span></div><div class="line"><span class="comment">## 49294 SPOT_ID 1/0/T/F/TRUE/FALSE AFFX-BioB-3_at literal data</span></div><div class="line"><span class="comment">## 49295 SPOT_ID 1/0/T/F/TRUE/FALSE AFFX-BioB-5_at literal data</span></div><div class="line"><span class="comment">## 49296 SPOT_ID 1/0/T/F/TRUE/FALSE AFFX-BioB-M_at literal data</span></div><div class="line"><span class="comment">## 49297 SPOT_ID 1/0/T/F/TRUE/FALSE AFFX-BioC-3_at literal data</span></div><div class="line"><span class="comment">## 49298 SPOT_ID 1/0/T/F/TRUE/FALSE AFFX-BioC-5_at literal data</span></div><div class="line"><span class="comment">## ..... ....... .................. .............. ............</span></div><div class="line"><span class="comment">## See problems(...) for more details.</span></div><div class="line">glioMA</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 49395 features, 12 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: GSM2072905 GSM2072906 ... GSM2072916 (12 total)</span></div><div class="line"><span class="comment">##   varLabels: title geo_accession ... treated with:ch1 (35 total)</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 11715100_at 11715101_s_at ... AFFX-TrpnX-M_at</span></div><div class="line"><span class="comment">##     (49395 total)</span></div><div class="line"><span class="comment">##   fvarLabels: ID GeneChip Array ... SPOT_ID (36 total)</span></div><div class="line"><span class="comment">##   fvarMetadata: Column Description labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">##   pubMedIds: 27746144 </span></div><div class="line"><span class="comment">## Annotation: GPL15207</span></div></pre></td></tr></table></figure>
<p>通过以上操作中，我们可以看到使用这个对象（数据集）可以查看使用LXR-623处理能够影响基因ABCA1的表达，文献对应的PubMed ID为27746144。</p>
<h2 id="arrayexpress检索并获取embl-ebi的arrayexpress">ArrayExpress检索并获取EMBL-EBI的ArrayExpress</h2>
<p><a href="https://www.ebi.ac.uk/arrayexpress/" target="_blank" rel="external">ArrayExpress Archive of Functional Genomics Data</a>储存了高通量功能基因组实验的数据，并为其他研究者再次使用这些数据提供方便。最近，ArrayExpress还从NCBI GEO导入了所有表达数据。</p>
<p><em>ArrayExpress</em>包支持使用<code>queryAE</code>直接查询EMBL-EBI数据库，现在我们来试一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ArrayExpress)</div><div class="line">sets = queryAE(keywords = <span class="string">"glioblastoma"</span>, species = <span class="string">"homo+sapiens"</span>)</div><div class="line">dim(sets)</div><div class="line"><span class="comment">## [1] 546   8</span></div><div class="line">sets[<span class="number">5</span>:<span class="number">7</span>,-c(<span class="number">7</span>,<span class="number">8</span>)]</div><div class="line"><span class="comment">##                      ID Raw Processed ReleaseDate PubmedID      Species</span></div><div class="line"><span class="comment">## E-MTAB-6881 E-MTAB-6881  no        no  2019-04-30       NA Homo sapiens</span></div><div class="line"><span class="comment">## E-MTAB-6882 E-MTAB-6882  no        no  2019-04-29       NA Homo sapiens</span></div><div class="line"><span class="comment">## E-MTAB-7490 E-MTAB-7490 yes        no  2019-03-22       NA Homo sapiens</span></div></pre></td></tr></table></figure>
<p><code>PubmedID</code>这一列显示的是Pubmed号，有的有，有的没有（没有的用NA标记），通过<code>getAE</code>函数则可以获取原始数据，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">initdir = dir()</div><div class="line">if (!file.exists(&quot;E-MTAB-5797.sdrf.txt&quot;)) nano = getAE(&quot;E-MTAB-5797&quot;)</div></pre></td></tr></table></figure>
<p>上面的代码将与该实验相关的数据下载到当前文件夹，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">afterget = dir()</div><div class="line">setdiff(afterget, initdir)</div><div class="line">##  [1] &quot;9406922003_R01C01_Grn.idat&quot; &quot;9406922003_R01C01_Red.idat&quot;</div><div class="line">##  [3] &quot;9406922003_R02C01_Grn.idat&quot; &quot;9406922003_R02C01_Red.idat&quot;</div><div class="line">##  [5] &quot;9406922003_R03C02_Grn.idat&quot; &quot;9406922003_R03C02_Red.idat&quot;</div><div class="line">##  [7] &quot;9406922003_R04C02_Grn.idat&quot; &quot;9406922003_R04C02_Red.idat&quot;</div><div class="line">##  [9] &quot;9406922003_R05C01_Grn.idat&quot; &quot;9406922003_R05C01_Red.idat&quot;</div><div class="line">## [11] &quot;A-MEXP-2255.adf.txt&quot;        &quot;E-MTAB-5797.idf.txt&quot;       </div><div class="line">## [13] &quot;E-MTAB-5797.raw.1.zip&quot;      &quot;E-MTAB-5797.sdrf.txt&quot;</div></pre></td></tr></table></figure>
<p>在接下来的部分里，我们会讲解一下如何导入并处理这些数据。</p>
<h1 id="summarizedexperiment实例">SummarizedExperiment实例</h1>
<p>在使用微阵列的时代，目标核酸的是由微阵列里面的探针决定的。而在现在广泛使用的短读长测序方法则为目标核酸的定量提供了更大的灵活性。因此，Bioconductor针对大规模的基因检测设计了一种更加灵活的容器。这种容器的一个重要思路就是我们仅仅通过基因组的坐标就能对我们的目标核酸进行识别。这种容器容器可以使用户们仅仅使用基因组坐标就能方便地分析各种数据。通用方法<code>subsetByOverlaps</code>可以与<code>SummarizedExperiment</code>实例配合使用来实现这一目标。</p>
<h2 id="常规思路">常规思路</h2>
<p>应用于<code>SummarizedExperiment</code>的方法要比<code>ExpressionSet</code>的多，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div></pre></td><td class="code"><pre><div class="line">methods(class=<span class="string">"SummarizedExperiment"</span>)</div><div class="line"><span class="comment">##   [1] !=                [                 [[               </span></div><div class="line"><span class="comment">##   [4] [[&lt;-              [&lt;-               %in%             </span></div><div class="line"><span class="comment">##   [7] &lt;                 &lt;=                ==               </span></div><div class="line"><span class="comment">##  [10] &gt;                 &gt;=                $$                </span></div><div class="line"><span class="comment">##  [13] $$&lt;-               aggregate         anyDuplicated    </span></div><div class="line"><span class="comment">##  [16] anyNA             append            as.character     </span></div><div class="line"><span class="comment">##  [19] as.complex        as.data.frame     as.env           </span></div><div class="line"><span class="comment">##  [22] as.integer        as.list           as.logical       </span></div><div class="line"><span class="comment">##  [25] as.matrix         as.numeric        as.raw           </span></div><div class="line"><span class="comment">##  [28] assay             assay&lt;-           assayNames       </span></div><div class="line"><span class="comment">##  [31] assayNames&lt;-      assays            assays&lt;-         </span></div><div class="line"><span class="comment">##  [34] bindROWS          by                c                </span></div><div class="line"><span class="comment">##  [37] cbind             coerce            coerce&lt;-         </span></div><div class="line"><span class="comment">##  [40] colData           colData&lt;-         countOverlaps    </span></div><div class="line"><span class="comment">##  [43] dim               dimnames          dimnames&lt;-       </span></div><div class="line"><span class="comment">##  [46] duplicated        elementMetadata   elementMetadata&lt;-</span></div><div class="line"><span class="comment">##  [49] eval              expand            expand.grid      </span></div><div class="line"><span class="comment">##  [52] extractROWS       FactorToClass     findOverlaps     </span></div><div class="line"><span class="comment">##  [55] head              intersect         is.na            </span></div><div class="line"><span class="comment">##  [58] length            lengths           match            </span></div><div class="line"><span class="comment">##  [61] mcols             mcols&lt;-           merge            </span></div><div class="line"><span class="comment">##  [64] mergeROWS         metadata          metadata&lt;-       </span></div><div class="line"><span class="comment">##  [67] mstack            names             names&lt;-          </span></div><div class="line"><span class="comment">##  [70] NROW              overlapsAny       parallelSlotNames</span></div><div class="line"><span class="comment">##  [73] pcompare          rank              rbind            </span></div><div class="line"><span class="comment">##  [76] realize           relist            rename           </span></div><div class="line"><span class="comment">##  [79] rep               rep.int           replaceROWS      </span></div><div class="line"><span class="comment">##  [82] rev               rowData           rowData&lt;-        </span></div><div class="line"><span class="comment">##  [85] ROWNAMES          rowRanges         rowRanges&lt;-      </span></div><div class="line"><span class="comment">##  [88] selfmatch         seqlevelsInUse    setdiff          </span></div><div class="line"><span class="comment">##  [91] setequal          shiftApply        show             </span></div><div class="line"><span class="comment">##  [94] sort              split             split&lt;-          </span></div><div class="line"><span class="comment">##  [97] subset            subsetByOverlaps  summary          </span></div><div class="line"><span class="comment">## [100] table             tail              tapply           </span></div><div class="line"><span class="comment">## [103] transform         union             unique           </span></div><div class="line"><span class="comment">## [106] updateObject      values            values&lt;-         </span></div><div class="line"><span class="comment">## [109] window            window&lt;-          with             </span></div><div class="line"><span class="comment">## [112] xtabs             xtfrm            </span></div><div class="line"><span class="comment">## see '?methods' for accessing help and source code</span></div></pre></td></tr></table></figure>
<p>不过其中的一些关键方法的使用类似于<code>ExpressionSet</code>中的方法，如下所示：</p>
<ul>
<li><code>assay()</code>: 获取原始数值量化结果，这里需要注意的是，该方法支持多个分析结果，使用<code>assays()</code>可以以列表的形式获取多个结果</li>
<li><code>colData()</code>: 获取样本信息（从名字上来看，就是列信息）。</li>
<li><code>rowData()</code>: 获取表达值信息。当表达值信息是使用基因组坐标来表示时，则可以使用 <code>rowRanges()</code> 。</li>
<li><code>metadata()</code>: 获取可能包含相关实验的任何相关元数据的列表。</li>
</ul>
<h2 id="rna-seq实验">RNA-seq实验</h2>
<p>现在我们使用 <code>airway</code> 包来演示一下 <code>SummarizedExperiment</code>的概念，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(airway)</div><div class="line">data(airway)</div><div class="line">airway</div><div class="line"><span class="comment">## class: RangedSummarizedExperiment </span></div><div class="line"><span class="comment">## dim: 64102 8 </span></div><div class="line"><span class="comment">## metadata(1): ''</span></div><div class="line"><span class="comment">## assays(1): counts</span></div><div class="line"><span class="comment">## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></div><div class="line"><span class="comment">## colData names(9): SampleName cell ... Sample BioSample</span></div></pre></td></tr></table></figure>
<p>元数据(Metadata)是以列表的形式储存的，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">metadata(airway)</div><div class="line"><span class="comment">## [[1]]</span></div><div class="line"><span class="comment">## Experiment data</span></div><div class="line"><span class="comment">##   Experimenter name: Himes BE </span></div><div class="line"><span class="comment">##   Laboratory: NA </span></div><div class="line"><span class="comment">##   Contact information:  </span></div><div class="line"><span class="comment">##   Title: RNA-Seq transcriptome profiling identifies CRISPLD2 as a glucocorticoid responsive gene that modulates cytokine function in airway smooth muscle cells. </span></div><div class="line"><span class="comment">##   URL: http://www.ncbi.nlm.nih.gov/pubmed/24926665 </span></div><div class="line"><span class="comment">##   PMIDs: 24926665 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   Abstract: A 226 word abstract is available. Use 'abstract' method.</span></div></pre></td></tr></table></figure>
<p>这个数据集中的表达矩阵有64102行，它们的使用ENSEMBL数据库的命名方法定量了外显子的表达量，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">rowRanges(airway)</div><div class="line"><span class="comment">## GRangesList object of length 64102:</span></div><div class="line"><span class="comment">## $$ENSG00000000003</span></div><div class="line"><span class="comment">## GRanges object with 17 ranges and 2 metadata columns:</span></div><div class="line"><span class="comment">##        seqnames            ranges strand |   exon_id       exon_name</span></div><div class="line"><span class="comment">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;</span></div><div class="line"><span class="comment">##    [1]        X 99883667-99884983      - |    667145 ENSE00001459322</span></div><div class="line"><span class="comment">##    [2]        X 99885756-99885863      - |    667146 ENSE00000868868</span></div><div class="line"><span class="comment">##    [3]        X 99887482-99887565      - |    667147 ENSE00000401072</span></div><div class="line"><span class="comment">##    [4]        X 99887538-99887565      - |    667148 ENSE00001849132</span></div><div class="line"><span class="comment">##    [5]        X 99888402-99888536      - |    667149 ENSE00003554016</span></div><div class="line"><span class="comment">##    ...      ...               ...    ... .       ...             ...</span></div><div class="line"><span class="comment">##   [13]        X 99890555-99890743      - |    667156 ENSE00003512331</span></div><div class="line"><span class="comment">##   [14]        X 99891188-99891686      - |    667158 ENSE00001886883</span></div><div class="line"><span class="comment">##   [15]        X 99891605-99891803      - |    667159 ENSE00001855382</span></div><div class="line"><span class="comment">##   [16]        X 99891790-99892101      - |    667160 ENSE00001863395</span></div><div class="line"><span class="comment">##   [17]        X 99894942-99894988      - |    667161 ENSE00001828996</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 722 sequences (1 circular) from an unspecified genome</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## ...</span></div><div class="line"><span class="comment">## &lt;64101 more elements&gt;</span></div></pre></td></tr></table></figure>
<p>在以前使用微阵列的时候，我们可能习惯于基因水平的定量信息。但在这个案例中，我们使用的外显子水平的定量信息，因此这里需要一个特定的计算方法将其转换为基因的定量信息。例如，基因ORMDL3在ENSEMBL中的标记符为ENSG00000172057，那么在<code>SummarizedExperiment</code>对象中，它所对应的坐标如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">rowRanges(airway)$ENSG00000172057</div><div class="line"><span class="comment">## GRanges object with 20 ranges and 2 metadata columns:</span></div><div class="line"><span class="comment">##        seqnames            ranges strand |   exon_id       exon_name</span></div><div class="line"><span class="comment">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;</span></div><div class="line"><span class="comment">##    [1]       17 38077294-38078938      - |    549057 ENSE00001316037</span></div><div class="line"><span class="comment">##    [2]       17 38077296-38077570      - |    549058 ENSE00002684279</span></div><div class="line"><span class="comment">##    [3]       17 38077929-38078002      - |    549059 ENSE00002697088</span></div><div class="line"><span class="comment">##    [4]       17 38078351-38078552      - |    549060 ENSE00002718599</span></div><div class="line"><span class="comment">##    [5]       17 38078530-38078938      - |    549061 ENSE00001517665</span></div><div class="line"><span class="comment">##    ...      ...               ...    ... .       ...             ...</span></div><div class="line"><span class="comment">##   [16]       17 38081876-38083094      - |    549072 ENSE00001517669</span></div><div class="line"><span class="comment">##   [17]       17 38082049-38082592      - |    549073 ENSE00002700064</span></div><div class="line"><span class="comment">##   [18]       17 38083070-38083099      - |    549074 ENSE00002688012</span></div><div class="line"><span class="comment">##   [19]       17 38083320-38083482      - |    549075 ENSE00002700595</span></div><div class="line"><span class="comment">##   [20]       17 38083737-38083854      - |    549076 ENSE00001283878</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 722 sequences (1 circular) from an unspecified genome</span></div></pre></td></tr></table></figure>
<p>另外，<code>GenomicRanges</code>的工作类似于<code>rowRanges()</code> 。 如果要查看在这个列表中外显子坐标是否存在重叠区域，可以使用 <code>reduce()</code>方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">reduce(rowRanges(airway)$ENSG00000172057)</div><div class="line"><span class="comment">## GRanges object with 8 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames            ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]       17 38077294-38078938      -</span></div><div class="line"><span class="comment">##   [2]       17 38079365-38079516      -</span></div><div class="line"><span class="comment">##   [3]       17 38080283-38080478      -</span></div><div class="line"><span class="comment">##   [4]       17 38081008-38081058      -</span></div><div class="line"><span class="comment">##   [5]       17 38081422-38081624      -</span></div><div class="line"><span class="comment">##   [6]       17 38081876-38083099      -</span></div><div class="line"><span class="comment">##   [7]       17 38083320-38083482      -</span></div><div class="line"><span class="comment">##   [8]       17 38083737-38083854      -</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 722 sequences (1 circular) from an unspecified genome</span></div></pre></td></tr></table></figure>
<p>这表明从外显子集合投身到基因组中的信息有可能包含8个被转录的亚区域。关于如何将这些信息转换为基因水平的定量信息，可以参考其它章节 <a href="http://genomicsclass.github.io/book/pages/rnaseq_gene_level.html" target="_blank" rel="external">elsewhere in the genomicsclass book.</a></p>
<p>除了对表达值进行详细地注释外，我们还需要<code>colData()</code>来处理样本信息。其中 <code>$</code> 符号可以从列中提取样本信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">names(colData(airway))</div><div class="line"><span class="comment">## [1] "SampleName" "cell"       "dex"        "albut"      "Run"       </span></div><div class="line"><span class="comment">## [6] "avgLength"  "Experiment" "Sample"     "BioSample"</span></div><div class="line">table(airway$dex) <span class="comment"># main treatment factor</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   trt untrt </span></div><div class="line"><span class="comment">##     4     4</span></div></pre></td></tr></table></figure>
<h2 id="处理由illumina-450k甲基化芯片生成的数据">处理由Illumina 450k甲基化芯片生成的数据</h2>
<p><code>SummarizedExperiment</code> 类设计出来用于储存所有类似的芯片数据或短序列读长测序数据。前面我们提到的 <code>getAE</code> 用于从ArrayExpress提取了一批文件，这批文件是胶质母细胞瘤组织中的甲基化定量数据。</p>
<p>而样本信息储存在 <code>sdrf.txt</code> 文件中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(data.table)</div><div class="line">sd5797 = fread(<span class="string">"E-MTAB-5797.sdrf.txt"</span>)</div><div class="line">head(sd5797[,c(<span class="number">3</span>,<span class="number">16</span>,<span class="number">18</span>)])</div><div class="line"><span class="comment">##    Characteristics[age] Label                   Performer</span></div><div class="line"><span class="comment">## 1:                   58   Cy3 IntegraGen SA, Evry, France</span></div><div class="line"><span class="comment">## 2:                   58   Cy5 IntegraGen SA, Evry, France</span></div><div class="line"><span class="comment">## 3:                   72   Cy3 IntegraGen SA, Evry, France</span></div><div class="line"><span class="comment">## 4:                   72   Cy5 IntegraGen SA, Evry, France</span></div><div class="line"><span class="comment">## 5:                   70   Cy3 IntegraGen SA, Evry, France</span></div><div class="line"><span class="comment">## 6:                   70   Cy5 IntegraGen SA, Evry, France</span></div></pre></td></tr></table></figure>
<p>原始表达数据储存在 <code>idat</code> 文件中。现在我们使用 <code>mnfi</code> 包中的<code>read.metharray()</code> 函数将其提取出来，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(minfi)</div><div class="line">pref = unique(substr(dir(patt=<span class="string">"idat"</span>),<span class="number">1</span>,<span class="number">17</span>)) <span class="comment"># find the prefix strings</span></div><div class="line">raw = read.metharray(pref)</div><div class="line">raw</div><div class="line"><span class="comment">## class: RGChannelSet </span></div><div class="line"><span class="comment">## dim: 622399 5 </span></div><div class="line"><span class="comment">## metadata(0):</span></div><div class="line"><span class="comment">## assays(2): Green Red</span></div><div class="line"><span class="comment">## rownames(622399): 10600313 10600322 ... 74810490 74810492</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(5): 9406922003_R01C01 9406922003_R02C01 9406922003_R03C02</span></div><div class="line"><span class="comment">##   9406922003_R04C02 9406922003_R05C01</span></div><div class="line"><span class="comment">## colData names(0):</span></div><div class="line"><span class="comment">## Annotation</span></div><div class="line"><span class="comment">##   array: IlluminaHumanMethylation450k</span></div><div class="line"><span class="comment">##   annotation: ilmn12.hg19</span></div></pre></td></tr></table></figure>
<p>现在已经出现了很多算法可以将原始数值转换为生物学上可以解释的相对甲基化数值。在这个案例中我们使用分位数归一化算法(quantile normalization)来将红色和绿色的信号转换为相对甲基化数值(M)，并且估计<code>SummarizedExperiment</code> 实例中的拷贝数(CN, copy number)，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">glioMeth = preprocessQuantile(raw) <span class="comment"># generate SummarizedExperiment</span></div><div class="line"><span class="comment"># Require IlluminaHumanMethylation450kmanifest packages</span></div><div class="line"><span class="comment"># BiocManager::install("IlluminaHumanMethylation450kmanifest")</span></div><div class="line"><span class="comment"># BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19")</span></div><div class="line"><span class="comment">## [preprocessQuantile] Mapping to genome.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Fixing outliers.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Quantile normalizing.</span></div><div class="line">glioMeth</div><div class="line"><span class="comment">## class: GenomicRatioSet </span></div><div class="line"><span class="comment">## dim: 485512 5 </span></div><div class="line"><span class="comment">## metadata(0):</span></div><div class="line"><span class="comment">## assays(2): M CN</span></div><div class="line"><span class="comment">## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(5): 9406922003_R01C01 9406922003_R02C01 9406922003_R03C02</span></div><div class="line"><span class="comment">##   9406922003_R04C02 9406922003_R05C01</span></div><div class="line"><span class="comment">## colData names(3): xMed yMed predictedSex</span></div><div class="line"><span class="comment">## Annotation</span></div><div class="line"><span class="comment">##   array: IlluminaHumanMethylation450k</span></div><div class="line"><span class="comment">##   annotation: ilmn12.hg19</span></div><div class="line"><span class="comment">## Preprocessing</span></div><div class="line"><span class="comment">##   Method: Raw (no normalization or bg correction)</span></div><div class="line"><span class="comment">##   minfi version: 1.31.1</span></div><div class="line"><span class="comment">##   Manifest version: 0.4.0</span></div></pre></td></tr></table></figure>
<p>在后面部分中，我们会来讲解一下如何处理这些数据。</p>
<h1 id="外部数据储存hdf5array-savehdf5summarizedexperiment">外部数据储存—HDF5Array, saveHDF5SummarizedExperiment</h1>
<h2 id="测试内存占用">测试内存占用</h2>
<p>在常规的交互式使用中，R数据都是驻留在内存中。我们可以使用 <code>gc()</code>函数来估计一下某个任务占用的内存。Ncell所占用的空间用于处理语言结构，例如解析树(parse tree)和名称空间(namespaces)，Vcell的空间用于储存会话中加载的数字和字符数据，现在我们打开R，查看一下，如下所示：</p>
<p>专用于Ncell的空间用于处理语言结构，如解析树和名称空间，而专用于Vcell的空间用于存储会话中加载的数字和字符数据。在我的MacBook Air上，R的香草启动收益率</p>
<p>Space devoted to Ncells is used to deal with language constructs such as parse trees and namespaces, while space devoted to Vcells is used to store numerical and character data loaded in the session. On my macbook air, a vanilla startup of R yields</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; gc()</div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">10410962</span> <span class="number">556.1</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">38615493</span> <span class="number">294.7</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div></pre></td></tr></table></figure>
<p>当我们加载了 <code>airway</code> 包后，再查看一下，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; suppressMessages(&#123;library(airway)&#125;)</div><div class="line">&gt; gc()</div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells 10410859 556.0   18810631 1004.6  16638654 888.6</div><div class="line">Vcells 38613228 294.6  128861398  983.2 128847953 983.1</div></pre></td></tr></table></figure>
<p>现在加载 <code>airway</code> 这个 SummarizedExperiment实例，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; data(airway)</div><div class="line">&gt; gc()</div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells 11229539 599.8   18810631 1004.6  16638654 888.6</div><div class="line">Vcells 42534679 324.6  128861398  983.2 128847953 983.1</div><div class="line">&gt; dim(airway)</div><div class="line">[1] 64102     8</div></pre></td></tr></table></figure>
<p>随着加载数据的增多，内存的占用也在增大。使用 <code>rm()</code> 函数可以清除一些不使用的对象，从而释放内存，不过在实际分析过程中很少这样做，或者说这样做效率并不高。</p>
<h2 id="hdf5用于外部存储">HDF5用于外部存储</h2>
<p><a href="https://blog.csdn.net/mzpmzk/article/details/89188968" target="_blank" rel="external">HDF5</a> 的全称是Hierarchical Data Format Version 5，即层次性数据格式第五版，它是一种广泛用于储存数值型的数据结构，并且提供了各种科学编程语言的接口。<code>HDF5Array</code> 包简化了管理大量数据型数组的操作，它的使用如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(airway)</div><div class="line"><span class="keyword">library</span>(HDF5Array)  <span class="comment"># setup for external serialization</span></div><div class="line"><span class="comment">## Loading required package: rhdf5</span></div><div class="line">data(airway)</div><div class="line">airass = assay(airway)  <span class="comment"># obtain numerical data, then save as HDF5</span></div><div class="line">href = writeHDF5Array(airass, <span class="string">"airass.h5"</span>, <span class="string">"airway"</span>)</div></pre></td></tr></table></figure>
<p>现在我们可以发现，当我们访问相同的数值型数据时，内存的占用并没有增长，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">&gt; gc()  <span class="comment"># after attaching HDF5Array package</span></div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">11246377</span> <span class="number">600.7</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">42828456</span> <span class="number">326.8</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div><div class="line">&gt; myd = HDF5Array(<span class="string">"airass.h5"</span>, <span class="string">"airway"</span>)  <span class="comment"># get reference to data</span></div><div class="line">&gt; gc()</div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">11246488</span> <span class="number">600.7</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">42828777</span> <span class="number">326.8</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div><div class="line">&gt; dim(myd)</div><div class="line">[<span class="number">1</span>] <span class="number">64102</span>     <span class="number">8</span></div></pre></td></tr></table></figure>
<p>这样运行起来一切还都正常，不过我们更常用 <code>SummarizedExperiment</code> 容器来操作airway数据，也就是我们下面要讲解的内容。</p>
<h2 id="summarizedexperiment支持hdf5">SummarizedExperiment支持HDF5</h2>
<p>针对某个特定的 SummarizedExperiment 实例，<code>saveHDF5SummarizedExperiment</code>能将数据与元素数据汇集在一起，控制内存占用，同时保留容器中丰度的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">saveHDF5SummarizedExperiment(airway, <span class="string">"externalAirway"</span>, replace=<span class="literal">TRUE</span>)</div><div class="line">newse = loadHDF5SummarizedExperiment(<span class="string">"externalAirway"</span>)</div><div class="line">newse</div><div class="line"><span class="comment">## class: RangedSummarizedExperiment </span></div><div class="line"><span class="comment">## dim: 64102 8 </span></div><div class="line"><span class="comment">## metadata(1): ''</span></div><div class="line"><span class="comment">## assays(1): counts</span></div><div class="line"><span class="comment">## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></div><div class="line"><span class="comment">## colData names(9): SampleName cell ... Sample BioSample</span></div><div class="line">assay(newse[c(<span class="string">"ENSG00000000005"</span>, <span class="string">"LRG_99"</span>), </div><div class="line">        which(newse$dex == <span class="string">"trt"</span>)]) <span class="comment"># use familiar subsetting</span></div><div class="line"><span class="comment">## &lt;2 x 4&gt; DelayedMatrix object of type "integer":</span></div><div class="line"><span class="comment">##                 SRR1039509 SRR1039513 SRR1039517 SRR1039521</span></div><div class="line"><span class="comment">## ENSG00000000005          0          0          0          0</span></div><div class="line"><span class="comment">## LRG_99                   0          0          0          0</span></div></pre></td></tr></table></figure>
<p>在这个案例中，处理SummarizedExperiment元数据的操作抵抗了外部化这些数据的优势，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">&gt; gc() <span class="comment"># after loading SummarizedExperiment</span></div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">11252592</span> <span class="number">601.0</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">42840344</span> <span class="number">326.9</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div><div class="line">&gt; gc() <span class="comment"># after loading SummarizedExperiment</span></div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">11252592</span> <span class="number">601.0</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">42840344</span> <span class="number">326.9</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div><div class="line">&gt; newse = loadHDF5SummarizedExperiment(<span class="string">"externalAirway"</span>)</div><div class="line">&gt; gc()</div><div class="line">           used  (Mb) gc trigger   (Mb)  max used  (Mb)</div><div class="line">Ncells <span class="number">11252922</span> <span class="number">601.0</span>   <span class="number">18810631</span> <span class="number">1004.6</span>  <span class="number">16638654</span> <span class="number">888.6</span></div><div class="line">Vcells <span class="number">44883257</span> <span class="number">342.5</span>  <span class="number">128861398</span>  <span class="number">983.2</span> <span class="number">128847953</span> <span class="number">983.1</span></div></pre></td></tr></table></figure>
<p>对那些大量的数据，支持HDF5的SummarizedExperiment会在内存不够的情况下，提供灵活的计算。</p>
<h1 id="genomicfiles某个特定类型的文件家族">GenomicFiles某个特定类型的文件家族</h1>
<p><code>GenomicFiles</code> 是一个用于管理文件集合的包。当我们不想整体对某批数据进行解析和建模时，这个包就比较重要了，它可以导入某些文件的部分内容。</p>
<p><code>GenomicFiles</code> 类的实例有<code>rowRanges</code> 和 <code>colData</code> 方法。我们可以在 <code>GenomicFiles</code> 实例的 <code>colData</code> 组件中提取有关样本或实验的元数据，并且可以使用 <code>rowRanges</code> 组件来查看感兴趣的基因组间隔信息。</p>
<h2 id="bam-文件集">BAM 文件集</h2>
<p>现在我们使用一个案例来说明一下这个文件集。我们所使用的案例是由 <a href="http://dx.doi.org/10.1016/j.cell.2012.12.023" target="_blank" rel="external">Zarnack团队</a>在2013年进行的一个HNRNPC敲除后的RNA-seq测序项目。在这个项目中，一共有8个Hela细胞样本，其中4个是野生型，4个使用了RNAi敲低了HNRNPC，这个基因位于14号染色体上。</p>
<p><code>RNAseqData.HNRNPC.bam.chr14</code> 中已经包含了比对好的读长。BAM文件被储存在了一个名为 <code>RNAseqData.HNRNPC.bam.chr14_BAMFILES</code> 的向量中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line"><span class="keyword">library</span>(GenomicFiles)</div><div class="line">gf = GenomicFiles(files=RNAseqData.HNRNPC.bam.chr14_BAMFILES)</div><div class="line">gf</div><div class="line"><span class="comment">## GenomicFiles object with 0 ranges and 8 files: </span></div><div class="line"><span class="comment">## files: ERR127306_chr14.bam, ERR127307_chr14.bam, ..., ERR127304_chr14.bam, ERR127305_chr14.bam </span></div><div class="line"><span class="comment">## detail: use files(), rowRanges(), colData(), ...</span></div></pre></td></tr></table></figure>
<p>我们可以通过将感兴趣的区域绑定到对象中，从而增强一个文件的信息，现在我们使用 <code>GRanges</code> 来检索HNRNPC，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hn = GRanges(&quot;chr14&quot;, IRanges(21677296, 21737638), strand=&quot;-&quot;)</div><div class="line">rowRanges(gf) = hn</div></pre></td></tr></table></figure>
<p>如果要提取与感兴趣的区域重叠的比对信息(alignments)，我们可以使用<code>GenomicFiles</code>包中的 <code>reduceByRange</code> 方法。我们需要定义一个映射函数来加入到 <code>reduceByRange</code> 函数中。这个映射函数我们在下面会定义为 <code>MAP()</code> ， 它的参数有两个，其中 <code>r</code> 表示感兴趣的范围，<code>f</code> 指的是，被解析的文件中比对信息的重叠范围。在<code>MAP()</code> 函数中我们还使用了 <code>readGAlignmentPairs</code> 函数，这是因为我们霜肆处理配对的末端测序数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line">MAP = <span class="keyword">function</span>(r, f) </div><div class="line">    readGAlignmentPairs(f, param=ScanBamParam(which=r))</div><div class="line">ali = reduceByRange(gf, MAP=MAP)</div><div class="line"><span class="comment">## Warning in .make_GAlignmentPairs_from_GAlignments(gal, strandMode = strandMode, :   4 alignments with ambiguous pairing were dumped.</span></div><div class="line"><span class="comment">##     Use 'getDumpedAlignments()' to retrieve them from the dump environment.</span></div><div class="line">sapply(ali[[<span class="number">1</span>]], length)</div><div class="line"><span class="comment">## ERR127306 ERR127307 ERR127308 ERR127309 ERR127302 ERR127303 ERR127304 </span></div><div class="line"><span class="comment">##      2711      3160      2946      2779        86        98       158 </span></div><div class="line"><span class="comment">## ERR127305 </span></div><div class="line"><span class="comment">##       141</span></div></pre></td></tr></table></figure>
<p>在这个案例中，我们大致比较了一下这8个样本的目标范围，对于前4个样本来说，有更多的读长与HNRNPC区域比对上了，因此我们可以知道前4个样本是野生型细胞，后4个样本有较少的读长与HNRNPC区域比对上，这表明它们是敲低的HNRNPC。这里我们注意的是，敲低的效率并不是100%的，也就是说，我们仍然有可能会检测到有HNRNPC的一些转录本。</p>
<h2 id="bed-文件集">BED 文件集</h2>
<p><code>erma</code> 这个包设计用来演示在内存不足的情况下，基于表观遗传学中各种细胞类型数据分析，如何打包大量的数据。 这个包创建的基础容器是对 <code>GenomicFiles</code> 的简单扩展，它使用 <code>makeErmaSet()</code> 函数进行构建，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(erma)</div><div class="line">erset = makeErmaSet()</div><div class="line"><span class="comment">## <span class="doctag">NOTE:</span> input data had non-ASCII characters replaced by ' '.</span></div><div class="line">erset</div><div class="line"><span class="comment">## ErmaSet object with 0 ranges and 31 files: </span></div><div class="line"><span class="comment">## files: E002_25_imputed12marks_mnemonics.bed.gz, E003_25_imputed12marks_mnemonics.bed.gz, ..., E088_25_imputed12marks_mnemonics.bed.gz, E096_25_imputed12marks_mnemonics.bed.gz </span></div><div class="line"><span class="comment">## detail: use files(), rowRanges(), colData(), ... </span></div><div class="line"><span class="comment">## cellTypes() for type names; data(short_celltype) for abbr.</span></div></pre></td></tr></table></figure>
<p>我们可以使用 <code>colData</code> 来查看样本信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div></pre></td><td class="code"><pre><div class="line">colData(erset)</div><div class="line"><span class="comment">## (showing narrow slice of  31 x 95  DataFrame)   narrDF with 31 rows and 6 columns</span></div><div class="line"><span class="comment">##      Epigenome.ID..EID.          GROUP  Epigenome.Mnemonic     ANATOMY</span></div><div class="line"><span class="comment">##             &lt;character&gt;    &lt;character&gt;         &lt;character&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">## E002               E002            ESC             ESC.WA7         ESC</span></div><div class="line"><span class="comment">## E003               E003            ESC              ESC.H1         ESC</span></div><div class="line"><span class="comment">## E021               E021           iPSC         IPSC.DF.6.9        IPSC</span></div><div class="line"><span class="comment">## E032               E032   HSC &amp; B-cell        BLD.CD19.PPC       BLOOD</span></div><div class="line"><span class="comment">## E033               E033 Blood &amp; T-cell         BLD.CD3.CPC       BLOOD</span></div><div class="line"><span class="comment">## ...                 ...            ...                 ...         ...</span></div><div class="line"><span class="comment">## E071               E071          Brain        BRN.HIPP.MID       BRAIN</span></div><div class="line"><span class="comment">## E072               E072          Brain         BRN.INF.TMP       BRAIN</span></div><div class="line"><span class="comment">## E073               E073          Brain BRN.DL.PRFRNTL.CRTX       BRAIN</span></div><div class="line"><span class="comment">## E088               E088          Other             LNG.FET        LUNG</span></div><div class="line"><span class="comment">## E096               E096          Other                 LNG        LUNG</span></div><div class="line"><span class="comment">##                TYPE            SEX</span></div><div class="line"><span class="comment">##         &lt;character&gt;    &lt;character&gt;</span></div><div class="line"><span class="comment">## E002 PrimaryCulture         Female</span></div><div class="line"><span class="comment">## E003 PrimaryCulture           Male</span></div><div class="line"><span class="comment">## E021 PrimaryCulture           Male</span></div><div class="line"><span class="comment">## E032    PrimaryCell           Male</span></div><div class="line"><span class="comment">## E033    PrimaryCell        Unknown</span></div><div class="line"><span class="comment">## ...             ...            ...</span></div><div class="line"><span class="comment">## E071  PrimaryTissue           Male</span></div><div class="line"><span class="comment">## E072  PrimaryTissue          Mixed</span></div><div class="line"><span class="comment">## E073  PrimaryTissue          Mixed</span></div><div class="line"><span class="comment">## E088  PrimaryTissue Female/Unknown</span></div><div class="line"><span class="comment">## E096  PrimaryTissue         Female</span></div><div class="line"><span class="comment">## use colnames() for full set of metadata attributes.</span></div></pre></td></tr></table></figure>
<p>我们可以使用熟悉的简捷方式来查看样本元数据的表格信息：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">table(erset$ANATOMY)</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##    BLOOD    BRAIN      ESC      FAT     IPSC    LIVER     LUNG     SKIN </span></div><div class="line"><span class="comment">##       15        7        2        1        1        1        2        1 </span></div><div class="line"><span class="comment">## VASCULAR </span></div><div class="line"><span class="comment">##        1</span></div></pre></td></tr></table></figure>
<p>因此我们可以在R中通过轻量级的接口（仅限于关于示例和文件路径的元数据）来访问大量的BED文件集。如果这些文件是用tabix建立索引的，那么我们就可以非常快速地针对性地检索特定范围的数据，我们可以使用 <code>stateProfile</code> 函数来说明这一点，如下所示：</p>
<blockquote>
<p>注：tabix是第一个能够对排序好的TAB分割数据进行处理，例如GEF，BED，PSL，SAM与SQL的输出数据等，它能快速地检索特定区域的重合部分。具体的使用方法可以参考本文的参考资料。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">stateProfile(erset[,<span class="number">26</span>:<span class="number">31</span>], shortCellType=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">## Scale for 'y' is already present. Adding another scale for 'y', which</span></div><div class="line"><span class="comment">## will replace the existing scale.</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/dataman2019-lkstpr-1.png" alt="plot of chunk lkstpr">
<p class="caption">plot of chunk lkstpr</p>
</div>
<p>这张图描述在chr9的正义链上，基因IL33的启动子区域的表观遗传状态中的的解剖位点之间的变异。需要注意的是，胎儿肺样本似乎说明，与成人肺样本在这个区域的静默相比，胎儿肺样本则表现出增强子活性。。</p>
<h1 id="管理大量的dna变异信息">管理大量的DNA变异信息</h1>
<h2 id="vcf背景">VCF背景</h2>
<p>处理大量SNP基因型（或小indels）的最常用方法就是处理VCF文件（Variant Call Format）。从Wikipedia或相关文档，例如<a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf" target="_blank" rel="external">spec</a>和<a href="http://vcftools.sourceforge.net/VCF-poster.pdf" target="_blank" rel="external">diagram</a>很容易查看VCF文件的内容。VCF文件的基本设计中含有一个头部信息，这个头部信息提供了关于VCF文件的元数据，并且包括了每个基因组的突变体(variant)信息。这些突变体信息记录了突变体的性质（那些已经检测出来与参考基因组的差异信息），以及包括用于描述存在于突变体结构(variant configuration)中的样本字段(field)。一些突变体是已经确定的，而一些突变体则无法确定，在这种情况下，基因型的可能性就会被记录下来。突变体也许是阶段性的(phased)，这就意味着，在某条具体的染色体上，突变的位置也有可能不同。</p>
<blockquote>
<p>后两句话不太理解，这里放上原文：Some variants are “called”, others may be uncertain and in this case genotype likelihoods are recorded. Variants may also be <em>phased</em>, meaning that it is possible to locate different variant events on a given chromosome.</p>
</blockquote>
<p><code>VariantAnnotation</code> 包定义了处理这种格式的基础工具。<code>Rsamtools</code> 包有一个处理这些文件的tabix索引程序，因此也可以使用这个包。</p>
<h2 id="基因组vcf">*1000基因组VCF</h2>
<p>DNA突变体的一个明显案例就是<a href="http://www.internationalgenome.org/" target="_blank" rel="external">1000 genomes project</a>。Bioconductr的 <code>ldblock</code> 包中有一个工具可以处理VCF文件，它能处理压缩的，由tabix建立索引的VCF文件，这个项目位于Amzon的云上，操作如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ldblock)</div><div class="line">sta = stack1kg()</div><div class="line">sta</div><div class="line"><span class="comment">## VcfStack object with 22 files and 2548 samples</span></div><div class="line"><span class="comment">## Seqinfo object with 22 sequences from an unspecified genome; no seqlengths </span></div><div class="line"><span class="comment">## use 'readVcfStack()' to extract VariantAnnotation VCF.</span></div></pre></td></tr></table></figure>
<p>这里注意一下，创建的基因组标记为<code>b37</code>。这是一个不太寻常的特征，它绑定到了VCF突变体数据中的元数据里，无法被改变。</p>
<p>我们将在此文件中添加有关样本地理来源的信息。正好 <code>ph525x</code> 这个包中就包含这些地理信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ph525x)</div><div class="line">data(sampInfo_1kg)</div><div class="line">rownames(sampInfo_1kg) = sampInfo_1kg[,<span class="number">1</span>]</div><div class="line">cd = sampInfo_1kg[ colnames(sta), ]</div><div class="line">colData(sta) = DataFrame(cd)</div></pre></td></tr></table></figure>
<p>注：这一部分的代码运行出错，经检索后，可能是由于网络问题。</p>
<h2 id="输入并处理vcf数据">输入并处理VCF数据</h2>
<p>现在我们使用 <code>GenomicFiles</code> 包中的 <code>readVcfStack</code> 函数来提取VCF中的储存的内容。我们提取的内容是针对ORMDL3的编码区，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(erma)</div><div class="line">orm = range(genemodel(<span class="string">"ORMDL3"</span>)) <span class="comment"># quick lookup</span></div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line">genome(orm) = <span class="string">"b37"</span>  <span class="comment"># must agree with VCF</span></div><div class="line">seqlevelsStyle(orm) = <span class="string">"NCBI"</span>  <span class="comment"># use integer chromosome codes</span></div><div class="line">ormRead = readVcfStack(sta, orm)</div><div class="line">ormRead</div><div class="line"><span class="comment">## class: CollapsedVCF </span></div><div class="line"><span class="comment">## dim: 3 2548 </span></div><div class="line"><span class="comment">## rowRanges(vcf):</span></div><div class="line"><span class="comment">##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></div><div class="line"><span class="comment">## info(vcf):</span></div><div class="line"><span class="comment">##   DataFrame with 12 columns: AF, AC, NS, AN, EAS_AF, EUR_AF, AFR_AF, AM...</span></div><div class="line"><span class="comment">## info(header(vcf)):</span></div><div class="line"><span class="comment">##              Number Type    Description                                   </span></div><div class="line"><span class="comment">##    AF        A      Float   Estimated allele frequency in the range (0,1) </span></div><div class="line"><span class="comment">##    AC        A      Integer Total number of alternate alleles in called...</span></div><div class="line"><span class="comment">##    NS        1      Integer Number of samples with data                   </span></div><div class="line"><span class="comment">##    AN        1      Integer Total number of alleles in called genotypes   </span></div><div class="line"><span class="comment">##    EAS_AF    A      Float   Allele frequency in the EAS populations cal...</span></div><div class="line"><span class="comment">##    EUR_AF    A      Float   Allele frequency in the EUR populations cal...</span></div><div class="line"><span class="comment">##    AFR_AF    A      Float   Allele frequency in the AFR populations cal...</span></div><div class="line"><span class="comment">##    AMR_AF    A      Float   Allele frequency in the AMR populations cal...</span></div><div class="line"><span class="comment">##    SAS_AF    A      Float   Allele frequency in the SAS populations cal...</span></div><div class="line"><span class="comment">##    VT        .      String  indicates what type of variant the line rep...</span></div><div class="line"><span class="comment">##    EX_TARGET 0      Flag    indicates whether a variant is within the e...</span></div><div class="line"><span class="comment">##    DP        1      Integer Approximate read depth; some reads may have...</span></div><div class="line"><span class="comment">## geno(vcf):</span></div><div class="line"><span class="comment">##   SimpleList of length 1: GT</span></div><div class="line"><span class="comment">## geno(header(vcf)):</span></div><div class="line"><span class="comment">##       Number Type   Description    </span></div><div class="line"><span class="comment">##    GT 1      String Phased Genotype</span></div></pre></td></tr></table></figure>
<p><code>CollapsedVCF</code> 类用于管理输入的基因型数据。现在我们检索3个突变体的信息。我们将得到的结果的子集转换为 <code>VariantTools</code> 所代表的<code>VRanges</code> 形式，从而可以更加清楚地了解到内容，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(VariantTools)</div><div class="line">vr = as(ormRead[,<span class="number">1</span>:<span class="number">5</span>], <span class="string">"VRanges"</span>)</div><div class="line">vr[<span class="number">1</span>:<span class="number">3</span>,]</div><div class="line"><span class="comment">## VRanges object with 3 ranges and 14 metadata columns:</span></div><div class="line"><span class="comment">##                     seqnames            ranges strand         ref</span></div><div class="line"><span class="comment">##                        &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG       17          38078372      *           T</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T       17 38078393-38078395      *         TGG</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT       17          38078398      *           G</span></div><div class="line"><span class="comment">##                                  alt     totalDepth       refDepth</span></div><div class="line"><span class="comment">##                     &lt;characterOrRle&gt; &lt;integerOrRle&gt; &lt;integerOrRle&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG               TG           &lt;NA&gt;           &lt;NA&gt;</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T                T           &lt;NA&gt;           &lt;NA&gt;</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT              GGT           &lt;NA&gt;           &lt;NA&gt;</span></div><div class="line"><span class="comment">##                           altDepth   sampleNames softFilterMatrix |</span></div><div class="line"><span class="comment">##                     &lt;integerOrRle&gt; &lt;factorOrRle&gt;         &lt;matrix&gt; |</span></div><div class="line"><span class="comment">##    17:38078372_T/TG           &lt;NA&gt;       HG00096                  |</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T           &lt;NA&gt;       HG00096                  |</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT           &lt;NA&gt;       HG00096                  |</span></div><div class="line"><span class="comment">##                          QUAL            AF            AC        NS</span></div><div class="line"><span class="comment">##                     &lt;numeric&gt; &lt;NumericList&gt; &lt;IntegerList&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG      &lt;NA&gt;          0.01            61      2548</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T      &lt;NA&gt;          0.04           191      2548</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT      &lt;NA&gt;          0.05           272      2548</span></div><div class="line"><span class="comment">##                            AN        EAS_AF        EUR_AF        AFR_AF</span></div><div class="line"><span class="comment">##                     &lt;integer&gt; &lt;NumericList&gt; &lt;NumericList&gt; &lt;NumericList&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG      5096             0             0          0.04</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T      5096             0          0.01          0.13</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT      5096          0.01          0.04          0.13</span></div><div class="line"><span class="comment">##                            AMR_AF        SAS_AF              VT EX_TARGET</span></div><div class="line"><span class="comment">##                     &lt;NumericList&gt; &lt;NumericList&gt; &lt;CharacterList&gt; &lt;logical&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG             0             0           INDEL     FALSE</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T          0.01          0.01           INDEL     FALSE</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT          0.02          0.03           INDEL     FALSE</span></div><div class="line"><span class="comment">##                            DP          GT</span></div><div class="line"><span class="comment">##                     &lt;integer&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##    17:38078372_T/TG    240289         0|0</span></div><div class="line"><span class="comment">##   17:38078393_TGG/T    222931         0|0</span></div><div class="line"><span class="comment">##   17:38078398_G/GGT    218409         0|0</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div><div class="line"><span class="comment">##   hardFilters: NULL</span></div></pre></td></tr></table></figure>
<p>从上面结果我们可以看出来，我们从VCF中提取了完整的信息，如果要列表出某个人的基因型，可以按下面进行操作，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">table(vr[which(sampleNames(vr)==<span class="string">"HG00096"</span>),]$GT)</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## 0|0 </span></div><div class="line"><span class="comment">##   3</span></div></pre></td></tr></table></figure>
<p>总结</p>
<ul>
<li>VCF文件储存队列(cohorts)的突变体信息，并且通过<code>VariantAnnotation</code>包可以导入这样的文件。</li>
<li>染色体特异性的VCF文件可以储存在VcfStack类中。</li>
<li><code>colData</code>可以绑定到VcfStack文件中，并将样本信息与其基因型对应起来。</li>
<li>VariantTools中的VRanges类可以用于生成丰富的元数据表示的单个样本-突变体表格(metadata-rich variant-by-individual tables)。</li>
</ul>
<h1 id="多组学multiassayexperiment以tcga为例说明">多组学：MultiAssayExperiment，以TCGA为例说明</h1>
<p><a href="https://cancergenome.nih.gov/" target="_blank" rel="external">癌基因组计划(The Cancer Genome Atlas)</a> 收集了来自于14000个人的数据，提供了29个不同肿瘤位点的基因组信息。我们可以通过位于CUNY的<a href="http://waldronlab.org/" target="_blank" rel="external">Levi Waldron实验室</a>实验室下载那些关于多型神经胶质细胞瘤的公共数据。Ｒ中的<code>MultiAssayExperiment</code> 类对象已经储存于Amazon S3中，并且<a href="http://tinyurl.com/MAEOurls" target="_blank" rel="external">Goolge表格</a>提供了详细地说明信息与链接。</p>
<h2 id="检索神经胶质细胞瘤数据">检索神经胶质细胞瘤数据</h2>
<p>现在我们来检索一下关于GBM的数据，其些软件可能要使用 <code>updateObject</code>，操作如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(curatedTCGAData)</div><div class="line">gbm = curatedTCGAData(<span class="string">"GBM"</span>, c(<span class="string">"Mutation"</span>, <span class="string">"CNASNP"</span>, <span class="string">"RNASeq2GeneNorm"</span>,</div><div class="line">   <span class="string">"Methylation_methyl27-20160128"</span>), dry.run=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## snapshotDate(): 2019-08-13</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## see ?curatedTCGAData and browseVignettes('curatedTCGAData') for documentation</span></div><div class="line"><span class="comment">## downloading 0 resources</span></div><div class="line"><span class="comment">## loading from cache</span></div><div class="line"><span class="comment">## harmonizing input:</span></div><div class="line"><span class="comment">##   removing 6320 sampleMap rows not in names(experiments)</span></div><div class="line"><span class="comment">##   removing 4 colData rownames not in sampleMap 'primary'</span></div><div class="line">gbm</div><div class="line"><span class="comment">## A MultiAssayExperiment object of 3 listed</span></div><div class="line"><span class="comment">##  experiments with user-defined names and respective classes. </span></div><div class="line"><span class="comment">##  Containing an ExperimentList class object of length 3: </span></div><div class="line"><span class="comment">##  [1] GBM_CNASNP-20160128: RaggedExperiment with 602338 rows and 1104 columns </span></div><div class="line"><span class="comment">##  [2] GBM_Mutation-20160128: RaggedExperiment with 22073 rows and 290 columns </span></div><div class="line"><span class="comment">##  [3] GBM_RNASeq2GeneNorm-20160128: SummarizedExperiment with 20501 rows and 166 columns </span></div><div class="line"><span class="comment">## Features: </span></div><div class="line"><span class="comment">##  experiments() - obtain the ExperimentList instance </span></div><div class="line"><span class="comment">##  colData() - the primary/phenotype DataFrame </span></div><div class="line"><span class="comment">##  sampleMap() - the sample availability DataFrame </span></div><div class="line"><span class="comment">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment </span></div><div class="line"><span class="comment">##  *Format() - convert into a long or wide DataFrame </span></div><div class="line"><span class="comment">##  assays() - convert ExperimentList to a SimpleList of matrices</span></div></pre></td></tr></table></figure>
<p>因此，单一的R变量可以用于处理（子集）593个人的3种不同的分析方法。成分分析(Constituent assays)含有ExpressionSet，SummarizedExperiment和RaggedExperiment。<code>RaggedExperiment</code>是它自己的包，可以通过<code>vignette</code>函数来查看使用方法。</p>
<p>为了了解GBM数据的范围和完整性，我们可以使用<code>UpSet</code>图来查看一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">upsetSamples(gbm)</div><div class="line"><span class="comment">## Loading required namespace: UpSetR</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104094334.png">

</div>
<p>这显示了绝大多数参与者提供的拷贝数变异和基于阵列的表达数据，但是很少有人提供RNA-seq或蛋白质组(PRRA，反相蛋白质组分析)数据。</p>
<h2 id="处理tcga突变数据">处理TCGA突变数据</h2>
<p>突变数据的复杂性为统一描述这种异质性数据提出了基本的挑战，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">mut = experiments(gbm)[[<span class="string">"GBM_Mutation-20160128"</span>]]</div><div class="line">mut</div><div class="line"><span class="comment">## class: RaggedExperiment </span></div><div class="line"><span class="comment">## dim: 22073 290 </span></div><div class="line"><span class="comment">## assays(74): Hugo_Symbol Entrez_Gene_Id ... OREGANNO_ID</span></div><div class="line"><span class="comment">##   OREGANNO_Values</span></div><div class="line"><span class="comment">## rownames: NULL</span></div><div class="line"><span class="comment">## colnames(290): TCGA-02-0003-01A-01D-1490-08</span></div><div class="line"><span class="comment">##   TCGA-02-0033-01A-01D-1490-08 ... TCGA-81-5911-01A-12D-1845-08</span></div><div class="line"><span class="comment">##   TCGA-87-5896-01A-01D-1696-08</span></div><div class="line"><span class="comment">## colData names(0):</span></div></pre></td></tr></table></figure>
<p>一个长度为74的向量已经给出了突变的“阵列”(assay)名称。通常意义上来说，这些都不是数组，而是不同个体之间可能存在着的不同突变特征或背景（注：这里的assay理解不清楚，后面理解清楚后再修改，原文已经列出）。</p>
<blockquote>
<p>The names of the ‘assays’ of mutations are given in a vector of length 74. These are not assays in the usual sense, but characteristics or contexts of mutations that may vary between individuals.</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">head(assayNames(mut))</div><div class="line"><span class="comment">## [1] "Hugo_Symbol"            "Entrez_Gene_Id"        </span></div><div class="line"><span class="comment">## [3] "Center"                 "NCBI_Build"            </span></div><div class="line"><span class="comment">## [5] "Variant_Classification" "Variant_Type"</span></div></pre></td></tr></table></figure>
<h3 id="突变的位置和基因关联性">突变的位置和基因关联性</h3>
<p>突变信息包含在了一个<code>GRanges</code>结构中，它记录了突变的坐标，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">rowRanges(mut)</div><div class="line"><span class="comment">## GRanges object with 22073 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           seqnames    ranges strand</span></div><div class="line"><span class="comment">##              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##       [1]        1   1430871      +</span></div><div class="line"><span class="comment">##       [2]        1 154148652      +</span></div><div class="line"><span class="comment">##       [3]        1 161206281      +</span></div><div class="line"><span class="comment">##       [4]        1 230846235      +</span></div><div class="line"><span class="comment">##       [5]       10 123810032      +</span></div><div class="line"><span class="comment">##       ...      ...       ...    ...</span></div><div class="line"><span class="comment">##   [22069]        X  77112989      +</span></div><div class="line"><span class="comment">##   [22070]        X 119243159      +</span></div><div class="line"><span class="comment">##   [22071]        X 131348336      +</span></div><div class="line"><span class="comment">##   [22072]        X 138643011      +</span></div><div class="line"><span class="comment">##   [22073]        X 150891145      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 24 sequences from 37 genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p>我们来查看一下哪些基因的突变频率最高，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">sort(table(assay(mut, <span class="string">"Hugo_Symbol"</span>)), decreasing=<span class="literal">TRUE</span>)[<span class="number">1</span>:<span class="number">6</span>]</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Unknown     TTN    EGFR    TP53    PTEN   MUC16 </span></div><div class="line"><span class="comment">##     126     121     102     101      93      68</span></div></pre></td></tr></table></figure>
<h3 id="突变类型">突变类型</h3>
<p>我们可以通过表格形式来查看突变类型，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">table(as.character(assay(mut, <span class="string">"Variant_Classification"</span>)))</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##        Frame_Shift_Del        Frame_Shift_Ins           In_Frame_Del </span></div><div class="line"><span class="comment">##                    566                    217                    214 </span></div><div class="line"><span class="comment">##           In_Frame_Ins      Missense_Mutation      Nonsense_Mutation </span></div><div class="line"><span class="comment">##                     28                  14213                    851 </span></div><div class="line"><span class="comment">##       Nonstop_Mutation                 Silent            Splice_Site </span></div><div class="line"><span class="comment">##                     17                   5514                    382 </span></div><div class="line"><span class="comment">## Translation_Start_Site </span></div><div class="line"><span class="comment">##                     71</span></div></pre></td></tr></table></figure>
<p>哪些基因的碱基的删除导致了移码突变呢？</p>
<p>我们可以使用一个矩阵下标来找到前10个发生移码突变的基因，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">sort(table(assay(mut, <span class="string">"Hugo_Symbol"</span>)[assay(mut, <span class="string">"Variant_Classification"</span>)==</div><div class="line">  <span class="string">"Frame_Shift_Del"</span>]),decreasing=<span class="literal">TRUE</span>)[<span class="number">1</span>:<span class="number">10</span>]</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##    PTEN     NF1    ATRX     RB1   CDC27    TP53     TTN Unknown    CHD8 </span></div><div class="line"><span class="comment">##      16      10       7       7       6       6       4       4       3 </span></div><div class="line"><span class="comment">##    OGDH </span></div><div class="line"><span class="comment">##       3</span></div></pre></td></tr></table></figure>
<p>总结</p>
<ul>
<li>MultiAssayExperiment统一了各种微阵列数据，并收集了一个队列的不同样本信息。</li>
<li>TCGA肿瘤特异性数据集已经被序列化为MultiAssayExperiments ，可以从Amzon S3作为公共数据。</li>
<li>含有复杂注释的异质性数据(Idiosyncratic data)可能通过RaggedAssay结构进行管理；这些结构已经被证明在处理突变和拷贝数变异方面发挥了强大的作用。</li>
</ul>
<h1 id="云端管理策略cgc与gdc概念">云端管理策略：CGC与GDC概念</h1>
<p>在前面部分，我介绍了如何将肿瘤的TCGA数据收集在一个队列对象中，该对象统一管理大量的基因组数据。在这一部分里，我们想说明如何通过单个的R连接对象来访问TCGA的完整数据并用于交互式分析。如果要使用这一部分中的命令，用户需要向由 Institute for Systems Biology管理的Cancer Genomics Cloud 获取权利。可以向[ISB project administrators] (https://www.systemsbiology.org/research/cancer-genomics-cloud/)联系，获取账户。</p>
<p>在下面的代码，我们使用r-db-stats团队维护的<code>bigrquery</code>包来创建一个连接到 <a href="https://cloud.google.com/bigquery/" target="_blank" rel="external">Google BigQuery</a>,的DBI连接。然后我们列出可用的表，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(bigrquery)</div><div class="line"><span class="keyword">library</span>(dplyr)</div><div class="line"><span class="keyword">library</span>(magrittr)</div><div class="line">tcgaCon = DBI::dbConnect(dbi_driver(), project=<span class="string">"isb-cgc"</span>, </div><div class="line">     dataset=<span class="string">"TCGA_hg38_data_v0"</span>, billing = Sys.getenv(<span class="string">"CGC_BILLING"</span>)) </div><div class="line">dbListTables(tcgaCon)</div><div class="line"><span class="comment">##  [1] "Copy_Number_Segment_Masked"  "DNA_Methylation"            </span></div><div class="line"><span class="comment">##  [3] "DNA_Methylation_chr1"        "DNA_Methylation_chr10"      </span></div><div class="line"><span class="comment">##  [5] "DNA_Methylation_chr11"       "DNA_Methylation_chr12"      </span></div><div class="line"><span class="comment">##  [7] "DNA_Methylation_chr13"       "DNA_Methylation_chr14"      </span></div><div class="line"><span class="comment">##  [9] "DNA_Methylation_chr15"       "DNA_Methylation_chr16"      </span></div><div class="line"><span class="comment">## [11] "DNA_Methylation_chr17"       "DNA_Methylation_chr18"      </span></div><div class="line"><span class="comment">## [13] "DNA_Methylation_chr19"       "DNA_Methylation_chr2"       </span></div><div class="line"><span class="comment">## [15] "DNA_Methylation_chr20"       "DNA_Methylation_chr21"      </span></div><div class="line"><span class="comment">## [17] "DNA_Methylation_chr22"       "DNA_Methylation_chr3"       </span></div><div class="line"><span class="comment">## [19] "DNA_Methylation_chr4"        "DNA_Methylation_chr5"       </span></div><div class="line"><span class="comment">## [21] "DNA_Methylation_chr6"        "DNA_Methylation_chr7"       </span></div><div class="line"><span class="comment">## [23] "DNA_Methylation_chr8"        "DNA_Methylation_chr9"       </span></div><div class="line"><span class="comment">## [25] "DNA_Methylation_chrX"        "DNA_Methylation_chrY"       </span></div><div class="line"><span class="comment">## [27] "Protein_Expression"          "RNAseq_Gene_Expression"     </span></div><div class="line"><span class="comment">## [29] "Somatic_Mutation"            "Somatic_Mutation_Jun2017"   </span></div><div class="line"><span class="comment">## [31] "miRNAseq_Expression"         "miRNAseq_Isoform_Expression"</span></div></pre></td></tr></table></figure>
<p>现在我们使用<code>dplyr</code>来对GBM中的突变类型做一个汇总，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">tcgaCon %&gt;% tbl(<span class="string">"Somatic_Mutation"</span>) %&gt;% dplyr::filter(project_short_name==<span class="string">"TCGA-GBM"</span>) %&gt;% </div><div class="line">      dplyr::select(Variant_Classification, Hugo_Symbol) %&gt;% group_by(Variant_Classification) %&gt;%</div><div class="line">      summarise(n=n())</div><div class="line"><span class="comment">## # Source:   lazy query [?? x 2]</span></div><div class="line"><span class="comment">## # Database: BigQueryConnection</span></div><div class="line"><span class="comment">##    Variant_Classification     n</span></div><div class="line"><span class="comment">##                     &lt;chr&gt; &lt;int&gt;</span></div><div class="line"><span class="comment">##  1                  3'UTR  4166</span></div><div class="line"><span class="comment">##  2                 Intron  3451</span></div><div class="line"><span class="comment">##  3                3'Flank   423</span></div><div class="line"><span class="comment">##  4       Nonstop_Mutation    58</span></div><div class="line"><span class="comment">##  5                    IGR     3</span></div><div class="line"><span class="comment">##  6                    RNA  1338</span></div><div class="line"><span class="comment">##  7            Splice_Site   955</span></div><div class="line"><span class="comment">##  8        Frame_Shift_Del  1487</span></div><div class="line"><span class="comment">##  9      Missense_Mutation 53054</span></div><div class="line"><span class="comment">## 10 Translation_Start_Site    60</span></div><div class="line"><span class="comment">## # ... with more rows</span></div></pre></td></tr></table></figure>
<p>这从本质上证明了，我们可以使用单个R变量（在本例中是 <code>tcgaCon</code> ）的操作来查看整个TCGA数据的概念。构成这个操作的基础设施是由Google提供的，但是这种基本思路也可以在其它环境中进行重现。</p>
<p>我们能设想一个类似于 <code>SummarizedExperiment</code> 样的接口来连接这种数据储存的？这是没问题的，可以参考Github上的<code>seByTumor</code> 函数，它位于<code>shwetagopaul92/restfulSE</code> 包（这个包正在进行Bioconductor的评估，不过可以通过Github来进行安装）。</p>
<p>Institute for Systems Biology的癌症基因组去试点项目(Cancer Genomics Cloud pilot instance)的介绍到此为止。其它的云试点项目则是由 <a href="http://www.cancergenomicscloud.org/" target="_blank" rel="external">Seven Bridges Genomics</a> 和 <a href="https://software.broadinstitute.org/firecloud/" target="_blank" rel="external">Broad Institute</a>创建的。这些项目目前正在努力形成一个云端框架，从而整合大量的通用基因组学数据，从而减少基因组学数据的传输。 <a href="https://grants.nih.gov/grants/guide/rfa-files/RFA-HG-17-011.html" target="_blank" rel="external">RFA</a> 提供了非常丰富的信息。</p>
<h1 id="总体总结">总体总结</h1>
<p>我们已经回顾了几个实验案例的基因组数据管理方式，包括这几类：</p>
<ul>
<li>基于微阵列的基因表达检测</li>
<li>基于微阵列的DNA甲基化检测</li>
<li>RNA-seq比对和由此衍生的基因表达测量</li>
<li>全基因组基因分型</li>
<li>肿瘤突变评估</li>
</ul>
<p>数据来源包括：</p>
<ul>
<li>本地文件</li>
<li>机构数据库 (NCBI GEO, EMBL ArrayExpress)</li>
<li>高通量外部数据储存 (HDF5)</li>
<li>云端分布式数据储存 (Google BigQuery or Amazon S3)</li>
</ul>
<p>重要的管理原则包括：</p>
<ul>
<li>统一的正式对象类(ExpressionSet，Summary)中的分析，样本和实验元数据。</li>
<li>当处理一个队列的重叠子集时，要协调分析集的合并 (MultiAssayExperiment)。</li>
<li>在特征过滤或样本过滤对错中的自同态特征。</li>
<li>使用Granges作为查询的基因组坐标的过滤工具。</li>
<li>通过使用关键来源信息（例如构建一个参考基因组）标记对象来检测其一致性。</li>
</ul>
<p>这是一个在分子分析技术、数据存储和检索领域正在进行重大创新的时代。制定基准和支持策略的选择是一个非常有价值但不简单的过程。这一章已经向您介绍了一系列概念和解决方案，其中许多内容已经应用于本课程中要进行讲解的分析案例中。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/dataman2019.html" target="_blank" rel="external">Management of genome-scale data (2019, Bioc 3.9)</a></li>
<li><a href="https://davetang.org/muse/2013/02/22/using-tabix/" target="_blank" rel="external">Using tabix</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><span class="space">&hellip;</span><a class="page-number" href="/page/24/">24</a><a class="extend next" rel="next" href="/page/2/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">231</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">113</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共930.0k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
