<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Hexo, NexT" />










<meta property="og:type" content="website">
<meta property="og:title" content="RVDSD的个人笔记本">
<meta property="og:url" content="http://rvdsd.top/page/3/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RVDSD的个人笔记本">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/page/3/"/>





  <title>RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/" itemprop="url">GDAS004-Bioconductor与基因组级数据分析简介</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-04T12:00:00+08:00">
                2019-09-04
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,657
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  15
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="前言">前言</h1>
<p>在生命科学研究领域的相关人员，经常使用电子表格来记录和管理实验数据。针对这个问题，我们将会提出一个使用R和Bioconductor管理基因组级实验数据的原则。使用基于R的软件来管理数据似乎并不太常见。通常来说，R仅用于编写程序，构建统计模型。但是，针对复杂生物实验数据的多方面需求这种情况，Bioconductor已经非常着手以统一的方式来详细地注释这些数据。Bioconductor的这种做法可以让我们导入初步分析的结果（例如多个微阵列的扫描结果以及多个测序结果的比对信息），将这些初步的结果与有关实验证的元数据，以及检测的样本信息结合起来，从而将所有的信息都保存在一个R变量中。通过周全的设计该变量的数据结构，程序可以对这些数据进行各种操作，最终获得更加有效的管理和分析环境。</p>
<p>在这一节中，我们会通过一些实验数据来演示一下基本的操作方法，从而，逐步过滤到更加全面的表示方法（</p>
<p>我们将通过对Bioconductor中所表示的实验数据的一些非常基本的表示法，逐步走向更先进的综合表示法，从而理解这种数据操作方法所表现出来的高效率。</p>
<h2 id="案例说明-youtube上有相应视频">案例说明 (Youtube上有相应视频)</h2>
<p>1.0 Distinct components from an experiment.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSE5859Subset)</div><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line">.oldls = ls()</div><div class="line">data(GSE5859Subset)</div><div class="line">.newls = ls()</div></pre></td></tr></table></figure>
<p>我们导入了一个<code>GSE5859Subset</code>的实验数据，现在我们使用 <code>setdiff()</code>函数来比较一下导入前后文件的区别，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">newstuff = setdiff(.newls, .oldls)</div><div class="line">newstuff</div><div class="line"><span class="comment">## [1] "geneAnnotation" "geneExpression" "sampleInfo"</span></div></pre></td></tr></table></figure>
<p>从结果中我们可以发现，当导入了 <code>GSE5859Subset</code>实验数据中，这个数据中含有3个文件。它们3个分别属于不同的类，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">cls = lapply(newstuff, <span class="keyword">function</span>(x) class(get(x)))</div><div class="line">names(cls) = newstuff</div><div class="line">cls</div><div class="line"><span class="comment">## $$geneAnnotation</span></div><div class="line"><span class="comment">## [1] "data.frame"</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$geneExpression</span></div><div class="line"><span class="comment">## [1] "matrix"</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$sampleInfo</span></div><div class="line"><span class="comment">## [1] "data.frame"</span></div></pre></td></tr></table></figure>
<p>我们可以按照种族绘制绘制出它们不同时间的图形，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(date~factor(ethnicity), data=sampleInfo, ylab=&quot;chip date&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png" alt="plot of chunk lket">
<p class="caption">plot of chunk lket</p>
</div>
<p>假如我们想查看一下BRCA2在不同种族中的表达情况，我们可以按以下操作进行：</p>
<ul>
<li>我们需要知道<code>geneExpression</code> 这个矩阵中哪一行代表BRCA2的数据，并且我们要知道<code>geneAnnotation</code>哪一列提供了BRCA2的信息（探针名与基因名的对应关系）；</li>
<li>我们需要知道 <code>geneExpression</code> 这个矩阵中的顺序与<code>samkpleInfo</code>是一致的，要知道<code>sampleInfo</code>中的信息与<code>geneExpression</code>中的标签的对应关系。</li>
</ul>
<p>因此，我为了将基因表达矩阵与样本信息联系起来，我们需要对三个数据进行操作（分别为表达矩阵，样本信息，基因注释信息），从而确保这些信息能够匹配起来，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">all.equal(colnames(geneExpression), sampleInfo$filename)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">all.equal(rownames(geneExpression), geneAnnotation$PROBEID)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">ind = which(geneAnnotation$SYMBOL==<span class="string">"BRCA2"</span>)</div><div class="line">boxplot(geneExpression[ind,]~sampleInfo$ethnicity, ylab=<span class="string">"BRCA2 by hgfocus"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-dochks-1.png" alt="plot of chunk dochks">
<p class="caption">plot of chunk dochks</p>
</div>
<p>虽然我们绘制出了图形，但是这略微复杂，为了降低操作难度，Bioconductor定义了一个名为<code>ExpressionSet</code>数据结构。</p>
<h3 id="expressionset数据结构">ExpressionSet数据结构</h3>
<p>我们使用简单的命令就能构建起ExpressionSet这种数据结构，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">es1 = ExpressionSet(geneExpression)</div><div class="line">pData(es1) = sampleInfo</div><div class="line">fData(es1) = geneAnnotation</div><div class="line">es1</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 24 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: 107 122 ... 159 (24 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename group</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData</span></div><div class="line"><span class="comment">##   featureNames: 1 30 ... 12919 (8793 total)</span></div><div class="line"><span class="comment">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</span></div><div class="line"><span class="comment">##   fvarMetadata: labelDescription</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation:</span></div></pre></td></tr></table></figure>
<p>上面的<code>es</code> 这个单一变量就浓缩了整个实验的数据，将原来的三个数据变为了一个。现在我们使用<code>es1</code>就能直接绘图，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(es1$date~es1$ethnicity)</div></pre></td></tr></table></figure>
<p>现在我们已经了解了<code>ExpressionSet</code>容器的设计，你自己也能构建一个ExpressionSet实例。我们将会在下面的实验中，使用一个更加广泛的数据表示方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSE5859)</div><div class="line">data(GSE5859)</div><div class="line">e</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 8793 features, 208 samples </span></div><div class="line"><span class="comment">##   element names: exprs, se.exprs </span></div><div class="line"><span class="comment">## protocolData</span></div><div class="line"><span class="comment">##   sampleNames: GSM25349.CEL.gz GSM25350.CEL.gz ...</span></div><div class="line"><span class="comment">##     GSM136729.CEL.gz (208 total)</span></div><div class="line"><span class="comment">##   varLabels: ScanDate</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: 1 2 ... 208 (208 total)</span></div><div class="line"><span class="comment">##   varLabels: ethnicity date filename</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData: none</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">## Annotation: hgfocus</span></div></pre></td></tr></table></figure>
<h3 id="元数据管理metadata-management">元数据管理(Metadata management)</h3>
<p>及时地了解关于实验意图和方法的信息非常有用。GSE5859中使用的一些微阵列数据就是类来源于一篇发表于2004年（PubmedID为15269782）的论文。在联网的情况下，我们可以将论文的信息导入R，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(annotate)</div><div class="line">p = pmid2MIAME(<span class="string">"15269782"</span>)</div><div class="line">p</div><div class="line"><span class="comment">## Experiment data</span></div><div class="line"><span class="comment">##   Experimenter name: Morley M </span></div><div class="line"><span class="comment">##   Laboratory: NA </span></div><div class="line"><span class="comment">##   Contact information:  </span></div><div class="line"><span class="comment">##   Title: Genetic analysis of genome-wide variation in human gene expression. </span></div><div class="line"><span class="comment">##   URL:  </span></div><div class="line"><span class="comment">##   PMIDs: 15269782 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   Abstract: A 167 word abstract is available. Use 'abstract' method.</span></div></pre></td></tr></table></figure>
<p>现在查看你R环境中的<code>e</code>变量，使用<code>experimentData(e) &lt;- p</code>后，就能够将PubmedID信息添加到<code>e</code>中。使用<code>experimentData(e)</code>也能发现<code>e</code>中的信息改变了，也含有Pubmed ID信息。</p>
<h1 id="关于expressionset的why与how">关于ExpressionSet的Why与How</h1>
<p>现在我们通过创建以及使用<code>ExpressionSet</code> 实例已经快速了解了这种数据结构的特点。我们现在将要了解一下，为什么要这样设计这种数据结构，以及如何使用这些数据结构。</p>
<p>设计<code>ExpressionSet</code> 这种数据结构的初衷有很多，以下几点最为关键：</p>
<ul>
<li>降低复杂分析，注释以及样本信息编程的难度；</li>
<li>支持特征值（其实就是表达值）和样本信息的简单的过滤；</li>
<li>在将分析数据传递给分析函数时控制内存占用。</li>
</ul>
<p>以上就是为什么要设计ExpressionSet这种数据结构。关于如何使用这种结构，这些内容就要涉及一些设计模式问题，在R中，与面向对象编程相关类被称为S4类。</p>
<h2 id="面向对象编程与s4类">面向对象编程与S4类</h2>
<p>面向对象编程(OOP, Object-oriented programming)是计算机编程一个学科，它有助于降低程序的复杂性和维护要求，并能增强程度的互动性。OOP有着各种解决方式，即使在R中，也可以进行OOP，这就是所谓的S4方法。在R中，OOP的基本思想就是一个对象有些特征：</p>
<ul>
<li>将不同类型的数据集中放到一个统一的结构中，</li>
<li>它被标记为某个正式指定类的成员，并且可以正式地位于相关对象的系统中，</li>
<li>以类特异性方式对泛型方法进行响应。</li>
</ul>
<p>我们可以使用<code>getClass</code>函数查看一个<code>ExpresionSet</code>的类结构，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"ExpressionSet"</span>)</div><div class="line"><span class="comment">## Class "ExpressionSet" [package "Biobase"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                                </span></div><div class="line"><span class="comment">## Name:      experimentData          assayData          phenoData</span></div><div class="line"><span class="comment">## Class:              MIAME          AssayData AnnotatedDataFrame</span></div><div class="line"><span class="comment">##                                                                </span></div><div class="line"><span class="comment">## Name:         featureData         annotation       protocolData</span></div><div class="line"><span class="comment">## Class: AnnotatedDataFrame          character AnnotatedDataFrame</span></div><div class="line"><span class="comment">##                          </span></div><div class="line"><span class="comment">## Name:   .__classVersion__</span></div><div class="line"><span class="comment">## Class:           Versions</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "eSet", directly</span></div><div class="line"><span class="comment">## Class "VersionedBiobase", by class "eSet", distance 2</span></div><div class="line"><span class="comment">## Class "Versioned", by class "eSet", distance 3</span></div></pre></td></tr></table></figure>
<p>从上面我们可以看到，一共有7个组件定义了该类。只有<code>annoataion</code>这个组件占据了一个单独的R类型，也就是<code>character</code>类型。至于其它的组件，都是S4类的一个实例（或者说是更多的R数据类型）。</p>
<p>举例说明一下，<code>MIAME</code>类的定义如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"MIAME"</span>)</div><div class="line"><span class="comment">## Class "MIAME" [package "Biobase"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:               name               lab           contact</span></div><div class="line"><span class="comment">## Class:         character         character         character</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:              title          abstract               url</span></div><div class="line"><span class="comment">## Class:         character         character         character</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:          pubMedIds           samples    hybridizations</span></div><div class="line"><span class="comment">## Class:         character              list              list</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:       normControls     preprocessing             other</span></div><div class="line"><span class="comment">## Class:              list              list              list</span></div><div class="line"><span class="comment">##                         </span></div><div class="line"><span class="comment">## Name:  .__classVersion__</span></div><div class="line"><span class="comment">## Class:          Versions</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "MIAxE", directly</span></div><div class="line"><span class="comment">## Class "characterORMIAME", directly</span></div><div class="line"><span class="comment">## Class "Versioned", by class "MIAxE", distance 2</span></div></pre></td></tr></table></figure>
<p>因此，我们可以发现，S4可以让我们定义包含各种信息的新数据结。我们虽然可以使用R中非常通用的列表结构来实现这一目的（就是说实现封装不同信息），但是S4允许我们能保证有新结构的属性，以便为这些结构定义的程序不必“检查”它们正在操作什么。这些检查过程内置在了类系统中。</p>
<h2 id="数据框dataframe-表格数据的增强型结构">数据框(DataFrame): 表格数据的增强型结构</h2>
<p>数据框是R中非常基础的数据结构。Bioconductor中引入了<code>DataFrame</code>类，它可以实现其它一些额外的功能，后面我们会提到。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(S4Vectors)</div><div class="line">sa2 = DataFrame(sampleInfo)</div><div class="line">sa2</div><div class="line"><span class="comment">## DataFrame with 24 rows and 4 columns</span></div><div class="line"><span class="comment">##     ethnicity       date         filename     group</span></div><div class="line"><span class="comment">##      &lt;factor&gt;     &lt;Date&gt;      &lt;character&gt; &lt;numeric&gt;</span></div><div class="line"><span class="comment">## 1         ASN 2005-06-23 GSM136508.CEL.gz         1</span></div><div class="line"><span class="comment">## 2         ASN 2005-06-27 GSM136530.CEL.gz         1</span></div><div class="line"><span class="comment">## 3         ASN 2005-06-27 GSM136517.CEL.gz         1</span></div><div class="line"><span class="comment">## 4         ASN 2005-10-28 GSM136576.CEL.gz         1</span></div><div class="line"><span class="comment">## 5         ASN 2005-10-07 GSM136566.CEL.gz         1</span></div><div class="line"><span class="comment">## ...       ...        ...              ...       ...</span></div><div class="line"><span class="comment">## 20        ASN 2005-06-27 GSM136524.CEL.gz         0</span></div><div class="line"><span class="comment">## 21        ASN 2005-06-23 GSM136514.CEL.gz         0</span></div><div class="line"><span class="comment">## 22        ASN 2005-10-07 GSM136563.CEL.gz         0</span></div><div class="line"><span class="comment">## 23        ASN 2005-10-07 GSM136564.CEL.gz         0</span></div><div class="line"><span class="comment">## 24        ASN 2005-10-07 GSM136572.CEL.gz         0</span></div></pre></td></tr></table></figure>
<p><code>show</code>方法描述顶部和底部的行数据，提供有关列数据类型的信息，并且能够比<code>data.frame</code>数据结构更有效地管理列中的复杂数据类所，下面是类定义：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(<span class="string">"DataFrame"</span>)</div><div class="line"><span class="comment">## Class "DataFrame" [package "S4Vectors"]</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Slots:</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:           rownames             nrows          listData</span></div><div class="line"><span class="comment">## Class: character_OR_NULL           integer              list</span></div><div class="line"><span class="comment">##                                                             </span></div><div class="line"><span class="comment">## Name:        elementType   elementMetadata          metadata</span></div><div class="line"><span class="comment">## Class:         character DataTable_OR_NULL              list</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Extends: </span></div><div class="line"><span class="comment">## Class "DataTable", directly</span></div><div class="line"><span class="comment">## Class "SimpleList", directly</span></div><div class="line"><span class="comment">## Class "DataTable_OR_NULL", by class "DataTable", distance 2</span></div><div class="line"><span class="comment">## Class "List", by class "SimpleList", distance 2</span></div><div class="line"><span class="comment">## Class "Vector", by class "SimpleList", distance 3</span></div><div class="line"><span class="comment">## Class "Annotated", by class "SimpleList", distance 4</span></div></pre></td></tr></table></figure>
<h2 id="源于多样本ngs数据的成熟数据表示方法">源于多样本NGS数据的成熟数据表示方法</h2>
<p>短读长测序产生了远比微阵列多得多的数据。测序实验也可以通过更加多样化的手段进行分析，因此测序实验能够在不同粒度级别(at various levels of granularity)提供有效的输入。</p>
<p>处理RNA-seq数据的典型流程包括：</p>
<ol style="list-style-type: decimal">
<li>展示每个周期中碱基所调用的色谱（2008年后就不再提供了）；</li>
<li>碱基的读取质量值，它们通常以FASTQ格式进行储存；</li>
<li>通过基因组或转录本比对后，对各个基因统计的结果；</li>
<li>基因组极的多个样本的数据储存，用于推断感兴趣的生物学问题。通常一个样本对应两个FASTQ文件。</li>
</ol>
<h3 id="bioconductor中的bam文件">Bioconductor中的BAM文件</h3>
<p><code>RNAseqData.HNRNPC.bam.chr14</code> 中含有一些BAM文件，这些BAM文件是8个Hela细胞系样本的双端测序数据，其中4个样本通过RNAi的方式敲低了异质核核糖核蛋白C1和C2（hnRNPC,heterogeneous nuclear ribonucleoproteins C1 and C2)。这个家庭的核蛋白参与pre-mRNA的加工。现在我们演示一下，如何从BAM文件中来统计与HNRNPC有关的读长数目(reads count)。</p>
<p>BAM文件储存在<code>RNAseqData.HNRNPC.bam.chr14</code> 包中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">bfp = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line">length(bfp)</div><div class="line"><span class="comment">## [1] 8</span></div><div class="line">bfp[<span class="number">1</span>]</div><div class="line"><span class="comment">##                                                                                                                ERR127306 </span></div><div class="line"><span class="comment">## "/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RNAseqData.HNRNPC.bam.chr14/extdata/ERR127306_chr14.bam"</span></div></pre></td></tr></table></figure>
<p>我们使用<code>Rsamtools</code>包以一个变量的形式管理这些信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Rsamtools)</div><div class="line">bfl = BamFileList(file=bfp)</div><div class="line">bfl</div><div class="line"><span class="comment">## BamFileList of length 8</span></div><div class="line"><span class="comment">## names(8): ERR127306 ERR127307 ERR127308 ... ERR127303 ERR127304 ERR127305</span></div></pre></td></tr></table></figure>
<p>BAM文件包括关于读长(reads)与基因组比对后的有限数量的元数据。这里我们以<code>BamFileList</code>实例来说明一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">seqinfo(bfl)</div><div class="line"><span class="comment">## Seqinfo object with 93 sequences from an unspecified genome:</span></div><div class="line"><span class="comment">##   seqnames              seqlengths isCircular genome</span></div><div class="line"><span class="comment">##   chr1                   249250621       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr10                  135534747       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr11                  135006516       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr11_gl000202_random      40103       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chr12                  133851895       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ...                          ...        ...    ...</span></div><div class="line"><span class="comment">##   chrUn_gl000247             36422       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrUn_gl000248             39786       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrUn_gl000249             38502       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrX                   155270560       &lt;NA&gt;   &lt;NA&gt;</span></div><div class="line"><span class="comment">##   chrY                    59373566       &lt;NA&gt;   &lt;NA&gt;</span></div></pre></td></tr></table></figure>
<p>我们看到基因”没有指定“，但是我们碰巧知道参考基因组<code>hg19</code>用于进入基因序列的比对。使用<code>[genome(bfl) &lt;- “hg19”</code>可以纠正这点，但是还存在着一个bug。这个bug应该修正一下，以便可以标记使用不同的参考基因组来进行比对或纠正错误，注：这段不太理解，原文如下：</p>
<blockquote>
<p>We see that the genome is “unspecified”, but we happen to know that the reference build ‘hg19’ was used to define the genomic sequence for alignment. [genome(bfl) &lt;- “hg19” should rectify this but at present there is a bug. This should be fixed so that attempts to compare alignments using different references can be flagged or throw errors.]</p>
</blockquote>
<p>为了验证是否成功敲除，我们将检查编码HNRNPC的基因组附近的序列的比对情况。稍后，我们将展示如何使用Bioconductor注释工具获取基因地址；不过这个地址仅用于hg19基因组，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">hnrnpcLoc = GRanges(<span class="string">"chr14"</span>, IRanges(<span class="number">21677296</span>, <span class="number">21737638</span>))</div><div class="line">hnrnpcLoc</div><div class="line"><span class="comment">## GRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames               ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]    chr14 [21677296, 21737638]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p>这标识了HNRNPC编码基因为hg19注释的染色体14的区域。评估表达式量化的一种简单方法是使用GenomicAlignmentspackage中的summary izeOverlaps。此计数方法将隐式使用多核计算。</p>
<p>上面的代码标识了14号染色体中编码HNRNPC附近的序列。评估基因表达量化的一种简单方法就是使用<code>GenomicAlignmentspackage</code>包中的<code>summarizeOverlaps</code> 函数，这个计算方法隐式地使用了多核计算，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line"><span class="keyword">library</span>(BiocParallel)</div><div class="line">register(SerialParam())</div><div class="line">hnse = summarizeOverlaps(hnrnpcLoc,bfl)</div></pre></td></tr></table></figure>
<h3 id="练习">练习</h3>
<ol style="list-style-type: decimal">
<li>检查<code>hnse</code>中的<code>count</code>组件内容。与HNRNPC敲除相对应的有关量化的指标是什么？</li>
<li>向<code>hnse</code>的<code>colData</code>组件中添加一个变量<code>condition</code>，用于指示敲除株与野生株。</li>
<li>具有后缀306和307的样本是单个对照的技术重复；308和309同样如此。因此使用了6个不同的生物学样本。现在向<code>colData</code>中添加一个变量，用于区别这6个样本。</li>
</ol>
<h2 id="expressionset设计的一些复杂性进阶">ExpressionSet设计的一些复杂性(进阶)</h2>
<p>生物信息学家是在一个不断变化的环境中工作。自Bioconductor建立以来，计算机的容量和通量已经大幅度增加。R语言也性现了很大的变量。在早期的时候，大家都比较注意内存的使用，因为当时在R中，向一个函数传递参数时，会复制内存中的对象， 这种操作比较低效。而“环境”类中的R对象则不会发生复制，这就导致了使用环境来储存大量的数据用于分析。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">class(slot(es1, <span class="string">"assayData"</span>))</div><div class="line"><span class="comment">## [1] "environment"</span></div><div class="line">ae = slot(es1, <span class="string">"assayData"</span>)</div><div class="line">ae</div><div class="line"><span class="comment">## &lt;environment: 0x7f813b4dc698&gt;</span></div></pre></td></tr></table></figure>
<p>“环境”是R中一种非常特殊的对象类型，可以将它视为对表达值矩阵引用的容器。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">ls(ae)</div><div class="line"><span class="comment">## [1] "exprs"</span></div><div class="line">dim(ae$exprs)</div><div class="line"><span class="comment">## [1] 8793   24</span></div><div class="line">dim(exprs(es1))</div><div class="line"><span class="comment">## [1] 8793   24</span></div><div class="line">object.size(ae)</div><div class="line"><span class="comment">## 56 bytes</span></div><div class="line">object.size(ae$exprs)</div><div class="line"><span class="comment">## 2253824 bytes</span></div></pre></td></tr></table></figure>
<h3 id="练习可选">练习（可选）</h3>
<p>请解释以下运行结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">e1 = new.env()</div><div class="line">l1 = list()</div><div class="line">e1$x = <span class="number">5</span></div><div class="line">l1$x = <span class="number">5</span></div><div class="line">all.equal(e1$x, l1$x)</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">f = <span class="keyword">function</span>(z) &#123;z$x = <span class="number">4</span>; z$x&#125;</div><div class="line">all.equal(f(e1), f(l1))</div><div class="line"><span class="comment">## [1] TRUE</span></div><div class="line">all.equal(e1$x, l1$x)  <span class="comment"># different from before</span></div><div class="line"><span class="comment">## [1] "Mean relative difference: 0.25"</span></div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">&gt; e1 = new.env()</div><div class="line">&gt; l1 = list()</div><div class="line">&gt; e1$x = <span class="number">5</span></div><div class="line">&gt; l1$x = <span class="number">5</span></div><div class="line">&gt; all.equal(e1$x, l1$x)</div><div class="line">[<span class="number">1</span>] <span class="literal">TRUE</span></div><div class="line">&gt; <span class="comment">## [1] TRUE</span></div><div class="line">&gt; f = <span class="keyword">function</span>(z) &#123;z$x = <span class="number">4</span>; z$x&#125;</div><div class="line">&gt; all.equal(f(e1), f(l1))</div><div class="line">[<span class="number">1</span>] <span class="literal">TRUE</span></div><div class="line">&gt; <span class="comment">## [1] TRUE</span></div><div class="line">&gt; all.equal(e1$x, l1$x)  <span class="comment"># different from before</span></div><div class="line">[<span class="number">1</span>] <span class="string">"Mean relative difference: 0.25"</span></div><div class="line">&gt; <span class="comment">## [1] "Mean relative difference: 0.25"</span></div></pre></td></tr></table></figure>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_mgt_gsd.html" target="_blank" rel="external">Management of genome-scale data</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/" itemprop="url">GDAS003-Bioconductor与基因组级数据分析简介</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-03T12:00:00+08:00">
                2019-09-03
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,445
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  13
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>本篇笔记主要是介绍了Bioconductor与基因组级数据分析的关系。</p>
<h2 id="r语言r包与仓库repositories">R语言，R包与仓库(repositories)</h2>
<p>学习这一系列课程的前提是你已经有了一定的R语言基因。并且我们推荐使用RStudio作为R语言的使用环境。如果你跳过前面的1x-4x部分，直接进入5x，可以先参考一下其它的R语言教程，例如 <a href="http://tryr.codeschool.com/" target="_blank" rel="external">Try R</a> 。</p>
<h2 id="why-r">Why R?</h2>
<p>Bioconductor的基础就是R。使用R有3个原因：</p>
<ul>
<li>许多统计学家与生物统计学家使用R语言，并且他们使用R语言设计了许多算法可以帮助我们理解复杂的实验数据。</li>
<li>R语言有着高度的互操作性(interoperable)，促进了使用其它语言编写软件组件的重用。</li>
<li>R可移植到在各种计算机上的操作系统里(Linux、MacOSX、Windows)，初学者可以使用这些平台中的任一个来学习。</li>
</ul>
<p>总之，R的易用性和在统计学和“数据科学”中扮演着核心角色，因此当生物学家和统计学家面对基因组级的实验数据时，选择R作为处理工具就是自然而然的事情了。Bioconductor项目开始于2001年，自那以后，它就与基因组级生物学中不断增长的数据量和复杂性保持着同步。</p>
<h2 id="函数式面向对象程序设计">函数式面向对象程序设计</h2>
<p>R语言是一种混合了函数式编程与面向对象的混合型语言。</p>
<p>在R中进行函数式编程时，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">square = <span class="keyword">function</span>(x) x^<span class="number">2</span></div></pre></td></tr></table></figure>
<p>上面就是一个典型的函数式编程。我们定义了一个函数<code>square()</code>，它会计算输入的数据字的平方，即<code>^2</code>，例如当我们输入3时，结果就是在R中，所有的计算都是通过函数进行的。</p>
<p>在进行面向对象的编程时，我们所关注的生就是构建数据结构，定义能够操作这些结构化数据的方法。这是一种接近于人正常思维的编程思想。最早出现在20世纪90年代。现在我们以Bioconductor中的一个案例说明一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">class(Homo.sapiens)</div><div class="line">methods(class=class(Homo.sapiens))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(Homo.sapiens)</div><div class="line">&gt; class(Homo.sapiens)</div><div class="line">[<span class="number">1</span>] <span class="string">"OrganismDb"</span></div><div class="line">attr(,<span class="string">"package"</span>)</div><div class="line">[<span class="number">1</span>] <span class="string">"OrganismDbi"</span></div><div class="line">&gt; methods(class=class(Homo.sapiens))</div><div class="line"> [<span class="number">1</span>] asBED                 asGFF                 cds                   cdsBy                </div><div class="line"> [<span class="number">5</span>] coerce&lt;-              columns               dbconn                dbfile               </div><div class="line"> [<span class="number">9</span>] disjointExons         distance              exons                 exonsBy              </div><div class="line">[<span class="number">13</span>] extractUpstreamSeqs   fiveUTRsByTranscript  genes                 getTxDbIfAvailable   </div><div class="line">[<span class="number">17</span>] intronsByTranscript   isActiveSeq           isActiveSeq&lt;-         keys                 </div><div class="line">[<span class="number">21</span>] keytypes              mapIds                mapToTranscripts      metadata             </div><div class="line">[<span class="number">25</span>] microRNAs             promoters             resources             select               </div><div class="line">[<span class="number">29</span>] selectByRanges        selectRangesById      seqinfo               show                 </div><div class="line">[<span class="number">33</span>] taxonomyId            threeUTRsByTranscript transcripts           transcriptsBy        </div><div class="line">[<span class="number">37</span>] tRNAs                 TxDb                  TxDb&lt;-               </div><div class="line">see <span class="string">'?methods'</span> <span class="keyword">for</span> accessing help and <span class="keyword">source</span> code</div></pre></td></tr></table></figure>
<p>在这个案例中，我们可以看到：</p>
<ol style="list-style-type: decimal">
<li><code>OrganismDb</code>是一个类(<code>class</code>)；</li>
<li><code>Homo.sapiens</code>是<code>OrganismDb</code>这个类的一个实例(instance)；</li>
<li>某个类的所有方法都可以应用到这个类的所有实例中去（<code>methods(class=class(Homo.sapiens))</code>就是查看这个类的方法）；</li>
<li>每个方法的实现是通过R中的函数进行的（我的理解就是，方法其实本质上就是函数）。并且函数的运行依赖于它自身的参数。</li>
</ol>
<p>其中需要特别注意的是juncture(交界区)的方法，即<code>genes</code>，<code>exons</code>，<code>transcripts</code>，这三个方法输出基因组基本组成成分的一些信息。上面我们列出的所有方法都可以应用于<code>OrganismDb</code>类定义的一些实例，也就是一些模式生物，例如小鼠(Mus musculus)，酒酿酵母(S. cerevisiae)和秀丽隐杆线虫(C. elegans)。</p>
<h2 id="r包模块化与持续集成continuous-integration">R包，模块化与持续集成(Continuous* *Integration)</h2>
<p>这一部分在阅读的时候可以跳过。</p>
<h3 id="r包的结构">R包的结构</h3>
<p>我们可以通过编写R代码来用R执行面向对象的函数式编程。一种基本的方法就是创建“脚本”，用于定义数据的导入和分析的基本流程。当脚本以仅定义函数和数据结构的方式编写时，我们就可以将这些脚本打包，从而方便地发布，供其他有类似数据处理问题的用户使用。</p>
<p>R软件的<a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html" target="_blank" rel="external">打包协议</a>规定了如何将R和其他语言的源代码与元数据以及文档一起组织起来，以促进简便的测试和发布的流程。例如，它义些文档的包的早期版本具有如下的目录和文件布局：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">├── DESCRIPTION  (文本文件，用于说明提供的元数据，授权信息)</div><div class="line">├── NAMESPACE    (定义输入与输出)</div><div class="line">├── R            (文件夹，包含了一些R代码)</div><div class="line">├── README.md    (可选，Github上的说明内容)</div><div class="line">├── data         (文件夹，含有示例数据)</div><div class="line">├── man          (文件夹，包的详细文档)</div><div class="line">├── tests        (文件夹，正式软件测试代码)</div><div class="line">└── vignettes    (文件夹，含有高质量的文档)</div><div class="line">    ├── biocOv1.Rmd</div><div class="line">    ├── biocOv1.html</div></pre></td></tr></table></figure>
<p>打包协议文档“Writing R Extensions”提供了完整的详细信息。使用R命令 <code>R CMD build [foldername]</code> 将对程度包文件夹的内容进行打包，从而创建可使用<code>R CMD install[archivename]</code> 添加到R安装的文件。不过现在都使用R Studio来完成这些任务。</p>
<h3 id="模块化与包的相互依赖">模块化与包的相互依赖</h3>
<p>打包协议可以帮助我们隔离执行有限操作集的软件，并识别随时间固有变化的程序集合的版本。没有客观的方法来确定一组操作是否适合打包。一些非常有用的包仅执行少量任务，而其他包则具有非常广泛的应用范围。重要的是打包的概念允许软件的模块化。这在两个方面很重要：范围和时间。范围的模块化对于允许并行独立开发解决不同问题的软件工具非常重要。时间上的模块化对于识别行为稳定的软件版本是很重要的。</p>
<h3 id="持续集成测试包的正确性和互操作性">持续集成：测试包的正确性和互操作性</h3>
<p>下列的图片就是开发Bioconductor分支的一个构建报告的<a href="http://bioconductor.org/checkResults/3.6/bioc-LATEST/" target="_blank" rel="external">快照</a>：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191105142824.png">

</div>
<p>在上面的表格中，一共有6列，其中最后1列标明了“Installed Pkgs”，其中2000表示用于Linux平台，这个数据在不同平台之间有所不同，并且通常会伴随着开发分支的时间而增加。</p>
<h2 id="汇总">汇总</h2>
<p>Bioconductor的核心开发者小组致力于开发数据结构，使用户能够方便地处理基因组和基因组规模的数据。用于设计的结构是为了支持基因组规模生物学实验的主要目标：</p>
<ul>
<li>解析由微阵列或测序仪产生的大规模数据集。</li>
<li>对（相对）原始数据进行预处理，以支持可靠的统计解释。</li>
<li>将分析量化与样本信息数据进行结合，用于检验分子过程和生物体水平特征（例如生长、疾病状态）之间关系的假设。</li>
<li>在本课程里，我们将回顾你可以在自己的研究中用于执行这些相关任务的无明和功能。</li>
</ul>
<h1 id="x的基因前提和概述">5x的基因前提和概述</h1>
<p>你知道如何使用R来操作和分析数据后，并且还对统计建模有不错的理解后。Bioconductor项目就会证明，在进行基因组级规模的计算生物操作时（但非所有），R是一个有效的工具。</p>
<p>将Bioconductor与处理基因组数据的其它软件系统的不同之处在于：</p>
<ul>
<li>使用面向对象的设计理解统一了基因组实验中出现的不同数据类型；</li>
<li>致力于基因组注释的可互操作结构，从核苷酸到人口规模；</li>
<li>发布和开发周期的持续集成原则，在多个广泛使用的计算平台上进行日常测试。我们这个为期四周的模块化学习目标旨在针对基因组规模数据分析的方方面面建立起对该系统使用的专业能力。</li>
</ul>
<p>525.5x课程主要为了四个主要部分，每部分花费一周时间来学习，内容如下：</p>
<ul>
<li>动机和技术：我们为什么要检测，以及如何检测数据，以及我们如何使用R语言来管理数据。</li>
<li>基因组注释：尤其是要注意基因组坐标中的区间（ranges，有的译为“范围”）在识别基因组结构方面的作用。</li>
<li>基因组级数据的预处理概念，重要如何通过Bioconductor来实现。</li>
<li>使用Bioconductor来做基因组级数据的假设检验。</li>
</ul>
<p>Some of the fundamental concepts that distinguish Bioconductor from other software systems addressing genome-scale data are</p>
<p>本章的不同小节将简要描述一下这些概念，以及说明如何进行计算。</p>
<h2 id="动机与技术">动机与技术</h2>
<p>“What we measure and why”这个视频（这个视频位于Youtube上）给出了基本生物过程的示意图，我们可以通过计算来研究这些过程。我们注意到，对有机体的生命过程至关重要的所有蛋白质的编码序列都位于有机体的基因组DNA中。研究生物体之间的差异， 以及生物体内的某些变化（例如肿瘤的发展），通常涉及基因组DNA序列的计算。</p>
<p>Bioconductor提供了处理许多生物体基因组DNA序列的工具。一种处理序列的基本方法计算就是给定一个“参考序列”，然后我们计算另外一条序列与参考序列之间的差异。</p>
<h3 id="参考序列的获取">参考序列的获取</h3>
<p>使用Bioconductor很容易处理人类(Homo sapiens)的参考序列。现在我们来看一下17号染色体，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">Hsapiens$chr17</div><div class="line"><span class="comment">##   81195210-letter "DNAString" instance</span></div><div class="line"><span class="comment">## seq: AAGCTTCTCACCCTGTTCCTGCATAGATAATTGC...GGTGTGGGTGTGGTGTGTGGGTGTGGGTGTGGT</span></div></pre></td></tr></table></figure>
<p>我们注意到：</p>
<ul>
<li>我们通过了一个R包来提供了序列。</li>
<li>包的名称表明了这个是来源于UCSC的参考基因组hg19。</li>
<li>我们使用美元符号 <code>$</code> 提取了染色体的序列。</li>
</ul>
<h3 id="表示dna突变体">表示DNA突变体</h3>
<p>与参考序列有单个偏离的标准表示方法就是使用VCF格式的文件（ <a href="http://samtools.github.io/hts-specs/VCFv4.1.pdf" target="_blank" rel="external">Variant Call Format</a>）。<code>VariantAnnotation</code>包就含有这样一个案例。我们有两个某些DNA突变体的两个高质量表示方法(high-level representations)，即VCF内容中的摘要，以及突变体本身在基因组上的地址，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div></pre></td><td class="code"><pre><div class="line">fl &lt;- system.file(<span class="string">"extdata"</span>, <span class="string">"ex2.vcf"</span>, package=<span class="string">"VariantAnnotation"</span>) </div><div class="line">vcf &lt;- readVcf(fl, <span class="string">"hg19"</span>)</div><div class="line">vcf</div><div class="line"><span class="comment">## class: CollapsedVCF </span></div><div class="line"><span class="comment">## dim: 5 3 </span></div><div class="line"><span class="comment">## rowRanges(vcf):</span></div><div class="line"><span class="comment">##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></div><div class="line"><span class="comment">## info(vcf):</span></div><div class="line"><span class="comment">##   DataFrame with 6 columns: NS, DP, AF, AA, DB, H2</span></div><div class="line"><span class="comment">## info(header(vcf)):</span></div><div class="line"><span class="comment">##       Number Type    Description                </span></div><div class="line"><span class="comment">##    NS 1      Integer Number of Samples With Data</span></div><div class="line"><span class="comment">##    DP 1      Integer Total Depth                </span></div><div class="line"><span class="comment">##    AF A      Float   Allele Frequency           </span></div><div class="line"><span class="comment">##    AA 1      String  Ancestral Allele           </span></div><div class="line"><span class="comment">##    DB 0      Flag    dbSNP membership, build 129</span></div><div class="line"><span class="comment">##    H2 0      Flag    HapMap2 membership         </span></div><div class="line"><span class="comment">## geno(vcf):</span></div><div class="line"><span class="comment">##   SimpleList of length 4: GT, GQ, DP, HQ</span></div><div class="line"><span class="comment">## geno(header(vcf)):</span></div><div class="line"><span class="comment">##       Number Type    Description      </span></div><div class="line"><span class="comment">##    GT 1      String  Genotype         </span></div><div class="line"><span class="comment">##    GQ 1      Integer Genotype Quality </span></div><div class="line"><span class="comment">##    DP 1      Integer Read Depth       </span></div><div class="line"><span class="comment">##    HQ 2      Integer Haplotype Quality</span></div><div class="line">rowRanges(vcf)</div><div class="line"><span class="comment">## GRanges object with 5 ranges and 5 metadata columns:</span></div><div class="line"><span class="comment">##                  seqnames             ranges strand | paramRangeID</span></div><div class="line"><span class="comment">##                     &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt;</span></div><div class="line"><span class="comment">##        rs6054257       20 [  14370,   14370]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##     20:17330_T/A       20 [  17330,   17330]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##        rs6040355       20 [1110696, 1110696]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##   20:1230237_T/.       20 [1230237, 1230237]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##        microsat1       20 [1234567, 1234569]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##                             REF                ALT      QUAL      FILTER</span></div><div class="line"><span class="comment">##                  &lt;DNAStringSet&gt; &lt;DNAStringSetList&gt; &lt;numeric&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##        rs6054257              G                  A        29        PASS</span></div><div class="line"><span class="comment">##     20:17330_T/A              T                  A         3         q10</span></div><div class="line"><span class="comment">##        rs6040355              A                G,T        67        PASS</span></div><div class="line"><span class="comment">##   20:1230237_T/.              T                           47        PASS</span></div><div class="line"><span class="comment">##        microsat1            GTC             G,GTCT        50        PASS</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们需要注意：</p>
<ul>
<li>演示中的数据已经包含在了包中，它仅用于说明和测试</li>
<li>变量 <code>vcf</code> 向用户简洁地展示了信息</li>
<li>使用<code>rowRanges</code>函数提取了突变体在hg19基因组中的位置，并与标签一同显示</li>
</ul>
<h3 id="检测基因表达">检测基因表达</h3>
<p>我们将以观察模式生物酿酒酵母(Sacchomyces Cerevisiae)中基因表达的检测数值来结束这一部分的讨论。现在我们来看一个实验，这个实验检测了酿酒酵母生殖周期中一系列时间点上的全基因组mRNA丰度。我们会使用R包来管理这些数据，并且我们使用特殊的Bioconductor定义 的数据结构来访问有关的实验和结果信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(yeastCC)</div><div class="line">data(spYCCES)</div><div class="line">spYCCES</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 6178 features, 77 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: cln3_40 cln3_30 ... elu_390 (77 total)</span></div><div class="line"><span class="comment">##   varLabels: syncmeth time</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData: none</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">##   pubMedIds: 9843569 </span></div><div class="line"><span class="comment">## Annotation:</span></div><div class="line">experimentData(spYCCES)</div><div class="line"><span class="comment">## Experiment data</span></div><div class="line"><span class="comment">##   Experimenter name: Spellman PT </span></div><div class="line"><span class="comment">##   Laboratory: Department of Genetics, Stanford University Medical Center, Stanford, California 94306-5120, USA. </span></div><div class="line"><span class="comment">##   Contact information:  </span></div><div class="line"><span class="comment">##   Title: Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. </span></div><div class="line"><span class="comment">##   URL:  </span></div><div class="line"><span class="comment">##   PMIDs: 9843569 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   Abstract: A 150 word abstract is available. Use 'abstract' method.</span></div></pre></td></tr></table></figure>
<p>经过一段时间的学习，你就会在接下来的几周内成为这方面的高手，你可以将整个细胞周期中参与调控的基因随时间变化的过程绘制出来。</p>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/biointro-lkycc2-1.png" alt="plot of chunk lkycc2">
<p class="caption">plot of chunk lkycc2</p>
</div>
<p>需要注意：</p>
<ul>
<li>关于实验的元数据已经绑定到R包的数据中（通过 <code>experimentData</code>可以可看PubmedID和摘要）；</li>
<li>我们使用简单的语法就能选择这些复杂实验设计中的信息；在这个案例中，我们使用<code>spYCCES[, spYCCES$syncmeth==&quot;alpha&quot;]</code> 就能提取出控制alpha pheromone的基因；</li>
<li>R的绘图工具支持通用的绘图注释与增强功能 (plot annotation and enhancemetn)；</li>
<li>统计模型工具能够直接帮我们区分周期与非周期的基因。</li>
</ul>
<h2 id="结束语">结束语</h2>
<p>你即将学习一些关于基因组结构和检测它们的分子生物学技术的高水平讲座。当你遇到这些概念时，请你记住那些与理解结构以及数据处理过程相关的计算思路与方法。在Bioconductor中找到相应的计算工具，并熟练掌握这些工具。如果你找不到它们，请通知我们，或者是如果满足某些需求的工具不存在，你也可以选择构建它们。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://rafalab.github.io/pages/harvardx.html" target="_blank" rel="external">HarvardX Biomedical Data Science Open Online Training</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/biointro.html" target="_blank" rel="external">Bioconductor for genome-scale data – quick intro</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/02/Genomics Data Analysis Series/GDAS002-Bioconductor备忘录/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/02/Genomics Data Analysis Series/GDAS002-Bioconductor备忘录/" itemprop="url">GDAS002-Bioconductor备忘录</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-02T12:00:00+08:00">
                2019-09-02
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,455
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  12
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这篇笔记的主要内容是关于Bioconductor的一些备忘录(cheat sheet)，作者总结的不错，我就按原文翻译了下来，由于一些知识点并没有学到，因此未翻译，后面学到后，再回头过来翻译。</p>
<h2 id="安装">安装</h2>
<p>安装细节参考http://bioconductor.org/install/，具体代码如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">if (!requireNamespace(&quot;BiocManager&quot;))</div><div class="line">    install.packages(&quot;BiocManager&quot;)</div><div class="line">BiocManager::install()</div><div class="line">BiocManager::install(c(&quot;package1&quot;,&quot;package2&quot;)</div><div class="line">BiocManager::valid() # are packages up to date?</div><div class="line"></div><div class="line"># what Bioc version is release right now?</div><div class="line">http://bioconductor.org/bioc-version</div><div class="line"># what Bioc versions are release/devel?</div><div class="line">http://bioconductor.org/js/versions.js</div></pre></td></tr></table></figure>
<h2 id="r语言帮助">R语言帮助</h2>
<p>简单的帮忙如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">?函数名</div><div class="line">?&quot;eSet-class&quot; # 如果要查看类的帮助，需要后缀&apos;-class&apos;</div><div class="line">help(package=&quot;foo&quot;,help_type=&quot;html&quot;) # 使用web浏览器来查看帮助文档</div><div class="line">vignette(&quot;topic&quot;)</div><div class="line">browseVignettes(package=&quot;package&quot;) # 某包的简单案例</div></pre></td></tr></table></figure>
<p>高级用户的帮助功能</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">functionName # 直接输入函数名（不带括号），则显示函数代码</div><div class="line">getMethod(method,&quot;class&quot;)  # 显示某类的方法代码</div><div class="line">selectMethod(method, &quot;class&quot;) # 查看继承来的方法</div><div class="line">showMethods(classes=&quot;class&quot;) # 显示某类的所有方法show all methods for class</div><div class="line">methods(class=&quot;GRanges&quot;) # 此函数限于R &gt;=3.2</div><div class="line">?&quot;functionName,class-method&quot; # 用于查看S4对象的方法文档</div><div class="line">?&quot;plotMA,data.frame-method&quot; # 需要加载geneplotter包，from library(geneplotter)</div><div class="line">?&quot;method.class&quot; # 查看S3对象的方法 </div><div class="line">?&quot;plot.lm&quot;</div><div class="line">sessionInfo() # 获取R语言版本的相关帮助信息 </div><div class="line">packageVersion(&quot;foo&quot;) # 查看某个包的版本</div></pre></td></tr></table></figure>
<p>Bioconductor相关的论坛： https://support.bioconductor.org</p>
<p>如果你使用的Rstudio，则可以使用<code>?</code>或<code>help</code>来查看帮忙文档，如果使用是命令行模式（例如Linux），则可以使用下面的代码通过浏览器来查看帮助文档（我使用的Rstudio，因此未尝试以下代码）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">alias rhelp=&quot;Rscript -e &apos;args &lt;- commandArgs(TRUE); help(args[2], package=args[3], help_type=\&quot;html\&quot;); Sys.sleep(5)&apos; --args&quot;</div></pre></td></tr></table></figure>
<h2 id="debugging-r">debugging R</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">traceback() # 查看哪一步导致了R错误 </div><div class="line"># debug a function</div><div class="line">debug(myFunction) # 逐行遍历函数中的代码 </div><div class="line">undebug(myFunction) # 停止debugging</div><div class="line">debugonce(myFunction) # 功能如上，但是不需要undebug()</div><div class="line"># 在写代码的过程中，可以将函数browser()放到一个关系的节点上</div><div class="line"># this plus devtools::load_all() can be useful for programming</div><div class="line"># to jump in function on error:</div><div class="line">options(error=recover)</div><div class="line"># turn that behavior off:</div><div class="line">options(error=NULL)</div><div class="line"># debug, e.g. estimateSizeFactors from DESeq2...</div><div class="line"># debugging an S4 method is more difficult; this gives you a peek inside:</div><div class="line">trace(estimateSizeFactors, browser, exit=browser, signature=&quot;DESeqDataSet&quot;)</div></pre></td></tr></table></figure>
<h2 id="显示某个类的包特异方法">显示某个类的包特异方法</h2>
<p>以下两个段代码实现的功能大致相同：查看特定的某个包中，某个特定类对象的操作方法。给定类的对象的操作方法。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">intersect(sapply(strsplit(as.character(methods(class=&quot;DESeqDataSet&quot;)), &quot;,&quot;), `[`, 1), ls(&quot;package:DESeq2&quot;))</div><div class="line">sub(&quot;Function: (.*) \\(package .*\\)&quot;,&quot;\\1&quot;,grep(&quot;Function&quot;,showMethods(classes=&quot;DESeqDataSet&quot;, </div><div class="line">    where=getNamespace(&quot;DESeq2&quot;), printTo=FALSE), value=TRUE))</div></pre></td></tr></table></figure>
<h2 id="注释annotations">注释Annotations</h2>
<p>这里需要先更新一下AnnotationHub，现在我们查看一下AnnotationHub的某个案例：</p>
<p>https://master.bioconductor.org/packages/release/bioc/html/AnnotationHub.html</p>
<p>以下代码是如何使用数据库包，以及<code>biomart</code>，先看<a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html" target="_blank" rel="external">AnnotationDbi</a>：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"># 某用包个注释包</div><div class="line">library(AnnotationDbi)</div><div class="line">library(org.Hs.eg.db) # 这个是人类注释包，即Homo.sapiens</div><div class="line">columns(org.Hs.eg.db)</div><div class="line">keytypes(org.Hs.eg.db) # columns与keytypes返回的结果一样，即有什么样的注释内容</div><div class="line">k &lt;- head(keys(org.Hs.eg.db, keytype=&quot;ENTREZID&quot;))</div><div class="line"></div><div class="line"># 返回一个字符向量（1对1的关系），查看mapIds则可以查看多个选项</div><div class="line">res &lt;- mapIds(org.Hs.eg.db, keys=k, column=&quot;ENSEMBL&quot;, keytype=&quot;ENTREZID&quot;)</div><div class="line"></div><div class="line"># 产生1对多的映射关系 </div><div class="line">res &lt;- select(org.Hs.eg.db, keys=k,</div><div class="line">  columns=c(&quot;ENTREZID&quot;,&quot;ENSEMBL&quot;,&quot;SYMBOL&quot;),</div><div class="line">  keytype=&quot;ENTREZID&quot;)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html" target="_blank" rel="external">biomaRt</a></p>
<p>这个包主要用于各种基因ID的转换。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"># 使用biomart包，将一个注释与另外一个注释进行映射</div><div class="line">library(biomaRt)</div><div class="line">m &lt;- useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)</div><div class="line">map &lt;- getBM(mart = m,</div><div class="line">  attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene&quot;),</div><div class="line">  filters = &quot;ensembl_gene_id&quot;, </div><div class="line">  values = some.ensembl.genes)</div></pre></td></tr></table></figure>
<h2 id="genomicranges包">GenomicRanges包</h2>
<p><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" target="_blank" rel="external">GenomicRanges</a>这个包用于分析高通量测序数据，它能有效地表壳与操作基因组注释信息(genomic annotations)与比对结果(alignments)。GenomicRanges包定义了一个通用容器，此容器用于储存与操作基因间隔(genomic intervals)与变量(variables)。在GenomicAlignments包和SummarizedExperiment包中定义了一个更加特殊的容器，GenomicAlignments包中的这个更加特殊的容器可以根据一个参考基因组来表示和操作一些短比对结果(short alignments)，SummarizedExperiment包中容器可以构建一个实验的矩阵样汇总信息。</p>
<p>此包的使用如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">library(GenomicRanges)</div><div class="line">z &lt;- GRanges(&quot;chr1&quot;,IRanges(1000001,1001000),strand=&quot;+&quot;)</div><div class="line">start(z)</div><div class="line">end(z)</div><div class="line">width(z)</div><div class="line">strand(z)</div><div class="line">mcols(z) # the &apos;metadata columns&apos;, any information stored alongside each range</div><div class="line">ranges(z) # gives the IRanges</div><div class="line">seqnames(z) # the chromosomes for each ranges</div><div class="line">seqlevels(z) # the possible chromosomes</div><div class="line">seqlengths(z) # the lengths for each chromosome</div></pre></td></tr></table></figure>
<h3 id="intra-range-methods">Intra-range methods</h3>
<p>不清楚功能，后面补充。</p>
<p>Affects ranges independently</p>
<table>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>shift</td>
<td>moves left/right</td>
</tr>
<tr class="even">
<td>narrow</td>
<td>narrows by relative position within range</td>
</tr>
<tr class="odd">
<td>resize</td>
<td>resizes to width, fixing start for +, end for -</td>
</tr>
<tr class="even">
<td>flank</td>
<td>returns flanking ranges to the left +, or right -</td>
</tr>
<tr class="odd">
<td>promoters</td>
<td>similar to flank</td>
</tr>
<tr class="even">
<td>restrict</td>
<td>restricts ranges to a start and end position</td>
</tr>
<tr class="odd">
<td>trim</td>
<td>trims out of bound ranges</td>
</tr>
<tr class="even">
<td>+/-</td>
<td>expands/contracts by adding/subtracting fixed amount</td>
</tr>
<tr class="odd">
<td>*</td>
<td>zooms in (positive) or out (negative) by multiples</td>
</tr>
</tbody>
</table>
<h3 id="inter-range-methods">Inter-range methods</h3>
<p>Affects ranges as a group</p>
<table>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>range</td>
<td>one range, leftmost start to rightmost end</td>
</tr>
<tr class="even">
<td>reduce</td>
<td>cover all positions with only one range</td>
</tr>
<tr class="odd">
<td>gaps</td>
<td>uncovered positions within range</td>
</tr>
<tr class="even">
<td>disjoin</td>
<td>breaks into discrete ranges based on original starts/ends</td>
</tr>
</tbody>
</table>
<h3 id="nearest-methods">Nearest methods</h3>
<p>Given two sets of ranges, <code>x</code> and <code>subject</code>, for each range in <code>x</code>, returns…</p>
<table style="width:100%;">
<colgroup>
<col width="22%">
<col width="77%">
</colgroup>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nearest</td>
<td>index of the nearest neighbor range in subject</td>
</tr>
<tr class="even">
<td>precede</td>
<td>index of the range in subject that is directly preceded by the range in x</td>
</tr>
<tr class="odd">
<td>follow</td>
<td>index of the range in subject that is directly followed by the range in x</td>
</tr>
<tr class="even">
<td>distanceToNearest</td>
<td>distances to its nearest neighbor in subject (Hits object)</td>
</tr>
<tr class="odd">
<td>distance</td>
<td>distances to nearest neighbor (integer vector)</td>
</tr>
</tbody>
</table>
<p>A Hits object can be accessed with <code>queryHits</code>, <code>subjectHits</code> and <code>mcols</code> if a distance is associated.</p>
<h3 id="set-methods">set methods</h3>
<p>If <code>y</code> is a GRangesList, then use <code>punion</code>, etc. All functions have default <code>ignore.strand=FALSE</code>, so are strand specific.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">union(x,y) </div><div class="line">intersect(x,y)</div><div class="line">setdiff(x,y)</div></pre></td></tr></table></figure>
<h3 id="overlaps">Overlaps</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">x %over% y  # logical vector of which x overlaps any in y</div><div class="line">fo &lt;- findOverlaps(x,y) # returns a Hits object</div><div class="line">queryHits(fo)   # which in x</div><div class="line">subjectHits(fo) # which in y</div></pre></td></tr></table></figure>
<h3 id="seqnames-and-seqlevels">Seqnames and seqlevels</h3>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html" target="_blank" rel="external">GenomicRanges</a> and <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html" target="_blank" rel="external">GenomeInfoDb</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">gr.sub &lt;- gr[seqlevels(gr) == &quot;chr1&quot;]</div><div class="line">seqlevelsStyle(x) &lt;- &quot;UCSC&quot; # convert to &apos;chr1&apos; style from &quot;NCBI&quot; style &apos;1&apos;</div></pre></td></tr></table></figure>
<h2 id="sequences">Sequences</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Biostrings.html" target="_blank" rel="external">Biostrings</a>是一个操作序列或序列集的包，简要操作可以参考<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/BiostringsQuickOverview.pdf" target="_blank" rel="external">Biostrings Quick Overview PDF</a>，关于一些物种的，缩写的信息可以参考<a href="http://genomicsclass.github.io/book/pages/annoCheat.html" target="_blank" rel="external">cheat sheet for annotation</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">dnastringset &lt;- getSeq(Hsapiens, granges) # returns a DNAStringSet</div><div class="line"># also Views() for Bioconductor &gt;= 3.1</div></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(Biostrings)</div><div class="line">dnastringset &lt;- readDNAStringSet(&quot;transcripts.fa&quot;)</div></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">substr(dnastringset, 1, 10) # to character string</div><div class="line">subseq(dnastringset, 1, 10) # returns DNAStringSet</div><div class="line">Views(dnastringset, 1, 10) # lightweight views into object</div><div class="line">complement(dnastringset)</div><div class="line">reverseComplement(dnastringset)</div><div class="line">matchPattern(&quot;ACGTT&quot;, dnastring) # also countPattern, also works on Hsapiens/genome</div><div class="line">vmatchPattern(&quot;ACGTT&quot;, dnastringset) # also vcountPattern</div><div class="line">letterFrequecy(dnastringset, &quot;CG&quot;) # how many C&apos;s or G&apos;s</div><div class="line"># also letterFrequencyInSlidingView</div><div class="line">alphabetFrequency(dnastringset, as.prob=TRUE)</div><div class="line"># also oligonucleotideFrequency, dinucleotideFrequency, trinucleotideFrequency</div><div class="line"># transcribe/translate for imitating biological processes</div></pre></td></tr></table></figure>
<h2 id="测序数据">测序数据</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsamtools.html" target="_blank" rel="external">Rsamtools</a>中的 <code>scanBam</code> 返回BAM文件中的原始数值：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(Rsamtools)</div><div class="line">which &lt;- GRanges(&quot;chr1&quot;,IRanges(1000001,1001000))</div><div class="line">what &lt;- c(&quot;rname&quot;,&quot;strand&quot;,&quot;pos&quot;,&quot;qwidth&quot;,&quot;seq&quot;)</div><div class="line">param &lt;- ScanBamParam(which=which, what=what)</div><div class="line"># for more BamFile functions/details see ?BamFile</div><div class="line"># yieldSize for chunk-wise access</div><div class="line">bamfile &lt;- BamFile(&quot;/path/to/file.bam&quot;)</div><div class="line">reads &lt;- scanBam(bamfile, param=param)</div><div class="line">res &lt;- countBam(bamfile, param=param) </div><div class="line"># for more sophisticated counting modes</div><div class="line"># see summarizeOverlaps() below</div><div class="line"></div><div class="line"># quickly check chromosome names</div><div class="line">seqinfo(BamFile(&quot;/path/to/file.bam&quot;))</div><div class="line"></div><div class="line"># DNAStringSet is defined in the Biostrings package</div><div class="line"># see the Biostrings Quick Overview PDF</div><div class="line">dnastringset &lt;- scanFa(fastaFile, param=granges)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html" target="_blank" rel="external">GenomicAlignments</a> 返回一个Bioconductor对象(GRanges-based)</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line">ga &lt;- readGAlignments(bamfile) # single-end</div><div class="line">ga &lt;- readGAlignmentPairs(bamfile) # paired-end</div></pre></td></tr></table></figure>
<h2 id="转录本数据库">转录本数据库</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html" target="_blank" rel="external">GenomicFeatures</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div></pre></td><td class="code"><pre><div class="line"># get a transcript database, which stores exon, trancript, and gene information</div><div class="line">library(GenomicFeatures)</div><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line"></div><div class="line"># or build a txdb from GTF file (e.g. downloadable from Ensembl FTP site)</div><div class="line">txdb &lt;- makeTranscriptDbFromGFF(&quot;file.GTF&quot;, format=&quot;gtf&quot;)</div><div class="line"></div><div class="line"># or build a txdb from Biomart (however, not as easy to reproduce later)</div><div class="line">txdb &lt;- makeTranscriptDbFromBiomart(biomart = &quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)</div><div class="line"></div><div class="line"># in Bioconductor &gt;= 3.1, also makeTxDbFromGRanges</div><div class="line"></div><div class="line"># saving and loading</div><div class="line">saveDb(txdb, file=&quot;txdb.sqlite&quot;)</div><div class="line">loadDb(&quot;txdb.sqlite&quot;)</div><div class="line"></div><div class="line"># extracting information from txdb</div><div class="line">g &lt;- genes(txdb) # GRanges, just start to end, no exon/intron information</div><div class="line">tx &lt;- transcripts(txdb) # GRanges, similar to genes()</div><div class="line">e &lt;- exons(txdb) # GRanges for each exon</div><div class="line">ebg &lt;- exonsBy(txdb, by=&quot;gene&quot;) # exons grouped in a GRangesList by gene</div><div class="line">ebt &lt;- exonsBy(txdb, by=&quot;tx&quot;) # similar but by transcript</div><div class="line"></div><div class="line"># then get the transcript sequence</div><div class="line">txSeq &lt;- extractTranscriptSeqs(Hsapiens, ebt)</div></pre></td></tr></table></figure>
<h2 id="根据范围ranges和实验汇总信息">根据范围(ranges)和实验汇总信息</h2>
<p><code>SummarizedExperiment</code>是高维数据的储存类， 它与<code>GRanges</code>或<code>GRangesList</code>配合使用（使用每个基因外显子的read计数）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line">fls &lt;- list.files(pattern=&quot;*.bam$&quot;)</div><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line">ebg &lt;- exonsBy(txdb, by=&quot;gene&quot;)</div><div class="line"># see yieldSize argument for restricting memory</div><div class="line">bf &lt;- BamFileList(fls)</div><div class="line">library(BiocParallel)</div><div class="line">register(MulticoreParam(4))</div><div class="line"># lots of options in the man page</div><div class="line"># singleEnd, ignore.strand, inter.features, fragments, etc.</div><div class="line">se &lt;- summarizeOverlaps(ebg, bf)</div><div class="line"></div><div class="line"># operations on SummarizedExperiment</div><div class="line">assay(se) # the counts from summarizeOverlaps</div><div class="line">colData(se)</div><div class="line">rowRanges(se)</div></pre></td></tr></table></figure>
<p><a href="https://combine-lab.github.io/salmon/" target="_blank" rel="external">Salmon</a>是一个定量工具，如果数据占有GC依赖的数据，那么就添加上<code>--gcBias</code>参数，接着使用 <a href="http://bioconductor.org/pacakges/tximport" target="_blank" rel="external">tximport</a>，以下是使用案例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">coldata &lt;- read.table(&quot;samples.txt&quot;)</div><div class="line">rownames(coldata) &lt;- coldata$id</div><div class="line">files &lt;- coldata$files; names(files) &lt;- coldata$id</div><div class="line">txi &lt;- tximport(files, type=&quot;salmon&quot;, tx2gene=tx2gene)</div><div class="line">dds &lt;- DESeqDataSetFromTximport(txi, coldata, ~condition)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsubread.html" target="_blank" rel="external">Rsubread</a>包中的<code>featureCounts</code>函数也可以用于read计算，速度可能更快。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(Rsubread)</div><div class="line">res &lt;- featureCounts(files, annot.ext=&quot;annotation.gtf&quot;,</div><div class="line">  isGTFAnnotationFile=TRUE,</div><div class="line">  GTF.featureType=&quot;exon&quot;,</div><div class="line">  GTF.attrType=&quot;gene_id&quot;)</div><div class="line">res$counts</div></pre></td></tr></table></figure>
<h2 id="rna-seq分析方法">RNA-seq分析方法</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html" target="_blank" rel="external">DESeq2</a>：My preferred pipeline for DESeq2 users is to start with a lightweight transcript abundance quantifier such as <a href="https://combine-lab.github.io/salmon/" target="_blank" rel="external">Salmon</a> and to use <a href="http://bioconductor.org/packages/tximport" target="_blank" rel="external">tximport</a>, followed by <code>DESeqDataSetFromTximport</code>.</p>
<p>Here, <code>coldata</code> is a <em>data.frame</em> with <code>group</code> as a column.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">library(DESeq2)</div><div class="line"># from tximport</div><div class="line">dds &lt;- DESeqDataSetFromTximport(txi, coldata, ~ group)</div><div class="line"># from SummarizedExperiment</div><div class="line">dds &lt;- DESeqDataSet(se, ~ group)</div><div class="line"># from count matrix</div><div class="line">dds &lt;- DESeqDataSetFromMatrix(counts, coldata, ~ group)</div><div class="line"># minimal filtering helps keep things fast </div><div class="line"># one can set &apos;n&apos; to e.g. min(5, smallest group sample size)</div><div class="line">keep &lt;- rowSums(counts(dds) &gt;= 10) &gt;= n </div><div class="line">dds &lt;- dds[keep,]</div><div class="line">dds &lt;- DESeq(dds)</div><div class="line">res &lt;- results(dds) # no shrinkage of LFC, or:</div><div class="line">res &lt;- lfcShrink(dds, coef = 2, type=&quot;apeglm&quot;) # shrink LFCs</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank" rel="external">edgeR</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"># this chunk from the Quick start in the edgeR User Guide</div><div class="line">library(edgeR) </div><div class="line">y &lt;- DGEList(counts=counts,group=group)</div><div class="line">keep &lt;- filterByExpr(y)</div><div class="line">y &lt;- y[keep,]</div><div class="line">y &lt;- calcNormFactors(y)</div><div class="line">design &lt;- model.matrix(~group)</div><div class="line">y &lt;- estimateDisp(y,design)</div><div class="line">fit &lt;- glmFit(y,design)</div><div class="line">lrt &lt;- glmLRT(fit)</div><div class="line">topTags(lrt)</div><div class="line"># or use the QL methods:</div><div class="line">qlfit &lt;- glmQLFit(y,design)</div><div class="line">qlft &lt;- glmQLFTest(qlfit)</div><div class="line">topTags(qlft)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html" target="_blank" rel="external">limma-voom</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">library(limma)</div><div class="line">design &lt;- model.matrix(~ group)</div><div class="line">y &lt;- DGEList(counts)</div><div class="line">keep &lt;- filterByExpr(y)</div><div class="line">y &lt;- y[keep,]</div><div class="line">y &lt;- calcNormFactors(y)</div><div class="line">v &lt;- voom(y,design)</div><div class="line">fit &lt;- lmFit(v,design)</div><div class="line">fit &lt;- eBayes(fit)</div><div class="line">topTable(fit)</div></pre></td></tr></table></figure>
<p>更多有关RNA-seq操作的包可以参考：<a href="http://www.bioconductor.org/packages/release/BiocViews.html#___RNASeq" target="_blank" rel="external">Many more RNA-seq packages</a></p>
<h2 id="表达数据集expression-set">表达数据集Expression set</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(Biobase)</div><div class="line">data(sample.ExpressionSet)</div><div class="line">e &lt;- sample.ExpressionSet</div><div class="line">exprs(e)</div><div class="line">pData(e)</div><div class="line">fData(e)</div></pre></td></tr></table></figure>
<h2 id="下载geo数据库">下载GEO数据库</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(GEOquery)</div><div class="line">e &lt;- getGEO(&quot;GSE9514&quot;)</div></pre></td></tr></table></figure>
<h2 id="芯片分析">芯片分析</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">library(affy)</div><div class="line">library(limma)</div><div class="line">phenoData &lt;- read.AnnotatedDataFrame(&quot;sample-description.csv&quot;)</div><div class="line">eset &lt;- justRMA(&quot;/celfile-directory&quot;, phenoData=phenoData)</div><div class="line">design &lt;- model.matrix(~ Disease, pData(eset))</div><div class="line">fit &lt;- lmFit(eset, design)</div><div class="line">efit &lt;- eBayes(fit)</div><div class="line">topTable(efit, coef=2)</div></pre></td></tr></table></figure>
<h2 id="icobra包">iCOBRA包</h2>
<p>iCOBRA包用于计算和可视化排序数据以及二分类匹配后的数据。例如，iCOBRA包的一个典型用途就是用于评估基因表达实验中不同组之间的比较方法，我们可以视之为一个排序问题（评估正确的效应量(effect size)以及按照显著性进行排列的基因），还是一个二分类比较问题（binary assignment problem）（例如将基因分为差异表达和非差异表达）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">library(iCOBRA)</div><div class="line">cd &lt;- COBRAData(pval=pval.df, padj=padj.df, score=score.df, truth=truth.df)</div><div class="line">cp &lt;- calculate_performance(cd, binary_truth = &quot;status&quot;, cont_truth = &quot;logFC&quot;)</div><div class="line">cobraplot &lt;- prepare_data_for_plot(cp)</div><div class="line">plot_fdrtprcurve(cobraplot)</div><div class="line"># interactive shiny app:</div><div class="line">COBRAapp(cd)</div></pre></td></tr></table></figure>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><p><a href="http://genomicsclass.github.io/book/pages/annoCheat.html" target="_blank" rel="external">Genomic annotation in Bioconductor: Cheat sheet</a></p></li>
<li><p><a href="http://rafalab.github.io/pages/harvardx.html" target="_blank" rel="external">HarvardX Biomedical Data Science Open Online Training</a></p></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/01/Genomics Data Analysis Series/GDAS001-安装Bioconductor与获取帮助/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/01/Genomics Data Analysis Series/GDAS001-安装Bioconductor与获取帮助/" itemprop="url">GDAS001-安装Bioconductor与获取帮助</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-01T12:00:00+08:00">
                2019-09-01
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,400
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  6
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="安装bioconductor">安装Bioconductor</h2>
<p>使用以下代码安装Bioconductor，直接粘贴到R的控制台即可。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">source(&quot;http://bioconductor.org/biocLite.R&quot;)</div><div class="line">biocLite()</div></pre></td></tr></table></figure>
<p>这两行代码会安装核心的Bioconductor包。如果要再安装其它的包，则要使用<code>biocLite()</code>函数，并且使用字符串参数指定包的名称，例如我们安装以下的两个包：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">biocLite(c(&quot;genefilter&quot;,&quot;geneplotter&quot;))</div></pre></td></tr></table></figure>
<p>在使用的时候，就跟常规的R包一样了，例如<code>library()</code>加载即可，例如 <code>library(genefilter)</code>。</p>
<h2 id="获取帮助">获取帮助</h2>
<p>R中的帮助函数很多， 如果在命令行中直接输入 <code>help.start()</code> 就能查看R帮助起始页面。</p>
<h2 id="查看函数帮助文档">查看函数帮助文档</h2>
<p>通常情况，使用<code>?</code>后面添加函数名称，回车则可以查看此函数的帮助文档。使用<code>example(函数)</code>则能查看该函数的帮助文档中Examples部分中的案例，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">?mean</div><div class="line">?mad</div><div class="line">example(mad)</div><div class="line">example(boxplot)</div></pre></td></tr></table></figure>
<p>输入函数名称，但不添加任何参数，则显示此函数的所有代码。</p>
<p>在函数的帮助文档中，我们可以看到许多内容，包括关于此函数的描述，用法，参数，细节，参考文献，使用案例等。</p>
<h2 id="包帮助">包帮助</h2>
<p>如果想查看某个包中所有函数的帮助文档，那么此时可以直接查看该函数的包文档，如果想使用Web浏览器来查看这些文档，那么需要添加以下参数（R默认的IDE）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">help(package=&quot;genefilter&quot;, help_type=&quot;html&quot;)</div></pre></td></tr></table></figure>
<p>如果使用的Rstudio，那么上述命令与下面命令的功能是一样的：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">help(package=&quot;genefilter&quot;)</div></pre></td></tr></table></figure>
<p>一种查看某个包中的有哪些函数的快速方法是使用双冒号<code>::</code>，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(geneplotter)</div><div class="line">geneplotter::</div></pre></td></tr></table></figure>
<h2 id="对象帮助">对象帮助</h2>
<p>如果要在R中查看某个对象的帮助信息，可以按如下操作：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">class(6)</div><div class="line">?numeric</div><div class="line">?&quot;numeric-class&quot;</div></pre></td></tr></table></figure>
<p>不过有的时候，在一些包中，某个构建的函数与此函数构建的对象的帮助文档是一样的，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(Biobase)</div><div class="line">?ExpressionSet</div><div class="line">?&quot;ExpressionSet-class&quot;</div></pre></td></tr></table></figure>
<p>快速查看某个对象的方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">methods(class=&quot;ExpressionSet&quot;)</div><div class="line">methods(class=&quot;lm&quot;)</div></pre></td></tr></table></figure>
<p>使用<code>method()</code>函数则能查看由某个类定义的实例的所有方法：</p>
<p>R has good capabilities for self-description. Classes can be formally linked to methods that operate usefully on their instances. The methods available can be listed using the <code>methods</code> function.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">data(sample.ExpressionSet)</div><div class="line">methods(class=class(sample.ExpressionSet))</div><div class="line">##  [1] [                [[               [[&lt;-             $$               </div><div class="line">##  [5] $$&lt;-              abstract         annotation       annotation&lt;-    </div><div class="line">##  [9] as.data.frame    assayData        assayData&lt;-      classVersion    </div><div class="line">## [13] classVersion&lt;-   coerce           combine          description     </div><div class="line">## [17] description&lt;-    dim              dimnames         dimnames&lt;-      </div><div class="line">## [21] dims             esApply          experimentData   experimentData&lt;-</div><div class="line">## [25] exprs            exprs&lt;-          fData            fData&lt;-         </div><div class="line">## [29] featureData      featureData&lt;-    featureNames     featureNames&lt;-  </div><div class="line">## [33] fvarLabels       fvarLabels&lt;-     fvarMetadata     fvarMetadata&lt;-  </div><div class="line">## [37] initialize       isCurrent        isVersioned      KEGG2heatmap    </div><div class="line">## [41] KEGGmnplot       makeDataPackage  Makesense        notes           </div><div class="line">## [45] notes&lt;-          pData            pData&lt;-          phenoData       </div><div class="line">## [49] phenoData&lt;-      preproc          preproc&lt;-        protocolData    </div><div class="line">## [53] protocolData&lt;-   pubMedIds        pubMedIds&lt;-      rowMedians      </div><div class="line">## [57] rowQ             sampleNames      sampleNames&lt;-    show            </div><div class="line">## [61] storageMode      storageMode&lt;-    updateObject     updateObjectTo  </div><div class="line">## [65] varLabels        varLabels&lt;-      varMetadata      varMetadata&lt;-   </div><div class="line">## [69] write.exprs     </div><div class="line">## see &apos;?methods&apos; for accessing help and source code</div></pre></td></tr></table></figure>
<h2 id="源代码">源代码</h2>
<p>You can find the source code for many functions by typing out the name of the function without () and pressing enter.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">read.csv</div><div class="line"><span class="comment">## function (file, header = TRUE, sep = ",", quote = "\\"", dec = ".", </span></div><div class="line"><span class="comment">##     fill = TRUE, comment.char = "", ...) </span></div><div class="line"><span class="comment">## read.table(file = file, header = header, sep = sep, quote = quote, </span></div><div class="line"><span class="comment">##     dec = dec, fill = fill, comment.char = comment.char, ...)</span></div><div class="line"><span class="comment">## &lt;bytecode: 0x7f8e35b19200&gt;</span></div><div class="line"><span class="comment">## &lt;environment: namespace:utils&gt;</span></div></pre></td></tr></table></figure>
<p>我们从上面可以看出来，<code>read.csv()</code>函数只是打包了(wraps)了<code>read.table()</code>函数。</p>
<p>有的时候，你需要指定一个特定的类，才能查看该方法的源代码，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div></pre></td><td class="code"><pre><div class="line">plotMA</div><div class="line"><span class="comment">## nonstandardGenericFunction for "plotMA" defined from package "BiocGenerics"</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## function (object, ...) </span></div><div class="line"><span class="comment">## &#123;</span></div><div class="line"><span class="comment">##     standardGeneric("plotMA")</span></div><div class="line"><span class="comment">## &#125;</span></div><div class="line"><span class="comment">## &lt;environment: 0x7f8e338381b8&gt;</span></div><div class="line"><span class="comment">## Methods may be defined for arguments: object</span></div><div class="line"><span class="comment">## Use  showMethods("plotMA")  for currently available ones.</span></div><div class="line">showMethods(<span class="string">"plotMA"</span>)</div><div class="line"><span class="comment">## Function: plotMA (package BiocGenerics)</span></div><div class="line"><span class="comment">## object="ANY"</span></div><div class="line"><span class="comment">## object="data.frame"</span></div><div class="line">getMethod(<span class="string">"plotMA"</span>,<span class="string">"data.frame"</span>)</div><div class="line"><span class="comment">## Method Definition:</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## function (object, ...) </span></div><div class="line"><span class="comment">## &#123;</span></div><div class="line"><span class="comment">##     .local &lt;- function (object, ylim = NULL, colNonSig = "gray32", </span></div><div class="line"><span class="comment">##         colSig = "red3", colLine = "#ff000080", log = "x", cex = 0.45, </span></div><div class="line"><span class="comment">##         xlab = "mean expression", ylab = "log fold change", ...) </span></div><div class="line"><span class="comment">##     &#123;</span></div><div class="line"><span class="comment">##         if (!(ncol(object) == 3 &amp; inherits(object[[1]], "numeric") &amp; </span></div><div class="line"><span class="comment">##             inherits(object[[2]], "numeric") &amp; inherits(object[[3]], </span></div><div class="line"><span class="comment">##             "logical"))) &#123;</span></div><div class="line"><span class="comment">##             stop("When called with a data.frame, plotMA expects the data frame to have 3 columns, two numeric ones for mean and log fold change, and a logical one for significance.")</span></div><div class="line"><span class="comment">##         &#125;</span></div><div class="line"><span class="comment">##         colnames(object) &lt;- c("mean", "lfc", "sig")</span></div><div class="line"><span class="comment">##         object = subset(object, mean != 0)</span></div><div class="line"><span class="comment">##         py = object$lfc</span></div><div class="line"><span class="comment">##         if (is.null(ylim)) </span></div><div class="line"><span class="comment">##             ylim = c(-1, 1) * quantile(abs(py[is.finite(py)]), </span></div><div class="line"><span class="comment">##                 probs = 0.99) * 1.1</span></div><div class="line"><span class="comment">##         plot(object$mean, pmax(ylim[1], pmin(ylim[2], py)), log = log, </span></div><div class="line"><span class="comment">##             pch = ifelse(py &lt; ylim[1], 6, ifelse(py &gt; ylim[2], </span></div><div class="line"><span class="comment">##                 2, 16)), cex = cex, col = ifelse(object$sig, </span></div><div class="line"><span class="comment">##                 colSig, colNonSig), xlab = xlab, ylab = ylab, </span></div><div class="line"><span class="comment">##             ylim = ylim, ...)</span></div><div class="line"><span class="comment">##         abline(h = 0, lwd = 4, col = colLine)</span></div><div class="line"><span class="comment">##     &#125;</span></div><div class="line"><span class="comment">##     .local(object, ...)</span></div><div class="line"><span class="comment">## &#125;</span></div><div class="line"><span class="comment">## &lt;environment: namespace:geneplotter&gt;</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Signatures:</span></div><div class="line"><span class="comment">##         object      </span></div><div class="line"><span class="comment">## target  "data.frame"</span></div><div class="line"><span class="comment">## defined "data.frame"</span></div></pre></td></tr></table></figure>
<h2 id="简要案例vignettes">简要案例Vignettes</h2>
<p>“Vignettes”包含在R包中，并且是Bioconductor包的标准组件之一。这个函数通常会通过“块”代码的形式，来显示此包中函数的使用案例。</p>
<p>在Bioconductor的官网有PDF格式或R代码的Vignettes。此外，通过在R中来调用vignette能够保证正在使用的包的正确版本。以下代码就是列出某个特定包的vignettes名字，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">vignette(package=&quot;Biobase&quot;)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">Vignettes <span class="keyword">in</span> package ‘Biobase’:</div><div class="line"></div><div class="line">ExpressionSetIntroduction</div><div class="line">                  An introduction to Biobase and</div><div class="line">                  ExpressionSets (<span class="keyword">source</span>, pdf)</div><div class="line">esApply           esApply Introduction (<span class="keyword">source</span>, pdf)</div><div class="line">BiobaseDevelopment</div><div class="line">                  Notes <span class="keyword">for</span> eSet developers (<span class="keyword">source</span>,</div><div class="line">                  pdf)</div></pre></td></tr></table></figure>
<p>进行一步调用 <code>vignette</code> 能够浏览PDF格式的内容：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">vignette(&quot;ExpressionSetIntroduction&quot;)</div></pre></td></tr></table></figure>
<p>也可以使用Web浏览器的方式来查看vignette，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">browseVignettes(package=&quot;Biobase&quot;)</div></pre></td></tr></table></figure>
<h2 id="分析帮助">分析帮助</h2>
<p><a href="http://www.bioconductor.org/help/workflows/" target="_blank" rel="external">workflows repository</a> 这个库中包含了大量的可计算文档(computable documents)，以及描述了如何执行某些常见的分析，这些数据来源于已经发布的数据。</p>
<h2 id="总结">总结</h2>
<ul>
<li>Bioconductor上的所有R函数都有自己的文档，以及运行案例；</li>
<li>所有的Bioconductor包都有自己的vignettes，它描述了该包中的函数如何配合使用，并且是如何实现包的目标；</li>
<li>Bioconductor包含许多workflow documents，我们可以通过这些工作流程文档来进行学习；</li>
<li>R内置了许多基础函数用于完成基础的统计分析；</li>
<li>R’s <a href="https://stat.ethz.ch/mailman/listinfo/r-help" target="_blank" rel="external">mailing list for elementary questions</a> 与Bioconductor’s <a href="https://support.bioconductor.org/" target="_blank" rel="external">support site</a> 上含有非常多的信息，可以用于进一步的学习。</li>
</ul>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/installing_Bioconductor_finding_help.html" target="_blank" rel="external">Installing Bioconductor and finding help</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/01/Genomics Data Analysis Series/GDAS000-基因组数据分析系列笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/01/Genomics Data Analysis Series/GDAS000-基因组数据分析系列笔记/" itemprop="url">GDAS000-基因组数据分析系列笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-01T12:00:00+08:00">
                2019-09-01
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  385
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>以前在网上看到了哈佛大学的<a href="http://genomicsclass.github.io/book/" target="_blank" rel="external">PH525x系列课程</a>，这个系列教程主要是生物信息学相关的内容，一共是17章。这17章内容可以大致分为三部分。</p>
<p>第一部分：第1章到10章主要是讲的与生物信息学有关的统计学基础，参考书是《Data Analysis for the life sciences》，<a href="http://rvdsd.top/categories/Data-Analysis-for-the-life-sciences/">第1章到第10章</a>的笔记已经做完。</p>
<p>第二部分：第11章到第17章，涉及生物信息学的应用部分，例如Bioconductor上的几个包使用等，根据官网信息，这部分是《Genomics Data Analysis Series》，本来是想按照<a href="http://genomicsclass.github.io/book/" target="_blank" rel="external">PH525x series - Biomedical Data Science</a> 这个系列课程的目录来做笔记的，但是这里面的顺序并不太好，有部分重复，后来又发现了<a href="http://rafalab.github.io/pages/harvardx.html" target="_blank" rel="external">HarvardX Biomedical Data Science Open Online Training的课程目录</a>，这个就比较友好了，里面有视频，也有代码。因此就按后者的顺序来做笔记。</p>
<p>第三部分：一些使用案例，内容是《Using Python for Research》，这个后面也会有笔记。</p>
<h2 id="主要内容">主要内容</h2>
<p>基因组数据分析系列笔记主要分为以下几部分：</p>
<h3 id="第一部分bioconductor介绍">第一部分：Bioconductor介绍</h3>
<p>这一部分涉及Bioconductor的基本情况，包括注释，基因组分析与基因分析（Introduction to Bioconductor: Annotation and Analysis of Genomes and Genomic Assays，注：我是区别不了Analysis of Genomes与Genomic Assays）。</p>
<h3 id="第二部分使用biconducotor重现高通量数据的分析结果">第二部分：使用Biconducotor重现高通量数据的分析结果</h3>
<h3 id="第三部分功能基因组的案例分析">第三部分：功能基因组的案例分析</h3>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/28/DAL/DALS028_Batch_Effect04_FactorAnalysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/28/DAL/DALS028_Batch_Effect04_FactorAnalysis/" itemprop="url">DALS028-批次效应04-因子分析</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-28T12:00:00+08:00">
                2019-08-28
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Data-Analysis-for-the-life-sciences/" itemprop="url" rel="index">
                    <span itemprop="name">Data Analysis for the life sciences</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,816
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  17
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第10章批次效应的第3小节，这一部分的主要内容涉及批次效应(Batch Effects)的校正，有关这一部分Rmarkdown文档参见作者的<a href="https://github.com/genomicsclass/labs/blob/master/batch/factor_analysis.Rmd" target="_blank" rel="external">Github</a>。</p>
<h2 id="因子分析">因子分析</h2>
<p>在我们介绍另一种用于批量效应校正的统计方法之前，我们先介绍这种方法的核心统计思想：因子分析(Factor Analysis)。因子分子早是在一个多世纪前出现的。卡尔·皮尔森(Karl Pearson)指出，当计算学生之间不同科目的数据时，这些不同科目之间存在着相关性。为了解释这种现象，他提出了一个模型，该模型有一个能解释这种相关性的因子，对于每个学生来说，这个因子在各个学科中都是通用的(common)，如下所示： <span class="math display">\[
Y_ij = \alpha_i W_1 + \varepsilon_{ij}
\]</span></p>
<p>其中，<span class="math inline">\(Y_{ij}\)</span> 表示每个人 <span class="math inline">\(i\)</span> 的科目 <span class="math inline">\(j\)</span> 的成绩，而$_{i} $表示学生 <span class="math inline">\(i\)</span> 获得好成绩的能力。</p>
<p>在这个案例中，<span class="math inline">\(W_{1}\)</span> 是一个常数。这里我们使用一个稍微复杂的情况来说明因子分析，这种情况就类似于批次效应。我们生成一个随机的 <span class="math inline">\(N \times 6\)</span> 矩阵 <span class="math inline">\(Y\)</span> 来表示我们的研究对象，其中6表示不同的学科来表示 <span class="math inline">\(N\)</span> 表示 <span class="math inline">\(N\)</span> 个不同小孩。我们生成数据的方式就是，那些有一定相关性的学科。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">library(MASS)</div><div class="line">library(rafalib)</div><div class="line">n &lt;- 250</div><div class="line">p &lt;- 6</div><div class="line">set.seed(1)</div><div class="line">g &lt;- mvrnorm(n,c(0,0),matrix(c(1,0.5,0.5,1),2,2))</div><div class="line">Ystem &lt;- g[,1] + matrix(rnorm(n*p/2,0,0.65),n,p/2)</div><div class="line">Yhum &lt;- g[,2] + matrix(rnorm(n*p/2,0,0.65),n,p/2)</div><div class="line">Y &lt;- cbind(Ystem,Yhum)</div><div class="line">colnames(Y) &lt;- c(&quot;Math&quot;,&quot;Science&quot;,&quot;CS&quot;,&quot;Eng&quot;,&quot;Hist&quot;,&quot;Classics&quot;)</div></pre></td></tr></table></figure>
<h3 id="样本相关性">样本相关性</h3>
<p>我们要注意一下，这6个学科有着高度的相关性，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">round(cor(Y),2)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; round(cor(Y),<span class="number">2</span>)</div><div class="line">         Math Science   CS  Eng Hist Classics</div><div class="line">Math     <span class="number">1.00</span>    <span class="number">0.67</span> <span class="number">0.64</span> <span class="number">0.34</span> <span class="number">0.29</span>     <span class="number">0.28</span></div><div class="line">Science  <span class="number">0.67</span>    <span class="number">1.00</span> <span class="number">0.65</span> <span class="number">0.29</span> <span class="number">0.29</span>     <span class="number">0.26</span></div><div class="line">CS       <span class="number">0.64</span>    <span class="number">0.65</span> <span class="number">1.00</span> <span class="number">0.35</span> <span class="number">0.30</span>     <span class="number">0.29</span></div><div class="line">Eng      <span class="number">0.34</span>    <span class="number">0.29</span> <span class="number">0.35</span> <span class="number">1.00</span> <span class="number">0.71</span>     <span class="number">0.72</span></div><div class="line">Hist     <span class="number">0.29</span>    <span class="number">0.29</span> <span class="number">0.30</span> <span class="number">0.71</span> <span class="number">1.00</span>     <span class="number">0.68</span></div><div class="line">Classics <span class="number">0.28</span>    <span class="number">0.26</span> <span class="number">0.29</span> <span class="number">0.72</span> <span class="number">0.68</span>     <span class="number">1.00</span></div></pre></td></tr></table></figure>
<p>我们将使用图片来展示一下相关性，从下图我们可以发现，我们可以把学生分为STEM组与人文组（注：STEM与人文学科(humanities )其实就相当于中国的理科和文科，在这个案例中，STEM指的是科学(Science)，计算机(CS)和数学(Math)，人文学科指，古典学(Classics)，历史学(Hist)和英语语言文字(Eng)）。</p>
<p>在下图中，高度相关性的学科用红色表示，不相关的学科用白色表示，负相关的学科用蓝色表示，代码如下所示：</p>
<figure class="highlight plain"><figcaption><span>correlation_images,fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">library(RColorBrewer)</div><div class="line">mypar(1,2)</div><div class="line">cols=colorRampPalette(rev(brewer.pal(11,&quot;RdBu&quot;)))(100)</div><div class="line">eps = matrix(rnorm(n*p),n,p)</div><div class="line">par(mar = c(8.1, 8.1, 3.5, 2.1))</div><div class="line">image(1:ncol(Y),1:ncol(Y),cor(Y)[,6:1],xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=cols,xlab=&quot;&quot;,ylab=&quot;&quot;,zlim=c(-1,1),main=&quot;Actual Data&quot;)</div><div class="line">axis(1,1:ncol(Y),colnames(Y),las=2)</div><div class="line">axis(2,1:ncol(Y),rev(colnames(Y)),las=2)</div><div class="line">image(1:ncol(Y),1:ncol(Y),cor(eps)[,6:1],xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=cols,xlab=&quot;&quot;,ylab=&quot;&quot;,zlim=c(-1,1),main=&quot;Independet Data&quot;)</div><div class="line">axis(1,1:ncol(Y),colnames(Y),las=2)</div><div class="line">axis(2,1:ncol(Y),rev(colnames(Y)),las=2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917125200.jpeg">

</div>
<p>从上图中，我们可以看到以下信息：所有学科之间都有相关性，这表明学生成绩的背景有着某些隐藏因素（例如学术能力），正是这种能力导致了这些学科表现的相关性，因为在一个科目中获得高分的学生往往在其它科目也能获得高分。我们还能看到，这种相关性在STEM类学科之间更强，以及在人文学科之间更强。这就说明了，还有可能存在着另一种隐藏因素，这种因素决定了学生擅长STEM还是人文学科。现在我们使用一个统计模型来解释这种思路。</p>
<h3 id="因子模型">因子模型</h3>
<p>基于前面的图形展示结果，我们假设这些现象背后有2个隐藏因素 <span class="math inline">\(\mathbf{W}_1\)</span> 和 <span class="math inline">\(\mathbf{W}_2\)</span> ，我们用这两个因素来解释所观察到的相关性结构，现在我们使用下面的公式来建模： <span class="math display">\[
Y_{ij} = \alpha_{i,1} W_{1,j} + \alpha_{i,2} W_{2,j} + \varepsilon_{ij}
\]</span></p>
<p>参数解释： <span class="math inline">\(\alpha_{i,1}\)</span> 表示学生 <span class="math inline">\(i\)</span> 的整体学术能力，<span class="math inline">\(\alpha_{i,2}\)</span> 表示学生 <span class="math inline">\(i\)</span> 的这种学术能力在STEM与人文学科之间的差异。现在我们的问题就是，如何估计 <span class="math inline">\(W\)</span> 和 <span class="math inline">\(\alpha\)</span> ？</p>
<h3 id="因子分析与pca">因子分析与PCA</h3>
<p>结果表明，在某些假设下，前两个主成分是对于 <span class="math inline">\(W_{1}\)</span> 和 <span class="math inline">\(W_{2}\)</span> 的最优估计，因此我们可以按以下方式对它们进行估计：</p>
<p>It turns out that under certain assumptions, the first two principal components are optimal estimates for <span class="math inline">\(W_1\)</span> and <span class="math inline">\(W_2\)</span>. So we can estimate them like this:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">s &lt;- svd(Y)</div><div class="line">What &lt;- t(s$v[,1:2])</div><div class="line">colnames(What)&lt;-colnames(Y)</div><div class="line">round(What,2)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; round(What,<span class="number">2</span>)</div><div class="line">      Math Science    CS  Eng Hist</div><div class="line">[<span class="number">1</span>,]  <span class="number">0.36</span>    <span class="number">0.36</span>  <span class="number">0.36</span> <span class="number">0.47</span> <span class="number">0.43</span></div><div class="line">[<span class="number">2</span>,] -<span class="number">0.44</span>   -<span class="number">0.49</span> -<span class="number">0.42</span> <span class="number">0.34</span> <span class="number">0.34</span></div><div class="line">     Classics</div><div class="line">[<span class="number">1</span>,]     <span class="number">0.45</span></div><div class="line">[<span class="number">2</span>,]     <span class="number">0.39</span></div></pre></td></tr></table></figure>
<p>正如预期，第一个因子接近于一个常数，它有助于解释所有学科之间观察到的相关性，而第二个因子能解释STEM和人文学科之间的不表现。现在我们就以下模型中使用这些估计值： <span class="math display">\[
Y_{ij} = \alpha_{i,1} \hat{W}_{1,j} + \alpha_{i,2} \hat{W}_{2,j} + \varepsilon_{ij}
\]</span></p>
<p>我们现在就可以拟合模型，并注意一下，它能解释多大比例的变异：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">fit = s$u[,1:2]%*% (s$d[1:2]*What)</div><div class="line">var(as.vector(fit))/var(as.vector(Y))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; fit = s$u[,<span class="number">1</span>:<span class="number">2</span>]%*% (s$d[<span class="number">1</span>:<span class="number">2</span>]*What)</div><div class="line">&gt; var(as.vector(fit))/var(as.vector(Y))</div><div class="line">[<span class="number">1</span>] <span class="number">0.7880933</span></div></pre></td></tr></table></figure>
<p>我们在这里学习到的是，当我们有一些相关的单位或数据时，并不适合使用标准的线性模型。我们需要以某种试来解释观察到的结果。而因子分子则是实现这一目标的强大方法。</p>
<h3 id="因子分析的常规用法">因子分析的常规用法</h3>
<p>在高通量数据中，我们经常会看到相关结构。例如，我们在下图中展示了样本之间复杂的相关性。下面的这张图是按日期排序后，基因表达实验的数据展示结果：</p>
<figure class="highlight plain"><figcaption><span>gene_expression_correlations, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">library(Biobase)</div><div class="line">library(GSE5859)</div><div class="line">data(GSE5859)</div><div class="line">n &lt;- nrow(pData(e))</div><div class="line">o &lt;- order(pData(e)$date)</div><div class="line">Y=exprs(e)[,o]</div><div class="line">cors=cor(Y-rowMeans(Y))</div><div class="line">cols=colorRampPalette(rev(brewer.pal(11,&quot;RdBu&quot;)))(100)</div><div class="line">mypar()</div><div class="line">image(1:n,1:n,cors,col=cols,xlab=&quot;samples&quot;,ylab=&quot;samples&quot;,zlim=c(-1,1))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917141729.jpeg">

</div>
<p>使用两个因子并不足以对观察到的相关结构进行建模。但是，我们可以使用一个更加普遍的因子模型：</p>
<p><span class="math display">\[
Y_{ij} = \sum_{k=1}^K \alpha_{i,k} W_{j,k} + \varepsilon_{ij}
\]</span></p>
<p>我们可以使用PCA来估计<span class="math inline">\(\mathbf{W}_1,\dots,\mathbf{W}_K\)</span>。但是，在选择 <span class="math inline">\(k\)</span> 方面并不容易，这就是我们要研究问题的核心。在下面的章节里，我们会介绍一下，如何使用探索性数据分析来辅助我们找到这个 <span class="math inline">\(k\)</span> 。</p>
<h2 id="练习">练习</h2>
<p>P448</p>
<h2 id="使用因子分析来对批次效应建模">使用因子分析来对批次效应建模</h2>
<p>继续使用前面的数据集：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(GSE5859Subset)</div><div class="line">data(GSE5859Subset)</div></pre></td></tr></table></figure>
<p>我们之前展示过一些图形，那些图形中包括了性别效应与月份效应的基因子集，但是，现在我们使用一张图片来展示样本与样本之间的相关性（所有基因），并且展示出数据的复杂结构，如下所示：</p>
<figure class="highlight plain"><figcaption><span>correlation_image, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">library(rafalib)</div><div class="line">library(RColorBrewer)</div><div class="line">library(genefilter)</div><div class="line">sex &lt;- sampleInfo$group</div><div class="line">batch &lt;- factor(format(sampleInfo$date,&quot;%m&quot;))</div><div class="line">chr &lt;- geneAnnotation$CHR</div><div class="line">tt&lt;-rowttests(geneExpression,batch)</div><div class="line">ind1 &lt;- which(chr==&quot;chrY&quot;) #real differences</div><div class="line">ind2 &lt;- setdiff(c(order(tt$dm)[1:25],order(-tt$dm)[1:25]),ind1)</div><div class="line">set.seed(1)</div><div class="line">ind0 &lt;- setdiff(sample(seq(along=tt$dm),50),c(ind2,ind1))</div><div class="line">geneindex&lt;-c(ind2,ind0,ind1)</div><div class="line">mat&lt;-geneExpression[geneindex,]</div><div class="line">mat &lt;- mat -rowMeans(mat)</div><div class="line">icolors &lt;- colorRampPalette(rev(brewer.pal(11,&quot;RdYlBu&quot;)))(100)</div><div class="line">mypar(1,2)</div><div class="line">image(t(mat),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=icolors)</div><div class="line">y &lt;- geneExpression - rowMeans(geneExpression)</div><div class="line">image(1:ncol(y),1:ncol(y),cor(y),col=icolors,zlim=c(-1,1),</div><div class="line">       xaxt=&quot;n&quot;,xlab=&quot;&quot;,yaxt=&quot;n&quot;,ylab=&quot;&quot;)</div><div class="line">axis(2,1:ncol(y),sex,las=2)</div><div class="line">axis(1,1:ncol(y),sex,las=2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917142426.jpeg">

</div>
<p>我们已经看到了即假设月份能够解释批次效应效应，并且使用线性模型对批次效应进行校正后的方法，这种方法表现很好。但是，这种方法仍然有改进的空间。这有可能是由于月份仅仅是背后一些隐藏因子的外面表现，或者是说是一些实际上导致结构或样本相关性的因子。</p>
<h3 id="什么是一个批次">什么是一个批次</h3>
<p>现在我画出每个样本的日期图，其中颜色表示月份，如下所示：</p>
<figure class="highlight plain"><figcaption><span>what_is_batch, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">times &lt;-sampleInfo$date </div><div class="line">mypar(1,1)</div><div class="line">o=order(times)</div><div class="line">plot(times[o],pch=21,bg=as.numeric(batch)[o],ylab=&quot;Date&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917142913.jpeg">

</div>
<p>我们注意到，每个月不止有一天。天这个时间单位也会产生影响吗？我们可以使用PCA和EDA来尝试回答一下这个问题。以下是按日期排序后的第一个主成分图：</p>
<figure class="highlight plain"><figcaption><span>PC1_versus_time, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">s &lt;- svd(y)</div><div class="line">mypar(1,1)</div><div class="line">o&lt;-order(times)</div><div class="line">cols &lt;- as.numeric( batch)</div><div class="line">plot(s$v[o,1],pch=21,cex=1.25,bg=cols[o],ylab=&quot;First PC&quot;,xlab=&quot;Date order&quot;)</div><div class="line">legend(&quot;topleft&quot;,c(&quot;Month 1&quot;,&quot;Month 2&quot;),col=1:2,pch=16,box.lwd=0)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917143100.jpeg">

</div>
<p>“天”似乎与第1主成分高度相关，它可以在很大程度上解释变异：</p>
<figure class="highlight plain"><figcaption><span>variance_explained, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">mypar(1,1)</div><div class="line">plot(s$d^2/sum(s$d^2),ylab=&quot;% variance explained&quot;,xlab=&quot;Principal component&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917143153.jpeg">

</div>
<p>进一步的研究显示，前6个左右的主成分(PC)似乎都是由日期驱动的：</p>
<figure class="highlight plain"><figcaption><span>PCs_stratified_by_time, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">mypar(3,4)</div><div class="line">for(i in 1:12)&#123;</div><div class="line">  days &lt;- gsub(&quot;2005-&quot;,&quot;&quot;,times)  </div><div class="line">  boxplot(split(s$v[,i],gsub(&quot;2005-&quot;,&quot;&quot;,days)))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917143257.jpeg">

</div>
<p>如果我们从数据中将前6个主成分移除后，我们使用t检验来看一下结果是什么样子：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">D &lt;- s$d; D[1:4]&lt;-0 #take out first 2</div><div class="line">cleandat &lt;- sweep(s$u,2,D,&quot;*&quot;)%*%t(s$v)</div><div class="line">res &lt;-rowttests(cleandat,factor(sex))</div></pre></td></tr></table></figure>
<p>事实上，这确实消除了批次效应，但是这似乎也消除了我们感兴趣的大部分生物学效应。事实上，没有基因的Q值再小于0.1了，如下所示：</p>
<figure class="highlight plain"><figcaption><span>pval_hist_and_volcano_after_removing_PCs, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">library(qvalue)</div><div class="line">mypar(1,2)</div><div class="line">hist(res$p.value[which(!chr%in%c(&quot;chrX&quot;,&quot;chrY&quot;) )],main=&quot;&quot;,ylim=c(0,1300))</div><div class="line">plot(res$dm,-log10(res$p.value))</div><div class="line">points(res$dm[which(chr==&quot;chrX&quot;)],-log10(res$p.value[which(chr==&quot;chrX&quot;)]),col=1,pch=16)</div><div class="line">points(res$dm[which(chr==&quot;chrY&quot;)],-log10(res$p.value[which(chr==&quot;chrY&quot;)]),col=2,pch=16,</div><div class="line">       xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;chrX&quot;,&quot;chrY&quot;),col=1:2,pch=16)</div><div class="line">qvals &lt;- qvalue(res$p.value)$qvalue</div><div class="line">index &lt;- which(qvals&lt;0.1)</div><div class="line">cat(&quot;Total genes with q-value &lt; 0.1: &quot;,length(index),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrY: &quot;, sum(chr[index]==&quot;chrY&quot;,na.rm=TRUE),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrX: &quot;, sum(chr[index]==&quot;chrX&quot;,na.rm=TRUE),sep=&quot;&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917143622.jpeg">

</div>
<p>这种情况似乎就是过度校正了，因为我们现在恢复了更少的Y染色体基因，并且p值的直方图中小p值的数目更小了，使得分布变得不再均匀。因此性别因素可能与前几个主成分(PC)有关，这就出现了”把婴儿和洗澡水一起倒掉“的情况。</p>
<h3 id="sva">SVA</h3>
<p>前面我们刚提到了我们所遇到的问题，即我们遇到了过度校正，以及消除了与感兴趣结果相关的变异，解决这个问题的方法就是同时使用感兴趣变量的协变量以及那些被认为是批次的模型共同进行拟合。代理变量分析(SVA, Surrogate Variable Analysis)就是这样的一种方法。</p>
<p>SVA的基本思想就是先估计因子，但要注意不要包括感兴趣的结果。为此，SVA使用了一种交互式方法，其中的每一行都赋予一个权重，这个权重能够量化基因与替代变量（而不是感兴趣结果）唯一相关的概率。然后在SVD中使用这些权重，将更高的权重赋予给感兴趣的结果无关，且与批次有关的行。下面是这两次迭代的计算过程。三个图片显示的是，数据乘以权重（对于基因的子集）、权重，以及估计的第一因子，代码如下所示：</p>
<figure class="highlight plain"><figcaption><span>illustration_of_sva,fig.height</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div></pre></td><td class="code"><pre><div class="line">library(sva)</div><div class="line">library(limma)</div><div class="line">mod &lt;- model.matrix(~sex)</div><div class="line">cind &lt;- order( as.Date(sampleInfo$date) )</div><div class="line">dates &lt;- gsub(&quot;2005-&quot;,&quot;&quot;,sampleInfo$date)</div><div class="line">weights=rep(1,nrow(y))</div><div class="line">par(mar = c(4.1, 2.1, 3.5, 2.1), </div><div class="line">    mgp = c(1.5, 0.5, 0))</div><div class="line">layout(matrix(c(1:6),nrow=2,byrow=TRUE),widths=c(5,1.5,5))</div><div class="line">for(b in 1:2)&#123;</div><div class="line">  image(1:ncol(mat),1:nrow(mat),t(mat[,cind]*weights[geneindex]),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=icolors,xlab=&quot;&quot;,ylab=&quot;&quot;)</div><div class="line">  axis(side=1,seq(along=dates),dates[cind],las=2)</div><div class="line">  abline(v=12.5)</div><div class="line">  </div><div class="line">  svafit &lt;- sva(y,mod,B=b,n.sv=5)</div><div class="line">  weights = svafit$pprob.gam*(1-svafit$pprob.b)</div><div class="line">  </div><div class="line">  surrogate &lt;- svd( y*weights)$v[,1]#Weighted SVD</div><div class="line">  </div><div class="line">  image(matrix(weights[geneindex],nrow=1),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=brewer.pal(9,&quot;Blues&quot;))</div><div class="line">  plot(surrogate[cind],bg=sex[cind]+1,pch=21,xlab=&quot;&quot;,xaxt=&quot;n&quot;,ylab=&quot;Surrogate variable&quot;,ylim=c(-.5,.5),cex=1.5)</div><div class="line">  axis(side=1,seq(along=dates),dates[cind],las=2)</div><div class="line">  abline(v=12.5)</div><div class="line">  text(1,0.5,&quot;June&quot;)</div><div class="line">  text(13.5,0.5,&quot;Oct&quot;)</div><div class="line">  legend(&quot;bottomright&quot;,c(&quot;0&quot;,&quot;1&quot;),col=c(1,2),pch=16)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917153302.jpeg">

</div>
<p>这个算法会多次迭代这个过程（由<code>B</code>参数控制），并返回替代变量的估计值，这就类似于因子分子中的隐藏因子。在R中，我们使用<code>sva()</code>函数来实现这个过程。在这个案例里，SVA会为我们选择代理值或因子的数量，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">library(limma)</div><div class="line">svafit &lt;- sva(geneExpression,mod)</div><div class="line">svaX&lt;-model.matrix(~sex+svafit$sv)</div><div class="line">lmfit &lt;- lmFit(geneExpression,svaX)</div><div class="line">tt&lt;- lmfit$coef[,2]*sqrt(lmfit$df.residual)/(2*lmfit$sigma)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; svafit &lt;- sva(geneExpression,mod)</div><div class="line">Number of significant surrogate variables is:  <span class="number">5</span> </div><div class="line">Iteration (out of <span class="number">5</span> ):<span class="number">1</span>  <span class="number">2</span>  <span class="number">3</span>  <span class="number">4</span>  <span class="number">5</span></div></pre></td></tr></table></figure>
<p>与之前的方法相比，这种方法有了提升：</p>
<figure class="highlight plain"><figcaption><span>pval_hist_and_volcano_sva, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">res &lt;- data.frame(dm= -lmfit$coef[,2],</div><div class="line">                  p.value=2*(1-pt(abs(tt),lmfit$df.residual[1]) ) )</div><div class="line">mypar(1,2)</div><div class="line">hist(res$p.value[which(!chr%in%c(&quot;chrX&quot;,&quot;chrY&quot;) )],main=&quot;&quot;,ylim=c(0,1300))</div><div class="line">plot(res$dm,-log10(res$p.value))</div><div class="line">points(res$dm[which(chr==&quot;chrX&quot;)],-log10(res$p.value[which(chr==&quot;chrX&quot;)]),col=1,pch=16)</div><div class="line">points(res$dm[which(chr==&quot;chrY&quot;)],-log10(res$p.value[which(chr==&quot;chrY&quot;)]),col=2,pch=16,</div><div class="line">       xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;chrX&quot;,&quot;chrY&quot;),col=1:2,pch=16)</div><div class="line">qvals &lt;- qvalue(res$p.value)$qvalue</div><div class="line">index &lt;- which(qvals&lt;0.1)</div><div class="line">cat(&quot;Total genes with q-value &lt; 0.1: &quot;,length(index),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrY: &quot;, sum(chr[index]==&quot;chrY&quot;,na.rm=TRUE),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrX: &quot;, sum(chr[index]==&quot;chrX&quot;,na.rm=TRUE),sep=&quot;&quot;)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; cat(<span class="string">"Total genes with q-value &lt; 0.1: "</span>,length(index),<span class="string">"\n"</span>,</div><div class="line">+     <span class="string">"Number of selected genes on chrY: "</span>, sum(chr[index]==<span class="string">"chrY"</span>,na.rm=<span class="literal">TRUE</span>),<span class="string">"\n"</span>,</div><div class="line">+     <span class="string">"Number of selected genes on chrX: "</span>, sum(chr[index]==<span class="string">"chrX"</span>,na.rm=<span class="literal">TRUE</span>),sep=<span class="string">""</span>)</div><div class="line">Total genes with q-value &lt; <span class="number">0.1</span>: <span class="number">14</span></div><div class="line">Number of selected genes on chrY: <span class="number">5</span></div><div class="line">Number of selected genes on chrX: <span class="number">8</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917154043.jpeg">

</div>
<p>现在我们通过图形来展示SVA实现了什么，下面是原始数据集分解为性别效应，代理变量和独立噪声的</p>
<p>为了可视化SVA实现了什么，下面是原始数据集通过算法分解为性别效果、代理变量和算法估计的独立噪声的可视化结果：</p>
<figure class="highlight plain"><figcaption><span>different_sources_of_var, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">Batch&lt;- lmfit$coef[geneindex,3:7]%*%t(svaX[,3:7])</div><div class="line">Signal&lt;-lmfit$coef[geneindex,1:2]%*%t(svaX[,1:2])</div><div class="line">error &lt;- geneExpression[geneindex,]-Signal-Batch</div><div class="line">##demean for plot</div><div class="line">Signal &lt;-Signal-rowMeans(Signal)</div><div class="line">mat &lt;- geneExpression[geneindex,]-rowMeans(geneExpression[geneindex,])</div><div class="line">mypar(1,4,mar = c(2.75, 4.5, 2.6, 1.1))</div><div class="line">image(t(mat),col=icolors,zlim=c(-5,5),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;)</div><div class="line">image(t(Signal),col=icolors,zlim=c(-5,5),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;)</div><div class="line">image(t(Batch),col=icolors,zlim=c(-5,5),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;)</div><div class="line">image(t(error),col=icolors,zlim=c(-5,5),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190917154225.jpeg">

</div>
<h2 id="练习-1">练习</h2>
<p>P459</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/27/DAL/DALS027_Batch_Effect03_AdjustBatchEffect/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/27/DAL/DALS027_Batch_Effect03_AdjustBatchEffect/" itemprop="url">DALS027-批次效应03-校正批次效应</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-27T12:00:00+08:00">
                2019-08-27
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Data-Analysis-for-the-life-sciences/" itemprop="url" rel="index">
                    <span itemprop="name">Data Analysis for the life sciences</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,004
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  13
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第10章批次效应的第3小节，这一部分的主要内容涉及批次效应(Batch Effects)的校正，有关这一部分Rmarkdown文档参见作者的<a href="https://github.com/genomicsclass/labs/blob/master/batch/adjusting_with_linear_models.Rmd" target="_blank" rel="external">Github</a>。</p>
<h2 id="使用情境">使用情境</h2>
<p>在这一部分里，我们要了解如何对批次效应进行校正。</p>
<h3 id="数据案例">数据案例</h3>
<p>为了说明我们可以使用统计方法来校正批次效应，我们将创建一个数据示例，其中感兴趣的结果与批次之间有某种程度的混杂，但不完全混杂。为了辅助说明和评估我们所示的方法，我们还将选择一个结果用于说明我们期望对哪些基因应该是差异表达的。也就是说，我们想让性别作为我们感兴趣的变量，并且期望 Y 染色体上的基因表达有所差异。我们或许能看到，来源于 X 染色体上的基因是差异表达的，因为有些基因在 X 染色体上并没有失活，而是正常表达，这些数据都包含在以下的数据集中，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">##available from course github repository</div><div class="line">library(GSE5859Subset)</div><div class="line">data(GSE5859Subset)</div></pre></td></tr></table></figure>
<p>我们可以看到，性别与月份有相关性：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">month &lt;- format(sampleInfo$date,&quot;%m&quot;)</div><div class="line">table(sampleInfo$group, month)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; month &lt;- format(sampleInfo$date,<span class="string">"%m"</span>)</div><div class="line">&gt; table(sampleInfo$group, month)</div><div class="line">   month</div><div class="line">    <span class="number">06</span> <span class="number">10</span></div><div class="line">  <span class="number">0</span>  <span class="number">9</span>  <span class="number">3</span></div><div class="line">  <span class="number">1</span>  <span class="number">3</span>  <span class="number">9</span></div></pre></td></tr></table></figure>
<p>为了说明混杂现在，我们会挑选一些基因绘制热图。现在我们挑选这些基因：(1)所有Y染色体上的基因；(2)某些与批次效应有效的基因；(3)一些随机选择的基因。下面是热图，其中红色表示高表达基因，蓝色表示低表达基因，黄色表示中间表达基因。每列是一个样本，每行就是随机选择的基因，代码如下所示：</p>
<figure class="highlight plain"><figcaption><span>image_of_subset, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(rafalib)</div><div class="line">library(RColorBrewer)</div><div class="line">library(genefilter)</div><div class="line">batch &lt;- factor(format(sampleInfo$date,&quot;%m&quot;))</div><div class="line">chr &lt;- geneAnnotation$CHR</div><div class="line">tt&lt;-rowttests(geneExpression,batch)</div><div class="line">ind1 &lt;- which(chr==&quot;chrY&quot;) ##real differences</div><div class="line">ind2 &lt;- setdiff(c(order(tt$dm)[1:25],order(-tt$dm)[1:25]),ind1)</div><div class="line">set.seed(1)</div><div class="line">ind0 &lt;- setdiff(sample(seq(along=tt$dm),50),c(ind2,ind1))</div><div class="line">geneindex&lt;-c(ind2,ind0,ind1)</div><div class="line">mat&lt;-geneExpression[geneindex,]</div><div class="line">mat &lt;- mat -rowMeans(mat)</div><div class="line">icolors &lt;- colorRampPalette(rev(brewer.pal(11,&quot;RdYlBu&quot;)))(100)</div><div class="line">mypar(1,1)</div><div class="line">image(1:24,1:nrow(mat), t(mat),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,col=icolors,xlab=&quot;&quot;,ylab=&quot;&quot;)</div><div class="line">axis(3,1:24,rep(c(&quot;F&quot;,&quot;M&quot;),each=12),cex.axis=0.5)</div><div class="line">axis(1,1:24,month,cex.axis=0.5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916180846.jpeg">

</div>
<p>在上图中，前12列是女性（用1表示），后12列是男性（用0表示）。我们可以看到，一些Y染色体上的基因趋向顶部，其中在女性数据中这些是蓝色，在男性数据中这是红色。我们也可以看到一些基因与月份相关，这些基因集中趋向于底部。一些基因在六月份低表达（June，6），一些基因在10月份高表达（October，10），还有一些基因正好相反。月份效应并没有性别效应明显，但是这种效应确实存在。</p>
<p>在接下来的内容里，我们会模拟实际分析中的一些典型分析方法。我们会假设一种情况，在这种情况里，我们假设不知道哪些基因在男女性别有差异，然后我们要找出这些差异基因，通过比较我们所期望的相关性，来评估一下这些方法的优劣。这里需要注意一下，我们虽然只绘制出其中一部分基因，但是我们所分析的基因数目却是8793个。</p>
<h3 id="评估图与总结">评估图与总结</h3>
<p>为了说明我们所用方法的优劣，我们将假设常染色体上的基因（这些基因不在染色体X和染色体Y上）有可能是假阳性，位于Y染色体上的基因是真阳性。染色体X上的基因可以是假阳性，可以是真阳性。上述的这些设计就在让我们评估特异性(specificity)与灵敏性(sensitivity)。由于在实际分析中，我们很少能知道“真相”，因此我们在这里做的评估是很在实际分析中实现的。因此模拟常常用于评估目的：我们知道真相，因为分析所用的数据是我们自己构建的。但是，模拟也会有无法捕获真实实验数据细微差异的风险。相比之下，这个数据集就是一个实验性质的数据集。</p>
<p>在接下来的部分中，我们将使用p值的直方图来评估一下批次效应校正程序的特异性(specificity)，特异性表示低假阳性率。因为常染色体应该不会出现由于差异导致的差异表达，我们应该会看到一个平坦的p值直方图。为了评估灵敏性(sensitivity,低假阳性)，我们会报告那些位于X染色体和Y染色体上我们拒绝零假设时，所计算出的基因的数目。我们还会生成一个火山图，用一条水平虚线将那些有显著性差异的基因与无显著性差异的基因区分开来，用于突出显示X染色体和Y染色体的基因。以下是一个应用原始t检验得到的结果，以及q值小于0.1的基因，如下所示：</p>
<figure class="highlight plain"><figcaption><span>pvalue_hist_and_volcano_plots, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">library(qvalue)</div><div class="line">res &lt;- rowttests(geneExpression,as.factor( sampleInfo$group ))</div><div class="line">mypar(1,2)</div><div class="line">hist(res$p.value[which(!chr%in%c(&quot;chrX&quot;,&quot;chrY&quot;) )],main=&quot;&quot;,ylim=c(0,1300))</div><div class="line">plot(res$dm,-log10(res$p.value))</div><div class="line">points(res$dm[which(chr==&quot;chrX&quot;)],-log10(res$p.value[which(chr==&quot;chrX&quot;)]),col=1,pch=16)</div><div class="line">points(res$dm[which(chr==&quot;chrY&quot;)],-log10(res$p.value[which(chr==&quot;chrY&quot;)]),col=2,pch=16,</div><div class="line">       xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;chrX&quot;,&quot;chrY&quot;),col=1:2,pch=16)</div><div class="line">qvals &lt;- qvalue(res$p.value)$qvalue</div><div class="line">index &lt;- which(qvals&lt;0.1)</div><div class="line">abline(h=-log10(max(res$p.value[index])))</div><div class="line">cat(&quot;Total genes with q-value &lt; 0.1: &quot;,length(index),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrY: &quot;, sum(chr[index]==&quot;chrY&quot;,na.rm=TRUE),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrX: &quot;, sum(chr[index]==&quot;chrX&quot;,na.rm=TRUE),sep=&quot;&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916180856.jpeg">

</div>
<p>很明显，我们注意到直方图并不平坦的。相反，低p值的条状过高（原文：low p-values are over-represented）。此外，最终列表上超过一半的基因是常染色体。我们现在介绍两个统计学方法来尝试一下解决这个问题。</p>
<h3 id="使用线性模型来校正批次效应">使用线性模型来校正批次效应</h3>
<p>我们已经观察到了处理时间(processing date)对基因表达有着某种效应。我们因此会通过将此效应添加到一个模型中来对其进行校正(adjust)。当我们对两组数据进行t检验时，它就相当于对基因 <span class="math inline">\(j\)</span> 进行以下线性模型拟合：</p>
<p><span class="math display">\[
Y_{ij} = \alpha_j + x_i \beta_{j} + \varepsilon_{ij}
\]</span></p>
<p>其中，研究对象 <span class="math inline">\(i\)</span> 如果是女性，那么 <span class="math inline">\(x_{i}=1\)</span>, 如果是男性则为0，<span class="math inline">\(\beta_{j}\)</span> 表示基因 <span class="math inline">\(j\)</span> 的差异，<span class="math inline">\(\varepsilon_{ij}\)</span> 表示组内变异。现在我们的问题是什么？</p>
<p>在线性模型那一章节中，我们假设错误项(error term)是独立的。我们知道，对于所有的基因来说，这是不可能的，因为我们知道10月份数据中的错误项比6月份中的错误项更相似。我们可以在线性模型中添加一项，用于校正这种错误，如下所示：</p>
<p><span class="math display">\[
Y_{ij} = \alpha_j + x_i \beta_{j} + z_i \gamma_j+\varepsilon_{ij}.
\]</span></p>
<p>在这个公式中，如果样本 <span class="math inline">\(i\)</span> 是在10月份处理的，那么 <span class="math inline">\(z_{i}=1\)</span> ，否则为0，另外， <span class="math inline">\(\gamma_{j}\)</span> 是基因<span class="math inline">\(j\)</span>的月份效应。这就是 一个线性模型如何比t检验具有更大灵活性的一个例子。</p>
<p>我们构建一个含有批次的模型矩阵，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">sex &lt;- sampleInfo$group</div><div class="line">X &lt;- model.matrix(~sex+batch)</div></pre></td></tr></table></figure>
<p>现在我们对每个基因拟合一个模型。例如，我们要注意在原始模型和那些含有批次校正模型之间的差异，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">j &lt;- 7635</div><div class="line">y &lt;- geneExpression[j,]</div><div class="line">X0 &lt;- model.matrix(~sex) </div><div class="line">fit &lt;- lm(y~X0-1)</div><div class="line">summary(fit)$coef</div><div class="line">X &lt;- model.matrix(~sex+batch)</div><div class="line">fit &lt;- lm(y~X)</div><div class="line">summary(fit)$coef</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">&gt; summary(fit)$coef</div><div class="line">                Estimate Std. Error   t value     Pr(&gt;|t|)</div><div class="line">X0(Intercept)  <span class="number">6.9555747</span>  <span class="number">0.2166035</span> <span class="number">32.112008</span> <span class="number">5.611901e-20</span></div><div class="line">X0sex         -<span class="number">0.6556865</span>  <span class="number">0.3063237</span> -<span class="number">2.140502</span> <span class="number">4.365102e-02</span></div><div class="line">&gt; X &lt;- model.matrix(~sex+batch)</div><div class="line">&gt; fit &lt;- lm(y~X)</div><div class="line">&gt; summary(fit)$coef</div><div class="line">               Estimate Std. Error    t value     Pr(&gt;|t|)</div><div class="line">(Intercept)  <span class="number">7.26329968</span>  <span class="number">0.1605560</span> <span class="number">45.2384140</span> <span class="number">2.036006e-22</span></div><div class="line">Xsex        -<span class="number">0.04023663</span>  <span class="number">0.2427379</span> -<span class="number">0.1657616</span> <span class="number">8.699300e-01</span></div><div class="line">Xbatch10    -<span class="number">1.23089977</span>  <span class="number">0.2427379</span> -<span class="number">5.0709009</span> <span class="number">5.070727e-05</span></div></pre></td></tr></table></figure>
<p>我们然后将这个新模型用于每个基因，例如我们可以使用<code>sapply()</code>函数来恢复估计的系数和p值，如下所示：</p>
<figure class="highlight plain"><figcaption><span>pvalue_hist_and_volcano_plots2, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">res &lt;- t( sapply(1:nrow(geneExpression),function(j)&#123;</div><div class="line">  y &lt;- geneExpression[j,]</div><div class="line">  fit &lt;- lm(y~X-1)</div><div class="line">  summary(fit)$coef[2,c(1,4)]</div><div class="line">&#125; ) )</div><div class="line">##turn into data.frame so we can use the same code for plots as above</div><div class="line">res &lt;- data.frame(res)</div><div class="line">names(res) &lt;- c(&quot;dm&quot;,&quot;p.value&quot;)</div><div class="line">mypar(1,2)</div><div class="line">hist(res$p.value[which(!chr%in%c(&quot;chrX&quot;,&quot;chrY&quot;) )],main=&quot;&quot;,ylim=c(0,1300))</div><div class="line">plot(res$dm,-log10(res$p.value))</div><div class="line">points(res$dm[which(chr==&quot;chrX&quot;)],-log10(res$p.value[which(chr==&quot;chrX&quot;)]),col=1,pch=16)</div><div class="line">points(res$dm[which(chr==&quot;chrY&quot;)],-log10(res$p.value[which(chr==&quot;chrY&quot;)]),col=2,pch=16,</div><div class="line">       xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;chrX&quot;,&quot;chrY&quot;),col=1:2,pch=16)</div><div class="line">qvals &lt;- qvalue(res$p.value)$qvalue</div><div class="line">index &lt;- which(qvals&lt;0.1)</div><div class="line">abline(h=-log10(max(res$p.value[index])))</div><div class="line">cat(&quot;Total genes with q-value &lt; 0.1: &quot;,length(index),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrY: &quot;, sum(chr[index]==&quot;chrY&quot;,na.rm=TRUE),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrX: &quot;, sum(chr[index]==&quot;chrX&quot;,na.rm=TRUE),sep=&quot;&quot;)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916181644.jpeg">

</div>
<p>从图上我们可以看到，特异性（假阳性较少）有了很大的提高，同时灵敏性也没有过多损失（我们仍然能发现许多Y染色体上的基因）。但是，我们仍然能够从直方图中看到一些偏差，在后面的部分里，我们可以看到月份并不能完美地解释批次效应，应该有更好的估计方法。</p>
<h3 id="注意计算效率">注意计算效率</h3>
<p>在上面代码中，当我们重复计算 <span class="math inline">\((X^\top X)^{-1}\)</span>，以及将其应用于每个基因时，设计矩阵并没发生改变。相反，我们可以在一次矩阵代数计算中实现该计算，然后通过将 <span class="math inline">\((X^\top X)^{-1}X^\top Y\)</span> 与表示所有基因的 <span class="math inline">\(Y\)</span> 列相乘来获得所有的 <span class="math inline">\(\beta\)</span> 。<code>limma</code>包的设计就是基于这个思路（利用了QR分解）。我们来看一下这样的计算有多快：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(limma)</div><div class="line">X &lt;- model.matrix(~sex+batch)</div><div class="line">fit &lt;- lmFit(geneExpression,X)</div></pre></td></tr></table></figure>
<p>每个基因的估计回归系数如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">dim( fit$coef)</div><div class="line"># &gt; dim( fit$coef)</div><div class="line"># [1] 8793    3</div></pre></td></tr></table></figure>
<p>我们对每个基因都有一个估计。如果要获得其中一个p值，我们就必须要构建以下的比值(ratio)：</p>
<p>We have one estimate for each gene. To obtain p-values for one of these, we have to construct the ratios:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">k &lt;- 2 ##second coef</div><div class="line">ses &lt;- fit$stdev.unscaled[,k]*fit$sigma</div><div class="line">ttest &lt;- fit$coef[,k]/ses</div><div class="line">pvals &lt;- 2*pt(-abs(ttest),fit$df)</div></pre></td></tr></table></figure>
<h2 id="combat">Combat</h2>
<p><a href="http://biostatistics.oxfordjournals.org/content/8/1/118.short" target="_blank" rel="external">Combat</a> 是一种基于线性模型的算法，常用于校正批次效应。这种方法能够对估计值拟合一个分层模型，并且移除行的特定批次效应。Combat使用了一个模块块的方法(a modular approach)。第一步就是删除那些被认为是批次效应的内容，如下所示：</p>
<figure class="highlight plain"><figcaption><span>message</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(sva) #available from Bioconductor</div><div class="line">mod &lt;- model.matrix(~sex)</div><div class="line">cleandat &lt;- ComBat(geneExpression,batch,mod)</div></pre></td></tr></table></figure>
<p>然后，将剩下的结果用于拟合我们感兴趣的变量：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">res&lt;-genefilter::rowttests(cleandat,factor(sex))</div></pre></td></tr></table></figure>
<p>在这种情况下，与我们通过拟合简单线性模型得到的结果相比，通过上面计算得到的结果缺乏特异性（原文：the results are less specific than what we obtain by fitting the simple linear model）：</p>
<figure class="highlight plain"><figcaption><span>pvalue_hist_and_volcano_plots3, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">mypar(1,2)</div><div class="line">hist(res$p.value[which(!chr%in%c(&quot;chrX&quot;,&quot;chrY&quot;) )],main=&quot;&quot;,ylim=c(0,1300))</div><div class="line">plot(res$dm,-log10(res$p.value))</div><div class="line">points(res$dm[which(chr==&quot;chrX&quot;)],-log10(res$p.value[which(chr==&quot;chrX&quot;)]),col=1,pch=16)</div><div class="line">points(res$dm[which(chr==&quot;chrY&quot;)],-log10(res$p.value[which(chr==&quot;chrY&quot;)]),col=2,pch=16,</div><div class="line">       xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;chrX&quot;,&quot;chrY&quot;),col=1:2,pch=16)</div><div class="line">qvals &lt;- qvalue(res$p.value)$qvalue</div><div class="line">index &lt;- which(qvals&lt;0.1)</div><div class="line">abline(h=-log10(max(res$p.value[index])))</div><div class="line">cat(&quot;Total genes with q-value &lt; 0.1: &quot;,length(index),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrY: &quot;, sum(chr[index]==&quot;chrY&quot;,na.rm=TRUE),&quot;\n&quot;,</div><div class="line">    &quot;Number of selected genes on chrX: &quot;, sum(chr[index]==&quot;chrX&quot;,na.rm=TRUE),sep=&quot;&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916181815.jpeg">

</div>
<h2 id="练习">练习</h2>
<p>P442</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/26/DAL/DALS026_Batch_Effect02_DiscoteryBatchEffect/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/26/DAL/DALS026_Batch_Effect02_DiscoteryBatchEffect/" itemprop="url">DALS026-批次效应02-发现批次效应</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-26T12:00:00+08:00">
                2019-08-26
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Data-Analysis-for-the-life-sciences/" itemprop="url" rel="index">
                    <span itemprop="name">Data Analysis for the life sciences</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,117
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第10章批次效应的第2小节，这一部分的主要内容涉及批次效应(Batch Effects)的发现与处理，有关这一部分Rmarkdown文档参见作者的<a href="https://github.com/genomicsclass/labs/tree/master/batch/eda_with_pca.Rmd" target="_blank" rel="external">Github</a>。</p>
<p>前面我们已经介绍过了PCA，我们在实际运用过程中就可以使用PCA来进行探索性数据分析了。为了说明这一点，我们将会使用一个实际中的数据集，出于我们教学的目的，这个数据还是原始数据，并没有经过数据清洗。我们首先从公共数据库中下载这个原始数据，接着你要做的就是预处理这些数据，并使用Bioconductor中的一个包进行后续的数据分析。</p>
<h2 id="基因表达数据">基因表达数据</h2>
<p>第一步，下载数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line"><span class="comment"># devtools::install_github("genomicsclass/GSE5859")</span></div><div class="line"><span class="keyword">library</span>(GSE5859) <span class="comment">##Available from GitHub</span></div><div class="line">data(GSE5859)</div></pre></td></tr></table></figure>
<p>注：这里书本中有所出入，</p>
<p>第二步，数据探索。这里先从探索样本相关性矩阵开始，我们注意到有一对样本的相关性是1.这就问题时着有一个样本被上传到了公共数据库2次，但是上传后的名称不同，下面的代码将会剔除这个样本，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">cors &lt;- cor(exprs(e))</div><div class="line">Pairs=which(abs(cors)&gt;<span class="number">0.9999</span>,arr.ind=<span class="literal">TRUE</span>)</div><div class="line">out = Pairs[which(Pairs[,<span class="number">1</span>]&lt;Pairs[,<span class="number">2</span>]),,drop=<span class="literal">FALSE</span>]</div><div class="line"><span class="keyword">if</span>(length(out[,<span class="number">2</span>])&gt;<span class="number">0</span>) e=e[,-out[<span class="number">2</span>]]</div></pre></td></tr></table></figure>
<p>还可以移除控制组的探针名，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">out &lt;- grep(<span class="string">"AFFX"</span>,featureNames(e))</div><div class="line">e &lt;- e[-out,]</div></pre></td></tr></table></figure>
<p>第三步，处理数据。现在我们创建一个去趋势(detrended)基因表达数据矩阵，并从样本注释表中提取日期与结果，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">y &lt;- exprs(e)-rowMeans(exprs(e))</div><div class="line">dates &lt;- pData(e)$date</div><div class="line">eth &lt;- pData(e)$ethnicity</div></pre></td></tr></table></figure>
<p>原始数据集中不包括样本信息中的性别。但是，我们为了出于教学的目的，我们使用下面的数据来研究一下这个性别问题，也就是说，在下面的代码中，我们会展示一下如何预测每个样本的性别。这种计算的基本思想就是观察Y染色体中基因的中位数基因表达水平。男性的这个数字应该更高。为此，我们需要上传一个注释包，用于提供这个实验中所用平台的一些特征信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">annotation(e)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; annotation(e)</div><div class="line">[<span class="number">1</span>] <span class="string">"hgfocus"</span></div></pre></td></tr></table></figure>
<p>此时，我们需要下载并安装<code>hgfocus.db</code>包，然后提取染色体位置的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">map2gene &lt;- mapIds(hgfocus.db, keys=featureNames(e),</div><div class="line">                   column=<span class="string">"ENTREZID"</span>, keytype=<span class="string">"PROBEID"</span>,</div><div class="line">                   multiVals=<span class="string">"first"</span>)</div><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">map2chr &lt;- mapIds(Homo.sapiens, keys=map2gene,</div><div class="line">                  column=<span class="string">"TXCHROM"</span>, keytype=<span class="string">"ENTREZID"</span>,</div><div class="line">                  multiVals=<span class="string">"first"</span>)</div><div class="line">chryexp &lt;- colMeans(y[which(unlist(map2chr)==<span class="string">"chrY"</span>),])</div></pre></td></tr></table></figure>
<p>如果我们创建Y染色体中位数基因表达水平的直方图，我们就能清楚地看到差异，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">mypar()</div><div class="line">hist(chryexp)</div></pre></td></tr></table></figure>
<p>现在我们预测一下性别，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sex &lt;- factor(ifelse(chryexp&lt;<span class="number">0</span>,<span class="string">"F"</span>,<span class="string">"M"</span>))</div></pre></td></tr></table></figure>
<h3 id="计算主成分">计算主成分</h3>
<p>现在我们计算一下主成分，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">s &lt;- svd(y)</div><div class="line">dim(s$v)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; s &lt;- svd(y)</div><div class="line">&gt; dim(s$v)</div><div class="line">[<span class="number">1</span>] <span class="number">207</span> <span class="number">207</span></div></pre></td></tr></table></figure>
<p>我们也可以使用<code>prcomp()</code>函数来创建一个主成分对象。</p>
<h3 id="变异解释">变异解释</h3>
<p>第一步就是解释由结构(structure)诱导的样本的相关性程度是什么样的，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RColorBrewer)</div><div class="line">cols=colorRampPalette(rev(brewer.pal(<span class="number">11</span>,<span class="string">"RdBu"</span>)))(<span class="number">100</span>)</div><div class="line">image ( cor(y) ,col=cols,zlim=c(-<span class="number">1</span>,<span class="number">1</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916131005.jpeg">

</div>
<p>上图显示了相关性。每个小单元络的i与j代表了样本i和样本j的相关性。红色表示高，白色表示0，蓝色表示低。</p>
<p>这时我们使用了术语结构(structure)，结构是指，如果样本实际上是相互独立时我们所能发现的偏差。从上面图片上我们可以明显看到，样本实际上有分组，也就是说，不同组内的样本相关性比组外的更强。</p>
<p>我们需要生成一个简单的探索性图来确定我们㙘多少主成分来描述这种结构，这就是方差解释图。 如果数据是独立的，那么方差的解释就如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">y0 &lt;- matrix( rnorm( nrow(y)*ncol(y) ) , nrow(y), ncol(y) )</div><div class="line">d0 &lt;- svd(y0)$d</div><div class="line">plot(d0^<span class="number">2</span>/sum(d0^<span class="number">2</span>),ylim=c(<span class="number">0</span>,<span class="number">.25</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916133405.jpeg">

</div>
<p>事实上，我们看到的方差解释图是下面的这个样子：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">plot(s$d^<span class="number">2</span>/sum(s$d^<span class="number">2</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916133542.jpeg">

</div>
<p>至少有20个PC高于我们对独立数据的预期。我们下一步就是尝试用检测的变量来解释这些PC。这些PC是由种族(ethnicity)，性别(sex)还是时间(date)或其它的因素驱动的吗？</p>
<h3 id="mds图">MDS图</h3>
<p>如前面一样，我们可以先用MDS图来探索一下数据，回答上述提到的问题。探索感兴趣的变量和PC之间的关系的一种方法就是使用颜色来表示这些变量，例如，以下是使用表示了颜色来标记种族(ethnicity)这个变量，通过前两个PC来展示的数据结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">cols = as.numeric(eth)</div><div class="line">mypar()</div><div class="line">plot(s$v[,<span class="number">1</span>],s$v[,<span class="number">2</span>],col=cols,pch=<span class="number">16</span>,</div><div class="line">     xlab=<span class="string">"PC1"</span>,ylab=<span class="string">"PC2"</span>)</div><div class="line">legend(<span class="string">"bottomleft"</span>,levels(eth),col=seq(along=levels(eth)),pch=<span class="number">16</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916134050.jpeg">

</div>
<p>从上图来看，第1 PC与种族(ethnicity)有着很强的联系。但是，我们也会看到一些橙色的点有的形成了亚簇(subslusters)。我们从以前的预分析中也知道，种族(ethnicity)和预处理时间(preprocessing date)有相关性，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">year = factor(format(dates,<span class="string">"%y"</span>))</div><div class="line">table(year,eth)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; table(year,eth)</div><div class="line">    eth</div><div class="line">year ASN CEU HAN</div><div class="line">  <span class="number">02</span>   <span class="number">0</span>  <span class="number">32</span>   <span class="number">0</span></div><div class="line">  <span class="number">03</span>   <span class="number">0</span>  <span class="number">54</span>   <span class="number">0</span></div><div class="line">  <span class="number">04</span>   <span class="number">0</span>  <span class="number">13</span>   <span class="number">0</span></div><div class="line">  <span class="number">05</span>  <span class="number">80</span>   <span class="number">3</span>   <span class="number">0</span></div><div class="line">  <span class="number">06</span>   <span class="number">2</span>   <span class="number">0</span>  <span class="number">23</span></div></pre></td></tr></table></figure>
<p>因此，我们通过与前面相同的图来研究一下，日期作为主要变异来源的可能性，在下面的图形中，我们使用颜色来表示年，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">cols = as.numeric(year)</div><div class="line">mypar()</div><div class="line">plot(s$v[,<span class="number">1</span>],s$v[,<span class="number">2</span>],col=cols,pch=<span class="number">16</span>,</div><div class="line">     xlab=<span class="string">"PC1"</span>,ylab=<span class="string">"PC2"</span>)</div><div class="line">legend(<span class="string">"bottomleft"</span>,levels(year),col=seq(along=levels(year)),pch=<span class="number">16</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916140347.jpeg">

</div>
<p>从上面图形我们看到，日期（以年为单位）也与第1PC非常相关。那么到底是哪个变量促进了这种情况呢？由于存在着很高的混杂(confounding)效应，现在还不清楚是哪个因素起了重要作用。但是，在解决这个问题方面，我们还会进一步探索数据。</p>
<h3 id="pc箱线图">PC箱线图</h3>
<p>样本之间的相关性的图形可以展示出一个复杂的结构，它似乎有5个以的因素（其中一个是年份）参与其中。很明显，这种复杂程度远非种族(ethnicity)这一个因素能解释的。我们此时还探索一下月份之间的相关性，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">month &lt;- format(dates,<span class="string">"%y%m"</span>)</div><div class="line">length( unique(month))</div><div class="line">variable &lt;- as.numeric(month)</div><div class="line">mypar(<span class="number">2</span>,<span class="number">2</span>)</div><div class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:<span class="number">4</span>)&#123;</div><div class="line">  boxplot(split(s$v[,i],variable),las=<span class="number">2</span>,range=<span class="number">0</span>)</div><div class="line">  stripchart(split(s$v[,i],variable),add=<span class="literal">TRUE</span>,vertical=<span class="literal">TRUE</span>,pch=<span class="number">1</span>,cex=<span class="number">.5</span>,col=<span class="number">1</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>一共有21个月份，下图是按照不同月份划分场次层后，根据第1PC到第4PC来绘制的箱线图，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916141037.jpeg">

</div>
<p>从上图我们可以看到，月份与第1PC有着非常强烈的相关性，即使按照种族(ethnicity)和其它因素来划分层次，也是如此。我们要知道，2002-2004之间处理的样本都来自于同一种族群体。在这样的情况下，我们可以使用方差分析来查看一下PC与哪个月份相关，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">corr &lt;- sapply(<span class="number">1</span>:ncol(s$v),<span class="keyword">function</span>(i)&#123;</div><div class="line">  fit &lt;- lm(s$v[,i]~as.factor(month))</div><div class="line">  <span class="keyword">return</span>( summary(fit)$adj.r.squared )</div><div class="line">&#125;)</div><div class="line">mypar()</div><div class="line">plot(seq(along=corr), corr, xlab=<span class="string">"PC"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916141913.jpeg">

</div>
<p>从上图我们可以看到，第1 PC的相关性非常强，而对于前20左右的PC，相关性较强。我们还可以计算一下月份内与月份之间的F统计量，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">Fstats&lt;- sapply(<span class="number">1</span>:ncol(s$v),<span class="keyword">function</span>(i)&#123;</div><div class="line">  fit &lt;- lm(s$v[,i]~as.factor(month))</div><div class="line">  Fstat &lt;- summary(aov(fit))[[<span class="number">1</span>]][<span class="number">1</span>,<span class="number">4</span>]</div><div class="line">  <span class="keyword">return</span>(Fstat)</div><div class="line">&#125;)</div><div class="line">mypar()</div><div class="line">plot(seq(along=Fstats),sqrt(Fstats))</div><div class="line">p &lt;- length(unique(month))</div><div class="line">abline(h=sqrt(qf(<span class="number">0.995</span>,p-<span class="number">1</span>,ncol(s$v)-<span class="number">1</span>)))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190916142249.jpeg">

</div>
<p>上图展示的是，方差分析的F平方根，用于解释PC与月份的关系。</p>
<p>至此为止，我们就了解了如何使用PCA联合EDA来做为一个强大的功能检测并理解批次效应的。在后面的部分里，我们将会了解如何使用PC用于因子分析(factor analysis)，用于改进模型估计。</p>
<h2 id="练习">练习</h2>
<p>P431</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/25/DAL/DALS025_Batch_Effect01_Introduction/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/25/DAL/DALS025_Batch_Effect01_Introduction/" itemprop="url">DALS025-批次效应01-什么是批次效应</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-25T12:00:00+08:00">
                2019-08-25
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Data-Analysis-for-the-life-sciences/" itemprop="url" rel="index">
                    <span itemprop="name">Data Analysis for the life sciences</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,824
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  17
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第10章批次效应的第1小节，这一部分的主要内容涉及批次效应(Batch Effects)的介绍，有关批次效应的前言部分Rmarkdown文档参见作者的<a href="https://github.com/genomicsclass/labs/blob/master/batch/intro_to_batch_effects.Rmd" target="_blank" rel="external">Github</a>。</p>
<h2 id="什么是批次效应">什么是批次效应</h2>
<p>高通量研究中一个经常被忽视的问题就是批次效应(batch effects)，批次效应受到当时检测的实验室条件、试剂批次和人员差异的影响。当批次效应与我们目标结果混淆并惬以不正确的结果时，这就成了一个主要问题。在这一章中，我们将详细地描述批次效应：对于批次效应如何检测、解释、建模和调整。</p>
<p>批次效应是基因组学研究中面临的最大挑战，尤其是在精确医学这个背景下。大多数情况（但并非全部）下，高通量技术已经被报道存在着一种形式或另外一种形式的批次效应[Leek et al. (2010) Nature Reviews Genetics 11, 733-739]。但是，批次效应并非基因组学所特有的。实际上， 在1972年一篇文献中，Mj Youden就在对物理常数的经验估计的背景下提到了批次效应。他指出，物理常数“存在着主观估计的特征”，以及如何在不同实验室之间变化的。例如，在下面的Table1中，Youden就展示了来自于不同实验室又的天文单的估计值。这些报告包括对离差(spread)（现在我们称之为置信区间）的估计。</p>
<figure class="highlight plain"><figcaption><span>astronomical_units,echo</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div></pre></td><td class="code"><pre><div class="line">library(rafalib)</div><div class="line">library(downloader)</div><div class="line">##Download the data from</div><div class="line">url &lt;- &quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/astronomicalunit.csv&quot;</div><div class="line">filename &lt;- tempfile() </div><div class="line">if (!file.exists(filename)) download(url, destfile=filename)</div><div class="line"></div><div class="line">dat &lt;- read.csv(filename)</div><div class="line">year &lt;-  jitter(dat[,2]) ##add jitter so points are not on top of each other</div><div class="line"></div><div class="line">##Use color to denote the labs that reported more than one measurement</div><div class="line">labs &lt;- as.character(dat[,1]) ##what lab did it</div><div class="line">labs[ !labs%in%c(&quot;Jodrell Bank&quot;,&quot;Spencer Jones&quot;)] &lt;- &quot;Others&quot;</div><div class="line">labs &lt;- factor(labs, levels=c(&quot;Others&quot;,&quot;Spencer Jones&quot;,&quot;Jodrell Bank&quot;))</div><div class="line">cols=as.numeric(labs)</div><div class="line"></div><div class="line">current &lt;- 92.956039 ##this is the current estimate in millions of mph</div><div class="line"></div><div class="line">mypar()</div><div class="line">plot(year, dat[,3], ylim=c(min(dat[,4]),max(dat[,5])), pch=16, col=cols, </div><div class="line">     xlab=&quot;Year&quot;,ylab=&quot;Astronomical unit (millions of miles)&quot;)</div><div class="line">for(i in 1:nrow(dat))</div><div class="line">  lines(c(year[i],year[i]),c(dat[i,4],dat[i,5]),col=cols[i],lwd=3)</div><div class="line">legend(&quot;topright&quot;, legend=levels(labs), col=seq_along( labs ) ,cex=0.75, lty=1,pch=16)</div><div class="line">abline(h=current,lty=2)</div><div class="line">text(1905,current,&quot;Current estimate&quot;,pos=3)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914105414.jpeg">

</div>
<p>从不同实验室之间的可变性以及报道的界限（可以理解为置信区间）来看，都不能解释这种可变性，这就清楚地表明不同实验室之间，而非某个实验室内部存在着某些效应。这种类型的变异就是我们所谓的批次效应。请注意，有些实验室报告了两个估计值（紫色和橙色），这样我们就在相同的实验室也看到了批次效应。 我们可以使用统计学符号来精确地描述这个问题。我们使用下面的公式来表示这些检测值： <span class="math display">\[
Y_{i,j} = 
\mu + \varepsilon_{i,j}, j=1,\dots,N
\]</span></p>
<p>其中，<span class="math inline">\(Y_{i,j}\)</span> 表示第 <span class="math inline">\(i\)</span> 个实验室的第 <span class="math inline">\(j\)</span> 个检测值，而 <span class="math inline">\(\mu\)</span> 表示真实的物理常数，<span class="math inline">\(\varepsilon_{i,j}\)</span> 表示独立的检测误差。为了解释可变性，我们引入了 <span class="math inline">\(\varepsilon_{i,j}\)</span> ，随后我们根据数据来计算标准误。就像本书前面提到的那样，我们使用 <span class="math inline">\(N\)</span> 值均值来估计物理物理常数，如下所示： <span class="math display">\[
\bar{Y}_i = 
\frac{1}{N} \sum_{i=1}^{N} Y_{i,j}
\]</span></p>
<p>再来构建一个置信区间： <span class="math display">\[
\bar{Y}_i 
 \pm 2 s_i / \sqrt{N} \mbox{ with }
s_i^2= 
\frac{1}{N-1} \sum_{i=1}^N (Y_{i,j} - 
\bar{Y}_i)^2
\]</span></p>
<p>但是，这个置信区间太小，它无法覆盖批次效应的变异，下面是一个更加合适的模型： <span class="math display">\[
Y_{i,j} = \mu +
\gamma_i + \varepsilon_{i,j}, j=1, \dots, N
\]</span></p>
<p>其中 <span class="math inline">\(\gamma_i\)</span> 表示一个实验室特定的偏差或批次效应(batch effect)。从图片上我们可以明显看出来，实验室之间 <span class="math inline">\(\gamma\)</span> 的变化要远大于一个实验室内部 <span class="math inline">\(\varepsilon\)</span> 的变化。用统计学的术语来描述这个问题就是，<span class="math inline">\(\mu\)</span> 和 <span class="math inline">\(\gamma\)</span> 无法被识别。我们可以估计 <span class="math inline">\(\mu_i+\gamma_i\)</span> ，但是无法区分开。我们可以将 <span class="math inline">\(\gamma\)</span> 视为一个随机变量。在这个案例中，每个实验室都有一个错误项 <span class="math inline">\(\gamma_i\)</span> ，这一项在贯穿于该实验室的所有检测值，在每个检测值中，这一项是相同的，但是实验室与实验室间的这一项则不同。因此，这个问题就可以用以下方程表示： <span class="math display">\[
 s_i / \sqrt{N} \mbox{ with } 
 s_i^2= 
\frac{1}{N-1} \sum_{i=1}^N (Y_{ij} - 
\bar{Y}_i)^2
\]</span></p>
<p>这是对标准差的低估，因为它没有解释由 <span class="math inline">\(\gamma\)</span> 导致的实验室内的相关性（这一句不懂，原文如下：</p>
<blockquote>
<p>Under this interpretation the problem is that:is an underestimate of the standard error since it does not account for the within lab correlation induced by <span class="math inline">\(\gamma\)</span>.</p>
</blockquote>
<p>如果我们假设 <span class="math inline">\(\gamma=0\)</span> ，那么利用来自几个实验室的数据，我们实际上可以估计出 <span class="math inline">\(\gamma\)</span> 。或者说我们可以将它们视为随机效应，简单地将它们当成一个新的估计值，并使用所有的检测值来计算其标准差。以下是我们将报告中的均值值视为随机观察值的置信区间：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">avg &lt;- mean(dat[,3])</div><div class="line">se &lt;- sd(dat[,3]) / sqrt(nrow(dat))</div><div class="line">cat(&quot;95% confidence interval is: [&quot;,avg-1.96*se,&quot;,&quot;, avg+1.96*se,&quot;]&quot;)</div><div class="line">cat(&quot;which does include the current estimate is:&quot;,current)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; cat(<span class="string">"95% confidence interval is: ["</span>,avg-<span class="number">1.96</span>*se,<span class="string">","</span>, avg+<span class="number">1.96</span>*se,<span class="string">"]"</span>)</div><div class="line"><span class="number">95</span>% confidence interval is: [ <span class="number">92.8727</span> , <span class="number">92.98542</span> ]&gt; cat(<span class="string">"which does include the current estimate is:"</span>,current)</div><div class="line">which does include the current estimate is: <span class="number">92.95604</span></div></pre></td></tr></table></figure>
<p>Youden的论文还包括最近关于光速以及策略常数的批次效应估计的案例。在这一章里，我们只展示高通量生物数据中批次效应的广泛性和复杂性。</p>
<p>## 混杂</p>
<p>书中有一个术语，即<code>confounding</code>，这里译为<code>混杂</code>。这一部分内容的Rmarkdown文档可以参考作者的<a href="https://github.com/genomicsclass/labs/blob/master/batch/confounding.Rmd" target="_blank" rel="external">Github</a>。</p>
<p>当批次效应与我们的目标结果混杂时，就会导致严重的𨤹。这里我们描述一下混杂(confounding)，以及它与我们数据解释的关系。</p>
<p>我们从这本书或者说从任何其它数据分析课程中尝到最重要的思想之一就是“相关不等于因果”。这句话多数情况就是真的，一个常见的案例就是混杂(confounding)。简单地讲，当我们观察到 <span class="math inline">\(X\)</span> 和 <span class="math inline">\(Y\)</span> 之间存在着相关(correlation)或关联(association)时，往往就会存在着混杂，但严格来说，这是因为 <span class="math inline">\(X\)</span> 和 <span class="math inline">\(Y\)</span> 都依赖于一个无关的变量 <span class="math inline">\(Z\)</span> 。这里我们会描述一个Simposon悖论，这是一个基于一个著名的法律案件的案例，接着，我们还会提到一个在高通量生物学研究中的一个混杂案例。</p>
<h3 id="案例之simpson悖论">案例之Simpson悖论</h3>
<p>加州大学伯克分校(UCB)1973年的入学数据显示，在录取的学生中，44%是男性，30%是女性，显著男性更多，以下是这些数据：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">library(dagdata)</div><div class="line">data(admissions)</div><div class="line">admissions$total=admissions$Percent*admissions$Number/100</div><div class="line">##percent men get in</div><div class="line">sum(admissions$total[admissions$Gender==1]/sum(admissions$Number[admissions$Gender==1]))</div><div class="line">##percent women get in</div><div class="line">sum(admissions$total[admissions$Gender==0]/sum(admissions$Number[admissions$Gender==0]))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; sum(admissions$total[admissions$Gender==<span class="number">1</span>]/sum(admissions$Number[admissions$Gender==<span class="number">1</span>]))</div><div class="line">[<span class="number">1</span>] <span class="number">0.4451951</span></div><div class="line">&gt; <span class="comment">##percent women get in</span></div><div class="line">&gt; sum(admissions$total[admissions$Gender==<span class="number">0</span>]/sum(admissions$Number[admissions$Gender==<span class="number">0</span>]))</div><div class="line">[<span class="number">1</span>] <span class="number">0.3033351</span></div></pre></td></tr></table></figure>
<p>卡方检验的结果明确拒绝零假设，即录取的性别是独立的，两者不受影响（也就是说，男女录取公平），如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">##make a 2 x 2 table</div><div class="line">index = admissions$Gender==1</div><div class="line">men = admissions[index,]</div><div class="line">women = admissions[!index,]</div><div class="line">menYes = sum(men$Number*men$Percent/100)</div><div class="line">menNo = sum(men$Number*(1-men$Percent/100))</div><div class="line">womenYes = sum(women$Number*women$Percent/100)</div><div class="line">womenNo = sum(women$Number*(1-women$Percent/100))</div><div class="line">tab = matrix(c(menYes,womenYes,menNo,womenNo),2,2)</div><div class="line">print(chisq.test(tab)$p.val)</div></pre></td></tr></table></figure>
<p>p值如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; print(chisq.test(tab)$p.val)</div><div class="line">[<span class="number">1</span>] <span class="number">9.139492e-22</span></div></pre></td></tr></table></figure>
<p>但经过更仔细的观察会发现一个自相矛盾的结果，以下是按专业划分后，不同性别的录取百分比：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">y=cbind(admissions[1:6,c(1,3)],admissions[7:12,3])</div><div class="line">colnames(y)[2:3]=c(&quot;Male&quot;,&quot;Female&quot;)</div><div class="line">y</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; y</div><div class="line">  Major Male Female</div><div class="line"><span class="number">1</span>     A   <span class="number">62</span>     <span class="number">82</span></div><div class="line"><span class="number">2</span>     B   <span class="number">63</span>     <span class="number">68</span></div><div class="line"><span class="number">3</span>     C   <span class="number">37</span>     <span class="number">34</span></div><div class="line"><span class="number">4</span>     D   <span class="number">33</span>     <span class="number">35</span></div><div class="line"><span class="number">5</span>     E   <span class="number">28</span>     <span class="number">24</span></div><div class="line"><span class="number">6</span>     <span class="literal">F</span>    <span class="number">6</span>      <span class="number">7</span></div></pre></td></tr></table></figure>
<p>从结果中我们可以发现，不同专业之间没有性别偏见。</p>
<p>但是，我们在前面使用了卡方检验发现，入学和性别之间存在着某种关系。然而，当我们的数据按照不同专业进行分组时，这种依赖性似乎消失了，这是怎么一回事呢？</p>
<p>这就是Simpson悖论的一个案例。</p>
<p>我们上面进行的卡方检验表明，入学和性别之间存在依赖关系。然而，当数据按专业分组时，这种依赖性似乎消失了。到底怎么回事？现在我们绘制一个能显示申请专业的人，以及最终进入这个专业读书的人的百分比的图形，用于说明男女在录取比例方面的问题，如下所示：</p>
<figure class="highlight plain"><figcaption><span>hard_major_confounding, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">y=cbind(admissions[1:6,5],admissions[7:12,5])</div><div class="line">y=sweep(y,2,colSums(y),&quot;/&quot;)*100</div><div class="line">x=rowMeans(cbind(admissions[1:6,3],admissions[7:12,3]))</div><div class="line">library(rafalib)</div><div class="line">mypar()</div><div class="line">matplot(x,y,xlab=&quot;percent that gets in the major&quot;,</div><div class="line">        ylab=&quot;percent that applies to major&quot;,</div><div class="line">        col=c(&quot;blue&quot;,&quot;red&quot;),cex=1.5)</div><div class="line">legend(&quot;topleft&quot;,c(&quot;Male&quot;,&quot;Female&quot;),col=c(&quot;blue&quot;,&quot;red&quot;),pch=c(&quot;1&quot;,&quot;2&quot;),box.lty=0)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914162851.jpeg">

</div>
<p>从图片上我们可以发现，男生更想申请那些“容易”的专业。也就是说，男生这个变量与“容易”专业这个变量发生了混杂。</p>
<h3 id="混杂的图形解释">混杂的图形解释</h3>
<p>在这一部分里，我们将混杂图形化。在下面的图形中，每个字母表示一个学生。被录取的人用绿色表示，字母表示专业。在第1张图中，所有相同性别的学生都被放在一起，我们可以发现，男生中绿色的比较更大，如下所示：</p>
<figure class="highlight plain"><figcaption><span>simpsons_paradox_illustration, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div></pre></td><td class="code"><pre><div class="line">###make data for plot</div><div class="line">library(rafalib)</div><div class="line">mypar()</div><div class="line">CEX=0.5</div><div class="line">NC &lt;- 70</div><div class="line">tmp=rowSums(tab)</div><div class="line">FNC &lt;- round(NC*tmp[2]/tmp[1])</div><div class="line">SCALE &lt;- 1</div><div class="line">makematrix&lt;-function(x,n,addx=0,addy=0)&#123;</div><div class="line">  m&lt;-ceiling(length(x)/n)</div><div class="line">  expand.grid(1:n+addx,addy+1:m)[seq(along=x),] </div><div class="line">&#125;</div><div class="line">males&lt;- sapply(1:6,function(i)&#123;</div><div class="line">  tot=admissions[i,2]*SCALE</div><div class="line">  p=admissions[i,3]/100</div><div class="line">  x=rep(c(0,1),round(tot*c(1-p,p)))</div><div class="line">&#125;)</div><div class="line">allmales&lt;-Reduce(c,males)</div><div class="line">females&lt;- sapply(7:12,function(i)&#123;</div><div class="line">  tot=admissions[i,2]*SCALE</div><div class="line">  p=admissions[i,3]/100</div><div class="line">  rep(c(0,1),round(tot*c(1-p,p)))</div><div class="line">&#125;)</div><div class="line">allfemales&lt;-Reduce(c,females)</div><div class="line">mypar(1,1)</div><div class="line">malepoints &lt;- makematrix(allmales,NC)</div><div class="line">femalepoints &lt;- makematrix(allfemales,FNC,NC+NC/10)</div><div class="line">NR &lt;- max(c(malepoints[,2],femalepoints[,2]))</div><div class="line">plot(0,type=&quot;n&quot;,xlim=c(min(malepoints[,1]),max(femalepoints[,1])),ylim=c(0,NR),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,xlab=&quot;&quot;,ylab=&quot;&quot;)</div><div class="line">PCH=LETTERS[rep(1:6,sapply(males,length))]</div><div class="line">o&lt;-order(-allmales)</div><div class="line">points(malepoints,col=2-allmales[o],pch=PCH[o],cex=CEX)</div><div class="line">PCH=LETTERS[rep(1:6,sapply(females,length))]</div><div class="line">o&lt;-order(-allfemales)</div><div class="line">points(femalepoints,col=2-allfemales[o],pch=PCH[o],cex=CEX)</div><div class="line">abline(v=NC+NC/20)</div><div class="line">axis(side=3,c(NC/2,NC+NC/2),c(&quot;Male&quot;,&quot;Female&quot;),tick=FALSE)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914163227.jpeg">

</div>
<p>现在我们按照专业对不同性别的学生进行分级。这里我们要注意，大多数被录取的学生（绿色）来源于两个容易的专业，即A和B，如下所示：</p>
<figure class="highlight plain"><figcaption><span>simpsons_paradox_illustration2, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">mypar()</div><div class="line">malepoints &lt;- vector(&quot;list&quot;,length(males))</div><div class="line">femalepoints &lt;- vector(&quot;list&quot;,length(males))</div><div class="line">N&lt;- length(males)</div><div class="line"> </div><div class="line">ADDY &lt;- vector(&quot;numeric&quot;,N+1)</div><div class="line">for(i in 1:N)&#123;</div><div class="line">  malepoints[[i]] &lt;- makematrix(males[[i]],NC,0,ADDY[i])</div><div class="line">  femalepoints[[i]] &lt;- makematrix(females[[i]],FNC,NC+NC/10,ADDY[i])</div><div class="line">   ADDY[i+1] &lt;- max(malepoints[[i]][,2],femalepoints[[i]][,2])+1</div><div class="line">&#125;</div><div class="line">plot(0,type=&quot;n&quot;,</div><div class="line">     xlim=c( min(sapply(malepoints,function(x)min(x[,1]))),max(sapply(femalepoints,function(x)max(x[,1])))),</div><div class="line">  ylim=c(0,max(sapply(femalepoints,function(x)max(x[,2])))),xaxt=&quot;n&quot;,yaxt=&quot;n&quot;,xlab=&quot;&quot;,ylab=&quot;&quot;)</div><div class="line">          </div><div class="line">for(i in 1:N)&#123;</div><div class="line">  points(malepoints[[i]],col=2+sort(-males[[i]]),pch=LETTERS[i],cex=CEX)</div><div class="line">  points(femalepoints[[i]],col=2+sort(-females[[i]]),pch=LETTERS[i],cex=CEX)</div><div class="line">  if(i&gt;1) abline(h=ADDY[i])</div><div class="line">  &#125;</div><div class="line">abline(v=NC+NC/20)</div><div class="line">axis(side=3,c(NC/2,NC+FNC/2),c(&quot;Male&quot;,&quot;Female&quot;),tick=FALSE)</div><div class="line">axis(side=2,ADDY[-1]/2+ADDY[-length(ADDY)]/2,LETTERS[1:N],tick=FALSE,las=1)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914163301.jpeg">

</div>
<h3 id="分层后的均值">分层后的均值</h3>
<p>在下图中，我们可以看到，我们根据专业设定条件或分层后，然后再看差异，也就是说，当我们控制了混杂项后，这就混杂效应就消失了，如下所示：</p>
<figure class="highlight plain"><figcaption><span>admission_by_major, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">y=cbind(admissions[1:6,3],admissions[7:12,3])</div><div class="line">matplot(1:6,y,xaxt=&quot;n&quot;,xlab=&quot;major&quot;,ylab=&quot;percent&quot;,col=c(&quot;blue&quot;,&quot;red&quot;),cex=1.5)</div><div class="line">axis(1,1:6,LETTERS[1:6])</div><div class="line">legend(&quot;topright&quot;,c(&quot;Male&quot;,&quot;Female&quot;),col=c(&quot;blue&quot;,&quot;red&quot;),pch=c(&quot;1&quot;,&quot;2&quot;),</div><div class="line">       box.lty=0)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914163912.jpeg">

</div>
<p>事实上，在不同专业录取的学生中，性别差异仅表示为，男生比女生高3.5%，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">mean(y[,1]-y[,2])</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; mean(y[,<span class="number">1</span>]-y[,<span class="number">2</span>])</div><div class="line">[<span class="number">1</span>] -<span class="number">3.5</span></div></pre></td></tr></table></figure>
<h3 id="棒球中的simpson悖论">棒球中的Simpson悖论</h3>
<p>在棒球比赛的统计中我们也经常看到Simpson悖论，现在我们来看一个有名的案例，在1995年和1996年，David Justice的平均击球率高于Derek Jeter，但是Jeter的击球率却高于整体平均水平，如下所示：</p>
<table>
<thead>
<tr class="header">
<th></th>
<th>1995</th>
<th>1996</th>
<th>Combined</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Derek Jeter</td>
<td>12/48 (.250)</td>
<td>183/582 (.314)</td>
<td>195/630 (.310)</td>
</tr>
<tr class="even">
<td>David Justice</td>
<td>104/411 (.253)</td>
<td>45/140 (.321)</td>
<td>149/551 (.270)</td>
</tr>
</tbody>
</table>
<p>这里的混杂项就是比赛场次，Jeter的比赛场次数目更多，但是结果却是Justice击球率更高。</p>
<h3 id="混杂之高通量案例">混杂之高通量案例</h3>
<p>为了描述我们在生物学中遇到的混杂问题，我们将会使用<a href="https://www.ncbi.nlm.nih.gov/pubmed/17206142" target="_blank" rel="external">《Common genetic variants account for differences in gene expression among ethnic groups》</a>这篇文献中的数据集，这篇文献指出，两个不同种族的血液大约有50%的差异基因，现在我们来下载这个数据集：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(Biobase) ##available from Bioconductor</div><div class="line">library(genefilter) </div><div class="line">load(&quot;GSE5859.rda&quot;) ##available from github</div></pre></td></tr></table></figure>
<p>我们使用Bioconductor中的函数<code>exprs</code>和<code>pData</code>就能提取基因的表达数据与样本信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">geneExpression = exprs(e)</div><div class="line">sampleInfo = pData(e)</div></pre></td></tr></table></figure>
<p>需要注意，样本按照不同的时间进行了处理，如下所示：</p>
<p>Note that some samples were processed at different times.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">head(sampleInfo)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; head(sampleInfo)</div><div class="line">  ethnicity       date        filename</div><div class="line"><span class="number">1</span>       CEU <span class="number">2003</span>-<span class="number">02</span>-<span class="number">04</span> GSM25349.CEL.gz</div><div class="line"><span class="number">2</span>       CEU <span class="number">2003</span>-<span class="number">02</span>-<span class="number">04</span> GSM25350.CEL.gz</div><div class="line"><span class="number">3</span>       CEU <span class="number">2002</span>-<span class="number">12</span>-<span class="number">17</span> GSM25356.CEL.gz</div><div class="line"><span class="number">4</span>       CEU <span class="number">2003</span>-<span class="number">01</span>-<span class="number">30</span> GSM25357.CEL.gz</div><div class="line"><span class="number">5</span>       CEU <span class="number">2003</span>-<span class="number">01</span>-<span class="number">03</span> GSM25358.CEL.gz</div><div class="line"><span class="number">6</span>       CEU <span class="number">2003</span>-<span class="number">01</span>-<span class="number">16</span> GSM25359.CEL.gz</div></pre></td></tr></table></figure>
<p>日期是一个无关的变量，它理论上不影响这个数据集中基因的表达值，然而，正如我们在前面分析中看到的那样，这个日期似乎也起了一些作用，因此我们在这里就来讲一下这个日期的因素。 我们可以明显看到，日期和种族这两个因素几乎完全混杂了：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">year = factor( format(sampleInfo$date,&quot;%y&quot;) )</div><div class="line">tab = table(year,sampleInfo$ethnicity)</div><div class="line">print(tab)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; print(tab)</div><div class="line">    </div><div class="line">year ASN CEU HAN</div><div class="line">  <span class="number">02</span>   <span class="number">0</span>  <span class="number">32</span>   <span class="number">0</span></div><div class="line">  <span class="number">03</span>   <span class="number">0</span>  <span class="number">54</span>   <span class="number">0</span></div><div class="line">  <span class="number">04</span>   <span class="number">0</span>  <span class="number">13</span>   <span class="number">0</span></div><div class="line">  <span class="number">05</span>  <span class="number">80</span>   <span class="number">3</span>   <span class="number">0</span></div><div class="line">  <span class="number">06</span>   <span class="number">2</span>   <span class="number">0</span>  <span class="number">24</span></div></pre></td></tr></table></figure>
<p>通过t检验以及生成的火山图我们可以发现，不同种族之间似乎存在着数千个差异基因。但是，当我们仅仅对2002年和2003年的CEU人群进行比较发现，我们又发现了数千个差异基因：</p>
<figure class="highlight plain"><figcaption><span>volcano_plots, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">library(genefilter)</div><div class="line">##remove control genes</div><div class="line">out &lt;- grep(&quot;AFFX&quot;,rownames(geneExpression))</div><div class="line">eth &lt;- sampleInfo$ethnicity</div><div class="line">ind&lt;- which(eth%in%c(&quot;CEU&quot;,&quot;ASN&quot;))</div><div class="line">res1 &lt;- rowttests(geneExpression[-out,ind],droplevels(eth[ind]))</div><div class="line">ind &lt;- which(year%in%c(&quot;02&quot;,&quot;03&quot;) &amp; eth==&quot;CEU&quot;)</div><div class="line">res2 &lt;- rowttests(geneExpression[-out,ind],droplevels(year[ind]))</div><div class="line">XLIM &lt;- max(abs(c(res1$dm,res2$dm)))*c(-1,1)</div><div class="line">YLIM &lt;- range(-log10(c(res1$p,res2$p)))</div><div class="line">mypar(1,2)</div><div class="line">plot(res1$dm,-log10(res1$p),xlim=XLIM,ylim=YLIM,</div><div class="line">     xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;,main=&quot;Populations&quot;)</div><div class="line">plot(res2$dm,-log10(res2$p),xlim=XLIM,ylim=YLIM,</div><div class="line">     xlab=&quot;Effect size&quot;,ylab=&quot;-log10(p-value)&quot;,main=&quot;2003 v 2002&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190914165820.jpeg">

</div>
<h2 id="练习">练习</h2>
<p>P419</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/24/DAL/DALS024_Basic_Machine_Learning03_Class_Prediction/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/24/DAL/DALS024_Basic_Machine_Learning03_Class_Prediction/" itemprop="url">DALS024-机器学习03-分类预测</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-24T12:00:00+08:00">
                2019-08-24
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Data-Analysis-for-the-life-sciences/" itemprop="url" rel="index">
                    <span itemprop="name">Data Analysis for the life sciences</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  6,035
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  27
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第9章机器学习的第3小节，这一部分的主要内容涉及类预测(Class Prediction)，这一部分相关的Rmarkdown文档参见作者的<a href="https://github.com/genomicsclass/labs/blob/master/ml/machine_learning.Rmd" target="_blank" rel="external">Github</a>。</p>
<p>在这一部分中我们主要介绍分类预测(class prediction)。实际上，许多人将分类预测称为机器学习，有的时候我们会交替使用这两个术语。我们会对这个庞杂的主题做一个非常简单的介绍，重点关注一些具体的案例。</p>
<p>我们这里使用的案例来源于统计学的经典书籍，即Trevor Hastie, Robert Tibshirani and Jerome Friedman的《<em>The Elements of Statistical Learning: Data Mining, Inference, and Prediction</em>》。与回归中的推断类似，机器学习(ML)也是研究结果 <span class="math inline">\(Y\)</span> 和协变量 <span class="math inline">\(X\)</span> 之间的关系。在ML中，我们将 <span class="math inline">\(X\)</span> 称为预测因子或特征值，ML和推断之间的主要区别在于，在ML中，我们主要研究使用 <span class="math inline">\(X\)</span> 来预测 <span class="math inline">\(Y\)</span> 。关于统计模型，我在常规的推断中，我们主要用来估计和解释模型参数，但在ML中，统计模型只是我们达到目的的手段，即预测 <span class="math inline">\(Y\)</span> 。</p>
<p>这里我们介绍理解ML的主要概念，以及两个具体的算法：回归(regression)和k近似算法(kNN,k nearest neighbors)。我们需要知道，绐中学习有几十种流行的算法，我们这里不一一列举。</p>
<p>在前面部分里，我们介绍了非常简单的单一预测因子案例。但是，大多数与这种案例相关的案例往往不止一个预测因子。为了说明这个问题，我们现在再介绍一个案例，其中 <span class="math inline">\(X\)</span> 是一个二维数据，<span class="math inline">\(Y\)</span> 是一个二分类结果。这个案例来源于Hastie, Tibshirani 和 Friedman的书中，在这个案例中， <span class="math inline">\(X\)</span> 和 <span class="math inline">\(Y\)</span> 并不线性关系。在下面的图形中，使用不 的颜色表示了实际的 <span class="math inline">\(f(x_1,x_2)=E(Y \mid X_1=x_1,X_2=x_2)\)</span> 值。下面的代码用于生成一个相对复杂的条件概率函数。我们随后会使用测试数据集与训练数据集。在下图中，我们使用红色来表示 <span class="math inline">\(f(x_1,x_2)\)</span>中接近于1的数据，使用蓝色表示接近于0的数据，中间过渡态使用黄色表示，如下所示：</p>
<figure class="highlight plain"><figcaption><span>conditional_prob, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div></pre></td><td class="code"><pre><div class="line">library(rafalib)</div><div class="line">library(RColorBrewer)</div><div class="line">hmcol &lt;- colorRampPalette(rev(brewer.pal(11, &quot;Spectral&quot;)))(100)</div><div class="line">mycols=c(hmcol[1],hmcol[100])</div><div class="line">set.seed(1)</div><div class="line">##create covariates and outcomes</div><div class="line">##outcomes are alwasy 50 0s and 50 1s</div><div class="line">s2=0.15</div><div class="line">##pick means to create a non linear conditional expectation</div><div class="line">library(MASS)</div><div class="line">M0 &lt;- mvrnorm(10,c(1,0),s2*diag(2)) ##generate 10 means</div><div class="line">M1 &lt;- rbind(mvrnorm(3,c(1,1),s2*diag(2)),</div><div class="line">            mvrnorm(3,c(0,1),s2*diag(2)),</div><div class="line">            mvrnorm(4,c(0,0),s2*diag(2)))</div><div class="line">###funciton to generate random pairs</div><div class="line">s&lt;- sqrt(1/5)</div><div class="line">N=200</div><div class="line">makeX &lt;- function(M,n=N,sigma=s*diag(2))&#123;</div><div class="line">  z &lt;- sample(1:10,n,replace=TRUE) ##pick n at random from above 10</div><div class="line">  m &lt;- M[z,] ##these are the n vectors (2 components)</div><div class="line">  return(t(apply(m,1,function(mu) mvrnorm(1,mu,sigma)))) ##the final values</div><div class="line">&#125;</div><div class="line">###create the training set and the test set</div><div class="line">x0 &lt;- makeX(M0)##the final values for y=0 (green)</div><div class="line">testx0 &lt;- makeX(M0)</div><div class="line">x1 &lt;- makeX(M1)</div><div class="line">testx1 &lt;-makeX(M1)</div><div class="line">x &lt;- rbind(x0,x1) ##one matrix with everything</div><div class="line">test &lt;- rbind(testx0,testx1)</div><div class="line">y &lt;- c(rep(0,N),rep(1,N)) #the outcomes</div><div class="line">ytest &lt;- c(rep(0,N),rep(1,N))</div><div class="line">cols &lt;- mycols[c(rep(1,N),rep(2,N))]</div><div class="line">colstest &lt;- cols</div><div class="line">##Create a grid so we can predict all of X,Y</div><div class="line">GS &lt;- 150 ##grid size is GS x GS</div><div class="line">XLIM &lt;- c(min(c(x[,1],test[,1])),max(c(x[,1],test[,1])))</div><div class="line">tmpx &lt;- seq(XLIM[1],XLIM[2],len=GS)</div><div class="line">YLIM &lt;- c(min(c(x[,2],test[,2])),max(c(x[,2],test[,2])))</div><div class="line">tmpy &lt;- seq(YLIM[1],YLIM[2],len=GS)</div><div class="line">newx &lt;- expand.grid(tmpx,tmpy) #grid used to show color contour of predictions</div><div class="line">###Bayes rule: best possible answer</div><div class="line">p &lt;- function(x)&#123; ##probability of Y given X</div><div class="line">  p0 &lt;- mean(dnorm(x[1],M0[,1],s)*dnorm(x[2],M0[,2],s))</div><div class="line">  p1 &lt;- mean(dnorm(x[1],M1[,1],s)*dnorm(x[2],M1[,2],s))</div><div class="line">  p1/(p0+p1)</div><div class="line">&#125;</div><div class="line">###Create the bayesrule prediction</div><div class="line">bayesrule &lt;- apply(newx,1,p)</div><div class="line">colshat &lt;- bayesrule</div><div class="line">colshat &lt;- hmcol[floor(bayesrule*100)+1]</div><div class="line">mypar()</div><div class="line">plot(x,type=&quot;n&quot;,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">points(newx,col=colshat,pch=16,cex=0.35)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190912170040.jpeg">

</div>
<p>如果我们将那些 <span class="math inline">\(E(Y \mid X=x)&gt;0.5\)</span> 的点用红色表示，剩下的点用蓝色表示，我们就能看到一条明显的分界线，它将0与1的区域分开了，如下所示：</p>
<figure class="highlight plain"><figcaption><span>bayes_rule,fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">mypar()</div><div class="line">colshat[bayesrule&gt;=0.5] &lt;- mycols[2]</div><div class="line">colshat[bayesrule&lt;0.5] &lt;- mycols[1]</div><div class="line">plot(x,type=&quot;n&quot;,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">points(newx,col=colshat,pch=16,cex=0.35)</div><div class="line">contour(tmpx,tmpy,matrix(round(bayesrule),GS,GS),levels=c(1,2),</div><div class="line">        add=TRUE,drawlabels=FALSE)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190912170527.jpeg">

</div>
<p>通过上面的图形我们并没有看到“真相”(truth)。大多数的ML方法涉及估计的 <span class="math inline">\(f(x)\)</span> 。通常第一步就是将一个样本作为参数，也就是训练集(training set)，用它来估计 <span class="math inline">\(f(x)\)</span> 。我们将回顾一下两种具体的ML技术。首先，我们需要回顾一下我们用评估这些方法性能的主要概念。</p>
<h3 id="训练集">训练集</h3>
<p>在第一张图中，我们创建了一个训练集和一个测试集，现在我们画出来，如下所示：</p>
<figure class="highlight plain"><figcaption><span>test_train, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">#x, test, cols, and coltest were created in code that was not shown</div><div class="line">#x is training x1 and x2, test is test x1 and x2</div><div class="line">#cols (0=blue, 1=red) are training observations</div><div class="line">#coltests are test observations</div><div class="line">mypar(1,2)</div><div class="line">plot(x,pch=21,bg=cols,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">plot(test,pch=21,bg=colstest,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190912171217.jpeg">

</div>
<p>从上面我们可以看到，训练集（左侧）与测试集（右侧）有着相似的全局属性，因为它们是用相同的随机谈量生成的（蓝色点都趋向分布于右下角），但是它们的构建的过程还是不同的。原因在于，我们创建测试集和训练集的原因是通过测试与用于拟合模型或训练算法的数据不同的数据来检测过度训练。 我们将在下面看到它的重要性。</p>
<h4 id="利用回归进行预测">利用回归进行预测</h4>
<p>关于ML问题的第一个简单方法就是拟合一个双变量线性回归模型：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">##x and y were created in the code (not shown) for the first plot</div><div class="line">#y is outcome for the training set</div><div class="line">X1 &lt;- x[,1] ##these are the covariates</div><div class="line">X2 &lt;- x[,2] </div><div class="line">fit1 &lt;- lm(y~X1+X2)</div></pre></td></tr></table></figure>
<p>一旦我们有了这些拟合的数据，我们就能使用 <span class="math inline">\(\hat{f}(x_1,x_2)=\hat{\beta}_0 + \hat{\beta}_1x_1 +\hat{\beta}_2 x_2\)</span> 来估计 <span class="math inline">\(f(x_1,x_2)\)</span> 。为了提供一个实际的预测结果，我们仅仅预测当 <span class="math inline">\(\hat{f}(x_1,x_2)&gt;0.5\)</span> 时结果为1（这一段不懂）。我们现在检测一个在训练集与测试集中的错误率，并绘制出边界区域，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">##prediction on train</div><div class="line">yhat &lt;- predict(fit1)</div><div class="line">yhat &lt;- as.numeric(yhat&gt;0.5)</div><div class="line">cat(&quot;Linear regression prediction error in train:&quot;,1-mean(yhat==y),&quot;\n&quot;)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">Linear regression prediction error <span class="keyword">in</span> train: <span class="number">0.295</span></div></pre></td></tr></table></figure>
<p>我们使用<code>predict()</code>函数就能很快地从任意数据集中来获得预测值，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">yhat &lt;- predict(fit1,newdata=data.frame(X1=newx[,1],X2=newx[,2]))</div></pre></td></tr></table></figure>
<p>现在我们生成图片，用于展示我们预测的1和0在图片上的分布，以及边界。我们还可以使用<code>predict()</code>函数从我们的测试集中生成预测数据。需要注意的是，我们无法在测试集中拟合模型：</p>
<figure class="highlight plain"><figcaption><span>regression_prediction, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">colshat &lt;- yhat</div><div class="line">colshat[yhat&gt;=0.5] &lt;- mycols[2]</div><div class="line">colshat[yhat&lt;0.5] &lt;- mycols[1]</div><div class="line">m &lt;- -fit1$coef[2]/fit1$coef[3] #boundary slope</div><div class="line">b &lt;- (0.5 - fit1$coef[1])/fit1$coef[3] #boundary intercept</div><div class="line">##prediction on test</div><div class="line">yhat &lt;- predict(fit1,newdata=data.frame(X1=test[,1],X2=test[,2]))</div><div class="line">yhat &lt;- as.numeric(yhat&gt;0.5)</div><div class="line">cat(&quot;Linear regression prediction error in test:&quot;,1-mean(yhat==ytest),&quot;\n&quot;)</div><div class="line">plot(test,type=&quot;n&quot;,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">abline(b,m)</div><div class="line">points(newx,col=colshat,pch=16,cex=0.35)</div><div class="line">##test was created in the code (not shown) for the first plot</div><div class="line">points(test,bg=cols,pch=21)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190912173537.jpeg">

</div>
<p>在上图中，我们使用 <span class="math inline">\(X_{1}\)</span>和 <span class="math inline">\(X_{2}\)</span> 作为预测因子估计了1的概率，预测的结果将高于0.5的数据标注成了红色，低于0.5的标注为了蓝色。</p>
<p>从计算结果来年地，训练集与测试集的错误率非常相似。因此我们可以相信似乎没有过度训练。这并不奇怪，因为我们使用了2参数模型来拟合400个数据点。不过需要注意的是，边界是一个直线。因为我们为这些数据拟合了一个平面，所以这里没有其它选择。线性回归方法过于僵化。这种僵化会让它稳定，并且避免过度训练。，但是线性回归也不能适合对 <span class="math inline">\(Y\)</span> 和 <span class="math inline">\(X\)</span> 之间的非线性关系进行拟合。我们之前在平滑部分中看到了这些东西。下一个ML技术将会达到我们之间平滑处理的那种效果。</p>
<h4 id="knn">kNN</h4>
<p>kNN的全称是K-nearest neighbors，即<code>k最近邻</code>，这种算法类似于微区间平滑处理，但是kNN更适合于多维数据。总的来说，只要给定我们想要估计的任意点 <span class="math inline">\(x\)</span> ，我们会寻找k个最近的点，然后取这些点的平均值。这就会估计 <span class="math inline">\(f(x_1,x_2)\)</span> , 跟微区间平滑处理生成一个条曲线类似。我们现在通过 <span class="math inline">\(k\)</span> 来控制灵活性。这时我们比较一下 <span class="math inline">\(k=1\)</span> 和 <span class="math inline">\(k=100\)</span> 时的计算结果：</p>
<figure class="highlight plain"><figcaption><span>message</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">library(class)</div><div class="line">mypar(2,2)</div><div class="line">for(k in c(1,100))&#123;</div><div class="line">  ##predict on train</div><div class="line">  yhat &lt;- knn(x,x,y,k=k)</div><div class="line">  cat(&quot;KNN prediction error in train:&quot;,1-mean((as.numeric(yhat)-1)==y),&quot;\n&quot;)</div><div class="line">  ##make plot</div><div class="line">  yhat &lt;- knn(x,test,y,k=k)</div><div class="line">  cat(&quot;KNN prediction error in test:&quot;,1-mean((as.numeric(yhat)-1)==ytest),&quot;\n&quot;)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">&gt; mypar(<span class="number">2</span>,<span class="number">2</span>)</div><div class="line">&gt; <span class="keyword">for</span>(k <span class="keyword">in</span> c(<span class="number">1</span>,<span class="number">100</span>))&#123;</div><div class="line">+   <span class="comment">##predict on train</span></div><div class="line">+   yhat &lt;- knn(x,x,y,k=k)</div><div class="line">+   cat(<span class="string">"KNN prediction error in train:"</span>,<span class="number">1</span>-mean((as.numeric(yhat)-<span class="number">1</span>)==y),<span class="string">"\n"</span>)</div><div class="line">+   <span class="comment">##make plot</span></div><div class="line">+   yhat &lt;- knn(x,test,y,k=k)</div><div class="line">+   cat(<span class="string">"KNN prediction error in test:"</span>,<span class="number">1</span>-mean((as.numeric(yhat)-<span class="number">1</span>)==ytest),<span class="string">"\n"</span>)</div><div class="line">+ &#125;</div><div class="line">KNN prediction error <span class="keyword">in</span> train: <span class="number">0</span> </div><div class="line">KNN prediction error <span class="keyword">in</span> test: <span class="number">0.375</span> </div><div class="line">KNN prediction error <span class="keyword">in</span> train: <span class="number">0.2425</span> </div><div class="line">KNN prediction error <span class="keyword">in</span> test: <span class="number">0.2825</span></div></pre></td></tr></table></figure>
<p>为了说明，当我们设定 <span class="math inline">\(k=1\)</span>时，训练集中没有错误，以及 <span class="math inline">\(k=100\)</span> 时错误升高的原因，我们用图片直观地展示一下上面的结果，如下所示：</p>
<figure class="highlight plain"><figcaption><span>knn, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">library(class)</div><div class="line">mypar(2,2)</div><div class="line">for(k in c(1,100))&#123;</div><div class="line">  ##predict on train</div><div class="line">  yhat &lt;- knn(x,x,y,k=k)</div><div class="line">  ##make plot</div><div class="line">  yhat &lt;- knn(x,newx,y,k=k)</div><div class="line">  colshat &lt;- mycols[as.numeric(yhat)]</div><div class="line">  plot(x,type=&quot;n&quot;,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">  points(newx,col=colshat,cex=0.35,pch=16)</div><div class="line">  contour(tmpx,tmpy,matrix(as.numeric(yhat),GS,GS),levels=c(1,2),</div><div class="line">          add=TRUE,drawlabels=FALSE)</div><div class="line">  points(x,bg=cols,pch=21)</div><div class="line">  title(paste(&quot;Train: KNN (&quot;,k,&quot;)&quot;,sep=&quot;&quot;))</div><div class="line">  </div><div class="line">  plot(test,type=&quot;n&quot;,xlab=&quot;X1&quot;,ylab=&quot;X2&quot;,xlim=XLIM,ylim=YLIM)</div><div class="line">  points(newx,col=colshat,cex=0.35,pch=16)</div><div class="line">  contour(tmpx,tmpy,matrix(as.numeric(yhat),GS,GS),levels=c(1,2),</div><div class="line">          add=TRUE,drawlabels=FALSE)</div><div class="line">  points(test,bg=cols,pch=21)</div><div class="line">  title(paste(&quot;Test: KNN (&quot;,k,&quot;)&quot;,sep=&quot;&quot;))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190912174226.jpeg">

</div>
<p>从图上我们可以发现，当 <span class="math inline">\(k=1\)</span> 时，在训练集中并没有出现错误，因为每个点都是它最接近的点，这个就是它自身。但是，我们可以看到一些蓝色的岛（由几个点构成的区域）在红色区域中，一旦我们将数据集移向测试集，就会出现一些错误。当 <span class="math inline">\(k=100\)</span> 时，我们没有这个问题（也就是说红蓝区域分得很开），我们可以看到，错误率比线性回归有着明显的降低。我们还看到，我们估计的 <span class="math inline">\(f(x_1,x_2)\)</span> 比较接近于真实情况。</p>
<h4 id="贝叶斯规则">贝叶斯规则</h4>
<p>在这一部分里，我们会比较了不同 <span class="math inline">\(k\)</span> 值下的训练集与测试集。我们还会比较当我们知道 <span class="math inline">\(\mbox{E}(Y \mid X_1=x1,X_2=x_2)\)</span> 时的错误率，这也就是所谓的贝叶斯规则(Bayes Rule)。</p>
<p>我们先来计算一下错误率，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">library(class)</div><div class="line">###Bayes Rule</div><div class="line">yhat &lt;- apply(test,1,p)</div><div class="line">cat(&quot;Bayes rule prediction error in train&quot;,1-mean(round(yhat)==y),&quot;\n&quot;)</div><div class="line">bayes.error=1-mean(round(yhat)==y)</div><div class="line">train.error &lt;- rep(0,16)</div><div class="line">test.error &lt;- rep(0,16)</div><div class="line">for(k in seq(along=train.error))&#123;</div><div class="line">  ##predict on train</div><div class="line">  yhat &lt;- knn(x,x,y,k=2^(k/2))</div><div class="line">  train.error[k] &lt;- 1-mean((as.numeric(yhat)-1)==y)</div><div class="line">  ##prediction on test    </div><div class="line">  yhat &lt;- knn(x,test,y,k=2^(k/2))</div><div class="line">  test.error[k] &lt;- 1-mean((as.numeric(yhat)-1)==y)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>然后绘制出不同 <span class="math inline">\(k\)</span> 值下的错误率。我们还以一条水平线来展示贝叶斯规则错误率，如下所示</p>
<figure class="highlight plain"><figcaption><span>bayes_rule2, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">ks &lt;- 2^(seq(along=train.error)/2)</div><div class="line">mypar()</div><div class="line">plot(ks,train.error,type=&quot;n&quot;,xlab=&quot;K&quot;,ylab=&quot;Prediction Error&quot;,log=&quot;x&quot;,</div><div class="line">     ylim=range(c(test.error,train.error)))</div><div class="line">lines(ks,train.error,type=&quot;b&quot;,col=4,lty=2,lwd=2)</div><div class="line">lines(ks,test.error,type=&quot;b&quot;,col=5,lty=3,lwd=2)</div><div class="line">abline(h=bayes.error,col=6)</div><div class="line">legend(&quot;bottomright&quot;,c(&quot;Train&quot;,&quot;Test&quot;,&quot;Bayes&quot;),col=c(4,5,6),lty=c(2,3,1),box.lwd=0)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190913013135.jpeg">

</div>
<p>在上图中，粉色是训练集的错误率，绿色的是测试集的错误率。黄色是贝叶斯规则的错误率。</p>
<p>我们要知道，错误率是一个随机变量，它有着标准差。在下面的部分里，我们会提到交叉验证，这种方法有助于降低一些变异（这里我自己的理解就是错误率的变异）。然而即使将这些变异降低，从图中我们就可以看出来，当 <span class="math inline">\(k\)</span> 低于20时就会出现过拟合(over-fitting)，当 <span class="math inline">\(k\)</span> 超过100时就会出现低拟合(under-fitting)。</p>
<h2 id="交叉验证">交叉验证</h2>
<p>这里我们描述一下交叉验证(cross-validation)，交叉验证是机器学习中有关方法评估的一个基础工具，它能在一项预测或机器学习任务中进行参数选择。假设我们有一组许多特征值的观测值，并且每个观测值都与一个标签关联。我们将这个集合称为我们的训练集。我们的任务就是通过从训练数据中学习模式来预测任何新样本的标签。对于一个具体的例子来说，例如我们会将每个基因看作是一个特征值，然后我们再来计算一组没有标签的数据（测试数据集），看一下这组数据是新样本中的哪些组织类型。</p>
<p>如果我们选择了一个可调参数的机器学习算法，那么我们必须要有一个策略来这个参数选择一个最佳值。我们可以尝试着先计算一批数据，例如可以把一些已知样本的数据当作训练集，然后算法会计算出这些值产生的错误数据，然后我们会选择在我们的训练集中表现最好的值。</p>
<p>现在我们使用前面提到过的组织基要因表达数据集来看一下，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(tissuesGeneExpression)</div><div class="line">data(tissuesGeneExpression)</div></pre></td></tr></table></figure>
<p>为了说明我们这么做的目的，我们把那些样本数目较少的组织剔除掉，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">table(tissue)</div><div class="line">ind &lt;- which(tissue != &quot;placenta&quot;)</div><div class="line">y &lt;- tissue[ind]</div><div class="line">X &lt;- t( e[,ind] )</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; table(tissue)</div><div class="line">tissue</div><div class="line"> cerebellum       colon endometrium hippocampus      kidney       liver    placenta </div><div class="line">         <span class="number">38</span>          <span class="number">34</span>          <span class="number">15</span>          <span class="number">31</span>          <span class="number">39</span>          <span class="number">26</span>           <span class="number">6</span></div></pre></td></tr></table></figure>
<p>我们把<code>placenta(胎盘)</code>这个组织去掉了，现在我们使用kNN法来进行归类，先使用 <span class="math inline">\(k=5\)</span> 这个参数试一下。当我们把这个参数用于训练集和测试集时，我们在预测训练集中的组织时，平均误差是多少呢？</p>
<p>计算过程如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(class)</div><div class="line">pred &lt;- knn(train =  X, test = X, cl=y, k=5)</div><div class="line">mean(y != pred)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; mean(y != pred)</div><div class="line">[<span class="number">1</span>] <span class="number">0</span></div></pre></td></tr></table></figure>
<p>当我们使用 <span class="math inline">\(k=5\)</span> 这个参数时，没有发现错误，如果是 <span class="math inline">\(k=1\)</span> 呢？如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">pred &lt;- knn(train=X, test=X, cl=y, k=1)</div><div class="line">mean(y != pred)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; mean(y != pred)</div><div class="line">[<span class="number">1</span>] <span class="number">0</span></div></pre></td></tr></table></figure>
<p>当我们试图通过对观测值进行分类用于训练(train)模型时，这一过程可能有误导性。事实时，对于kNN法来说，使用 <span class="math inline">\(k=1\)</span> 这个参数时，总是能在训练集中得到0个分类错误，因为我们使用的是数据本身。了解算法是否可靠的方法就让它对未见过的样本进行预测。类似地，如果我们想知道可调参数的最佳值是什么，我们查看不同的参数值在那些不在训练集中的样本上表现如何。</p>
<p>交叉验证是机器学习中广泛使用的一种方法，它解决了训练集和测试集的问题，同时它仍然可以使用所有的数据用于检测预测的准确性。它通过将所有的数据分散成一定数据的折叠(fold)（注：这里有关交叉验证的问题，可以参考以前的文章<a href="http://rvdsd.top/2018/07/05/StatQuest/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B022-%E4%BA%A4%E5%8F%89%E9%AA%8C%E8%AF%81/">《StatQuest学习笔记22——交叉验证》</a>）。如果我们有 <span class="math inline">\(N\)</span> 个折叠，那么算法的第一步就是使用 <span class="math inline">\((N-1)\)</span> 个折叠来训练算法，并且剩下一个折叠用于测试算法的准确性。然后重复 <span class="math inline">\(N\)</span> 次，直到所有的折叠都在测试集中一样被使用。如果我们有M个参数需要进行尝试，那么我们需要在外部循环中完成这个过程，因为我们需要总共拟合 <span class="math inline">\(N \times M\)</span>次。</p>
<p>在R中，我们使用<code>caret</code>包中的<code>createFolds()</code>函数来实现这个过程,在下面的案例中，我们使用5个折叠来计算我们的基因表达数据，这个数字与组织的数目比较接近。此外，<code>createFold()</code>函数中有个参数<code>k</code>，这里不要与kNN算法中的<code>k</code>混淆，它们只是相同的字符，具体意义不一样，它们完全不相关。<code>createFolds()</code>会寻问用户要创建多少个折叠，也就是上文提到的 <span class="math inline">\(N\)</span> 。而<code>knn()</code>中的参数<code>k</code>则是说明在一个新的样本中，使用多少个最接近的观测值。现在我们创建10个折叠，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">library(caret)</div><div class="line">set.seed(1)</div><div class="line">idx &lt;- createFolds(y, k=10)</div><div class="line">sapply(idx, length)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; sapply(idx, length)</div><div class="line">Fold01 Fold02 Fold03 Fold04 Fold05 Fold06 Fold07 Fold08 Fold09 Fold10 </div><div class="line">    <span class="number">18</span>     <span class="number">19</span>     <span class="number">17</span>     <span class="number">17</span>     <span class="number">18</span>     <span class="number">20</span>     <span class="number">19</span>     <span class="number">19</span>     <span class="number">20</span>     <span class="number">16</span></div></pre></td></tr></table></figure>
<p>折叠会以数字索引列表的形式返回，因此数据的第一个折叠是：</p>
<p>The folds are returned as a list of numeric indices. The first fold of data is therefore:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">y[idx[[1]]] ##the labels</div><div class="line">head( X[idx[[1]], 1:3] ) ##the genes (only showing the first 3 genes...)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; y[idx[[<span class="number">1</span>]]] <span class="comment">##the labels</span></div><div class="line"> [<span class="number">1</span>] <span class="string">"kidney"</span>      <span class="string">"kidney"</span>      <span class="string">"hippocampus"</span> <span class="string">"hippocampus"</span> <span class="string">"hippocampus"</span> <span class="string">"cerebellum"</span> </div><div class="line"> [<span class="number">7</span>] <span class="string">"cerebellum"</span>  <span class="string">"cerebellum"</span>  <span class="string">"colon"</span>       <span class="string">"colon"</span>       <span class="string">"colon"</span>       <span class="string">"colon"</span>      </div><div class="line">[<span class="number">13</span>] <span class="string">"kidney"</span>      <span class="string">"kidney"</span>      <span class="string">"endometrium"</span> <span class="string">"endometrium"</span> <span class="string">"liver"</span>       <span class="string">"liver"</span>      </div><div class="line">&gt; head( X[idx[[<span class="number">1</span>]], <span class="number">1</span>:<span class="number">3</span>] ) <span class="comment">##the genes (only showing the first 3 genes...)</span></div><div class="line">                1007_s_at  1053_at   117_at</div><div class="line">GSM12075.CEL.gz  <span class="number">9.966782</span> <span class="number">6.060069</span> <span class="number">7.644452</span></div><div class="line">GSM12098.CEL.gz  <span class="number">9.945652</span> <span class="number">5.927861</span> <span class="number">7.847192</span></div><div class="line">GSM21214.cel.gz <span class="number">10.955428</span> <span class="number">5.776781</span> <span class="number">7.493743</span></div><div class="line">GSM21218.cel.gz <span class="number">10.757734</span> <span class="number">5.984170</span> <span class="number">8.525524</span></div><div class="line">GSM21230.cel.gz <span class="number">11.496114</span> <span class="number">5.760156</span> <span class="number">7.787561</span></div><div class="line">GSM87086.cel.gz  <span class="number">9.798633</span> <span class="number">5.862426</span> <span class="number">7.279199</span></div></pre></td></tr></table></figure>
<p>我们可以看到，事实上组织在10个折叠中表现非常平均：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sapply(idx, function(i) table(y[i]))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; sapply(idx, <span class="keyword">function</span>(i) table(y[i]))</div><div class="line">            Fold01 Fold02 Fold03 Fold04 Fold05 Fold06 Fold07 Fold08 Fold09 Fold10</div><div class="line">cerebellum       <span class="number">3</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">3</span></div><div class="line">colon            <span class="number">4</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">4</span>      <span class="number">3</span></div><div class="line">endometrium      <span class="number">2</span>      <span class="number">2</span>      <span class="number">1</span>      <span class="number">1</span>      <span class="number">1</span>      <span class="number">2</span>      <span class="number">1</span>      <span class="number">2</span>      <span class="number">2</span>      <span class="number">1</span></div><div class="line">hippocampus      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">4</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span></div><div class="line">kidney           <span class="number">4</span>      <span class="number">4</span>      <span class="number">3</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span>      <span class="number">4</span></div><div class="line">liver            <span class="number">2</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">2</span>      <span class="number">2</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">3</span>      <span class="number">2</span></div></pre></td></tr></table></figure>
<p>因为不同组织的表达谱不一样，因此使用所有的基因来预测组织非常容易。为了说明这种算法的原理，我们这里只使用二维数据来进行预测，现在我们使用<code>cmdscale()</code>函数进行降低处理：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar()</div><div class="line">Xsmall &lt;- cmdscale(dist(X))</div><div class="line">plot(Xsmall,col=as.fumeric(y))</div><div class="line">legend(<span class="string">"topleft"</span>,levels(factor(y)),fill=seq_along(levels(factor(y))))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190913180841.jpeg">

</div>
<p>现在我们在单个折叠上试一下kNN法。我们使用<code>Xsmall</code>中的样本（不用第1个样本），用<code>knn()</code>函数计算一下。我们使用<code>-idx[[1]]</code>移除第1个样本，这样将剩下来的样本当作测试集。参数<code>cl</code>用于指定真分类（true classificaiton）或者是训练集的标签（这里指的是组织）。现在我们使用5个观测值来为我们的kNN算法指定分类，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">pred &lt;- knn(train=Xsmall[ -idx[[1]] , ], test=Xsmall[ idx[[1]], ], cl=y[ -idx[[1]] ], k=5)</div><div class="line">table(true=y[ idx[[1]] ], pred)</div><div class="line">mean(y[ idx[[1]] ] != pred)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">&gt; table(true=y[ idx[[<span class="number">1</span>]] ], pred)</div><div class="line">             pred</div><div class="line">true          cerebellum colon endometrium hippocampus kidney liver</div><div class="line">  cerebellum           <span class="number">2</span>     <span class="number">0</span>           <span class="number">0</span>           <span class="number">1</span>      <span class="number">0</span>     <span class="number">0</span></div><div class="line">  colon                <span class="number">0</span>     <span class="number">4</span>           <span class="number">0</span>           <span class="number">0</span>      <span class="number">0</span>     <span class="number">0</span></div><div class="line">  endometrium          <span class="number">0</span>     <span class="number">0</span>           <span class="number">1</span>           <span class="number">0</span>      <span class="number">1</span>     <span class="number">0</span></div><div class="line">  hippocampus          <span class="number">1</span>     <span class="number">0</span>           <span class="number">0</span>           <span class="number">2</span>      <span class="number">0</span>     <span class="number">0</span></div><div class="line">  kidney               <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>           <span class="number">0</span>      <span class="number">4</span>     <span class="number">0</span></div><div class="line">  liver                <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>           <span class="number">0</span>      <span class="number">0</span>     <span class="number">2</span></div><div class="line">&gt; mean(y[ idx[[<span class="number">1</span>]] ] != pred)</div><div class="line">[<span class="number">1</span>] <span class="number">0.1666667</span></div></pre></td></tr></table></figure>
<p>现在我们出现了一些分类错误，我们计算一下剩余折叠，看一下情况怎么样？</p>
<p>Now we have some misclassifications. How well do we do for the rest of the folds?</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">for (i in 1:10) &#123;</div><div class="line">  pred &lt;- knn(train=Xsmall[ -idx[[i]] , ], test=Xsmall[ idx[[i]], ], cl=y[ -idx[[i]] ], k=5)</div><div class="line">  print(paste0(i,&quot;) error rate: &quot;, round(mean(y[ idx[[i]] ] != pred),3)))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:<span class="number">10</span>) &#123;</div><div class="line">+   pred &lt;- knn(train=Xsmall[ -idx[[i]] , ], test=Xsmall[ idx[[i]], ], cl=y[ -idx[[i]] ], k=<span class="number">5</span>)</div><div class="line">+   print(paste0(i,<span class="string">") error rate: "</span>, round(mean(y[ idx[[i]] ] != pred),<span class="number">3</span>)))</div><div class="line">+ &#125;</div><div class="line">[<span class="number">1</span>] <span class="string">"1) error rate: 0.111"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"2) error rate: 0.105"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"3) error rate: 0.118"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"4) error rate: 0.118"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"5) error rate: 0.278"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"6) error rate: 0.1"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"7) error rate: 0.105"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"8) error rate: 0.158"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"9) error rate: 0.15"</span></div><div class="line">[<span class="number">1</span>] <span class="string">"10) error rate: 0.312"</span></div></pre></td></tr></table></figure>
<p>所以，我们会看到每个折叠都会发生一些变化，其错误率徘徊在0.1-0.3之间。但是，<span class="math inline">\(k=5\)</span> 是最佳的参数吗？为了研究 <span class="math inline">\(k\)</span> 的最佳数值，我们需要创建一个外部循环，我们会在其中尝试不同的 <span class="math inline">\(k\)</span> 值，然后计算出所有折叠的平均测试集误差。我们将会尝试从1到12的每个 <span class="math inline">\(k\)</span> 值。不过这里我们不使用<code>for</code>循环，我们将使用<code>sapply</code>，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">set.seed(1)</div><div class="line">ks &lt;- 1:12</div><div class="line">res &lt;- sapply(ks, function(k) &#123;</div><div class="line">  ##try out each version of k from 1 to 12</div><div class="line">  res.k &lt;- sapply(seq_along(idx), function(i) &#123;</div><div class="line">    ##loop over each of the 10 cross-validation folds</div><div class="line">    ##predict the held-out samples using k nearest neighbors</div><div class="line">    pred &lt;- knn(train=Xsmall[ -idx[[i]], ],</div><div class="line">                test=Xsmall[ idx[[i]], ],</div><div class="line">                cl=y[ -idx[[i]] ], k = k)</div><div class="line">    ##the ratio of misclassified samples</div><div class="line">    mean(y[ idx[[i]] ] != pred)</div><div class="line">  &#125;)</div><div class="line">  ##average over the 10 folds</div><div class="line">  mean(res.k)</div><div class="line">&#125;)</div></pre></td></tr></table></figure>
<p>对于每个 <span class="math inline">\(k\)</span> 值，我们都有一个来自于交叉验证的关联测试集错误率，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">res</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; res</div><div class="line"> [<span class="number">1</span>] <span class="number">0.1978212</span> <span class="number">0.1703423</span> <span class="number">0.1882933</span> <span class="number">0.1750989</span> <span class="number">0.1613291</span> <span class="number">0.1500791</span> <span class="number">0.1552670</span> <span class="number">0.1884813</span></div><div class="line"> [<span class="number">9</span>] <span class="number">0.1822020</span> <span class="number">0.1763197</span> <span class="number">0.1761318</span> <span class="number">0.1813197</span></div></pre></td></tr></table></figure>
<p>我们可以绘制出每个 <span class="math inline">\(k\)</span> 值下的错误率图，它可以帮助我们查看哪个区域中的错误率最小，如下所示：</p>
<figure class="highlight plain"><figcaption><span>misclassification_error, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">plot(ks, res, type=&quot;o&quot;,ylab=&quot;misclassification error&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190913184146.jpeg">

</div>
<p>我们要记住一点就是，训练集是一个随机变量，因为我们生成折叠的程序涉随机数的生成，因此，在产生“最好”的 <span class="math inline">\(k\)</span> 值时，这个 <span class="math inline">\(k\)</span> 值的产生过程也是一个随机变量。如果我们有新的训练集，并且如果我们再创建折叠，那么我们有可能会得到一个新的优化的 <span class="math inline">\(k\)</span> 值。</p>
<p>最后，为了说明基因表达数据能够很好的预测组织类型，我们使用5维数据，而非是2维数据来展示这个过程，毕竟5维数据的信息量更丰富，如下所示：</p>
<figure class="highlight plain"><figcaption><span>misclassification_error2, fig.cap</span></figcaption><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">Xsmall &lt;- cmdscale(dist(X),k=5)</div><div class="line">set.seed(1)</div><div class="line">ks &lt;- 1:12</div><div class="line">res &lt;- sapply(ks, function(k) &#123;</div><div class="line">  res.k &lt;- sapply(seq_along(idx), function(i) &#123;</div><div class="line">    pred &lt;- knn(train=Xsmall[ -idx[[i]], ],</div><div class="line">                test=Xsmall[ idx[[i]], ],</div><div class="line">                cl=y[ -idx[[i]] ], k = k)</div><div class="line">    mean(y[ idx[[i]] ] != pred)</div><div class="line">  &#125;)</div><div class="line">  mean(res.k)</div><div class="line">&#125;)</div><div class="line">plot(ks, res, type=&quot;o&quot;,ylim=c(0,0.20),ylab=&quot;misclassification error&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190913184233.jpeg">

</div>
<p>上图显示的是，使用5维数据后，错误分配与 <span class="math inline">\(k\)</span> 值变化的关系。</p>
<p>重要提示：我们使用<code>cmdscale()</code>函数来计算整个数据集，用于创建一个较小的数据集来说明kNN的计算过程。然而在真正的机器学习应用中，这样的处理会低估小样本测试数据集的错误，但是，使用未标记的完全数据集进行降维会改善这种情况。一个种更安全的做法就是为每个折叠分别转换数据（这句不懂， 我个人的猜测就是，对每个折叠进行计算，把几个折叠中的每一个都当作测试集，而不是将几个折叠放一块当作测试集），方法就是仅用训练集来计算旋转和降维，并将其应用于测试集。</p>
<p>最后一段实在没读懂，这里放原文：</p>
<blockquote>
<p>However, in a real machine learning application, this may result in an underestimation of test set error for small sample sizes, where dimension reduction using the unlabeled full dataset gives a boost in performance. A safer choice would have been to transform the data separately for each fold, by calculating a rotation and dimension reduction using the training set only and applying this to the test set.</p>
</blockquote>
<h2 id="练习">练习</h2>
<p>P407</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/2/"><i class="fa fa-angle-left"></i></a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/4/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">243</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">117</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共963.7k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
