<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Hexo, NexT" />










<meta property="og:type" content="website">
<meta property="og:title" content="RVDSD的个人笔记本">
<meta property="og:url" content="http://rvdsd.top/page/3/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RVDSD的个人笔记本">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/page/3/"/>





  <title>RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">学习过程中的输出</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/09/03/Tools/Venn图/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/09/03/Tools/Venn图/" itemprop="url">Venn图</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-09-03T12:00:00+08:00">
                2018-09-03
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  58
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="在线维恩图工具">在线维恩图工具</h2>
<ol style="list-style-type: decimal">
<li>Bioinformatics $ Evolutionary Genomics</li>
</ol>
<p>http://bioinformatics.psb.ugent.be/webtools/Venn/</p>
<ol start="2" style="list-style-type: decimal">
<li>VENNY2.1</li>
</ol>
<p>http://bioinfogp.cnb.csic.es/tools/venny/index.html</p>
<ol start="3" style="list-style-type: decimal">
<li>Hello jvenn</li>
</ol>
<p>http://jvenn.toulouse.inra.fr/app/example.html</p>
<ol start="4" style="list-style-type: decimal">
<li>Gene List Venn Diagram</li>
</ol>
<p>http://genevenn.sourceforge.net/</p>
<ol start="5" style="list-style-type: decimal">
<li>BioVeen</li>
</ol>
<p>http://www.biovenn.nl/index.php</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/30/R/20个R语言习题/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/30/R/20个R语言习题/" itemprop="url">20个R语言习题</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-30T12:00:00+08:00">
                2018-08-30
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/R/" itemprop="url" rel="index">
                    <span itemprop="name">R</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,748
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  26
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="原题目">原题目</h2>
<p><a href="http://www.bio-info-trainee.com/3409.html" target="_blank" rel="external">生信人的20个R语言习题。</a></p>
<h2 id="题目代码">题目代码</h2>
<h3 id="安装r语言包">1. 安装R语言包</h3>
<p>安装以下包：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">数据包： ALL, CLL, pasilla, airway </div><div class="line">软件包：limma，DESeq2，clusterProfiler </div><div class="line">工具包：reshape2</div><div class="line">绘图包：ggplot2</div></pre></td></tr></table></figure>
<p>代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">source</span>(<span class="string">"http://bioconductor.org/biocLite.R"</span>)</div><div class="line">biocLite(<span class="string">"ALL"</span>)</div><div class="line">biocLite(<span class="string">"CLL"</span>)</div><div class="line">biocLite(<span class="string">"pasilla"</span>)</div><div class="line">biocLite(<span class="string">"airway"</span>)</div><div class="line">biocLite(<span class="string">"limma"</span>)</div><div class="line">biocLite(<span class="string">"DESeq2"</span>)</div><div class="line">biocLite(<span class="string">"clusterProfiler "</span>)</div><div class="line">biocLite(<span class="string">"airway"</span>)</div><div class="line"></div><div class="line"><span class="keyword">library</span>(ALL)</div><div class="line"><span class="keyword">library</span>(CLL)</div><div class="line"><span class="keyword">library</span>(pasilla)</div><div class="line"><span class="keyword">library</span>(airway)</div><div class="line"><span class="keyword">library</span>(DESeq2)</div><div class="line"><span class="keyword">library</span>(clusterProfiler )</div><div class="line"><span class="keyword">library</span>(reshape2)</div><div class="line"><span class="keyword">library</span>(ggplot2)</div></pre></td></tr></table></figure>
<h3 id="expressionset对象">2. ExpressionSet对象</h3>
<h1 id="注补充expressionset知识">注：补充ExpressionSet知识</h1>
<p>CLL包中里面有一个<code>sCLLex</code>数据，它是ExpressionSet格式的，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">data(<span class="string">"sCLLex"</span>)</div><div class="line">class(sCLLex)</div><div class="line"><span class="comment"># [1] "ExpressionSet"</span></div><div class="line"><span class="comment"># attr(,"package")</span></div><div class="line"><span class="comment"># [1] "Biobase"</span></div></pre></td></tr></table></figure>
<p>现在查看一下这个对象，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; sCLLex</div><div class="line">ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">12625</span> features, <span class="number">22</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: CLL11.CEL CLL12.CEL <span class="keyword">...</span> CLL9.CEL (<span class="number">22</span> total)</div><div class="line">  varLabels: SampleID Disease</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData: none</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: hgu95av2</div></pre></td></tr></table></figure>
<p>看一下这个数据的文档，可以得到这些信息：</p>
<p>这个数据集中有22个样本，12625个基因；使用<code>exprs</code>可以查看这个数据集的表达水平；使用<code>phenoData</code>可以查看这个数据集的分组信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">data_expression &lt;- exprs(sCLLex) </div><div class="line"><span class="comment">#提取出来表达矩阵，赋给data_expression</span></div><div class="line"></div><div class="line">head(data_expression)</div><div class="line"><span class="comment"># CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL CLL15.CEL CLL16.CEL CLL17.CEL</span></div><div class="line"><span class="comment"># 1000_at    5.743132  6.219412  5.523328  5.340477  5.229904  4.920686  5.325348</span></div><div class="line"><span class="comment"># 1001_at    2.285143  2.291229  2.287986  2.295313  2.662170  2.278040  2.350796</span></div><div class="line"><span class="keyword">...</span></div><div class="line"></div><div class="line">phenoData(sCLLex)</div><div class="line"><span class="comment"># 查看分组信息</span></div><div class="line"><span class="comment"># An object of class 'AnnotatedDataFrame'</span></div><div class="line"><span class="comment"># sampleNames: CLL11.CEL CLL12.CEL ... CLL9.CEL (22 total)</span></div><div class="line"><span class="comment"># varLabels: SampleID Disease</span></div><div class="line"><span class="comment"># varMetadata: labelDescription</span></div></pre></td></tr></table></figure>
<p>使用<code>View(data_expression)</code>可以直观看出来这个表达矩阵，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180905/hacL6JB59C.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>再来查看一个<code>sCCLex</code>这个数据集的其它信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div></pre></td><td class="code"><pre><div class="line">sampleNames(sCLLex)</div><div class="line"><span class="comment">#查看样本编号</span></div><div class="line"><span class="comment"># [1] "CLL11.CEL" "CLL12.CEL" "CLL13.CEL" "CLL14.CEL" "CLL15.CEL" "CLL16.CEL"</span></div><div class="line"><span class="comment"># [7] "CLL17.CEL" "CLL18.CEL" "CLL19.CEL" "CLL20.CEL" "CLL21.CEL" "CLL22.CEL"</span></div><div class="line"><span class="comment"># [13] "CLL23.CEL" "CLL24.CEL" "CLL2.CEL"  "CLL3.CEL"  "CLL4.CEL"  "CLL5.CEL" </span></div><div class="line"><span class="comment"># [19] "CLL6.CEL"  "CLL7.CEL"  "CLL8.CEL"  "CLL9.CEL" </span></div><div class="line"></div><div class="line">varMetadata(sCLLex) </div><div class="line"><span class="comment">#查看所有表型变量</span></div><div class="line"><span class="comment"># labelDescription</span></div><div class="line"><span class="comment"># SampleID          Sample ID</span></div><div class="line"><span class="comment"># Disease  Stable/Progressive</span></div><div class="line"></div><div class="line">data_phenotype=pData(sCLLex) </div><div class="line"><span class="comment">#提取表型信息</span></div><div class="line"><span class="comment"># data_phenotype</span></div><div class="line"><span class="comment"># SampleID  Disease</span></div><div class="line"><span class="comment"># CLL11.CEL    CLL11 progres.</span></div><div class="line"><span class="comment"># CLL12.CEL    CLL12   stable</span></div><div class="line"><span class="comment"># CLL13.CEL    CLL13 progres.</span></div><div class="line"><span class="comment"># CLL14.CEL    CLL14 progres.</span></div><div class="line"><span class="comment"># CLL15.CEL    CLL15 progres.</span></div><div class="line"><span class="comment"># CLL16.CEL    CLL16 progres.</span></div><div class="line"><span class="comment"># CLL17.CEL    CLL17   stable</span></div><div class="line"><span class="comment"># CLL18.CEL    CLL18   stable</span></div><div class="line"><span class="comment"># CLL19.CEL    CLL19 progres.</span></div><div class="line"><span class="comment"># CLL20.CEL    CLL20   stable</span></div><div class="line"><span class="comment"># CLL21.CEL    CLL21 progres.</span></div><div class="line"><span class="comment"># CLL22.CEL    CLL22   stable</span></div><div class="line"><span class="comment"># CLL23.CEL    CLL23 progres.</span></div><div class="line"><span class="comment"># CLL24.CEL    CLL24   stable</span></div><div class="line"><span class="comment"># CLL2.CEL      CLL2   stable</span></div><div class="line"><span class="comment"># CLL3.CEL      CLL3 progres.</span></div><div class="line"><span class="comment"># CLL4.CEL      CLL4 progres.</span></div><div class="line"><span class="comment"># CLL5.CEL      CLL5 progres.</span></div><div class="line"><span class="comment"># CLL6.CEL      CLL6 progres.</span></div><div class="line"><span class="comment"># CLL7.CEL      CLL7 progres.</span></div><div class="line"><span class="comment"># CLL8.CEL      CLL8 progres.</span></div><div class="line"><span class="comment"># CLL9.CEL      CLL9   stable</span></div><div class="line"></div><div class="line">featureNames(sCLLex)[<span class="number">1</span>:<span class="number">10</span>] </div><div class="line"><span class="comment">#查看基因芯片编码</span></div><div class="line"><span class="comment"># [1] "1000_at"   "1001_at"   "1002_f_at" "1003_s_at" "1004_at"   "1005_at"  </span></div><div class="line"><span class="comment"># [7] "1006_at"   "1007_s_at" "1008_f_at" "1009_at" </span></div><div class="line"></div><div class="line">featureNames(sCLLex) %&gt;% unique() %&gt;% length() </div><div class="line"><span class="comment"># 查看是否有重复的编码</span></div><div class="line"><span class="comment"># [1] 12625</span></div></pre></td></tr></table></figure>
<p>上面的<code>%&gt;%</code>这个符号相当于Linux中的管道命令。</p>
<p>接着，提取表型信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">pdata &lt;- pData(sCLLex)</div><div class="line">head(pdata)</div><div class="line"><span class="comment"># SampleID  Disease</span></div><div class="line"><span class="comment"># CLL11.CEL    CLL11 progres.</span></div><div class="line"><span class="comment"># CLL12.CEL    CLL12   stable</span></div><div class="line"><span class="comment"># CLL13.CEL    CLL13 progres.</span></div><div class="line"><span class="comment"># CLL14.CEL    CLL14 progres.</span></div><div class="line"><span class="comment"># CLL15.CEL    CLL15 progres.</span></div><div class="line"><span class="comment"># CLL16.CEL    CLL16 progres.</span></div><div class="line"></div><div class="line">group_list=as.character(pdata$Disease) </div><div class="line"><span class="comment">#从数据框中只要表型信息</span></div><div class="line"></div><div class="line">table(group_list) </div><div class="line"><span class="comment">#统计表型信息</span></div><div class="line"><span class="comment"># group_list</span></div><div class="line"><span class="comment"># progres.   stable </span></div><div class="line"><span class="comment"># 14        8</span></div></pre></td></tr></table></figure>
<p>绘制芯片数据的质量值，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">y &lt;- melt(as.data.frame(data_expression))</div><div class="line">p &lt;- ggplot(data=y,aes(x=variable,y=value))</div><div class="line">p &lt;- p + geom_boxplot(aes(fill=variable))</div><div class="line">p &lt;- p + theme(axis.text.x = element_text(angle=<span class="number">45</span>,hjust=<span class="number">1</span>,vjust=<span class="number">1</span>))</div><div class="line">p &lt;- p + xlab(<span class="string">"分组信息"</span>) + ylab(<span class="string">"表达值"</span>) + guides(fill=<span class="literal">FALSE</span>)</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190302161815.png">

</div>
<h3 id="了解strheadhelp函数">3.了解str,head,help函数</h3>
<p>略，第2题中已经有了。</p>
<h3 id="了解并安装hgu95av2.db">4.了解并安装hgu95av2.db</h3>
<h1 id="注补充db数据格式的相关知识">注：补充db数据格式的相关知识</h1>
<p>hgu95av2.db是一个注释包，它为hgu95av2平台的芯片提供注释，这个包中有很多注释文件，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">ls(<span class="string">"package:hgu95av2.db"</span>)</div><div class="line"><span class="comment"># [1] "hgu95av2"              "hgu95av2.db"           "hgu95av2_dbconn"      </span></div><div class="line"><span class="comment"># [4] "hgu95av2_dbfile"       "hgu95av2_dbInfo"       "hgu95av2_dbschema"    </span></div><div class="line"><span class="comment"># [7] "hgu95av2ACCNUM"        "hgu95av2ALIAS2PROBE"   "hgu95av2CHR"          </span></div><div class="line"><span class="comment"># [10] "hgu95av2CHRLENGTHS"    "hgu95av2CHRLOC"        "hgu95av2CHRLOCEND"    </span></div><div class="line"><span class="comment"># [13] "hgu95av2ENSEMBL"       "hgu95av2ENSEMBL2PROBE" "hgu95av2ENTREZID"     </span></div><div class="line"><span class="comment"># [16] "hgu95av2ENZYME"        "hgu95av2ENZYME2PROBE"  "hgu95av2GENENAME"     </span></div><div class="line"><span class="comment"># [19] "hgu95av2GO"            "hgu95av2GO2ALLPROBES"  "hgu95av2GO2PROBE"     </span></div><div class="line"><span class="comment"># [22] "hgu95av2MAP"           "hgu95av2MAPCOUNTS"     "hgu95av2OMIM"         </span></div><div class="line"><span class="comment"># [25] "hgu95av2ORGANISM"      "hgu95av2ORGPKG"        "hgu95av2PATH"         </span></div><div class="line"><span class="comment"># [28] "hgu95av2PATH2PROBE"    "hgu95av2PFAM"          "hgu95av2PMID"         </span></div><div class="line"><span class="comment"># [31] "hgu95av2PMID2PROBE"    "hgu95av2PROSITE"       "hgu95av2REFSEQ"       </span></div><div class="line"><span class="comment"># [34] "hgu95av2SYMBOL"        "hgu95av2UNIGENE"       "hgu95av2UNIPROT"</span></div></pre></td></tr></table></figure>
<p>安装hgu95av2.db，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">biocLite(<span class="string">"hgu95av2.db"</span>)</div><div class="line"><span class="keyword">library</span>(hgu95av2.db)</div></pre></td></tr></table></figure>
<h3 id="理解totable的用法">5.理解toTable的用法</h3>
<p>原是描述：理解 head(toTable(hgu95av2SYMBOL)) 的用法，找到 TP53 基因对应的探针ID。</p>
<p>先看一下<code>hgu95av2SYMBOL</code>是什么，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">?hgu95av2SYMBOL</div><div class="line"><span class="comment"># hgu95av2SYMBOL is an R object that provides mappings between manufacturer identifiers and gene abbreviations.</span></div></pre></td></tr></table></figure>
<p>大致意思就是说，hgu95av2SYMBOL是一个R对象，它描述厂家的标记符(identifiers)与基因缩写之间的映射关系。</p>
<p>再看一下<code>toTable</code>的用法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">?toTable</div><div class="line">These methods are part of the Bimap interface (see ?Bimap <span class="keyword">for</span> a quick overview of the Bimap objects and their interface).</div></pre></td></tr></table></figure>
<p>大致意思是说，toTable是一种能够以数据框的形式来操作一个Bimap对象的方法，也就是把Bimap对象转换为一个数据框，这些方法是Bimap interface方法的一部分。Bimap指的是一种映射关系，例如探针的编号与基因名称之间的映射，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div></pre></td><td class="code"><pre><div class="line">summary(hgu95av2SYMBOL)</div><div class="line"><span class="comment"># 查看hgu95av2SYMBOL大致信息</span></div><div class="line"><span class="comment"># SYMBOL map for chip hgu95av2 (object of class "ProbeAnnDbBimap")</span></div><div class="line"><span class="comment"># |</span></div><div class="line"><span class="comment">#   | Lkeyname: probe_id (Ltablename: probes)</span></div><div class="line"><span class="comment"># |    Lkeys: "1000_at", "1001_at", ... (total=12625/mapped=11471)</span></div><div class="line"><span class="comment"># |</span></div><div class="line"><span class="comment">#   | Rkeyname: symbol (Rtablename: gene_info)</span></div><div class="line"><span class="comment"># |    Rkeys: "A1BG", "A2M", ... (total=59682/mapped=8595)</span></div><div class="line"><span class="comment"># |</span></div><div class="line"><span class="comment">#   | direction: L --&gt; R</span></div><div class="line"></div><div class="line">gene_id &lt;- toTable(hgu95av2SYMBOL)</div><div class="line">head(gene_id) <span class="comment">#这是一个数据框</span></div><div class="line"><span class="comment"># probe_id  symbol</span></div><div class="line"><span class="comment"># 1   1000_at   MAPK3</span></div><div class="line"><span class="comment"># 2   1001_at    TIE1</span></div><div class="line"><span class="comment"># 3 1002_f_at CYP2C19</span></div><div class="line"><span class="comment"># 4 1003_s_at   CXCR5</span></div><div class="line"><span class="comment"># 5   1004_at   CXCR5</span></div><div class="line"><span class="comment"># 6   1005_at   DUSP1</span></div><div class="line">str(gene_id)</div><div class="line"><span class="comment"># 'data.frame':	11471 obs. of  2 variables:</span></div><div class="line"><span class="comment">#   $ probe_id: chr  "1000_at" "1001_at" "1002_f_at" "1003_s_at" ...</span></div><div class="line"><span class="comment"># $ symbol  : chr  "MAPK3" "TIE1" "CYP2C19" "CXCR5" ...</span></div><div class="line"></div><div class="line">filter(gene_id,symbol==<span class="string">"TP53"</span>)</div><div class="line"><span class="comment"># 找到 TP53 基因对应的探针ID</span></div><div class="line"><span class="comment"># probe_id symbol</span></div><div class="line"><span class="comment"># 1   1939_at   TP53</span></div><div class="line"><span class="comment"># 2 1974_s_at   TP53</span></div><div class="line"><span class="comment"># 3  31618_at   TP53</span></div><div class="line"></div><div class="line"><span class="keyword">library</span>(dplyr)</div><div class="line">filter(gene_id,symbol==<span class="string">"TP53"</span>)</div><div class="line"><span class="comment"># 找到 TP53 基因对应的探针ID</span></div><div class="line"><span class="comment"># probe_id symbol</span></div><div class="line"><span class="comment"># 1   1939_at   TP53</span></div><div class="line"><span class="comment"># 2 1974_s_at   TP53</span></div><div class="line"><span class="comment"># 3  31618_at   TP53</span></div></pre></td></tr></table></figure>
<h3 id="探针与基因的关系">6.探针与基因的关系</h3>
<p>原题目描述：</p>
<blockquote>
<p>理解探针与基因的对应关系，总共多少个基因，基因最多对应多少个探针，是哪些基因，是不是因为这些基因很长，所以在其上面设计多个探针呢？</p>
</blockquote>
<p>答：不管是Agilent芯片，还是Affymetrix芯片，上面设计的探针都非常短。最长的如Agilent芯片上的探针，往往都是60bp，但是往往一个基因的长度都好几Kb。因此一般多个探针对应一个基因，取最大表达值探针来作为基因的表达量。</p>
<h3 id="提取目的基因">7.提取目的基因</h3>
<p>原题目描述：</p>
<blockquote>
<p>第二步提取到的表达矩阵是12625个探针在22个样本的表达量矩阵，找到那些不在 <code>hgu95av2.db</code> 包收录的对应着SYMBOL的探针。</p>
</blockquote>
<p>芯片表达的数据存放在<code>data_expression</code>这个变量中，它有22个样本的12625个探针数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">class(data_expression)</div><div class="line"><span class="comment"># [1] "matrix"</span></div><div class="line">str(data_expression)</div><div class="line"><span class="comment"># num [1:12625, 1:22] 5.74 2.29 3.31 1.09 7.54 ...</span></div><div class="line"><span class="comment"># - attr(*, "dimnames")=List of 2</span></div><div class="line"><span class="comment"># ..$ : chr [1:12625] "1000_at" "1001_at" "1002_f_at" "1003_s_at" ...</span></div><div class="line"><span class="comment"># ..$ : chr [1:22] "CLL11.CEL" "CLL12.CEL" "CLL13.CEL" "CLL14.CEL" ...</span></div></pre></td></tr></table></figure>
<p>现在要解决的一个问题是：在<code>data_expression</code>中有一些基因没有收录在<code>hug95av2.db</code>包中，现在要找到这些基因，把它们都去掉。</p>
<p>思路：找到<code>hug95av2.db</code>包中能够与基因映射起来的探针，再找到这些探针对应的基因SYMBOL，然后在<code>data_expression</code>中取出这些基因即可。此时要使用到<code>hug95av2.db</code>包中的<code>hug95av2SYMBOL</code>这个文件，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">?hug95av2SYMBOL</div></pre></td></tr></table></figure>
<p>查阅其文档，了解到如下信息：hug95av2SYMBOL是一个R对象，它提供的是芯片生产厂家与基因缩写之间的映射信息。这个映射的信息主要依据Entrez Gene数据库。现在我们通过<code>mappedkeys()</code>这个函数，得到映射到基因上的探针信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">probe_map &lt;- hgu95av2SYMBOL</div><div class="line">length(probe_map) </div><div class="line"><span class="comment">#全部的探针数目</span></div><div class="line"><span class="comment"># [1] 12625</span></div><div class="line">probe_info &lt;- mappedkeys(probe_map)</div><div class="line">length(probe_info) </div><div class="line"><span class="comment">#探针与基因产生映射的数目</span></div><div class="line"><span class="comment"># [1] 11471</span></div><div class="line">gene_info &lt;- as.list(probe_map[probe_info])</div><div class="line"><span class="comment"># 转化为数据表</span></div><div class="line">length(gene_info)</div><div class="line"><span class="comment"># [1] 11471</span></div><div class="line">gene_symbol &lt;- toTable(probe_map[probe_info])</div><div class="line"><span class="comment"># 从hgu95av2SYMBOL文件中，取出有映射关系的探针，并生成数据框给gene_symbol</span></div><div class="line">head(gene_symbol)</div><div class="line"><span class="comment"># probe_id  symbol</span></div><div class="line"><span class="comment"># 1   1000_at   MAPK3</span></div><div class="line"><span class="comment"># 2   1001_at    TIE1</span></div><div class="line"><span class="comment"># 3 1002_f_at CYP2C19</span></div><div class="line"><span class="comment"># 4 1003_s_at   CXCR5</span></div><div class="line"><span class="comment"># 5   1004_at   CXCR5</span></div><div class="line"><span class="comment"># 6   1005_at   DUSP1</span></div></pre></td></tr></table></figure>
<p>其中，gene_symbol就是有映射关系的基因，它里面的数据是探针的编码以及探针对应的基因，在data_expression中它们挑出来即可。这里使用到了<code>mappedkeys()</code>函数，这个函数用于处理映射文件(bimap)，它的参数就是一个Bimap对象，例如上面的<code>hgu95av2SYMBOL</code>这个对象，关于Bimap对象，可以看这篇笔记《Bimap对象笔记》。</p>
<h3 id="过滤表达矩阵">8. 过滤表达矩阵</h3>
<p>原问题描述：过滤表达矩阵，删除那1165个没有对应基因名字的探针。这时再看一下原始数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">head(data_expression)</div><div class="line"><span class="comment"># CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL CLL15.CEL CLL16.CEL CLL17.CEL</span></div><div class="line"><span class="comment"># 1000_at    5.743132  6.219412  5.523328  5.340477  5.229904  4.920686  5.325348</span></div><div class="line"><span class="comment"># 1001_at    2.285143  2.291229  2.287986  2.295313  2.662170  2.278040  2.350796</span></div><div class="line"><span class="comment"># 1002_f_at  3.309294  3.318466  3.354423  3.327130  3.365113  3.568353  3.502440</span></div><div class="line"><span class="comment"># 1003_s_at  1.085264  1.117288  1.084010  1.103217  1.074243  1.073097  1.091264</span></div><div class="line"><span class="comment"># 1004_at    7.544884  7.671801  7.474025  7.152482  6.902932  7.368660  6.456285</span></div><div class="line"><span class="comment"># 1005_at    5.083793  7.610593  7.631311  6.518594  5.059087  4.855161  5.176975</span></div></pre></td></tr></table></figure>
<p>思路：在这个原始数据中，有12625个探针，下面通过<code>isNA</code>函数可以得到一个逻辑向量，其中<code>TRUE</code>表示没有映射关系的探针，<code>FALSE</code>表示有映射关系的探针。然后使用<code>seq</code>生成一个整数向量，并提取那些有映射关系的探针所在的列，提取原始数据即可，如下所示：data_filter &lt;- data_expression[rownames(data_expression)%in%probe_info]</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">data_filter &lt;- data_expression[rownames(data_expression)%<span class="keyword">in</span>%probe_info,]</div><div class="line"><span class="comment"># rownames(data_expression)是原始数据表达值的名称，这里提探针的编号</span></div><div class="line"><span class="comment"># probe_info是有映射的探针的编号</span></div><div class="line"><span class="comment"># A%in%B，表示A中有哪些成员在B中，有的返回TRUE，无的返回FALSE</span></div><div class="line"><span class="comment"># 这行命令表示，提取那些在probe_info中的探针的表达值，赋值给变量data_filter</span></div><div class="line"></div><div class="line">table(rownames(data_expression) %<span class="keyword">in</span>% probe_info)</div><div class="line"><span class="comment"># FALSE  TRUE </span></div><div class="line"><span class="comment"># 1154 11471 </span></div><div class="line"><span class="comment"># 从上面的数据可以看出来，原始表达值中有11471个数据在probe_info中</span></div><div class="line"></div><div class="line">table(rownames(data_expression) %<span class="keyword">in</span>% rownames(data_filter))</div><div class="line"><span class="comment"># FALSE  TRUE </span></div><div class="line"><span class="comment"># 1154 11471</span></div></pre></td></tr></table></figure>
<p>现在就得了过滤后的表达矩阵，即<code>data_filter</code>。</p>
<h3 id="整合表达矩阵">9.整合表达矩阵</h3>
<p>原问题描述：整合表达矩阵，多个探针对应一个基因的情况下，只保留在所有样本里面平均表达量最大的那个探针。</p>
<p>思路：基因芯片的的多个探针会针对同一个基因，因此会有重复，先看一下表达矩阵<code>data_filter</code>的样子：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/A5iBlKK20h.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>其中列是不同的样本，行是不同的探针，此时我们解决问题的思路应该是这个样子的：第一，把表达矩阵中的探针名称转换为基因名称；第二，计算每行的均值；第三，在相同基因的表达值中，取最大的那行。</p>
<p>第一步，要使用到的数据是<code>gene_symbol</code>，它的内容是，去掉了无映射关系的探针编号后，剩余的探针与相应基因的映射关系，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">head(gene_symbol)</div><div class="line"><span class="comment"># probe_id  symbol</span></div><div class="line"><span class="comment"># 1   1000_at   MAPK3</span></div><div class="line"><span class="comment"># 2   1001_at    TIE1</span></div><div class="line"><span class="comment"># 3 1002_f_at CYP2C19</span></div><div class="line"><span class="comment"># 4 1003_s_at   CXCR5</span></div><div class="line"><span class="comment"># 5   1004_at   CXCR5</span></div><div class="line"><span class="comment"># 6   1005_at   DUSP1</span></div></pre></td></tr></table></figure>
<p>现在将表达矩阵<code>data_filter</code>中的探针名称排序，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">data_filter_order &lt;- data_filter[order(rownames(data_filter)),]</div><div class="line"><span class="comment"># 对过滤后的表达值的数据根据探针名称进行排序</span></div><div class="line">gene_symbol_order &lt;- gene_symbol[order(gene_symbol$probe_id),]</div><div class="line"><span class="comment"># 对gene_symbol按照探针名称进行排序</span></div><div class="line">table(gene_symbol_order$probe_id == rownames(data_filter_order))</div><div class="line"><span class="comment"># 查看地表达值的矩阵的名称是否与gene_symbol中的探针名称是否一一对应</span></div><div class="line"><span class="comment"># TRUE </span></div><div class="line"><span class="comment"># 1147</span></div></pre></td></tr></table></figure>
<p>排序的目的在于，将探针与基因名一一对应起来，排序后的表达矩阵如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/hHGk50Hhcb.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第二步，计算每行的均值，使用apply函数即可，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">row_mean &lt;- apply(data_filter_order,<span class="number">1</span>,mean)</div><div class="line">head(row_mean)</div><div class="line"><span class="comment">#  100_g_at   1000_at   1001_at 1002_f_at 1003_s_at   1004_at </span></div><div class="line"><span class="comment">#  2.289762  5.456561  2.334098  3.402072  1.126168  7.439034</span></div></pre></td></tr></table></figure>
<p>第三步，保留相同基因中表达值最高的那一行，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">index_probe &lt;- <span class="keyword">function</span>(x)&#123;</div><div class="line">  names(x)[which.max(x)]</div><div class="line">&#125;</div><div class="line"><span class="comment"># 构建函数</span></div><div class="line"></div><div class="line">max_index &lt;- tapply(row_mean,gene_symbol$symbol,index_probe)</div><div class="line"><span class="comment"># index_probe这个函数的功能在于，用gene_symbol$symol（基因名）对row_mean进行分组，然后挑出相同基因中表达值最大的那个值所对应的探针编号</span></div><div class="line"></div><div class="line">data_filter_order_unique &lt;- data_filter_order[(gene_symbol_order$probe_id %<span class="keyword">in</span>% max_index),]</div><div class="line"><span class="comment"># gene_symbol_order与data_filter_order都按照探针名称进行排序，再挑出max_index中的探针哪些在gene_symbol_order$probe_id中</span></div><div class="line"></div><div class="line">unique_map_gene_probe &lt;- gene_symbol_order[(gene_symbol_order$probe_id %<span class="keyword">in</span>% rownames(data_filter_order_unique)),]</div><div class="line"><span class="comment"># 将data_filter_order_unique那些与gene_symbol_order$probe_id对应起来的探针与基因名提取出来</span></div><div class="line"></div><div class="line"><span class="comment">#table(unique_map_gene_probe$probe_id == rownames(data_filter_order_unique))</span></div><div class="line"><span class="literal">TRUE</span> </div><div class="line"><span class="number">8595</span> </div><div class="line"><span class="comment"># 可以发现，unique_map_gene_probe$probe_id中的探针顺序与data_filter_order_unique中的探针顺序是致的</span></div><div class="line"></div><div class="line">rownames(data_filter_order_unique) &lt;- unique_map_gene_probe$symbol</div><div class="line"><span class="comment"># 将最终过滤后的矩阵的名称换为基因名</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/gKC46jCJDE.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="更改矩阵名称">10.更改矩阵名称</h3>
<p>上题已经实现。</p>
<h3 id="绘制基本图形">11.绘制基本图形</h3>
<p>原题描述：对第10步得到的表达矩阵进行探索，先画第一个样本的所有基因的表达量的boxplot,hist,density ， 然后画所有样本的 这些图</p>
<p>思路：使用ggplot2绘图，要把数据转换为ggplot2可识别的形式，也就是第1列是基因名，第2列是分组信息，第3列是表达值，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ggplot2)</div><div class="line"><span class="keyword">library</span>(reshape2)</div><div class="line">data_draw &lt;- melt(data_filter_order_unique)</div><div class="line">colnames(data_draw) &lt;- c(<span class="string">"Gene"</span>,<span class="string">"Group"</span>,<span class="string">"Value"</span>)</div><div class="line">data_draw$status &lt;- rep(data_phenotype$Disease,each=nrow(data_filter_order_unique))</div></pre></td></tr></table></figure>
<h4 id="箱线图">箱线图</h4>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">p &lt;- ggplot(data=data_draw,aes(x=Group,y=Value,fill=status))+ geom_boxplot()</div><div class="line">p &lt;- p + theme(axis.text.x = element_text(angle=<span class="number">45</span>,hjust=<span class="number">1</span>,vjust=<span class="number">1</span>))</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/AC46gD52em.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="小提琴图">小提琴图</h4>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">p &lt;- ggplot(data=data_draw,aes(x=Group,y=Value,fill=status))+ geom_violin()</div><div class="line">p &lt;- p + theme(axis.text.x = element_text(angle=<span class="number">45</span>,hjust=<span class="number">1</span>,vjust=<span class="number">1</span>))</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/jKe8Cfj6c3.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="密度图">密度图</h4>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">p &lt;- ggplot(data=data_draw,aes(x=Value,fill=status))+ geom_histogram(bins=<span class="number">500</span>)</div><div class="line">p &lt;- p + facet_wrap(~Group,nrow=<span class="number">5</span>)</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/mEa4Ilm08h.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="密度图-1">密度图</h4>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">p &lt;- ggplot(data=data_draw,aes(x=Value,fill=status))+ geom_density()</div><div class="line">p &lt;- p + facet_wrap(~Group,nrow=<span class="number">5</span>)</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180906/cFHc7HAJbk.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="pca图">PCA图</h4>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">pca_data &lt;- prcomp(t(data_expression),scale=<span class="literal">TRUE</span>)</div><div class="line">pcx &lt;- data.frame(pca_data$x)</div><div class="line">pcr &lt;- cbind(samples=rownames(pcx),group_list, pcx) </div><div class="line">p  &lt;- ggplot(pcr, aes(PC1, PC2))+geom_point(size=<span class="number">5</span>, aes(color=group_list)) +</div><div class="line">  geom_text(aes(label=samples),hjust=-<span class="number">0.1</span>, vjust=-<span class="number">0.3</span>)</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180907/H04EBamdi5.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="聚类图">聚类图</h4>
<p>聚类的相关知识参考这篇笔记<a href="http://rvdsd.top/2018/08/19/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-17%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90%E7%AC%94%E8%AE%B0/">《聚类分析笔记》。</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">data_clust &lt;- t(data_expression)</div><div class="line">rownames(data_clust) &lt;- colnames(data_expression)</div><div class="line">data_clust_dist &lt;- dist(data_clust,method=<span class="string">"euclidean"</span>)</div><div class="line">hc &lt;-hclust(data_clust_dist,<span class="string">"ward"</span>)</div><div class="line"><span class="keyword">library</span>(factoextra)</div><div class="line">fviz_dend(hc, k = <span class="number">4</span>,</div><div class="line">          cex = <span class="number">0.5</span>,</div><div class="line">          k_colors = c(<span class="string">"#2E9FDF"</span>, <span class="string">"#00AFBB"</span>, <span class="string">"#E7B800"</span>, <span class="string">"#FC4E07"</span>),</div><div class="line">          color_labels_by_k = <span class="literal">TRUE</span>, rect = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180907/kk1caGGg9j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="统计学指标">12.统计学指标</h3>
<p>原题描述：理解统计学指标mean,median,max,min,sd,var,mad并计算出每个基因在所有样本的这些统计学指标，最后按照mad值排序，取top 50 mad值的基因，得到列表。</p>
<p>mean,median,max,min,sd,var,mad中除了mad外，其余的好理解，下面只介绍一下mad。</p>
<p>mad指的是<code>绝对中位数差(median absolute deviation)</code>，在R中的函数是<code>mad()</code>。它的用法如下所示： <figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">mad(x, center = median(x), constant = <span class="number">1.4826</span>, na.rm = <span class="literal">FALSE</span>, low = <span class="literal">FALSE</span>, high = <span class="literal">FALSE</span>)</div></pre></td></tr></table></figure></p>
<p>其中，<code>x</code>数值向量；<code>center</code>可选，是中心点，默认是中位数；<code>constant</code>是缩放因子(scale factor)，<code>na.rm</code>如果是TRUE，在计算前将x中的NA删除，<code>low</code>如果是TRUE，计算<code>lo-median</code>，也就是说，对于个数为偶数的样本，在最后计算中位数时不使用两个中间值的均值，而是采用其中较小的值。<code>high</code>如果为TRUE，计算‘hi-median’，也就是对于偶数样本，采用两个中间值的较大者作为中位数。</p>
<p>在R中，mad的计算公式为<code>constant * cMedian(abs(x - center))</code>，其中，center默认为<code>median(x)</code>，<code>cMedian</code>为“低”中位数或“高”中位数，constant默认值为<code>1.4826</code>，它近似等于<span class="math inline">\(1/\Phi^{-1}(3/4)=1/qnorm(3/4)\)</span>，这样就可以确保对于服从<span class="math inline">\(N(\mu,\sigma^2)\)</span>分布的<span class="math inline">\(X_{i}\)</span>和大的n值，存在以下一致性：</p>
<p><span class="math inline">\(E[mad(X_{1},...,X_{n})]=\sigma\)</span></p>
<p>这里取3/4是因为MAD包含标准正态累积分布函数的50%，具体的推导过程可以看这个帖子<a href="https://blog.csdn.net/horses/article/details/78749485" target="_blank" rel="external">《绝对中位差Median Absolute Deviation》</a>。</p>
<p>如果na.rm为TRUE，在执行计算之前，将会删除x中的NA值；否则，只要x中存在任何NA值，mad函数将会返回NA</p>
<p>我们来看一个案例：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">x &lt;- c(<span class="number">1</span>,<span class="number">2</span>,<span class="number">3</span>,<span class="number">5</span>,<span class="number">7</span>,<span class="number">8</span>)</div><div class="line">x_median &lt;- median(x)</div><div class="line">x_median</div><div class="line"><span class="comment"># [1] 4</span></div><div class="line">y &lt;- median(abs(x - x_median))</div><div class="line">y</div><div class="line"><span class="comment"># [1] 2.5</span></div><div class="line">mad(x)</div><div class="line"><span class="comment"># [1] 3.7065</span></div><div class="line"><span class="comment"># 这个值也就是y乘以1.4826</span></div><div class="line">mad(x)/y</div><div class="line"><span class="comment"># [1] 1.4826</span></div></pre></td></tr></table></figure>
<h3 id="提取-top-50-mad值的基因列表">13.提取 TOP 50 mad值的基因列表</h3>
<p>原是描述：根据第12步骤得到top 50 mad值的基因列表来取表达矩阵的子集，并且热图可视化子表达矩阵。试试看其它5种热图的包的不同效果，先看一下原表，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/3ABk79J8I4.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>从上面的这个表可以知道，行是基因名，列是样本名，现在进行计算，如下所示：</p>
<h1 id="注补充apply系列函数的相关知识">注：补充apply系列函数的相关知识</h1>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">data_mad &lt;- apply(data_filter_order_unique,<span class="number">1</span>,mad)</div><div class="line">head(data_mad)</div><div class="line"><span class="comment"># RABGGTA      MAPK3    CYP2C19      CXCR5      DUSP1    EIF2AK2 </span></div><div class="line"><span class="comment"># 0.03886242 0.22779776 0.08294023 0.38497146 1.43854685 0.24478690 </span></div><div class="line">data_mad_30 &lt;- head(sort(data_mad,decreasing = <span class="literal">TRUE</span>),<span class="number">30</span>)</div><div class="line">data_mad_30</div><div class="line"><span class="comment"># FAM30A  IGF2BP3      DMD     TCF7   SLAMF1      FOS      HBB   LGALS1    IGLC1 </span></div><div class="line"><span class="comment"># 3.982191 3.234011 3.071541 2.993352 2.944105 2.938078 2.732776 2.600604 2.590895 </span></div><div class="line"><span class="comment"># ZAP70     FCN1   LHFPL2      HBB   S100A8  GUSBP11   COBLL1    VIPR1    PCDH9 </span></div><div class="line"><span class="comment"># 2.579046 2.371590 2.317045 2.267136 2.220970 2.204212 2.179666 2.171912 2.144223 </span></div><div class="line"><span class="comment"># SLC25A1     JUND      IGH     H1FX  ZNF804A     OAS1     CCL3     GNLY     CYBB </span></div><div class="line"><span class="comment"># 2.115238 2.105234 2.090866 2.052788 1.986163 1.883509 1.862311 1.814364 1.811973 </span></div><div class="line"><span class="comment"># VAMP5   RNASE6     RGS2 </span></div><div class="line"><span class="comment"># 1.791017 1.775430 1.770151 </span></div><div class="line">data_mad_30_names &lt;- names(data_mad_30)</div><div class="line">data_mad_30_names</div><div class="line"><span class="comment"># [1] "FAM30A"  "IGF2BP3" "DMD"     "TCF7"    "SLAMF1"  "FOS"     "HBB"     "LGALS1" </span></div><div class="line"><span class="comment"># [9] "IGLC1"   "ZAP70"   "FCN1"    "LHFPL2"  "HBB"     "S100A8"  "GUSBP11" "COBLL1" </span></div><div class="line"><span class="comment"># [17] "VIPR1"   "PCDH9"   "SLC25A1" "JUND"    "IGH"     "H1FX"    "ZNF804A" "OAS1"   </span></div><div class="line"><span class="comment"># [25] "CCL3"    "GNLY"    "CYBB"    "VAMP5"   "RNASE6"  "RGS2" </span></div><div class="line">data_mad_30_heatmap &lt;- data_filter_order_unique[data_mad_30_names,]</div><div class="line"><span class="comment"># 提取mad 30的子集</span></div></pre></td></tr></table></figure>
<p>以下为提取后的数据集，这个数据集是一个矩阵，行是基因名，列是样本名，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/511Jlk60be.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>绘制这个数据集的热图，目前已经绘制热图的函数有：</p>
<h4 id="heatmap">heatmap</h4>
<p><code>heatmap(stat)</code>这个是stat包中自带的热图工具，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">heatmap(x, Rowv = <span class="literal">NULL</span>, Colv = <span class="keyword">if</span>(symm)<span class="string">"Rowv"</span> <span class="keyword">else</span> <span class="literal">NULL</span>,</div><div class="line">        distfun = dist, hclustfun = hclust,</div><div class="line">        reorderfun = <span class="keyword">function</span>(d, w) reorder(d, w),</div><div class="line">        add.expr, symm = <span class="literal">FALSE</span>, revC = identical(Colv, <span class="string">"Rowv"</span>),</div><div class="line">        scale = c(<span class="string">"row"</span>, <span class="string">"column"</span>, <span class="string">"none"</span>), na.rm = <span class="literal">TRUE</span>,</div><div class="line">        margins = c(<span class="number">5</span>, <span class="number">5</span>), ColSideColors, RowSideColors,</div><div class="line">        cexRow = <span class="number">0.2</span> + <span class="number">1</span>/log10(nr), cexCol = <span class="number">0.2</span> + <span class="number">1</span>/log10(nc),</div><div class="line">        labRow = <span class="literal">NULL</span>, labCol = <span class="literal">NULL</span>, main = <span class="literal">NULL</span>,</div><div class="line">        xlab = <span class="literal">NULL</span>, ylab = <span class="literal">NULL</span>,</div><div class="line">        keep.dendro = <span class="literal">FALSE</span>, verbose = getOption(<span class="string">"verbose"</span>), <span class="keyword">...</span>)</div></pre></td></tr></table></figure>
<p>如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">heatmap(data_mad_30_heatmap,scale=c(<span class="string">"row"</span>))</div><div class="line"><span class="comment"># 通常我们要查看某个基因在不同样本中的表达情况，就采用row进行均一化</span></div><div class="line"><span class="comment"># 上面的这个数据行是基因名，列是样本名，因此采用行均一化，就可以看相同的某个基因在不同样本中的表达情况</span></div><div class="line"><span class="comment"># 如果要想看一个样本中不同基因的表达情况，就要采用column进行均一化</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/CLEbGBG8db.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="heatmap.2">heatmap.2</h4>
<p><code>heatmap.2</code>是<code>gplots</code>包中的函数（文档太长，就不列了），绘制热图如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(gplots)</div><div class="line">heatmap.2(data_mad_30_heatmap,scale=<span class="string">"row"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/Ak1136CLHH.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="d3heatmap">d3heatmap</h4>
<p>d3heatmap包可用于生成交互式热图绘制,可通过以下代码生成：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">if</span> (!<span class="keyword">require</span>(<span class="string">"devtools"</span>))</div><div class="line">install.packages(<span class="string">"devtools"</span>)</div><div class="line">devtools::install_github(<span class="string">"rstudio/d3heatmap"</span>)</div><div class="line"><span class="keyword">library</span>(d3heatmap)</div><div class="line">d3heatmap(data_mad_30_heatmap, colors = <span class="string">"RdBu"</span>, k_row = <span class="number">4</span>, k_col = <span class="number">2</span>)</div><div class="line"><span class="comment"># colors是颜色，k_row是行的聚类数目，用不同的颜色标记了出来，k_col是列的聚类数目</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/c7kkhD44jc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="heatmap-1">Heatmap</h4>
<p>这个函数是<code>ComplexHeatmap</code>包中的函数。用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">install.packages(<span class="string">"ComplexHeatmap"</span>)</div><div class="line">biocLite(<span class="string">"ComplexHeatmap"</span>)</div><div class="line"><span class="keyword">library</span>(ComplexHeatmap)</div><div class="line">Heatmap(data_mad_30_heatmap,name=<span class="string">"Mad 30"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/EialA01eHB.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="pheatmap">pheatmap</h4>
<p>此函数是<code>pheatmap</code>包中的函数，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(pheatmap)</div><div class="line">pheatmap(data_mad_30_heatmap)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/A22A6m3dA9.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>上述几种绘制热图的函数是比较常用的，还有很多参数，需要自己调一下。</p>
<h3 id="不同统计指标的交集">14.不同统计指标的交集</h3>
<p>原题描述：取不同统计学指标mean,median,max,mean,sd,var,mad的各top50基因列表，使用UpSetR包来看他们之间的overlap情况。先计算一下各个统计学指标，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line">data_mean = tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,mean)),<span class="number">30</span>)</div><div class="line">data_median = tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,median)), <span class="number">30</span>)</div><div class="line">data_max &lt;- tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,max)),<span class="number">30</span>)</div><div class="line">data_min &lt;- tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,min)),<span class="number">30</span>)</div><div class="line">data_sd &lt;- tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,sd)),<span class="number">30</span>)</div><div class="line">data_var &lt;- tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,var)),<span class="number">30</span>)</div><div class="line">data_mad &lt;- tail(sort(apply(data_filter_order_unique,<span class="number">1</span>,mad)),<span class="number">30</span>) </div><div class="line"></div><div class="line">data_all &lt;- unique(c(names(data_mean),names(data_median),names(data_max),names(data_min),</div><div class="line">              names(data_sd),names(data_var),names(data_mad)))</div><div class="line"><span class="comment"># 取并集</span></div><div class="line"></div><div class="line">data_upset &lt;- data.frame(data_all,</div><div class="line">                         data_mean=ifelse(data_all %<span class="keyword">in</span>% names(data_mean) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_median=ifelse(e_all %<span class="keyword">in</span>% names(data_median) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_max=ifelse(e_all %<span class="keyword">in</span>% names(data_max) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_min=ifelse(e_all %<span class="keyword">in</span>% names(data_min) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_sd=ifelse(e_all %<span class="keyword">in</span>% names(data_sd) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_var=ifelse(e_all %<span class="keyword">in</span>% names(data_var) ,<span class="number">1</span>,<span class="number">0</span>),</div><div class="line">                         data_mad=ifelse(e_all %<span class="keyword">in</span>% names(data_mad) ,<span class="number">1</span>,<span class="number">0</span>))</div><div class="line"><span class="comment"># 生成一个数据框，内容是不同的统计量是在并集中的重合，如果在为1，不在为0</span></div><div class="line"></div><div class="line">install.packages(<span class="string">"UpSetR"</span>)</div><div class="line"><span class="keyword">library</span>(<span class="string">"UpSetR"</span>)</div><div class="line">upset(data_upset,nsets = <span class="number">7</span>,sets.bar.color = <span class="string">"#56B4E9"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/gLK22c6lCJ.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="提取表型数据">15.提取表型数据</h3>
<p>已经在前面完成。</p>
<h3 id="聚类">16. 聚类</h3>
<p>已经在第11题中完成。</p>
<h3 id="pca">17.PCA</h3>
<p>已经在第11题中完成。</p>
<h3 id="批量t检验">18.批量t检验</h3>
<h1 id="注这里要补充一下统计学的相关知识">注：这里要补充一下统计学的相关知识。</h1>
<p>代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div></pre></td><td class="code"><pre><div class="line">group_list</div><div class="line"><span class="comment"># [1] "progres." "stable"   "progres." "progres." "progres." "progres." "stable"  </span></div><div class="line"><span class="comment"># [8] "stable"   "progres." "stable"   "progres." "stable"   "progres." "stable"  </span></div><div class="line"><span class="comment"># [15] "stable"   "progres." "progres." "progres." "progres." "progres." "progres."</span></div><div class="line"><span class="comment"># [22] "stable"  </span></div><div class="line">gl=as.factor(group_list) </div><div class="line"><span class="comment">#将表型数据因子化</span></div><div class="line"></div><div class="line">group1 = which(group_list == levels(gl)[<span class="number">1</span>]) </div><div class="line"><span class="comment"># levels(group_list)[1]返回第一个因子progres</span></div><div class="line"><span class="comment"># 从group_list中选出progres的元素</span></div><div class="line"><span class="comment"># 用which来获取他们所在的索引</span></div><div class="line"></div><div class="line">group2 = which(group_list == levels(gl)[<span class="number">2</span>]) </div><div class="line"><span class="comment">#返回第二个因子stable</span></div><div class="line"></div><div class="line">data_t1 = data_expression[, group1] </div><div class="line"><span class="comment">#将表型为progres的样本选出来</span></div><div class="line">data_t2 = data_expression[, group2] </div><div class="line"><span class="comment">#将表型为stable的样本选出来</span></div><div class="line">data_t = cbind(data_t1, data_t2) </div><div class="line"></div><div class="line">pvals = apply(data_expression, <span class="number">1</span>, <span class="keyword">function</span>(x)&#123;</div><div class="line">  t.test(as.numeric(x)~group_list)$p.value</div><div class="line">&#125;)</div><div class="line">p.adj = p.adjust(pvals, method = <span class="string">"BH"</span>)  </div><div class="line"></div><div class="line">data_mean_1 = rowMeans(data_t1) </div><div class="line"><span class="comment">#progres是对照组</span></div><div class="line">data_mean_2 = rowMeans(data_t2) </div><div class="line"><span class="comment">#stable是使用药物处理后的——处理组</span></div><div class="line">logFC = data_mean_2-data_mean_1</div><div class="line"></div><div class="line">DEG_t.test = cbind(data_mean_1, data_mean_2, log2FC, pvals, p.adj)</div><div class="line">DEG_t.test=DEG_t.test[order(DEG_t.test[,<span class="number">4</span>]),] <span class="comment">#从小到大排序</span></div><div class="line">DEG_t.test=as.data.frame(DEG_t.test)</div><div class="line">head(DEG_t.test)</div><div class="line"><span class="comment"># data_mean_1 data_mean_2     log2FC        pvals     p.adj</span></div><div class="line"><span class="comment"># 36129_at    7.875615    8.791753  0.9161377 1.629755e-05 0.2057566</span></div><div class="line"><span class="comment"># 37676_at    6.622749    7.965007  1.3422581 4.058944e-05 0.2436177</span></div><div class="line"><span class="comment"># 33791_at    7.616197    5.786041 -1.8301554 6.965416e-05 0.2436177</span></div><div class="line"><span class="comment"># 39967_at    4.456446    2.152471 -2.3039752 8.993339e-05 0.2436177</span></div><div class="line"><span class="comment"># 34594_at    5.988866    7.058738  1.0698718 9.648226e-05 0.2436177</span></div><div class="line"><span class="comment"># 32198_at    4.157971    3.407405 -0.7505660 2.454557e-04 0.3516678</span></div></pre></td></tr></table></figure>
<h3 id="limma">19.limma</h3>
<p>原题描述：使用limma包对表达矩阵及样本分组信息进行差异分析，得到差异分析表格，重点看logFC和P值，画个火山图(就是logFC和-log10(P值)的散点图。)。</p>
<h1 id="注这里需要补充一下基因芯片原理的相关知识">注：这里需要补充一下基因芯片原理的相关知识。</h1>
<p>limma包是常用的分析芯片数据的R包，分析过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">design1=model.matrix(~factor(group_list)) </div><div class="line"><span class="comment"># 这是一个设计矩阵</span></div><div class="line">colnames(design1)=levels(factor(group_list))</div><div class="line">rownames(design1)=colnames(data_expression)</div><div class="line"></div><div class="line">fit1 = lmFit(data_expression,design1)</div><div class="line">fit1=eBayes(fit1)</div><div class="line">options(digits = <span class="number">3</span>) <span class="comment">#设置结果的小数位数为3</span></div><div class="line">mtx1 = topTable(fit1,coef=<span class="number">2</span>,adjust=<span class="string">'BH'</span>,n=<span class="literal">Inf</span>) <span class="comment">#coef要么必须等于2， 要么是个字符串；关于adjust的设置，说明书中13.1章有描述，BH是最流行的设置</span></div><div class="line"><span class="comment"># topTable 默认显示前10个基因的统计数据；使用选项n可以设置,n=Inf就是不设上限，全部输出</span></div><div class="line">DEG_mtx1 = na.omit(mtx1) <span class="comment">#去除缺失值</span></div><div class="line">head(DEG_mtx1)</div></pre></td></tr></table></figure>
<p>解释一下代码：<code>design1=model.matrix(~factor(group_list))</code>这是一个设计矩阵，关于设计矩阵的一些知识，可以看这篇笔记<a href="http://rvdsd.top/2018/06/07/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B005-%E7%BA%BF%E6%80%A7%E6%A8%A1%E5%9E%8B/">《线性模型》</a>。</p>
<h1 id="注这里需要补充一下线性代数的相关知识">注：这里需要补充一下线性代数的相关知识。</h1>
<p><code>lmFit</code>函数是用于构建一个线性模型；它的参数一个是表达矩阵，一个是分组对象</p>
<p><code>eBays</code>函数是用于构建一个微阵列线性模型。</p>
<h3 id="火山图">20.火山图</h3>
<p>设定绘图数据的阈值，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">DEG=DEG_mtx1</div><div class="line">logFC_cutoff &lt;- with(DEG,mean(abs(logFC)) + <span class="number">2</span>*sd(abs(logFC)) ) </div><div class="line">DEG$result = as.factor(ifelse(DEG$P.Value &lt; <span class="number">0.05</span> &amp; abs(DEG$logFC) &gt; logFC_cutoff,</div><div class="line">                              ifelse(DEG$logFC &gt; logFC_cutoff ,<span class="string">'UP'</span>,<span class="string">'DOWN'</span>),<span class="string">'NOT'</span>))</div><div class="line">this_tile &lt;- paste0(<span class="string">'Cutoff for logFC is '</span>,round(logFC_cutoff,<span class="number">3</span>), <span class="comment">#round保留小数位数</span></div><div class="line">                    <span class="string">'\nThe number of up gene is '</span>,nrow(DEG[DEG$result ==<span class="string">'UP'</span>,]) ,</div><div class="line">                    <span class="string">'\nThe number of down gene is '</span>,nrow(DEG[DEG$result ==<span class="string">'DOWN'</span>,])</div><div class="line">)</div><div class="line">ggplot(data=DEG, aes(x=logFC, y=-log10(P.Value), color=result)) +</div><div class="line">  geom_point(alpha=<span class="number">0.4</span>, size=<span class="number">1.75</span>) +</div><div class="line">  theme_set(theme_set(theme_bw(base_size=<span class="number">20</span>)))+</div><div class="line">  xlab(<span class="string">"log2 fold change"</span>) + ylab(<span class="string">"-log10 p-value"</span>) +</div><div class="line">  ggtitle( this_tile ) + theme(plot.title = element_text(size=<span class="number">15</span>,hjust = <span class="number">0.5</span>))+</div><div class="line">  scale_colour_manual(values = c(<span class="string">'blue'</span>,<span class="string">'black'</span>,<span class="string">'red'</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180909/ekdCeiBe9e.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="结论">结论</h2>
<p>与R语言有关的这20道题中涉及的知识不少，代码背后的数理统计与线性代数知识目前比较欠缺，需要补充。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://github.com/bioconductor-china/basic" target="_blank" rel="external">ExpressionSet 对象简单讲解</a></li>
<li><a href="http://www.bio-info-trainee.com/3409.html" target="_blank" rel="external">生信人的20个R语言习题</a></li>
<li><a href="http://www.bio-info-trainee.com/bioconductor_China/software/limma.html" target="_blank" rel="external">用limma对芯片数据做差异分析</a></li>
<li><a href="http://bio-info-trainee.com/tmp/basic_visualization_for_expression_matrix.html" target="_blank" rel="external">basic visualization for expression matrix</a></li>
<li><a href="https://mp.weixin.qq.com/s/gOmwLbQle7NdcUTx2HRoEg" target="_blank" rel="external">R语言20练习题【完整版】</a></li>
<li><a href="https://blog.csdn.net/horses/article/details/78749485" target="_blank" rel="external">绝对中位差Median Absolute Deviation</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/23/Tools/BioStar笔记之SRA/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/23/Tools/BioStar笔记之SRA/" itemprop="url">Biostar Handbook笔记之SRA</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-23T12:00:00+08:00">
                2018-08-23
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,834
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="什么是sra">什么是SRA</h2>
<p>SRA的全称是Short Read Archive，它是由NCBI提供的一个数据储存服务。从这个名称上来看，貌似是储存测序的原始数据的一个数据库。SRA中的数据组织形式是按照层级结构组织的，不用的数据前面有不同的前缀，如下所示：</p>
<ol style="list-style-type: decimal">
<li>NCBI的生物项目(BioProject)的命名为<code>PRJN****</code>，例如<code>PRJNA257197</code>，这个数据是对一个单独研究机制的整体描述，一个生物项目含有很多样本与数据集；</li>
<li>NCBI的生物样本(BioSample)，它的命名为<code>SAM****</code>，或者是<code>SRS****</code>，例如<code>SAMN03254300</code>，这个数据指的是生物样本的来源，每个生物学样本都有一个单独的BioSample编号，以及一组独特的特征。</li>
<li>SRA实验，命名方式为<code>SRX****</code>，这是针对某个特定样本的一个测序文库。</li>
<li>SRA run，命名方式为<code>SRR****</code>或<code>ERR****</code>，例如<code>SRR1553610</code>，它对应的是某个测序文库的数据文件。</li>
</ol>
<h2 id="下载sra数据文件">下载SRA数据文件</h2>
<p>下载SRA主要有两种方式，一种是通过<code>sratoolkit</code>这个工具包，另外一种就是通过普通的下载工具下载，例如迅雷（我下载过，速度还不错）。</p>
<h3 id="sratoolkit安装">sratoolkit安装</h3>
<p>现在讲一下sratoolkit这个工具包的安装，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<p>或者直接使用conda来安装，非常方便，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">conda install sratoolkit</div></pre></td></tr></table></figure>
<h3 id="下载数据">下载数据</h3>
<h4 id="下载fastq格式文件">下载fastq格式文件</h4>
<p>sratoolkit中常用的下载命令有两个，如下所示：</p>
<p><code>fastq-dump</code>，下载FASTQ格式的文件；</p>
<p><code>sam-dump</code>下载SAM比对格式的文件；</p>
<p>fastq-dump的典型用法如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">fastq-dump SRR1553607</div></pre></td></tr></table></figure>
<p>查看下载结果，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/sratoolkit$ head -4 SRR1553607.fastq </div><div class="line">@SRR1553607.1 1 length=202</div><div class="line">GTTAGCGTTGTTGATCGCGACGCAACAACTGGTAAAGAATCTGGAAGAAGGATATCAGTTCAAACGCTCAAGCGAGATGATGGATATTTTTGAACGACTCATCATTCGTCACATCCCGTTCAACCTGATTGAGTCGTTCAAAAATATCCATCATCTCGCTTGAGCGTTTGAACTGATATCCTTCTTCCAGATTCTTTACCAG</div><div class="line">+SRR1553607.1 1 length=202</div><div class="line">BB@FFFFFHHHHHJJJJJJJJJJJJJJJJJJJGHIJJJJJJJJJHHHHHFFFFFEEEEEEEEEDDDDDDDDDDDDDDDDDEDDDDEDEEEDDDDDDDDDDDCCCFFFFFHHHHHJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJIJJJIJJHIJJHHHFFFFFEDDEFEEEDDDDDDDDDDDDDEDDDDDD</div></pre></td></tr></table></figure>
<p>有的数据文件是双端测序文件，此时我们通过添加<code>--split-files</code>参数下载双端测序的文件，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">fastq-dump --split-files SRR1553607</div></pre></td></tr></table></figure>
<p>下载结果如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<p>fastq-dump还可以只下载一部分数据，例如我闪只下载前10000个reads，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">fastq-dump --split-files -X 10000 SRR1553607</div></pre></td></tr></table></figure>
<p>不过，这种情况下，fastq-dump会先下载全部数据，然后从中提取前10000个reads，如下已经下载了全部的序列，直接运行即可，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/sratoolkit$ ls -lh</div><div class="line">total 171M</div><div class="line">drwxr-xr-x 5 biotest biotest 4.0K 3月   7  2017 sratoolkit.2.8.2-ubuntu64</div><div class="line">-rw-r--r-- 1 biotest biotest  73M 8月  30 11:37 sratoolkit.2.8.2-ubuntu64.tar.gz</div><div class="line">-rw-r--r-- 1 biotest biotest 2.6M 8月  30 12:23 SRR1553607_1.fastq</div><div class="line">-rw-r--r-- 1 biotest biotest 2.6M 8月  30 12:23 SRR1553607_2.fastq</div><div class="line">-rw-r--r-- 1 biotest biotest  93M 8月  30 12:15 SRR1553607.fastq</div></pre></td></tr></table></figure>
<p>从上面可以看出，有两个文件的大小是2.6M，用wc统计一下，发现确实是1000个reads（4000除以4，因为一条reads占4行）。</p>
<h4 id="下载sra格式文件">下载sra格式文件</h4>
<p>通常情况下，下载sra格式的文件可以使用<code>curl</code>，如下所示</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">curl -O ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR197/SRR1972739/SRR1972739.sra</div></pre></td></tr></table></figure>
<p>查看下载结果：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/LKEGL2djmd.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>prefetch也能下载该文件，不过是fastq格式的，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="meta">#</span> 先删除原来的下载文件</div><div class="line"><span class="meta">#</span> rm SRR1972739.sra</div><div class="line">prefetch SRR1972739</div><div class="line"><span class="meta">#</span> 注意，后续不带sra</div></pre></td></tr></table></figure>
<p>查看下载文件：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/kGaC1h314f.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>当通过fastq-dump下载文件时，全部的数据默认情况下都位于<code>~/ncbi/public/sra/</code>中，如下所示</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/sratoolkit$ ls -lh ~/ncbi/public/sra</div><div class="line">total 110M</div><div class="line">-rw-r--r-- 1 biotest biotest 25M 8月  30 12:15 SRR1553607.sra</div><div class="line">-rw-r--r-- 1 biotest biotest 86M 8月  30 12:33 SRR1972739.sra</div></pre></td></tr></table></figure>
<h2 id="sra中的spot">SRA中的spot</h2>
<p>当我们在SRA中检索时，会发现这样的信息，如下所示</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/meB6E1cAJj.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>SRA中呈现的结果是<code>Spots</code>，而非<code>reads</code>，其实这两个术语描述的是一种东西。但是如果真要区分这两个东西，可以从Illumina测序的原理上来理解，在测试时，最终的序列是靠相机来读取的，一个spot上有一串碱基，，spot就是这串碱基（就是桥式PCR一端的结合位点）所在的位置，因此spot中包括了这些信息：接头，测序引物，目标序列片段，index，这是从技术层面来讲的，如果从生物学层面来讲，我们所说的reads其实就是指的目标序列，是去掉了接头，测序引物，index后的序列信息。但在很多情况下，spots跟reads所表达的意思是一样的。</p>
<h2 id="run的信息">run的信息</h2>
<p><code>sra-stat</code>可以获取数据的XML报告，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sra-stat --xml --quick SRR1553610</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/8iJaG6idKh.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>查看一个PacBio数据集的read长度，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sra-stat --xml --statistics SRR4237168</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/efKGcFld96.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="下载多个sra-runs信息">下载多个SRA runs信息</h2>
<p>先检索某个生物项目，例如<code>PRJNA257197</code>，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db sra -query PRJNA257197</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/hdba7Cch58.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>从结果中可以发现，有891个run，再看一下网页版：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/5J9BAEaDJC.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在，我们把这些信息都导出到一个文件中，即<code>info.csv</code>，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db sra -query PRJNA257197 | efetch -format runinfo &gt; info.csv</div></pre></td></tr></table></figure>
<p>现在显示前5行的第1到8列，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">cut -d ',' -f 1-8 info.csv  |head -5</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/b03jBje1La.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>在网页版中，也能导出这些信息，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/gjg36IfEJ2.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>打开，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/LiaE658K52.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="文本的切割与组合">文本的切割与组合</h3>
<p>前面已经看了简单的文本切割，现在再复习一下。</p>
<ol style="list-style-type: decimal">
<li>只显示<code>info.csv</code>文件的第1列：</li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">cat info.csv |cut -f 1 -d ',' |head</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/A67BBG0f3g.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<ol start="2" style="list-style-type: decimal">
<li>提取第1列信息，满足开头是<code>SRR</code>的字段，将这些信息导致入到<code>ids.txt</code>这个文件中：</li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/m7HG4HKLFg.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>使用fastq-dump下载sra的run文件，并只取其前10000个reads，并分割（双端测序）如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">fastq-dump -X 10000 --split-files SRR1972918</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/0CK2lA2Ikc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>使用<code>xargs</code>命令生成新的文本信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">cat ids.txt | xargs -n 1 echo This is number $1</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/DHcd59ACC7.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="xargs">xargs</h4>
<p>xargs命令的功能是：将上游输入的数据转化为特定的命令，以上面的命令为例解释一下<code>cat ids.txt | xargs -n 1 echo This is number $1</code>，这个命令是将<code>ids.txt</code>中的SRR号作为<code>echo This is number $1</code>的输入，由于idx。txt这个文件中一共有10行，那么这个<code>-n 1</code>此时就表示，把每一行的信息作为参数传递给<code>THis is number $1</code>，如果我们把<code>-n 1</code>改为<code>-n 2</code>，那么就是下面的情况：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/9jhDjbJ40i.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>可以发现，一次传递了2个参数的内容。</p>
<p>xargs后面还有东西，也就是<code>echo This is number $1</code>，这个<code>echo</code>好理解，把它当成输出某些信息的函数即可，这个<code>$1</code>表示脚本运行中的第1个参数，具体的内容参考以前的笔记<a href="http://rvdsd.top/2018/05/02/Linux/Linux-shell04-%E5%A4%84%E7%90%86%E7%94%A8%E6%88%B7%E8%BE%93%E5%85%A5/">《Shell学习笔记（4）——处理用户输入》</a>。</p>
<p>通过这个<code>xargs</code>这个函数，我们就可以下载批量的fastq格式的文件了，现在把ids.txt中所列的文件下载下来，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">mkdir download </div><div class="line">mv ids.txt downolad</div><div class="line">cd download</div><div class="line">cat ids.txt |xargs -n 1 fastq-dump -X 10000 -split-files $1</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/F4eJ0kfL4j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>过滤多个条件的记录，如下所示：</p>
<p>这里提取的是2014-08-19，并以SRR开头的文件名，显示前3列：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">cat info.csv | grep "2014-08-19"|cut -f 1 -d ','| grep "SRR" |head -10</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180830/KiACF4gfI1.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="获取sra中全部runs信息">获取SRA中全部runs信息</h2>
<p>下载2018年1月的全部SRA的信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db sra -query '"2018/01"[Publication Date]' | efetch -format runinfo &gt; 2018-01.csv</div></pre></td></tr></table></figure>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/22/Tools/BioStar笔记之Entrez/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/22/Tools/BioStar笔记之Entrez/" itemprop="url">Biostar Handbook笔记之Entrez</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-22T12:00:00+08:00">
                2018-08-22
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,872
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="什么是entrez">什么是Entrez</h2>
<p>Entrez是NCBI的一个检索工具，它整合了PubMed数据库中的生物学文献中的信息，包括DNA序列，蛋白质序列，结构，基因组，基因表达的信息等。Entrez是一个法语单词，读作“En-trez”。</p>
<h2 id="entrez的使用">Entrez的使用</h2>
<p>NCBI提供了Entrez的一个WEB API，称为<a href="https://www.ncbi.nlm.nih.gov/books/NBK25500/" target="_blank" rel="external">Entrez E-utils</a>，还有一套工具，称为<a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/" target="_blank" rel="external">Entrez Direct</a>。其中Entrez Direct可以在Linux下运行，通过命令行的形式访问NCBi的数据库。</p>
<h2 id="ncbi的数据组织形式">NCBI的数据组织形式</h2>
<p>NCBI上的数据是随着生物学的研究而不断更新的。例如我们以<code>Ebola virus 1976</code>为关键词在NCBI的Nucleotide中检索一下，单击第二行的链接，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180829/F61bhef8kd.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>我们会看到<a href="https://www.ncbi.nlm.nih.gov/nuccore/AF086833.2" target="_blank" rel="external">如下信息</a>：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180829/mdEHmK4lcE.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在解释一下这个信息：</p>
<ol style="list-style-type: decimal">
<li><code>AF086833</code>是登录号(accession number)，它代表了一个完整的数据记录，即使这个数据的版本再更改，它也会保持不变，具有唯一性。</li>
<li><code>VERSION AF086833.2</code>表示版本号，它与登录号比较相似，只是在登录号的后面用了一个号和一个数字表示，即<code>.2</code>。版本号的逗号前面部分不变，但是，如果序列发生改变，只改变版本号的逗号后面部分。</li>
</ol>
<p>自从2015年，NCBI已经不再使用原本的GI号了，一个数据对应一个GI号，不过一些老的文献中可能还会使用GI或gi号，这里需要注意一下。</p>
<h2 id="entrez-e-utils-web-api的使用">Entrez E-utils Web API的使用</h2>
<p>我们可以通过一个特定的URL使用Entrez Web API接口来访问NCBI的数据，这个接口的构成形式是这个样子的：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">https://service.nih.gov?param1=value1&amp;param2=value2&amp;param3=value3</div></pre></td></tr></table></figure>
<p>由于<code>&amp;</code>符号在shell中是一个特殊符号，因为我需要使用一个单引号，或者是转义字符来正常使用<code>&amp;</code>，转义字符的使用是<code>\&amp;</code>。</p>
<p>现在我们使用这种方法来获得<code>AF086833.2</code>的数据，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">curl -s https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2\&amp;db=nuccore\&amp;rettype=fasta | head</div><div class="line"><span class="meta">#</span> 或者是下面的这个样子</div><div class="line">curl -s 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2&amp;db=nuccore&amp;rettype=fasta' | head</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ curl -s https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2\&amp;db=nuccore\&amp;rettype=fasta | head</div><div class="line"><span class="meta">&gt;</span>AF086833.2 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATAACTATGAGGAAGATTAATAA</div><div class="line">TTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTT</div><div class="line">CAGAGCCACATCACAAAGATAGAGAACAACCTAGGTCTCCGAAGGGAGCAAGGGCATCAGTGTGCTCAGT</div><div class="line">TGAAAATCCCTTGTCAACACCTAGGTCTTATCACATCACAAGTTCCACCTCAGACTCTGCAGGGTGATCC</div><div class="line">AACAACCTTAATAGAAACATTATTGTTAAAGGACAGCATTAGTTCACAGTCAAACAAGCAAGATTGAGAA</div><div class="line">TTAACCTTGGTTTTGAACTTGAACACTTAGGGGATTGAAGATTCAACAACCCTAAAGCTTGGGGTAAAAC</div><div class="line">ATTGGAAATAGTTAAAAGACAAATTGCTCGGAATCACAAAATTCCGAGTATGGATTCTCGTCCTCAGAAA</div><div class="line">ATCTGGATGGCGCCGAGTCTCACTGAATCTGACATGGATTACCACAAGATCTTGACAGCAGGTCTGTCCG</div><div class="line">TTCAACAGGGGATTGTTCGGCAAAGAGTCATCCCAGTGTATCAAGTAAACAATCTTGAAGAAATTTGCCA</div></pre></td></tr></table></figure>
<h2 id="entrez-direct的使用">Entrez Direct的使用</h2>
<p>Entrez Direct简称EDirect，它算是Entrez E-utils的本地版本.</p>
<h3 id="安装">安装</h3>
<p>它的安装如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">sudo apt install ncbi-entrez-direct</div><div class="line"><span class="meta">#</span> 如果安装不了，可以直接从官网下载</div><div class="line"><span class="meta">#</span> curl ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/edirect.zip -o edirect.zip</div></pre></td></tr></table></figure>
<h3 id="下载序列">下载序列</h3>
<p>现在我们先看一下下面的这个命令<code>efetch</code>，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">efetch -db=nuccore -format=gb -id=AF086833 | head</div></pre></td></tr></table></figure>
<p>fetch的英文意思是拿来，从这个命令的命名上大概能猜出来，这个命令是用来检索一些信息的。</p>
<p>通过<code>efetch --help</code>看了这个命令的说明后，我们知道，<code>-db=nuccore</code>表示检索的是核酸（为什么使用nuccore，后来我试了一下<code>-db=nucleotide</code>也可以)还不清楚。<code>-format=gb</code>表示显示的是GeneBank的格式，<code>id=AF086833</code>表示登录号。</p>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ </div><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ efetch -db=nuccore -format=gb -id=AF086833 | head</div><div class="line">LOCUS       AF086833               18959 bp    cRNA    linear   VRL 13-FEB-2012</div><div class="line">DEFINITION  Ebola virus - Mayinga, Zaire, 1976, complete genome.</div><div class="line">ACCESSION   AF086833</div><div class="line">VERSION     AF086833.2</div><div class="line">KEYWORDS    .</div><div class="line">SOURCE      Ebola virus - Mayinga, Zaire, 1976 (EBOV-May)</div><div class="line">  ORGANISM  Ebola virus - Mayinga, Zaire, 1976</div><div class="line">            Viruses; ssRNA viruses; ssRNA negative-strand viruses;</div><div class="line">            Mononegavirales; Filoviridae; Ebolavirus.</div><div class="line">REFERENCE   1  (bases 1 to 18959)</div></pre></td></tr></table></figure>
<p>前面提到，<code>-db=nuccore</code>表示的是核酸，为什么不使用<code>-db=nucleotide</code>，我发现使用了一下，也能显示，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ efetch -db=nucleotide  -format=gb -id=AF086833 | head</div><div class="line">LOCUS       AF086833               18959 bp    cRNA    linear   VRL 13-FEB-2012</div><div class="line">DEFINITION  Ebola virus - Mayinga, Zaire, 1976, complete genome.</div><div class="line">ACCESSION   AF086833</div><div class="line">VERSION     AF086833.2</div><div class="line">KEYWORDS    .</div><div class="line">SOURCE      Ebola virus - Mayinga, Zaire, 1976 (EBOV-May)</div><div class="line">  ORGANISM  Ebola virus - Mayinga, Zaire, 1976</div><div class="line">            Viruses; ssRNA viruses; ssRNA negative-strand viruses;</div><div class="line">            Mononegavirales; Filoviridae; Ebolavirus.</div><div class="line">REFERENCE   1  (bases 1 to 18959)</div></pre></td></tr></table></figure>
<h3 id="下载不同格式的序列">下载不同格式的序列</h3>
<p>使用<code>efetch</code>功能，还能把<code>AF086833</code>的信息导入到不同格式的文件中去，但要注意<code>-format=格式</code>这个参数，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=gb -id=AF086833 &gt; AF086833.gb</div><div class="line"><span class="meta">#</span> download information output Genebank format files</div><div class="line"></div><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=fasta -id=AF086833 &gt; AF086833.fa</div><div class="line"><span class="meta">#</span> download information output fasta format files</div></pre></td></tr></table></figure>
<p>如果要下载多条序列，可以直接在<code>id=</code>参数后面添加就行了，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=fasta -id=AF086833,AF086833 &gt; test.fa</div><div class="line"><span class="meta">#</span> 这里只是说明一下下载多个序列的用法，因此id=后面接了两个相同的登录号</div></pre></td></tr></table></figure>
<h3 id="显示指定的碱基范围">显示指定的碱基范围</h3>
<p>efetch还可以显示指定的序列范围，例如下面就是显示<code>AF086833.2</code>这个序列的第100到300个碱基，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=100 -seq_stop=300</div><div class="line"><span class="meta">&gt;</span>AF086833.2:100-300 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">TTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTTCAGAGCCACATCACAAAGATAGAGAACAA</div><div class="line">CCTAGGTCTCCGAAGGGAGCAAGGGCATCAGTGTGCTCAGTTGAAAATCCCTTGTCAACACCTAGGTCTT</div><div class="line">ATCACATCACAAGTTCCACCTCAGACTCTGCAGGGTGATCCAACAACCTTAATAGAAACAT</div></pre></td></tr></table></figure>
<p>这里要使用两个参数，一个是<code>-seq_start=</code>，它是起始的碱基位置，一个是<code>-seq_stop=</code>，它是碱基的结束位置。</p>
<h3 id="显示互补链">显示互补链</h3>
<p>通过添加<code>-strand</code>参数可以显示互补链，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=1 -seq_stop=5 -strand=1</div><div class="line"><span class="meta">&gt;</span>AF086833.2:1-5 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">CGGAC</div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=1 -seq_stop=5 -strand=2</div><div class="line"><span class="meta">&gt;</span>AF086833.2:c5-1 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">GTCCG</div></pre></td></tr></table></figure>
<p>从上面的结果可以看出来，<code>strand=1</code>表示是正义链，也就是<code>CGGAC</code>，而<code>strand=2</code>表示的是互补链，也就是<code>CGGAC</code>的互补链，即<code>GTCCG</code>。</p>
<h2 id="entrez-direct检索">Entrez Direct检索</h2>
<p>如果我们有一个项目号(project accession)，例如<code>PRJNA257197</code>，我们就能通过EDirect来检索这个项目的一些信息。这个项目号存在于发表的文章中，或者是文件的附件里。现在我们利用<code>esearch</code>来检索一下，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">esearch -db=nucleotide -query PRJNA257197</div><div class="line">esearch -db protein -query PRJNA257197</div></pre></td></tr></table></figure>
<p>其中<code>-query</code>后面跟的是要检索的关系词。</p>
<p>结果如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ esearch -db nucleotide -query PRJNA257197</div><div class="line">&lt;ENTREZ_DIRECT&gt;</div><div class="line">  &lt;Db&gt;nucleotide&lt;/Db&gt;</div><div class="line">  &lt;WebEnv&gt;NCID_1_215173136_130.14.18.125_9001_1535597951_440051473_0MetA0_S_MegaStore&lt;/WebEnv&gt;</div><div class="line">  &lt;QueryKey&gt;1&lt;/QueryKey&gt;</div><div class="line">  &lt;Count&gt;249&lt;/Count&gt;</div><div class="line">  &lt;Step&gt;1&lt;/Step&gt;</div><div class="line">&lt;/ENTREZ_DIRECT&gt;</div><div class="line"></div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ esearch -db protein -query PRJNA257197</div><div class="line">&lt;ENTREZ_DIRECT&gt;</div><div class="line">  &lt;Db&gt;protein&lt;/Db&gt;</div><div class="line">  &lt;WebEnv&gt;NCID_1_215068217_130.14.22.215_9001_1535597965_849331092_0MetA0_S_MegaStore&lt;/WebEnv&gt;</div><div class="line">  &lt;QueryKey&gt;1&lt;/QueryKey&gt;</div><div class="line">  &lt;Count&gt;2240&lt;/Count&gt;</div><div class="line">  &lt;Step&gt;1&lt;/Step&gt;</div><div class="line">&lt;/ENTREZ_DIRECT&gt;</div></pre></td></tr></table></figure>
<p>在上面的代码中，我们使用了两条命令来检索<code>PRJNA257197</code>的信息，结果显示了这个项目的“环境”(environment)“，这个环境，我的理解就是，这个项目在Web上的一些配置信息，如下所示：</p>
<ol style="list-style-type: decimal">
<li><code>&lt;Db&gt;nucleotide&lt;/Db&gt;</code>说明这是一个核酸数据库；</li>
<li>剩下的看不懂，不知道是什么，可能是ip地址。</li>
<li><code>&lt;Count&gt;249&lt;/Count&gt;</code>表示有249个检索结果。</li>
</ol>
<h2 id="edirect提取核酸与蛋白信息">EDirect提取核酸与蛋白信息</h2>
<p>我们可以采用管道操作，将这个项目中的一些信息，例如核酸的信息下载下来，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db nucleotide -query PRJNA257197 | efetch -format=fasta &gt; genomes.fa</div></pre></td></tr></table></figure>
<p>也可以下载蛋白质的信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db protein -query PRJNA257197 | efetch -format=fasta &gt; proteins.fa</div></pre></td></tr></table></figure>
<p>查看一下相应的信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ ls</div><div class="line">AF086833.fa  AF086833.gb  genomes.fa  proteins.fa</div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ head -3 genomes.fa </div><div class="line"><span class="meta">&gt;</span>KR105345.1 Zaire ebolavirus isolate Ebola virus/H.sapiens-wt/SLE/2014/Makona-G6089.1, partial genome</div><div class="line">ATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTT</div><div class="line">GTTTCAGAGCCATATCACCAAGATAGAGAACAACCTAGGTCTCCGGAGGGGGCAAGGGCATCAGTGTGCT</div><div class="line"></div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ head -3 proteins.fa </div><div class="line"><span class="meta">&gt;</span>AKC37233.1 polymerase [Zaire ebolavirus]</div><div class="line">MATQHTQYPDARLSSPIVLDQCDLVTRACGLYSSYSLNPQLRNCKLPKHIYRLKYDVTVTKFLSDVPVAT</div><div class="line">LPIDFIVPILLKALSGNGFCPVEPRCQQFLDEIIKYTMQDALFLKYYLKNVGAQEDCVDDHFQEKILSSI</div></pre></td></tr></table></figure>
<h3 id="提取复杂信息">提取复杂信息</h3>
<p>Entrez Direct中有一个<code>xtract</code>工具，可以用于从XML文件中提取相应的信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db taxonomy -id 9606,7227,10090 -format xml | xtract -Pattern Taxon -first TaxId Scien tificName GenbankCommonName Division</div><div class="line">9606	human	Primates</div><div class="line">7227	fruit fly	Invertebrates</div><div class="line">10090	house mouse	Rodents</div></pre></td></tr></table></figure>
<p>总之，Entrez中有很多有用的工具，没事可以多看看文档。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>Biostar Handbook</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/19/BioStatistics/生物统计-16决策树/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/19/BioStatistics/生物统计-16决策树/" itemprop="url">决策树笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-19T12:02:00+08:00">
                2018-08-19
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,377
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  10
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="决策树的原理">决策树的原理</h2>
<p>决策树的原理可以参考这篇笔记<a href="http://rvdsd.top/2018/07/01/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B019-%E5%86%B3%E7%AD%96%E6%A0%91/">《StatQuest学习笔记19——决策树》</a>。</p>
<h2 id="决策树的计算过程">决策树的计算过程</h2>
<p>决策树(decision tree)用于回归时称为回归树(regression tree)，用于分类时则称为分类树(classification tree)。这里我们先看一批数据，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/Dh8fddjd9h.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>原始数据很多，有1000多行。这个数据来自于Flake and lawrence(2002)，可以在<a href="https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/regression/mg" target="_blank" rel="external">网上下载</a>，这且数据是做什么的，我也不清楚。这个数据集有6个自变量和1个因变量，一共有1385个观测值。书上说之所以用这个数据是因为这里的所有变量都是数量变量，因此这个数据集也可以用经典线性回归来做，从而可以对各种方法进行比较。</p>
<h3 id="线性回归">线性回归</h3>
<p>现在先对全部数据做一个简单的 线性回归，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">raw_dt &lt;- read.table(<span class="string">"https://raw.githubusercontent.com/20170505a/raw_data/master/mg.csv"</span>,header=<span class="literal">T</span>,sep=<span class="string">","</span>)</div><div class="line">result_dt &lt;- lm(y~.,raw_dt)</div><div class="line">summary(result_dt)</div><div class="line">pairs(raw_dt)<span class="comment">#查看不同变量之间的相关关系</span></div><div class="line"><span class="comment"># 或plot(raw_dt)</span></div><div class="line">cor(raw_dt)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; summary(result_dt)</div><div class="line"></div><div class="line">Call:</div><div class="line">lm(formula = y ~ ., data = raw_dt)</div><div class="line"></div><div class="line">Residuals:</div><div class="line">     Min       1Q   Median       3Q      Max </div><div class="line">-0.40866 -0.09516  0.01098  0.09645  0.40011 </div><div class="line"></div><div class="line">Coefficients:</div><div class="line">            Estimate Std. Error t value Pr(&gt;|t|)    </div><div class="line">(Intercept)  0.95582    0.07142  13.384  &lt; 2e-16 ***</div><div class="line">x1           0.28512    0.02773  10.282  &lt; 2e-16 ***</div><div class="line">x2           0.17263    0.02655   6.502 1.10e-10 ***</div><div class="line">x3          -0.37956    0.02769 -13.709  &lt; 2e-16 ***</div><div class="line">x4          -0.36756    0.02764 -13.298  &lt; 2e-16 ***</div><div class="line">x5           0.07165    0.02656   2.697  0.00707 ** </div><div class="line">x6           0.19003    0.02778   6.841 1.18e-11 ***</div><div class="line">---</div><div class="line">Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1</div><div class="line"></div><div class="line">Residual standard error: 0.1459 on 1378 degrees of freedom</div><div class="line">Multiple R-squared:  0.5868,	Adjusted R-squared:  0.585 </div><div class="line">F-statistic: 326.2 on 6 and 1378 DF,  p-value: &lt; 2.2e-16</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/egLil2EEmc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>mg数据各个变量之间的相关系数如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">&gt; cor(raw_dt)</div><div class="line">            y         x1         x2         x3         x4         x5         x6</div><div class="line">y   <span class="number">1.0000000</span>  <span class="number">0.2620280</span>  <span class="number">0.5441721</span> -<span class="number">0.5006887</span> -<span class="number">0.4355964</span>  <span class="number">0.5510561</span>  <span class="number">0.1511016</span></div><div class="line">x1  <span class="number">0.2620280</span>  <span class="number">1.0000000</span> -<span class="number">0.1465800</span> -<span class="number">0.6543690</span>  <span class="number">0.4333849</span>  <span class="number">0.4543362</span> -<span class="number">0.6277910</span></div><div class="line">x2  <span class="number">0.5441721</span> -<span class="number">0.1465800</span>  <span class="number">1.0000000</span> -<span class="number">0.1475594</span> -<span class="number">0.6539043</span>  <span class="number">0.4367584</span>  <span class="number">0.4515753</span></div><div class="line">x3 -<span class="number">0.5006887</span> -<span class="number">0.6543690</span> -<span class="number">0.1475594</span>  <span class="number">1.0000000</span> -<span class="number">0.1471907</span> -<span class="number">0.6537680</span>  <span class="number">0.4397873</span></div><div class="line">x4 -<span class="number">0.4355964</span>  <span class="number">0.4333849</span> -<span class="number">0.6539043</span> -<span class="number">0.1471907</span>  <span class="number">1.0000000</span> -<span class="number">0.1494223</span> -<span class="number">0.6524989</span></div><div class="line">x5  <span class="number">0.5510561</span>  <span class="number">0.4543362</span>  <span class="number">0.4367584</span> -<span class="number">0.6537680</span> -<span class="number">0.1494223</span>  <span class="number">1.0000000</span> -<span class="number">0.1498846</span></div><div class="line">x6  <span class="number">0.1511016</span> -<span class="number">0.6277910</span>  <span class="number">0.4515753</span>  <span class="number">0.4397873</span> -<span class="number">0.6524989</span> -<span class="number">0.1498846</span>  <span class="number">1.0000000</span></div></pre></td></tr></table></figure>
<p>从前面的线性回归结果可以看出来，<span class="math inline">\(R^2=0.5868\)</span>，这个数值并不太好，再加上两两散点图与相关系数的结果，我们看不出各个变量之间的任何模式。现在我们使用其它的几种算法建模方法，同时使用交叉验证（利用均方误差或标准化均方误差）对各种方法时行比较。</p>
<p>下面我们对各种方法都用10折交叉验证的方法来判断其结果的可靠性（关于10折交叉验证的一些知识，可以看这笔记<a href="http://rvdsd.top/2018/07/05/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B022-%E4%BA%A4%E5%8F%89%E9%AA%8C%E8%AF%81/">《StatQuest学习笔记22——交叉验证》</a>）。对于每种方法，按照随机建立的10个训练集通过计算建立10个模型，对测试集分别得到10个标准化均方误差(NMSE)，再得出10次平均的NMSE。令<span class="math inline">\(\bar{y}\)</span>为因变量均值，<span class="math inline">\(\hat{y}\)</span>为从训练集得到的模型对一个测试数据集的预测值，这里NMSE的定义如下所示：</p>
<p><span class="math inline">\(NMSE=\bar{(y-\hat{y})^2}/=\bar{(y-\bar{y})^2}=\sum(y-\hat{y})^2/\sum(y-\bar{y})^2\)</span></p>
<p>显然，如果什么模型都不用，仅仅用均值来做预测，那么NMSE应该等于1.所以，如果在回归时得到NMSE大于1，这个模型就很糟糕了，还不如没有模型。仅仅对于训练集来说，其NMSE等于<span class="math inline">\(1-R^2\)</span>，这里的<span class="math inline">\(R^2\)</span>为回归的确定系数，但是对于测试集来说，其NMSE与测试集回归的<span class="math inline">\(R^2\)</span>没有什么关系，交叉验证主要关心测试集的NMSE。</p>
<h3 id="构建训练集和测试集">构建训练集和测试集</h3>
<p>现在构建为交叉验证服务的10个训练集和测试集，具体过程是：随机把下标分配给1,2,…10这10个数字，也就是把数据下标随机分成10份，然后每次提取一份作为测试集，其它9份放在一起作为训练集，用模型进行拟合，记下结果和误差，如此下去，一共做10次，最后把误差平均起来。这里所用的随机选Z折下标集的函数如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">CV=<span class="keyword">function</span>(n,Z=<span class="number">10</span>,seed=<span class="number">1000</span>)&#123;</div><div class="line">  z=rep(<span class="number">1</span>:Z,ceiling(n/Z))[<span class="number">1</span>:n]</div><div class="line">  set.seed(seed)</div><div class="line">  z=sample(z,n)</div><div class="line">  mm=list()</div><div class="line">  <span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:Z)</div><div class="line">    mm[[i]]=(<span class="number">1</span>:n)[z==i]</div><div class="line">  <span class="keyword">return</span>(mm)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>这里的<code>n</code>是样本量，<code>Z</code>为折的数目（也就是训练集加测试集的数目，这里默认是10），<code>seed</code>为随机种子，这里设为了1000，转出为<code>Z</code>个下标集，<code>mm[[i]]</code>为第i（i=1,2,…,Z)个下标集，根据这个函数对这个mg数据集求10个下标集的代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">n&lt;-nrow(raw_dt)</div><div class="line">Z=<span class="number">10</span></div><div class="line">mm=CV(n,Z)</div><div class="line">D=<span class="number">1</span></div></pre></td></tr></table></figure>
<p>在这段代码中，<code>mm</code>存储了10个下标集，而D说明因变量是第一个变量，这些值在后面对每种方法都要用。</p>
<p>为了和其它方法进行地比较，先对数据mg求简单的线性回归的10折交叉验证的测试集的NMSE，代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; mean(MSE)</div><div class="line">[1] 0.4222689</div></pre></td></tr></table></figure>
<p>结果表明，测试集的NMSE为0.4222689</p>
<h2 id="决策树回归">决策树回归</h2>
<p>决策树(decision tree)用于回归时称为回归树(regression tree)，用于分类时则称为分类树(classification tree)，现在我们先用数据集mg做一棵回归树，然后解释其意义。这里要使用到<code>rpart</code>包中的<code>rpart()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rpart)</div><div class="line"><span class="keyword">library</span>(rpart.plot)</div><div class="line">a=rpart(y~.,raw_dt)</div><div class="line">rpart.plot(a,type=<span class="number">2</span>,faclen=<span class="number">0</span>)</div></pre></td></tr></table></figure>
<p>结果如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">&gt; a</div><div class="line">n= 1385 </div><div class="line"></div><div class="line">node), split, n, deviance, yval</div><div class="line">      * denotes terminal node</div><div class="line"></div><div class="line"> 1) root 1385 70.9545100 0.9301440  </div><div class="line">   2) x2&lt; 0.9894983 736 31.9699900 0.8017888  </div><div class="line">     4) x3&gt;=0.8423859 551 16.4573000 0.7288982  </div><div class="line">       8) x2&gt;=0.5178963 483 13.3868900 0.7068909  </div><div class="line">        16) x2&lt; 0.7999875 263  4.7064590 0.6422662  </div><div class="line">          32) x3&lt; 1.092533 185  3.3567710 0.6073434 *</div><div class="line">          33) x3&gt;=1.092533 78  0.5889238 0.7250958 *</div><div class="line">        17) x2&gt;=0.7999875 220  6.2689860 0.7841468 *</div><div class="line">       9) x2&lt; 0.5178963 68  1.1749050 0.8852150 *</div><div class="line">     5) x3&lt; 0.8423859 185  3.8660570 1.0188850  </div><div class="line">      10) x4&gt;=1.062353 71  1.8511830 0.9239636 *</div><div class="line">      11) x4&lt; 1.062353 114  0.9767496 1.0780020 *</div><div class="line">   3) x2&gt;=0.9894983 649 13.1077400 1.0757060  </div><div class="line">     6) x3&gt;=1.056185 206  3.5797200 0.9644841  </div><div class="line">      12) x4&gt;=0.9107852 23  0.5746155 0.7823564 *</div><div class="line">      13) x4&lt; 0.9107852 183  2.1462980 0.9873744 *</div><div class="line">     7) x3&lt; 1.056185 443  5.7947760 1.1274250 *</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/hG48FgC33C.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>计算的结果有2个，一个是文本描述的结果，一个是图形结果，这两个结果是等价的，我们可以了解这个决策树的构造过程。现在解释一下这个决策树。</p>
<p>决策树就像一棵倒长的树，有很多分叉，分叉点叫节点(node)，其中1号节点为根节点，也就是上图中最高的那个节点，输出的文本信息则是全部数据的信息，其中在1号节点上，n=1385，偏差(deviance)=70.9545100，在该节点的因变量的均值为0.9301440。</p>
<p>这里的偏差是批在每个节点的偏差，它的定义为：在该节点的数据中的因变量与其均值之差的平方和，即<span class="math inline">\(\sum_{i}(y_{i}-\bar{y})^2\)</span>这里回归树用偏差大小来选择变量，当然也可以用其它诸如残差平方和作为选择变量的准则。</p>
<p>上述的这些信息显示在上图中（不过只显示了均值），然后计算机程度在每个自变量中选择一个数值分割点，使得在该节点的整个数据在这个分割点比在其它分割点分成两部分之后的偏差之和都小，对每个自变量都选择这样的一个分割点，并对不同自变量的分割结果做比较，得到一个使得分割后总偏差最小的变量，这个变量就是该节点的分割变量（或称为拆分变量）。</p>
<p>根据输出的文本信息，在这个案例中，根节点的分割变量为x2，其分割点为0.9894983，x2小于该值的观测值分到左边（节点2），大于该值的则分到右边（节点3）。在每个节点都展示出观测值数目、偏差及因变量均值这三项内容（不过决策树的图形中只有均值），这样数据就分成了两个部分。然后，对每一部分数据，选择分割变量及其分割点的程序重新开始。如此下去，根据一些按照拟合程度和避免决策树无限制长下去的准则，到一定节点就不再分割，这些节点就是终节点（也称为叶节点），也就是文本信息中带有星号(<code>*</code>)的为叶节点。</p>
<h3 id="决策树的使用">决策树的使用</h3>
<p>使用决策树很简单，如果有新的（只知道自变量的）观测值，就可以根据决策树以及各个变量大小，从根节点一直走到一个终节点，那里因变量的均值就是该观测值的因变量预测值。决策树本身受数据影响较大，但许多决策树结果起来就可以构成很好的方法，例如boosting，bagging以及随机森林等组合方法就是以决策树作为基本决策单元构造出来的。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>复杂数据统计方法：基于R的应用.第三版.吴喜之</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/19/BioStatistics/生物统计-17聚类分析笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/19/BioStatistics/生物统计-17聚类分析笔记/" itemprop="url">聚类分析笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-19T12:02:00+08:00">
                2018-08-19
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  4,942
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  19
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>以前看过StatQuest的聚类分析笔记，可以参考这篇笔记<a href="http://rvdsd.top/2018/06/29/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B017-%E8%81%9A%E7%B1%BB/">《StatQuest学习笔记17——聚类》</a>。</p>
<h2 id="聚类的概率与类型">聚类的概率与类型</h2>
<h3 id="聚类分析法的概念">聚类分析法的概念</h3>
<p>聚类分析法(cluster analysis)是研究“物以类聚”的一种现代统计分析方法，聚类的目的是把分类对象按照一定的规则分成若干类，这些类不是事先给定的，而是根据数据的特征确定的，在同一类中这些对象在某种意义上趋向于彼此相似，而在不同类中对象趋于不相似。</p>
<h3 id="聚类分析法的类型">聚类分析法的类型</h3>
<p>在实际问题中，经常要对一些东西进行分类，例如在古生物研究中，通过挖掘出来的一些骨髓的形状和大小对它们进行科学的分类，在这里，骨髓的形状和大小是我们用来分类的依据，称为指标或变量，用<span class="math inline">\(X_{1},X_{2},\cdots,X_{p}\)</span>表示，其中，p是变量的个数，需要进行分类的骨髓为样品，用1,2,3,…n表示，其中n是样本的个数，聚类分析的数据结果见下表：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/1F0eg5gCJd.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>在聚类分析中，基本的思想是认为所研究的样本或变量存在着程度不同的相似性（亲疏关系）。于是，根据一批样品中的多个观测指标，具体找出一些能够度量样本之间相似程度的统计量，以这些统计量为划分类型的依据，把一些相似程度较大的样本聚合为一类，把另外一些彼此之间相似程度较大的样品又聚合为另一类，关系密切的聚合到一个小的分类单位，关系疏远的聚合到一个大的分类单位，直到把所有的样本都又聚合完毕，把不同类型一一划分出来，形成一个由小到大的分类系统，最后把整个分类系统画成一张聚类图，用它把所有样本间的亲疏关系表示出来。</p>
<p>通常根据分类对象的不同，可以将聚类分为两类：一类是对样品进行分类处理，叫Q型；一类是对变量进行分类处理，叫R型。Q型聚类又叫样品分类，就是对观测对象进行聚类，是根据被观测的对象各种特征进行分类。在生物信息学中，经常使用的是R型，例如有对照组1，2，3三个样本，实验组1，2，3这三个样本，把这6个样本把基因表达谱的数据进行聚类。</p>
<h3 id="聚类统计量">聚类统计量</h3>
<p>聚类分手析的基本原则是将有较大相似性的对象归为同一类，而将差异较大的个体归入不同的类。为了将样本聚类，就需要研究样品之间的关系，一种方法是将每一个样本看做p维空间的一个点，并在空间定义距离，距离较近的点归为一类，距离较远的点属于不同的类。对变量通常计算它们的相似系数，性质越接近的变量，它们的相似系数越接近于1或-1，彼此无关的变量的相似系数越接近于0。可进行聚类的统计量有距离和相似系数。距离有欧氏距离，马氏距离，兰氏距离，相似系数有夹角余弦，相关系数。</p>
<p>对样品进行聚类时，我们将把样品间的“靠近”程度用某种距离来刻画对指标的聚类，往往用某种相似系数来刻画。当选用n个样本，p个指标时，就可以得到一个<span class="math inline">\(n\times p\)</span>的数据矩阵，即<span class="math inline">\(X=(x_{ij})n \times p\)</span>，该距离的元素<span class="math inline">\(x_{ij}\)</span>表示第i个样本的第j个变量值。</p>
<h4 id="距离和相似系数">距离和相似系数</h4>
<p>对样品或变量进行分类时，我们常用距离和相似系数对样品或变量之间的相似性进行度量。距离常用来度量样本之间的相似性，而相似系数常用来度量变量间的相似性。</p>
<p>距离多用于样品的分类，信<span class="math inline">\(d_{ij}\)</span>表示样本<span class="math inline">\(x_{i}\)</span>和<span class="math inline">\(x_{j}\)</span>的距离，一般要求<span class="math inline">\(d_{ij}\)</span>满足以下四个条件，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/fDEh2L8bCJ.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>在聚类分析中并不严格要求定义的距离都满足这四条，一般来说前三条大部分是能满足的，一些不能满足(4)，但是在广义的角度上也称其为距离。</p>
<p>设<span class="math inline">\(x_{ij}(i=1,2,\cdots,n;j=1,2,\cdots,p)\)</span>为第i个样本的第j个指标的观测数据。即若每个样本有p个变量，则每个样本都可以看成p维空间中的一个点，n个样品就是p维空间中的n个点，定义<span class="math inline">\(d_{ij}\)</span>为样本xj和xj的距离，于是得到一个<span class="math inline">\(n\times n\)</span>的距离矩阵，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/hmL6biif2K.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>样品聚类都是基于此距离矩阵进行的，现在我们看一个简单的案例：</p>
<h4 id="距离案例a样本聚类">距离：案例A：样本聚类</h4>
<p>以下列举5个观察值，2个变量数据的平面散点图，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">x1 &lt;- c(<span class="number">5</span>,<span class="number">7</span>,<span class="number">3</span>,<span class="number">6</span>,<span class="number">6</span>)</div><div class="line">x2 &lt;- c(<span class="number">7</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">5</span>,<span class="number">6</span>)</div><div class="line">plot(x1,x2)</div><div class="line">text(x1,x2,labels=c(<span class="number">1</span>:<span class="number">5</span>),adj=-<span class="number">.5</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/6G46d2cDj3.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>由于只有2个变量，所以从散点图上就可以直观地将这5个样本分为几类，但当变量较多时，这种方法显著是不行的。为了计算平面上各点之间的距离<span class="math inline">\(d_{ij}\)</span>，在聚类分析中对连续型变量常用的距离有以下3种，如下所示：</p>
<h5 id="明氏距离minkowski">明氏距离(Minkowski):</h5>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/FE2B6E29Jk.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h5 id="马氏距离mahalanobi">马氏距离(Mahalanobi)</h5>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/egafd7I0j7.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h5 id="兰氏距离canberra">兰氏距离(Canberra)</h5>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/L8854g7kKJ.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>下面我们使用<code>dist()</code>函数计算一下案例A的欧氏距离与马氏距离，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">dist(x,mdehtod=<span class="string">"euclidean"</span>,diag=<span class="literal">FALSE</span>,upper=<span class="literal">FALSE</span>,p=<span class="number">2</span>)</div><div class="line"><span class="comment"># x是数据矩阵</span></div><div class="line"><span class="comment"># method是计算方法，包括euclidean,maximum,manhattan,canberra,binary,minkowski</span></div><div class="line"><span class="comment"># diag为是否包含对角线元素</span></div><div class="line"><span class="comment"># upper为是否需要上三角</span></div><div class="line"><span class="comment"># p为Minkowski距离的幂次</span></div></pre></td></tr></table></figure>
<p>案例A的计算结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line">&gt; X&lt;-as.matrix(data.frame(x1,x2))</div><div class="line">&gt; dist(X) <span class="comment"># 默认情况下是欧氏距离</span></div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span>                           </div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span>                  </div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span>         </div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span></div><div class="line">&gt; dist(X,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>) <span class="comment"># 默认情况下是欧氏距离</span></div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span>        <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0.000000</span> <span class="number">6.324555</span> <span class="number">5.385165</span> <span class="number">2.236068</span> <span class="number">1.414214</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span> <span class="number">0.000000</span> <span class="number">4.123106</span> <span class="number">4.123106</span> <span class="number">5.099020</span></div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span> <span class="number">0.000000</span> <span class="number">4.242641</span> <span class="number">5.000000</span></div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span> <span class="number">0.000000</span> <span class="number">1.000000</span></div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span> <span class="number">0.000000</span></div><div class="line">&gt; dist(X,method=<span class="string">"manhattan"</span>,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>) <span class="comment">#马氏距离</span></div><div class="line">  <span class="number">1</span> <span class="number">2</span> <span class="number">3</span> <span class="number">4</span> <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0</span> <span class="number">8</span> <span class="number">7</span> <span class="number">3</span> <span class="number">2</span></div><div class="line"><span class="number">2</span> <span class="number">8</span> <span class="number">0</span> <span class="number">5</span> <span class="number">5</span> <span class="number">6</span></div><div class="line"><span class="number">3</span> <span class="number">7</span> <span class="number">5</span> <span class="number">0</span> <span class="number">6</span> <span class="number">7</span></div><div class="line"><span class="number">4</span> <span class="number">3</span> <span class="number">5</span> <span class="number">6</span> <span class="number">0</span> <span class="number">1</span></div><div class="line"><span class="number">5</span> <span class="number">2</span> <span class="number">6</span> <span class="number">7</span> <span class="number">1</span> <span class="number">0</span></div><div class="line">&gt; dist(X,method=<span class="string">"minkowski"</span>,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>,p=<span class="number">2</span>)</div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span>        <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0.000000</span> <span class="number">6.324555</span> <span class="number">5.385165</span> <span class="number">2.236068</span> <span class="number">1.414214</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span> <span class="number">0.000000</span> <span class="number">4.123106</span> <span class="number">4.123106</span> <span class="number">5.099020</span></div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span> <span class="number">0.000000</span> <span class="number">4.242641</span> <span class="number">5.000000</span></div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span> <span class="number">0.000000</span> <span class="number">1.000000</span></div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span> <span class="number">0.000000</span></div></pre></td></tr></table></figure>
<h4 id="相似系数">相似系数</h4>
<p>对两个变量之间的相似程度可以用相似系数来刻画，用<span class="math inline">\(C_{ij}\)</span>表示第i个变量与第j个变量之间的相似系数，<span class="math inline">\(C_{ij}\)</span>的绝对值越接近1，表示指标i与指标j的关系越密切，<span class="math inline">\(C_{ij}\)</span>的绝对值越接近于0，表示指标i与指标j的关系越疏远，常用的相似系数有夹角余弦和相关系数，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/15HbGggiJm.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>距离和相似系数之间的转换，一般来说，距离越小，两样本之间关系越密切，而相似系数越大，两变量之间的关系越密切，为了聚类分析方便起见，可以用下面的通用公式得到变量间的距离，即<span class="math inline">\(d_{ij}^2=1-C_{ij}^2\)</span>。</p>
<h2 id="系统聚类法">系统聚类法</h2>
<h3 id="系数聚类法的基本思想">系数聚类法的基本思想</h3>
<p>确定了距离和相似系数后就要进行分类。分类有许多种方法，最常用的一种方法是在样品距离的基础上定义类与类之间的距离。首选将n个样品分成n类，每个样品自成成一类，然后每将将具有最小距离的两类合并，合并后重新计算类与类之间的距离，这个过程一直持续到将所有的样品归为一类为止，并把这个过程画成一张聚类图，由聚类图可以方便地进行分类。因为聚类图很像一张系统图，所以这种方法就叫层次聚类法(hierachical clustering method)。层次聚类方法是目前应用最多的一种方法。从上面的分析可以看出，虽然已经给出了计算样品之间的距离，但在实际计算过程中还要定义类与类之间的距离。如何定义类与类之间的距离，也有很多方法，不同的方法就产生了不同的层次聚类，常用的方法有六种：</p>
<ol style="list-style-type: decimal">
<li>最短距离法：类与类之间的距离等于两类最近样品之间的距离；</li>
<li>最长距离法：类与类之间的距离等于两类最远样品之间的距离；</li>
<li>类平均法：类与类之间的距离等于各类元素两两之间平方距离的平均；</li>
<li>重心法：类与类之间的距离定义为对应这两类重心之间的距离。对样品分类来说，每一类的类重心就是该类样品的均值。</li>
<li>中间距离法：最长距离夸大了类间距离，最短距离低估了类间距离，介于两者之间的距离法即为中间距离法，中间距离法则是采用了介于最远和最近之间的距离。</li>
<li>离差平方和法（Ward法）：基于方差分析的思想，如果类分得正确，同样样品之间的离差平方和应当较小，类与类之间的离差平方和应该较大。</li>
</ol>
<h3 id="层次聚类法的计算公式">层次聚类法的计算公式</h3>
<h4 id="最短距离法">最短距离法</h4>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/GjJ89mm9ak.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="最长距离法">最长距离法</h4>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/EllAL3e95K.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="重心法">重心法</h4>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/mi72GAm1Eh.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="中间距离法">中间距离法</h4>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/j94bFkD3Gb.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="类平均法">类平均法</h4>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/ikKEd47fIh.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="离差平方和法ward法">离差平方和法(Ward法)</h4>
<p>这种方法是Ward提出来的，因此称为Ward法，该方法的基本思想来自于方差分析，如果分类正确，同样样品的离差平方和应当较小，类与类的离差平方和应当较大。具体做法是先将n个样品各自归为一类，然后每次缩小一类，每缩小一类，离差平方和就要增大，选择使方差增加最小的两类合并，直接所有的样本都归为一类为止。计算过程如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/AHHeFa9cI9.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/0iaFhF3HA3.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>这六种层次聚类法的并类原则和过程完全相同，不同之处在于类与类之间的距离定义。当采用欧氏距离时，Lance和Williams于19678年将这些方法统一成如下的递推公式，即：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/HDddlmkB4d.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/df257DGccL.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="层次聚类法的基本步骤">层次聚类法的基本步骤</h4>
<ol style="list-style-type: decimal">
<li>计算n个样本两两的距离<span class="math inline">\(\{d_{ij}\}\)</span>，记作D；</li>
<li>构造n个类，每个类只包含一个样品；</li>
<li>合并距离最近的两类为一个新类；</li>
<li>计算新类与当前各类的距离，若类个数为1，进行第5步，否则回到第3步；</li>
<li>画聚类图；</li>
<li>决定类的个数和类。</li>
</ol>
<h4 id="层次聚类的函数">层次聚类的函数</h4>
<p>R中进行层次聚类的函数是<code>hclust()</code>，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">hclust(d,mehtod=<span class="string">"complete"</span>,<span class="keyword">...</span>)</div><div class="line"><span class="comment"># d为相似矩阵</span></div><div class="line"><span class="comment"># method是层次聚类方法，包括ward,single,complete,average,mcquitty,median,centroid</span></div></pre></td></tr></table></figure>
<p>现在使用案例A中的数据进行层次聚类。开始有5类，即每个样品自成一类，即<span class="math inline">\(\{G_{1},G_{2},G_{3},G_{4},G_{5}\}\)</span>，这5类之间的距离就等于5个样品之间的距离，距离矩阵记为<span class="math inline">\(D_{0}\)</span>，其中最小的元素为<span class="math inline">\(D_{0}(4,5)=1\)</span>，因此将<span class="math inline">\(G_{4}\)</span>和<span class="math inline">\(G_{5}\)</span>聚合并成一个新类，即<span class="math inline">\(G_{6}=\{G_{4},G_{5}\}\)</span>，然后再过计算<span class="math inline">\(G_{6}\)</span>与$G_{1},G_{2},G_{3}#之间的距离。</p>
<p>应用公式<span class="math inline">\(G_{1}(6,i)=min{D_{0}(4,i),D_{0}(5,i)}\)</span>，求其最近相邻的距离，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/Jf1chaHL76.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>同理可知，其它类与类之间的距离如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/kHk4C3Bm8j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在绘制层次聚类图，如下所示；</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; hc &lt;-hclust(dist(X),<span class="string">"single"</span>)</div><div class="line">&gt; cbind(hc$merge,hc$height)</div><div class="line">     [,<span class="number">1</span>] [,<span class="number">2</span>]     [,<span class="number">3</span>]</div><div class="line">[<span class="number">1</span>,]   -<span class="number">4</span>   -<span class="number">5</span> <span class="number">1.000000</span></div><div class="line">[<span class="number">2</span>,]   -<span class="number">1</span>    <span class="number">1</span> <span class="number">1.414214</span></div><div class="line">[<span class="number">3</span>,]   -<span class="number">2</span>    <span class="number">2</span> <span class="number">4.123106</span></div><div class="line">[<span class="number">4</span>,]   -<span class="number">3</span>    <span class="number">3</span> <span class="number">4.123106</span></div><div class="line">&gt; plot(hc)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/06aFAJl8CA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>ward法（采用欧氏距离），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">hc2 &lt;-hclust(dist(X),<span class="string">"ward"</span>)</div><div class="line">cbind(hc2$merge,hc$height)</div><div class="line">&gt; cbind(hc2$merge,hc$height)</div><div class="line">     [,<span class="number">1</span>] [,<span class="number">2</span>]     [,<span class="number">3</span>]</div><div class="line">[<span class="number">1</span>,]   -<span class="number">4</span>   -<span class="number">5</span> <span class="number">1.000000</span></div><div class="line">[<span class="number">2</span>,]   -<span class="number">1</span>    <span class="number">1</span> <span class="number">1.414214</span></div><div class="line">[<span class="number">3</span>,]   -<span class="number">2</span>   -<span class="number">3</span> <span class="number">4.123106</span></div><div class="line">[<span class="number">4</span>,]    <span class="number">2</span>    <span class="number">3</span> <span class="number">4.123106</span></div><div class="line">plot(hc2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/6lm49eiDef.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="案例分析">案例分析</h3>
<p>为了研究我们31个省、市、自治区2001年城镇居民生活消费的分布规律，根据调查资料作区域消费类型划分，其中变量如下所示：</p>
<p>X1：人均食品支出；</p>
<p>X2：人均衣着商品支出</p>
<p>X3：人均家庭设备用品及服务支出</p>
<p>X4：人均医疗保健支出</p>
<p>X5：人均交通和通信支出</p>
<p>X6：人均娱乐教育文化服务支出</p>
<p>X7：人均居住支出</p>
<p>X8：人均杂项商品中和服务支出</p>
<p>表格如下所示（完整的数据在github上）：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/JekccFJc4j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在使用六种聚类方法进行聚类，如下所示：</p>
<p>第1种：欧氏距离，euclidean</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">x_raw&lt;-read.table(<span class="string">"https://raw.githubusercontent.com/20170505a/raw_data/master/wbh_cluster_072.txt"</span>,header=<span class="literal">T</span>,sep=<span class="string">","</span>)<span class="comment"># 完整数据</span></div><div class="line">rownames(x_raw)&lt;-as.vector(x_raw$X)</div><div class="line">x_raw&lt;-x_raw[,-<span class="number">1</span>]</div><div class="line"></div><div class="line">x_dist1&lt;-dist(x_raw,method=<span class="string">"euclidean"</span>)</div><div class="line">hc1 &lt;-hclust(x_dist1,<span class="string">"single"</span>)</div><div class="line">fviz_dend(hc1) <span class="comment">#factoextra包</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/1CGDL9cH1I.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>还可以对分成几类进行颜色区分，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fviz_dend(hc1, k = <span class="number">4</span>, </div><div class="line">          cex = <span class="number">0.5</span>, </div><div class="line">          k_colors = c(<span class="string">"#2E9FDF"</span>, <span class="string">"#00AFBB"</span>, <span class="string">"#E7B800"</span>, <span class="string">"#FC4E07"</span>),</div><div class="line">          color_labels_by_k = <span class="literal">TRUE</span>, </div><div class="line">          rect = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/lHbiGFIeEi.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第二种：欧氏距离，complete</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc2 &lt;-hclust(x_dist1,<span class="string">"complete"</span>)</div><div class="line">fviz_dend(hc2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/46LK4cfLKm.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第三种：欧氏距离,euclidean,median，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc3 &lt;-hclust(x_dist1,<span class="string">"median"</span>)</div><div class="line">fviz_dend(hc3)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/bLH7a7Hk3e.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第四种：欧氏距离，euclidean，average，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc4 &lt;-hclust(x_dist1,<span class="string">"average"</span>)</div><div class="line">fviz_dend(hc4)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/h0dFL4DH7a.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第五种：欧氏距离，centroid，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc5 &lt;-hclust(x_dist1,<span class="string">"centroid"</span>)</div><div class="line">fviz_dend(hc5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/2m74EhedAA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第六种：欧氏距离，ward，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc6 &lt;-hclust(x_dist1,<span class="string">"ward"</span>)</div><div class="line">fviz_dend(hc6)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/fGbleIfG2c.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>划分类别，如下所示：：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fviz_dend(hc6, k = <span class="number">3</span>, </div><div class="line">          cex = <span class="number">0.5</span>, </div><div class="line">          k_colors = c(<span class="string">"#2E9FDF"</span>, <span class="string">"#00AFBB"</span>, <span class="string">"#E7B800"</span>),</div><div class="line">          color_labels_by_k = <span class="literal">TRUE</span>, </div><div class="line">          rect = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/03fegDE3Ig.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="kmeans聚类法">kmeans聚类法</h2>
<h3 id="kmeans聚类法的概念">kmeans聚类法的概念</h3>
<p>层级聚类法需要计算出不同样本或变量的距离，还要在聚类的每一步计算出类间距离，相应的计算量比较大，尤其是当样本量比较大的时候，非常消耗计算机资源。而kmeans是一类快速聚类法。kmeans法（K均值法）的具体算法至少包括以下三个步骤：</p>
<ol style="list-style-type: decimal">
<li>将所有的样本分成k个初始类；</li>
<li>通过欧氏距离将某个样本划入离中心最近的类中，并对获得的样本与失去样品的类重新计算中心坐标</li>
<li>重复步骤2，直到所有的样品都不能再分配为止。</li>
</ol>
<p>kmeans法和层级聚类法一样，都是以距离的远近亲疏为标准进行聚类的。但两者的不同之处在于：层级聚类对没的类数产生一系列的聚类结果，而K均值法只能产生指定类数的聚类结果。具体类数的确定，离不开实践经验的积累。有时候也要借助层级聚类法，以一部分样本为对象进行聚类，其结果为Kmeans法提供参考。</p>
<h3 id="kemans聚类的原理与计算">kemans聚类的原理与计算</h3>
<p>kmeans算法以k为参数，把n个对象分为k个聚类，以使聚类内具有较高的相似度，而聚类间的相似度较低。相似度的计算是根据一个聚类中对象的均值来进行的。kmeans算法的处理流程如下所示：</p>
<ol style="list-style-type: decimal">
<li>随机地选择k个对象，每个对象初始地代表了一个簇的平均值或中心，对剩余的每个对象，根据其与各个聚类中心的距离将其赋给最近的簇；</li>
<li>重新过计算每个簇的平均值作为聚类中心进行聚类。</li>
</ol>
<p>这个过程不断重复，直到准则函数收敛，通常，采用平方误差准则，其定义如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/6mcbdGI5m2.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>其中，E为数据中所有对象与相应聚类中心的均方差之和，p为代表对象空间中的一个点，<span class="math inline">\(m_{i}\)</span>为类<span class="math inline">\(C_{i}\)</span>的均值（p和<span class="math inline">\(m_{i}\)</span>都是多维的）。该式所示聚类的标准旨在使所有获得的聚类有以下特点：</p>
<p>各本身尽可能地紧凑，而各类之间尽可能地分开，kmean迭代图例如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/39JJc44BHe.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>根据聚类中的均值进行聚类划分的kmeans算法如下：</p>
<ol style="list-style-type: decimal">
<li>从n个数据对象中取任意k个对象作为初始簇中心；</li>
<li>循环下述流程3到4，直到每个聚类不再发生变化为止；</li>
<li>根据每个簇中对象的均值（中心对象），计算每个对象与这些中心对象的距离，并根据最小距离重新对相应对象进行划分；</li>
<li>重新计算每个（有变化）簇的均值。</li>
</ol>
<h3 id="kmeans">kmeans()</h3>
<p>R中的kmeans()函数用于进行kmeans聚类，用法为：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">kmeans(x,centers,..)</div><div class="line">其中，x为数据矩阵或数据框</div><div class="line">centers为聚类数或初始聚类中心</div></pre></td></tr></table></figure>
<h3 id="案例分析-1">案例分析</h3>
<p>kmeans算法的R语言实现及模拟。本例模拟正态随机变量<span class="math inline">\(x~N(\mu,\delta^2)\)</span>。</p>
<ol style="list-style-type: decimal">
<li>用R模拟1000个均值为0，标准差为0.3的正态分布随机数据，再把这些随机数转换为10个变量、100个对象的矩阵；</li>
<li>用同样的方法模拟1000个均值为1，标准差为0.3的正态分布随机数，再转化为10个变量、100个对象的矩阵；</li>
<li>把这两个矩阵合并为10个变量、200个样本的数据矩阵；</li>
<li>利用kmeans聚类法将其聚成两类，观察其效果，如下所示：</li>
</ol>
<p>先看一下层级聚类：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/gl6f4DCm0c.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>再看一下kmeans聚类：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(factoextra)</div><div class="line">cl&lt;-kmeans(x_norm,<span class="number">2</span>)</div><div class="line">fviz_cluster(cl,data=x_norm)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/Id6A8cAl0I.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>我们通过<code>fviz_nbclust</code>函数查看一下聚成几类比较好，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_norm2&lt;-scale(x_norm)</div><div class="line">fviz_nbclust(x_norm2, kmeans, method = <span class="string">"wss"</span>) + geom_vline(xintercept = <span class="number">4</span>, linetype = <span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/jKBHBmKiJg.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>从图上可以看出，聚成2类比较好，因为1到2变化幅度最大，2之后的变化幅度就非常小了。</p>
<p>可视化二：可以将不同的聚类呈现出来：</p>
<p>使用<code>silhouette</code>函数来计算一下要绘图的轮廓。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">x.res &lt;- kmeans(x_norm, <span class="number">2</span>)</div><div class="line">sil &lt;- silhouette(km.res$cluster, dist(x_norm))</div><div class="line">fviz_silhouette(sil)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/icH4baljaI.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="其它r语言聚类函数">其它R语言聚类函数</h2>
<p>eclust()是另外的一个聚类函数，此函数简化了聚类分析的工作流程，可以用于计算层次聚类和分区聚类，eclust()自动计算最佳聚类簇数。 自动提供Silhouette plot，可以结合ggplot2绘制图形，使用eclust()的K均值聚类如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_eclust &lt;- eclust(x_norm,<span class="string">"hclust"</span>)</div><div class="line">fviz_dend(x_eclust,rect=<span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/Dc1mFb1KEi.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在计算一下前面的那个案例，也就是各个省市的聚类，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_eclust2 &lt;-eclust(x_raw,<span class="string">"hclust"</span>,k=<span class="number">4</span>)</div><div class="line">fviz_dend(x_eclust2,rect=<span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/5cG9k0918F.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>各省市的Kmean聚类图，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<h2 id="代码汇总">代码汇总</h2>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><p>多元统计分析及R语言建模.王斌会</p></li>
<li><p><a href="https://blog.csdn.net/Anna_datahummingbird/article/details/79912348" target="_blank" rel="external">R语言聚类分析-kmeans聚类分析实战</a></p></li>
<li><p><a href="https://mp.weixin.qq.com/s/M_rJqcHz6HUlGIHVeN4RkQ" target="_blank" rel="external">R语言学习笔记之聚类分析</a></p></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/18/BioStatistics/生物统计-15判别分析/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/18/BioStatistics/生物统计-15判别分析/" itemprop="url">判别分析笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-18T12:02:00+08:00">
                2018-08-18
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,316
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  5
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>一般的回归是指因变量为定量变量的情况，而logistic回归主要处理因变量为只取两个值的定性变量情况。如果作为因变量的定性变量取多于两个值的时候，就需要使用线性判别分析(linear discriminant analysis)来处理。判别分析根据自变量来预测因变量的类型。</p>
<h2 id="判别分析原理">判别分析原理</h2>
<p>差别分析是利用部分已知类别的观测数据来建立模型，这样的数据称为训练集(training set)。我们这里介绍的线性判别分析所得到的模型是一组线性差别公式；每个公式是自变量的一个线性组合；公式的个数和类别数目一样。有了公式之后，就可以把每个未知类别的自变量观测值代入那些差别公式，算出和类别一样多的值；相应于哪一类的值最大，则把该观测值判为属于该类。</p>
<p>判别分析的原理并不复杂，可以简单地描述如下。</p>
<p>如果作为属性变量的因变量有k个预付（即观测值应该分成k个类），而自变量包含p个变量，那么每一个观测值就是p维空间的一个点。整个训练集的各个点（假定有n个点）就按照已知的类别在p维空间中形成共有n个点的k个点群。那么，对于一个未知类别的点，如果离哪一群近，就可以分到哪一集。关键的细节问题是如何定义“远近”或距离。关键是确定一些准则，使得每一个点群内部的点接近，而使各群之间尽量分开。不同的准则造就了不同的差别或分类方法。这里不讨论这些方法的数学细节，这里仅仅就是线性差别分析一种方法，介绍如何通过R得到结果及结果的解释。这里先把p维空间的点按照最能够分开各类的原则投影到二维空间，再根据二维空间这些点的分布得到判别函数。</p>
<p>如何来确定判别分析模型的效果呢？最基本的就是用一个没有参加建模的已知观测值类别的数据来验证。这样的数据集称为测试集或检验集(test set)。把通过模型预测的这些观测值的类别和它们已知的类别比较，就知道有多少划分错了，由此可以计算出误差来。</p>
<h2 id="案例分析">案例分析</h2>
<p>R自带了一个数据集，即<code>iris</code>。这是一个鸢尾花(iris)的数据，该数据给出了150个鸢尾花的萼片长(sepal length)、萼片宽(sepal width)、花瓣长(petal length)、花瓣宽(petal width)以及这些花的种类(Species)这五个变量。萼片和花瓣的长宽是四个定量变量，而各类为分类变量（有3个值，为setosa、versicolour、virginica）。现在，我们建立一个模型，根据萼片和花瓣这四个度量来把鸢尾花进行分类，这里的鸢尾花一共有3种，每种都有50个观测值。</p>
<p>现在我们在每种鸢尾花中也能耐选择一半，也就是25个作为测试集，另外25个用来建模。三种鸢尾花一起，训练集和测试集各有75个观测。由于选择测试集和训练集的过程是随机的，每次计算的结果也不相同，在R软件中，对于不同的成随机数种子，我们得到的结果也不尽相同。下面是对两个随机数种子得到的两个测试集判别结果，每个格子中的数目是相应的真实类的观测被划分到预测类的频数，如下所示：</p>
<p>随机种子数：set.seed(5)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(MASS)</div><div class="line">set.seed(<span class="number">5</span>)</div><div class="line">samp &lt;- c(sample(<span class="number">1</span>:<span class="number">50</span>,<span class="number">25</span>),sample(<span class="number">51</span>:<span class="number">100</span>,<span class="number">25</span>),sample(<span class="number">101</span>:<span class="number">150</span>,<span class="number">25</span>))</div><div class="line">a &lt;- lda(Species~.,data=iris,shuset=samp)</div><div class="line">summary(predict(a,iris[-samp,])$class)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; summary(predict(a,iris[-samp,])$class)</div><div class="line">    setosa versicolor  virginica </div><div class="line">        25         25         25</div></pre></td></tr></table></figure>
<p>随机种子数：set.seed(4)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(MASS)</div><div class="line">set.seed(<span class="number">4</span>)</div><div class="line">samp &lt;- c(sample(<span class="number">1</span>:<span class="number">50</span>,<span class="number">25</span>),sample(<span class="number">51</span>:<span class="number">100</span>,<span class="number">25</span>),sample(<span class="number">101</span>:<span class="number">150</span>,<span class="number">25</span>))</div><div class="line">a &lt;- lda(Species~.,data=iris,shuset=samp)</div><div class="line">summary(predict(a,iris[-samp,])$class)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; summary(predict(a,iris[-samp,])$class)</div><div class="line">    setosa versicolor  virginica </div><div class="line">        25         23         27</div></pre></td></tr></table></figure>
<p>由这两个结果可知，在第一个随机种子下（set.seed(5)）下，测试集中的每个类别完全划分正确，在第二个随机种子下（set.seed(4)），错误地划分了4个类别，出错率为4/75。</p>
<p>下图是在建立线性差别函数过程中，把75个训练集的四维空间点投影到二维空间的结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(mda)</div><div class="line">z&lt;-fda(Species~.,data=iris[samp,])$fit$fitted</div><div class="line">plot(z,pch=c(rep(<span class="number">19</span>,<span class="number">25</span>),rep(<span class="number">2</span>,<span class="number">25</span>),rep(<span class="number">8</span>,<span class="number">25</span>)),xlab=<span class="string">""</span>,ylab=<span class="string">""</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180819/HJDbc61Eim.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>可以看也，setosa（用实心圆点标记）离另外两种很远，而versicolor（用空心三角形标记）和virginica（用星号标记）这两类则离得比较近，容易划分错，我们从第二个随机种子（set.seed(4)）中也能看出来，出错的应时这两类。</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/18/Linux/Linux命令收集/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/18/Linux/Linux命令收集/" itemprop="url">Linux命令收集</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-18T12:00:00+08:00">
                2018-08-18
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  136
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="下载">下载</h3>
<p><code>curl</code>，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">curl http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff -o sc.gff</div></pre></td></tr></table></figure>
<p>这个命令是将某个URL的信息下载下来，并通过<code>-o</code>参数，将其输出到<code>sc.gff</code>文件中，其中,<code>-progress</code>是显示进度条，这个是选用的。</p>
<p><code>prefetch</code></p>
<h3 id="查看文件">查看文件</h3>
<p><code>ls</code></p>
<h2 id="文本相关">文本相关</h2>
<h3 id="文本统计">文本统计</h3>
<p><code>wc</code></p>
<h3 id="文本排序">文本排序</h3>
<p><code>sort</code></p>
<h3 id="去重">去重</h3>
<p><code>unique</code></p>
<h3 id="查看文本">查看文本</h3>
<p><code>head</code></p>
<p><code>tail</code></p>
<p><code>less</code></p>
<p><code>more</code></p>
<p><code>cat</code></p>
<h3 id="正则表达式">正则表达式</h3>
<p><code>grep</code></p>
<p><code>awk</code></p>
<p><code>sed</code></p>
<hr>
<h2 id="文件操作">文件操作</h2>
<h3 id="查找文件">查找文件</h3>
<p><code>find</code></p>
<p><code>locate</code></p>
<h3 id="压缩">压缩</h3>
<p><code>unzip</code></p>
<hr>
<h3 id="目录操作">目录操作</h3>
<p>切换目录：<code>cd</code></p>
<p>新建目录：<code>mkdir</code></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/17/Linux/Linux文档的查阅/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/17/Linux/Linux文档的查阅/" itemprop="url">Linux文档的查阅</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-17T12:00:00+08:00">
                2018-08-17
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  0
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/07/14/BioStatistics/0统计学书籍汇总/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/07/14/BioStatistics/0统计学书籍汇总/" itemprop="url">统计学书籍汇总</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-07-14T12:10:54+08:00">
                2018-07-14
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,385
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>我觉得统计学的知识点很碎，仅看一本书是不太容易学好统计学的，这篇笔记就介绍几本有关统计学的书，这些书是我看过的，也有正在看的，也有看了一部分太难理解放弃的，我大致按照由易到难的顺序列出了几本（当然还有很多优秀的书没看过，就没列出来），这些书可以在京东或淘宝上买到，大部分书也可以通过搜索引擎找到相应下载的地方，考虑到知识产权的问题，这里不提供PDF版的书籍下载地址。</p>
<h2 id="零基础">零基础</h2>
<h4 id="漫画统计学入门">《漫画统计学入门》</h4>
<p>作者: 拉里-戈尼克 / 沃尔科特-史密斯 出版社: 辽宁教育出版社 ，译者: 梁杰，出版年: 2002-1</p>
<p>这本书就是以漫画的形式把统计学原理给讲了一下，非常适合零基础的人入门，没有复杂的公式，可以从直观上克服对统计学的那种畏惧心理，但是如果要深入理解统计学，这本书远远不够。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/205H40fcDI.jpg?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="漫画统计学">《漫画统计学》</h4>
<p>作者: [日] 高桥信 ，出版社: 科学出版社，原作名: マンガでわかる統計学，译者: 陈刚 ，出版年: 2009-8</p>
<p>这本书跟《漫画统计学入门》类似，这本书只讲了统计学的一些基本原理，这本书是一个叫欧姆社系列漫画书中的一本，其余有关统计的书还有《漫画统计学之因子分析》，《漫画统计学之回归分析》，与数学有关的书有《漫画微分方程》《漫画线性代数》。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/jJ5kj241AA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="女士品茶">《女士品茶》</h4>
<p>作者: 萨尔斯伯格 出版社: 中国统计出版社 副标题: 20世纪统计怎样变革了科学 原作名: The Lady Tasting Tea: How Statistics Revolutionized Science in the Twentieth Century 译者: 邱东等 出版年: 2004-11-01</p>
<p>这本书以某位喝茶的英国女士的假设学说为起点，写了数理统计学历史中的一些大牛，例如戈塞物，费歇尔，皮尔逊等一大批牛人的统计学故事，这本书其实不算作是统计学书，更像是统计学的科普书。对于提高统计学方面的兴趣的思维还是比较有好处的。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/fd14HLgh90.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="入门">入门</h2>
<h4 id="白话统计">《白话统计》</h4>
<p>《白话统计》，冯国双，2018，电子工业出版社。</p>
<p>2018年的新书，没有复杂的公式，非常简单，能把统计学原理以最浅显的语言给讲出来，比较适合那些没有太多数学基础的同学，不过要深入理解统计学，这本书还是不够。我觉得这本书更适合作为统计学教材（例如孙振球的《医学统计学》）的辅助用书。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/6I06L1J3KF.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="深入浅出统计学">《深入浅出统计学》</h4>
<p>《深入浅出统计学》，DawnGriffiths. 深入浅出统计学[M]. 电子工业出版社, 2012.</p>
<p>本书涵盖的知识点包括：信息可视化、概率计算、几何分布、二项分布及泊松分布、正态分布、统计抽样、置信区间的构建、假设检验、卡方分布、相关于回归等等， 没有太多数学公式，很好理解。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/3cB4B0BaaG.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="深入浅出数据分析">《深入浅出数据分析》</h4>
<p>作者MichaelMilton. 深入浅出数据分析[M]. 电子工业出版社, 2010.</p>
<p>这本书可以与《深入浅出统计学》配合使用，可以看作是前一本的升级版。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/biI99liiCc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="行为科学统计">《行为科学统计》</h4>
<p>《行为科学统计》，作者是 [美] F·J·Gravetter / [美] L·B·Wallnau，中文貌似最新的只有第7版，第7版是2008年出版的，有点早了。可以看英文的，英文已经出到第10版了，名字是《Statistics for the Behavioral Sciences》(10th)，第10版是2016年出的。</p>
<p>这本书本来就是为社会学，心理学这些数学基础比较薄弱的学科的学生写的，同样适用于生物学（毕竟数学基础也很薄弱），读起来并不困难。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/C4H4egjCel.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/bD7L8j6Fc1.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="统计学基础">统计学基础</h2>
<p>这一部分的书可以与后面提到的有关R语言的书配合看。 ####《医学统计学》</p>
<p>国内的医学统计学书都差不多，去京东上挑一本销量比较大的就行，从头到尾看一下，把统计学的知识理一下，不懂的地方可以找其他的统计学书补充一下，再用R把书中的案例给计算一下就行，我平时看的是孙振球的《医学统计学》。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/f7Bkh3d67g.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="统计思维程序员数学之概率统计">《统计思维——程序员数学之概率统计》</h4>
<p>Allen B.Downey著，张建锋，陈钢 译，人民邮电出版社，书不长，很快就可以读完，对几个统计学基础的概念介绍了一下。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/ihfdkGijm1.png?imageslim" alt="mark"></p>
<h4 id="从数据到结论">《从数据到结论》</h4>
<p>作者吴喜之，把一些统计学的概念与原理讲得很清楚，公式不多，比较简单。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/lJjm024lbL.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="从概念到数据分析">《从概念到数据分析》</h4>
<p>2008年高等教育出版社出版的图书，作者是吴喜之 ，里面使用R语言把一些基本的统计学概念给计算了一下，对于理解统计学很有帮助，公式不多，书很薄，很快就能看完，可以与他的《从数据到结论》结合起来看。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/L4fl1BfGgC.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="医学统计学基础与典型错误辨析">《医学统计学基础与典型错误辨析》</h4>
<p>作者胡良平，书比较老，但我觉得比较有价值，作者收集了医学上常见的一些统计学错误，由于医学生（或者说是生物背景的学生）的统计学基础很多并不是很好，因此经常出现一些统计学的错误，这本书就挑了不少这方面的论文，主要是中文论文里的错误，有精力了可以看看。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/AGbm6IKEC4.png?imageslim" alt="mark"></p>
<h4 id="数理统计学教程">《数理统计学教程》</h4>
<p>陈希孺的书，公式比较多，适合对统计学想深入了解的同学，需要一定的数学基础。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/5CKG09eGje.png?imageslim" alt="mark"></p>
<h4 id="统计学习基础数据挖掘推理与预测">《统计学习基础——数据挖掘、推理与预测》</h4>
<p>原书名是《The Elements of Statistical Learning》，副标题是《Data Mining, Inference, and Prediction, Second Edition》，据说是统计学领域的经典之作，不过貌似需要一定的数学基础，这本书并不太好理解。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/3lb11FI4k8.png?imageslim" alt="mark"> 英文版的封面如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/6FghfK6hbg.png?imageslim" alt="mark"></p>
<h2 id="统计学与r">统计学与R</h2>
<p>这一部分的书可以与一些统计学的书配合看。</p>
<h4 id="r语言统计入门">《R语言统计入门》</h4>
<p>原名为Introductory Statistics with R second Edition，常用的统计方法与R实现，没多少公式，很好理解。书的前面部分是对R操作的一些简介，后面就是利用R在常规的统计学中的应用。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/B4IHmb7eF5.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="r语言与统计分析">《R语言与统计分析》</h4>
<p>作者汤银才，书中有很多案例，使用R语言把统计学中常见的方法给演示了出来，适合有一些统计学基础的同学，书中的R代码都很短，很好理解，适合那些又想学R语言，又想学统计学的同学，不过书比较老，2008年出的，但统计学的方法基本上变化不大，比前面的那本略难。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/04h5clAH89.png?imageslim" alt="mark"></p>
<h4 id="多元统计分析与r语言建模">《多元统计分析与R语言建模》</h4>
<p>作者王赋会，介绍了常见的一些统计学方法原理与R语言实现。这些统计学方法包括广义线性模型、判别分析、聚类分析、主成分分析、因子分析、对应分析、典型相关分析等，虽然有公式，但我觉得通过查资料，还是能理解的。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/cjKAg78m7C.png?imageslim" alt="mark"></p>
<h4 id="复杂数据统计方法基于r的应用">《复杂数据统计方法——基于R的应用》</h4>
<p>作者吴喜之，书中的数据都是从国外网站上下载的，这本书是2012年出版的，介绍了一些常见的经典统计学方法以及机器学习方法，公式不多。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/FlLcg4AlJm.png?imageslim" alt="mark"></p>
<h2 id="网上的一些资料">网上的一些资料</h2>
<p>这一部分的资料主要是网上的一些不错的资料，有两个。</p>
<h4 id="医学统计学及sas应用之r语言实现">《医学统计学及SAS应用之R语言实现》</h4>
<p>这个资料最初是我从丁香园上看到的，作者不详，只知道是一名留日的医学博士，并非统计学出身，他学了R语言之后，就把《医学统计学及SAS应用》这本书中的案例重新计算了一下，这个资料网上有，能搜到，只有代码与计算结果，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/7DFiLjJ2Ci.png?imageslim" alt="mark"> 如果看这个资料，需要配合看《医学统计学及SAS应用》。 <strong>《医学统计学及SAS应用》</strong>的作者是王炳顺，这本书主要讲的是医学统计学中的一些案例与SAS的计算过程，如果要看这本书，只需要看书中的统计学原理、案例与SAS的计算结果就行，SAS的使用可以不看，可以自己使用R语言把书中的案例计算一下，然后对比一下自己的计算结果与SAS的计算结果，以及计算结果如何解读。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/5DDiECCh5D.png?imageslim" alt="mark"></p>
<h4 id="医学统计分析与r语言">《医学统计分析与R语言》</h4>
<p>作者梁雪枫，光明日报出版社。这本书里面对医学统计学介绍的极为详细，对数学与统计学基础也要求比较高，如果没有相应的基础读起来比较吃力。这本书很奇怪的是，在京东与淘宝上都搜不到，我还是在豆瓣的评论中找到了下载地址，如果想对统计学有进一步了解的同学，可以看看这本书。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/Ie62E0bfab.png?imageslim" alt="mark"></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/2/"><i class="fa fa-angle-left"></i></a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/14/">14</a><a class="extend next" rel="next" href="/page/4/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">134</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">13</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">37</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_uv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共513.4k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
