<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="GEO,公共数据库挖掘," />










<meta name="description" content="前言 此篇笔记主要是学习生信技术树Jimmy在B站上的视频教程。 背景知识 GEO的全称是Gene Expression Omnibus，数据库地址是：https://www.ncbi.nlm.nih.gov/geo/。 GEO是一个公共的基因组数据库，含有大量的芯片以及二代测序数据。在挖掘GEO上的数据集时，常规流程就是选择GSE号，下载表达矩阵，得到基因集，使用数据库进行注释，PPI等">
<meta name="keywords" content="GEO,公共数据库挖掘">
<meta property="og:type" content="article">
<meta property="og:title" content="GEO笔记">
<meta property="og:url" content="http://rvdsd.top/2019/06/02/BioItem/GEO笔记/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 此篇笔记主要是学习生信技术树Jimmy在B站上的视频教程。 背景知识 GEO的全称是Gene Expression Omnibus，数据库地址是：https://www.ncbi.nlm.nih.gov/geo/。 GEO是一个公共的基因组数据库，含有大量的芯片以及二代测序数据。在挖掘GEO上的数据集时，常规流程就是选择GSE号，下载表达矩阵，得到基因集，使用数据库进行注释，PPI等">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630180025.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630180621.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630181024.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630181741.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719171511.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719175830.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719182201.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719191931.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719194204.png">
<meta property="og:updated_time" content="2019-07-19T11:42:19.727Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GEO笔记">
<meta name="twitter:description" content="前言 此篇笔记主要是学习生信技术树Jimmy在B站上的视频教程。 背景知识 GEO的全称是Gene Expression Omnibus，数据库地址是：https://www.ncbi.nlm.nih.gov/geo/。 GEO是一个公共的基因组数据库，含有大量的芯片以及二代测序数据。在挖掘GEO上的数据集时，常规流程就是选择GSE号，下载表达矩阵，得到基因集，使用数据库进行注释，PPI等">
<meta name="twitter:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630180025.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/06/02/BioItem/GEO笔记/"/>





  <title>GEO笔记 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/06/02/BioItem/GEO笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GEO笔记</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-06-02T12:00:00+08:00">
                2019-06-02
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信实战项目/" itemprop="url" rel="index">
                    <span itemprop="name">生信实战项目</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,593
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>此篇笔记主要是学习生信技术树Jimmy在<a href="https://www.bilibili.com/video/av37568990" target="_blank" rel="external">B站上的视频教程</a>。</p>
<h2 id="背景知识">背景知识</h2>
<p>GEO的全称是Gene Expression Omnibus，数据库地址是：https://www.ncbi.nlm.nih.gov/geo/。</p>
<p>GEO是一个公共的基因组数据库，含有大量的芯片以及二代测序数据。在挖掘GEO上的数据集时，常规流程就是选择GSE号，下载表达矩阵，得到基因集，使用数据库进行注释，PPI等网络等。</p>
<p>GEO里面的数据前缀有4种，如下所示：</p>
<ul>
<li>GEO Platform (GPL)</li>
<li>GEO Sample (GSM)</li>
<li>GEO Series (GSE)</li>
<li>GEO Dataset (GDS)</li>
</ul>
<p>其中GSE开头的是表达矩阵，GPL开头的是测序的平台信息。一篇文章可以有一个或者多个GSE数据集，一个GSE里面可以有一个或者多个GSM样本。多个研究的GSM样本可以根据研究目的整合为一个GDS，不过GDS本身用的很少。而每个数据集都有着自己对应的芯片平台，就是GPL。</p>
<h2 id="下载数据">下载数据</h2>
<p>这个数据已经被其他人挖掘过发了文章，文章信息如下所示：</p>
<p>Wang Q L , Chen X , Zhang M H , et al. Identification of hub genes and pathways associated with retinoblastoma based on co-expression network analysis[J]. Genetics and Molecular Research, 2015, 14(4):16151-16161.</p>
<p>现在去GEO官网下载数据：</p>
<p>代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">rm(list = ls())</div><div class="line">setwd(<span class="string">"D:\\GEO_test"</span>)</div><div class="line"><span class="keyword">library</span>(GEOquery)</div><div class="line">eSet &lt;- getGEO(<span class="string">'GSE42872'</span>, destdir=<span class="string">"."</span>, </div><div class="line">                 AnnotGPL = <span class="literal">F</span>,</div><div class="line">                 getGPL = <span class="literal">F</span>)</div><div class="line">save(eSet, file=<span class="string">'GSE42872_eSet.Rdata'</span>)</div></pre></td></tr></table></figure>
<p>这里用到了<code>getGEO</code>函数，这个函数的参数如下所示：</p>
<ol style="list-style-type: decimal">
<li><code>destdir</code>：下载的地址，使用<code>.</code>则下载到当前的工作目录；</li>
<li><code>AnnotGPL</code>：是否下载注释的信息。</li>
<li><code>getGPL</code>：若为TRUE，则下载GPL注释文件；若为FALSE，则不下载。默认为TRUE。</li>
<li><code>GSEMatrix</code>：若为TRUE，则下载Matrix文件；若为FALSE，则下载SOFT文件。默认为TRUE。</li>
</ol>
<p>随后，<code>D:\GEO_test</code>目录下就会出现下载后的表达矩阵，名称为<code>GSE42872_series_matrix.txt.gz</code>。</p>
<p>GSE24673</p>
<p>下载后的这个<code>eSet</code>文件信息如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line">&gt; eSet</div><div class="line">$GSE24673_series_matrix.txt.gz</div><div class="line">ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">28869</span> features, <span class="number">11</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: GSM607938 GSM607939 <span class="keyword">...</span> GSM607948 (<span class="number">11</span> total)</div><div class="line">  varLabels: title geo_accession <span class="keyword">...</span> tissue:ch1 (<span class="number">40</span> total)</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData: none</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: GPL6244 </div><div class="line"></div><div class="line">&gt; class(eSet)</div><div class="line">[<span class="number">1</span>] <span class="string">"list"</span></div><div class="line">&gt; eSet[[<span class="number">1</span>]]</div><div class="line">ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">28869</span> features, <span class="number">11</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: GSM607938 GSM607939 <span class="keyword">...</span> GSM607948 (<span class="number">11</span> total)</div><div class="line">  varLabels: title geo_accession <span class="keyword">...</span> tissue:ch1 (<span class="number">40</span> total)</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData: none</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: GPL6244</div></pre></td></tr></table></figure>
<p>从上面信息可以知道：</p>
<ol style="list-style-type: decimal">
<li><code>eSet</code>是一个列表(list)；</li>
<li><code>eSet</code>这个列表只有一个元素。</li>
</ol>
<h2 id="获取表达矩阵">获取表达矩阵</h2>
<h3 id="提取表达矩阵">提取表达矩阵</h3>
<p>使用<code>exprs</code>函数可以获取表达矩阵，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">raw_profile_data &lt;- eSet[[<span class="number">1</span>]]</div><div class="line">exprSet = exprs(raw_profile_data)</div></pre></td></tr></table></figure>
<p>表达矩阵如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630180025.png">

</div>
<h3 id="获取样本信息">获取样本信息</h3>
<p>使用<code>sampleNames</code>函数可以获取样本信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sampleNames(raw_profile_data)</div></pre></td></tr></table></figure>
<p>运行后如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; sampleNames(raw_profile_data)</div><div class="line"> [<span class="number">1</span>] <span class="string">"GSM607938"</span> <span class="string">"GSM607939"</span> <span class="string">"GSM607940"</span> <span class="string">"GSM607941"</span> <span class="string">"GSM607942"</span> <span class="string">"GSM607943"</span> <span class="string">"GSM607944"</span> <span class="string">"GSM607945"</span></div><div class="line"> [<span class="number">9</span>] <span class="string">"GSM607946"</span> <span class="string">"GSM607947"</span> <span class="string">"GSM607948"</span></div></pre></td></tr></table></figure>
<p>使用<code>pData</code>可以获取分组信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">pData &lt;- pData(raw_profile_data)</div></pre></td></tr></table></figure>
<p>运行如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630180621.png">

</div>
<p>从title这一列就可以知道，里面有几个是干预组，有几个是健康组。</p>
<h2 id="表达矩阵过滤">表达矩阵过滤</h2>
<p>这一部分主要是将探针名称转化为基因名。</p>
<h4 id="转换探针名">转换探针名</h4>
<p>表达矩阵的列的名称是探针名，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630181024.png">

</div>
<p>现在将探针名转换为基因名，从前面的<code>eSet</code>数据集中可以获取以下信息：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">&gt; eSet</div><div class="line">$GSE24673_series_matrix.txt.gz</div><div class="line">ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">assayData: <span class="number">28869</span> features, <span class="number">11</span> samples </div><div class="line">  element names: exprs </div><div class="line">protocolData: none</div><div class="line">phenoData</div><div class="line">  sampleNames: GSM607938 GSM607939 <span class="keyword">...</span> GSM607948 (<span class="number">11</span> total)</div><div class="line">  varLabels: title geo_accession <span class="keyword">...</span> tissue:ch1 (<span class="number">40</span> total)</div><div class="line">  varMetadata: labelDescription</div><div class="line">featureData: none</div><div class="line">experimentData: use <span class="string">'experimentData(object)'</span></div><div class="line">Annotation: GPL6244</div></pre></td></tr></table></figure>
<p>从<code>Annotation: GPL6244</code>这行可以知道，这个数据集是用<code>GPL6244</code>这个平台测的，同时在GEO的官网也能查到这个信息，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190630181741.png">

</div>
<p>现有的一些常用平台都对应的有相应的R包，如下所示：</p>
<table>
<thead>
<tr class="header">
<th></th>
<th>gpl</th>
<th>organism</th>
<th>bioc_package</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>GPL32</td>
<td>Mus musculus</td>
<td>mgu74a</td>
</tr>
<tr class="even">
<td>2</td>
<td>GPL33</td>
<td>Mus musculus</td>
<td>mgu74b</td>
</tr>
<tr class="odd">
<td>3</td>
<td>GPL34</td>
<td>Mus musculus</td>
<td>mgu74c</td>
</tr>
<tr class="even">
<td>6</td>
<td>GPL74</td>
<td>Homo sapiens</td>
<td>hcg110</td>
</tr>
<tr class="odd">
<td>7</td>
<td>GPL75</td>
<td>Mus musculus</td>
<td>mu11ksuba</td>
</tr>
<tr class="even">
<td>8</td>
<td>GPL76</td>
<td>Mus musculus</td>
<td>mu11ksubb</td>
</tr>
<tr class="odd">
<td>9</td>
<td>GPL77</td>
<td>Mus musculus</td>
<td>mu19ksuba</td>
</tr>
<tr class="even">
<td>10</td>
<td>GPL78</td>
<td>Mus musculus</td>
<td>mu19ksubb</td>
</tr>
<tr class="odd">
<td>11</td>
<td>GPL79</td>
<td>Mus musculus</td>
<td>mu19ksubc</td>
</tr>
<tr class="even">
<td>12</td>
<td>GPL80</td>
<td>Homo sapiens</td>
<td>hu6800</td>
</tr>
<tr class="odd">
<td>13</td>
<td>GPL81</td>
<td>Mus musculus</td>
<td>mgu74av2</td>
</tr>
<tr class="even">
<td>14</td>
<td>GPL82</td>
<td>Mus musculus</td>
<td>mgu74bv2</td>
</tr>
<tr class="odd">
<td>15</td>
<td>GPL83</td>
<td>Mus musculus</td>
<td>mgu74cv2</td>
</tr>
<tr class="even">
<td>16</td>
<td>GPL85</td>
<td>Rattus norvegicus</td>
<td>rgu34a</td>
</tr>
<tr class="odd">
<td>17</td>
<td>GPL86</td>
<td>Rattus norvegicus</td>
<td>rgu34b</td>
</tr>
<tr class="even">
<td>18</td>
<td>GPL87</td>
<td>Rattus norvegicus</td>
<td>rgu34c</td>
</tr>
<tr class="odd">
<td>19</td>
<td>GPL88</td>
<td>Rattus norvegicus</td>
<td>rnu34</td>
</tr>
<tr class="even">
<td>20</td>
<td>GPL89</td>
<td>Rattus norvegicus</td>
<td>rtu34</td>
</tr>
<tr class="odd">
<td>22</td>
<td>GPL91</td>
<td>Homo sapiens</td>
<td>hgu95av2</td>
</tr>
<tr class="even">
<td>23</td>
<td>GPL92</td>
<td>Homo sapiens</td>
<td>hgu95b</td>
</tr>
<tr class="odd">
<td>24</td>
<td>GPL93</td>
<td>Homo sapiens</td>
<td>hgu95c</td>
</tr>
<tr class="even">
<td>25</td>
<td>GPL94</td>
<td>Homo sapiens</td>
<td>hgu95d</td>
</tr>
<tr class="odd">
<td>26</td>
<td>GPL95</td>
<td>Homo sapiens</td>
<td>hgu95e</td>
</tr>
<tr class="even">
<td>27</td>
<td>GPL96</td>
<td>Homo sapiens</td>
<td>hgu133a</td>
</tr>
<tr class="odd">
<td>28</td>
<td>GPL97</td>
<td>Homo sapiens</td>
<td>hgu133b</td>
</tr>
<tr class="even">
<td>29</td>
<td>GPL98</td>
<td>Homo sapiens</td>
<td>hu35ksuba</td>
</tr>
<tr class="odd">
<td>30</td>
<td>GPL99</td>
<td>Homo sapiens</td>
<td>hu35ksubb</td>
</tr>
<tr class="even">
<td>31</td>
<td>GPL100</td>
<td>Homo sapiens</td>
<td>hu35ksubc</td>
</tr>
<tr class="odd">
<td>32</td>
<td>GPL101</td>
<td>Homo sapiens</td>
<td>hu35ksubd</td>
</tr>
<tr class="even">
<td>36</td>
<td>GPL201</td>
<td>Homo sapiens</td>
<td>hgfocus</td>
</tr>
<tr class="odd">
<td>37</td>
<td>GPL339</td>
<td>Mus musculus</td>
<td>moe430a</td>
</tr>
<tr class="even">
<td>38</td>
<td>GPL340</td>
<td>Mus musculus</td>
<td>mouse4302</td>
</tr>
<tr class="odd">
<td>39</td>
<td>GPL341</td>
<td>Rattus norvegicus</td>
<td>rae230a</td>
</tr>
<tr class="even">
<td>40</td>
<td>GPL342</td>
<td>Rattus norvegicus</td>
<td>rae230b</td>
</tr>
<tr class="odd">
<td>41</td>
<td>GPL570</td>
<td>Homo sapiens</td>
<td>hgu133plus2</td>
</tr>
<tr class="even">
<td>42</td>
<td>GPL571</td>
<td>Homo sapiens</td>
<td>hgu133a2</td>
</tr>
<tr class="odd">
<td>43</td>
<td>GPL886</td>
<td>Homo sapiens</td>
<td>hgug4111a</td>
</tr>
<tr class="even">
<td>44</td>
<td>GPL887</td>
<td>Homo sapiens</td>
<td>hgug4110b</td>
</tr>
<tr class="odd">
<td>45</td>
<td>GPL1261</td>
<td>Mus musculus</td>
<td>mouse430a2</td>
</tr>
<tr class="even">
<td>49</td>
<td>GPL1352</td>
<td>Homo sapiens</td>
<td>u133x3p</td>
</tr>
<tr class="odd">
<td>50</td>
<td>GPL1355</td>
<td>Rattus norvegicus</td>
<td>rat2302</td>
</tr>
<tr class="even">
<td>51</td>
<td>GPL1708</td>
<td>Homo sapiens</td>
<td>hgug4112a</td>
</tr>
<tr class="odd">
<td>54</td>
<td>GPL2891</td>
<td>Homo sapiens</td>
<td>h20kcod</td>
</tr>
<tr class="even">
<td>55</td>
<td>GPL2898</td>
<td>Rattus norvegicu</td>
<td>adme16cod</td>
</tr>
<tr class="odd">
<td>60</td>
<td>GPL3921</td>
<td>Homo sapiens</td>
<td>hthgu133a</td>
</tr>
<tr class="even">
<td>63</td>
<td>GPL4191</td>
<td>Homo sapiens</td>
<td>h10kcod</td>
</tr>
<tr class="odd">
<td>64</td>
<td>GPL5689</td>
<td>Homo sapiens</td>
<td>hgug4100a</td>
</tr>
<tr class="even">
<td>65</td>
<td>GPL6097</td>
<td>Homo sapiens</td>
<td>illuminaHumanv1</td>
</tr>
<tr class="odd">
<td>66</td>
<td>GPL6102</td>
<td>Homo sapiens</td>
<td>illuminaHumanv2</td>
</tr>
<tr class="even">
<td>67</td>
<td>GPL6244</td>
<td>Homo sapiens</td>
<td>hugene10sttranscriptcluster</td>
</tr>
<tr class="odd">
<td>68</td>
<td>GPL6947</td>
<td>Homo sapiens</td>
<td>illuminaHumanv3</td>
</tr>
<tr class="even">
<td>69</td>
<td>GPL8300</td>
<td>Homo sapiens</td>
<td>hgu95av2</td>
</tr>
<tr class="odd">
<td>70</td>
<td>GPL8490</td>
<td>Homo sapiens</td>
<td>IlluminaHumanMethylation27k</td>
</tr>
<tr class="even">
<td>71</td>
<td>GPL10558</td>
<td>Homo sapiens</td>
<td>illuminaHumanv4</td>
</tr>
<tr class="odd">
<td>72</td>
<td>GPL11532</td>
<td>Homo sapiens</td>
<td>hugene11sttranscriptcluster</td>
</tr>
<tr class="even">
<td>73</td>
<td>GPL13497</td>
<td>Homo sapiens</td>
<td>HsAgilentDesign026652</td>
</tr>
<tr class="odd">
<td>74</td>
<td>GPL13534</td>
<td>Homo sapiens</td>
<td>IlluminaHumanMethylation450k</td>
</tr>
<tr class="even">
<td>75</td>
<td>GPL13667</td>
<td>Homo sapiens</td>
<td>hgu219</td>
</tr>
<tr class="odd">
<td>76</td>
<td>GPL15380</td>
<td>Homo sapiens</td>
<td>GGHumanMethCancerPanelv1</td>
</tr>
<tr class="even">
<td>77</td>
<td>GPL15396</td>
<td>Homo sapiens</td>
<td>hthgu133b</td>
</tr>
<tr class="odd">
<td>78</td>
<td>GPL17897</td>
<td>Homo sapiens</td>
<td>hthgu133a</td>
</tr>
</tbody>
</table>
<p>如果是比较少见的平台，那么可以通过下面的方式来下载相应的平台探针信息：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># gpl &lt;- getGEO('GPL6244',destdir = ".")</span></div></pre></td></tr></table></figure>
<p>只是这种方式下载起来比较慢。</p>
<p>此时将探针转换为基因名，我们前面提到，这个数据使用的是<code>GPL6244</code>平台来测的（上表第67行），它对应的R包是<code>hugene10sttranscriptcluster</code>，此时安装这个包并加载即可：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">biocLite(<span class="string">"hugene10sttranscriptcluster.db"</span>)</div><div class="line"><span class="keyword">library</span>(hugene10sttranscriptcluster.db)</div><div class="line">ls(<span class="string">'package:hugene10sttranscriptcluster.db'</span>)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">&gt; ls(<span class="string">'package:hugene10sttranscriptcluster.db'</span>)</div><div class="line"> [<span class="number">1</span>] <span class="string">"hugene10sttranscriptcluster"</span>              <span class="string">"hugene10sttranscriptcluster.db"</span>          </div><div class="line"> [<span class="number">3</span>] <span class="string">"hugene10sttranscriptcluster_dbconn"</span>       <span class="string">"hugene10sttranscriptcluster_dbfile"</span>      </div><div class="line"> [<span class="number">5</span>] <span class="string">"hugene10sttranscriptcluster_dbInfo"</span>       <span class="string">"hugene10sttranscriptcluster_dbschema"</span>    </div><div class="line"> [<span class="number">7</span>] <span class="string">"hugene10sttranscriptclusterACCNUM"</span>        <span class="string">"hugene10sttranscriptclusterALIAS2PROBE"</span>  </div><div class="line"> [<span class="number">9</span>] <span class="string">"hugene10sttranscriptclusterCHR"</span>           <span class="string">"hugene10sttranscriptclusterCHRLENGTHS"</span>   </div><div class="line">[<span class="number">11</span>] <span class="string">"hugene10sttranscriptclusterCHRLOC"</span>        <span class="string">"hugene10sttranscriptclusterCHRLOCEND"</span>    </div><div class="line">[<span class="number">13</span>] <span class="string">"hugene10sttranscriptclusterENSEMBL"</span>       <span class="string">"hugene10sttranscriptclusterENSEMBL2PROBE"</span></div><div class="line">[<span class="number">15</span>] <span class="string">"hugene10sttranscriptclusterENTREZID"</span>      <span class="string">"hugene10sttranscriptclusterENZYME"</span>       </div><div class="line">[<span class="number">17</span>] <span class="string">"hugene10sttranscriptclusterENZYME2PROBE"</span>  <span class="string">"hugene10sttranscriptclusterGENENAME"</span>     </div><div class="line">[<span class="number">19</span>] <span class="string">"hugene10sttranscriptclusterGO"</span>            <span class="string">"hugene10sttranscriptclusterGO2ALLPROBES"</span> </div><div class="line">[<span class="number">21</span>] <span class="string">"hugene10sttranscriptclusterGO2PROBE"</span>      <span class="string">"hugene10sttranscriptclusterMAP"</span>          </div><div class="line">[<span class="number">23</span>] <span class="string">"hugene10sttranscriptclusterMAPCOUNTS"</span>     <span class="string">"hugene10sttranscriptclusterOMIM"</span>         </div><div class="line">[<span class="number">25</span>] <span class="string">"hugene10sttranscriptclusterORGANISM"</span>      <span class="string">"hugene10sttranscriptclusterORGPKG"</span>       </div><div class="line">[<span class="number">27</span>] <span class="string">"hugene10sttranscriptclusterPATH"</span>          <span class="string">"hugene10sttranscriptclusterPATH2PROBE"</span>   </div><div class="line">[<span class="number">29</span>] <span class="string">"hugene10sttranscriptclusterPFAM"</span>          <span class="string">"hugene10sttranscriptclusterPMID"</span>         </div><div class="line">[<span class="number">31</span>] <span class="string">"hugene10sttranscriptclusterPMID2PROBE"</span>    <span class="string">"hugene10sttranscriptclusterPROSITE"</span>      </div><div class="line">[<span class="number">33</span>] <span class="string">"hugene10sttranscriptclusterREFSEQ"</span>        <span class="string">"hugene10sttranscriptclusterSYMBOL"</span>       </div><div class="line">[<span class="number">35</span>] <span class="string">"hugene10sttranscriptclusterUNIGENE"</span>       <span class="string">"hugene10sttranscriptclusterUNIPROT"</span></div></pre></td></tr></table></figure>
<p>这是一个R包的注释文件，它提供的是hugene10sttranscriptcluster这个芯片平台的详细信息，每两个月更新一次。这个包中含有的很多数据，上表已经列出来了，我们将探针转换为基因要使用的是文件是<code>hugene10sttranscriptclusterSYMBOL</code>这个，它的功能在于：将探针编号转换为基因名。</p>
<p>同理，<code>hugene10sttranscriptclusterUNIPROT</code>是将探针编号转换为蛋白名，</p>
<p>同时，<code>hugene10sttranscriptclusterREFSEQ</code>是将探针编号转换为转录组名，转换时使用的函数是<code>toTable()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">probe_trans_SYMBOL &lt;- toTable(hugene10sttranscriptclusterSYMBOL)</div><div class="line">probe_trans_UNIPROT &lt;- toTable(hugene10sttranscriptclusterUNIPROT)</div><div class="line">probe_trans_REFSEQ &lt;- toTable(hugene10sttranscriptclusterREFSEQ)</div><div class="line">head(probe_trans_SYMBOL)</div><div class="line">head(probe_trans_UNIPROT)</div><div class="line">head(probe_trans_REFSEQ)</div></pre></td></tr></table></figure>
<p>运行结果为：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; head(probe_trans_SYMBOL)</div><div class="line">  probe_id    symbol</div><div class="line"><span class="number">1</span>  <span class="number">7896759</span> LINC01128</div><div class="line"><span class="number">2</span>  <span class="number">7896761</span>    SAMD11</div><div class="line"><span class="number">3</span>  <span class="number">7896779</span>    KLHL17</div><div class="line"><span class="number">4</span>  <span class="number">7896798</span>   PLEKHN1</div><div class="line"><span class="number">5</span>  <span class="number">7896817</span>     ISG15</div><div class="line"><span class="number">6</span>  <span class="number">7896822</span>      AGRN</div><div class="line">&gt; head(probe_trans_UNIPROT)</div><div class="line">  probe_id uniprot_id</div><div class="line"><span class="number">1</span>  <span class="number">7896761</span>     Q96NU1</div><div class="line"><span class="number">2</span>  <span class="number">7896779</span>     Q6TDP4</div><div class="line"><span class="number">3</span>  <span class="number">7896798</span>     Q494U1</div><div class="line"><span class="number">4</span>  <span class="number">7896817</span>     P05161</div><div class="line"><span class="number">5</span>  <span class="number">7896822</span>     O00468</div><div class="line"><span class="number">6</span>  <span class="number">7896865</span>     Q6ZVT0</div><div class="line">&gt; head(probe_trans_REFSEQ)</div><div class="line">  probe_id accession</div><div class="line"><span class="number">1</span>  <span class="number">7896759</span> NR_015368</div><div class="line"><span class="number">2</span>  <span class="number">7896759</span> NR_047519</div><div class="line"><span class="number">3</span>  <span class="number">7896759</span> NR_047520</div><div class="line"><span class="number">4</span>  <span class="number">7896759</span> NR_047521</div><div class="line"><span class="number">5</span>  <span class="number">7896759</span> NR_047522</div><div class="line"><span class="number">6</span>  <span class="number">7896759</span> NR_047523</div></pre></td></tr></table></figure>
<p>我们这里只使用基因SYMBOL即可，也就是<code>probe_trans_SYMBOL &lt;- toTable(hugene10sttranscriptclusterSYMBOL)</code>这个语句。</p>
<p>由于一个基因对应好几个探针，因此转换后的基因名必然是小于探针名的，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">length(unique(probe_trans_SYMBOL$symbol))</div><div class="line">length(unique(probe_trans_SYMBOL$probe_id))</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">&gt; length(unique(probe_trans_SYMBOL$symbol))</div><div class="line">[<span class="number">1</span>] <span class="number">18834</span></div><div class="line">&gt; length(unique(probe_trans_SYMBOL$probe_id))</div><div class="line">[<span class="number">1</span>] <span class="number">19827</span></div></pre></td></tr></table></figure>
<p>从上面可以看出来，基因SYMBOL有18834个，探针有19827个。</p>
<p>现在我们将下载的表达谱中的探针转换为基因SYMBOL，先来看下表达谱，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719171511.png">

</div>
<p>现在的任务就是将左侧的探针编号转换为基因SYMBOL。</p>
<p>此时有两步操作：</p>
<p>第一步就是我们先把表达谱中能够在平台芯片注释文件中找到基因SYMBOL给挑出来，因为表达谱中的有的探针名没有对应的基因SYMBOL，不要这种探针的数据，筛选过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">table(rownames(exprSet) %<span class="keyword">in</span>% probe_trans_SYMBOL$probe_id)</div><div class="line">dim(exprSet)</div><div class="line">exprSet_trans_symbol &lt;- exprSet[rownames(exprSet) %<span class="keyword">in</span>% probe_trans_SYMBOL$probe_id,]</div><div class="line">dim(exprSet_trans_symbol)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; dim(exprSet)</div><div class="line">[<span class="number">1</span>] <span class="number">28869</span>    <span class="number">11</span></div><div class="line">&gt; exprSet_trans_symbol &lt;- exprSet[rownames(exprSet) %<span class="keyword">in</span>% probe_trans_SYMBOL$probe_id,]</div><div class="line">&gt; dim(exprSet_trans_symbol)</div><div class="line">[<span class="number">1</span>] <span class="number">19631</span>    <span class="number">11</span></div></pre></td></tr></table></figure>
<p>从结果中可以看出来，表达谱中有28869行，第1列就是探针名，探针中有19631个能找到相应的基因SYMBOL，因此表达谱经过第一步操作，就从原来的28869行缩小到了19631行，再强调一下这一步的目的，就是筛选表达谱数据，将那些表达谱中有探针号，但没有对应基因SYMBOL的数据给剔除掉。</p>
<p>第二步：将表达谱中的这些探针号与基因名对应起来，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">probe_trans_SYMBOL_filter &lt;- probe_trans_SYMBOL[match(rownames(exprSet_trans_symbol), probe_trans_SYMBOL$probe_id),]</div></pre></td></tr></table></figure>
<p>这里用到了<code>match()</code>函数，简单看一个案例：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">A&lt;-<span class="number">1</span>:<span class="number">10</span></div><div class="line">B&lt;-seq(<span class="number">5</span>,<span class="number">15</span>,<span class="number">2</span>)</div><div class="line">match(A,B) <span class="comment">#遍历A中的元素，看A的元素是否在B里面出现过，如果A中的元素在B中出现，就返回此元素在B中的索引号</span></div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; match(A,B) <span class="comment">#遍历A中的元素，看A的元素是否在B里面出现过，如果A中的元素在B中出现，就返回此元素在B中的索引号</span></div><div class="line"> [<span class="number">1</span>] <span class="literal">NA</span> <span class="literal">NA</span> <span class="literal">NA</span> <span class="literal">NA</span>  <span class="number">1</span> <span class="literal">NA</span>  <span class="number">2</span> <span class="literal">NA</span>  <span class="number">3</span> <span class="literal">NA</span></div></pre></td></tr></table></figure>
<p>因此上面代码的目的在于：将表达谱中的探针位置与转换后的探针文件<code>probe_trans_SYMBOL</code>中的位置对应起来，先看一下这两个文件的长度：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; length(rownames(exprSet_trans_symbol))</div><div class="line">[<span class="number">1</span>] <span class="number">19631</span></div><div class="line">&gt; length(probe_trans_SYMBOL$probe_id)</div><div class="line">[<span class="number">1</span>] <span class="number">19827</span></div><div class="line">&gt; length(probe_trans_SYMBOL_filter$symbol)</div><div class="line">[<span class="number">1</span>] <span class="number">19631</span></div></pre></td></tr></table></figure>
<p>经过前面的<code>match()</code>函数操作，我们就可以看到，探针的位置已经与基因名一一对应了，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719175830.png">

</div>
<p>左侧是探针对应的文件，右侧是表达谱数据，由于现在这两个文件的探针与基因的位置都是一一对应的，使用<code>merge()</code>函数或<code>cbind()</code>都可以，由于名称对应，直接改名也行，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">rownames(exprSet_trans_symbol) &lt;- probe_trans_SYMBOL_filter$symbol</div><div class="line">str(exprSet_trans_symbol)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; str(exprSet_trans_symbol)</div><div class="line"> num [<span class="number">1</span>:<span class="number">19631</span>, <span class="number">1</span>:<span class="number">11</span>] <span class="number">7.27</span> <span class="number">8.05</span> <span class="number">7.47</span> <span class="number">7.52</span> <span class="number">7.08</span> <span class="keyword">...</span></div><div class="line"> - attr(*, <span class="string">"dimnames"</span>)=List of <span class="number">2</span></div><div class="line">  ..$ : chr [<span class="number">1</span>:<span class="number">19631</span>] <span class="string">"LINC01128"</span> <span class="string">"SAMD11"</span> <span class="string">"KLHL17"</span> <span class="string">"PLEKHN1"</span> <span class="keyword">...</span></div><div class="line">  ..$ : chr [<span class="number">1</span>:<span class="number">11</span>] <span class="string">"GSM607938"</span> <span class="string">"GSM607939"</span> <span class="string">"GSM607940"</span> <span class="string">"GSM607941"</span> <span class="keyword">...</span></div></pre></td></tr></table></figure>
<p>此时，这个表达谱是有19631行数据，但是，由于芯片的设计，一个基因名有可能对应好几个探针编号，查看一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; length(unique(rownames(exprSet_trans_symbol)))</div><div class="line">[<span class="number">1</span>] <span class="number">18775</span></div></pre></td></tr></table></figure>
<p>从这个结果可以看出来，其实基因一共是18775个，比19631要少，因此我们还要继续剔除掉一些相同的基因名，此时的问题就是，如果有3行的基因相同，要剔掉哪个？按教程中的标准，剔除标准是：只保留在各个样本中均值最大的那个基因名，这个处理过程的示意图如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719182201.png">

</div>
<p>在上表中，一共有4个样本，6个基因，其中<code>Gene_5</code>有3个重复，<code>Gene_6</code>有4个重复，因此我们要剔除掉那些重复的基因名，假设我们按照这2个基因在不同样本中的最高均值这个条件来筛选，那么<code>Gene_5</code>只需要保留第1行，<code>Gene_6</code>只需要保留第3行就行，代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">exprSet_trans_symbol2 &lt;- as.data.frame(exprSet_trans_symbol)</div><div class="line">exprSet_trans_symbol2$names &lt;-rownames(exprSet_trans_symbol)</div><div class="line">result &lt;- exprSet_trans_symbol2 %&gt;% group_by(names) %&gt;% summarise_all(max)</div><div class="line">result2 &lt;- as.matrix(result[-<span class="number">1</span>])</div><div class="line">rownames(result2) &lt;- result$names</div><div class="line">final_data &lt;- result2</div><div class="line">boxplot(final_data)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719191931.png">

</div>
<p>现在提取分组信息，在开头的代码中，我们把分组信息储存在了<code>pData</code>变量中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(stringr)</div><div class="line">group_sample &lt;- pData$title</div><div class="line">group_clean &lt;- str_sub(group_sample,<span class="number">1</span>,<span class="number">6</span>)</div><div class="line">group_clean[<span class="number">10</span>:<span class="number">11</span>] &lt;- rep(<span class="string">"Health"</span>,<span class="number">2</span>)</div></pre></td></tr></table></figure>
<p>有了分组信息后，把原始表达谱再整理一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(reshape2)</div><div class="line">exprSet_L &lt;- melt(final_data)</div><div class="line">colnames(exprSet_L) &lt;- c(<span class="string">"probe"</span>, <span class="string">"Sample"</span>, <span class="string">"value"</span>)</div><div class="line">exprSet_L$group &lt;- rep(group_clean, each=nrow(final_data))</div></pre></td></tr></table></figure>
<p>绘图：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ggplot2)</div><div class="line">p &lt;- ggplot(exprSet_L, aes(x=Sample, y=value, fill=group))+geom_boxplot()</div><div class="line">p &lt;- p + theme(axis.text.x = element_text(angle = <span class="number">45</span>, hjust = <span class="number">0.5</span>, vjust = <span class="number">0.5</span>))</div><div class="line">p</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190719194204.png">

</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247486063&amp;idx=1&amp;sn=156bee5397e979722b36b78284188538&amp;chksm=9b484ad4ac3fc3c2d025b9e4bb1c3c8392839c08d84697754d7d95d041b539479a45f19cf5d5&amp;scene=21#wechat_redirect" target="_blank" rel="external">解读GEO数据存放规律及下载，一文就够</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/GEO/" <i class="fa fa-tag"></i> GEO</a>
          
            <a href="/tags/公共数据库挖掘/" <i class="fa fa-tag"></i> 公共数据库挖掘</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/05/28/Experiment/siRNA干扰实验流程/" rel="next" title="siRNA干扰实验流程">
                <i class="fa fa-chevron-left"></i> siRNA干扰实验流程
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/06/14/Python/Python学习笔记（19）-Python与生物信息学 - 副本/" rel="prev" title="Python学习笔记（19）-Python与生物信息学">
                Python学习笔记（19）-Python与生物信息学 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">223</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">13</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">109</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#背景知识"><span class="nav-number">2.</span> <span class="nav-text">背景知识</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#下载数据"><span class="nav-number">3.</span> <span class="nav-text">下载数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#获取表达矩阵"><span class="nav-number">4.</span> <span class="nav-text">获取表达矩阵</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#提取表达矩阵"><span class="nav-number">4.1.</span> <span class="nav-text">提取表达矩阵</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#获取样本信息"><span class="nav-number">4.2.</span> <span class="nav-text">获取样本信息</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#表达矩阵过滤"><span class="nav-number">5.</span> <span class="nav-text">表达矩阵过滤</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#转换探针名"><span class="nav-number">5.0.1.</span> <span class="nav-text">转换探针名</span></a></li></ol></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">6.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共890.3k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
