<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Linux,RNA-seq,基因差异," />










<meta name="description" content="前言 这是曾健明生物信息教程的第二部分，原始数据位于/public/study/mRNAseq/tair/中，为方便起见，可以添加到系统变量，`alias rnaseq=“cd /public/study/mRNAseq/tair/”``这是一个拟南芥转录组数据，整个流程的操作可看博客：http://www.bio-info-trainee.com/2809.html。 练习所用数据介绍 数">
<meta name="keywords" content="Linux,RNA-seq,基因差异">
<meta property="og:type" content="article">
<meta property="og:title" content="拟南芥转录组项目的实战">
<meta property="og:url" content="http://rvdsd.top/2018/04/20/生信-拟南芥转录组项目的实战/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 这是曾健明生物信息教程的第二部分，原始数据位于/public/study/mRNAseq/tair/中，为方便起见，可以添加到系统变量，`alias rnaseq=“cd /public/study/mRNAseq/tair/”``这是一个拟南芥转录组数据，整个流程的操作可看博客：http://www.bio-info-trainee.com/2809.html。 练习所用数据介绍 数">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://www.bio-info-trainee.com/wp-content/uploads/2017/10/overview-of-RNA-seq-technology.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/01.png">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/0F6d3ae0Ab.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/4i9FDihi44.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/4AaHhB7D73.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/j215KeB9bD.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/7gb25H9cfe.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180401/DC3BJcA0II.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180402/I8cg2DHA9c.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180403/3Ab775ecI6.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180403/gI545KkDa7.png?imageslim">
<meta property="og:updated_time" content="2018-04-26T03:41:20.173Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="拟南芥转录组项目的实战">
<meta name="twitter:description" content="前言 这是曾健明生物信息教程的第二部分，原始数据位于/public/study/mRNAseq/tair/中，为方便起见，可以添加到系统变量，`alias rnaseq=“cd /public/study/mRNAseq/tair/”``这是一个拟南芥转录组数据，整个流程的操作可看博客：http://www.bio-info-trainee.com/2809.html。 练习所用数据介绍 数">
<meta name="twitter:image" content="http://www.bio-info-trainee.com/wp-content/uploads/2017/10/overview-of-RNA-seq-technology.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2018/04/20/生信-拟南芥转录组项目的实战/"/>





  <title>拟南芥转录组项目的实战 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">学习过程中的输出</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/04/20/生信-拟南芥转录组项目的实战/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">拟南芥转录组项目的实战</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-04-20T12:00:00+08:00">
                2018-04-20
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信实战项目/" itemprop="url" rel="index">
                    <span itemprop="name">生信实战项目</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,004
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  13
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>这是曾健明生物信息教程的第二部分，原始数据位于<code>/public/study/mRNAseq/tair/</code>中，为方便起见，可以添加到系统变量，`alias rnaseq=“cd /public/study/mRNAseq/tair/”``这是一个拟南芥转录组数据，整个流程的操作可看博客：<a href="http://www.bio-info-trainee.com/2809.html" class="uri" target="_blank" rel="external">http://www.bio-info-trainee.com/2809.html</a>。</p>
<h2 id="练习所用数据介绍">练习所用数据介绍</h2>
<p>数据来自于发表在Nature commmunication 上的一篇文章 “<a href="https://www.nature.com/articles/ncomms15120" target="_blank" rel="external">Temporal dynamics of gene expression and histone marks at the Arabidopsis shoot meristem during flowerin</a>”。原文用RNA-Seq的方式研究在开花阶段,芽分生组织在不同时期的基因表达变化。</p>
<p>原文的流程是： TopHat -&gt; SummarizeOverlaps -&gt; Deseq2 -&gt; AmiGO 其中比对的参考基因组为TAIR10 ver.24 ，并且屏蔽了ribosomal RNA regions。 Deseq2只计算至少在一个时间段的FPKM的count &gt; 1 的基因。 数据存放在<a href="http://www.ebi.ac.uk/arrayexpress" class="uri" target="_blank" rel="external">http://www.ebi.ac.uk/arrayexpress</a>/, ID为E-MTAB-5130。</p>
<p>实验设计： 4个时间段（0,1,2,3），分别有4个生物学重复，一共有16个样品。</p>
<h2 id="相关概念">相关概念</h2>
<blockquote>
<p>转录组测序的研究对象为特定细胞在某一功能状态下所能转录出来的所有 RNA 的总和，包括 mRNA 和非编码 RNA 。通过转录组测序，能够全面获得物种特定组织或器官的转录本信息，从而进行转录本结构研究、变异研究、基因表达水平研究以及全新转录本发现等研究。</p>
</blockquote>
<blockquote>
<p>其中，基因表达水平的探究是转录组领域最热门的方向，利用转录组数据来识别转录本和表达定量，是转录组数据的核心作用。由于这个作用，他可以不依赖其他组学信息，单独成为一个产品项目RNA-seq测序。所以很多时候转录组测序会与RNA-seq混为一谈。</p>
</blockquote>
<blockquote>
<p>现在RNA-seq数据使用广泛，但是没有一套流程可以解决所有的问题。比较值得关注的RNA-seq分析中的重要的步骤包括：实验设计，质控，read比对，表达定量，可视化，差异表达，识别可变剪切，功能注释，融合基因检测，eQTL定位等。</p>
</blockquote>
<h3 id="流程如下图所示">流程如下图所示：</h3>
<div class="figure">
<img src="http://www.bio-info-trainee.com/wp-content/uploads/2017/10/overview-of-RNA-seq-technology.png">

</div>
<h2 id="分析流程">分析流程</h2>
<h3 id="第1步下载数据">第1步：下载数据</h3>
<p>由于数据已经放在了服务器，如下所示，不用下载，直接在服务器里面分析，这一步略过。 <img src="http://ortxpw68f.bkt.clouddn.com/01.png"> 这几个数据是： 1. 转录本：Arabidopsis_thaliana.TAIR10.28.cdna.all.fa.gz 2. 基因组：Arabidopsis_thaliana.TAIR10.28.dna.genome.fa.gz 3. 注释文件：Arabidopsis_thaliana.TAIR10.28.dna.genome.fa.gz 4. 注释文件：Arabidopsis_thaliana.TAIR10.28.gtf.gz 5. sample_fastq：测序的原始数据 6. sample_sra：呈交给SRA数据库的数据（解压后就是sample_fastq）</p>
<h3 id="第2步基因定量">第2步：基因定量</h3>
<h4 id="第1种方法salmon方法">第1种方法：salmon方法</h4>
<p><a href="https://combine-lab.github.io/salmon/" target="_blank" rel="external">Salmon软件</a>：对RNA-seq数据进行转录本定量的工具，Salmon使用了一种新的算法（主要是specifically, coupling the concept of quasi-mapping with a two-phase inference procedure），该工具能够快速地进行转录本的定量，并且所耗内存极小。此工具进行运算的过程使用了一种可靠的模型，此模式考虑了实验特性以及RNA-seq中常见的偏倚。</p>
<h5 id="安装并添加环境变量">安装并添加环境变量</h5>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">wget https://github.com/COMBINE-lab/salmon/releases/download/v0.9.1/Salmon-0.9.1_linux_x86_64.tar.gz</div><div class="line"></div><div class="line">export PATH=$PATH:/trainee/home/jlzhang/download/Salmon-latest_linux_x86_64/bin</div></pre></td></tr></table></figure>
<h5 id="第1步建立索引如下所示">第1步：建立索引，如下所示：</h5>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">salmon index -t Arabidopsis_thaliana.TAIR10.28.cdna.all.fa.gz -i /trainee/home/jlzhang/arab/</div></pre></td></tr></table></figure>
<p>建立索引的过程如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/0F6d3ae0Ab.png?imageslim" alt="mark"></p>
<p>切换到生成索引的目录<code>/trainee/home/jlzhang/arab/</code>，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/4i9FDihi44.png?imageslim" alt="mark"></p>
<p>注：输出目录要写要自己名下，如果是公共目录，则无权限写入，会出错</p>
<h5 id="第2步数据定量">第2步：数据定量</h5>
<p>salmon不进行比对，直接进行定量</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">index = /trainee/home/jlzhang/arab</div><div class="line"># 索引文件夹</div><div class="line">for fn in ERR1698&#123;194..209&#125;;</div><div class="line">do</div><div class="line">    sample=`basename $&#123;fn&#125;`</div><div class="line">    echo &quot;Proicessin sample $&#123;sample&#125;&quot;</div><div class="line">salmon quant -i $index -l A -1 /public/study/mRNAseq/tair/sample_fastq/$&#123;sample&#125;_1.fastq.gz -2 /public/study/mRNAseq/tair/sample_fastq/$&#123;sample&#125;_2.fastq.gz  -p 5 -o quants/$&#123;sample&#125;_quant</div><div class="line">done</div></pre></td></tr></table></figure>
<p>注：上述命名-l A 后面没有“”符号。</p>
<p>运行上述命令，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/4AaHhB7D73.png?imageslim" alt="mark"> 整个过程大概有7分钟，计算结束后，会在quants目录（<code>/trainee/home/jlzhang/quants/ERR1698194_quant</code>）下生成各个样本的目录，以及相应的数据，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/j215KeB9bD.png?imageslim" alt="mark"> 其中<code>quant.sf</code>文件是基因的定量数据，打开，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/7gb25H9cfe.png?imageslim" alt="mark"></p>
<p>为方便查看，用表格表示，如下所示：</p>
<table>
<thead>
<tr class="header">
<th>Name</th>
<th>Length</th>
<th>EffectiveLength</th>
<th>TPM</th>
<th>NumReads</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ATMG00010.1</td>
<td>462</td>
<td>299.317</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>ATMG00030.1</td>
<td>324</td>
<td>165.898</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>ATMG00040.1</td>
<td>948</td>
<td>784.352</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>ATMG00050.1</td>
<td>396</td>
<td>234.567</td>
<td>10.33454</td>
<td>13</td>
</tr>
<tr class="odd">
<td>ATMG00060.1</td>
<td>542</td>
<td>378.761</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>ATMG00070.1</td>
<td>573</td>
<td>409.668</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>ATMG00080.1</td>
<td>540</td>
<td>376.77</td>
<td>68.35285</td>
<td>138.1076</td>
</tr>
<tr class="even">
<td>ATMG00090.1</td>
<td>1671</td>
<td>1507.288</td>
<td>0.785473</td>
<td>6.349101</td>
</tr>
<tr class="odd">
<td>ATMG00100.1</td>
<td>73</td>
<td>15.492</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>这于<code>quant.sf</code>文件的说明，在salmon的官网上已经有详细介绍，如下所示：</p>
<ul>
<li>Name — This is the name of the target transcript provided in the input transcript database (FASTA file).</li>
<li>Length — This is the length of the target transcript in nucleotides.</li>
<li>EffectiveLength — This is the computed effective length of the target transcript. It takes into account all factors being modeled that will effect the probability of sampling fragments from this transcript, including the fragment length distribution and sequence-specific and gc-fragment bias (if they are being modeled).</li>
<li>TPM — This is salmon’s estimate of the relative abundance of this transcript in units of Transcripts Per Million (TPM). TPM is the recommended relative abundance measure to use for downstream analysis.</li>
<li>NumReads — This is salmon’s estimate of the number of reads mapping to each transcript that was quantified. It is an “estimate” insofar as it is the expected number of reads that have originated from each transcript given the structure of the uniquely mapping and multi-mapping reads and the relative abundance estimates for each transcript.S</li>
</ul>
<p>翻译成中文就是：</p>
<ul>
<li>Name：这个目标转录组的名称（源于转录组数据库的fasta文件）；</li>
<li>Length：目标转录组的碱基长度；</li>
<li>EffectiveLenght：这是目标转录组的有效长度。它考虑了所有会影响该转录本采样片段概率的因素，其中包括长度分页，序列特异性和GC片段偏好。</li>
<li>TPM：这是salmon对转录本相对丰度的估计，其单位是百分率（TPM，Transcripts Per Million）。TPM可以用于下游的定量分析。</li>
<li>NumReads：salmon对回贴（mapping）到每个转录本的reads数的估计。</li>
</ul>
<p>到这里，用salmon进行定量的过程已经结束，此处并没有进行不同的样本间的差异基因分析。接着看第2种方法进行定量，即用subread进行定量。</p>
<h4 id="第2种方法subread方法">第2种方法：subread方法</h4>
<p>subread是一个reads比对工具。官网是：<a href="http://subread.sourceforge.net/" class="uri" target="_blank" rel="external">http://subread.sourceforge.net/</a></p>
<h5 id="第1步建立索引">第1步：建立索引</h5>
<p>subread构建索引时需要解压fa文件，因此先将服务器中的参考基因组复制到自己的目录下，然后建立索引，如下所示： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">mkdir index_file</div><div class="line">cp /public/study/mRNAseq/tair/Arabidopsis_thaliana.TAIR10.28.dna.genome.fa.gz  ~/index_file/</div><div class="line">cd index_file</div><div class="line">gunzip Arabidopsis_thaliana.TAIR10.28.dna.genome.fa.gz</div><div class="line">/public/biosoft/subread/subread-1.5.3-Linux-x86_64/bin/subread-buildindex  -o athal_index Arabidopsis_thaliana.TAIR10.28.dna.genome.fa</div></pre></td></tr></table></figure></p>
<p>建立后的文件如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180401/DC3BJcA0II.png?imageslim" alt="mark"></p>
<h5 id="第2步比对">第2步：比对</h5>
<p>建立一个脚本<code>subjunc.sh</code>，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">#! /bin/bash</div><div class="line">subjunc=&quot;/public/biosoft/subread/subread-1.5.3-Linux-x86_64/bin/subjunc&quot;; </div><div class="line">index=&apos;/trainee/home/jlzhang/index_file&apos;;</div><div class="line">for fn in ERR1698&#123;194..209&#125;;</div><div class="line">do</div><div class="line">    sample=`basename $&#123;fn&#125;`</div><div class="line">    echo &quot;Processin sample $&#123;sampe&#125;&quot; </div><div class="line">    $subjunc -i $index -r  /public/study/mRNAseq/tair/sample_fastq/$&#123;sample&#125;_1.fastq.gz -R /public/study/mRNAseq/tair/sample_fastq/$&#123;sample&#125;_2.fastq.gz -T 5 -o /trainee/home/jlzhang/quant2/$&#123;sample&#125;_subjunc.bam</div><div class="line">done</div></pre></td></tr></table></figure>
<p>运行，即<code>bash subjunc.sh</code>，这个过程大概需要一段时间，运行结果如下所示：</p>
<h5 id="第3步定量与表达矩阵">第3步：定量与表达矩阵</h5>
<p>用featureCount进行定量：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">/public/biosoft/subread/subread-1.5.3-Linux-x86_64/bin/featureCounts -T 5 -p -t exon -g gene_name -a /trainee/home/jlzhang/download/Arabidopsis_thaliana.TAIR10.28.gtf -o  counts.txt   *.bam</div></pre></td></tr></table></figure>
<p>注：服务器中的Arabidopsis_thaliana.TAIR10.28.gtf.gz是压缩文件，需要解压后才能使用。</p>
<p>featureCounts的参数：</p>
<ul>
<li>T 多线程，5表示5个线程</li>
<li>p 双端测序</li>
<li>t 默认将Exon作为一个feature</li>
<li>g 从注释文件中提取Meta-features信息用于read coutn，默认是gene_id</li>
<li>a 输入GTF/GFF基因组注释文件</li>
<li>o 输出文件</li>
</ul>
<p>比对后的结果如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180402/I8cg2DHA9c.png?imageslim" alt="mark"></p>
<h5 id="第4步分组矩阵">第4步：分组矩阵</h5>
<p>构建分组矩阵：</p>
<p><code>/trainee/home/huanggy/RNAseq/data/fastq/E-MTAB-5130.sdrf.txt</code>这个文件如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180403/3Ab775ecI6.png?imageslim" alt="mark"> 其中第1列和第32列都是样本名称，第32列是测序的样本名称，这个有用，第36列则是分组的信息，因此提取分组信息，代码如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">tail -n +2 /trainee/home/huanggy/RNAseq/data/fastq/E-MTAB-5130.sdrf.txt | cut -f 32,36 |sort -u</div></pre></td></tr></table></figure>
<p><code>sort -u</code>则是表示排序，并去重（这个是双端测序，因此在提取分组信息方面，需要去重），运行完上述代码后，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180403/gI545KkDa7.png?imageslim" alt="mark"></p>
<p>tail -n +2表示是从文件的第2行开始显示文件的内容</p>
<h3 id="第4步其他比对方法">第4步：其他比对方法</h3>
<h1 id="第5步表达矩阵的normalization方法">第5步：表达矩阵的normalization方法</h1>
<blockquote>
<p>总reads数矫正(TC)：基因计数除以与其land相关的回贴的reads数（mapped reads或文库大小），再乘以数据集所有样本的平均总计数。</p>
</blockquote>
<blockquote>
<p>上四分位数(UQ)：与TC在原理上非常相似，在计算归一化因子时，总reads数被不同于0的计数的上四分位数所取代。</p>
</blockquote>
<blockquote>
<p>中位数(MED)：类似于TC，在计算标准化因子时，总reads数由与0不同的中位数计数代替。也就是说，中位数被计算为所有运行的基因计数的中值。</p>
</blockquote>
<blockquote>
<p>DESeq：这种标准化方法包含在DESeq Bioconducter包中，并基于大多数基因不是DE的假设。该方法基于一个负二项分布模型，通过局部回归将方差和均值联系起来，并给出了一个给出标度因子的实现方法。在DESeq包中，使用EstimatateSizeFactorsForMatrixfunction，可以为每次运行计算标度因子。将基因计数除以每个比例因子后，计算DESeq值作为所有运行的重标度基因计数的总和。</p>
</blockquote>
<blockquote>
<p>加权截断均值：TMM（edgeR）：与Deseq类似，在去除高表达基因和差异最大的基因后，TMM也是要找到一个加权系数，使剩余的基因在被矫正后差异倍数可能小。TMM的加权系数是基于两两样本比较后推算获得的（也就是两组样本的比较，将产生与这次比较相关的加权系数）。然后将所有基因除以这个加权系数，从而保证大部分表达量居中的基因表达量最相似。</p>
</blockquote>
<blockquote>
<p>分位数（Q）：首先在微阵列数据的背景下提出，这种标准化方法包括匹配泳道间基因计数的分布。</p>
</blockquote>
<blockquote>
<p>每千碱基每百万映射读取(RPKM)：这种方法最初是为了促进样本中基因之间的比较，并结合样本之间和样本内的标准化。该方法通过归一化总转录物长度和测序读数的数量来量化来自RNA-Seq数据的基因表达。</p>
</blockquote>
<h3 id="第6步差异分析">第6步：差异分析</h3>
<p>在这一步中，作者原博客中直接给出了差异分析的脚本，直接下载即可，脚本文件一共有3个，分别为<code>run_DEG.R</code>,<code>exprSet.txt</code>,<code>group_info.txt</code>，其中<code>run_DEG.R</code>是最主要的脚本文件，它是用R语言写成的,<code>exprSet.txt</code>是所有样本的表达矩阵，这是原始的数据，并没有经过Normalization，<code>group_info.txt</code>是分组信息，直接输入下面的命令即可：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">Rscript ../run_DEG.R -e exprSet.txt -g group_info.txt -c <span class="string">'Day1-Day0'</span> -s counts  -m DESeq2</div></pre></td></tr></table></figure>
<p>如果是转录组的raw counts数据，就选择 -s counts（这个案例中就是这个选项），如果是芯片等normalization好的表达矩阵数据，用默认参数即可，对于转录组的raw counts数据，有DEseq2包和edgeR包可供选择。对于芯片等normalization好的表达矩阵数据，有limma和t.test供选择，下面是举例说明：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"># Rscript run_DEG.R -e airway.expression.txt -g airway.group.txt -c &apos;trt-untrt&apos; -s counts -m DESeq2</div><div class="line"># Rscript run_DEG.R -e airway.expression.txt -g airway.group.txt -c &apos;trt-untrt&apos; -s counts -m edgeR</div><div class="line"># Rscript run_DEG.R -e sCLLex.expression.txt -g sCLLex.group.txt -c &apos;progres.-stable&apos;</div><div class="line"># Rscript run_DEG.R -e sCLLex.expression.txt -g sCLLex.group.txt -c &apos;progres.-stable&apos; -m t.test</div></pre></td></tr></table></figure>
<h3 id="第7步富集分析">第7步：富集分析</h3>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://github.com/twbattaglia/RNAseq-workflow" target="_blank" rel="external">Differential Gene Expression using RNA-Seq (Workflow)</a></li>
<li><a href="http://www.bio-info-trainee.com/2809.html" target="_blank" rel="external">一个植物转录组项目的实战</a></li>
<li><a href="http://biocluster.ucr.edu/~rkaundal/workshops/R_mar2016/RNAseq.html" target="_blank" rel="external">NGS sequence analysis with R / Bioconductor: RNA-Seq workflow</a></li>
<li><a href="https://www.jianshu.com/p/7564eca62354" target="_blank" rel="external">基因表达分析（前传）-准备count矩阵</a></li>
<li>http://www.360doc.com/content/17/0921/09/2856980_688849117.shtml</li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Linux/" <i class="fa fa-tag"></i> Linux</a>
          
            <a href="/tags/RNA-seq/" <i class="fa fa-tag"></i> RNA-seq</a>
          
            <a href="/tags/基因差异/" <i class="fa fa-tag"></i> 基因差异</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/03/31/生信-Jupyter nook主题设置/" rel="next" title="Jupyter notebook主题的设置">
                <i class="fa fa-chevron-left"></i> Jupyter notebook主题的设置
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/04/21/生信-Linux环境变量/" rel="prev" title="Linux环境变量">
                Linux环境变量 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">59</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">8</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">52</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#练习所用数据介绍"><span class="nav-number">2.</span> <span class="nav-text">练习所用数据介绍</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#相关概念"><span class="nav-number">3.</span> <span class="nav-text">相关概念</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#流程如下图所示"><span class="nav-number">3.1.</span> <span class="nav-text">流程如下图所示：</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#分析流程"><span class="nav-number">4.</span> <span class="nav-text">分析流程</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#第1步下载数据"><span class="nav-number">4.1.</span> <span class="nav-text">第1步：下载数据</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#第2步基因定量"><span class="nav-number">4.2.</span> <span class="nav-text">第2步：基因定量</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#第1种方法salmon方法"><span class="nav-number">4.2.1.</span> <span class="nav-text">第1种方法：salmon方法</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#安装并添加环境变量"><span class="nav-number">4.2.1.1.</span> <span class="nav-text">安装并添加环境变量</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#第1步建立索引如下所示"><span class="nav-number">4.2.1.2.</span> <span class="nav-text">第1步：建立索引，如下所示：</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#第2步数据定量"><span class="nav-number">4.2.1.3.</span> <span class="nav-text">第2步：数据定量</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#第2种方法subread方法"><span class="nav-number">4.2.2.</span> <span class="nav-text">第2种方法：subread方法</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#第1步建立索引"><span class="nav-number">4.2.2.1.</span> <span class="nav-text">第1步：建立索引</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#第2步比对"><span class="nav-number">4.2.2.2.</span> <span class="nav-text">第2步：比对</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#第3步定量与表达矩阵"><span class="nav-number">4.2.2.3.</span> <span class="nav-text">第3步：定量与表达矩阵</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#第4步分组矩阵"><span class="nav-number">4.2.2.4.</span> <span class="nav-text">第4步：分组矩阵</span></a></li></ol></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#第4步其他比对方法"><span class="nav-number">4.3.</span> <span class="nav-text">第4步：其他比对方法</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#第5步表达矩阵的normalization方法"><span class="nav-number"></span> <span class="nav-text">第5步：表达矩阵的normalization方法</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#第6步差异分析"><span class="nav-number">0.1.</span> <span class="nav-text">第6步：差异分析</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#第7步富集分析"><span class="nav-number">0.2.</span> <span class="nav-text">第7步：富集分析</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">1.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_uv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共191.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
