<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor,基因组注释,KEGG,GO," />










<meta name="description" content="前言 这一部分内容涉及R中使用人类基因且，内含子，外显子，转录本，AnnotationHub，基因组的注释包，GO分析，KEGG分析等，笔记末尾的参考文献是原文。 基础注释资源与发现 在这一部分里，我们将回顾Bioconductor中用于处理和注释基因组序列的一些工具。我们将研究参考基因组序列，转录本和基因，并以基因通路(gene pathway)作为结束。我们学习这一部分的最终目标就是使用">
<meta name="keywords" content="Bioconductor,基因组注释,KEGG,GO">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS009-Bioconductor中的基因组注释">
<meta property="og:url" content="http://rvdsd.top/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 这一部分内容涉及R中使用人类基因且，内含子，外显子，转录本，AnnotationHub，基因组的注释包，GO分析，KEGG分析等，笔记末尾的参考文献是原文。 基础注释资源与发现 在这一部分里，我们将回顾Bioconductor中用于处理和注释基因组序列的一些工具。我们将研究参考基因组序列，转录本和基因，并以基因通路(gene pathway)作为结束。我们学习这一部分的最终目标就是使用">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_annoOverview-getp-1.png">
<meta property="og:updated_time" content="2019-11-13T08:51:12.610Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS009-Bioconductor中的基因组注释">
<meta name="twitter:description" content="前言 这一部分内容涉及R中使用人类基因且，内含子，外显子，转录本，AnnotationHub，基因组的注释包，GO分析，KEGG分析等，笔记末尾的参考文献是原文。 基础注释资源与发现 在这一部分里，我们将回顾Bioconductor中用于处理和注释基因组序列的一些工具。我们将研究参考基因组序列，转录本和基因，并以基因通路(gene pathway)作为结束。我们学习这一部分的最终目标就是使用">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_annoOverview-getp-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/"/>





  <title>GDAS009-Bioconductor中的基因组注释 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS009-Bioconductor中的基因组注释</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-09T12:00:00+08:00">
                2019-09-09
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  6,586
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  31
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>这一部分内容涉及R中使用人类基因且，内含子，外显子，转录本，AnnotationHub，基因组的注释包，GO分析，KEGG分析等，笔记末尾的参考文献是原文。</p>
<h2 id="基础注释资源与发现">基础注释资源与发现</h2>
<p>在这一部分里，我们将回顾Bioconductor中用于处理和注释基因组序列的一些工具。我们将研究参考基因组序列，转录本和基因，并以基因通路(gene pathway)作为结束。我们学习这一部分的最终目标就是使用注释信息来帮助我们对基因组实验进行可靠的解释。Bioconductor的基本目标就是更加方便地有关基因组结构和功能的信息统计统计分析程序。</p>
<h3 id="注释概念的层次结构">注释概念的层次结构</h3>
<p>Bioconductor包括许多不同类型的基因组注释。我们可以在层次结构中来理解这些注释资源。</p>
<ul>
<li>最基因的注释就是某个物种的参考基因组序列。它总是按照核苷酸的线性方式排列成染色体（例如参考基因组。</li>
<li>在此之上就是将染色体序列排列到感兴趣的区域中。最感兴趣的区域就是基因，但是注释中也含有其它的信息，例如SNP或CpG位点。基因具有内部结构，即被转录的部分和未被转录的部分。“基因模式”定义了在基因组坐标中的标记和布置这些结构的方式。</li>
<li>在感兴趣的区域(regions of interest)的理念下，我们还定义了面向平台的注释(platform-oriented annotation)。这处类型的注释通常首先是由厂家提供的，但随着研究的进行，对这些平台中最初有歧义信息进行了确认和更新，从而完善了这些注释内容。密歇根大学的<a href="http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp" target="_blank" rel="external">brainarray project</a> 说明了affymetrix阵列注释的过程。我们将在本节最后讨论面向平台注释的问题。</li>
<li>在此之是是将区域（通常是基因或基因的产物）组成成具有共同结构或功能特性的组。例如在细胞中共同被发现的，或者是被鉴定为在生物学过程中协同作用的基因组（我的理解就是GO分析，KEGG分析这一类）。</li>
</ul>
<h3 id="发现可用的参考基因组">发现可用的参考基因组</h3>
<p>Bioconductor已经包含了注释包的合成，将它这一层次结构上的所有元素都带了可编程环境中。参考基因组序列是使用Biostrings和BSgenome包中的工具进行管理的，<code>available.genomes</code> 函数能够列出构建好的人和现在各种模式生物的参考基因组，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">library(Biostrings)</div><div class="line">ag = available.genomes()</div><div class="line">length(ag)</div><div class="line">## [1] 87</div><div class="line">head(ag)</div><div class="line">## [1] &quot;BSgenome.Alyrata.JGI.v1&quot;                </div><div class="line">## [2] &quot;BSgenome.Amellifera.BeeBase.assembly4&quot;  </div><div class="line">## [3] &quot;BSgenome.Amellifera.UCSC.apiMel2&quot;       </div><div class="line">## [4] &quot;BSgenome.Amellifera.UCSC.apiMel2.masked&quot;</div><div class="line">## [5] &quot;BSgenome.Athaliana.TAIR.04232008&quot;       </div><div class="line">## [6] &quot;BSgenome.Athaliana.TAIR.TAIR9&quot;</div></pre></td></tr></table></figure>
<h3 id="参考基因组的版本很重要">参考基因组的版本很重要</h3>
<p>不同物种的参考基因组是从头构建的，然后随着算法和测序数据的不断改进而进一步完善。对人类而言，基因组研究联盟（Genome Research Consortium）于2009年构建了37号版本，并于2013年构建了38号版本。</p>
<p>一旦参考基因组构建完成，就哦可以很轻松地对某个物种进行信息丰富的基因组序列分析，因为人们可以专注于那引起已知含有等位基因多样性的区域。</p>
<p>The reference build for an organism is created de novo and then refined as algorithms and sequenced data improve. For humans, the Genome Research Consortium signed off on build 37 in 2009, and on build 38 in 2013.</p>
<p>需要注意的是，基因组序列包含有很长的名称，这个名称里包括版本信息。这样命名的方式就是为了避免与不同版本的参考基因组混淆。在LiftOver这节视频里，我们就展示了如何使用UCSC的<code>liftOver</code>工具与<code>rtracklayer</code>包中的接口对接，从而实现不同版本的基因组坐标转化的过程。</p>
<p>为了帮助用户避免混淆从不同参考基因组坐标上收集分析来的数据，我们提供了一个”基因组“标签，这个标签填充了大多关于序列的信息。在随后的部分里，我们会看到一些案例。用于序列比对的软件可以检查被比对上的序列的兼容标签，从而有助于确保有意义的结果。</p>
<h2 id="h.-sapiens的参考基因序列">H. sapiens的参考基因序列</h2>
<p>通过安装并添加一个单独的R包就能获取智人(Homo sapiens)的参考序列。这个程序包定义了一个<code>Hsapiens</code>对象，试剂公司对象是染色体序列的来源，但是当对其进行单独显示时，它会提供相关序列数据来源的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">Hsapiens</div><div class="line"><span class="comment">## Human genome:</span></div><div class="line"><span class="comment">## # organism: Homo sapiens (Human)</span></div><div class="line"><span class="comment">## # provider: UCSC</span></div><div class="line"><span class="comment">## # provider version: hg19</span></div><div class="line"><span class="comment">## # release date: Feb. 2009</span></div><div class="line"><span class="comment">## # release name: Genome Reference Consortium GRCh37</span></div><div class="line"><span class="comment">## # 93 sequences:</span></div><div class="line"><span class="comment">## #   chr1                  chr2                  chr3                 </span></div><div class="line"><span class="comment">## #   chr4                  chr5                  chr6                 </span></div><div class="line"><span class="comment">## #   chr7                  chr8                  chr9                 </span></div><div class="line"><span class="comment">## #   chr10                 chr11                 chr12                </span></div><div class="line"><span class="comment">## #   chr13                 chr14                 chr15                </span></div><div class="line"><span class="comment">## #   ...                   ...                   ...                  </span></div><div class="line"><span class="comment">## #   chrUn_gl000235        chrUn_gl000236        chrUn_gl000237       </span></div><div class="line"><span class="comment">## #   chrUn_gl000238        chrUn_gl000239        chrUn_gl000240       </span></div><div class="line"><span class="comment">## #   chrUn_gl000241        chrUn_gl000242        chrUn_gl000243       </span></div><div class="line"><span class="comment">## #   chrUn_gl000244        chrUn_gl000245        chrUn_gl000246       </span></div><div class="line"><span class="comment">## #   chrUn_gl000247        chrUn_gl000248        chrUn_gl000249       </span></div><div class="line"><span class="comment">## # (use 'seqnames()' to see all the sequence names, use the '$' or '[['</span></div><div class="line"><span class="comment">## # operator to access a given sequence)</span></div><div class="line">head(genome(Hsapiens))  <span class="comment"># see the tag</span></div><div class="line"><span class="comment">##   chr1   chr2   chr3   chr4   chr5   chr6 </span></div><div class="line"><span class="comment">## "hg19" "hg19" "hg19" "hg19" "hg19" "hg19"</span></div></pre></td></tr></table></figure>
<p>我们使用 <code>$</code> 符号来获取17号染色体的序列，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">Hsapiens$chr17</div><div class="line">##   81195210-letter &quot;DNAString&quot; instance</div><div class="line">## seq: AAGCTTCTCACCCTGTTCCTGCATAGATAATTGC...GGTGTGGGTGTGGTGTGTGGGTGTGGGTGTGGT</div></pre></td></tr></table></figure>
<h2 id="参考序列的转录本和基因">参考序列的转录本和基因</h2>
<h3 id="ucsc注释">UCSC注释</h3>
<p><code>TxDb</code>包家族和数据对象管理了转录本和基因模式信息。我们可以认为这些信息来源于UCSC基因组浏览器的注释表，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene # abbreviate</div><div class="line">txdb</div><div class="line">## TxDb object:</div><div class="line">## # Db type: TxDb</div><div class="line">## # Supporting package: GenomicFeatures</div><div class="line">## # Data source: UCSC</div><div class="line">## # Genome: hg19</div><div class="line">## # Organism: Homo sapiens</div><div class="line">## # Taxonomy ID: 9606</div><div class="line">## # UCSC Table: knownGene</div><div class="line">## # Resource URL: http://genome.ucsc.edu/</div><div class="line">## # Type of Gene ID: Entrez Gene ID</div><div class="line">## # Full dataset: yes</div><div class="line">## # miRBase build ID: GRCh37</div><div class="line">## # transcript_nrow: 82960</div><div class="line">## # exon_nrow: 289969</div><div class="line">## # cds_nrow: 237533</div><div class="line">## # Db created by: GenomicFeatures package from Bioconductor</div><div class="line">## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)</div><div class="line">## # GenomicFeatures version at creation time: 1.21.30</div><div class="line">## # RSQLite version at creation time: 1.0.0</div><div class="line">## # DBSCHEMAVERSION: 1.1</div></pre></td></tr></table></figure>
<p>我们使用 <code>genes()</code> 来获取Entrez Gene ID的地址，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">ghs = genes(txdb)</div><div class="line">ghs</div><div class="line">## GRanges object with 23056 ranges and 1 metadata column:</div><div class="line">##         seqnames                 ranges strand |     gene_id</div><div class="line">##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</div><div class="line">##       1    chr19 [ 58858172,  58874214]      - |           1</div><div class="line">##      10     chr8 [ 18248755,  18258723]      + |          10</div><div class="line">##     100    chr20 [ 43248163,  43280376]      - |         100</div><div class="line">##    1000    chr18 [ 25530930,  25757445]      - |        1000</div><div class="line">##   10000     chr1 [243651535, 244006886]      - |       10000</div><div class="line">##     ...      ...                    ...    ... .         ...</div><div class="line">##    9991     chr9 [114979995, 115095944]      - |        9991</div><div class="line">##    9992    chr21 [ 35736323,  35743440]      + |        9992</div><div class="line">##    9993    chr22 [ 19023795,  19109967]      - |        9993</div><div class="line">##    9994     chr6 [ 90539619,  90584155]      + |        9994</div><div class="line">##    9997    chr22 [ 50961997,  50964905]      - |        9997</div><div class="line">##   -------</div><div class="line">##   seqinfo: 93 sequences (1 circular) from hg19 genome</div></pre></td></tr></table></figure>
<p>我们也可以使用合适的标识符进行信息过滤。现在我们提取两个不同基因的外显子，这些外显子由其Entrez基因ID标明，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div></pre></td><td class="code"><pre><div class="line">exons(txdb, columns=c(<span class="string">"EXONID"</span>, <span class="string">"TXNAME"</span>, <span class="string">"GENEID"</span>),</div><div class="line">                  filter=list(gene_id=c(<span class="number">100</span>, <span class="number">101</span>)))</div><div class="line"><span class="comment">## GRanges object with 39 ranges and 3 metadata columns:</span></div><div class="line"><span class="comment">##        seqnames                 ranges strand |    EXONID</span></div><div class="line"><span class="comment">##           &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span></div><div class="line"><span class="comment">##    [1]    chr10 [135075920, 135076737]      - |    144421</span></div><div class="line"><span class="comment">##    [2]    chr10 [135077192, 135077269]      - |    144422</span></div><div class="line"><span class="comment">##    [3]    chr10 [135080856, 135080921]      - |    144423</span></div><div class="line"><span class="comment">##    [4]    chr10 [135081433, 135081570]      - |    144424</span></div><div class="line"><span class="comment">##    [5]    chr10 [135081433, 135081622]      - |    144425</span></div><div class="line"><span class="comment">##    ...      ...                    ...    ... .       ...</span></div><div class="line"><span class="comment">##   [35]    chr20   [43254210, 43254325]      - |    256371</span></div><div class="line"><span class="comment">##   [36]    chr20   [43255097, 43255240]      - |    256372</span></div><div class="line"><span class="comment">##   [37]    chr20   [43257688, 43257810]      - |    256373</span></div><div class="line"><span class="comment">##   [38]    chr20   [43264868, 43264929]      - |    256374</span></div><div class="line"><span class="comment">##   [39]    chr20   [43280216, 43280376]      - |    256375</span></div><div class="line"><span class="comment">##                                  TXNAME          GENEID</span></div><div class="line"><span class="comment">##                         &lt;CharacterList&gt; &lt;CharacterList&gt;</span></div><div class="line"><span class="comment">##    [1] uc009ybi.3,uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [2]            uc009ybi.3,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [3] uc009ybi.3,uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [4]                       uc009ybi.3             101</span></div><div class="line"><span class="comment">##    [5]            uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    ...                              ...             ...</span></div><div class="line"><span class="comment">##   [35]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [36]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [37]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [38]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [39]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 93 sequences (1 circular) from hg19 genome</span></div></pre></td></tr></table></figure>
<h3 id="ensembl注释">ENSEMBL注释</h3>
<p><a href="http://www.ensembl.org/index.html" target="_blank" rel="external">Ensembl home主页</a>上写道：Ensembl创建，整合和发布研究基因组的参考数据库和工具。该项目位于 <a href="https://www.ebi.ac.uk/" target="_blank" rel="external">欧洲分子生物学实验室</a>，该实验室的数据库支持其注释资源可以与Bioconductor兼容。</p>
<p><code>ensembldb</code> 包含有一个简要说明，其内容如下所示：</p>
<p>ensembldb包提供了一些函数，这些函数用于创建和使用以转录本为中心的注释数据库/包。使用注释数据库的Perl API可以从Ensembl 1中直接获取这些数据。<code>TxDb</code> 包的功能和数据类似于<code>GenomicFeatures</code>包，另外，除了从数据库检索所有的基因/转录本模型和注释外，ensembldb包还提供了一个过滤框架，用于检索特定条目的注释，例如位于染色体区域上的某编码基因或某LincRNA转录模式的特定条目。从1.7版本开始，由ensembldb创建的EnsDb数据库还包含蛋白质注释数据库（参考<a href="http://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html#org35014ed" target="_blank" rel="external">第11节</a>：数据库而已和可用属性/列的概述）。有关蛋白质注释的信息请参考蛋白质的vignette，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ensembldb)</div><div class="line"><span class="keyword">library</span>(EnsDb.Hsapiens.v75)</div><div class="line">names(listTables(EnsDb.Hsapiens.v75))</div><div class="line"><span class="comment">##  [1] "gene"           "tx"             "tx2exon"        "exon"          </span></div><div class="line"><span class="comment">##  [5] "chromosome"     "protein"        "uniprot"        "protein_domain"</span></div><div class="line"><span class="comment">##  [9] "entrezgene"     "metadata"</span></div></pre></td></tr></table></figure>
<p>举例说明如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div></pre></td><td class="code"><pre><div class="line">edb = EnsDb.Hsapiens.v75  <span class="comment"># abbreviate</span></div><div class="line">txs &lt;- transcripts(edb, filter = GenenameFilter(<span class="string">"ZBTB16"</span>),</div><div class="line">                   columns = c(<span class="string">"protein_id"</span>, <span class="string">"uniprot_id"</span>, <span class="string">"tx_biotype"</span>))</div><div class="line">txs</div><div class="line"><span class="comment">## GRanges object with 20 ranges and 5 metadata columns:</span></div><div class="line"><span class="comment">##                   seqnames                 ranges strand |      protein_id</span></div><div class="line"><span class="comment">##                      &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##               ...      ...                    ...    ... .             ...</span></div><div class="line"><span class="comment">##   ENST00000392996       11 [113931229, 114121374]      + | ENSP00000376721</span></div><div class="line"><span class="comment">##   ENST00000539918       11 [113935134, 114118066]      + | ENSP00000445047</span></div><div class="line"><span class="comment">##   ENST00000545851       11 [114051488, 114118018]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ENST00000535379       11 [114107929, 114121279]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ENST00000535509       11 [114117512, 114121198]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##                     uniprot_id              tx_biotype           tx_id</span></div><div class="line"><span class="comment">##                    &lt;character&gt;             &lt;character&gt;     &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953  ZBT16_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL7_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL6_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL5_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 F5H6C3_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##               ...          ...                     ...             ...</span></div><div class="line"><span class="comment">##   ENST00000392996 F5H5Y7_HUMAN          protein_coding ENST00000392996</span></div><div class="line"><span class="comment">##   ENST00000539918         &lt;NA&gt; nonsense_mediated_decay ENST00000539918</span></div><div class="line"><span class="comment">##   ENST00000545851         &lt;NA&gt;    processed_transcript ENST00000545851</span></div><div class="line"><span class="comment">##   ENST00000535379         &lt;NA&gt;    processed_transcript ENST00000535379</span></div><div class="line"><span class="comment">##   ENST00000535509         &lt;NA&gt;         retained_intron ENST00000535509</span></div><div class="line"><span class="comment">##                     gene_name</span></div><div class="line"><span class="comment">##                   &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##               ...         ...</span></div><div class="line"><span class="comment">##   ENST00000392996      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000539918      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000545851      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000535379      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000535509      ZBTB16</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from GRCh37 genome</span></div></pre></td></tr></table></figure>
<h2 id="你的数据将会成他人的注释导入导出">你的数据将会成他人的注释：导入/导出</h2>
<p>ENCODE项目很地说明了今天的实验是明天的注释。你应该以同样的方式考虑自己的实验（当然，要使实验成为可靠且持久的注释，它必须解决有关基因组结构或功能的重要问题，并且必须使用适当的，能正确执行的实验流程。需要注意，ENCODE能够非常明确地将实验流程链接到数据）。</p>
<p>例如，我们来看一个雌激素受体结合数据，它是由ENCODE发布的一个narrowPeak 数据。它的碱基是用ascii文本表示的，因此可以很容易地导入为一组文本数据。如果记录的字段有一定的规律性，则可以将文件作为表格导入。</p>
<p>但是，我们不仅是想导入数据，还想将导入的数据作为可计算的对象。我们认识到arrowePeak和bedGraph格式之间的联系后，我们就可以立即将其导入GRanges中。</p>
<p>为了说明这一点，我们在ERBS包中找到narrowPeak原始数据文件的路径，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">f1 = dir(system.file(<span class="string">"extdata"</span>,package=<span class="string">"ERBS"</span>), full=<span class="literal">TRUE</span>)[<span class="number">1</span>]</div><div class="line">readLines(f1, <span class="number">4</span>) <span class="comment"># look at a few lines</span></div><div class="line"><span class="comment">## [1] "chrX\t1509354\t1512462\t5\t0\t.\t157.92\t310\t32.000000\t1991"    </span></div><div class="line"><span class="comment">## [2] "chrX\t26801421\t26802448\t6\t0\t.\t147.38\t310\t32.000000\t387"   </span></div><div class="line"><span class="comment">## [3] "chr19\t11694101\t11695359\t1\t0\t.\t99.71\t311.66\t32.000000\t861"</span></div><div class="line"><span class="comment">## [4] "chr19\t4076892\t4079276\t4\t0\t.\t84.74\t310\t32.000000\t1508"</span></div></pre></td></tr></table></figure>
<p>使用<code>import</code>命令非常简单，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rtracklayer)</div><div class="line">imp = import(f1, format=<span class="string">"bedGraph"</span>)</div><div class="line">imp</div><div class="line"><span class="comment">## GRanges object with 1873 ranges and 7 metadata columns:</span></div><div class="line"><span class="comment">##          seqnames               ranges strand |     score       NA.</span></div><div class="line"><span class="comment">##             &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##      [1]     chrX [ 1509355,  1512462]      * |         5         0</span></div><div class="line"><span class="comment">##      [2]     chrX [26801422, 26802448]      * |         6         0</span></div><div class="line"><span class="comment">##      [3]    chr19 [11694102, 11695359]      * |         1         0</span></div><div class="line"><span class="comment">##      [4]    chr19 [ 4076893,  4079276]      * |         4         0</span></div><div class="line"><span class="comment">##      [5]     chr3 [53288568, 53290767]      * |         9         0</span></div><div class="line"><span class="comment">##      ...      ...                  ...    ... .       ...       ...</span></div><div class="line"><span class="comment">##   [1869]    chr19 [11201120, 11203985]      * |      8701         0</span></div><div class="line"><span class="comment">##   [1870]    chr19 [ 2234920,  2237370]      * |       990         0</span></div><div class="line"><span class="comment">##   [1871]     chr1 [94311336, 94313543]      * |      4035         0</span></div><div class="line"><span class="comment">##   [1872]    chr19 [45690614, 45691210]      * |     10688         0</span></div><div class="line"><span class="comment">##   [1873]    chr19 [ 6110100,  6111252]      * |      2274         0</span></div><div class="line"><span class="comment">##               NA.1      NA.2      NA.3      NA.4      NA.5</span></div><div class="line"><span class="comment">##          &lt;logical&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##      [1]      &lt;NA&gt;    157.92       310        32      1991</span></div><div class="line"><span class="comment">##      [2]      &lt;NA&gt;    147.38       310        32       387</span></div><div class="line"><span class="comment">##      [3]      &lt;NA&gt;     99.71    311.66        32       861</span></div><div class="line"><span class="comment">##      [4]      &lt;NA&gt;     84.74       310        32      1508</span></div><div class="line"><span class="comment">##      [5]      &lt;NA&gt;      78.2   299.505        32      1772</span></div><div class="line"><span class="comment">##      ...       ...       ...       ...       ...       ...</span></div><div class="line"><span class="comment">##   [1869]      &lt;NA&gt;      8.65     7.281   0.26576      2496</span></div><div class="line"><span class="comment">##   [1870]      &lt;NA&gt;      8.65    26.258  1.995679      1478</span></div><div class="line"><span class="comment">##   [1871]      &lt;NA&gt;      8.65    12.511   1.47237      1848</span></div><div class="line"><span class="comment">##   [1872]      &lt;NA&gt;      8.65     6.205         0       298</span></div><div class="line"><span class="comment">##   [1873]      &lt;NA&gt;      8.65    17.356  2.013228       496</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></div><div class="line">genome(imp)  <span class="comment"># genome identifier tag not set, but you should set it</span></div><div class="line"><span class="comment">##  chrX chr19  chr3 chr17  chr8 chr11 chr16  chr1  chr2  chr6  chr9  chr7 </span></div><div class="line"><span class="comment">##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA </span></div><div class="line"><span class="comment">##  chr5 chr12 chr20 chr21 chr22 chr18 chr10 chr14 chr15  chr4 chr13 </span></div><div class="line"><span class="comment">##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA</span></div></pre></td></tr></table></figure>
<p>我们可以通过一次获取GRanges。元数据列中还有一些其他字段用于指定名称，但是如果我们只对范围感兴趣，除了添加基因组元数据以防止与不兼容的坐标中记录的数据非法组合外，我们就完成了这个任务（这一段不太理解，原文如下）：</p>
<blockquote>
<p>We obtain a GRanges in one stroke. There are some additional fields in the metadata columns whose names should be specified, but if we are interested only in the ranges, we are done, with the exception of adding the genome metadata to protect against illegitimate combination with data recorded in an incompatible coordinate system.</p>
</blockquote>
<p>为了与其他得养家或系统进行交流，我们有两个主要选择。我们可以将GRanges保存为<code>RData</code>对象，轻松地传递给另外一个R用户使用。或者，我们歌词采用其他标准格式进行导出。例如，如果我们仅对间隔地址和绑定的得分感兴趣，则仅保存为“bed”格式就足够了，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">export(imp, <span class="string">"demoex.bed"</span>)  <span class="comment"># implicit format choice</span></div><div class="line">cat(readLines(<span class="string">"demoex.bed"</span>, n=<span class="number">5</span>), sep=<span class="string">"\n"</span>)</div><div class="line"><span class="comment">## chrX	1509354	1512462	.	5	.</span></div><div class="line"><span class="comment">## chrX	26801421	26802448	.	6	.</span></div><div class="line"><span class="comment">## chr19	11694101	11695359	.	1	.</span></div><div class="line"><span class="comment">## chr19	4076892	4079276	.	4	.</span></div><div class="line"><span class="comment">## chr3	53288567	53290767	.	9	.</span></div></pre></td></tr></table></figure>
<p>我们已经进行了导入，建模和导入实验数据之间的“往返”，该实验数据可以与其他数据集成在一起，从而增进生物学的理解。</p>
<p>我们需要注意的是，注释在某种程度上是永久正确的，它与在知识边界上的研究进展乏味地隔离开来。我们已经看到了，甚至人类染色体的参考序列也受到了修订。在使用ERBS包时，我们将未知的实验结果视为定义ER结合位点从而进入潜在的生物学解释。不确定性，峰鉴定的可变质量，尚未得到明确估计，但应该是这个样子。</p>
<p>Bioconductor已经尽力致力于这种情况的多个方面。我们维护软件先前版本和注释的存档，以便可以检查或修改过去的工作。我们每年会两次更新中疏注释资源，以确保正在进行的工作以及获得新知识的稳定性。而且，我们已经简化了导入和创建实验数据和注释数据的表示形式。</p>
<h2 id="annotationhub">AnnotationHub</h2>
<p><code>AnnotationHub</code>包用于获取GRanges或其它的合适设计的容器，用于机构设计的容器，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(AnnotationHub)</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Attaching package: 'AnnotationHub'</span></div><div class="line"><span class="comment">## The following object is masked from 'package:Biobase':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     cache</span></div><div class="line">ah = AnnotationHub()</div><div class="line"><span class="comment">## snapshotDate(): 2017-10-27</span></div><div class="line">ah</div><div class="line"><span class="comment">## AnnotationHub with 42282 records</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # $$dataprovider: BroadInstitute, Ensembl, UCSC, ftp://ftp.ncbi.nlm.nih....</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos ta...</span></div><div class="line"><span class="comment">## # $$rdataclass: GRanges, BigWigFile, FaFile, TwoBitFile, Rle, ChainFile,...</span></div><div class="line"><span class="comment">## # additional mcols(): taxonomyid, genome, description,</span></div><div class="line"><span class="comment">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,</span></div><div class="line"><span class="comment">## #   tags, rdatapath, sourceurl, sourcetype </span></div><div class="line"><span class="comment">## # retrieve records with, e.g., 'object[["AH2"]]' </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##             title                                                         </span></div><div class="line"><span class="comment">##   AH2     | Ailuropoda_melanoleuca.ailMel1.69.dna.toplevel.fa             </span></div><div class="line"><span class="comment">##   AH3     | Ailuropoda_melanoleuca.ailMel1.69.dna_rm.toplevel.fa          </span></div><div class="line"><span class="comment">##   AH4     | Ailuropoda_melanoleuca.ailMel1.69.dna_sm.toplevel.fa          </span></div><div class="line"><span class="comment">##   AH5     | Ailuropoda_melanoleuca.ailMel1.69.ncrna.fa                    </span></div><div class="line"><span class="comment">##   AH6     | Ailuropoda_melanoleuca.ailMel1.69.pep.all.fa                  </span></div><div class="line"><span class="comment">##   ...       ...                                                           </span></div><div class="line"><span class="comment">##   AH58988 | org.Flavobacterium_piscicida.eg.sqlite                        </span></div><div class="line"><span class="comment">##   AH58989 | org.Bacteroides_fragilis_YCH46.eg.sqlite                      </span></div><div class="line"><span class="comment">##   AH58990 | org.Pseudomonas_mendocina_ymp.eg.sqlite                       </span></div><div class="line"><span class="comment">##   AH58991 | org.Salmonella_enterica_subsp._enterica_serovar_Typhimurium...</span></div><div class="line"><span class="comment">##   AH58992 | org.Acinetobacter_baumannii.eg.sqlite</span></div></pre></td></tr></table></figure>
<p>我们可以通过AnnotationHub获得许多与HepG2细胞系相关的实验数据对象，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">query(ah, <span class="string">"HepG2"</span>)</div><div class="line"><span class="comment">## AnnotationHub with 440 records</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # $$dataprovider: UCSC, BroadInstitute, Pazar</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens, NA</span></div><div class="line"><span class="comment">## # $$rdataclass: GRanges, BigWigFile</span></div><div class="line"><span class="comment">## # additional mcols(): taxonomyid, genome, description,</span></div><div class="line"><span class="comment">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,</span></div><div class="line"><span class="comment">## #   tags, rdatapath, sourceurl, sourcetype </span></div><div class="line"><span class="comment">## # retrieve records with, e.g., 'object[["AH22246"]]' </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##             title                                                         </span></div><div class="line"><span class="comment">##   AH22246 | pazar_CEBPA_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22249 | pazar_CTCF_HEPG2_Schmidt_20120522.csv                         </span></div><div class="line"><span class="comment">##   AH22273 | pazar_HNF4A_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22309 | pazar_STAG1_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22348 | wgEncodeAffyRnaChipFiltTransfragsHepg2CytosolLongnonpolya.b...</span></div><div class="line"><span class="comment">##   ...       ...                                                           </span></div><div class="line"><span class="comment">##   AH41564 | E118-H4K5ac.imputed.pval.signal.bigwig                        </span></div><div class="line"><span class="comment">##   AH41691 | E118-H4K8ac.imputed.pval.signal.bigwig                        </span></div><div class="line"><span class="comment">##   AH41818 | E118-H4K91ac.imputed.pval.signal.bigwig                       </span></div><div class="line"><span class="comment">##   AH46971 | E118_15_coreMarks_mnemonics.bed.gz                            </span></div><div class="line"><span class="comment">##   AH49484 | E118_RRBS_FractionalMethylation.bigwig</span></div></pre></td></tr></table></figure>
<p><code>query</code> 方法可以使用过滤字符串的向量。要限制对寻址组蛋白H4K5的注释资源的响应，只需要添加该标签，如下所示(To limit response to annotation resources addressing the histone H4K5, simply add that tag)：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">query(ah, c(<span class="string">"HepG2"</span>, <span class="string">"H4K5"</span>))</div><div class="line"><span class="comment">## AnnotationHub with 1 record</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # names(): AH41564</span></div><div class="line"><span class="comment">## # $$dataprovider: BroadInstitute</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens</span></div><div class="line"><span class="comment">## # $$rdataclass: BigWigFile</span></div><div class="line"><span class="comment">## # $$rdatadateadded: 2015-05-08</span></div><div class="line"><span class="comment">## # $$title: E118-H4K5ac.imputed.pval.signal.bigwig</span></div><div class="line"><span class="comment">## # $$description: Bigwig File containing -log10(p-value) signal tracks fr...</span></div><div class="line"><span class="comment">## # $$taxonomyid: 9606</span></div><div class="line"><span class="comment">## # $$genome: hg19</span></div><div class="line"><span class="comment">## # $$sourcetype: BigWig</span></div><div class="line"><span class="comment">## # $$sourceurl: http://egg2.wustl.edu/roadmap/data/byFileType/signal/cons...</span></div><div class="line"><span class="comment">## # $$sourcesize: 226630905</span></div><div class="line"><span class="comment">## # $$tags: c("EpigenomeRoadMap", "signal", "consolidatedImputed",</span></div><div class="line"><span class="comment">## #   "H4K5ac", "E118", "ENCODE2012", "LIV.HEPG2.CNCR", "HepG2</span></div><div class="line"><span class="comment">## #   Hepatocellular Carcinoma Cell Line") </span></div><div class="line"><span class="comment">## # retrieve record with 'object[["AH41564"]]'</span></div></pre></td></tr></table></figure>
<h2 id="the-orgdb基因注释图">The OrgDb基因注释图</h2>
<p>那些命名为<code>org.*.ge.db</code> 的包含在基因水平上链接到位置，蛋白产物标识符，KEGG途径和GO term，PMIDs以及其它注释资源的标识符的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(org.Hs.eg.db)</div><div class="line">keytypes(org.Hs.eg.db) <span class="comment"># columns() gives same answer</span></div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" </span></div><div class="line"><span class="comment">##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    </span></div><div class="line"><span class="comment">##  [9] "EVIDENCEALL"  "GENENAME"     "GO"           "GOALL"       </span></div><div class="line"><span class="comment">## [13] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    </span></div><div class="line"><span class="comment">## [17] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        </span></div><div class="line"><span class="comment">## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      </span></div><div class="line"><span class="comment">## [25] "UNIGENE"      "UNIPROT"</span></div><div class="line">head(select(org.Hs.eg.db, keys=<span class="string">"ORMDL3"</span>, keytype=<span class="string">"SYMBOL"</span>, </div><div class="line">   columns=<span class="string">"PMID"</span>))</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">##   SYMBOL     PMID</span></div><div class="line"><span class="comment">## 1 ORMDL3 11042152</span></div><div class="line"><span class="comment">## 2 ORMDL3 12093374</span></div><div class="line"><span class="comment">## 3 ORMDL3 12477932</span></div><div class="line"><span class="comment">## 4 ORMDL3 14702039</span></div><div class="line"><span class="comment">## 5 ORMDL3 15489334</span></div><div class="line"><span class="comment">## 6 ORMDL3 16169070</span></div></pre></td></tr></table></figure>
<h2 id="基因集和通路资源">基因集和通路资源</h2>
<h3 id="基因本体论">基因本体论</h3>
<p><a href="http://www.geneontology.org" target="_blank" rel="external">Gene Ontology</a> (GO)是一种广泛使用的结构化词汇，它组织了基因和基因产物在以下方面的内容：</p>
<ul>
<li>生物过程</li>
<li>分子功能</li>
<li>细胞组分。</li>
</ul>
<p>这套词汇本身旨在与所有生物有关。它采用有向无环图的形式，其中term作为节点，使用<code>is-a</code>和<code>part-of</code>关系作构成了大多数链接。</p>
<p>将生物体特定基因链接到基因本体中的术语的注释与词汇表本身是分开的，并且涉及不同类型的证据。这些记录都在Bioconductor的注释包中。</p>
<p>我们可以使用<code>GO.db</code>包来快速地访问GO词汇，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GO.db)</div><div class="line">GO.db <span class="comment"># metadata</span></div><div class="line"><span class="comment">## GODb object:</span></div><div class="line"><span class="comment">## | GOSOURCENAME: Gene Ontology</span></div><div class="line"><span class="comment">## | GOSOURCEURL: ftp://ftp.geneontology.org/pub/go/godatabase/archive/latest-lite/</span></div><div class="line"><span class="comment">## | GOSOURCEDATE: 2017-Nov01</span></div><div class="line"><span class="comment">## | Db type: GODb</span></div><div class="line"><span class="comment">## | package: AnnotationDbi</span></div><div class="line"><span class="comment">## | DBSCHEMA: GO_DB</span></div><div class="line"><span class="comment">## | GOEGSOURCEDATE: 2017-Nov6</span></div><div class="line"><span class="comment">## | GOEGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | DBSCHEMAVERSION: 2.1</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Please see: help('select') for usage information</span></div></pre></td></tr></table></figure>
<p>使用<code>AnnotationDbi</code>包中的<code>keys</code>，<code>columns</code>和<code>select</code>函数也很容易在地id与不同terms之间进行映射，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">k5 = keys(GO.db)[<span class="number">1</span>:<span class="number">5</span>]</div><div class="line">cgo = columns(GO.db)</div><div class="line">select(GO.db, keys=k5, columns=cgo[<span class="number">1</span>:<span class="number">3</span>])</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line"><span class="comment">##         GOID</span></div><div class="line"><span class="comment">## 1 GO:0000001</span></div><div class="line"><span class="comment">## 2 GO:0000002</span></div><div class="line"><span class="comment">## 3 GO:0000003</span></div><div class="line"><span class="comment">## 4 GO:0000006</span></div><div class="line"><span class="comment">## 5 GO:0000007</span></div><div class="line"><span class="comment">##                                                                                                                                                                                                                                                                                      DEFINITION</span></div><div class="line"><span class="comment">## 1                                                                                                       The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.</span></div><div class="line"><span class="comment">## 2                                                                                                                                             The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.</span></div><div class="line"><span class="comment">## 3                                                                                                                                                                  The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.</span></div><div class="line"><span class="comment">## 4                                      Enables the transfer of zinc ions (Zn2+) from one side of a membrane to the other, probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations.</span></div><div class="line"><span class="comment">## 5 Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.</span></div><div class="line"><span class="comment">##   ONTOLOGY</span></div><div class="line"><span class="comment">## 1       BP</span></div><div class="line"><span class="comment">## 2       BP</span></div><div class="line"><span class="comment">## 3       BP</span></div><div class="line"><span class="comment">## 4       MF</span></div><div class="line"><span class="comment">## 5       MF</span></div></pre></td></tr></table></figure>
<p>词汇表的图形结构被编码在<code>SQLite</code>数据库的表中。我们可以使用<code>RSQLite</code>接口对此进行查询，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">con = GO_dbconn()</div><div class="line">dbListTables(con)</div><div class="line"><span class="comment">##  [1] "go_bp_offspring" "go_bp_parents"   "go_cc_offspring"</span></div><div class="line"><span class="comment">##  [4] "go_cc_parents"   "go_mf_offspring" "go_mf_parents"  </span></div><div class="line"><span class="comment">##  [7] "go_obsolete"     "go_ontology"     "go_synonym"     </span></div><div class="line"><span class="comment">## [10] "go_term"         "map_counts"      "map_metadata"   </span></div><div class="line"><span class="comment">## [13] "metadata"        "sqlite_stat1"</span></div></pre></td></tr></table></figure>
<p>以下查询提示了一些内部标识符：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term limit 5"</span>)</div><div class="line"><span class="comment">##   _id      go_id                                        term</span></div><div class="line"><span class="comment">## 1  30 GO:0000001                   mitochondrion inheritance</span></div><div class="line"><span class="comment">## 2  32 GO:0000002            mitochondrial genome maintenance</span></div><div class="line"><span class="comment">## 3  33 GO:0000003                                reproduction</span></div><div class="line"><span class="comment">## 4  37 GO:0042254                         ribosome biogenesis</span></div><div class="line"><span class="comment">## 5  38 GO:0044183 protein binding involved in protein folding</span></div></pre></td></tr></table></figure>
<p>我们可以将 <code>mitochondrion inheritance</code> term追溯到父项和祖父母项，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">dbGetQuery(con, <span class="string">"select * from go_bp_parents where _id=30"</span>)</div><div class="line"><span class="comment">##   _id _parent_id relationship_type</span></div><div class="line"><span class="comment">## 1  30      26537              is_a</span></div><div class="line"><span class="comment">## 2  30      26540              is_a</span></div><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term where _id=26616"</span>)</div><div class="line"><span class="comment">##     _id      go_id</span></div><div class="line"><span class="comment">## 1 26616 GO:0048387</span></div><div class="line"><span class="comment">##                                                              term</span></div><div class="line"><span class="comment">## 1 negative regulation of retinoic acid receptor signaling pathway</span></div><div class="line">dbGetQuery(con, <span class="string">"select * from go_bp_parents where _id=26616"</span>)</div><div class="line"><span class="comment">##     _id _parent_id    relationship_type</span></div><div class="line"><span class="comment">## 1 26616       8389                 is_a</span></div><div class="line"><span class="comment">## 2 26616      26614                 is_a</span></div><div class="line"><span class="comment">## 3 26616      26613 negatively_regulates</span></div><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term where _id=5932"</span>)</div><div class="line"><span class="comment">##    _id      go_id                    term</span></div><div class="line"><span class="comment">## 1 5932 GO:0019237 centromeric DNA binding</span></div></pre></td></tr></table></figure>
<p>将 “mitochondrion inheritance” 视为过程“mitochondrion distribution”和 “organelle inheritance”在概念上的精练是有意义的，这两个term在数据库中被为父项。</p>
<p>可以使用 <code>GO_dbschema()</code>来查看整个数据库模式。</p>
<h3 id="kegg">KEGG</h3>
<p>自Bioconductor诞生以来，KEGG的注释就能在Bioconductor中人使用了，但KEGG的数据库使用权限已经进行了更改。当我们使用<code>KEGG.db</code>加载后会出现以下信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(KEGG.db)</div><div class="line">KEGG.db contains mappings based on older data because the original</div><div class="line">  resource was removed from the the public domain before the most</div><div class="line">  recent update was produced. This package should now be considered</div><div class="line">  deprecated and future versions of Bioconductor may not have it</div><div class="line">  available.  Users who want more current data are encouraged to look</div><div class="line">  at the KEGGREST or reactome.db packages</div></pre></td></tr></table></figure>
<p>因此我们可以关注KEGGREST这个包，它需要联网。这是一个非常有用的，基于Entrez标识符的工具。现在我们查询一下BRCA2的信息（它的EntrezID为675），如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(KEGGREST)</div><div class="line">brca2K = keggGet(&quot;hsa:675&quot;)</div><div class="line">names(brca2K[[1]])</div><div class="line">##  [1] &quot;ENTRY&quot;      &quot;NAME&quot;       &quot;DEFINITION&quot; &quot;ORTHOLOGY&quot;  &quot;ORGANISM&quot;  </div><div class="line">##  [6] &quot;PATHWAY&quot;    &quot;DISEASE&quot;    &quot;BRITE&quot;      &quot;POSITION&quot;   &quot;MOTIF&quot;     </div><div class="line">## [11] &quot;DBLINKS&quot;    &quot;STRUCTURE&quot;  &quot;AASEQ&quot;      &quot;NTSEQ&quot;</div></pre></td></tr></table></figure>
<p>我们也可以通过<code>keggGet</code>函数来获取构成通路模式的基因列表，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">brpat = keggGet(<span class="string">"path:hsa05212"</span>)</div><div class="line">names(brpat[[<span class="number">1</span>]])</div><div class="line"><span class="comment">##  [1] "ENTRY"       "NAME"        "DESCRIPTION" "CLASS"       "PATHWAY_MAP"</span></div><div class="line"><span class="comment">##  [6] "DISEASE"     "DRUG"        "ORGANISM"    "GENE"        "COMPOUND"   </span></div><div class="line"><span class="comment">## [11] "KO_PATHWAY"  "REFERENCE"</span></div><div class="line">brpat[[<span class="number">1</span>]]$GENE[seq(<span class="number">1</span>,<span class="number">132</span>,<span class="number">2</span>)] <span class="comment"># entrez gene ids</span></div><div class="line"><span class="comment">##  [1] "3845"  "5290"  "5293"  "5291"  "5295"  "5296"  "8503"  "9459" </span></div><div class="line"><span class="comment">##  [9] "5879"  "5880"  "5881"  "4790"  "5970"  "207"   "208"   "10000"</span></div><div class="line"><span class="comment">## [17] "1147"  "3551"  "8517"  "572"   "598"   "842"   "369"   "673"  </span></div><div class="line"><span class="comment">## [25] "5894"  "5604"  "5594"  "5595"  "5599"  "5602"  "5601"  "5900" </span></div><div class="line"><span class="comment">## [33] "5898"  "5899"  "10928" "998"   "7039"  "1950"  "1956"  "2064" </span></div><div class="line"><span class="comment">## [41] "2475"  "6198"  "6199"  "3716"  "6774"  "6772"  "7422"  "1029" </span></div><div class="line"><span class="comment">## [49] "1019"  "1021"  "595"   "5925"  "1869"  "1870"  "1871"  "7157" </span></div><div class="line"><span class="comment">## [57] "1026"  "1647"  "4616"  "10912" "581"   "578"   "1643"  "51426"</span></div><div class="line"><span class="comment">## [65] "7040"  "7042"</span></div></pre></td></tr></table></figure>
<p>KEGGREST还有许多值得研究的地方，例如还可以查询BRCA2（人类）关于胰腺癌途径的静态图像，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(png)</div><div class="line"><span class="keyword">library</span>(grid)</div><div class="line">brpng = keggGet(<span class="string">"hsa05212"</span>, <span class="string">"image"</span>)</div><div class="line">grid.raster(brpng)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_annoOverview-getp-1.png" alt="plot of chunk getp">
<p class="caption">plot of chunk getp</p>
</div>
<h3 id="其它本体">其它本体</h3>
<p><code>rols</code>包含有与EMBL-EBI连接的接口 <a href="https://www.ebi.ac.uk/ols/index" target="_blank" rel="external">Ontology Lookup Service</a>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rols)</div><div class="line">oo = Ontologies()</div><div class="line">oo</div><div class="line"><span class="comment">## Object of class 'Ontologies' with 198 entries</span></div><div class="line"><span class="comment">##    GENEPIO, MP ... SEPIO, SIBO</span></div><div class="line">oo[[<span class="number">1</span>]]</div><div class="line"><span class="comment">## Ontology: Genomic Epidemiology Ontology (genepio)  </span></div><div class="line"><span class="comment">##   The Genomic Epidemiology Ontology (GenEpiO) covers vocabulary</span></div><div class="line"><span class="comment">##   necessary to identify, document and research foodborne pathogens</span></div><div class="line"><span class="comment">##   and associated outbreaks.</span></div><div class="line"><span class="comment">##    Loaded: 2017-04-10 Updated: 2017-10-20 Version: 2017-04-09 </span></div><div class="line"><span class="comment">##    4351 terms  137 properties  38 individuals</span></div></pre></td></tr></table></figure>
<p>为了控制查询检索中涉及的网络流量，搜索分为几个阶段，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">glis = OlsSearch(<span class="string">"glioblastoma"</span>)</div><div class="line">glis</div><div class="line"><span class="comment">## Object of class 'OlsSearch':</span></div><div class="line"><span class="comment">##   query: glioblastoma </span></div><div class="line"><span class="comment">##   requested: 20 (out of 502)</span></div><div class="line"><span class="comment">##   response(s): 0</span></div><div class="line">res = olsSearch(glis)</div><div class="line">dim(res)</div><div class="line"><span class="comment">## NULL</span></div><div class="line">resdf = as(res, <span class="string">"data.frame"</span>) <span class="comment"># get content</span></div><div class="line">resdf[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]</div><div class="line"><span class="comment">##                                                     id</span></div><div class="line"><span class="comment">## 1 ncit:class:http://purl.obolibrary.org/obo/NCIT_C3058</span></div><div class="line"><span class="comment">## 2     omit:http://purl.obolibrary.org/obo/OMIT_0007102</span></div><div class="line"><span class="comment">## 3    ordo:class:http://www.orpha.net/ORDO/Orphanet_360</span></div><div class="line"><span class="comment">## 4   hp:class:http://purl.obolibrary.org/obo/HP_0100843</span></div><div class="line"><span class="comment">##                                           iri   short_form        label</span></div><div class="line"><span class="comment">## 1   http://purl.obolibrary.org/obo/NCIT_C3058   NCIT_C3058 Glioblastoma</span></div><div class="line"><span class="comment">## 2 http://purl.obolibrary.org/obo/OMIT_0007102 OMIT_0007102 Glioblastoma</span></div><div class="line"><span class="comment">## 3      http://www.orpha.net/ORDO/Orphanet_360 Orphanet_360 Glioblastoma</span></div><div class="line"><span class="comment">## 4   http://purl.obolibrary.org/obo/HP_0100843   HP_0100843 Glioblastoma</span></div><div class="line">resdf[<span class="number">1</span>,<span class="number">5</span>]  <span class="comment"># full description for one instance</span></div><div class="line"><span class="comment">## [[1]]</span></div><div class="line"><span class="comment">## [1] "The most malignant astrocytic tumor (WHO grade IV).  It is composed of poorly differentiated neoplastic astrocytes and it is characterized by the presence of cellular polymorphism, nuclear atypia, brisk mitotic activity, vascular thrombosis, microvascular proliferation and necrosis. It typically affects adults and is preferentially located in the cerebral hemispheres. It may develop from diffuse astrocytoma WHO grade II or anaplastic astrocytoma (secondary glioblastoma, IDH-mutant), but more frequently, it manifests after a short clinical history de novo, without evidence of a less malignant precursor lesion (primary glioblastoma, IDH- wildtype). (Adapted from WHO)"</span></div></pre></td></tr></table></figure>
<p><code>ontologyIndex</code> 包支持导入开放生物本体(OBO, Open Biological Ontologies)格式的数据，并含有用于查询和可视化本体系统高效的工具。</p>
<h3 id="通用基因集管理">通用基因集管理</h3>
<p><code>GSEABase</code> 包有一个用于管理基因集和集合的优秀工具。我们可以从<a href="http://software.broadinstitute.org/gsea/msigdb/search.jsp" target="_blank" rel="external">MSigDb</a>中导入胶质母细胞瘤相关的基因集来说明一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSEABase)</div><div class="line">glioG = getGmt(system.file(<span class="string">"gmt/glioSets.gmt"</span>, package=<span class="string">"ph525x"</span>))</div><div class="line"><span class="comment">## Warning in readLines(con, ...): incomplete final line found on '/</span></div><div class="line"><span class="comment">## Library/Frameworks/R.framework/Versions/3.4/Resources/library/ph525x/gmt/</span></div><div class="line"><span class="comment">## glioSets.gmt'</span></div><div class="line">glioG</div><div class="line"><span class="comment">## GeneSetCollection</span></div><div class="line"><span class="comment">##   names: BALDWIN_PRKCI_TARGETS_UP, BEIER_GLIOMA_STEM_CELL_DN, ..., ZHENG_GLIOBLASTOMA_PLASTICITY_UP (47 total)</span></div><div class="line"><span class="comment">##   unique identifiers: ADA, AQP9, ..., ZFP28 (3671 total)</span></div><div class="line"><span class="comment">##   types in collection:</span></div><div class="line"><span class="comment">##     geneIdType: NullIdentifier (1 total)</span></div><div class="line"><span class="comment">##     collectionType: NullCollection (1 total)</span></div><div class="line">head(geneIds(glioG[[<span class="number">1</span>]]))</div><div class="line"><span class="comment">## [1] "ADA"      "AQP9"     "ATP2B4"   "ATP6V1G1" "CBX6"     "CCDC165"</span></div></pre></td></tr></table></figure>
<h2 id="模式生物的统一自我描述方法">模式生物的统一，自我描述方法</h2>
<p><code>OrganismDb</code>包简化了对注释的访问。还可以针对<code>TxDb</code>和<code>org.[Nn].eg.db</code>进行直接查询，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">class(Homo.sapiens)</div><div class="line"><span class="comment">## [1] "OrganismDb"</span></div><div class="line"><span class="comment">## attr(,"package")</span></div><div class="line"><span class="comment">## [1] "OrganismDbi"</span></div><div class="line">Homo.sapiens</div><div class="line"><span class="comment">## OrganismDb Object:</span></div><div class="line"><span class="comment">## # Includes GODb Object:  GO.db </span></div><div class="line"><span class="comment">## # With data about:  Gene Ontology </span></div><div class="line"><span class="comment">## # Includes OrgDb Object:  org.Hs.eg.db </span></div><div class="line"><span class="comment">## # Gene data about:  Homo sapiens </span></div><div class="line"><span class="comment">## # Taxonomy Id:  9606 </span></div><div class="line"><span class="comment">## # Includes TxDb Object:  TxDb.Hsapiens.UCSC.hg19.knownGene </span></div><div class="line"><span class="comment">## # Transcriptome data about:  Homo sapiens </span></div><div class="line"><span class="comment">## # Based on genome:  hg19 </span></div><div class="line"><span class="comment">## # The OrgDb gene id ENTREZID is mapped to the TxDb gene id GENEID .</span></div><div class="line">tx = transcripts(Homo.sapiens)</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line">keytypes(Homo.sapiens)</div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "CDSID"        "CDSNAME"     </span></div><div class="line"><span class="comment">##  [5] "DEFINITION"   "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"</span></div><div class="line"><span class="comment">##  [9] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></div><div class="line"><span class="comment">## [13] "EXONID"       "EXONNAME"     "GENEID"       "GENENAME"    </span></div><div class="line"><span class="comment">## [17] "GO"           "GOALL"        "GOID"         "IPI"         </span></div><div class="line"><span class="comment">## [21] "MAP"          "OMIM"         "ONTOLOGY"     "ONTOLOGYALL" </span></div><div class="line"><span class="comment">## [25] "PATH"         "PFAM"         "PMID"         "PROSITE"     </span></div><div class="line"><span class="comment">## [29] "REFSEQ"       "SYMBOL"       "TERM"         "TXID"        </span></div><div class="line"><span class="comment">## [33] "TXNAME"       "UCSCKG"       "UNIGENE"      "UNIPROT"</span></div><div class="line">columns(Homo.sapiens)</div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "CDSCHROM"     "CDSEND"      </span></div><div class="line"><span class="comment">##  [5] "CDSID"        "CDSNAME"      "CDSSTART"     "CDSSTRAND"   </span></div><div class="line"><span class="comment">##  [9] "DEFINITION"   "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"</span></div><div class="line"><span class="comment">## [13] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></div><div class="line"><span class="comment">## [17] "EXONCHROM"    "EXONEND"      "EXONID"       "EXONNAME"    </span></div><div class="line"><span class="comment">## [21] "EXONRANK"     "EXONSTART"    "EXONSTRAND"   "GENEID"      </span></div><div class="line"><span class="comment">## [25] "GENENAME"     "GO"           "GOALL"        "GOID"        </span></div><div class="line"><span class="comment">## [29] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    </span></div><div class="line"><span class="comment">## [33] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        </span></div><div class="line"><span class="comment">## [37] "PROSITE"      "REFSEQ"       "SYMBOL"       "TERM"        </span></div><div class="line"><span class="comment">## [41] "TXCHROM"      "TXEND"        "TXID"         "TXNAME"      </span></div><div class="line"><span class="comment">## [45] "TXSTART"      "TXSTRAND"     "TXTYPE"       "UCSCKG"      </span></div><div class="line"><span class="comment">## [49] "UNIGENE"      "UNIPROT"</span></div></pre></td></tr></table></figure>
<h2 id="面向平台的注释">面向平台的注释</h2>
<p>通过在NCBI GEO的<a href="https://www.ncbi.nlm.nih.gov/geo/browse/?view=platforms&amp;sort=series&amp;display=20&amp;page=892" target="_blank" rel="external">GPL信息页面</a> 上对信息进行排序，我们就可以看到最常用的寡核苷阵列平台（数据库中有4760个系列）就是Affy Human Genome U133 plus 2.0 array (GPL 570)。我们可以使用<code>hgu133plus2.db</code>对这些数据进行注释，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(hgu133plus2.db)</div><div class="line"><span class="comment">## </span></div><div class="line">hgu133plus2.db</div><div class="line"><span class="comment">## ChipDb object:</span></div><div class="line"><span class="comment">## | DBSCHEMAVERSION: 2.1</span></div><div class="line"><span class="comment">## | Db type: ChipDb</span></div><div class="line"><span class="comment">## | Supporting package: AnnotationDbi</span></div><div class="line"><span class="comment">## | DBSCHEMA: HUMANCHIP_DB</span></div><div class="line"><span class="comment">## | ORGANISM: Homo sapiens</span></div><div class="line"><span class="comment">## | SPECIES: Human</span></div><div class="line"><span class="comment">## | MANUFACTURER: Affymetrix</span></div><div class="line"><span class="comment">## | CHIPNAME: Human Genome U133 Plus 2.0 Array</span></div><div class="line"><span class="comment">## | MANUFACTURERURL: http://www.affymetrix.com/support/technical/byproduct.affx?product=hg-u133-plus</span></div><div class="line"><span class="comment">## | EGSOURCEDATE: 2015-Sep27</span></div><div class="line"><span class="comment">## | EGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | CENTRALID: ENTREZID</span></div><div class="line"><span class="comment">## | TAXID: 9606</span></div><div class="line"><span class="comment">## | GOSOURCENAME: Gene Ontology</span></div><div class="line"><span class="comment">## | GOSOURCEURL: ftp://ftp.geneontology.org/pub/go/godatabase/archive/latest-lite/</span></div><div class="line"><span class="comment">## | GOSOURCEDATE: 20150919</span></div><div class="line"><span class="comment">## | GOEGSOURCEDATE: 2015-Sep27</span></div><div class="line"><span class="comment">## | GOEGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | KEGGSOURCENAME: KEGG GENOME</span></div><div class="line"><span class="comment">## | KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes</span></div><div class="line"><span class="comment">## | KEGGSOURCEDATE: 2011-Mar15</span></div><div class="line"><span class="comment">## | GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)</span></div><div class="line"><span class="comment">## | GPSOURCEURL: ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19</span></div><div class="line"><span class="comment">## | GPSOURCEDATE: 2010-Mar22</span></div><div class="line"><span class="comment">## | ENSOURCEDATE: 2015-Jul16</span></div><div class="line"><span class="comment">## | ENSOURCENAME: Ensembl</span></div><div class="line"><span class="comment">## | ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta</span></div><div class="line"><span class="comment">## | UPSOURCENAME: Uniprot</span></div><div class="line"><span class="comment">## | UPSOURCEURL: http://www.uniprot.org/</span></div><div class="line"><span class="comment">## | UPSOURCEDATE: Thu Oct  1 23:31:58 2015</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Please see: help('select') for usage information</span></div></pre></td></tr></table></figure>
<p>这个资源（以及ChipDb类的所有实例）的基本目的是在探针集(probeset)识别符和更高层次的基因组注释之间进行映射。</p>
<p>有关探针的详细信息（探针集的组成部分）已经由那些后缀为<code>probe</code>的文件包提供，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(hgu133plus2probe)</div><div class="line">head(hgu133plus2probe)</div><div class="line">##                    sequence    x   y Probe.Set.Name</div><div class="line">## 1 CACCCAGCTGGTCCTGTGGATGGGA  718 317      1007_s_at</div><div class="line">## 2 GCCCCACTGGACAACACTGATTCCT 1105 483      1007_s_at</div><div class="line">## 3 TGGACCCCACTGGCTGAGAATCTGG  584 901      1007_s_at</div><div class="line">## 4 AAATGTTTCCTTGTGCCTGCTCCTG  192 205      1007_s_at</div><div class="line">## 5 TCCTTGTGCCTGCTCCTGTACTTGT  844 979      1007_s_at</div><div class="line">## 6 TGCCTGCTCCTGTACTTGTCCTCAG  537 971      1007_s_at</div><div class="line">##   Probe.Interrogation.Position Target.Strandedness</div><div class="line">## 1                         3330           Antisense</div><div class="line">## 2                         3443           Antisense</div><div class="line">## 3                         3512           Antisense</div><div class="line">## 4                         3563           Antisense</div><div class="line">## 5                         3570           Antisense</div><div class="line">## 6                         3576           Antisense</div><div class="line">dim(hgu133plus2probe)</div><div class="line">## [1] 604258      6</div></pre></td></tr></table></figure>
<p>将探针集标识符映射到基因水平的信息可以提示一些有意思的歧视，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">select(hgu133plus2.db, keytype=<span class="string">"PROBEID"</span>, </div><div class="line">  columns=c(<span class="string">"SYMBOL"</span>, <span class="string">"GENENAME"</span>, <span class="string">"PATH"</span>, <span class="string">"MAP"</span>), keys=<span class="string">"1007_s_at"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">##     PROBEID  SYMBOL                                    GENENAME PATH</span></div><div class="line"><span class="comment">## 1 1007_s_at    DDR1 discoidin domain receptor tyrosine kinase 1 &lt;NA&gt;</span></div><div class="line"><span class="comment">## 2 1007_s_at MIR4640                               microRNA 4640 &lt;NA&gt;</span></div><div class="line"><span class="comment">##       MAP</span></div><div class="line"><span class="comment">## 1 6p21.33</span></div><div class="line"><span class="comment">## 2 6p21.33</span></div></pre></td></tr></table></figure>
<p>显然，该探针集合可以用于mRNA和miRNA丰度的定量。作为稳定的检查，我们可以看到，不同的符号映射到了相同的细胞带（最后一句不懂，原文为： As a sanity check we see that the distinct symbols map to the same cytoband）。</p>
<h2 id="总结">总结</h2>
<p>我们现在已经拥有了含有从核酸到通路水平的许多数据。通过Bioconductor.org上的View就可以查看现有的一些资源。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_annoOverview.html" target="_blank" rel="external">Genomic annotation in Bioconductor: The general situation</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
            <a href="/tags/基因组注释/" <i class="fa fa-tag"></i> 基因组注释</a>
          
            <a href="/tags/KEGG/" <i class="fa fa-tag"></i> KEGG</a>
          
            <a href="/tags/GO/" <i class="fa fa-tag"></i> GO</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/" rel="next" title="GDAS008-IRanges和GRanges">
                <i class="fa fa-chevron-left"></i> GDAS008-IRanges和GRanges
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/" rel="prev" title="GDAS010-生物学变异与技术变异">
                GDAS010-生物学变异与技术变异 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">249</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">118</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#基础注释资源与发现"><span class="nav-number">2.</span> <span class="nav-text">基础注释资源与发现</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#注释概念的层次结构"><span class="nav-number">2.1.</span> <span class="nav-text">注释概念的层次结构</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#发现可用的参考基因组"><span class="nav-number">2.2.</span> <span class="nav-text">发现可用的参考基因组</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#参考基因组的版本很重要"><span class="nav-number">2.3.</span> <span class="nav-text">参考基因组的版本很重要</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#h.-sapiens的参考基因序列"><span class="nav-number">3.</span> <span class="nav-text">H. sapiens的参考基因序列</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考序列的转录本和基因"><span class="nav-number">4.</span> <span class="nav-text">参考序列的转录本和基因</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#ucsc注释"><span class="nav-number">4.1.</span> <span class="nav-text">UCSC注释</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#ensembl注释"><span class="nav-number">4.2.</span> <span class="nav-text">ENSEMBL注释</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#你的数据将会成他人的注释导入导出"><span class="nav-number">5.</span> <span class="nav-text">你的数据将会成他人的注释：导入/导出</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#annotationhub"><span class="nav-number">6.</span> <span class="nav-text">AnnotationHub</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#the-orgdb基因注释图"><span class="nav-number">7.</span> <span class="nav-text">The OrgDb基因注释图</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#基因集和通路资源"><span class="nav-number">8.</span> <span class="nav-text">基因集和通路资源</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#基因本体论"><span class="nav-number">8.1.</span> <span class="nav-text">基因本体论</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#kegg"><span class="nav-number">8.2.</span> <span class="nav-text">KEGG</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#其它本体"><span class="nav-number">8.3.</span> <span class="nav-text">其它本体</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#通用基因集管理"><span class="nav-number">8.4.</span> <span class="nav-text">通用基因集管理</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#模式生物的统一自我描述方法"><span class="nav-number">9.</span> <span class="nav-text">模式生物的统一，自我描述方法</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#面向平台的注释"><span class="nav-number">10.</span> <span class="nav-text">面向平台的注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#总结"><span class="nav-number">11.</span> <span class="nav-text">总结</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">12.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共988.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
