<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言In the data management overview, we applied summarizeOverlaps to the set of BAM files in RNAseqData.HNRNPC.bam.chr14.  Let’s do it again to focus on the object that is returned. BAM files to Summari">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS006-深入理解SummarizedExperiment类">
<meta property="og:url" content="http://rvdsd.top/2019/09/05/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言In the data management overview, we applied summarizeOverlaps to the set of BAM files in RNAseqData.HNRNPC.bam.chr14.  Let’s do it again to focus on the object that is returned. BAM files to Summari">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkbo-1.png">
<meta property="og:updated_time" content="2019-11-06T14:17:47.401Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS006-深入理解SummarizedExperiment类">
<meta name="twitter:description" content="前言In the data management overview, we applied summarizeOverlaps to the set of BAM files in RNAseqData.HNRNPC.bam.chr14.  Let’s do it again to focus on the object that is returned. BAM files to Summari">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/05/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/"/>





  <title>GDAS006-深入理解SummarizedExperiment类 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/05/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS006-深入理解SummarizedExperiment类</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-05T12:00:00+08:00">
                2019-09-05
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,044
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  7
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h1><p>In the data management overview, we applied <code>summarizeOverlaps</code> to the set of BAM files in RNAseqData.HNRNPC.bam.chr14.  Let’s do it again to focus on the object that is returned.</p>
<h2 id="BAM-files-to-SummarizedExperiment-for-a-single-region"><a href="#BAM-files-to-SummarizedExperiment-for-a-single-region" class="headerlink" title="BAM files to SummarizedExperiment for a single region"></a>BAM files to SummarizedExperiment for a single region</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">bfp = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line">library(Rsamtools)</div><div class="line">bfl = BamFileList(file=bfp)</div><div class="line">hnrnpcLoc = GRanges(&quot;chr14&quot;, IRanges(21677296, 21737638))</div><div class="line">library(GenomicAlignments)</div><div class="line">library(BiocParallel)</div><div class="line">register(SerialParam())</div><div class="line">hnse = summarizeOverlaps(hnrnpcLoc,bfl)</div><div class="line">hnse</div><div class="line">## class: RangedSummarizedExperiment </div><div class="line">## dim: 1 8 </div><div class="line">## metadata(0):</div><div class="line">## assays(1): counts</div><div class="line">## rownames: NULL</div><div class="line">## rowData names(0):</div><div class="line">## colnames(8): ERR127306 ERR127307 ... ERR127304 ERR127305</div><div class="line">## colData names(0):</div></pre></td></tr></table></figure>
<p><code>hnse</code> is an instance of RangedSummarizedExperiment.  This class is like an ExpressionSet but has more facilities (and more modern facilities) for managing assay summaries and metadata. A visual schematic follows:</p>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png" alt="plot of chunk lkseee"></p>
<p>Effective use of SummarizedExperiment instances involves learning about methods that have been defined for them. In order to get at the read/region overlap counts for HNRNPC, we apply the <code>assay</code> method:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">assay(hnse)</div><div class="line">##      ERR127306 ERR127307 ERR127308 ERR127309 ERR127302 ERR127303 ERR127304</div><div class="line">## [1,]      5422      6320      5896      5558       172       196       316</div><div class="line">##      ERR127305</div><div class="line">## [1,]       282</div></pre></td></tr></table></figure>
<p>This is a bare-bones representation of the result.  The sample identifiers have been propagated to column names of the matrix of counts, but information on the region examined is lost in this display.</p>
<h2 id="Metadata-opportunities-in-the-SummarizedExperiment"><a href="#Metadata-opportunities-in-the-SummarizedExperiment" class="headerlink" title="Metadata opportunities in the SummarizedExperiment"></a>Metadata opportunities in the SummarizedExperiment</h2><p>The <code>hnse</code> object has a little more information.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">rowRanges(hnse)</div><div class="line">## GRanges object with 1 range and 0 metadata columns:</div><div class="line">##       seqnames               ranges strand</div><div class="line">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</div><div class="line">##   [1]    chr14 [21677296, 21737638]      *</div><div class="line">##   -------</div><div class="line">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div><div class="line">seqinfo(hnse)</div><div class="line">## Seqinfo object with 1 sequence from an unspecified genome; no seqlengths:</div><div class="line">##   seqnames seqlengths isCircular genome</div><div class="line">##   chr14            NA         NA   &lt;NA&gt;</div><div class="line">metadata(hnse)</div><div class="line">## list()</div></pre></td></tr></table></figure>
<p>We can do better.  We will set up the analysis differently so that the output is more comprehensive and self-describing.</p>
<h1 id="Making-SummarizedExperiment-more-effective-by-enriching-the-inputs-to-summarizing-methods"><a href="#Making-SummarizedExperiment-more-effective-by-enriching-the-inputs-to-summarizing-methods" class="headerlink" title="Making SummarizedExperiment more effective by enriching the inputs to summarizing methods"></a>Making SummarizedExperiment more effective by enriching the inputs to summarizing methods</h1><h2 id="Defining-regions-of-interest-with-metadata"><a href="#Defining-regions-of-interest-with-metadata" class="headerlink" title="Defining regions of interest, with metadata"></a>Defining regions of interest, with metadata</h2><p>We have seen that it is sufficient to define a single GRanges to drive <code>summarizeOverlaps</code> over a set of BAM files.  We’d like to preserve more metadata about the regions examined. We’ll use the TxDb infrastructure, to be described in more detail later, to get a structure defining gene regions on chr14. We’ll also use the Homo.sapiens annotation package to add gene symbols.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div></pre></td><td class="code"><pre><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line">gr14 = genes(txdb, vals=list(tx_chrom=&quot;chr14&quot;))</div><div class="line">## Warning: &apos;elementLengths&apos; is deprecated.</div><div class="line">## Use &apos;elementNROWS&apos; instead.</div><div class="line">## See help(&quot;Deprecated&quot;)</div><div class="line"></div><div class="line">## Warning: &apos;elementLengths&apos; is deprecated.</div><div class="line">## Use &apos;elementNROWS&apos; instead.</div><div class="line">## See help(&quot;Deprecated&quot;)</div><div class="line">gr14$symbol = mapIds(Homo.sapiens, keys=gr14$gene_id, keytype=&quot;ENTREZID&quot;,</div><div class="line">   column=&quot;SYMBOL&quot;)</div><div class="line">## &apos;select()&apos; returned 1:1 mapping between keys and columns</div><div class="line">gr14</div><div class="line">## GRanges object with 781 ranges and 2 metadata columns:</div><div class="line">##             seqnames                 ranges strand |     gene_id</div><div class="line">##                &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</div><div class="line">##       10001    chr14 [ 71050957,  71067384]      - |       10001</div><div class="line">##   100113389    chr14 [ 45580078,  45580176]      + |   100113389</div><div class="line">##   100113391    chr14 [ 20794600,  20794698]      - |   100113391</div><div class="line">##   100124539    chr14 [ 91592770,  91592896]      + |   100124539</div><div class="line">##   100126297    chr14 [101507700, 101507781]      + |   100126297</div><div class="line">##         ...      ...                    ...    ... .         ...</div><div class="line">##        9870    chr14 [ 75127955,  75179807]      - |        9870</div><div class="line">##        9878    chr14 [ 21945335,  21967319]      + |        9878</div><div class="line">##        9895    chr14 [102829300, 102968818]      + |        9895</div><div class="line">##        9950    chr14 [ 93260650,  93306304]      + |        9950</div><div class="line">##        9985    chr14 [ 24641234,  24649463]      + |        9985</div><div class="line">##                  symbol</div><div class="line">##             &lt;character&gt;</div><div class="line">##       10001        MED6</div><div class="line">##   100113389    SNORD127</div><div class="line">##   100113391    SNORD126</div><div class="line">##   100124539    SNORA11B</div><div class="line">##   100126297      MIR300</div><div class="line">##         ...         ...</div><div class="line">##        9870       AREL1</div><div class="line">##        9878        TOX4</div><div class="line">##        9895      TECPR2</div><div class="line">##        9950      GOLGA5</div><div class="line">##        9985        REC8</div><div class="line">##   -------</div><div class="line">##   seqinfo: 93 sequences (1 circular) from hg19 genome</div></pre></td></tr></table></figure>
<h2 id="Defining-the-sample-characteristics-for-the-BAM-files"><a href="#Defining-the-sample-characteristics-for-the-BAM-files" class="headerlink" title="Defining the sample characteristics for the BAM files"></a>Defining the sample characteristics for the BAM files</h2><p>We have three distinct preparations, one for control and two for knockdown. We will use the GenomicFiles infrastructure to bind the sample level metadata.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">char = rep(c(&quot;hela_wt&quot;, &quot;hela_hkd&quot;), each=4)</div><div class="line">bff = GenomicFiles(files=path(bfl))</div><div class="line">colData(bff)$condition = char</div><div class="line">sid = c(1,1,1,1,2,2,3,3)</div><div class="line">bff$sample = sid</div><div class="line">bff</div><div class="line">## GenomicFiles object with 0 ranges and 8 files: </div><div class="line">## files: ERR127306_chr14.bam, ERR127307_chr14.bam, ..., ERR127304_chr14.bam, ERR127305_chr14.bam </div><div class="line">## detail: use files(), rowRanges(), colData(), ...</div></pre></td></tr></table></figure>
<h2 id="Comparing-read-overlaps-preserving-metadata"><a href="#Comparing-read-overlaps-preserving-metadata" class="headerlink" title="Comparing read overlaps, preserving metadata"></a>Comparing read overlaps, preserving metadata</h2><p>We’ll look at 5 genes, including HNRNPC.  After computing, we bind the sample-level data back into the result.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">hnse = summarizeOverlaps(gr14[c(1:4,305)],files(bff))</div><div class="line">colData(hnse) = cbind(colData(hnse), colData(bff))</div><div class="line">hnse</div><div class="line">## class: RangedSummarizedExperiment </div><div class="line">## dim: 5 8 </div><div class="line">## metadata(0):</div><div class="line">## assays(1): counts</div><div class="line">## rownames(5): 10001 100113389 100113391 100124539 3183</div><div class="line">## rowData names(2): gene_id symbol</div><div class="line">## colnames(8): ERR127306 ERR127307 ... ERR127304 ERR127305</div><div class="line">## colData names(2): condition sample</div><div class="line">assay(hnse)</div><div class="line">##           ERR127306 ERR127307 ERR127308 ERR127309 ERR127302 ERR127303</div><div class="line">## 10001           114       156       129       144       175       213</div><div class="line">## 100113389         0         0         0         0         0         0</div><div class="line">## 100113391         0         0         0         0         0         0</div><div class="line">## 100124539         0         0         0         0         0         0</div><div class="line">## 3183           2711      3160      2948      2779        86        98</div><div class="line">##           ERR127304 ERR127305</div><div class="line">## 10001           210       165</div><div class="line">## 100113389         0         0</div><div class="line">## 100113391         0         0</div><div class="line">## 100124539         0         0</div><div class="line">## 3183            158       141</div></pre></td></tr></table></figure>
<p>Note that row identifiers are now present with the count matrix.</p>
<p>A simple sanity check:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(2,2))</div><div class="line">for (i in 2:5) &#123;</div><div class="line">  boxplot(assay(hnse)[i,]~hnse$condition, ylab=rowRanges(hnse)$symbol[i])</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkbo-1.png" alt="plot of chunk lkbo"></p>
<h1 id="Converting-from-ExpressionSet"><a href="#Converting-from-ExpressionSet" class="headerlink" title="Converting from ExpressionSet"></a>Converting from ExpressionSet</h1><p>This is easy, but more work will be needed to allow subsetting of array probes based on genomic range queries.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div></pre></td><td class="code"><pre><div class="line">library(ALL)</div><div class="line">data(ALL)</div><div class="line">allse = makeSummarizedExperimentFromExpressionSet(ALL)</div><div class="line">allse</div><div class="line">## class: RangedSummarizedExperiment </div><div class="line">## dim: 12625 128 </div><div class="line">## metadata(3): experimentData annotation protocolData</div><div class="line">## assays(1): exprs</div><div class="line">## rownames(12625): 1000_at 1001_at ... AFFX-YEL021w/URA3_at</div><div class="line">##   AFFX-YEL024w/RIP1_at</div><div class="line">## rowData names(0):</div><div class="line">## colnames(128): 01005 01010 ... 83001 LAL4</div><div class="line">## colData names(21): cod diagnosis ... f.u date.last.seen</div><div class="line">rowRanges(allse)</div><div class="line">## Warning: &apos;elementLengths&apos; is deprecated.</div><div class="line">## Use &apos;elementNROWS&apos; instead.</div><div class="line">## See help(&quot;Deprecated&quot;)</div><div class="line">## GRangesList object of length 12625:</div><div class="line">## $$1000_at </div><div class="line">## GRanges object with 0 ranges and 0 metadata columns:</div><div class="line">##    seqnames    ranges strand</div><div class="line">##       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</div><div class="line">## </div><div class="line">## $$1001_at </div><div class="line">## GRanges object with 0 ranges and 0 metadata columns:</div><div class="line">##      seqnames ranges strand</div><div class="line">## </div><div class="line">## $$1002_f_at </div><div class="line">## GRanges object with 0 ranges and 0 metadata columns:</div><div class="line">##      seqnames ranges strand</div><div class="line">## </div><div class="line">## ...</div><div class="line">## &lt;12622 more elements&gt;</div><div class="line">## -------</div><div class="line">## seqinfo: no sequences</div></pre></td></tr></table></figure>
<h1 id="Summary"><a href="#Summary" class="headerlink" title="Summary"></a>Summary</h1><p>The RangedSummarizedExperiment class instantiates some of the key principles of Bioconductor data structure design:</p>
<ul>
<li>Assay data and metadata on sample characteristics (colData) are bound together in a coordinated way</li>
<li>Matrix-like subsetting works directly on both assay and sample data</li>
<li>Range-based subsetting works for assay components addressible by genomic coordinates</li>
<li>Arbitrary metadata on assay features can be provided in the mcols(rowRanges(se))</li>
<li>Arbitrary general metadata can be provided through <code>metadata(se)&lt;-</code></li>
</ul>
<p>We’ll learn more about adaptations of SummarizedExperiment to perform specifically for multistage processing and analysis of RNA-seq experiments later in the course.</p>
<hr>
<p>​     <a href="http://genomicsclass.github.io/book/" target="_blank" rel="external">PH525x</a>,  Rafael Irizarry and Michael Love,  <a href="https://github.com/genomicsclass/labs/blob/master/LICENSE" target="_blank" rel="external">MIT License</a>             </p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/04/Genomics Data Analysis Series/GDAS005-管理基因组级数据/" rel="next" title="GDAS004-管理基因组级数据">
                <i class="fa fa-chevron-left"></i> GDAS004-管理基因组级数据
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/05/Genomics Data Analysis Series/GDAS005-基因组实验中的数据结构与管理/" rel="prev" title="GDAS005-基因组实验中的数据结构与管理">
                GDAS005-基因组实验中的数据结构与管理 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">236</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">113</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#BAM-files-to-SummarizedExperiment-for-a-single-region"><span class="nav-number">1.1.</span> <span class="nav-text">BAM files to SummarizedExperiment for a single region</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Metadata-opportunities-in-the-SummarizedExperiment"><span class="nav-number">1.2.</span> <span class="nav-text">Metadata opportunities in the SummarizedExperiment</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Making-SummarizedExperiment-more-effective-by-enriching-the-inputs-to-summarizing-methods"><span class="nav-number">2.</span> <span class="nav-text">Making SummarizedExperiment more effective by enriching the inputs to summarizing methods</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Defining-regions-of-interest-with-metadata"><span class="nav-number">2.1.</span> <span class="nav-text">Defining regions of interest, with metadata</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Defining-the-sample-characteristics-for-the-BAM-files"><span class="nav-number">2.2.</span> <span class="nav-text">Defining the sample characteristics for the BAM files</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Comparing-read-overlaps-preserving-metadata"><span class="nav-number">2.3.</span> <span class="nav-text">Comparing read overlaps, preserving metadata</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Converting-from-ExpressionSet"><span class="nav-number">3.</span> <span class="nav-text">Converting from ExpressionSet</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Summary"><span class="nav-number">4.</span> <span class="nav-text">Summary</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共938.3k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
