<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Hexo, NexT" />










<meta property="og:type" content="website">
<meta property="og:title" content="RVDSD的个人笔记本">
<meta property="og:url" content="http://rvdsd.top/page/5/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RVDSD的个人笔记本">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/page/5/"/>





  <title>RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">学习过程中的输出</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/22/BioTools/BioStar笔记之Entrez/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/22/BioTools/BioStar笔记之Entrez/" itemprop="url">Biostar Handbook笔记之Entrez</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-22T12:00:00+08:00">
                2018-08-22
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,872
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="什么是entrez">什么是Entrez</h2>
<p>Entrez是NCBI的一个检索工具，它整合了PubMed数据库中的生物学文献中的信息，包括DNA序列，蛋白质序列，结构，基因组，基因表达的信息等。Entrez是一个法语单词，读作“En-trez”。</p>
<h2 id="entrez的使用">Entrez的使用</h2>
<p>NCBI提供了Entrez的一个WEB API，称为<a href="https://www.ncbi.nlm.nih.gov/books/NBK25500/" target="_blank" rel="external">Entrez E-utils</a>，还有一套工具，称为<a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/" target="_blank" rel="external">Entrez Direct</a>。其中Entrez Direct可以在Linux下运行，通过命令行的形式访问NCBi的数据库。</p>
<h2 id="ncbi的数据组织形式">NCBI的数据组织形式</h2>
<p>NCBI上的数据是随着生物学的研究而不断更新的。例如我们以<code>Ebola virus 1976</code>为关键词在NCBI的Nucleotide中检索一下，单击第二行的链接，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180829/F61bhef8kd.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>我们会看到<a href="https://www.ncbi.nlm.nih.gov/nuccore/AF086833.2" target="_blank" rel="external">如下信息</a>：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180829/mdEHmK4lcE.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在解释一下这个信息：</p>
<ol style="list-style-type: decimal">
<li><code>AF086833</code>是登录号(accession number)，它代表了一个完整的数据记录，即使这个数据的版本再更改，它也会保持不变，具有唯一性。</li>
<li><code>VERSION AF086833.2</code>表示版本号，它与登录号比较相似，只是在登录号的后面用了一个号和一个数字表示，即<code>.2</code>。版本号的逗号前面部分不变，但是，如果序列发生改变，只改变版本号的逗号后面部分。</li>
</ol>
<p>自从2015年，NCBI已经不再使用原本的GI号了，一个数据对应一个GI号，不过一些老的文献中可能还会使用GI或gi号，这里需要注意一下。</p>
<h2 id="entrez-e-utils-web-api的使用">Entrez E-utils Web API的使用</h2>
<p>我们可以通过一个特定的URL使用Entrez Web API接口来访问NCBI的数据，这个接口的构成形式是这个样子的：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">https://service.nih.gov?param1=value1&amp;param2=value2&amp;param3=value3</div></pre></td></tr></table></figure>
<p>由于<code>&amp;</code>符号在shell中是一个特殊符号，因为我需要使用一个单引号，或者是转义字符来正常使用<code>&amp;</code>，转义字符的使用是<code>\&amp;</code>。</p>
<p>现在我们使用这种方法来获得<code>AF086833.2</code>的数据，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">curl -s https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2\&amp;db=nuccore\&amp;rettype=fasta | head</div><div class="line"><span class="meta">#</span> 或者是下面的这个样子</div><div class="line">curl -s 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2&amp;db=nuccore&amp;rettype=fasta' | head</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ curl -s https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2\&amp;db=nuccore\&amp;rettype=fasta | head</div><div class="line"><span class="meta">&gt;</span>AF086833.2 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATAACTATGAGGAAGATTAATAA</div><div class="line">TTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTT</div><div class="line">CAGAGCCACATCACAAAGATAGAGAACAACCTAGGTCTCCGAAGGGAGCAAGGGCATCAGTGTGCTCAGT</div><div class="line">TGAAAATCCCTTGTCAACACCTAGGTCTTATCACATCACAAGTTCCACCTCAGACTCTGCAGGGTGATCC</div><div class="line">AACAACCTTAATAGAAACATTATTGTTAAAGGACAGCATTAGTTCACAGTCAAACAAGCAAGATTGAGAA</div><div class="line">TTAACCTTGGTTTTGAACTTGAACACTTAGGGGATTGAAGATTCAACAACCCTAAAGCTTGGGGTAAAAC</div><div class="line">ATTGGAAATAGTTAAAAGACAAATTGCTCGGAATCACAAAATTCCGAGTATGGATTCTCGTCCTCAGAAA</div><div class="line">ATCTGGATGGCGCCGAGTCTCACTGAATCTGACATGGATTACCACAAGATCTTGACAGCAGGTCTGTCCG</div><div class="line">TTCAACAGGGGATTGTTCGGCAAAGAGTCATCCCAGTGTATCAAGTAAACAATCTTGAAGAAATTTGCCA</div></pre></td></tr></table></figure>
<h2 id="entrez-direct的使用">Entrez Direct的使用</h2>
<p>Entrez Direct简称EDirect，它算是Entrez E-utils的本地版本.</p>
<h3 id="安装">安装</h3>
<p>它的安装如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">sudo apt install ncbi-entrez-direct</div><div class="line"><span class="meta">#</span> 如果安装不了，可以直接从官网下载</div><div class="line"><span class="meta">#</span> curl ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/edirect.zip -o edirect.zip</div></pre></td></tr></table></figure>
<h3 id="下载序列">下载序列</h3>
<p>现在我们先看一下下面的这个命令<code>efetch</code>，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">efetch -db=nuccore -format=gb -id=AF086833 | head</div></pre></td></tr></table></figure>
<p>fetch的英文意思是拿来，从这个命令的命名上大概能猜出来，这个命令是用来检索一些信息的。</p>
<p>通过<code>efetch --help</code>看了这个命令的说明后，我们知道，<code>-db=nuccore</code>表示检索的是核酸（为什么使用nuccore，后来我试了一下<code>-db=nucleotide</code>也可以)还不清楚。<code>-format=gb</code>表示显示的是GeneBank的格式，<code>id=AF086833</code>表示登录号。</p>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ </div><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ efetch -db=nuccore -format=gb -id=AF086833 | head</div><div class="line">LOCUS       AF086833               18959 bp    cRNA    linear   VRL 13-FEB-2012</div><div class="line">DEFINITION  Ebola virus - Mayinga, Zaire, 1976, complete genome.</div><div class="line">ACCESSION   AF086833</div><div class="line">VERSION     AF086833.2</div><div class="line">KEYWORDS    .</div><div class="line">SOURCE      Ebola virus - Mayinga, Zaire, 1976 (EBOV-May)</div><div class="line">  ORGANISM  Ebola virus - Mayinga, Zaire, 1976</div><div class="line">            Viruses; ssRNA viruses; ssRNA negative-strand viruses;</div><div class="line">            Mononegavirales; Filoviridae; Ebolavirus.</div><div class="line">REFERENCE   1  (bases 1 to 18959)</div></pre></td></tr></table></figure>
<p>前面提到，<code>-db=nuccore</code>表示的是核酸，为什么不使用<code>-db=nucleotide</code>，我发现使用了一下，也能显示，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/python_zjl$ efetch -db=nucleotide  -format=gb -id=AF086833 | head</div><div class="line">LOCUS       AF086833               18959 bp    cRNA    linear   VRL 13-FEB-2012</div><div class="line">DEFINITION  Ebola virus - Mayinga, Zaire, 1976, complete genome.</div><div class="line">ACCESSION   AF086833</div><div class="line">VERSION     AF086833.2</div><div class="line">KEYWORDS    .</div><div class="line">SOURCE      Ebola virus - Mayinga, Zaire, 1976 (EBOV-May)</div><div class="line">  ORGANISM  Ebola virus - Mayinga, Zaire, 1976</div><div class="line">            Viruses; ssRNA viruses; ssRNA negative-strand viruses;</div><div class="line">            Mononegavirales; Filoviridae; Ebolavirus.</div><div class="line">REFERENCE   1  (bases 1 to 18959)</div></pre></td></tr></table></figure>
<h3 id="下载不同格式的序列">下载不同格式的序列</h3>
<p>使用<code>efetch</code>功能，还能把<code>AF086833</code>的信息导入到不同格式的文件中去，但要注意<code>-format=格式</code>这个参数，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=gb -id=AF086833 &gt; AF086833.gb</div><div class="line"><span class="meta">#</span> download information output Genebank format files</div><div class="line"></div><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=fasta -id=AF086833 &gt; AF086833.fa</div><div class="line"><span class="meta">#</span> download information output fasta format files</div></pre></td></tr></table></figure>
<p>如果要下载多条序列，可以直接在<code>id=</code>参数后面添加就行了，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="meta">$</span> efetch -db=nuccore -format=fasta -id=AF086833,AF086833 &gt; test.fa</div><div class="line"><span class="meta">#</span> 这里只是说明一下下载多个序列的用法，因此id=后面接了两个相同的登录号</div></pre></td></tr></table></figure>
<h3 id="显示指定的碱基范围">显示指定的碱基范围</h3>
<p>efetch还可以显示指定的序列范围，例如下面就是显示<code>AF086833.2</code>这个序列的第100到300个碱基，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=100 -seq_stop=300</div><div class="line"><span class="meta">&gt;</span>AF086833.2:100-300 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">TTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTTCAGAGCCACATCACAAAGATAGAGAACAA</div><div class="line">CCTAGGTCTCCGAAGGGAGCAAGGGCATCAGTGTGCTCAGTTGAAAATCCCTTGTCAACACCTAGGTCTT</div><div class="line">ATCACATCACAAGTTCCACCTCAGACTCTGCAGGGTGATCCAACAACCTTAATAGAAACAT</div></pre></td></tr></table></figure>
<p>这里要使用两个参数，一个是<code>-seq_start=</code>，它是起始的碱基位置，一个是<code>-seq_stop=</code>，它是碱基的结束位置。</p>
<h3 id="显示互补链">显示互补链</h3>
<p>通过添加<code>-strand</code>参数可以显示互补链，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=1 -seq_stop=5 -strand=1</div><div class="line"><span class="meta">&gt;</span>AF086833.2:1-5 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">CGGAC</div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db=nuccore -format=fasta -id=AF086833 -seq_start=1 -seq_stop=5 -strand=2</div><div class="line"><span class="meta">&gt;</span>AF086833.2:c5-1 Ebola virus - Mayinga, Zaire, 1976, complete genome</div><div class="line">GTCCG</div></pre></td></tr></table></figure>
<p>从上面的结果可以看出来，<code>strand=1</code>表示是正义链，也就是<code>CGGAC</code>，而<code>strand=2</code>表示的是互补链，也就是<code>CGGAC</code>的互补链，即<code>GTCCG</code>。</p>
<h2 id="entrez-direct检索">Entrez Direct检索</h2>
<p>如果我们有一个项目号(project accession)，例如<code>PRJNA257197</code>，我们就能通过EDirect来检索这个项目的一些信息。这个项目号存在于发表的文章中，或者是文件的附件里。现在我们利用<code>esearch</code>来检索一下，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">esearch -db=nucleotide -query PRJNA257197</div><div class="line">esearch -db protein -query PRJNA257197</div></pre></td></tr></table></figure>
<p>其中<code>-query</code>后面跟的是要检索的关系词。</p>
<p>结果如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ esearch -db nucleotide -query PRJNA257197</div><div class="line">&lt;ENTREZ_DIRECT&gt;</div><div class="line">  &lt;Db&gt;nucleotide&lt;/Db&gt;</div><div class="line">  &lt;WebEnv&gt;NCID_1_215173136_130.14.18.125_9001_1535597951_440051473_0MetA0_S_MegaStore&lt;/WebEnv&gt;</div><div class="line">  &lt;QueryKey&gt;1&lt;/QueryKey&gt;</div><div class="line">  &lt;Count&gt;249&lt;/Count&gt;</div><div class="line">  &lt;Step&gt;1&lt;/Step&gt;</div><div class="line">&lt;/ENTREZ_DIRECT&gt;</div><div class="line"></div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ esearch -db protein -query PRJNA257197</div><div class="line">&lt;ENTREZ_DIRECT&gt;</div><div class="line">  &lt;Db&gt;protein&lt;/Db&gt;</div><div class="line">  &lt;WebEnv&gt;NCID_1_215068217_130.14.22.215_9001_1535597965_849331092_0MetA0_S_MegaStore&lt;/WebEnv&gt;</div><div class="line">  &lt;QueryKey&gt;1&lt;/QueryKey&gt;</div><div class="line">  &lt;Count&gt;2240&lt;/Count&gt;</div><div class="line">  &lt;Step&gt;1&lt;/Step&gt;</div><div class="line">&lt;/ENTREZ_DIRECT&gt;</div></pre></td></tr></table></figure>
<p>在上面的代码中，我们使用了两条命令来检索<code>PRJNA257197</code>的信息，结果显示了这个项目的“环境”(environment)“，这个环境，我的理解就是，这个项目在Web上的一些配置信息，如下所示：</p>
<ol style="list-style-type: decimal">
<li><code>&lt;Db&gt;nucleotide&lt;/Db&gt;</code>说明这是一个核酸数据库；</li>
<li>剩下的看不懂，不知道是什么，可能是ip地址。</li>
<li><code>&lt;Count&gt;249&lt;/Count&gt;</code>表示有249个检索结果。</li>
</ol>
<h2 id="edirect提取核酸与蛋白信息">EDirect提取核酸与蛋白信息</h2>
<p>我们可以采用管道操作，将这个项目中的一些信息，例如核酸的信息下载下来，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db nucleotide -query PRJNA257197 | efetch -format=fasta &gt; genomes.fa</div></pre></td></tr></table></figure>
<p>也可以下载蛋白质的信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">esearch -db protein -query PRJNA257197 | efetch -format=fasta &gt; proteins.fa</div></pre></td></tr></table></figure>
<p>查看一下相应的信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ ls</div><div class="line">AF086833.fa  AF086833.gb  genomes.fa  proteins.fa</div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ head -3 genomes.fa </div><div class="line"><span class="meta">&gt;</span>KR105345.1 Zaire ebolavirus isolate Ebola virus/H.sapiens-wt/SLE/2014/Makona-G6089.1, partial genome</div><div class="line">ATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTT</div><div class="line">GTTTCAGAGCCATATCACCAAGATAGAGAACAACCTAGGTCTCCGGAGGGGGCAAGGGCATCAGTGTGCT</div><div class="line"></div><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ head -3 proteins.fa </div><div class="line"><span class="meta">&gt;</span>AKC37233.1 polymerase [Zaire ebolavirus]</div><div class="line">MATQHTQYPDARLSSPIVLDQCDLVTRACGLYSSYSLNPQLRNCKLPKHIYRLKYDVTVTKFLSDVPVAT</div><div class="line">LPIDFIVPILLKALSGNGFCPVEPRCQQFLDEIIKYTMQDALFLKYYLKNVGAQEDCVDDHFQEKILSSI</div></pre></td></tr></table></figure>
<h3 id="提取复杂信息">提取复杂信息</h3>
<p>Entrez Direct中有一个<code>xtract</code>工具，可以用于从XML文件中提取相应的信息，如下所示：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">biotest@biotest-VirtualBox:~/biostar/EDirect$ efetch -db taxonomy -id 9606,7227,10090 -format xml | xtract -Pattern Taxon -first TaxId Scien tificName GenbankCommonName Division</div><div class="line">9606	human	Primates</div><div class="line">7227	fruit fly	Invertebrates</div><div class="line">10090	house mouse	Rodents</div></pre></td></tr></table></figure>
<p>总之，Entrez中有很多有用的工具，没事可以多看看文档。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>Biostar Handbook</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/19/BioStatistics/生物统计-16决策树/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/19/BioStatistics/生物统计-16决策树/" itemprop="url">决策树笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-19T12:02:00+08:00">
                2018-08-19
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,377
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  10
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="决策树的原理">决策树的原理</h2>
<p>决策树的原理可以参考这篇笔记<a href="http://rvdsd.top/2018/07/01/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B019-%E5%86%B3%E7%AD%96%E6%A0%91/">《StatQuest学习笔记19——决策树》</a>。</p>
<h2 id="决策树的计算过程">决策树的计算过程</h2>
<p>决策树(decision tree)用于回归时称为回归树(regression tree)，用于分类时则称为分类树(classification tree)。这里我们先看一批数据，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/Dh8fddjd9h.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>原始数据很多，有1000多行。这个数据来自于Flake and lawrence(2002)，可以在<a href="https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/regression/mg" target="_blank" rel="external">网上下载</a>，这且数据是做什么的，我也不清楚。这个数据集有6个自变量和1个因变量，一共有1385个观测值。书上说之所以用这个数据是因为这里的所有变量都是数量变量，因此这个数据集也可以用经典线性回归来做，从而可以对各种方法进行比较。</p>
<h3 id="线性回归">线性回归</h3>
<p>现在先对全部数据做一个简单的 线性回归，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">raw_dt &lt;- read.table(<span class="string">"https://raw.githubusercontent.com/20170505a/raw_data/master/mg.csv"</span>,header=<span class="literal">T</span>,sep=<span class="string">","</span>)</div><div class="line">result_dt &lt;- lm(y~.,raw_dt)</div><div class="line">summary(result_dt)</div><div class="line">pairs(raw_dt)<span class="comment">#查看不同变量之间的相关关系</span></div><div class="line"><span class="comment"># 或plot(raw_dt)</span></div><div class="line">cor(raw_dt)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; summary(result_dt)</div><div class="line"></div><div class="line">Call:</div><div class="line">lm(formula = y ~ ., data = raw_dt)</div><div class="line"></div><div class="line">Residuals:</div><div class="line">     Min       1Q   Median       3Q      Max </div><div class="line">-0.40866 -0.09516  0.01098  0.09645  0.40011 </div><div class="line"></div><div class="line">Coefficients:</div><div class="line">            Estimate Std. Error t value Pr(&gt;|t|)    </div><div class="line">(Intercept)  0.95582    0.07142  13.384  &lt; 2e-16 ***</div><div class="line">x1           0.28512    0.02773  10.282  &lt; 2e-16 ***</div><div class="line">x2           0.17263    0.02655   6.502 1.10e-10 ***</div><div class="line">x3          -0.37956    0.02769 -13.709  &lt; 2e-16 ***</div><div class="line">x4          -0.36756    0.02764 -13.298  &lt; 2e-16 ***</div><div class="line">x5           0.07165    0.02656   2.697  0.00707 ** </div><div class="line">x6           0.19003    0.02778   6.841 1.18e-11 ***</div><div class="line">---</div><div class="line">Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1</div><div class="line"></div><div class="line">Residual standard error: 0.1459 on 1378 degrees of freedom</div><div class="line">Multiple R-squared:  0.5868,	Adjusted R-squared:  0.585 </div><div class="line">F-statistic: 326.2 on 6 and 1378 DF,  p-value: &lt; 2.2e-16</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/egLil2EEmc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>mg数据各个变量之间的相关系数如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">&gt; cor(raw_dt)</div><div class="line">            y         x1         x2         x3         x4         x5         x6</div><div class="line">y   <span class="number">1.0000000</span>  <span class="number">0.2620280</span>  <span class="number">0.5441721</span> -<span class="number">0.5006887</span> -<span class="number">0.4355964</span>  <span class="number">0.5510561</span>  <span class="number">0.1511016</span></div><div class="line">x1  <span class="number">0.2620280</span>  <span class="number">1.0000000</span> -<span class="number">0.1465800</span> -<span class="number">0.6543690</span>  <span class="number">0.4333849</span>  <span class="number">0.4543362</span> -<span class="number">0.6277910</span></div><div class="line">x2  <span class="number">0.5441721</span> -<span class="number">0.1465800</span>  <span class="number">1.0000000</span> -<span class="number">0.1475594</span> -<span class="number">0.6539043</span>  <span class="number">0.4367584</span>  <span class="number">0.4515753</span></div><div class="line">x3 -<span class="number">0.5006887</span> -<span class="number">0.6543690</span> -<span class="number">0.1475594</span>  <span class="number">1.0000000</span> -<span class="number">0.1471907</span> -<span class="number">0.6537680</span>  <span class="number">0.4397873</span></div><div class="line">x4 -<span class="number">0.4355964</span>  <span class="number">0.4333849</span> -<span class="number">0.6539043</span> -<span class="number">0.1471907</span>  <span class="number">1.0000000</span> -<span class="number">0.1494223</span> -<span class="number">0.6524989</span></div><div class="line">x5  <span class="number">0.5510561</span>  <span class="number">0.4543362</span>  <span class="number">0.4367584</span> -<span class="number">0.6537680</span> -<span class="number">0.1494223</span>  <span class="number">1.0000000</span> -<span class="number">0.1498846</span></div><div class="line">x6  <span class="number">0.1511016</span> -<span class="number">0.6277910</span>  <span class="number">0.4515753</span>  <span class="number">0.4397873</span> -<span class="number">0.6524989</span> -<span class="number">0.1498846</span>  <span class="number">1.0000000</span></div></pre></td></tr></table></figure>
<p>从前面的线性回归结果可以看出来，<span class="math inline">\(R^2=0.5868\)</span>，这个数值并不太好，再加上两两散点图与相关系数的结果，我们看不出各个变量之间的任何模式。现在我们使用其它的几种算法建模方法，同时使用交叉验证（利用均方误差或标准化均方误差）对各种方法时行比较。</p>
<p>下面我们对各种方法都用10折交叉验证的方法来判断其结果的可靠性（关于10折交叉验证的一些知识，可以看这笔记<a href="http://rvdsd.top/2018/07/05/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B022-%E4%BA%A4%E5%8F%89%E9%AA%8C%E8%AF%81/">《StatQuest学习笔记22——交叉验证》</a>）。对于每种方法，按照随机建立的10个训练集通过计算建立10个模型，对测试集分别得到10个标准化均方误差(NMSE)，再得出10次平均的NMSE。令<span class="math inline">\(\bar{y}\)</span>为因变量均值，<span class="math inline">\(\hat{y}\)</span>为从训练集得到的模型对一个测试数据集的预测值，这里NMSE的定义如下所示：</p>
<p><span class="math inline">\(NMSE=\bar{(y-\hat{y})^2}/=\bar{(y-\bar{y})^2}=\sum(y-\hat{y})^2/\sum(y-\bar{y})^2\)</span></p>
<p>显然，如果什么模型都不用，仅仅用均值来做预测，那么NMSE应该等于1.所以，如果在回归时得到NMSE大于1，这个模型就很糟糕了，还不如没有模型。仅仅对于训练集来说，其NMSE等于<span class="math inline">\(1-R^2\)</span>，这里的<span class="math inline">\(R^2\)</span>为回归的确定系数，但是对于测试集来说，其NMSE与测试集回归的<span class="math inline">\(R^2\)</span>没有什么关系，交叉验证主要关心测试集的NMSE。</p>
<h3 id="构建训练集和测试集">构建训练集和测试集</h3>
<p>现在构建为交叉验证服务的10个训练集和测试集，具体过程是：随机把下标分配给1,2,…10这10个数字，也就是把数据下标随机分成10份，然后每次提取一份作为测试集，其它9份放在一起作为训练集，用模型进行拟合，记下结果和误差，如此下去，一共做10次，最后把误差平均起来。这里所用的随机选Z折下标集的函数如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">CV=<span class="keyword">function</span>(n,Z=<span class="number">10</span>,seed=<span class="number">1000</span>)&#123;</div><div class="line">  z=rep(<span class="number">1</span>:Z,ceiling(n/Z))[<span class="number">1</span>:n]</div><div class="line">  set.seed(seed)</div><div class="line">  z=sample(z,n)</div><div class="line">  mm=list()</div><div class="line">  <span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:Z)</div><div class="line">    mm[[i]]=(<span class="number">1</span>:n)[z==i]</div><div class="line">  <span class="keyword">return</span>(mm)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>这里的<code>n</code>是样本量，<code>Z</code>为折的数目（也就是训练集加测试集的数目，这里默认是10），<code>seed</code>为随机种子，这里设为了1000，转出为<code>Z</code>个下标集，<code>mm[[i]]</code>为第i（i=1,2,…,Z)个下标集，根据这个函数对这个mg数据集求10个下标集的代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">n&lt;-nrow(raw_dt)</div><div class="line">Z=<span class="number">10</span></div><div class="line">mm=CV(n,Z)</div><div class="line">D=<span class="number">1</span></div></pre></td></tr></table></figure>
<p>在这段代码中，<code>mm</code>存储了10个下标集，而D说明因变量是第一个变量，这些值在后面对每种方法都要用。</p>
<p>为了和其它方法进行地比较，先对数据mg求简单的线性回归的10折交叉验证的测试集的NMSE，代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; mean(MSE)</div><div class="line">[1] 0.4222689</div></pre></td></tr></table></figure>
<p>结果表明，测试集的NMSE为0.4222689</p>
<h2 id="决策树回归">决策树回归</h2>
<p>决策树(decision tree)用于回归时称为回归树(regression tree)，用于分类时则称为分类树(classification tree)，现在我们先用数据集mg做一棵回归树，然后解释其意义。这里要使用到<code>rpart</code>包中的<code>rpart()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rpart)</div><div class="line"><span class="keyword">library</span>(rpart.plot)</div><div class="line">a=rpart(y~.,raw_dt)</div><div class="line">rpart.plot(a,type=<span class="number">2</span>,faclen=<span class="number">0</span>)</div></pre></td></tr></table></figure>
<p>结果如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">&gt; a</div><div class="line">n= 1385 </div><div class="line"></div><div class="line">node), split, n, deviance, yval</div><div class="line">      * denotes terminal node</div><div class="line"></div><div class="line"> 1) root 1385 70.9545100 0.9301440  </div><div class="line">   2) x2&lt; 0.9894983 736 31.9699900 0.8017888  </div><div class="line">     4) x3&gt;=0.8423859 551 16.4573000 0.7288982  </div><div class="line">       8) x2&gt;=0.5178963 483 13.3868900 0.7068909  </div><div class="line">        16) x2&lt; 0.7999875 263  4.7064590 0.6422662  </div><div class="line">          32) x3&lt; 1.092533 185  3.3567710 0.6073434 *</div><div class="line">          33) x3&gt;=1.092533 78  0.5889238 0.7250958 *</div><div class="line">        17) x2&gt;=0.7999875 220  6.2689860 0.7841468 *</div><div class="line">       9) x2&lt; 0.5178963 68  1.1749050 0.8852150 *</div><div class="line">     5) x3&lt; 0.8423859 185  3.8660570 1.0188850  </div><div class="line">      10) x4&gt;=1.062353 71  1.8511830 0.9239636 *</div><div class="line">      11) x4&lt; 1.062353 114  0.9767496 1.0780020 *</div><div class="line">   3) x2&gt;=0.9894983 649 13.1077400 1.0757060  </div><div class="line">     6) x3&gt;=1.056185 206  3.5797200 0.9644841  </div><div class="line">      12) x4&gt;=0.9107852 23  0.5746155 0.7823564 *</div><div class="line">      13) x4&lt; 0.9107852 183  2.1462980 0.9873744 *</div><div class="line">     7) x3&lt; 1.056185 443  5.7947760 1.1274250 *</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/hG48FgC33C.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>计算的结果有2个，一个是文本描述的结果，一个是图形结果，这两个结果是等价的，我们可以了解这个决策树的构造过程。现在解释一下这个决策树。</p>
<p>决策树就像一棵倒长的树，有很多分叉，分叉点叫节点(node)，其中1号节点为根节点，也就是上图中最高的那个节点，输出的文本信息则是全部数据的信息，其中在1号节点上，n=1385，偏差(deviance)=70.9545100，在该节点的因变量的均值为0.9301440。</p>
<p>这里的偏差是批在每个节点的偏差，它的定义为：在该节点的数据中的因变量与其均值之差的平方和，即<span class="math inline">\(\sum_{i}(y_{i}-\bar{y})^2\)</span>这里回归树用偏差大小来选择变量，当然也可以用其它诸如残差平方和作为选择变量的准则。</p>
<p>上述的这些信息显示在上图中（不过只显示了均值），然后计算机程度在每个自变量中选择一个数值分割点，使得在该节点的整个数据在这个分割点比在其它分割点分成两部分之后的偏差之和都小，对每个自变量都选择这样的一个分割点，并对不同自变量的分割结果做比较，得到一个使得分割后总偏差最小的变量，这个变量就是该节点的分割变量（或称为拆分变量）。</p>
<p>根据输出的文本信息，在这个案例中，根节点的分割变量为x2，其分割点为0.9894983，x2小于该值的观测值分到左边（节点2），大于该值的则分到右边（节点3）。在每个节点都展示出观测值数目、偏差及因变量均值这三项内容（不过决策树的图形中只有均值），这样数据就分成了两个部分。然后，对每一部分数据，选择分割变量及其分割点的程序重新开始。如此下去，根据一些按照拟合程度和避免决策树无限制长下去的准则，到一定节点就不再分割，这些节点就是终节点（也称为叶节点），也就是文本信息中带有星号(<code>*</code>)的为叶节点。</p>
<h3 id="决策树的使用">决策树的使用</h3>
<p>使用决策树很简单，如果有新的（只知道自变量的）观测值，就可以根据决策树以及各个变量大小，从根节点一直走到一个终节点，那里因变量的均值就是该观测值的因变量预测值。决策树本身受数据影响较大，但许多决策树结果起来就可以构成很好的方法，例如boosting，bagging以及随机森林等组合方法就是以决策树作为基本决策单元构造出来的。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>复杂数据统计方法：基于R的应用.第三版.吴喜之</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/19/BioStatistics/生物统计-17聚类分析笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/19/BioStatistics/生物统计-17聚类分析笔记/" itemprop="url">聚类分析笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-19T12:02:00+08:00">
                2018-08-19
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,852
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  24
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>以前看过StatQuest的聚类分析笔记，可以参考这篇笔记<a href="http://rvdsd.top/2018/06/29/BioStatistics/%E7%94%9F%E7%89%A9%E7%BB%9F%E8%AE%A1-StatQuest%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B017-%E8%81%9A%E7%B1%BB/">《StatQuest学习笔记17——聚类》</a>。</p>
<h2 id="聚类的概率与类型">聚类的概率与类型</h2>
<h3 id="聚类分析法的概念">聚类分析法的概念</h3>
<p>聚类分析法(cluster analysis)是研究“物以类聚”的一种现代统计分析方法，聚类的目的是把分类对象按照一定的规则分成若干类，这些类不是事先给定的，而是根据数据的特征确定的，在同一类中这些对象在某种意义上趋向于彼此相似，而在不同类中对象趋于不相似。</p>
<h3 id="聚类分析法的类型">聚类分析法的类型</h3>
<p>在实际问题中，经常要对一些东西进行分类，例如在古生物研究中，通过挖掘出来的一些骨髓的形状和大小对它们进行科学的分类，在这里，骨髓的形状和大小是我们用来分类的依据，称为指标或变量，用<span class="math inline">\(X_{1},X_{2},\cdots,X_{p}\)</span>表示，其中，p是变量的个数，需要进行分类的骨髓为样品，用1,2,3,…n表示，其中n是样本的个数，聚类分析的数据结果见下表：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/1F0eg5gCJd.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>在聚类分析中，基本的思想是认为所研究的样本或变量存在着程度不同的相似性（亲疏关系）。于是，根据一批样品中的多个观测指标，具体找出一些能够度量样本之间相似程度的统计量，以这些统计量为划分类型的依据，把一些相似程度较大的样本聚合为一类，把另外一些彼此之间相似程度较大的样品又聚合为另一类，关系密切的聚合到一个小的分类单位，关系疏远的聚合到一个大的分类单位，直到把所有的样本都又聚合完毕，把不同类型一一划分出来，形成一个由小到大的分类系统，最后把整个分类系统画成一张聚类图，用它把所有样本间的亲疏关系表示出来。</p>
<p>通常根据分类对象的不同，可以将聚类分为两类：一类是对样品进行分类处理，叫Q型；一类是对变量进行分类处理，叫R型。Q型聚类又叫样品分类，就是对观测对象进行聚类，是根据被观测的对象各种特征进行分类。在生物信息学中，经常使用的是R型，例如有对照组1，2，3三个样本，实验组1，2，3这三个样本，把这6个样本把基因表达谱的数据进行聚类。</p>
<h3 id="聚类统计量">聚类统计量</h3>
<p>聚类分手析的基本原则是将有较大相似性的对象归为同一类，而将差异较大的个体归入不同的类。为了将样本聚类，就需要研究样品之间的关系，一种方法是将每一个样本看做p维空间的一个点，并在空间定义距离，距离较近的点归为一类，距离较远的点属于不同的类。对变量通常计算它们的相似系数，性质越接近的变量，它们的相似系数越接近于1或-1，彼此无关的变量的相似系数越接近于0。可进行聚类的统计量有距离和相似系数。距离有欧氏距离，马氏距离，兰氏距离，相似系数有夹角余弦，相关系数。</p>
<p>对样品进行聚类时，我们将把样品间的“靠近”程度用某种距离来刻画对指标的聚类，往往用某种相似系数来刻画。当选用n个样本，p个指标时，就可以得到一个<span class="math inline">\(n\times p\)</span>的数据矩阵，即<span class="math inline">\(X=(x_{ij})n \times p\)</span>，该距离的元素<span class="math inline">\(x_{ij}\)</span>表示第i个样本的第j个变量值。</p>
<h4 id="距离和相似系数">距离和相似系数</h4>
<p>对样品或变量进行分类时，我们常用距离和相似系数对样品或变量之间的相似性进行度量。距离常用来度量样本之间的相似性，而相似系数常用来度量变量间的相似性。</p>
<p>距离多用于样品的分类，信<span class="math inline">\(d_{ij}\)</span>表示样本<span class="math inline">\(x_{i}\)</span>和<span class="math inline">\(x_{j}\)</span>的距离，一般要求<span class="math inline">\(d_{ij}\)</span>满足以下四个条件，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/fDEh2L8bCJ.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>在聚类分析中并不严格要求定义的距离都满足这四条，一般来说前三条大部分是能满足的，一些不能满足(4)，但是在广义的角度上也称其为距离。</p>
<p>设<span class="math inline">\(x_{ij}(i=1,2,\cdots,n;j=1,2,\cdots,p)\)</span>为第i个样本的第j个指标的观测数据。即若每个样本有p个变量，则每个样本都可以看成p维空间中的一个点，n个样品就是p维空间中的n个点，定义<span class="math inline">\(d_{ij}\)</span>为样本xj和xj的距离，于是得到一个<span class="math inline">\(n\times n\)</span>的距离矩阵，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/hmL6biif2K.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>样品聚类都是基于此距离矩阵进行的，现在我们看一个简单的案例：</p>
<h4 id="距离案例a样本聚类">距离：案例A：样本聚类</h4>
<p>以下列举5个观察值，2个变量数据的平面散点图，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">x1 &lt;- c(<span class="number">5</span>,<span class="number">7</span>,<span class="number">3</span>,<span class="number">6</span>,<span class="number">6</span>)</div><div class="line">x2 &lt;- c(<span class="number">7</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">5</span>,<span class="number">6</span>)</div><div class="line">plot(x1,x2)</div><div class="line">text(x1,x2,labels=c(<span class="number">1</span>:<span class="number">5</span>),adj=-<span class="number">.5</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190703142825.png">

</div>
<p>由于只有2个变量，所以从散点图上就可以直观地将这5个样本分为几类，但当变量较多时，这种方法显著是不行的。为了计算平面上各点之间的距离<span class="math inline">\(d_{ij}\)</span>，在聚类分析中对连续型变量常用的距离有以下3种，如下所示：</p>
<h5 id="明氏距离minkowski">明氏距离(Minkowski):</h5>
<p><span class="math inline">\(d_{i j}(q)=\left[\sum_{k=1}^{p}\left(x_{i k}-x_{j k}\right)^{q}\right]^{\frac{1}{q}}\)</span></p>
<p>当<code>q=1</code>时，<span class="math inline">\(d_{i j}(1)=\sum_{k=1}^{p}\left|x_{i k}-x_{j k}\right|\)</span>，称为绝对值距离(manhattan);</p>
<p>当<code>q=2</code>时，<span class="math inline">\(d_{i j}(2)=\left[\sum_{k=1}^{p}\left(x_{i k}-x_{j k}\right)^{2}\right]^{\frac{1}{2}}\)</span>，称为欧氏距离(euclidean)；</p>
<p>当<span class="math inline">\(q=\infty\)</span>时，<span class="math inline">\(d_{i j}(\infty) \max _{1 \leqslant k \leqslant D}\left|x_{i k}-x_{j k}\right|\)</span>，称为切比雪夫距离(maximun)。</p>
<h5 id="马氏距离mahalanobi">马氏距离(Mahalanobi)：</h5>
<p><span class="math inline">\(d_{i j}(M)=\left(x_{i}-x_{j}\right)^{\prime} \Sigma^{-1}\left(x_{i}-x_{j}\right)\)</span></p>
<p>其中：<span class="math inline">\(x_{i}\)</span>为样本<span class="math inline">\(i\)</span>的p什指标组成的行徽号<span class="math inline">\(\sum\)</span>为协方差阵。</p>
<p>优点：马氏距离既排除了各指标间的相关性干扰，又消除了各指标的量纲。</p>
<p>缺点：样本协方差中阵在聚类过程中不变，这点不合理。</p>
<h5 id="兰氏距离canberra">兰氏距离(Canberra)</h5>
<p><span class="math inline">\(d_{i j}(L W)=\frac{1}{p} \sum_{k=1}^{p} \frac{\left|x_{i k}-x_{j k}\right|}{x_{i k}+x_{j k}} \quad\left(x_{i j}&gt;0\right)\)</span></p>
<p>下面我们使用<code>dist()</code>函数计算一下案例A的欧氏距离与马氏距离，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">dist(x,mdehtod=<span class="string">"euclidean"</span>,diag=<span class="literal">FALSE</span>,upper=<span class="literal">FALSE</span>,p=<span class="number">2</span>)</div><div class="line"><span class="comment"># x是数据矩阵</span></div><div class="line"><span class="comment"># method是计算方法，包括euclidean,maximum,manhattan,canberra,binary,minkowski</span></div><div class="line"><span class="comment"># diag为是否包含对角线元素</span></div><div class="line"><span class="comment"># upper为是否需要上三角</span></div><div class="line"><span class="comment"># p为Minkowski距离的幂次</span></div></pre></td></tr></table></figure>
<p>案例A的计算结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line">&gt; X&lt;-as.matrix(data.frame(x1,x2))</div><div class="line">&gt; dist(X) <span class="comment"># 默认情况下是欧氏距离</span></div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span>                           </div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span>                  </div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span>         </div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span></div><div class="line">&gt; dist(X,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>) <span class="comment"># 默认情况下是欧氏距离</span></div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span>        <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0.000000</span> <span class="number">6.324555</span> <span class="number">5.385165</span> <span class="number">2.236068</span> <span class="number">1.414214</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span> <span class="number">0.000000</span> <span class="number">4.123106</span> <span class="number">4.123106</span> <span class="number">5.099020</span></div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span> <span class="number">0.000000</span> <span class="number">4.242641</span> <span class="number">5.000000</span></div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span> <span class="number">0.000000</span> <span class="number">1.000000</span></div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span> <span class="number">0.000000</span></div><div class="line">&gt; dist(X,method=<span class="string">"manhattan"</span>,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>) <span class="comment">#马氏距离</span></div><div class="line">  <span class="number">1</span> <span class="number">2</span> <span class="number">3</span> <span class="number">4</span> <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0</span> <span class="number">8</span> <span class="number">7</span> <span class="number">3</span> <span class="number">2</span></div><div class="line"><span class="number">2</span> <span class="number">8</span> <span class="number">0</span> <span class="number">5</span> <span class="number">5</span> <span class="number">6</span></div><div class="line"><span class="number">3</span> <span class="number">7</span> <span class="number">5</span> <span class="number">0</span> <span class="number">6</span> <span class="number">7</span></div><div class="line"><span class="number">4</span> <span class="number">3</span> <span class="number">5</span> <span class="number">6</span> <span class="number">0</span> <span class="number">1</span></div><div class="line"><span class="number">5</span> <span class="number">2</span> <span class="number">6</span> <span class="number">7</span> <span class="number">1</span> <span class="number">0</span></div><div class="line">&gt; dist(X,method=<span class="string">"minkowski"</span>,diag=<span class="literal">T</span>,upper=<span class="literal">T</span>,p=<span class="number">2</span>)</div><div class="line">         <span class="number">1</span>        <span class="number">2</span>        <span class="number">3</span>        <span class="number">4</span>        <span class="number">5</span></div><div class="line"><span class="number">1</span> <span class="number">0.000000</span> <span class="number">6.324555</span> <span class="number">5.385165</span> <span class="number">2.236068</span> <span class="number">1.414214</span></div><div class="line"><span class="number">2</span> <span class="number">6.324555</span> <span class="number">0.000000</span> <span class="number">4.123106</span> <span class="number">4.123106</span> <span class="number">5.099020</span></div><div class="line"><span class="number">3</span> <span class="number">5.385165</span> <span class="number">4.123106</span> <span class="number">0.000000</span> <span class="number">4.242641</span> <span class="number">5.000000</span></div><div class="line"><span class="number">4</span> <span class="number">2.236068</span> <span class="number">4.123106</span> <span class="number">4.242641</span> <span class="number">0.000000</span> <span class="number">1.000000</span></div><div class="line"><span class="number">5</span> <span class="number">1.414214</span> <span class="number">5.099020</span> <span class="number">5.000000</span> <span class="number">1.000000</span> <span class="number">0.000000</span></div></pre></td></tr></table></figure>
<h4 id="相似系数">相似系数</h4>
<p>对两个变量之间的相似程度可以用相似系数来刻画，用<span class="math inline">\(C_{ij}\)</span>表示第i个变量与第j个变量之间的相似系数，<span class="math inline">\(C_{ij}\)</span>的绝对值越接近1，表示指标i与指标j的关系越密切，<span class="math inline">\(C_{ij}\)</span>的绝对值越接近于0，表示指标i与指标j的关系越疏远，常用的相似系数有夹角余弦和相关系数，如下所示：</p>
<h5 id="夹角余弦">夹角余弦：</h5>
<p><span class="math display">\[
C_{i j}(1)=\frac{\sum_{k=1}^{n} x_{k i} x_{k j}}{\left[\left(\sum_{k=1}^{n} x_{k i}^{2}\right)\left(\sum_{k=1}^{n} x_{k j}^{2}\right)\right]^{\frac{1}{2}}}
\]</span></p>
<h5 id="相关系数">相关系数：</h5>
<p><span class="math display">\[
C_{i j}(2)=\frac{\sum_{k=1}^{n}\left(x_{k i}-\overline{x}_{i}\right)\left(x_{k i}-\overline{x}_{j}\right)}{\sqrt{\sum_{k=1}^{n}\left(x_{k i}-\overline{x}_{i}\right)^{2} \sum_{k=1}^{n}\left(x_{k j}-\overline{x}_{j}\right)^{2}}}
\]</span></p>
<p>相似系数之间的转换，一般来说，距离越小，两样本之间关系越密切，而相似系数越大，两变量之间的关系越密切，为了聚类分析方便起见，可以用下面的通用公式得到变量间的距离，即<span class="math inline">\(d_{ij}^2=1-C_{ij}^2\)</span>。</p>
<h2 id="系统聚类法">系统聚类法</h2>
<h3 id="系数聚类法的基本思想">系数聚类法的基本思想</h3>
<p>确定了距离和相似系数后就要进行分类。分类有许多种方法，最常用的一种方法是在样品距离的基础上定义类与类之间的距离。首选将n个样品分成n类，每个样品自成成一类，然后每将将具有最小距离的两类合并，合并后重新计算类与类之间的距离，这个过程一直持续到将所有的样品归为一类为止，并把这个过程画成一张聚类图，由聚类图可以方便地进行分类。因为聚类图很像一张系统图，所以这种方法就叫层次聚类法(hierachical clustering method)。层次聚类方法是目前应用最多的一种方法。从上面的分析可以看出，虽然已经给出了计算样品之间的距离，但在实际计算过程中还要定义类与类之间的距离。如何定义类与类之间的距离，也有很多方法，不同的方法就产生了不同的层次聚类，常用的方法有六种：</p>
<ol style="list-style-type: decimal">
<li>最短距离法：类与类之间的距离等于两类最近样品之间的距离；</li>
<li>最长距离法：类与类之间的距离等于两类最远样品之间的距离；</li>
<li>类平均法：类与类之间的距离等于各类元素两两之间平方距离的平均；</li>
<li>重心法：类与类之间的距离定义为对应这两类重心之间的距离。对样品分类来说，每一类的类重心就是该类样品的均值。</li>
<li>中间距离法：最长距离夸大了类间距离，最短距离低估了类间距离，介于两者之间的距离法即为中间距离法，中间距离法则是采用了介于最远和最近之间的距离。</li>
<li>离差平方和法（Ward法）：基于方差分析的思想，如果类分得正确，同样样品之间的离差平方和应当较小，类与类之间的离差平方和应该较大。</li>
</ol>
<h3 id="层次聚类法的计算公式">层次聚类法的计算公式</h3>
<h4 id="最短距离法">最短距离法</h4>
<p>该法使用<span class="math inline">\(D_{k}(p, q)=\min \left\{d_{i j} | i \in G_{p}, j \in G_{q}\right\}\)</span>来刻画<span class="math inline">\(G_p\)</span>与类<span class="math inline">\(G_{1}\)</span>中最临近的两个样本的距离。若类<span class="math inline">\(G_{p}\)</span>与类<span class="math inline">\(G_{q}\)</span>合并为<span class="math inline">\(G_{r}\)</span>，则<span class="math inline">\(G_{r}\)</span>与<span class="math inline">\(G_{s}\)</span>的距离为： <span class="math display">\[
D_{k}(r, s)=\min \left\{D_{k}(p, s), D_{k}(q, s)\right\}
\]</span></p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190703143117.png">

</div>
<h4 id="最长距离法">最长距离法</h4>
<p>该法用<span class="math inline">\(D_{k}(p, q)=\max \left\{d_{i j} | i \in G_{p}, j \in G_{q}\right\}\)</span>来刻画<span class="math inline">\(G_p\)</span>与类<span class="math inline">\(G_{1}\)</span>中最临近的两个样本的距离。</p>
<p>若类<span class="math inline">\(G_{p}\)</span>与类<span class="math inline">\(G_{q}\)</span>合并为<span class="math inline">\(G_{r}\)</span>，则<span class="math inline">\(G_{r}\)</span>与<span class="math inline">\(G_{s}\)</span>的距离为： <span class="math display">\[
D_{k}(r, s)=\max \left\{D_{k}(p, s), D_{k}(q, s)\right\}
\]</span> <img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190703143250.png"></p>
<h4 id="重心法">重心法</h4>
<p>在样本空间中，一个类用它的重心（即该类样本的均值）作代表较为合理，类与类之间的距离就用重心之间的距离来表示。</p>
<p>设样本之间的距离用欧氏距离，若若类<span class="math inline">\(G_{p}\)</span>与类<span class="math inline">\(G_{q}\)</span>合并为<span class="math inline">\(G_{r}\)</span>，它们各有<span class="math inline">\(n_{p}\)</span>、<span class="math inline">\(n_{q}\)</span>、<span class="math inline">\(n_{r}\left(n_{r}=n_{p}+n_{q}\right)\)</span>个样本，它们的重心用<span class="math inline">\(\overline{x}_{P}\)</span>、<span class="math inline">\(\overline{x}_{q}\)</span>和<span class="math inline">\(\overline{x}_{r}\)</span>表示，显然，<span class="math inline">\(\overline{x}_{r}=\frac{1}{n_{r}}\left(n_{p} \overline{x}_{P}+n_{q} \overline{x}_{q}\right)\)</span>，某一类<span class="math inline">\(G_{k}\)</span>重心为<span class="math inline">\(\overline{x}_{k}\)</span>，它与新类<span class="math inline">\(G_{r}\)</span>的距离是： <span class="math display">\[
D^{2}(k, r)=\left(\overline{x}_{k}-\overline{x}_{r}\right)^{\prime}\left(\overline{x}_{k}-\overline{x}_{r}\right)
\]</span> 其递推公式为： <span class="math display">\[
D^{2}(k, r)=\frac{n_{p}}{n_{r}} D^{2}(k, p)+\frac{n_{q}}{n_{r}} D^{2}(k, q)-\frac{n_{p}}{n_{r}} \cdot \frac{n_{q}}{n_{r}} D^{2}(p, q)
\]</span></p>
<h4 id="中间距离法">中间距离法</h4>
<p>该方法是对最短距离法和最长距离法的一个折中，即令类间距离的递推公式为： <span class="math display">\[
(当G_{r}=\left\{G_{p}, G_{q}\right\})\quad D_{k r}^{2}=\frac{1}{2} D_{k p}^{2}+\frac{1}{2} D_{k q}^{2}-\frac{1}{4} D_{p q}^{2}
\]</span></p>
<h4 id="类平均法">类平均法</h4>
<p>将两类之间的距离平方案底为这两类元素两两之间的平均来计算距离，即： <span class="math display">\[
D^{2}(k, r)=\frac{1}{n_{k} n_{r}} \sum_{i \in G_{k}} \sum_{G_{r}} d_{i j}^{2}=\frac{1}{n_{k} n_{r}}\left(\sum_{i \in G_{k} j \in G_{p}} d_{i j}^{2}+\sum_{i \in G_{k} j \in G_{q}} d_{i j}^{2}\right)
\]</span> 其递推公式为： <span class="math display">\[
D^{2}(k, r)=\frac{n_{p}}{n_{r}} D^{2}(k, p)+\frac{n_{q}}{n_{r}} D^{2}(k, q)
\]</span></p>
<h4 id="离差平方和法ward法">离差平方和法(Ward法)</h4>
<p>这种方法是Ward提出来的，因此称为Ward法，该方法的基本思想来自于方差分析，如果分类正确，同样样品的离差平方和应当较小，类与类的离差平方和应当较大。具体做法是先将n个样品各自归为一类，然后每次缩小一类，每缩小一类，离差平方和就要增大，选择使方差增加最小的两类合并，直接所有的样本都归为一类为止。计算过程如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/AHHeFa9cI9.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/0iaFhF3HA3.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>这六种层次聚类法的并类原则和过程完全相同，不同之处在于类与类之间的距离定义。当采用欧氏距离时，Lance和Williams于19678年将这些方法统一成如下的递推公式，即：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/HDddlmkB4d.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180823/df257DGccL.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="层次聚类法的基本步骤">层次聚类法的基本步骤</h4>
<ol style="list-style-type: decimal">
<li>计算n个样本两两的距离<span class="math inline">\(\{d_{ij}\}\)</span>，记作D；</li>
<li>构造n个类，每个类只包含一个样品；</li>
<li>合并距离最近的两类为一个新类；</li>
<li>计算新类与当前各类的距离，若类个数为1，进行第5步，否则回到第3步；</li>
<li>画聚类图；</li>
<li>决定类的个数和类。</li>
</ol>
<h4 id="层次聚类的函数">层次聚类的函数</h4>
<p>R中进行层次聚类的函数是<code>hclust()</code>，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">hclust(d,mehtod=<span class="string">"complete"</span>,<span class="keyword">...</span>)</div><div class="line"><span class="comment"># d为相似矩阵</span></div><div class="line"><span class="comment"># method是层次聚类方法，包括ward,single,complete,average,mcquitty,median,centroid</span></div></pre></td></tr></table></figure>
<p>现在使用案例A中的数据进行层次聚类。开始有5类，即每个样品自成一类，即<span class="math inline">\(\{G_{1},G_{2},G_{3},G_{4},G_{5}\}\)</span>，这5类之间的距离就等于5个样品之间的距离，距离矩阵记为<span class="math inline">\(D_{0}\)</span>，其中最小的元素为<span class="math inline">\(D_{0}(4,5)=1\)</span>，因此将<span class="math inline">\(G_{4}\)</span>和<span class="math inline">\(G_{5}\)</span>聚合并成一个新类，即<span class="math inline">\(G_{6}=\{G_{4},G_{5}\}\)</span>，然后再过计算<span class="math inline">\(G_{6}\)</span>与$G_{1},G_{2},G_{3}#之间的距离。</p>
<p>应用公式<span class="math inline">\(G_{1}(6,i)=min{D_{0}(4,i),D_{0}(5,i)}\)</span>，求其最近相邻的距离，如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/Jf1chaHL76.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>同理可知，其它类与类之间的距离如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/kHk4C3Bm8j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在绘制层次聚类图，如下所示；</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; hc &lt;-hclust(dist(X),<span class="string">"single"</span>)</div><div class="line">&gt; cbind(hc$merge,hc$height)</div><div class="line">     [,<span class="number">1</span>] [,<span class="number">2</span>]     [,<span class="number">3</span>]</div><div class="line">[<span class="number">1</span>,]   -<span class="number">4</span>   -<span class="number">5</span> <span class="number">1.000000</span></div><div class="line">[<span class="number">2</span>,]   -<span class="number">1</span>    <span class="number">1</span> <span class="number">1.414214</span></div><div class="line">[<span class="number">3</span>,]   -<span class="number">2</span>    <span class="number">2</span> <span class="number">4.123106</span></div><div class="line">[<span class="number">4</span>,]   -<span class="number">3</span>    <span class="number">3</span> <span class="number">4.123106</span></div><div class="line">&gt; plot(hc)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/06aFAJl8CA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>ward法（采用欧氏距离），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">hc2 &lt;-hclust(dist(X),<span class="string">"ward"</span>)</div><div class="line">cbind(hc2$merge,hc$height)</div><div class="line">&gt; cbind(hc2$merge,hc$height)</div><div class="line">     [,<span class="number">1</span>] [,<span class="number">2</span>]     [,<span class="number">3</span>]</div><div class="line">[<span class="number">1</span>,]   -<span class="number">4</span>   -<span class="number">5</span> <span class="number">1.000000</span></div><div class="line">[<span class="number">2</span>,]   -<span class="number">1</span>    <span class="number">1</span> <span class="number">1.414214</span></div><div class="line">[<span class="number">3</span>,]   -<span class="number">2</span>   -<span class="number">3</span> <span class="number">4.123106</span></div><div class="line">[<span class="number">4</span>,]    <span class="number">2</span>    <span class="number">3</span> <span class="number">4.123106</span></div><div class="line">plot(hc2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/6lm49eiDef.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h3 id="案例分析">案例分析</h3>
<p>为了研究我们31个省、市、自治区2001年城镇居民生活消费的分布规律，根据调查资料作区域消费类型划分，其中变量如下所示：</p>
<p>X1：人均食品支出；</p>
<p>X2：人均衣着商品支出</p>
<p>X3：人均家庭设备用品及服务支出</p>
<p>X4：人均医疗保健支出</p>
<p>X5：人均交通和通信支出</p>
<p>X6：人均娱乐教育文化服务支出</p>
<p>X7：人均居住支出</p>
<p>X8：人均杂项商品中和服务支出</p>
<p>表格如下所示（完整的数据在github上）：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/JekccFJc4j.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在使用六种聚类方法进行聚类，如下所示：</p>
<p>第1种：欧氏距离，euclidean</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">x_raw&lt;-read.table(<span class="string">"https://raw.githubusercontent.com/20170505a/raw_data/master/wbh_cluster_072.txt"</span>,header=<span class="literal">T</span>,sep=<span class="string">","</span>)<span class="comment"># 完整数据</span></div><div class="line">rownames(x_raw)&lt;-as.vector(x_raw$X)</div><div class="line">x_raw&lt;-x_raw[,-<span class="number">1</span>]</div><div class="line"></div><div class="line">x_dist1&lt;-dist(x_raw,method=<span class="string">"euclidean"</span>)</div><div class="line">hc1 &lt;-hclust(x_dist1,<span class="string">"single"</span>)</div><div class="line">fviz_dend(hc1) <span class="comment">#factoextra包</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/1CGDL9cH1I.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>还可以对分成几类进行颜色区分，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fviz_dend(hc1, k = <span class="number">4</span>, </div><div class="line">          cex = <span class="number">0.5</span>, </div><div class="line">          k_colors = c(<span class="string">"#2E9FDF"</span>, <span class="string">"#00AFBB"</span>, <span class="string">"#E7B800"</span>, <span class="string">"#FC4E07"</span>),</div><div class="line">          color_labels_by_k = <span class="literal">TRUE</span>, </div><div class="line">          rect = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/lHbiGFIeEi.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第二种：欧氏距离，complete</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc2 &lt;-hclust(x_dist1,<span class="string">"complete"</span>)</div><div class="line">fviz_dend(hc2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/46LK4cfLKm.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第三种：欧氏距离,euclidean,median，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc3 &lt;-hclust(x_dist1,<span class="string">"median"</span>)</div><div class="line">fviz_dend(hc3)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/bLH7a7Hk3e.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第四种：欧氏距离，euclidean，average，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc4 &lt;-hclust(x_dist1,<span class="string">"average"</span>)</div><div class="line">fviz_dend(hc4)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/h0dFL4DH7a.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第五种：欧氏距离，centroid，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc5 &lt;-hclust(x_dist1,<span class="string">"centroid"</span>)</div><div class="line">fviz_dend(hc5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/2m74EhedAA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>第六种：欧氏距离，ward，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">hc6 &lt;-hclust(x_dist1,<span class="string">"ward"</span>)</div><div class="line">fviz_dend(hc6)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/fGbleIfG2c.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>划分类别，如下所示：：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fviz_dend(hc6, k = <span class="number">3</span>, </div><div class="line">          cex = <span class="number">0.5</span>, </div><div class="line">          k_colors = c(<span class="string">"#2E9FDF"</span>, <span class="string">"#00AFBB"</span>, <span class="string">"#E7B800"</span>),</div><div class="line">          color_labels_by_k = <span class="literal">TRUE</span>, </div><div class="line">          rect = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/03fegDE3Ig.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="kmeans聚类法">kmeans聚类法</h2>
<h3 id="kmeans聚类法的概念">kmeans聚类法的概念</h3>
<p>层级聚类法需要计算出不同样本或变量的距离，还要在聚类的每一步计算出类间距离，相应的计算量比较大，尤其是当样本量比较大的时候，非常消耗计算机资源。而kmeans是一类快速聚类法。kmeans法（K均值法）的具体算法至少包括以下三个步骤：</p>
<ol style="list-style-type: decimal">
<li>将所有的样本分成k个初始类；</li>
<li>通过欧氏距离将某个样本划入离中心最近的类中，并对获得的样本与失去样品的类重新计算中心坐标</li>
<li>重复步骤2，直到所有的样品都不能再分配为止。</li>
</ol>
<p>kmeans法和层级聚类法一样，都是以距离的远近亲疏为标准进行聚类的。但两者的不同之处在于：层级聚类对没的类数产生一系列的聚类结果，而K均值法只能产生指定类数的聚类结果。具体类数的确定，离不开实践经验的积累。有时候也要借助层级聚类法，以一部分样本为对象进行聚类，其结果为Kmeans法提供参考。</p>
<h3 id="kemans聚类的原理与计算">kemans聚类的原理与计算</h3>
<p>kmeans算法以k为参数，把n个对象分为k个聚类，以使聚类内具有较高的相似度，而聚类间的相似度较低。相似度的计算是根据一个聚类中对象的均值来进行的。kmeans算法的处理流程如下所示：</p>
<ol style="list-style-type: decimal">
<li>随机地选择k个对象，每个对象初始地代表了一个簇的平均值或中心，对剩余的每个对象，根据其与各个聚类中心的距离将其赋给最近的簇；</li>
<li>重新过计算每个簇的平均值作为聚类中心进行聚类。</li>
</ol>
<p>这个过程不断重复，直到准则函数收敛，通常，采用平方误差准则，其定义如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/6mcbdGI5m2.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>其中，E为数据中所有对象与相应聚类中心的均方差之和，p为代表对象空间中的一个点，<span class="math inline">\(m_{i}\)</span>为类<span class="math inline">\(C_{i}\)</span>的均值（p和<span class="math inline">\(m_{i}\)</span>都是多维的）。该式所示聚类的标准旨在使所有获得的聚类有以下特点：</p>
<p>各本身尽可能地紧凑，而各类之间尽可能地分开，kmean迭代图例如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/39JJc44BHe.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>根据聚类中的均值进行聚类划分的kmeans算法如下：</p>
<ol style="list-style-type: decimal">
<li>从n个数据对象中取任意k个对象作为初始簇中心；</li>
<li>循环下述流程3到4，直到每个聚类不再发生变化为止；</li>
<li>根据每个簇中对象的均值（中心对象），计算每个对象与这些中心对象的距离，并根据最小距离重新对相应对象进行划分；</li>
<li>重新计算每个（有变化）簇的均值。</li>
</ol>
<h3 id="kmeans">kmeans()</h3>
<p>R中的kmeans()函数用于进行kmeans聚类，用法为：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">kmeans(x,centers,..)</div><div class="line">其中，x为数据矩阵或数据框</div><div class="line">centers为聚类数或初始聚类中心</div></pre></td></tr></table></figure>
<h3 id="案例分析-1">案例分析</h3>
<p>kmeans算法的R语言实现及模拟。本例模拟正态随机变量<span class="math inline">\(x~N(\mu,\delta^2)\)</span>。</p>
<ol style="list-style-type: decimal">
<li>用R模拟1000个均值为0，标准差为0.3的正态分布随机数据，再把这些随机数转换为10个变量、100个对象的矩阵；</li>
<li>用同样的方法模拟1000个均值为1，标准差为0.3的正态分布随机数，再转化为10个变量、100个对象的矩阵；</li>
<li>把这两个矩阵合并为10个变量、200个样本的数据矩阵；</li>
<li>利用kmeans聚类法将其聚成两类，观察其效果，如下所示：</li>
</ol>
<p>先看一下层级聚类：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/gl6f4DCm0c.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>再看一下kmeans聚类：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(factoextra)</div><div class="line">cl&lt;-kmeans(x_norm,<span class="number">2</span>)</div><div class="line">fviz_cluster(cl,data=x_norm)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180828/Id6A8cAl0I.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>我们通过<code>fviz_nbclust</code>函数查看一下聚成几类比较好，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_norm2&lt;-scale(x_norm)</div><div class="line">fviz_nbclust(x_norm2, kmeans, method = <span class="string">"wss"</span>) + geom_vline(xintercept = <span class="number">4</span>, linetype = <span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/jKBHBmKiJg.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>从图上可以看出，聚成2类比较好，因为1到2变化幅度最大，2之后的变化幅度就非常小了。</p>
<p>可视化二：可以将不同的聚类呈现出来：</p>
<p>使用<code>silhouette</code>函数来计算一下要绘图的轮廓。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">x.res &lt;- kmeans(x_norm, <span class="number">2</span>)</div><div class="line">sil &lt;- silhouette(km.res$cluster, dist(x_norm))</div><div class="line">fviz_silhouette(sil)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/icH4baljaI.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="其它r语言聚类函数">其它R语言聚类函数</h2>
<p>eclust()是另外的一个聚类函数，此函数简化了聚类分析的工作流程，可以用于计算层次聚类和分区聚类，eclust()自动计算最佳聚类簇数。 自动提供Silhouette plot，可以结合ggplot2绘制图形，使用eclust()的K均值聚类如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_eclust &lt;- eclust(x_norm,<span class="string">"hclust"</span>)</div><div class="line">fviz_dend(x_eclust,rect=<span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/Dc1mFb1KEi.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>现在计算一下前面的那个案例，也就是各个省市的聚类，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">x_eclust2 &lt;-eclust(x_raw,<span class="string">"hclust"</span>,k=<span class="number">4</span>)</div><div class="line">fviz_dend(x_eclust2,rect=<span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180824/5cG9k0918F.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>各省市的Kmean聚类图，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"></div><div class="line"></div></pre></td></tr></table></figure>
<h2 id="代码汇总">代码汇总</h2>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><p>多元统计分析及R语言建模.王斌会</p></li>
<li><p><a href="https://blog.csdn.net/Anna_datahummingbird/article/details/79912348" target="_blank" rel="external">R语言聚类分析-kmeans聚类分析实战</a></p></li>
<li><p><a href="https://mp.weixin.qq.com/s/M_rJqcHz6HUlGIHVeN4RkQ" target="_blank" rel="external">R语言学习笔记之聚类分析</a></p></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/18/BioStatistics/生物统计-15判别分析/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/18/BioStatistics/生物统计-15判别分析/" itemprop="url">判别分析笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-18T12:02:00+08:00">
                2018-08-18
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,316
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  5
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>一般的回归是指因变量为定量变量的情况，而logistic回归主要处理因变量为只取两个值的定性变量情况。如果作为因变量的定性变量取多于两个值的时候，就需要使用线性判别分析(linear discriminant analysis)来处理。判别分析根据自变量来预测因变量的类型。</p>
<h2 id="判别分析原理">判别分析原理</h2>
<p>差别分析是利用部分已知类别的观测数据来建立模型，这样的数据称为训练集(training set)。我们这里介绍的线性判别分析所得到的模型是一组线性差别公式；每个公式是自变量的一个线性组合；公式的个数和类别数目一样。有了公式之后，就可以把每个未知类别的自变量观测值代入那些差别公式，算出和类别一样多的值；相应于哪一类的值最大，则把该观测值判为属于该类。</p>
<p>判别分析的原理并不复杂，可以简单地描述如下。</p>
<p>如果作为属性变量的因变量有k个预付（即观测值应该分成k个类），而自变量包含p个变量，那么每一个观测值就是p维空间的一个点。整个训练集的各个点（假定有n个点）就按照已知的类别在p维空间中形成共有n个点的k个点群。那么，对于一个未知类别的点，如果离哪一群近，就可以分到哪一集。关键的细节问题是如何定义“远近”或距离。关键是确定一些准则，使得每一个点群内部的点接近，而使各群之间尽量分开。不同的准则造就了不同的差别或分类方法。这里不讨论这些方法的数学细节，这里仅仅就是线性差别分析一种方法，介绍如何通过R得到结果及结果的解释。这里先把p维空间的点按照最能够分开各类的原则投影到二维空间，再根据二维空间这些点的分布得到判别函数。</p>
<p>如何来确定判别分析模型的效果呢？最基本的就是用一个没有参加建模的已知观测值类别的数据来验证。这样的数据集称为测试集或检验集(test set)。把通过模型预测的这些观测值的类别和它们已知的类别比较，就知道有多少划分错了，由此可以计算出误差来。</p>
<h2 id="案例分析">案例分析</h2>
<p>R自带了一个数据集，即<code>iris</code>。这是一个鸢尾花(iris)的数据，该数据给出了150个鸢尾花的萼片长(sepal length)、萼片宽(sepal width)、花瓣长(petal length)、花瓣宽(petal width)以及这些花的种类(Species)这五个变量。萼片和花瓣的长宽是四个定量变量，而各类为分类变量（有3个值，为setosa、versicolour、virginica）。现在，我们建立一个模型，根据萼片和花瓣这四个度量来把鸢尾花进行分类，这里的鸢尾花一共有3种，每种都有50个观测值。</p>
<p>现在我们在每种鸢尾花中也能耐选择一半，也就是25个作为测试集，另外25个用来建模。三种鸢尾花一起，训练集和测试集各有75个观测。由于选择测试集和训练集的过程是随机的，每次计算的结果也不相同，在R软件中，对于不同的成随机数种子，我们得到的结果也不尽相同。下面是对两个随机数种子得到的两个测试集判别结果，每个格子中的数目是相应的真实类的观测被划分到预测类的频数，如下所示：</p>
<p>随机种子数：set.seed(5)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(MASS)</div><div class="line">set.seed(<span class="number">5</span>)</div><div class="line">samp &lt;- c(sample(<span class="number">1</span>:<span class="number">50</span>,<span class="number">25</span>),sample(<span class="number">51</span>:<span class="number">100</span>,<span class="number">25</span>),sample(<span class="number">101</span>:<span class="number">150</span>,<span class="number">25</span>))</div><div class="line">a &lt;- lda(Species~.,data=iris,shuset=samp)</div><div class="line">summary(predict(a,iris[-samp,])$class)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; summary(predict(a,iris[-samp,])$class)</div><div class="line">    setosa versicolor  virginica </div><div class="line">        25         25         25</div></pre></td></tr></table></figure>
<p>随机种子数：set.seed(4)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(MASS)</div><div class="line">set.seed(<span class="number">4</span>)</div><div class="line">samp &lt;- c(sample(<span class="number">1</span>:<span class="number">50</span>,<span class="number">25</span>),sample(<span class="number">51</span>:<span class="number">100</span>,<span class="number">25</span>),sample(<span class="number">101</span>:<span class="number">150</span>,<span class="number">25</span>))</div><div class="line">a &lt;- lda(Species~.,data=iris,shuset=samp)</div><div class="line">summary(predict(a,iris[-samp,])$class)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">&gt; summary(predict(a,iris[-samp,])$class)</div><div class="line">    setosa versicolor  virginica </div><div class="line">        25         23         27</div></pre></td></tr></table></figure>
<p>由这两个结果可知，在第一个随机种子下（set.seed(5)）下，测试集中的每个类别完全划分正确，在第二个随机种子下（set.seed(4)），错误地划分了4个类别，出错率为4/75。</p>
<p>下图是在建立线性差别函数过程中，把75个训练集的四维空间点投影到二维空间的结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(mda)</div><div class="line">z&lt;-fda(Species~.,data=iris[samp,])$fit$fitted</div><div class="line">plot(z,pch=c(rep(<span class="number">19</span>,<span class="number">25</span>),rep(<span class="number">2</span>,<span class="number">25</span>),rep(<span class="number">8</span>,<span class="number">25</span>)),xlab=<span class="string">""</span>,ylab=<span class="string">""</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180819/HJDbc61Eim.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>可以看也，setosa（用实心圆点标记）离另外两种很远，而versicolor（用空心三角形标记）和virginica（用星号标记）这两类则离得比较近，容易划分错，我们从第二个随机种子（set.seed(4)）中也能看出来，出错的应时这两类。</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/18/Linux/Linux命令收集/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/18/Linux/Linux命令收集/" itemprop="url">Linux命令收集</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-18T12:00:00+08:00">
                2018-08-18
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  136
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="下载">下载</h3>
<p><code>curl</code>，用法如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">curl http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff -o sc.gff</div></pre></td></tr></table></figure>
<p>这个命令是将某个URL的信息下载下来，并通过<code>-o</code>参数，将其输出到<code>sc.gff</code>文件中，其中,<code>-progress</code>是显示进度条，这个是选用的。</p>
<p><code>prefetch</code></p>
<h3 id="查看文件">查看文件</h3>
<p><code>ls</code></p>
<h2 id="文本相关">文本相关</h2>
<h3 id="文本统计">文本统计</h3>
<p><code>wc</code></p>
<h3 id="文本排序">文本排序</h3>
<p><code>sort</code></p>
<h3 id="去重">去重</h3>
<p><code>unique</code></p>
<h3 id="查看文本">查看文本</h3>
<p><code>head</code></p>
<p><code>tail</code></p>
<p><code>less</code></p>
<p><code>more</code></p>
<p><code>cat</code></p>
<h3 id="正则表达式">正则表达式</h3>
<p><code>grep</code></p>
<p><code>awk</code></p>
<p><code>sed</code></p>
<hr>
<h2 id="文件操作">文件操作</h2>
<h3 id="查找文件">查找文件</h3>
<p><code>find</code></p>
<p><code>locate</code></p>
<h3 id="压缩">压缩</h3>
<p><code>unzip</code></p>
<hr>
<h3 id="目录操作">目录操作</h3>
<p>切换目录：<code>cd</code></p>
<p>新建目录：<code>mkdir</code></p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/08/17/Linux/Linux文档的查阅/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/08/17/Linux/Linux文档的查阅/" itemprop="url">Linux文档的查阅</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-08-17T12:00:00+08:00">
                2018-08-17
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  0
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/07/14/BioStatistics/0统计学与数学书籍汇总/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/07/14/BioStatistics/0统计学与数学书籍汇总/" itemprop="url">统计学与数学书籍汇总</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-07-14T12:10:54+08:00">
                2018-07-14
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,420
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>我觉得统计学的知识点很碎，仅看一本书是不太容易学好统计学的，这篇笔记就介绍几本有关统计学的书，这些书是我看过的，也有正在看的，也有看了一部分太难理解放弃的，我大致按照由易到难的顺序列出了几本（当然还有很多优秀的书没看过，就没列出来），这些书可以在京东或淘宝上买到，大部分书也可以通过搜索引擎找到相应下载的地方，考虑到知识产权的问题，这里不提供PDF版的书籍下载地址。</p>
<h2 id="零基础">零基础</h2>
<h4 id="漫画统计学入门">《漫画统计学入门》</h4>
<p>作者: 拉里-戈尼克 / 沃尔科特-史密斯 出版社: 辽宁教育出版社 ，译者: 梁杰，出版年: 2002-1</p>
<p>这本书就是以漫画的形式把统计学原理给讲了一下，非常适合零基础的人入门，没有复杂的公式，可以从直观上克服对统计学的那种畏惧心理，但是如果要深入理解统计学，这本书远远不够。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/205H40fcDI.jpg?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="漫画统计学">《漫画统计学》</h4>
<p>作者: [日] 高桥信 ，出版社: 科学出版社，原作名: マンガでわかる統計学，译者: 陈刚 ，出版年: 2009-8</p>
<p>这本书跟《漫画统计学入门》类似，这本书只讲了统计学的一些基本原理，这本书是一个叫欧姆社系列漫画书中的一本，其余有关统计的书还有《漫画统计学之因子分析》，《漫画统计学之回归分析》，与数学有关的书有《漫画微分方程》《漫画线性代数》。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/jJ5kj241AA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="女士品茶">《女士品茶》</h4>
<p>作者: 萨尔斯伯格 出版社: 中国统计出版社 副标题: 20世纪统计怎样变革了科学 原作名: The Lady Tasting Tea: How Statistics Revolutionized Science in the Twentieth Century 译者: 邱东等 出版年: 2004-11-01</p>
<p>这本书以某位喝茶的英国女士的假设学说为起点，写了数理统计学历史中的一些大牛，例如戈塞物，费歇尔，皮尔逊等一大批牛人的统计学故事，这本书其实不算作是统计学书，更像是统计学的科普书。对于提高统计学方面的兴趣的思维还是比较有好处的。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/fd14HLgh90.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="入门">入门</h2>
<h4 id="白话统计">《白话统计》</h4>
<p>《白话统计》，冯国双，2018，电子工业出版社。</p>
<p>2018年的新书，没有复杂的公式，非常简单，能把统计学原理以最浅显的语言给讲出来，比较适合那些没有太多数学基础的同学，不过要深入理解统计学，这本书还是不够。我觉得这本书更适合作为统计学教材（例如孙振球的《医学统计学》）的辅助用书。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/6I06L1J3KF.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="深入浅出统计学">《深入浅出统计学》</h4>
<p>《深入浅出统计学》，DawnGriffiths. 深入浅出统计学[M]. 电子工业出版社, 2012.</p>
<p>本书涵盖的知识点包括：信息可视化、概率计算、几何分布、二项分布及泊松分布、正态分布、统计抽样、置信区间的构建、假设检验、卡方分布、相关于回归等等， 没有太多数学公式，很好理解。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/3cB4B0BaaG.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="深入浅出数据分析">《深入浅出数据分析》</h4>
<p>作者MichaelMilton. 深入浅出数据分析[M]. 电子工业出版社, 2010.</p>
<p>这本书可以与《深入浅出统计学》配合使用，可以看作是前一本的升级版。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/biI99liiCc.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="行为科学统计">《行为科学统计》</h4>
<p>《行为科学统计》，作者是 [美] F·J·Gravetter / [美] L·B·Wallnau，中文貌似最新的只有第7版，第7版是2008年出版的，有点早了。可以看英文的，英文已经出到第10版了，名字是《Statistics for the Behavioral Sciences》(10th)，第10版是2016年出的。</p>
<p>这本书本来就是为社会学，心理学这些数学基础比较薄弱的学科的学生写的，同样适用于生物学（毕竟数学基础也很薄弱），读起来并不困难。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/C4H4egjCel.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/bD7L8j6Fc1.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h2 id="统计学基础">统计学基础</h2>
<p>这一部分的书可以与后面提到的有关R语言的书配合看。 ####《医学统计学》</p>
<p>国内的医学统计学书都差不多，去京东上挑一本销量比较大的就行，从头到尾看一下，把统计学的知识理一下，不懂的地方可以找其他的统计学书补充一下，再用R把书中的案例给计算一下就行，我平时看的是孙振球的《医学统计学》。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/f7Bkh3d67g.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="统计思维程序员数学之概率统计">《统计思维——程序员数学之概率统计》</h4>
<p>Allen B.Downey著，张建锋，陈钢 译，人民邮电出版社，书不长，很快就可以读完，对几个统计学基础的概念介绍了一下。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/ihfdkGijm1.png?imageslim" alt="mark"></p>
<h4 id="从数据到结论">《从数据到结论》</h4>
<p>作者吴喜之，把一些统计学的概念与原理讲得很清楚，公式不多，比较简单。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/lJjm024lbL.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="从概念到数据分析">《从概念到数据分析》</h4>
<p>2008年高等教育出版社出版的图书，作者是吴喜之 ，里面使用R语言把一些基本的统计学概念给计算了一下，对于理解统计学很有帮助，公式不多，书很薄，很快就能看完，可以与他的《从数据到结论》结合起来看。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/L4fl1BfGgC.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="医学统计学基础与典型错误辨析">《医学统计学基础与典型错误辨析》</h4>
<p>作者胡良平，书比较老，但我觉得比较有价值，作者收集了医学上常见的一些统计学错误，由于医学生（或者说是生物背景的学生）的统计学基础很多并不是很好，因此经常出现一些统计学的错误，这本书就挑了不少这方面的论文，主要是中文论文里的错误，有精力了可以看看。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/AGbm6IKEC4.png?imageslim" alt="mark"></p>
<h4 id="数理统计学教程">《数理统计学教程》</h4>
<p>陈希孺的书，公式比较多，适合对统计学想深入了解的同学，需要一定的数学基础。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/5CKG09eGje.png?imageslim" alt="mark"></p>
<h4 id="统计学习基础数据挖掘推理与预测">《统计学习基础——数据挖掘、推理与预测》</h4>
<p>原书名是《The Elements of Statistical Learning》，副标题是《Data Mining, Inference, and Prediction, Second Edition》，据说是统计学领域的经典之作，不过貌似需要一定的数学基础，这本书并不太好理解。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180920/3lb11FI4k8.png?imageslim" alt="mark"> 英文版的封面如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/6FghfK6hbg.png?imageslim" alt="mark"></p>
<h2 id="统计学与r">统计学与R</h2>
<p>这一部分的书可以与一些统计学的书配合看。</p>
<h4 id="r语言统计入门">《R语言统计入门》</h4>
<p>原名为Introductory Statistics with R second Edition，常用的统计方法与R实现，没多少公式，很好理解。书的前面部分是对R操作的一些简介，后面就是利用R在常规的统计学中的应用。</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/B4IHmb7eF5.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<h4 id="r语言与统计分析">《R语言与统计分析》</h4>
<p>作者汤银才，书中有很多案例，使用R语言把统计学中常见的方法给演示了出来，适合有一些统计学基础的同学，书中的R代码都很短，很好理解，适合那些又想学R语言，又想学统计学的同学，不过书比较老，2008年出的，但统计学的方法基本上变化不大，比前面的那本略难。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/04h5clAH89.png?imageslim" alt="mark"></p>
<h4 id="多元统计分析与r语言建模">《多元统计分析与R语言建模》</h4>
<p>作者王赋会，介绍了常见的一些统计学方法原理与R语言实现。这些统计学方法包括广义线性模型、判别分析、聚类分析、主成分分析、因子分析、对应分析、典型相关分析等，虽然有公式，但我觉得通过查资料，还是能理解的。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/cjKAg78m7C.png?imageslim" alt="mark"></p>
<h4 id="复杂数据统计方法基于r的应用">《复杂数据统计方法——基于R的应用》</h4>
<p>作者吴喜之，书中的数据都是从国外网站上下载的，这本书是2012年出版的，介绍了一些常见的经典统计学方法以及机器学习方法，公式不多。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/FlLcg4AlJm.png?imageslim" alt="mark"></p>
<h2 id="网上的一些资料">网上的一些资料</h2>
<p>这一部分的资料主要是网上的一些不错的资料，有两个。</p>
<h4 id="医学统计学及sas应用之r语言实现">《医学统计学及SAS应用之R语言实现》</h4>
<p>这个资料最初是我从丁香园上看到的，作者不详，只知道是一名留日的医学博士，并非统计学出身，他学了R语言之后，就把《医学统计学及SAS应用》这本书中的案例重新计算了一下，这个资料网上有，能搜到，只有代码与计算结果，如下所示： <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/7DFiLjJ2Ci.png?imageslim" alt="mark"> 如果看这个资料，需要配合看《医学统计学及SAS应用》。 <strong>《医学统计学及SAS应用》</strong>的作者是王炳顺，这本书主要讲的是医学统计学中的一些案例与SAS的计算过程，如果要看这本书，只需要看书中的统计学原理、案例与SAS的计算结果就行，SAS的使用可以不看，可以自己使用R语言把书中的案例计算一下，然后对比一下自己的计算结果与SAS的计算结果，以及计算结果如何解读。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/5DDiECCh5D.png?imageslim" alt="mark"></p>
<h4 id="医学统计分析与r语言">《医学统计分析与R语言》</h4>
<p>作者梁雪枫，光明日报出版社。这本书里面对医学统计学介绍的极为详细，对数学与统计学基础也要求比较高，如果没有相应的基础读起来比较吃力。这本书很奇怪的是，在京东与淘宝上都搜不到，我还是在豆瓣的评论中找到了下载地址，如果想对统计学有进一步了解的同学，可以看看这本书。 <img src="http://ortxpw68f.bkt.clouddn.com/professional/180927/Ie62E0bfab.png?imageslim" alt="mark"></p>
<h2 id="数学相关资源与书籍整理">数学相关资源与书籍整理</h2>
<p>结城浩的系列书籍，包括《数学女孩系列》，《程序员的数学》系列。</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/06/24/BioTools/各种基因名转换/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/06/24/BioTools/各种基因名转换/" itemprop="url">各种基因名笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-06-24T12:00:00+08:00">
                2018-06-24
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,757
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  17
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="基因名">基因名</h2>
<p>不同的数据据中的对于一个基因的描述不同，例如p53，它又叫P53，又叫TP53，或BCC7，或ENSG00000141510或OTTHUMG00000162125。通常来说，对p53的描述使用的是Official Gene Symbol（例如TP53）。</p>
<p>另外，在NCBI数据库中我们查找TP53，会出现下面这些信息：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190303173528.png">

</div>
<p>现在我们从第一行开始解释：</p>
<ol style="list-style-type: decimal">
<li><code>TP53 tumor protein p53 [ Homo sapiens (human) ]</code>：它表示这个基因是TP53肿瘤蛋白53，TP表示的是tumor protein，p表示的是protein，53的意思是这个蛋白的分子量是53kDa（其实当初在跑SDS-PAGE胶时，这个蛋白位于53kDa附近，就命名为p53，其实后来发现，这个蛋白中富含脯氨酸，在SDS-PAGE胶中迁移速度偏慢，它表示出来的分子量要比实际的分子量大，此蛋白的实际分子量是43.7kDa，不过由于命名太久了，也就将错就错了），后面的<code>Homo sapiens (human)</code>是人类的意思，它表示这个基因是人源的。</li>
<li><code>Gene ID</code>：7157，表示的是基因的entrez的系统代号，Entrez 是美国国家生物技术信息中心所提供的在线资源检索器。该资源将GenBank序列与其原始文献出处链接在一起。如果我们在Entrez中直接检索7157，也能直接搜索到TP53。Entrez ID实际上是指的Entrez gene ID，是对应于染色体上一个gene location的。每一个发现的基因都会被编制一个统一的编号，而Entrez ID是指的来自于NCBI旗下的Entrez gene数据库所使用的编号。每个基因的编号具有唯一性，包括不同种属生物间的同源基因编号也不相同，例如同样是TP53基因，人源TP53的Entrez ID为7157，zebrafish源的Entrez ID为30590，chichen源的Entrez ID为396200。EntrezID标记的基因具有唯一性，可以理解为它就是基因的身份证，通过一串数字就能区分每个基因。</li>
<li><code>Official Symbol</code>，这个是我们常用的这个基因的名称，即TP53，这个命名是由HGNC批准的，HGNC的全称是HUGO Gene Nomenclature Committee，其中HUGO的全称是Human Genome Organisation，即国际人类基因组组织，而HGNC则是HUGO基因命名委员会。</li>
<li><code>Official Full Name</code>，官方的全称。</li>
<li><code>Ensembl:ENSG00000141510</code>：ENSEMBL是一个由由英国Sanger研究所Wellcome基金会及欧洲分子生物学实验室所属分部欧洲生物信息学研究所共同协作运营的一个数据库。Ensembl数据库的基因名称后面有的还有个小数点，这个表示基因的版本号，没多少用处，在进行生信分析的时候，可以去掉。</li>
<li>MIM:191170。OMIM 为“0nline Mendelian Inheritance in Man”的简称，即在线《人类孟德尔遗传》，是一个关于人类基因和遗传紊乱的数据库（持续更新）。主要着眼于遗传性的基因疾病，包括文本信息和相关参考信息、序列纪录、图谱和相关其他数据库。其中MIM号就是在OMIM数据库中的基因编号。</li>
<li><code>Vega:OTTHUMG00000162125</code>。脊椎动物基因组注释（Vertebrate Genome Annotation，Vega）数据库已经被设计为浏览来自各种脊椎动物基因组的完结序列的人工注释的一个社区资源。它的核心数据库是基于Ensembl样式的模式，扩展到包含精选-特异的元数据。</li>
</ol>
<p>常用的基因名转换工具：</p>
<ol style="list-style-type: decimal">
<li><p>DAVID：https://david.ncifcrf.gov/</p></li>
<li><p>Absolute Gene ID Conversion Tools：http://bioinformatics.louisville.edu/abid/</p></li>
<li><p>bioDBnet：（https://biodbnet-abcc.ncifcrf.gov/db/db2db.php）</p></li>
<li><p>ENSEMBL biomart：（http://asia.ensembl.org/biomart/martview/bbd9fa2dd0eac95aa098be3514472b0e）</p></li>
</ol>
<h2 id="r包转换基因名">R包转换基因名</h2>
<p>使用R包来转换基因通常有三种方法。</p>
<h3 id="clusterprofiler包的bitr函数">clusterProfiler包的bitr函数</h3>
<p>bitr<code>是</code>clusterProfiler`包中的一个函数，它用于生物ID的转换，用法如下所示:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">bitr(geneID, fromType, toType, OrgDb, drop = <span class="literal">TRUE</span>)</div></pre></td></tr></table></figure>
<p>其中，geneID是输入的基因ID，fromType是输入的ID类型，<code>toType</code>是输出的ID类型，<code>OrgDb</code>注释的db文件，<code>drop</code>剔除NA数据与否。</p>
<p><code>bitr</code>中的各种基因ID参数名称如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; keytypes(org.Mm.eg.db)</div><div class="line"> [<span class="number">1</span>] <span class="string">"ACCNUM"</span>       <span class="string">"ALIAS"</span>        <span class="string">"ENSEMBL"</span>      <span class="string">"ENSEMBLPROT"</span>  <span class="string">"ENSEMBLTRANS"</span></div><div class="line"> [<span class="number">6</span>] <span class="string">"ENTREZID"</span>     <span class="string">"ENZYME"</span>       <span class="string">"EVIDENCE"</span>     <span class="string">"EVIDENCEALL"</span>  <span class="string">"GENENAME"</span>    </div><div class="line">[<span class="number">11</span>] <span class="string">"GO"</span>           <span class="string">"GOALL"</span>        <span class="string">"IPI"</span>          <span class="string">"MGI"</span>          <span class="string">"ONTOLOGY"</span>    </div><div class="line">[<span class="number">16</span>] <span class="string">"ONTOLOGYALL"</span>  <span class="string">"PATH"</span>         <span class="string">"PFAM"</span>         <span class="string">"PMID"</span>         <span class="string">"PROSITE"</span>     </div><div class="line">[<span class="number">21</span>] <span class="string">"REFSEQ"</span>       <span class="string">"SYMBOL"</span>       <span class="string">"UNIGENE"</span>      <span class="string">"UNIPROT"</span></div></pre></td></tr></table></figure>
<p>首先需要加载注释文件，小鼠使用<code>org.Mm.eg.db</code>文件，人类使用<code>org.Hs.eg.db</code>文件。如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">source</span>(<span class="string">"http://bioconductor.org/biocLite.R"</span>)</div><div class="line">options(BioC_mirror=<span class="string">"http://mirrors.ustc.edu.cn/bioc/"</span>)</div><div class="line">biocLite(<span class="string">"clusterProfiler"</span>)</div><div class="line">biocLite(<span class="string">'org.Hs.eg.db'</span>)</div><div class="line">biocLite(<span class="string">'org.Mn.eg.db'</span>)</div><div class="line"><span class="keyword">library</span>(<span class="string">"clusterProfiler"</span>)</div><div class="line">gene.df &lt;- bitr(gene, fromType = <span class="string">"ENTREZID"</span>, <span class="comment">#fromType是指你的数据ID类型是属于哪一类的</span></div><div class="line">                toType = c(<span class="string">"ENSEMBL"</span>, <span class="string">"SYMBOL"</span>), <span class="comment">#toType是指你要转换成哪种ID类型，可以写多种，也可以只写一种</span></div><div class="line">                OrgDb = org.Hs.eg.db)<span class="comment">#Orgdb是指对应的注释包是哪个</span></div><div class="line">head(gene.df)</div></pre></td></tr></table></figure>
<p>上面的代码运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; head(gene.df)</div><div class="line">  ENTREZID         ENSEMBL SYMBOL</div><div class="line"><span class="number">1</span>     <span class="number">4312</span> ENSG00000196611   MMP1</div><div class="line"><span class="number">2</span>     <span class="number">8318</span> ENSG00000093009  CDC45</div><div class="line"><span class="number">3</span>    <span class="number">10874</span> ENSG00000109255    NMU</div><div class="line"><span class="number">4</span>    <span class="number">55143</span> ENSG00000134690  CDCA8</div><div class="line"><span class="number">5</span>    <span class="number">55388</span> ENSG00000065328  MCM10</div><div class="line"><span class="number">6</span>      <span class="number">991</span> ENSG00000117399  CDC20</div></pre></td></tr></table></figure>
<h2 id="使用mget函数">使用mget函数</h2>
<p>如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(AnnotationDbi)</div><div class="line">mySymbols &lt;- mget(gene,</div><div class="line">                  org.Hs.egSYMBOL, <span class="comment">#这个是可以选择的，选择不同，转换的ID类型也不一样</span></div><div class="line">                  ifnotfound=<span class="literal">NA</span>)</div><div class="line">head(mySymbols)</div><div class="line">class(mySymbols)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">&gt; head(mySymbols)</div><div class="line">$`4312`</div><div class="line">[<span class="number">1</span>] <span class="string">"MMP1"</span></div><div class="line"></div><div class="line">$`8318`</div><div class="line">[<span class="number">1</span>] <span class="string">"CDC45"</span></div><div class="line"></div><div class="line">$`10874`</div><div class="line">[<span class="number">1</span>] <span class="string">"NMU"</span></div><div class="line"></div><div class="line">$`55143`</div><div class="line">[<span class="number">1</span>] <span class="string">"CDCA8"</span></div><div class="line"></div><div class="line">$`55388`</div><div class="line">[<span class="number">1</span>] <span class="string">"MCM10"</span></div><div class="line"></div><div class="line">$`991`</div><div class="line">[<span class="number">1</span>] <span class="string">"CDC20"</span></div><div class="line"></div><div class="line">&gt; class(mySymbols)</div><div class="line">[<span class="number">1</span>] <span class="string">"list"</span></div></pre></td></tr></table></figure>
<h3 id="select函数">select()函数</h3>
<p>如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">geneIDselect &lt;-select(org.Hs.eg.db, <span class="comment">#.db是这个芯片数据对应的注释包</span></div><div class="line">                      keys=gene,</div><div class="line">                      columns=c(<span class="string">"SYMBOL"</span>,<span class="string">"ENSEMBL"</span>,<span class="string">"GENENAME"</span>), <span class="comment">#clolumns参数是你要转换的ID类型是什么，这里选择三个。</span></div><div class="line">                      keytype=<span class="string">"ENTREZID"</span> )<span class="comment">#函数里面的keytype与keys参数是对应的，keys是你输入的那些数据，keytype是指这些数据是属于什么类型的数据。 </span></div><div class="line">head(geneIDselect)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">&gt; head(geneIDselect)</div><div class="line">  ENTREZID SYMBOL         ENSEMBL</div><div class="line"><span class="number">1</span>     <span class="number">4312</span>   MMP1 ENSG00000196611</div><div class="line"><span class="number">2</span>     <span class="number">8318</span>  CDC45 ENSG00000093009</div><div class="line"><span class="number">3</span>    <span class="number">10874</span>    NMU ENSG00000109255</div><div class="line"><span class="number">4</span>    <span class="number">55143</span>  CDCA8 ENSG00000134690</div><div class="line"><span class="number">5</span>    <span class="number">55388</span>  MCM10 ENSG00000065328</div><div class="line"><span class="number">6</span>      <span class="number">991</span>  CDC20 ENSG00000117399</div><div class="line">                                                     GENENAME</div><div class="line"><span class="number">1</span>                                   matrix metallopeptidase <span class="number">1</span></div><div class="line"><span class="number">2</span>                                      cell division cycle <span class="number">45</span></div><div class="line"><span class="number">3</span>                                                neuromedin U</div><div class="line"><span class="number">4</span>                            cell division cycle associated <span class="number">8</span></div><div class="line"><span class="number">5</span> minichromosome maintenance <span class="number">10</span> replication initiation factor</div><div class="line"><span class="number">6</span>                                      cell division cycle <span class="number">20</span></div></pre></td></tr></table></figure>
<h2 id="使用biomart包来转换基因名">使用<code>biomaRt</code>包来转换基因名</h2>
<p><code>biomaRt</code>工具包是一个连接<code>bioMart</code>数据库的R语言接口，能够通过这个软件包自由地链接到bioMart数据库，方便得读取该数据库的数据，并进一步进行数据处理，从名称中的<code>Mart</code>这个英文字符就可以知道，这是一个有数据中心类似功能的包（注：Mart在英文中有中心，集市的意思）。</p>
<p>bioMart是一个集成了生物学数据的大型集成数据库，包括Ensemble，Uniprot，NCBI，EBI，TAIR等常用的数据库。具体说来，使用这个工具包可以做以下几个工作：</p>
<ol style="list-style-type: decimal">
<li><p>查找某个基因在染色体上的位置。反之，给定染色体每一区间，返回该区间的基因s；</p></li>
<li><p>通过EntrezGene的ID查找到相关序列的GO注释。反之，给定相关的GO注释，获取相关的EntrezGene的ID；</p></li>
<li><p>通过EntrezGene的ID查找到相关序列的上游100bp序列（可能包含启动子等调控元件）；</p></li>
<li><p>查找人类染色体上每一段区域中已知的SNPs；</p></li>
<li><p>给定一组的序列ID，获得其中具体的序列。</p></li>
</ol>
<h3 id="biomart包中函数"><code>biomaRt</code>包中函数</h3>
<p>先了解几个这个包中的函数。</p>
<h4 id="usemart函数"><code>useMart()</code>函数</h4>
<p><code>useMart()</code>函数是<code>biomaRt</code>包中的函数，其功能在于选择BioMart数据库，如果要查看能使用BioMart的哪个数据库，可以使用<code>listMarts</code>函数，这个函数后面会有介绍。</p>
<p><code>useMart()</code>参数为：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">useMart(biomart, dataset, host=<span class="string">"www.ensembl.org"</span>,</div><div class="line">path=<span class="string">"/biomart/martservice"</span>, port=<span class="number">80</span>, archive=<span class="literal">FALSE</span>, ssl.verifypeer =</div><div class="line"><span class="literal">TRUE</span>, ensemblRedirect = <span class="literal">NULL</span>, version, verbose = <span class="literal">FALSE</span>)</div></pre></td></tr></table></figure>
<ol style="list-style-type: decimal">
<li><code>biomart</code>：BioMart所用的数据库名称；</li>
<li><code>dataset</code>：biomart参数选中的大数据中的小数据库，使用我们先选ENSMBL数据库，再选人的数据库，语句就是如下所示：</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(biomaRt)</div><div class="line"></div><div class="line">ensembl &lt;- useMart(biomart=<span class="string">"ensembl"</span>, dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</div><div class="line"><span class="comment"># Or as the same following command lines:</span></div><div class="line"><span class="comment"># ensembl &lt;- useMart("ensembl")</span></div><div class="line">ensembl</div><div class="line">class(ensembl)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(biomaRt)</div><div class="line">&gt; ensembl &lt;- useMart(biomart=<span class="string">"ensembl"</span>, dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</div><div class="line">&gt; <span class="comment"># Or as the same following command lines:</span></div><div class="line">&gt; <span class="comment"># ensembl &lt;- useMart("ensembl")</span></div><div class="line">&gt; ensembl</div><div class="line">Object of class <span class="string">'Mart'</span>:</div><div class="line">  Using the ENSEMBL_MART_ENSEMBL BioMart database</div><div class="line">  Using the hsapiens_gene_ensembl dataset</div><div class="line">&gt; class(ensembl)</div><div class="line">[<span class="number">1</span>] <span class="string">"Mart"</span></div><div class="line">attr(,<span class="string">"package"</span>)</div><div class="line">[<span class="number">1</span>] <span class="string">"biomaRt"</span></div></pre></td></tr></table></figure>
<p>从中可以发现，<code>ensembl</code>是一个<code>Mart</code>类型的对象。</p>
<h4 id="listmarts函数"><code>listMarts()</code>函数</h4>
<p><code>listMarts</code>：返回BioMart的数据库名称，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; listMarts()</div><div class="line">               biomart               version</div><div class="line"><span class="number">1</span> ENSEMBL_MART_ENSEMBL      Ensembl Genes <span class="number">97</span></div><div class="line"><span class="number">2</span>   ENSEMBL_MART_MOUSE      Mouse strains <span class="number">97</span></div><div class="line"><span class="number">3</span>     ENSEMBL_MART_SNP  Ensembl Variation <span class="number">97</span></div><div class="line"><span class="number">4</span> ENSEMBL_MART_FUNCGEN Ensembl Regulation <span class="number">97</span></div></pre></td></tr></table></figure>
<p>在使用<code>useMart()</code>创建一个查询对象时，可以使用<code>test_gene_id &lt;- useMart(biomart = &quot;ensembl&quot;)</code>这个语句来创建，它就相当于<code>test_gene_id &lt;- useMart(biomart = &quot;ENSEMBL_MART_ENSEMBL&quot;)</code>。</p>
<p>另外，<code>ENSEMBL_MART_MOUSE</code>是不同品系小鼠的数据库。<code>ENSEMBL_MART_SNP</code>这个从字面上应该是SNP数据库，<code>ENSEMBL_MART_FUNCGEN</code>应该是基因调控元件的一些数据库。</p>
<h4 id="listdatasets函数"><code>listDatasets()</code>函数</h4>
<p>当我们使用<code>useMart()</code>函数选择了总数据库（例如ensembl这个大数据为）后，就可以使用<code>listDatasets()</code>函数来查看可用的数据库的子数据库（小鼠的，还是人的），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">test_ensembl &lt;- useMart(<span class="string">"ensembl"</span>)</div><div class="line">list_dataset &lt;- listDatasets(test_ensembl)</div><div class="line">head(list_dataset)</div><div class="line">str(list_dataset)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">&gt; head(list_dataset)</div><div class="line">                       dataset                           description     version</div><div class="line"><span class="number">1</span> abrachyrhynchus_gene_ensembl Pink-footed goose genes (ASM259213v1) ASM259213v1</div><div class="line"><span class="number">2</span>     acalliptera_gene_ensembl      Eastern happy genes (fAstCal1.2)  fAstCal1.2</div><div class="line"><span class="number">3</span>   acarolinensis_gene_ensembl        Anole lizard genes (AnoCar2.0)   AnoCar2.0</div><div class="line"><span class="number">4</span>    acitrinellus_gene_ensembl        Midas cichlid genes (Midas_v5)    Midas_v5</div><div class="line"><span class="number">5</span>        ahaastii_gene_ensembl    Great spotted kiwi genes (aptHaa1)     aptHaa1</div><div class="line"><span class="number">6</span>    amelanoleuca_gene_ensembl                 Panda genes (ailMel1)     ailMel1</div><div class="line">&gt; str(list_dataset)</div><div class="line"><span class="string">'data.frame'</span>:	<span class="number">190</span> obs. of  <span class="number">3</span> variables:</div><div class="line"> $ dataset    : <span class="string">'AsIs'</span> chr  <span class="string">"abrachyrhynchus_gene_ensembl"</span> <span class="string">"acalliptera_gene_ensembl"</span> <span class="string">"acarolinensis_gene_ensembl"</span> <span class="string">"acitrinellus_gene_ensembl"</span> <span class="keyword">...</span></div><div class="line"> $ description: <span class="string">'AsIs'</span> chr  <span class="string">"Pink-footed goose genes (ASM259213v1)"</span> <span class="string">"Eastern happy genes (fAstCal1.2)"</span> <span class="string">"Anole lizard genes (AnoCar2.0)"</span> <span class="string">"Midas cichlid genes (Midas_v5)"</span> <span class="keyword">...</span></div><div class="line"> $ version    : <span class="string">'AsIs'</span> chr  <span class="string">"ASM259213v1"</span> <span class="string">"fAstCal1.2"</span> <span class="string">"AnoCar2.0"</span> <span class="string">"Midas_v5"</span> <span class="keyword">...</span></div></pre></td></tr></table></figure>
<p>从结果中我们可以看出来，使用<code>useMart()</code>函数创建的<code>ensemebl</code>这个变量是一个<code>Mart</code>类型的变量。</p>
<p>而<code>listDatasets(ensembl)</code>返回的结果则是查看在Ensembl这个数据库中，有多少能用的子数据库列表，从结果可以看出，有190个数据库能用，部分表格如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190710185051.png">

</div>
<p>其中第1列是数据库名，第2列是对数据库的描述，第3列是版本信息，现在来查看一下人类与小鼠的有关数据库，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">number_data &lt;- grep(<span class="string">"[Mm]ouse|[Hm]uman"</span>, list_dataset$description)</div><div class="line">list_dataset[number_data,]</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">&gt; list_dataset[number_data,]</div><div class="line">                     dataset                                      description         version</div><div class="line"><span class="number">76</span>     hsapiens_gene_ensembl                         Human genes (GRCh38.p12)      GRCh38.p12</div><div class="line"><span class="number">92</span>      mcaroli_gene_ensembl             Ryukyu mouse genes (CAROLI_EIJ_v1.1) CAROLI_EIJ_v1.1</div><div class="line"><span class="number">102</span>    mmurinus_gene_ensembl                     Mouse Lemur genes (Mmur_3.0)        Mmur_3.0</div><div class="line"><span class="number">103</span>   mmusculus_gene_ensembl                          Mouse genes (GRCm38.p6)       GRCm38.p6</div><div class="line"><span class="number">106</span>     mpahari_gene_ensembl              Shrew mouse genes (PAHARI_EIJ_v1.1) PAHARI_EIJ_v1.1</div><div class="line"><span class="number">107</span> mspicilegus_gene_ensembl                     Steppe mouse genes (MUSP714)         MUSP714</div><div class="line"><span class="number">108</span>    mspretus_gene_ensembl              Algerian mouse genes (SPRET_EiJ_v1)    SPRET_EiJ_v1</div><div class="line"><span class="number">135</span>    pbairdii_gene_ensembl Northern American deer mouse genes (HU_Pman_2.1)     HU_Pman_2.1</div></pre></td></tr></table></figure>
<p>从上面的结果我们可以知道，人类基因数据库只有1个，即<code>hsapiens_gene_ensembl</code>，这也是前面的那个使用<code>useMart</code>来要创建查询语句中的<code>dataset</code>参数，即<code>ensembl= useMart(&quot;ENSEMBL_MART_ENSEMBL&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)</code>。</p>
<p>而小鼠的数据库有7个，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190710191821.png">

</div>
<p>不过，就我们常用的小鼠而言，通常是C57BL/6小鼠，使用的数据库是<code>mmusculus_gene_ensembl</code>这个，也就是版本号为<code>GRCm38.p6</code>的数据库。</p>
<p>也可以使用<code>searchDatasets()</code>函数来检索子数据库的类型，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">searchDatasets(mart = ensembl, pattern = <span class="string">"mouse"</span>)</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; searchDatasets(mart = ensembl, pattern = <span class="string">"mouse"</span>)</div><div class="line">                     dataset                                      description         version</div><div class="line"><span class="number">92</span>      mcaroli_gene_ensembl             Ryukyu mouse genes (CAROLI_EIJ_v1.1) CAROLI_EIJ_v1.1</div><div class="line"><span class="number">106</span>     mpahari_gene_ensembl              Shrew mouse genes (PAHARI_EIJ_v1.1) PAHARI_EIJ_v1.1</div><div class="line"><span class="number">107</span> mspicilegus_gene_ensembl                     Steppe mouse genes (MUSP714)         MUSP714</div><div class="line"><span class="number">108</span>    mspretus_gene_ensembl              Algerian mouse genes (SPRET_EiJ_v1)    SPRET_EiJ_v1</div><div class="line"><span class="number">135</span>    pbairdii_gene_ensembl Northern American deer mouse genes (HU_Pman_2.1)     HU_Pman_2.1</div></pre></td></tr></table></figure>
<p>总之，如果我们要使用人数据库时，在构建一个检索对象时，就如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">ensembl= useMart(<span class="string">"ENSEMBL_MART_ENSEMBL"</span>,dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</div></pre></td></tr></table></figure>
<p>小鼠的话，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">mus = useMart(<span class="string">"ENSEMBL_MART_ENSEMBL"</span>,dataset=<span class="string">"mmusculus_gene_ensembl"</span>)</div></pre></td></tr></table></figure>
<h4 id="参数选择相关函数">参数选择相关函数</h4>
<p>在选中数据库后，就可以进行基因名的转换了，在转换之前，我们可以看一下要输入什么参数，以及能输出什么参数。</p>
<p><code>listFilters()</code>可以得到所有的输入结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; listFilters(ensembl)[<span class="number">1</span>:<span class="number">5</span>,]</div><div class="line">             name              description</div><div class="line"><span class="number">1</span> chromosome_name Chromosome/scaffold name</div><div class="line"><span class="number">2</span>           start                    Start</div><div class="line"><span class="number">3</span>             end                      End</div><div class="line"><span class="number">4</span>      band_start               Band Start</div><div class="line"><span class="number">5</span>        band_end                 Band End</div></pre></td></tr></table></figure>
<p><code>listAttributes</code>获得所有输出结果：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; listAttributes(ensembl)[<span class="number">1</span>:<span class="number">5</span>,]</div><div class="line">                           name                  description         page</div><div class="line"><span class="number">1</span>               ensembl_gene_id               Gene stable ID feature_page</div><div class="line"><span class="number">2</span>       ensembl_gene_id_version       Gene stable ID version feature_page</div><div class="line"><span class="number">3</span>         ensembl_transcript_id         Transcript stable ID feature_page</div><div class="line"><span class="number">4</span> ensembl_transcript_id_version Transcript stable ID version feature_page</div><div class="line"><span class="number">5</span>            ensembl_peptide_id            Protein stable ID feature_page</div></pre></td></tr></table></figure>
<p><code>searchFilters</code>与<code>searchAttributes</code>定向搜索想要结果的关键词：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">&gt; searchFilters(mart = ensembl, <span class="string">'entrez|hgnc'</span>)</div><div class="line">                         name                                                description</div><div class="line"><span class="number">19</span> with_entrezgene_trans_name                      With EntrezGene transcript name ID(s)</div><div class="line"><span class="number">25</span>                  with_hgnc                                     With HGNC Symbol ID(s)</div><div class="line"><span class="number">26</span>       with_hgnc_trans_name                            With HGNC transcript name ID(s)</div><div class="line"><span class="number">37</span>            with_entrezgene                                       With NCBI gene ID(s)</div><div class="line"><span class="number">76</span>      entrezgene_trans_name EntrezGene transcript name ID(s) [e.g. ARMCX5-GPRASP2-<span class="number">211</span>]</div><div class="line"><span class="number">82</span>                    hgnc_id                                 HGNC ID(s) [e.g. HGNC:<span class="number">100</span>]</div><div class="line"><span class="number">83</span>                hgnc_symbol                                 HGNC symbol(s) [e.g. A1BG]</div><div class="line"><span class="number">84</span>            hgnc_trans_name                 HGNC transcript name ID(s) [e.g. A1BG-<span class="number">201</span>]</div><div class="line"><span class="number">97</span>       entrezgene_accession                         NCBI gene accession(s) [e.g. A1BG]</div><div class="line"><span class="number">98</span>              entrezgene_id                                   NCBI gene ID(s) [e.g. <span class="number">1</span>]</div></pre></td></tr></table></figure>
<h4 id="getbm函数"><code>getBM()</code>函数</h4>
<p><code>getBM()</code>函数的参数如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">getBM(attributes, filters = <span class="string">""</span>, values = <span class="string">""</span>, mart, curl = <span class="literal">NULL</span>, </div><div class="line">checkFilters = <span class="literal">TRUE</span>, verbose = <span class="literal">FALSE</span>, uniqueRows = <span class="literal">TRUE</span>, bmHeader = <span class="literal">FALSE</span>,</div><div class="line">quote = <span class="string">"\""</span>)</div></pre></td></tr></table></figure>
<p>参数如下：</p>
<ul>
<li><code>attributes</code>：输出的数据；</li>
<li><code>filters</code>：输入数据的类型，例如<code>hgnc_symbol</code>就是说我们输入的基因名是HGNC协会指定的基因名。</li>
</ul>
<blockquote>
<p>HGNC是指雨果基因命名委员会，它负责制定人类基因命名标准，HGNC的全称是hugo gene nomenclature committee，而其中hugo的全称则是 Human Genome Organisation，即人类基因组组织。</p>
</blockquote>
<ul>
<li><code>values</code>：想要转换的基因名，例如在下面的案例中，我们想转换3个基因的名称，即<code>trans_gene_id &lt;- c(&quot;TP53&quot;, &quot;SMAD4&quot;, &quot;BRCA1&quot;)</code>。</li>
<li><code>mart</code>：检索的数据库。</li>
</ul>
<p>第一个案例：将HGNC基因名转换为<code>Entrez</code>名称，使用案例如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">trans_gene_id &lt;- c(<span class="string">"TP53"</span>, <span class="string">"SMAD4"</span>, <span class="string">"BRCA1"</span>)</div><div class="line">ensembl &lt;- useMart(biomart=<span class="string">"ensembl"</span>,dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</div><div class="line">result_gene_entrez &lt;- getBM(values = trans_gene_id,</div><div class="line">                      filters = <span class="string">'hgnc_symbol'</span>, </div><div class="line">                      mart = ensembl,</div><div class="line">                      attributes = c(<span class="string">'hgnc_symbol'</span>, <span class="string">'entrezgene_id'</span>))</div><div class="line">result_gene_entrez</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; result_gene_entrez</div><div class="line">  hgnc_symbol entrezgene_id</div><div class="line"><span class="number">1</span>       BRCA1           <span class="number">672</span></div><div class="line"><span class="number">2</span>       SMAD4          <span class="number">4089</span></div><div class="line"><span class="number">3</span>        TP53          <span class="number">7157</span></div></pre></td></tr></table></figure>
<p>第二个案例：将HGNC基因名转换为Ensembl名称，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">result_gene_emsembl &lt;- getBM(values = trans_gene_id,</div><div class="line">                      filters = <span class="string">'hgnc_symbol'</span>, </div><div class="line">                      mart = ensembl,</div><div class="line">                      attributes = c(<span class="string">'hgnc_symbol'</span>,</div><div class="line">                                     <span class="string">"ensembl_gene_id"</span>))</div><div class="line">result_gene_emsembl</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; result_gene_emsembl</div><div class="line">  hgnc_symbol ensembl_gene_id</div><div class="line"><span class="number">1</span>       BRCA1 ENSG00000012048</div><div class="line"><span class="number">2</span>       SMAD4 ENSG00000141646</div><div class="line"><span class="number">3</span>        TP53 ENSG00000141510</div></pre></td></tr></table></figure>
<p>如果是想输出多个结果，就在<code>attributes</code>中添加多个参数即可：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">trans_gene_id &lt;- c(<span class="string">"TP53"</span>, <span class="string">"SMAD4"</span>, <span class="string">"BRCA1"</span>)</div><div class="line">result_gene_emsembl2 &lt;- getBM(values = trans_gene_id,</div><div class="line">                      filters = <span class="string">'hgnc_symbol'</span>, </div><div class="line">                      mart = ensembl,</div><div class="line">                      attributes = c(<span class="string">'hgnc_symbol'</span>,</div><div class="line">                                     <span class="string">'entrezgene_id'</span>,</div><div class="line">                                     <span class="string">"ensembl_gene_id"</span>))</div><div class="line">result_gene_emsembl2</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; result_gene_emsembl2</div><div class="line">  hgnc_symbol entrezgene_id ensembl_gene_id</div><div class="line"><span class="number">1</span>       BRCA1           <span class="number">672</span> ENSG00000012048</div><div class="line"><span class="number">2</span>       SMAD4          <span class="number">4089</span> ENSG00000141646</div><div class="line"><span class="number">3</span>        TP53          <span class="number">7157</span> ENSG00000141510</div></pre></td></tr></table></figure>
<p>第三个案例：将HGNC基因名转换为转录本名称，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">trans_gene_id &lt;- c(<span class="string">"OTUD1"</span>, <span class="string">"SMAD4"</span>, <span class="string">"IRAK1"</span>,<span class="string">"EPHA2"</span>,<span class="string">"RAB32"</span>)</div><div class="line">result_gene_transcript &lt;- getBM(values = trans_gene_id,</div><div class="line">                      filters = <span class="string">'hgnc_symbol'</span>, </div><div class="line">                      mart = ensembl,</div><div class="line">                      attributes = c(<span class="string">'hgnc_symbol'</span>,</div><div class="line">                                     <span class="string">'entrezgene_id'</span>,</div><div class="line">                                     <span class="string">"ensembl_gene_id"</span>,</div><div class="line">                                     <span class="string">"ensembl_transcript_id"</span>,</div><div class="line">                                     <span class="string">"refseq_mrna"</span>))</div><div class="line">result_gene_transcript</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div></pre></td><td class="code"><pre><div class="line">&gt; result_gene_transcript</div><div class="line">   hgnc_symbol entrezgene_id ensembl_gene_id ensembl_transcript_id  refseq_mrna</div><div class="line"><span class="number">1</span>        EPHA2          <span class="number">1969</span> ENSG00000142627       ENST00000358432 NM_001329090</div><div class="line"><span class="number">2</span>        EPHA2          <span class="number">1969</span> ENSG00000142627       ENST00000358432    NM_004431</div><div class="line"><span class="number">3</span>        EPHA2          <span class="number">1969</span> ENSG00000142627       ENST00000462805             </div><div class="line"><span class="number">4</span>        EPHA2          <span class="number">1969</span> ENSG00000142627       ENST00000480202             </div><div class="line"><span class="number">5</span>        EPHA2          <span class="number">1969</span> ENSG00000142627       ENST00000461614             </div><div class="line"><span class="number">6</span>        IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000369980    NM_001569</div><div class="line"><span class="number">7</span>        IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000369974 NM_001025243</div><div class="line"><span class="number">8</span>        IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000444230             </div><div class="line"><span class="number">9</span>        IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000477274             </div><div class="line"><span class="number">10</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000455690             </div><div class="line"><span class="number">11</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000369973             </div><div class="line"><span class="number">12</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000444254             </div><div class="line"><span class="number">13</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000437278             </div><div class="line"><span class="number">14</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000393687 NM_001025242</div><div class="line"><span class="number">15</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000429936             </div><div class="line"><span class="number">16</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000443220             </div><div class="line"><span class="number">17</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000467236             </div><div class="line"><span class="number">18</span>       IRAK1          <span class="number">3654</span> ENSG00000184216       ENST00000463031             </div><div class="line"><span class="number">19</span>       OTUD1        <span class="number">220213</span> ENSG00000165312       ENST00000376495 NM_001145373</div><div class="line"><span class="number">20</span>       RAB32         <span class="number">10981</span> ENSG00000118508       ENST00000367495    NM_006834</div><div class="line"><span class="number">21</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000588860             </div><div class="line"><span class="number">22</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000398417             </div><div class="line"><span class="number">23</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000589076             </div><div class="line"><span class="number">24</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000590061             </div><div class="line"><span class="number">25</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000591914             </div><div class="line"><span class="number">26</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000592911             </div><div class="line"><span class="number">27</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000589941             </div><div class="line"><span class="number">28</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000588745             </div><div class="line"><span class="number">29</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000592186             </div><div class="line"><span class="number">30</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000589706             </div><div class="line"><span class="number">31</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000585448             </div><div class="line"><span class="number">32</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000591126             </div><div class="line"><span class="number">33</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000611848             </div><div class="line"><span class="number">34</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000590499             </div><div class="line"><span class="number">35</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000593223             </div><div class="line"><span class="number">36</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000586253             </div><div class="line"><span class="number">37</span>       SMAD4          <span class="number">4089</span> ENSG00000141646       ENST00000342988    NM_005359</div></pre></td></tr></table></figure>
<p>## <code>biomaRt</code>包转换代码总结</p>
<p>总结一下代码：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(biomaRt)</div><div class="line">trans_gene_id &lt;- c(<span class="string">"TP53"</span>, <span class="string">"SMAD4"</span>, <span class="string">"BRCA1"</span>)</div><div class="line">ensembl &lt;- useMart(biomart=<span class="string">"ensembl"</span>,dataset=<span class="string">"hsapiens_gene_ensembl"</span>)</div><div class="line">result_gene_entrez &lt;- getBM(values = trans_gene_id,</div><div class="line">                      filters = <span class="string">'hgnc_symbol'</span>, </div><div class="line">                      mart = ensembl,</div><div class="line">                      attributes = c(<span class="string">'hgnc_symbol'</span>, <span class="string">'entrezgene_id'</span>))</div><div class="line">result_gene_entrez</div></pre></td></tr></table></figure>
<h2 id="在线转换工具">在线转换工具</h2>
<ol style="list-style-type: decimal">
<li>g:profiler：https://biit.cs.ut.ee/gprofiler/gost</li>
</ol>
<p>这个貌似有R包，学习完补充笔记。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://blog.csdn.net/liufeichendi/article/details/78819884" target="_blank" rel="external">各类基因ID转换</a></li>
<li>简述几种Gene ID的转换方法<br>
</li>
<li><a href="https://www.douban.com/note/544090098/" target="_blank" rel="external">OMIM数据库</a></li>
<li><a href="https://vip.biotrainee.com/d/109-entrez-id" target="_blank" rel="external">浅谈Entrez ID</a></li>
<li><a href="https://blog.csdn.net/weixin_40739969/article/details/89354167" target="_blank" rel="external">Gene id 转换(三种基因各种id转换方法)</a></li>
<li><a href="https://mp.weixin.qq.com/s/wsiceQmNVveoggiqeDSlmQ" target="_blank" rel="external">如何让基因名称在多个数据库间随意转换？</a></li>
<li><a href="https://www.jianshu.com/p/f284e5a2069e" target="_blank" rel="external">g:profiler-富集分析/ID转换神器</a></li>
<li><a href="https://blog.csdn.net/jiuyizhizhu/article/details/5649834" target="_blank" rel="external">R语言biomaRt工具包（package）的使用</a></li>
<li><a href="https://www.jianshu.com/p/33093fc4f2ca" target="_blank" rel="external">biomaRt:基因类型转换R包使用</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/06/23/BioStatistics/生物统计-06非参数检验/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/06/23/BioStatistics/生物统计-06非参数检验/" itemprop="url">非参数检验笔记</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-06-23T12:02:00+08:00">
                2018-06-23
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,564
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  24
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>非参数检验(non-parametric test)是相对于参数检验(parametric test)而言的。如果总体分布为已知的数学形式，用参数检验，反之用非参数检验。当总体分布不能由已知的数学形式表达，没有总体参数时，就无法用参数检验，两个或多个正态总体方差不等，也不能用t检验或F检验的参数检验。对于不满足参数检验条件的数据，一是进行变量变换，使其满足参数检验条件，另外就是用非参数检验。非参检验对总体分布不作严格假定，又称任意分布检验(distribution-free test)，《医学统计学》（第三版，孙振球）书中采用的是秩转换的非参数检验，即将数值变量从小到大排列，再计算检验统计量。</p>
<p>非参数检验在总体分布未知时有很大的优越性，在分布未时，如果还假定总体诸如正态分布那样的已知分布，在进行统计推断时就可能产生错误。非参数检验总是比传统检验案例，但是在总体分布形式已经已知时，非参数检验就不如传统方法效率高，这是因为非参数方法利用的利息要少些。往往在传统方法可以拒绝零假设的情况下，非参数检验无法拒绝。用统计的术语来说，在总体分布已知时，传统方法有较大的功效(power)，效率要高；但非参数统计在总体分布未知时，效率往往比假定了错误总体分布时的传统方法要高，有时候要高很多。</p>
<h2 id="秩的概念">秩的概念</h2>
<p>非参数检验中的秩(rank)是最常用的概念，秩就是该数据按照升序排列后，每个观测值的位置。例如我们有下面的数据（样本量为10），其中第一行为数据本身，第二行就是它们的秩，可以看出，3最小，秩为1，13其次，秩为2，而最大的是98，秩为10，我们使用rank(x)表示x的秩，通常也用Ri表示观测值<span class="math inline">\(x_{i}\)</span>的秩，如下所示：</p>
<table>
<thead>
<tr class="header">
<th><span class="math inline">\(x_{i}\)</span></th>
<th>98</th>
<th>67</th>
<th>63</th>
<th>3</th>
<th>38</th>
<th>16</th>
<th>21</th>
<th>53</th>
<th>13</th>
<th>81</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="math inline">\(R_{i}=rank(x_{i})\)</span></td>
<td>10</td>
<td>8</td>
<td>7</td>
<td>1</td>
<td>5</td>
<td>3</td>
<td>4</td>
<td>6</td>
<td>2</td>
<td>9</td>
</tr>
</tbody>
</table>
<h2 id="非参数检验的原理">非参数检验的原理</h2>
<p>如果零假设中确定了一个中位数<span class="math inline">\(M_{0}\)</span>，那么样本点应该以同样的概率出现在<span class="math inline">\(M_{0}\)</span>的两边，也就是说，如果每个样本点送去零假设的中位数，那么得到的差值有些为正，有些为负，如果零假设正确，那么正负符号数目应该差不多，如果差得很多，这就意味着零假设有问题，可以拒绝零假设。这也就是所谓的<strong>符号检验(sign test)</strong>。</p>
<p>这里没有涉及任何总体分布，在零假设下，样本点和<span class="math inline">\(M_{0}\)</span>的差值中负号的个数（记为<span class="math inline">\(S^-\)</span>）应该服从二项分布Bin(n,α)，而正号的个数（记为<span class="math inline">\(S^+\)</span>）䚱服从二项分布Bin(n,1-α)，注意，对于已知对称的分布，由于中位数和均值相等，对中位数的检验等同于对均值的检验，我们通过一个例子来解释符号检验。</p>
<h2 id="单样本的关于总体中位数的符号检验">单样本的关于总体中位数的符号检验</h2>
<p>案例A，质量监督部门对商店里出售的某厂家的西洋参片进行了抽查，对于25包写明为净重100g的西洋参片的称重结果如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180819/EhJ1bCk5HA.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>这里的总体分布未知（我们只是假设未知，其实它的分布是可以计算出来的），用M表示总体中位数，容易计算出，样本中位数为m=98.36，因为人们怀疑厂家包装的西洋参片分量。由于对于重量的总体分布不清楚，决定对其进行符号检验，需要的检验如下所示：</p>
<p><span class="math display">\[H_{0}:M=100\iff H_{1}:M&lt;100\]</span></p>
<p>按照零假设，每个观测值（每包西洋参的净重）大于中位数<span class="math inline">\(M_{0}=100\)</span>的概率和小于100的概率都是0.5，这应该服从二项分布Bin(25,0.5)，容易计算出大于100的只有<span class="math inline">\(S^+=8\)</span>，这样的参数为n=25，p=0.5的二项分布变量小于或等于8的概率为0.05388，这就是p值，历引，对于显著性水平α=0.05，根据这个符号检验，我们没有足够的证据拒绝零假设。在R中计算的过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; pbinom(<span class="number">8</span>,<span class="number">25</span>,<span class="number">0.5</span>)</div><div class="line">[<span class="number">1</span>] <span class="number">0.05387607</span></div><div class="line"><span class="comment"># 或者按另外一种方法：</span></div><div class="line">&gt; raw &lt;-c(<span class="number">99.05</span>,<span class="number">100.25</span>,<span class="number">102.56</span>,<span class="number">99.15</span>,<span class="number">104.89</span>,<span class="number">101.86</span>,<span class="number">96.37</span>,<span class="number">96.79</span>,<span class="number">99.37</span>,</div><div class="line">+         <span class="number">96.9</span>,<span class="number">93.94</span>,<span class="number">92.97</span>,<span class="number">108.28</span>,<span class="number">96.86</span>,<span class="number">93.94</span>,<span class="number">98.27</span>,<span class="number">98.36</span>,<span class="number">100.81</span>,<span class="number">92.99</span>,</div><div class="line">+         <span class="number">103.72</span>,<span class="number">90.66</span>,<span class="number">98.24</span>,<span class="number">97.87</span>,<span class="number">99.21</span>,<span class="number">101.79</span>)</div><div class="line">&gt; pbinom(length(raw[raw&gt;=<span class="number">100</span>]),<span class="number">25</span>,<span class="number">0.5</span>)</div><div class="line">[<span class="number">1</span>] <span class="number">0.05387607</span></div></pre></td></tr></table></figure>
<p>一般来说，根据样本中位数和零假设中的待检验总体中位数<span class="math inline">\(M_{0}\)</span>的比较，我们的检验为下面三种形式之一（记样本中位数为m），如下所示：</p>
<div class="figure">
<img src="http://ortxpw68f.bkt.clouddn.com/professional/180819/3j1H5fkEb9.png?imageslim" alt="mark">
<p class="caption">mark</p>
</div>
<p>这里的关于中位数的符号检验概率，完全可以很容易地推广到关于总体α分位数的符号检验中，如果零假设中确定了一个α分位数<span class="math inline">\(Q_{\alpha0}\)</span></p>
<p>那么，样本点应该以概率α出现在<span class="math inline">\(Q_{\alpha0}\)</span>的左边，或者以概率1-α出现在<span class="math inline">\(Q_{\alpha0}\)</span>的右边，因此样本点减去<span class="math inline">\(Q_{\alpha0}\)</span>差值中铅的个数（记为<span class="math inline">\(S^-\)</span>）应该服从二项分布Bin(n,α)，而正号的个数（记为<span class="math inline">\(S^+\)</span>）应该服从二项分布Bin(n,1-α)。当α=0.5时，总体α分位数就是总体中位数。</p>
<h2 id="单样本的关于对称总体中位数的wilcoxon符号秩检验">单样本的关于对称总体中位数的Wilcoxon符号秩检验</h2>
<p>前面的符号检验仅仅用了样本点和零假设的中位数或α分位数的差的符号，我们可以看到，检验的结果p值大于0.05，但是这种检验方法没有利用这些差的大小，下面介绍的Wilcoxn符号秩检验（Wilcoxon signed-rank test）把差的绝对值的秩分别按照不同的符号相加作为其检验统计量，这里把对总体唯一假定是总体具有对称分布，这种检验方法我们可以看到，它的功效要比前面介绍的高（计算出来的p值小于0.05）。</p>
<h3 id="wilcoxon符号秩检验原理">Wilcoxon符号秩检验原理</h3>
<p>Wilcoxon符号秩检验的原理是这样的，假定<span class="math inline">\(x_{1},x_{2},\cdots,x_{n}\)</span>为来自连续对称总体的一个样本，如果我们的假设检验问题的零假设为中位数（均值）<span class="math inline">\(M=M_{0}\)</span>。那么，对于符号检验而言，只需计算<span class="math inline">\(x_{i}-M_{0}(i=1,2,\cdots,n)\)</span>中有多少正负符号，即可利用二项分布的概率来计算p值。但对于Wilcoxon符号秩检验，则要把<span class="math inline">\(|x_{i}-M_{0}|\)</span>排序，得到<span class="math inline">\(|x_{i}-M_{0}|\)</span>的秩。然后把<span class="math inline">\(x_{i}-M_{0}\)</span>的符号加到相应的秩上面。于是，就可以得到既带有正号的秩，又带有负号的秩，对带负号的秩的绝对值求和，即把满足<span class="math inline">\(x_{i}-M_{0}&lt;0\)</span>的<span class="math inline">\(|x_{i}-M_{0}|\)</span>的秩求和，并用<span class="math inline">\(W^+\)</span>表示。如果<span class="math inline">\(M_{0}\)</span>的确是中位娄和，那么<span class="math inline">\(W^-\)</span>和<span class="math inline">\(W^+\)</span>应该大体上差不多。如果<span class="math inline">\(W^-\)</span>或者<span class="math inline">\(W^+\)</span>过大或过小，则应该怀疑中位数<span class="math inline">\(M=M_{0}\)</span>的零假设。令<span class="math inline">\(W=min(W^-,W^+)\)</span>，则当W太小时，应该拒绝零假设。这个W就是Wilcoxon符号秩检验统计量。</p>
<h2 id="r中的非参数检验函数">R中的非参数检验函数</h2>
<p>在R中进行Wilcoxon检验的函数是<code>wilcoxon.test</code>，使用方法如下所示：</p>
<p>如果总体不服从正态分布，那么T检验就不再适用，此时我们可以利用非参数方法推断中位数。wilcoxon.test函数可实现符号秩检验，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">x &lt;- rnorm(50,mean=10,sd=5)</div><div class="line">wilcox.test(x,conf.int=T) #指定conf.int让函数返回中位数的置信区间</div><div class="line">wilcox.test(x,mu=1) #指定mu让函数返回中位数为10的检验结果</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">&gt; wilcox.test(x,conf.int=T) #指定conf.int让函数返回中位数的置信区间</div><div class="line"></div><div class="line">	Wilcoxon signed rank test with continuity correction</div><div class="line"></div><div class="line">data:  x</div><div class="line">V = 1271, p-value = 9.93e-10</div><div class="line">alternative hypothesis: true location is not equal to 0</div><div class="line">95 percent confidence interval:</div><div class="line">  9.31246 12.20277</div><div class="line">sample estimates:</div><div class="line">(pseudo)median </div><div class="line">       10.7765 </div><div class="line"></div><div class="line">&gt; wilcox.test(x,mu=1) #指定mu让函数返回中位数为10的检验结果</div><div class="line"></div><div class="line">	Wilcoxon signed rank test with continuity correction</div><div class="line"></div><div class="line">data:  x</div><div class="line">V = 1269, p-value = 1.121e-09</div><div class="line">alternative hypothesis: true location is not equal to 1</div></pre></td></tr></table></figure>
<h2 id="单个样本wilcoxon检验案例">单个样本Wilcoxon检验案例</h2>
<h4 id="案例a">案例A</h4>
<p>现在我们再看一下前面的那个案例A，使用Wilcoxon检验来看一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">&gt; raw &lt;-c(<span class="number">99.05</span>,<span class="number">100.25</span>,<span class="number">102.56</span>,<span class="number">99.15</span>,<span class="number">104.89</span>,<span class="number">101.86</span>,<span class="number">96.37</span>,<span class="number">96.79</span>,<span class="number">99.37</span>,</div><div class="line">+         <span class="number">96.9</span>,<span class="number">93.94</span>,<span class="number">92.97</span>,<span class="number">108.28</span>,<span class="number">96.86</span>,<span class="number">93.94</span>,<span class="number">98.27</span>,<span class="number">98.36</span>,<span class="number">100.81</span>,<span class="number">92.99</span>,</div><div class="line">+         <span class="number">103.72</span>,<span class="number">90.66</span>,<span class="number">98.24</span>,<span class="number">97.87</span>,<span class="number">99.21</span>,<span class="number">101.79</span>)</div><div class="line">&gt; wilcox.test(raw,mu=<span class="number">100</span>,alternative = <span class="string">"less"</span>)</div><div class="line"></div><div class="line">	Wilcoxon signed rank test with continuity correction</div><div class="line"></div><div class="line">data:  raw</div><div class="line">V = <span class="number">100</span>, p-value = <span class="number">0.04763</span></div><div class="line">alternative hypothesis: true location is less than <span class="number">100</span></div><div class="line"></div><div class="line">Warning message:</div><div class="line">In wilcox.test.default(raw, mu = <span class="number">100</span>, alternative = <span class="string">"less"</span>) :</div><div class="line">  无法精確計算带连结的p值</div></pre></td></tr></table></figure>
<h4 id="案例b">案例B</h4>
<p>现在再看一个案例B，下面是随机挑选了12袋包装上写的重量是15g的干酵母粉，并进行称重所得到的重量的数据，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">12.39,15.64,14,14.17,14.08,14.16,13.94,14.61,14.7,12.81,14.12,13.37</div></pre></td></tr></table></figure>
<p>由于总体中位数是14.1，而零假设的总体均值是15，因此我们可以怀疑这种袋装酵母粉不够分量，相应的关于中位数的检验如下所示：</p>
<p><span class="math display">\[H_{0}:M=15\iff H_{1}:M&lt;15\]</span></p>
<p>假定总体是对称分布的，因此可以应用Wilcoxon符号秩检验，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">&gt; raw_b &lt;- c(<span class="number">12.39</span>,<span class="number">15.64</span>,<span class="number">14</span>,<span class="number">14.17</span>,<span class="number">14.08</span>,<span class="number">14.16</span>,<span class="number">13.94</span>,<span class="number">14.61</span>,<span class="number">14.7</span>,<span class="number">12.81</span>,<span class="number">14.12</span>,<span class="number">13.37</span>)</div><div class="line">&gt; wilcox.test(raw_b,m=<span class="number">15</span>,alternative = <span class="string">"less"</span>)</div><div class="line"></div><div class="line">	Wilcoxon signed rank test</div><div class="line"></div><div class="line">data:  raw_b</div><div class="line">V = <span class="number">3</span>, p-value = <span class="number">0.001221</span></div><div class="line">alternative hypothesis: true location is less than <span class="number">15</span></div></pre></td></tr></table></figure>
<h4 id="案例c">案例C</h4>
<p>例8-2 已知某地正常人尿氟含量的中位数为45.30μmol/L 。今在该地某厂随机抽取12名工人，测得尿氟含量见表8-2第（1）栏。问该厂工人的尿氟含量是否高于当地正常人的尿氟含量？（《医学统计学》，第三版，孙振球，P134）</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">data82 &lt;- c(44.21,45.30,46.39,49.47,51.05,53.16,53.26,54.37,57.16,67.37,71.05,87.37)</div><div class="line">wilcox.test(data82-45.30)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; wilcox.test(data82-45.30)</div><div class="line"></div><div class="line">	Wilcoxon signed rank test with continuity correction</div><div class="line"></div><div class="line">data:  data82 - 45.3</div><div class="line">V = 65, p-value = 0.005099</div><div class="line">alternative hypothesis: true location is not equal to 0</div></pre></td></tr></table></figure>
<p>这个结果与书上有出入，书中提到，p值是小于0.005，则R的计算出来是0.005099，小于0.05，但SPSS的计算结果与R一致。</p>
<h2 id="两个独立样本wilcoxon检验mann-whitney-u检验">两个独立样本Wilcoxon检验（Mann-Whitney U检验）</h2>
<p>如果是比较两个正态总体均值，那么就是使用t检验（检验均值之差为<span class="math inline">\(\mu_{1}-\mu_{2}\)</span>），如果不知道两组数据是否服从哪种特定的总体分布，我们就要考虑比较两组数据的中位数大小，我们需要的唯一定义就是两个总体的分布有类似的形状，但不要求对称。这里介绍的检验的原理很简单，假定第一个样本有m个观测值<span class="math inline">\(x_{1},\cdots,x_{m}\)</span>，第二个有n个观测值，<span class="math inline">\(y_{1},\cdots,y_{n}\)</span>。把两个样本混合之后把这个m+n个观测值按照大小次序排序，然后记下每个观测值在混合排序下面的秩，之后分别把两个样本所得到的秩相加。记第一个样本观测值的秩的和为<span class="math inline">\(W_{x}\)</span>，而第二个样本秩的和为<span class="math inline">\(W_{Y}\)</span>。这两个值可以互相推算，称为Wilcoxon统计量。该统计量的分布和两个总体分布无关，，由此分布可以得到p值。直观上看，如果<span class="math inline">\(W_{x}\)</span>与<span class="math inline">\(W_{Y}\)</span>之中有一个显著地大（或显著地小），则可以选择拒绝零假设。这个检验注称为Wilcoxon秩和检验(Wilcoxon rank-sum test)，也称为Mann-Whitney检验或Mann-Whitney-Wilcoxon检验。之所以有两个名称，这是因为Wilcoxon统计量和由Mann-Whitney导出的检验统计量等价。Mann-Whitney统计量也是一对：<span class="math inline">\(W_{XY}\)</span>和<span class="math inline">\(W_{YX}\)</span>；定义为两个样本中满足<span class="math inline">\(x_{i}&gt;y_{i}\)</span>的数目，而<span class="math inline">\(W_{XY}\)</span>定义为两个样本中满足<span class="math inline">\(x_{i}&lt;y_{i}\)</span>的数目。检验时，一般使用<span class="math inline">\(W=min(W_{XY},W_{YX})\)</span>作为检验统计量。</p>
<p>案例A：（《医学统计学》，第三版，孙振球，P136）</p>
<p>对10例肺癌病人和12例矽肺0期工人用X光片测量肺门横径右侧距RD值（cm），结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">肺癌：2.78,3.23,4.2,4.87,5.12,6.21,7.18,8.05,8.56,9.6</div><div class="line">矽肺：3.23,3.5,4.04,4.15,4.28,4.34,4.47,4.64,4.75,4.82,4.95,5.1</div></pre></td></tr></table></figure>
<p>问肺癌病人的RD值是否高于矽肺0期工人的RD值？</p>
<p>思路：第一组数据的样本量为10，中位数是5.665，第二组数据的样本量为12，中位数是4.405，这两个中位数看上去有点差异，令这两组数据的总体中位数分别为<span class="math inline">\(M_{1}\)</span>和<span class="math inline">\(M_{2}\)</span>，现在进行检验，如下所示：</p>
<p><span class="math display">\[H_{0}:M_{1}=M_{2}\iff H_{1}:M_{1}&gt;M_{2}\]</span></p>
<p>计算过程如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">lc &lt;- c(2.78,3.23,4.2,4.87,5.12,6.21,7.18,8.05,8.56,9.6)</div><div class="line">si &lt;- c(3.23,3.5,4.04,4.15,4.28,4.34,4.47,4.64,4.75,4.82,4.95,5.1)</div><div class="line">wilcox.test(lc,si,correct=T)</div></pre></td></tr></table></figure>
<p>从结果中我们可以看到，p值等于0.04318，同时也得到中间结果<span class="math inline">\(W_{XY}=86.5\)</span>，因此哦可以说，对于大于p值的显著必不平都可以拒绝零假设，也就是说，这两组数据的中位数有显著性的差异。</p>
<h2 id="配对wilcoxon符号秩检验">配对Wilcoxon符号秩检验</h2>
<p>例8-1：对12份血清分别用原方法（检测时间20分钟）和新方法（检测时间10分钟）测谷-丙转氨酶，结果见表8-1的（2）、（3）栏。问两法所得结果有无差别？（《医学统计学》，第三版，孙振球，P132） <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">old &lt;- c(60,142,195,80,242,220,190,25,198,38,236,95)</div><div class="line">new &lt;- c(76,152,243,82,240,220,205,38,243,44,190,100)</div><div class="line">wilcox.test(old,new,paired = TRUE)</div></pre></td></tr></table></figure></p>
<p>结果如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">&gt; old &lt;- c(60,142,195,80,242,220,190,25,198,38,236,95)</div><div class="line">&gt; new &lt;- c(76,152,243,82,240,220,205,38,243,44,190,100)</div><div class="line">&gt; wilcox.test(old,new,paired = TRUE)</div><div class="line"></div><div class="line">	Wilcoxon signed rank test with continuity correction</div><div class="line"></div><div class="line">data:  old and new</div><div class="line">V = 11.5, p-value = 0.06175</div><div class="line">alternative hypothesis: true location shift is not equal to 0</div><div class="line"></div><div class="line">Warning messages:</div><div class="line">1: In wilcox.test.default(old, new, paired = TRUE) :</div><div class="line">  无法精確計算带连结的p值</div><div class="line">2: In wilcox.test.default(old, new, paired = TRUE) :</div><div class="line">  有0时无法計算精確的p值</div></pre></td></tr></table></figure>
<h2 id="有序的秩检验">有序的秩检验</h2>
<p>例8-4 39名吸烟工人和40名不吸烟工人的碳氧血红蛋白HbCO(%)含量见表8-6。问吸烟工人的HbCO(%)含量是否高于不吸烟工人的HbCO(%)含量？</p>
<h3 id="例8-4第1种方法wilcox.test检验">例8-4第1种方法：Wilcox.test检验：</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">smoke &lt;- c(1,8,16,10,4)</div><div class="line">no.smoke &lt;-c(2,23,11,4,0)</div><div class="line">rank.c &lt;- c(1:5)</div><div class="line">group1 &lt;- rep(rank.c,smoke)</div><div class="line">group2 &lt;- rep(rank.c,no.smoke)</div><div class="line">data84 &lt;- c(group1,group2)</div><div class="line">group.f &lt;-factor(c(rep(1,length(group1)),rep(2,length(group2))))</div><div class="line">wilcox.test(data84~group.f)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">&gt; wilcox.test(data84~group.f)</div><div class="line"></div><div class="line">        Wilcoxon rank sum test with continuity correction</div><div class="line"></div><div class="line">data:  data84 by group.f</div><div class="line">W = <span class="number">1137</span>, p-value = <span class="number">0.0002181</span></div><div class="line">alternative hypothesis: true location shift is not equal to <span class="number">0</span></div><div class="line"></div><div class="line">Warning message:</div><div class="line">In wilcox.test.default(x = c(<span class="number">1L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>, <span class="number">2L</span>,  :</div><div class="line">  无法精確計算带连结的p值</div></pre></td></tr></table></figure>
<h3 id="例8-4第2种方法kruskal检验">例8-4第2种方法：Kruskal检验：</h3>
<p>Kurskal-Wallis检验是Wilcoxon方法（其实是Mann-Whitney检验）用于多于两个样本的时候的升级版。当对两个样本进行比较的时候，Kurskal-Wallis检验与Mann-Whitney检验是等价的。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">kruskal.test(data84~group.f)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; kruskal.test(data84~group.f)</div><div class="line"></div><div class="line">        Kruskal-Wallis rank sum test</div><div class="line"></div><div class="line">data:  data84 by group.f</div><div class="line">Kruskal-Wallis chi-squared = <span class="number">13.707</span>, df = <span class="number">1</span>, p-value = <span class="number">0.0002137</span></div></pre></td></tr></table></figure>
<h3 id="例8-4第2种方法ridit检验">例8-4第2种方法：Ridit检验:</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">data84 &lt;- matrix(c(1,8,16,10,4,2,23,11,4,0),nrow=5)</div><div class="line">dimnames(data84) &lt;- list(cone=c(&quot;very low&quot;,&quot;low&quot;,&quot;median&quot;,&quot;a bit high&quot;,&quot;high&quot;),</div><div class="line">worker=c(&quot;smoker&quot;,&quot;no-smoker&quot;))</div><div class="line">if(!require(&quot;Ridit&quot;))&#123;</div><div class="line">  install.packages(&quot;Ridit&quot;)</div><div class="line">  library(Ridit)</div><div class="line">&#125;</div><div class="line">ridit(data84,2)</div><div class="line">chisq.test(data84)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div></pre></td><td class="code"><pre><div class="line">&gt; ridit(data84,<span class="number">2</span>)</div><div class="line"></div><div class="line">Ridit Analysis:</div><div class="line"></div><div class="line">Group   Label           Mean Ridit</div><div class="line">-----   -----           ----------</div><div class="line"><span class="number">1</span>       smoker          <span class="number">0.6159</span></div><div class="line"><span class="number">2</span>       no-smoker       <span class="number">0.387</span></div><div class="line"></div><div class="line">Reference: Total of all groups </div><div class="line">chi-squared = <span class="number">13.707</span>, df = <span class="number">1</span>, p-value = <span class="number">0.0002137</span></div><div class="line"></div><div class="line">&gt; chisq.test(data84)</div><div class="line"></div><div class="line">        Pearson<span class="string">'s Chi-squared test</span></div><div class="line"><span class="string"></span></div><div class="line"><span class="string">data:  data84</span></div><div class="line"><span class="string">X-squared = 15.079, df = 4, p-value = 0.004541</span></div><div class="line"><span class="string"></span></div><div class="line"><span class="string">Warning message:</span></div><div class="line"><span class="string">In chisq.test(data84) : Chi-squared近似算法有可能不准</span></div></pre></td></tr></table></figure>
<h2 id="例8-5">例8-5</h2>
<p>用三种药物杀灭钉螺，每批用200只活钉螺，用药后清点每批钉螺的死亡数、再计算死亡率（%），结果见表8-9。问三种药物杀灭钉螺的效果有无差别？</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">drug &lt;-rep(c(&quot;甲药&quot;,&quot;乙药&quot;,&quot;丙药&quot;),each=5)</div><div class="line">data &lt;- c(32.5,35.5,40.5,46,49,16,20.5,22.5,29,36,6.5,9.0,12.5,18,24)</div><div class="line">data85 &lt;- data.frame(drug,data)</div><div class="line">kruskal.test(data85$data~data85$drug)</div></pre></td></tr></table></figure>
<h2 id="例8-6">例8-6</h2>
<p>比较小白鼠接种三种不同菌型伤寒杆菌9D、11C和DSC1后存活日数，结果见表8-10。问小白鼠接种三种不同菌型伤寒杆菌的存活日数有无差别？ <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">mice &lt;- as.factor(c(rep(&quot;9D&quot;,10),rep(&quot;11C&quot;,9),rep(&quot;DSC1&quot;,11)))</div><div class="line">data86 &lt;- c(2,2,2,3,4,4,4,5,7,7,5,5,6,6,6,7,8,10,12,3,5,6,6,6,7,7,9,10,11,11)</div><div class="line">data86 &lt;- data.frame(mice,data)</div><div class="line">kruskal.test(data86$data~data86$mice)</div></pre></td></tr></table></figure></p>
<h2 id="例8-7">例8-7</h2>
<p>四种疾病患者痰液内嗜酸性白细胞的检查结果见表8-11。问四种疾病患者痰液内的嗜酸性白细胞有无差别？注：这道例题与《医学统计学及SAS应用》（上海交通大学）的9.11类似 <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">x1 &lt;- c(0,2,9,6)</div><div class="line">x2 &lt;- c(3,5,5,2)</div><div class="line">x3 &lt;- c(5,7,3,2)</div><div class="line">x4 &lt;- c(3,5,3,0)</div><div class="line"> </div><div class="line">freq &lt;- function(x)&#123;</div><div class="line">count &lt;-c()</div><div class="line">for(i in 1:4)&#123;</div><div class="line">count1 &lt;-c(rep(i,x[i]))</div><div class="line">count &lt;- append(count,count1)</div><div class="line">&#125;</div><div class="line">return(count)&#125;</div><div class="line">data87 &lt;-c(freq(x1),freq(x2),freq(x3),freq(x4))</div><div class="line">group &lt;- c(rep(1,sum(x1)),rep(2,sum(x2)),rep(3,sum(x3)),rep(4,sum(x4)))</div><div class="line">kruskal.test(data87~group)</div></pre></td></tr></table></figure></p>
<h2 id="例8-8">例8-8</h2>
<p>对例8-6资料（表8-10）作三个样本间的两两比较，Nemenyi检验。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div></pre></td><td class="code"><pre><div class="line"># 需要安装下列包：</div><div class="line">if(!require(&quot;pgirmess&quot;))&#123;</div><div class="line">  install.packages(&quot;pgirmess&quot;)</div><div class="line">  library(pgirmess)</div><div class="line">&#125;</div><div class="line">if(!require(&quot;coin&quot;))&#123;</div><div class="line">  install.packages(&quot;coin&quot;)</div><div class="line">  library(coin)</div><div class="line">&#125;</div><div class="line">if(!require(&quot;multcomp&quot;))&#123;</div><div class="line">  install.packages(&quot;multcomp&quot;)</div><div class="line">  library(multcomp)</div><div class="line">&#125;</div><div class="line"></div><div class="line">mice &lt;- as.factor(c(rep(&quot;9D&quot;,10),rep(&quot;11C&quot;,9),rep(&quot;DSC1&quot;,11)))</div><div class="line">data &lt;- c(2,2,2,3,4,4,4,5,7,7,5,5,6,6,6,7,8,10,12,3,5,6,6,6,7,7,9,10,11,11)</div><div class="line">data88 &lt;- data.frame(mice,data)</div><div class="line">kruskal.test(data~mice,data=data88)</div><div class="line">kruskalmc(data~mice, data=data88, probs=0.05) # 使用kruskalmc函数做两两比较，但此方法不能给出具体的值</div><div class="line"># 下面构建函数计算具体的p值</div><div class="line">mult &lt;- oneway_test(data ~ mice, data = data88,</div><div class="line">ytrafo = function(data) trafo(data, numeric_trafo = rank),</div><div class="line">xtrafo = function(data) trafo(data, factor_trafo = function(x)</div><div class="line">model.matrix(~x - 1) %*% t(contrMat(table(x), &quot;Tukey&quot;))),</div><div class="line">teststat = &quot;max&quot;, distribution = approximate(B = 90000)) </div><div class="line">pvalue(mult, method = &quot;single-step&quot;) #计算具体的p值</div></pre></td></tr></table></figure>
<h2 id="例8-9">例8-9</h2>
<p>8名受试对象在相同实验条件下分别接受4种不同频率声音的刺激，他们的反应率（%）资料见表8-12。问4种频率声音刺激的反应率是否有差别？ <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">freA &lt;- c(8.40,11.60,9.40,9.80,8.30,8.60,8.90,7.80)</div><div class="line">freB &lt;- c(9.60,12.70,9.10,8.70,8.00,9.80,9.00,8.20)</div><div class="line">freC &lt;- c(9.80,11.80,10.40,9.90,8.60,9.60,10.60,8.50)</div><div class="line">freD &lt;- c(11.70,12.00,9.80,12.00,8.60,10.60,11.40,10.80)</div><div class="line">matrix89 &lt;- matrix(c(freA,freB,freC,freD),nrow=8,dimnames=list(no=1:8,c(&quot;A&quot;,&quot;B&quot;,&quot;C&quot;,&quot;D&quot;)))</div><div class="line">friedman.test(matrix89)</div></pre></td></tr></table></figure></p>
<p>对两个以上样本进行比较的Kruskal-Wallis 秩和检验 kruskal.test() kruskal.test(formula, data, subset, na.action, …) 例子 为了比较属于同一类的四种不同食谱的营养效果，将25只老鼠随机的氛围4组，每组分别为8只，4只，7只和6只，采用食谱甲乙丙丁喂养，假设其他条件均值相同，12周后测得体重增加量如表所示，在alpha=0.05水平上，检验各食谱的营养效果是否有显著差异。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">food&lt;-data.frame(</div><div class="line">x=c(164, 190, 203, 205, 206, 214, 228, 257,</div><div class="line">185, 197, 201, 231,</div><div class="line">187, 212, 215, 220, 248, 265, 281,</div><div class="line">202, 204, 207, 227, 230, 276),</div><div class="line">g=factor(rep(1:4, c(8,4,7,6)))</div><div class="line">)</div><div class="line">kruskal.test(x~g, data=food)</div></pre></td></tr></table></figure>
<p>另一种写法 <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">kruskal.test(food$x, food$g)</div></pre></td></tr></table></figure></p>
<p>或 <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">A&lt;-c(164, 190, 203, 205, 206, 214, 228, 257)</div><div class="line">B&lt;-c(185, 197, 201, 231)</div><div class="line">C&lt;-c(187, 212, 215, 220, 248, 265, 281)</div><div class="line">D&lt;-c(202, 204, 207, 227, 230, 276)</div><div class="line">kruskal.test(list(A,B,C,D))</div></pre></td></tr></table></figure></p>
<p>　　利用R做Nemenyi的检验方法有人介绍过，SPSS里面没有现成的方法，需要编程，这里用R来做一下这个检验，原文是在丁香园引用的，参见<a href="http://www.dxy.cn/bbs/thread/26928446#26928446" target="_blank" rel="external">（R语言）秩和检验后两两比较方法之Nemenyi 检验及其他</a>这篇帖子。　　</p>
<p>原题：比较以下三组之间的数据，看有无差异<br>
第一组：293,409,392,213,409,57,97,244,254,352,168； 第二组：441,538,390,589,244,409,72,168,254,374 ； 第三组：807,833,409,914,380,883,254,993,667 ； 将上面三组数据录入data20140501，分两列，第一列为cd，第2列为group，存为csv格式。</p>
<p>源代码如下： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">datanemenyi &lt;- read.csv(&quot;D:/Tutorial/20151219Statistics/data/rdata/data20140501.csv&quot;) #导入源数据文件  </div><div class="line">str(datanemenyi) #查看数据结构</div><div class="line">datanemenyi$group &lt;- as.factor(datanemenyi$group);datanemenyi #将group转化为因子类型  </div><div class="line">kruskal.test(cd~group, data=datanemenyi)  # Kruskal-Wallis检验，类似于参数检验的ANOVA检验  </div><div class="line"></div><div class="line">library(pgirmess)</div><div class="line">library(coin)  </div><div class="line">library(multcomp) # 安装并加载要用到的包  </div><div class="line">kruskalmc(cd~group, data=datanemenyi, probs=0.05) # 使用kruskalmc函数做两两比较，此法不能给出精确p值  </div><div class="line"></div><div class="line">mult &lt;- oneway_test(cd ~ group, data = datanemenyi,</div><div class="line">   ytrafo = function(data) trafo(data, numeric_trafo = rank),</div><div class="line">  xtrafo = function(data) trafo(data, factor_trafo = function(x)</div><div class="line">  model.matrix(~x - 1) %*% t(contrMat(table(x), &quot;Tukey&quot;))),</div><div class="line">  teststat = &quot;max&quot;, distribution = approximate(B = 90000))</div><div class="line"></div><div class="line">pvalue(mult, method = &quot;single-step&quot;)</div></pre></td></tr></table></figure></p>
<p>这个结果与原代码的运行结果不同</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li>关于Nemenyi检验的方法是参照丁香园的，原贴地址如下，帖子的作者有把写过《医学统计学及SAS应用》之R语言实现这本电子书，下载地址如下:<a href="http://www.dxy.cn/bbs/thread/26928446#26928446" target="_blank" rel="external">Nemenyi检验方法</a></li>
<li><a href="http://d.dxy.cn/detail/6143147" target="_blank" rel="external">《医学统计学及SAS应用》之R语言实现</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/06/13/BioTools/高通量qPCR均一化方法总结/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/06/13/BioTools/高通量qPCR均一化方法总结/" itemprop="url">高通量qPCR均一化方法总结</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-06-13T12:00:00+08:00">
                2018-06-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  685
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  3
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>去年看到一篇讲miRNA的文献，这篇文献的信息如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">Zhang, B. J., et al. (2016). &quot;MicroRNA-23a Curbs Necrosis during Early T Cell Activation by Enforcing Intracellular Reactive Oxygen Species Equilibrium.&quot; Immunity 44(3): 568-581.</div></pre></td></tr></table></figure>
<p>这篇文章是杜克大学的李启靖教授于2016年发表在Immunity上的，这里只讲一下文章前面对miRNA高通量数据的筛选，常规的miRNA数据来源是芯片，测序，但这篇文献里面的miRNA数据表达谱是用RT-qPCR法做的（原理就是反转录部分采用了检测miRNA的加尾法，在qPCR部分使用了Labcyte的Echo550声波移液器与ABI ViiA7高通量qPCR仪)。一共做了300多个miRNA，文献附件中给出了每个样本的每个miRNA的相对表达值，与原始Ct值。在这里重复一下文献中的数据分析。</p>
<p>先看一下文献中最终的数据（只列了前5行），如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210137.png">

</div>
<p>再看一下相应的原始数据：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210208.png">

</div>
<p>从原始数据上大致可以猜出已经扣了内参（文献中提到了使用了4个内参，多内参均一化是采用几何平均数的方法来处理的）。</p>
<p>现在描述一下如何计算。</p>
<p>第一步，均一化原始数据，将原始数据导入R中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">setwd(<span class="string">"D:\netdisk\bioinfo.notes\MicroRNA-23a Curbs Necrosis during Early T Cel"</span>)</div><div class="line">rawdata_miRNA &lt;- read.delim(<span class="string">"clipboard"</span>,header=<span class="literal">T</span>)</div><div class="line">str(rawdata_miRNA)</div><div class="line">raw_array &lt;- as.matrix(rawdata_miRNA[,<span class="number">2</span>:<span class="number">11</span>])</div><div class="line">raw_norm &lt;- t(apply(raw_array,<span class="number">1</span>,scale))</div><div class="line">colnames(raw_norm) &lt;- colnames(rawdata_miRNA)[-<span class="number">1</span>]</div><div class="line">dimnames(raw_norm)[<span class="number">1</span>] &lt;- rawdata_miRNA[<span class="number">1</span>]</div><div class="line">write.csv(raw_norm,<span class="string">"Literature_norm.csv"</span>)</div></pre></td></tr></table></figure>
<p>第二步，打开导出的数据（这里只显示了前5行），如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210416.png">

</div>
<p>将计算结果与文献附件中的计算结果对比，下面是文献附件中的均一化后的结果：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210451.png">

</div>
<p>可以发现，计算结果与文献中的结果基本上是一致的，现在计算每组样本中每个miRNA相对表达量。</p>
<p>第三步，在计算每组样本中每个miRNA相对表达量的同时，要先求相同样本之间的平均数，例如第14天的有2个样本，第5天有2个样本，求它们平均数，有的只有1个样本，不用管（其中1个样本是3只小鼠的混合样本，具体的可以参考文献的Methods部分），计算结果如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210644.png">

</div>
<p>第四步，求相对表达量，直接以上面的数据的相反数为2的指数求幂即可，例如Naïve组的Let-7a是0.2983，则其相对表达量就是2^-0.2983=0.81321，再乘以100，同理，求出所有样本中miRNA的相对表达量，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210722.png">

</div>
<p>再看一下文献中的数据，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190325210743.png">

</div>
<p>可以发现两者的数据是一样的。</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/page/4/"><i class="fa fa-angle-left"></i></a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/4/">4</a><span class="page-number current">5</span><a class="page-number" href="/page/6/">6</a><span class="space">&hellip;</span><a class="page-number" href="/page/18/">18</a><a class="extend next" rel="next" href="/page/6/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">174</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">11</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">51</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>

<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:30px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118	</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共654.3k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
