<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Beyes,Bierachical," />










<meta name="description" content="前言 这一部分是《Data Analysis for the life sciences》的第7章统计模型的第2小节，这一部分的主要内容涉及贝叶斯统计的一些原理。相应的R markdown文档可以参考作者的Github，另外，如果想补充一些贝叶斯的相关知识，可以参考这本书《统计学关我什么事：生活中的极简统计学》，这是一本有关贝叶斯统计的科普书，公式不多。 贝叶斯统计 高通量数据的一个明显特点">
<meta name="keywords" content="Beyes,Bierachical">
<meta property="og:type" content="article">
<meta property="og:title" content="DALS019-统计模型2-贝叶斯分布与层次分析">
<meta property="og:url" content="http://rvdsd.top/2019/08/23/DAL/DALS019_StatisticalModels2_Beyes_Hierarchical/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 这一部分是《Data Analysis for the life sciences》的第7章统计模型的第2小节，这一部分的主要内容涉及贝叶斯统计的一些原理。相应的R markdown文档可以参考作者的Github，另外，如果想补充一些贝叶斯的相关知识，可以参考这本书《统计学关我什么事：生活中的极简统计学》，这是一本有关贝叶斯统计的科普书，公式不多。 贝叶斯统计 高通量数据的一个明显特点">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908153222.jpeg">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908211314.jpeg">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908225643.jpeg">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908230650.jpeg">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908232355.jpeg">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908233029.jpeg">
<meta property="og:updated_time" content="2019-09-08T15:35:20.398Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="DALS019-统计模型2-贝叶斯分布与层次分析">
<meta name="twitter:description" content="前言 这一部分是《Data Analysis for the life sciences》的第7章统计模型的第2小节，这一部分的主要内容涉及贝叶斯统计的一些原理。相应的R markdown文档可以参考作者的Github，另外，如果想补充一些贝叶斯的相关知识，可以参考这本书《统计学关我什么事：生活中的极简统计学》，这是一本有关贝叶斯统计的科普书，公式不多。 贝叶斯统计 高通量数据的一个明显特点">
<meta name="twitter:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908153222.jpeg">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/08/23/DAL/DALS019_StatisticalModels2_Beyes_Hierarchical/"/>





  <title>DALS019-统计模型2-贝叶斯分布与层次分析 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/08/23/DAL/DALS019_StatisticalModels2_Beyes_Hierarchical/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">DALS019-统计模型2-贝叶斯分布与层次分析</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T12:00:00+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物统计/" itemprop="url" rel="index">
                    <span itemprop="name">生物统计</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,055
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  21
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>这一部分是《Data Analysis for the life sciences》的第7章统计模型的第2小节，这一部分的主要内容涉及贝叶斯统计的一些原理。相应的R markdown文档可以参考作者的<a href="https://github.com/genomicsclass/labs/tree/master/modeling/bayes.Rmd" target="_blank" rel="external">Github</a>，另外，如果想补充一些贝叶斯的相关知识，可以参考这本书《统计学关我什么事：生活中的极简统计学》，这是一本有关贝叶斯统计的科普书，公式不多。</p>
<h2 id="贝叶斯统计">贝叶斯统计</h2>
<p>高通量数据的一个明显特点就是，虽然我们最终会报道一些特定的基因，但是我们还会观察到许多相关的结果。例如，我们检测会数千个基因或者是代表了蛋白结合位点的数千个峰图，或者是一些CpGs的甲基化水平。但是，我们这里使用的多数统计推断方法都是独立地处理每个特征值，并且几乎忽略了来自其它特征的数据。在这一部分里，我们将会了解到，如何通过对特征值的联合建模来进行统计。这里方法中使用最为广泛的就是层次模型(hierachical models)，我们会在后面的贝叶斯统计中进行解释。</p>
<h3 id="贝叶斯定理">贝叶斯定理</h3>
<p>先来看一个案例，如果我们有一种检测手段来检测囊性纤维化(cystic fibrosis)。假设这个检测手段的精确程度为99%，我们可以使用下面的公式来表示： <span class="math display">\[
\mbox{Prob}(+ \mid D=1)=0.99, \mbox{Prob}(- \mid D=0)=0.99
\]</span> 其中，<span class="math inline">\(+\)</span>表示阳性结果，<span class="math inline">\(D\)</span>表示检测的结果，其中<span class="math inline">\(1\)</span>表示得病，，<span class="math inline">\(0\)</span>表示不得病。</p>
<p>现在我们随机选择一个人进行检测，结果如果是阳性，那么这个人患病的概率是多大？也就是说要计算<span class="math inline">\(\mbox{Prb}(D=1|+)\)</span>的结果。囊性纤维化的发病率是每1/3900，也就是说，<span class="math inline">\(\mbox{Prob}(D=1|+)=0.0025\)</span>，为了计算出这个人患病的概率，我们就会使用到贝叶斯定理，贝叶斯定理公式如下所示： <span class="math display">\[
\mbox{Pr}(A \mid B)  =  \frac{\mbox{Pr}(B \mid A)\mbox{Pr}(A)}{\mbox{Pr}(B)}
\]</span> 这个公式就可以应用到我们的案例中，如下所示： <span class="math display">\[
\begin{align*}
\mbox{Prob}(D=1 \mid +) &amp; =  \frac{ P(+ \mid D=1) \cdot P(D=1)} {\mbox{Prob}(+)} \\
&amp; =  \frac{\mbox{Prob}(+ \mid D=1)\cdot P(D=1)} {\mbox{Prob}(+ \mid D=1) \cdot P(D=1) + \mbox{Prob}(+ \mid D=0) \mbox{Prob}( D=0)} 
\end{align*}
\]</span> 换成实际数字，则如下所示： <span class="math display">\[
\frac{0.99 \cdot 0.00025}{0.99 \cdot 0.00025 + 0.01 \cdot (.99975)}  =  0.02
\]</span> 也就是说，虽然这种检测手段有99%的精度，但是一个人的检测结果如果是阳性，那么这个人得病的概率只有0.02。这似乎有点反直觉。其原因就是，我们必须要考虑，当我们随机选择一个人时，这个人患上这种疾病时非常罕见的可能性。为了说明这一个，我们随便使用Monte Carlo模拟来计算一下。</p>
<h3 id="模拟">模拟</h3>
<p>下面的模拟旨在帮助你能够以可视化的形式来理解贝叶斯定理。我们首选从一个总体中随机选择1500人，其中患病的概率是5%，代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">set.seed(<span class="number">3</span>)</div><div class="line">prev &lt;- <span class="number">1</span>/<span class="number">20</span></div><div class="line"><span class="comment">##Later, we are arranging 1000 people in 80 rows and 20 columns</span></div><div class="line">M &lt;- <span class="number">50</span> ; N &lt;- <span class="number">30</span></div><div class="line"><span class="comment">##do they have the disease?</span></div><div class="line">d&lt;-rbinom(N*M,<span class="number">1</span>,p=prev)</div></pre></td></tr></table></figure>
<p>现在进行一项检测，这个检测的准确率是90%，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">accuracy &lt;- <span class="number">0.9</span></div><div class="line">test &lt;- rep(<span class="literal">NA</span>,N*M)</div><div class="line"><span class="comment">##do controls test positive?</span></div><div class="line">test[d==<span class="number">1</span>] &lt;- rbinom(sum(d==<span class="number">1</span>), <span class="number">1</span>, p=accuracy)</div><div class="line"><span class="comment">##do cases test positive?</span></div><div class="line">test[d==<span class="number">0</span>] &lt;- rbinom(sum(d==<span class="number">0</span>), <span class="number">1</span>, p=<span class="number">1</span>-accuracy)</div></pre></td></tr></table></figure>
<p>由于没有患病的人数要远远超过患病的人数，即使存在着极低的假阳性率，那么在检测为阳性的结果中，不患病人的也要多于患病的人，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">cols &lt;- c(<span class="string">"grey"</span>,<span class="string">"red"</span>)</div><div class="line">people &lt;- expand.grid(<span class="number">1</span>:M,N:<span class="number">1</span>)</div><div class="line">allcols &lt;- cols[d+<span class="number">1</span>] <span class="comment">##Cases will be red</span></div><div class="line">positivecols &lt;- allcols</div><div class="line">positivecols[test==<span class="number">0</span>] &lt;- <span class="literal">NA</span> <span class="comment">##remove non-positives</span></div><div class="line">negativecols &lt;- allcols</div><div class="line">negativecols[test==<span class="number">1</span>] &lt;- <span class="literal">NA</span> <span class="comment">##remove non-positives</span></div><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar()</div><div class="line">layout(matrix(c(<span class="number">1</span>,<span class="number">2</span>,<span class="number">1</span>,<span class="number">3</span>),<span class="number">2</span>,<span class="number">2</span>),width=c(<span class="number">0.35</span>,<span class="number">0.65</span>))</div><div class="line"><span class="comment">###plot of all people</span></div><div class="line">plot(people,col=allcols,pch=<span class="number">16</span>,xaxt=<span class="string">"n"</span>,yaxt=<span class="string">"n"</span>,xlab=<span class="string">""</span>,ylab=<span class="string">""</span>,</div><div class="line">     main=paste0(<span class="string">"Population: "</span>,round(mean(d)*<span class="number">100</span>),<span class="string">"% are red"</span>))</div><div class="line">plot(people,col=positivecols,pch=<span class="number">16</span>,xaxt=<span class="string">"n"</span>,yaxt=<span class="string">"n"</span>,xlab=<span class="string">""</span>,ylab=<span class="string">""</span>,</div><div class="line">     main=paste(<span class="string">"Tested Positive:"</span>,round(mean(d[test==<span class="number">1</span>])*<span class="number">100</span>),<span class="string">"% are red"</span>))</div><div class="line">plot(people,col=negativecols,pch=<span class="number">16</span>,xaxt=<span class="string">"n"</span>,yaxt=<span class="string">"n"</span>,xlab=<span class="string">""</span>,ylab=<span class="string">""</span>,</div><div class="line">     main=paste(<span class="string">"Tested Negative:"</span>,round(mean(d[test==<span class="number">0</span>])*<span class="number">100</span>,<span class="number">1</span>),<span class="string">"% are red"</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908153222.jpeg">

</div>
<p>现在解释一下上面的图：</p>
<p>顶图：红点表示患者。每个人接受检测的话，有90%的可能性结果是准确的。即<span class="math inline">\(\mbox{Pr}(D=1)\)</span>；</p>
<p>下左图：检测结果为阳性（无论结果对不对，这里面含有真阳性与假阳性）的人，即<span class="math inline">\(\mbox{Pr}(D=1|+)\)</span></p>
<p>下右图：检测结果为阴性的人，即<span class="math inline">\(\mbox{Pr}(D=0|+)\)</span></p>
<h3 id="贝叶斯的实际运用">贝叶斯的实际运用</h3>
<p>José Iglesias是一名职业棒球运行员，在2013年4月份，他开始了职业生成，他表表现得很好，成绩如下所示：</p>
<table>
<thead>
<tr class="header">
<th>Month</th>
<th>At Bats</th>
<th>H</th>
<th>AVG</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>April</td>
<td>20</td>
<td>9</td>
<td>.450</td>
</tr>
</tbody>
</table>
<p>平均击球率(battingaverage, AVG)统计是一种检测成功的方式。粗略地说，这个指标是告诉我们击球时的成功情况。<code>AVG=0.450</code>就表明，这个运动员在他击球的时间内（对应上面的<code>At Bats</code>）成功了45%，这是一个非常高的水平。为什么水平高呢，因为自1941年Ted Williams的<code>AVG=0.400</code>以来，还没有人能超过这个水平。为了说明层次模型的强大功能，我们将会在后面对Jose的数据进行预测。</p>
<p>在本书的前面部分到此为止，我们所到到的统计学技术可以被称为频率学派技术(frequentist techniques)，使用这种知识做出的结论就是能计算出一个置信区间。我们可以把击打(hitting)这个事件看作是一个二元结果，其成功的概论为<span class="math inline">\(p\)</span>。因此，如果成功率是0.450的话，那么击球次数是20次时，标准误为： <span class="math display">\[
\sqrt{\frac{.450 (1-.450)}{20}}=.111
\]</span> 也就是说，我们计算出的置信区间为<code>.450-.222</code> to <code>.450+.222</code>或<code>.228</code> to <code>.672</code>。</p>
<p>使用这种手段进行预测存在着两个问题。第一，实际用处不大。第二，成功率在0.450之间波动，也就是说这个人打破了Ted William的纪录。如果你自己关注棒球运动的话，打破Ted William纪录这种描述似乎是有问题，这是因为你隐含地使用了一个层次模型，这个模型会影响后面几年棒球的信息。在这里，我们对这种直觉进行量化。</p>
<p>首选，我们来研究一下前三个赛季里所有超过500次击打(at bats)的运动员的击球率的分布情况：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">tmpfile &lt;- tempfile()</div><div class="line">tmpdir &lt;- tempdir()</div><div class="line">download.file(<span class="string">"http://seanlahman.com/files/database/lahman-csv_2014-02-14.zip"</span>,tmpfile)</div><div class="line"><span class="comment">##this shows us files</span></div><div class="line">filenames &lt;- unzip(tmpfile,list=<span class="literal">TRUE</span>)</div><div class="line">players &lt;- read.csv(unzip(tmpfile,files=<span class="string">"Batting.csv"</span>,exdir=tmpdir),as.is=<span class="literal">TRUE</span>)</div><div class="line">unlink(tmpdir)</div><div class="line">file.remove(tmpfile)</div><div class="line"></div><div class="line"><span class="keyword">library</span>(dplyr)</div><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar(<span class="number">1</span>,<span class="number">3</span>)</div><div class="line"><span class="keyword">for</span>(y <span class="keyword">in</span> <span class="number">2010</span>:<span class="number">2012</span>)&#123;</div><div class="line">  dat &lt;- filter(players,yearID==y) %&gt;% mutate(AVG=H/AB) %&gt;% filter(AB&gt;<span class="number">500</span>)</div><div class="line">  hist(dat$AVG*<span class="number">1000</span>,xlab=<span class="string">"AVG"</span>,freq=<span class="literal">FALSE</span>,main=y,xlim=c(<span class="number">200</span>,<span class="number">360</span>))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908211314.jpeg">

</div>
<p>我们注意到，普通运行员的AVG为0.275，总体运动员的标准差为0.027。所以，我们看到，0.45这个数字与它们假设偏离非常大，因为这个数字离平均值的差距超过了6个标准差。Jose的这个数字是由于运气，还是说他确实是过去50年来最好的运动员？或者是这两者的结合？但是，他有多幸运，以及运动天赋有多高？如果我们确信他是由于运气出现的这个数字，我们应该把他送到相信他确实是0.45这个水平的球队里，并且有可能高估了他的潜力。</p>
<h3 id="层次模型"><strong>层次模型</strong></h3>
<p>层次模型为我们提供了如何观察到0.45的这个数字的数学描述。首先，我们会随机选择一个内在能力为<span class="math inline">\(\theta\)</span>运动员，然后，我们看到成功概率为<span class="math inline">\(\theta\)</span>的20个随机结果（这一段不太懂，原文为：The hierarchical model provides a mathematical description of how we came to see the observation of .450. First, we pick a player at random with an intrinsic ability summarized by, for example, <span class="math inline">\(\theta\)</span>, then we see 20 random outcomes with success probability <span class="math inline">\(\theta\)</span>.），其中： <span class="math display">\[
\begin{align*}
\theta &amp;\sim N(\mu, \tau^2) \mbox{ describes randomness in picking a player}\\
Y \mid \theta &amp;\sim N(\theta, \sigma^2) \mbox{ describes randomness in the performance of this particular player}
\end{align*}
\]</span> 我们要注意2个层次（这就是为什么要称为层次分析的原因）：</p>
<ol style="list-style-type: decimal">
<li>运动员与运动员之间的变异；</li>
<li>击球时，运气因素导致的变异。</li>
</ol>
<p>在贝叶斯框架中，第一个水平称为先验分布(prior distribution)，第二个水平称为采样分布(sampling distribution)。</p>
<p>现在我们使用贝叶斯模型来计算Jose的数据。假设我们想预测构成他真实击球平均水平<span class="math inline">\(\theta\)</span>的内在能力的话。使用层次模型就按下面的方法表示： <span class="math display">\[
\begin{align*}
\theta &amp;\sim N(.275, .027^2) \\
Y \mid \theta &amp;\sim N(\theta, .111^2) 
\end{align*}
\]</span> 我们现在就可以计算出一个后验分布(posterior distribution)来描述我们对<span class="math inline">\(\theta\)</span>的预测。这里我们可以使用贝叶斯规则的连续计算方法推导后验概率，后验概率是对给定观测数据参数<span class="math inline">\(\theta\)</span>的分布： <span class="math display">\[
\begin{align*}
f_{ \theta \mid Y} (\theta\mid Y) &amp;=
\frac{f_{Y\mid \theta}(Y\mid \theta) f_{\theta}(\theta)
}{f_Y(Y)}\\
&amp;= \frac{f_{Y\mid \theta}(Y\mid \theta) f_{\theta}(\theta)}
{\int_{\theta}f_{Y\mid \theta}(Y\mid \theta)f_{\theta}(\theta)}
\end{align*}
\]</span> 我们主要是对能够使后验概率<span class="math inline">\(f_{\theta\mid Y}(\theta\mid Y)\)</span>的值最大的<span class="math inline">\(\theta\)</span>感兴趣。在我们的案例中，我们可以看出后验概率服从正态分布，我们能计算出均值<span class="math inline">\(\mbox{E}(\theta\mid y)\)</span>，方差<span class="math inline">\(\mbox{var}(\theta\mid y)\)</span>，尤其，我们可以计算出这个分布的均值服从以下分布： <span class="math display">\[
\begin{align*}
\mbox{E}(\theta\mid y) &amp;= B \mu + (1-B) Y\\
&amp;= \mu + (1-B)(Y-\mu)\\
B &amp;= \frac{\sigma^2}{\sigma^2+\tau^2}
\end{align*}
\]</span> 这是一个总体均值<span class="math inline">\(\mu\)</span>和观测数据<span class="math inline">\(Y\)</span>的加权均值。其权重取决于总体<span class="math inline">\(\tau\)</span>的SD和我们观测数据<span class="math inline">\(\sigma\)</span>的SD。这个加权均值有时候也会被称为<code>shrinking</code>，因为它缩小(shrink)了对先验均值的估计，在Joes Iglesias的数据中，结果如下所示： <span class="math display">\[
\begin{align*}
\mbox{E}(\theta \mid Y=.450) &amp;= B \times .275 + (1 - B) \times .450 \\
&amp;= .275 + (1 - B)(.450 - .275) \\
B &amp;=\frac{.111^2}{.111^2 + .027^2} = 0.944\\
\mbox{E}(\theta \mid Y=450) &amp;\approx .285
\end{align*}
\]</span> 方差如下所示： <span class="math display">\[
\mbox{var}(\theta\mid y) = \frac{1}{1/\sigma^2+1/\tau^2}
= \frac{1}{1/.111^2 + 1/.027^2} = 0.00069
\]</span> 标准差因此是0.026。我们开始时，使用了传统频率学派的思路计算出的95%置信区间忽略了来自于其他运动员的数据，因此会单纯地认为Joes的数据是0.0450 ± 0.220。我们随后使用了贝叶斯方法，整合了来源于其他运动员的数据，以及前几年的数据，计算出了后验概率。这种计算思路实际上就是经验贝叶斯方法，因此我们使用了数据先构建了先验知识。从后验结果中我们可以知道，通过报告一个以均值为中心的区间来报告所谓的95%的可信区间，其发生的概率为95%，在这个案例中，其结果是0.285 ± 0.052。原文：From the posterior we can report what is called a 95% credible interval by reporting a region, centered at the mean, with a 95% chance of occurring. In our case, this turns out to be: 0.285 ± 0.052.</p>
<p>贝叶斯可信区间表明，如果其他的球队发现了0.45这个数字，我们应该考虑到Joses可能转会到其他球队，因为我们预测到了Jose的水平高于平均不水平。有意思的是，Red Sox在7月份的时候将Jose转会到了Detroit Tigers队。这里是Jose在接下来的5个月内的击球率：</p>
<table>
<thead>
<tr class="header">
<th>Month</th>
<th>At Bat</th>
<th>Hits</th>
<th>AVG</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>April</td>
<td>20</td>
<td>9</td>
<td>.450</td>
</tr>
<tr class="even">
<td>May</td>
<td>26</td>
<td>11</td>
<td>.423</td>
</tr>
<tr class="odd">
<td>June</td>
<td>86</td>
<td>34</td>
<td>.395</td>
</tr>
<tr class="even">
<td>July</td>
<td>83</td>
<td>17</td>
<td>.205</td>
</tr>
<tr class="odd">
<td>August</td>
<td>85</td>
<td>25</td>
<td>.294</td>
</tr>
<tr class="even">
<td>September</td>
<td>50</td>
<td>10</td>
<td>.200</td>
</tr>
<tr class="odd">
<td>Total w/o April</td>
<td>330</td>
<td>97</td>
<td>.293</td>
</tr>
</tbody>
</table>
<p>虽然这两个区间都包括了最终的击球平均值，但是贝叶斯可信区间提供了更精确的预测，尤其是，这种方法预测到Jose在本赛季的剩余时间里表现不佳。</p>
<h2 id="练习">练习</h2>
<p>P308</p>
<h2 id="层次模型-1">层次模型</h2>
<p>有关层次模型的内容可以参考作者的<a href="https://github.com/genomicsclass/labs/tree/master/modeling/hierarchical_models.Rmd" target="_blank" rel="external">Github</a>。</p>
<p>在这一部分里，我们会使用数据理论来描述在高通量数据分析中常用的方法。常规的思路就是构建一个两层的层次模型。一层用于描述样本/实验单元之间的变异，另外一层用于描述特征值之间的变异。这种分析方法类似于我们前面讲的棒球案例，即第一层用于描述不同运动员之间的变异，第二层用于描述一个运动员成功的随机性。我们这里𢪮的所有模型与方法都考虑了第一个变异水平，例如构建t检验的醋。第二个水平允许我们通过从所有的特征值里“借用(borrow)”信息用于对特征值进行统计推断，从而提供检验效能。</p>
<p>现在我们来看一个在基因表达数据中使用最为广泛的统计学方法。这个统计学方就是由<code>limma</code>Bioconductor包提供的。这个方法已经被用于改造分析RNAseq数据，例如edgeR<a href="https://www.ncbi.nlm.nih.gov/pubmed/19910308" target="_blank" rel="external">《edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.》</a>和DESeq2<a href="https://www.ncbi.nlm.nih.gov/pubmed/25516281" target="_blank" rel="external">《Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2》</a>。这两个包提供了t检验的替代方案，它们通过对方差进行建模从而极大地改善了统计功效。然而在棒球案例中，我们是对均值进行了建模，这是与那两种方法建模的不同之处。对方差建模需要更深的数学知识，但是思路是一样的。我们以一个案例来说明一下这种方法。</p>
<p>下图是一个火山图，它显示了使用t检验来分析数据的结果，显示了效应大小(effect size)和p值，其中使用了6个重复样本（对照组3个，干预组3个），其中有16个基因是人为设定的差异基因。只有这16个基因的备选假设为真，在火山图上它们标记为蓝色，代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(SpikeInSubset) <span class="comment">##Available from Bioconductor</span></div><div class="line">data(rma95)</div><div class="line"><span class="keyword">library</span>(genefilter)</div><div class="line">fac &lt;- factor(rep(<span class="number">1</span>:<span class="number">2</span>,each=<span class="number">3</span>))</div><div class="line">tt &lt;- rowttests(exprs(rma95),fac)</div><div class="line">smallp &lt;- with(tt, p.value &lt; <span class="number">.01</span>)</div><div class="line">spike &lt;- rownames(rma95) %<span class="keyword">in</span>% colnames(pData(rma95))</div><div class="line">cols &lt;- ifelse(spike,<span class="string">"dodgerblue"</span>,ifelse(smallp,<span class="string">"red"</span>,<span class="string">"black"</span>))</div><div class="line">with(tt, plot(-dm, -log10(p.value), cex=<span class="number">.8</span>, pch=<span class="number">16</span>,</div><div class="line">xlim=c(-<span class="number">1</span>,<span class="number">1</span>), ylim=c(<span class="number">0</span>,<span class="number">4.5</span>),</div><div class="line">xlab=<span class="string">"difference in means"</span>,</div><div class="line">col=cols))</div><div class="line">abline(h=<span class="number">2</span>,v=c(-<span class="number">.2</span>,<span class="number">.2</span>), lty=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908225643.jpeg">

</div>
<p>上图是使用t检验计算了两组样本的差异基因，其中Spiked-in基因使用蓝色。剩下的基因中，p小于的用红色标明。</p>
<p>在上面的火山图中，我们将y轴的截止值(cut-off)设为了4.5，但是有一个蓝点的p值小于<span class="math inline">\(10^{-6}\)</span>。但是，从这张图中我们会发现2点怪异之处。第一，按照5% FDR的标准，只有一个阳性结果是显著的，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">sum( p.adjust(tt$p.value,method = <span class="string">"BH"</span>)[spike] &lt; <span class="number">0.05</span>)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">&gt; sum( p.adjust(tt$p.value,method = <span class="string">"BH"</span>)[spike] &lt; <span class="number">0.05</span>)</div><div class="line">[<span class="number">1</span>] <span class="number">1</span></div></pre></td></tr></table></figure>
<p>这个结果与每组3个样本的低统计效能有关。第二，如果我们忽略掉统计推断，仅仅是基于t检验统计量的大小简单地对这些基因进行排序，那么我们会在任何大于1的排序列表中得到很多假阳性结果。例如，按照t检验统计量进行排序，位列前10名的基因中，有6个都是假阳性，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">table( top50=rank(tt$p.value)&lt;= <span class="number">10</span>, spike) <span class="comment">#t-stat and p-val rank is the same</span></div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; table( top50=rank(tt$p.value)&lt;= <span class="number">10</span>, spike) <span class="comment">#t-stat and p-val rank is the same</span></div><div class="line">       spike</div><div class="line">top50   <span class="literal">FALSE</span>  <span class="literal">TRUE</span></div><div class="line">  <span class="literal">FALSE</span> <span class="number">12604</span>    <span class="number">12</span></div><div class="line">  <span class="literal">TRUE</span>      <span class="number">6</span>     <span class="number">4</span></div></pre></td></tr></table></figure>
<p>在火山图中，我们注意到，大多数基因的效能大小都非常小，这说明，估计的标准误非常小，我们可以通过画图的手段来看一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">tt$s &lt;- apply(exprs(rma95), <span class="number">1</span>, <span class="keyword">function</span>(row)</div><div class="line">  sqrt(<span class="number">.5</span> * (var(row[<span class="number">1</span>:<span class="number">3</span>]) + var(row[<span class="number">4</span>:<span class="number">6</span>]) ) ) )</div><div class="line">with(tt, plot(s, -log10(p.value), cex=<span class="number">.8</span>, pch=<span class="number">16</span>,</div><div class="line">              log=<span class="string">"x"</span>,xlab=<span class="string">"estimate of standard deviation"</span>,</div><div class="line">              col=cols))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908230650.jpeg">

</div>
<p>在这里我们就可以看到层次模型的用处了。如果我们假设这些变异的分布在所有基因中，然后我们通过分布来“调整”那些“太小”估计值，就可以改善我们的计算结果。在本书的前面部分中，我们提到F分布与观测到的方差分布近似，即： <span class="math display">\[
s^2 \sim s_0^2 F_{d,d_0}
\]</span> 因为我们有数千个数据点，我们实际上可以检验一下这个假设，并且估计出参数<span class="math inline">\(s_{0}\)</span>和<span class="math inline">\(d_{0}\)</span>。这种估计方法指的就是经验贝叶斯统计，因为我们使用现有的数据（经验）就可以构建先验分布（贝叶斯方法）。</p>
<p>现在我们将前面的棒球案例应用到标准误的估计中。像以前一样，我们已经有了每个基因的观测值<span class="math inline">\(s_{g}\)</span>，这是一个采样分布，用它来作为先验分布。我们因此可以计算出方差<span class="math inline">\(\sigma_{g}^2\)</span>的后来又做分布，并且获得一个后验均值，细节可以参考文献<a href="https://www.ncbi.nlm.nih.gov/pubmed/16646809" target="_blank" rel="external">《Linear models and empirical bayes methods for assessing differential expression in microarray experiments.》</a>，均值如下所示： <span class="math display">\[
\mathrm{E}\left[\sigma_{g}^{2} | s_{g}\right]=\frac{d_{0} s_{0}^{2}+d s_{g}^{2}}{d_{0}+d}
\]</span> 与棒球案例一样，后验均值会降低我们观测到的方差<span class="math inline">\(s_{g}^2\)</span>偏向于全局方差<span class="math inline">\(s_{0}^2\)</span>，其权重取决于样本大小，以及含有自由度<span class="math inline">\(d\)</span>的样本数目，在这个案例中，就是取决于通过<span class="math inline">\(d_{0}\)</span>的先验分布形状。（原文：AAs in the baseball example, the posterior mean <em>shrinks</em> the observed variance <span class="math inline">\(s_g^2\)</span> towards the global variance <span class="math inline">\(s_0^2\)</span> and the weights depend on the sample size through the degrees of freedom <span class="math inline">\(d\)</span> and, in this case, the shape of the prior distribution through <span class="math inline">\(d_0\)</span>. ）</p>
<p>在上面的图形中，我们可以看到40个基因的方差估计是如何缩小(shrink)的：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(limma)</div><div class="line">fit &lt;- lmFit(rma95, model.matrix(~ fac))</div><div class="line">ebfit &lt;- ebayes(fit)</div><div class="line">n &lt;- <span class="number">40</span></div><div class="line">qs &lt;- seq(from=<span class="number">0</span>,to=<span class="number">.2</span>,length=n)</div><div class="line">idx &lt;- sapply(seq_len(n),<span class="keyword">function</span>(i) which(as.integer(cut(tt$s^<span class="number">2</span>,qs)) == i)[<span class="number">1</span>])</div><div class="line">idx &lt;- idx[!is.na(idx)]</div><div class="line">par(mar=c(<span class="number">5</span>,<span class="number">5</span>,<span class="number">2</span>,<span class="number">2</span>))</div><div class="line">plot(<span class="number">1</span>,<span class="number">1</span>,xlim=c(<span class="number">0</span>,<span class="number">.21</span>),ylim=c(<span class="number">0</span>,<span class="number">1</span>),type=<span class="string">"n"</span>,</div><div class="line">     xlab=<span class="string">"variance estimates"</span>,ylab=<span class="string">""</span>,yaxt=<span class="string">"n"</span>)</div><div class="line">axis(<span class="number">2</span>,at=c(<span class="number">.1</span>,<span class="number">.9</span>),c(<span class="string">"before"</span>,<span class="string">"after"</span>),las=<span class="number">2</span>)</div><div class="line">segments((tt$s^<span class="number">2</span>)[idx],rep(<span class="number">.1</span>,n),</div><div class="line">         ebfit$s2.post[idx],rep(<span class="number">.9</span>,n))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908232355.jpeg">

</div>
<p>上图显示的就是估计值如何向先验期望缩小的，40个基因包括了我们选择值的整个范围。</p>
<p>这种调整的一个重要方面就是使那些样本标准差接近于0的基因的样本偏差不再接近于0（向<span class="math inline">\(s_{0}\)</span>收缩）。我们现在就创建一个t检验的统计模型，用于替代使用这个后验均值或“收缩“(shrunken)后的方差估计值。我们称这种t检验模型为适度t检验（moderated t-test）。当我们使用适应t检验后从火山图上就能明显地看到其改进之处：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(limma)</div><div class="line">fit &lt;- lmFit(rma95, model.matrix(~ fac))</div><div class="line">ebfit &lt;- ebayes(fit)</div><div class="line">limmares &lt;- data.frame(dm=coef(fit)[,<span class="string">"fac2"</span>], p.value=ebfit$p.value[,<span class="string">"fac2"</span>])</div><div class="line">with(limmares, plot(dm, -log10(p.value),cex=<span class="number">.8</span>, pch=<span class="number">16</span>,</div><div class="line">                    col=cols,xlab=<span class="string">"difference in means"</span>,</div><div class="line">                    xlim=c(-<span class="number">1</span>,<span class="number">1</span>), ylim=c(<span class="number">0</span>,<span class="number">5</span>)))</div><div class="line">abline(h=<span class="number">2</span>,v=c(-<span class="number">.2</span>,<span class="number">.2</span>), lty=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190908233029.jpeg">

</div>
<p>这个火山图显示的就是使用适度t检验比较两组的差异基因结果。Spiked-in基因用蓝色进行了标注。剩下的基因中，p值小于的用红色标注。</p>
<p>现在我们来看一下排列前10的基因中假阳性的数目，这个数目就降为了2，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">table( top50=rank(limmares$p.value)&lt;= <span class="number">10</span>, spike)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">&gt; table( top50=rank(limmares$p.value)&lt;= <span class="number">10</span>, spike)</div><div class="line">       spike</div><div class="line">top50   <span class="literal">FALSE</span>  <span class="literal">TRUE</span></div><div class="line">  <span class="literal">FALSE</span> <span class="number">12608</span>     <span class="number">8</span></div><div class="line">  <span class="literal">TRUE</span>      <span class="number">2</span>     <span class="number">8</span></div></pre></td></tr></table></figure>
<h2 id="练习-1">练习</h2>
<p>P315</p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Beyes/" <i class="fa fa-tag"></i> Beyes</a>
          
            <a href="/tags/Bierachical/" <i class="fa fa-tag"></i> Bierachical</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/08/22/DAL/DALS017_InferenceForHighDimensionalData2_Principle/" rel="next" title="DALS017-高维数据推断2-统计原理">
                <i class="fa fa-chevron-left"></i> DALS017-高维数据推断2-统计原理
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/08/23/DAL/DALS018_StatisticalModels1_binomal_Possion_MLE/" rel="prev" title="DALS018-统计模型1-二项分布与泊松分布，MLE">
                DALS018-统计模型1-二项分布与泊松分布，MLE <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">214</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">12</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">106</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#贝叶斯统计"><span class="nav-number">2.</span> <span class="nav-text">贝叶斯统计</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#贝叶斯定理"><span class="nav-number">2.1.</span> <span class="nav-text">贝叶斯定理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#模拟"><span class="nav-number">2.2.</span> <span class="nav-text">模拟</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#贝叶斯的实际运用"><span class="nav-number">2.3.</span> <span class="nav-text">贝叶斯的实际运用</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#层次模型"><span class="nav-number">2.4.</span> <span class="nav-text">层次模型</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#练习"><span class="nav-number">3.</span> <span class="nav-text">练习</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#层次模型-1"><span class="nav-number">4.</span> <span class="nav-text">层次模型</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#练习-1"><span class="nav-number">5.</span> <span class="nav-text">练习</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共851.2k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
