<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言 原始的Rmd文档可以参考Github上。 IRanges 和 GRanges 对象是Bioconductor基础数据类型的核心成员，其中 IRanges 定义了 integer ranges ，它用于解决基因组上的位置问题，而 GRanges 则包含了染色体和DNA链的信息。这里我们简介绍一下这两个对象，以及如何操作 IRanges 和 GRanges 。 IRanges 第一步是加">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS008-IRanges和GRanges">
<meta property="og:url" content="http://rvdsd.top/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 原始的Rmd文档可以参考Github上。 IRanges 和 GRanges 对象是Bioconductor基础数据类型的核心成员，其中 IRanges 定义了 integer ranges ，它用于解决基因组上的位置问题，而 GRanges 则包含了染色体和DNA链的信息。这里我们简介绍一下这两个对象，以及如何操作 IRanges 和 GRanges 。 IRanges 第一步是加">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-unnamed-chunk-6-1.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104114940.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-lkir-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-dopr-1.png">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104142451.png">
<meta property="og:updated_time" content="2019-11-13T08:51:00.582Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS008-IRanges和GRanges">
<meta name="twitter:description" content="前言 原始的Rmd文档可以参考Github上。 IRanges 和 GRanges 对象是Bioconductor基础数据类型的核心成员，其中 IRanges 定义了 integer ranges ，它用于解决基因组上的位置问题，而 GRanges 则包含了染色体和DNA链的信息。这里我们简介绍一下这两个对象，以及如何操作 IRanges 和 GRanges 。 IRanges 第一步是加">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-unnamed-chunk-6-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/"/>





  <title>GDAS008-IRanges和GRanges | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS008-IRanges和GRanges</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-08T12:00:00+08:00">
                2019-09-08
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,178
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  24
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>原始的Rmd文档可以参考<a href="https://github.com/genomicsclass/labs/blob/master/biocintro_5x/bioc1_grangeOps.Rmd" target="_blank" rel="external">Github</a>上。</p>
<p><code>IRanges</code> 和 <code>GRanges</code> 对象是Bioconductor基础数据类型的核心成员，其中 <code>IRanges</code> 定义了 <code>integer ranges</code> ，它用于解决基因组上的位置问题，而 <code>GRanges</code> 则包含了染色体和DNA链的信息。这里我们简介绍一下这两个对象，以及如何操作 <code>IRanges</code> 和 <code>GRanges</code> 。</p>
<h2 id="iranges">IRanges</h2>
<p>第一步是加载<code>IRanges</code>包，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">library(IRanges)</div></pre></td></tr></table></figure>
<p><code>IRanges</code> 函数定义了区间范围(interval ranges)。如果你输入2个数据，那么就表示一个区间，例如 [5,10]={5,6,7,8,9,10}，它的宽度是6。另外，如果我们要查看一个 <code>IRanges</code> 定义的区间宽度，我们使用的<code>width()</code>函数，而非 <code>length()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">ir &lt;- IRanges(<span class="number">5</span>,<span class="number">10</span>)</div><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">start(ir)</div><div class="line"><span class="comment">## [1] 5</span></div><div class="line">end(ir)</div><div class="line"><span class="comment">## [1] 10</span></div><div class="line">width(ir)</div><div class="line"><span class="comment">## [1] 6</span></div><div class="line"><span class="comment"># for detailed information on the IRanges class:</span></div><div class="line"><span class="comment"># ?IRanges</span></div><div class="line">length(ir)</div><div class="line"><span class="comment"># [1] 1</span></div></pre></td></tr></table></figure>
<p>一个单独的<code>IRanges</code>对象可以含有多个范围。现在我们可以在一个向量中指定<code>start</code>，在另外一个向量中指定<code>end</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">IRanges(start=c(<span class="number">3</span>,<span class="number">5</span>,<span class="number">17</span>), end=c(<span class="number">10</span>,<span class="number">8</span>,<span class="number">20</span>))</div><div class="line"><span class="comment">## IRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]        17        20         4</span></div></pre></td></tr></table></figure>
<h3 id="内部范围intra-range操作">内部范围(Intra-range)操作</h3>
<p>现在我们继续使用单个范围[5,10]，我们查看内部范围(intra-range)<code>{…,−2,−1,0,1,2,…}</code>的一个数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># full details on the intra-range methods:</span></div><div class="line"><span class="comment"># ?"intra-range-methods"</span></div><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">shift(ir, -<span class="number">2</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3         8         6</span></div></pre></td></tr></table></figure>
<p>这里我们展示了应用于<code>ir</code>的多个不同操作的结果，结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div></pre></td><td class="code"><pre><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">narrow(ir, start=<span class="number">2</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         6        10         5</span></div><div class="line">narrow(ir, end=<span class="number">5</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5         9         5</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">TRUE</span>, both=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2         4         3</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">FALSE</span>, both=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]        11        13         3</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">TRUE</span>, both=<span class="literal">TRUE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2         7         6</span></div><div class="line">ir * <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         6         8         3</span></div><div class="line">ir * -<span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2        13        12</span></div><div class="line">ir + <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        12        10</span></div><div class="line">ir - <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         7         8         2</span></div><div class="line">resize(ir, <span class="number">1</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5         5         1</span></div></pre></td></tr></table></figure>
<p>我们使用下图展示了上面操作的结果。红色柱状部分显示了最初的<code>ir</code>范围。掌握这些操作的最好方法是在自己定义的范围内在控制台中亲自操作一下。</p>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<h3 id="内部范围方法">内部范围方法</h3>
<p>针对<code>intra-range</code>有着一系列的方法。还有一些方法是针对一系列范围的操作，它的输出取决于所有的范围，因此这些方法与<code>intra-range</code>方法不同，后者不会改变输出。现在我们一些案例来说明一下。<code>range()</code>函数提供了从最左侧到最右侧的整数输出，如下所示：</p>
<p>注：原文我不太理解，翻译的不精准，原文如下：</p>
<blockquote>
<p>There are also a set of <em>inter-range</em> methods. These are operations which work on a set of ranges, and the output depends on all the ranges, thus distinguishes these methods from the <em>intra-range</em> methods, for which the other ranges in the set do not change the output. This is best explained with some examples. The <code>range</code> function gives the integer range from the start of the leftmost range to the end of the rightmost range:</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># full details on the inter-range methods:</span></div><div class="line"><span class="comment"># ?"inter-range-methods"</span></div><div class="line">(ir &lt;- IRanges(start=c(<span class="number">3</span>,<span class="number">5</span>,<span class="number">17</span>), end=c(<span class="number">10</span>,<span class="number">8</span>,<span class="number">20</span>)))</div><div class="line"><span class="comment">## IRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]        17        20         4</span></div><div class="line">range(ir)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        20        18</span></div></pre></td></tr></table></figure>
<p><code>reduce</code> 函数能折叠这些范围，以便于整个可以只被输出范围中的一个范围覆盖，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">reduce(ir)</div><div class="line"><span class="comment">## IRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]        17        20         4</span></div></pre></td></tr></table></figure>
<p>注：ir原来有3个范围，其中[5,8]范围是在[3,10]之间，因此前者被折叠到了后者之中。</p>
<p><code>gaps</code> 函数能够给出上面两个范围的区间，但是它不覆盖任何<code>ir</code>中的范围，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">gaps(ir)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]        11        16         6</span></div></pre></td></tr></table></figure>
<p><code>disjoin</code> 会打断 <code>ir</code>的范围，并它们分成不同的范围，如下所示：</p>
<blockquote>
<p>disjoin返回一个disjoint对象，它会查看参数中端点的并集，返回不相交的对象。换话句讲，结果中由区间的最大长度范围组成，在该范围上，对象x中的重叠范围集合是相同的，并且至少为1。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">disjoin(ir)</div><div class="line"><span class="comment">## IRanges object with 4 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3         4         2</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]         9        10         2</span></div><div class="line"><span class="comment">##   [4]        17        20         4</span></div></pre></td></tr></table></figure>
<p>现在我们使用示意图来说明一下：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104114940.png">

</div>
<p>第一二行是原始的范围，第四行是经过disjoin处理后的范围。</p>
<h2 id="granges">GRanges</h2>
<p>GRanges对象包含了Iranges，它含有2个方面的重要信息：</p>
<ul>
<li>染色体信息（Bioconductor称为 <code>seqnames</code>）。</li>
<li>DNA链。</li>
</ul>
<p>链的信息可以指定为 <code>+</code> 号或 <code>-</code> ，或者是保留未指定的 <code>*</code> 。正义链的特征则是在编号线上从左到右的生物学方面，而负链特征则是从右到左。就IRanges而言，正链特征是从<code>start</code>到 <code>end</code> ,负链则是从 <code>end</code> 到 <code>start</code> 。这初看起来比较混乱，但这是必需的，因此 <code>width</code> 被定义为 <code>end - start + 1</code>，并且不允许有负的宽度范围。因为DNA有两条链，这两条链具有相反的方向性，因此当我们说DNA时，可以只指明其中的一条链。</p>
<p>通过IRnage、染色体名称和一条链，我们可以确定我们与另外一个研究人员所指的DNA分子具有相同的区域和链，因为我们使用的相同的基因组构建的。GRanges对象中还可以包括其它部分的信息，不过上述的两个信息则是最重要的 。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicRanges)</div></pre></td></tr></table></figure>
<p>现在我们创建一个虚构染色体<code>chrZ</code> 上的两个范围。我们会说这两个范围指的基因组<code>hg19</code>上的范围。因为我们没有将我们的基因组链接到数据库，因此我们可以指定一个在hg19中并不存在的染色体，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">gr &lt;- GRanges(<span class="string">"chrZ"</span>, IRanges(start=c(<span class="number">5</span>,<span class="number">10</span>),end=c(<span class="number">35</span>,<span class="number">45</span>)),</div><div class="line">              strand=<span class="string">"+"</span>, seqlengths=c(chrZ=<span class="number">100L</span>))</div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome</span></div><div class="line">genome(gr) &lt;- <span class="string">"hg19"</span></div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们可以调用 <code>seqnames</code> 和 <code>seqlengths</code> 来查看上面的信息：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">seqnames(gr)</div><div class="line"><span class="comment">## factor-Rle of length 2 with 1 run</span></div><div class="line"><span class="comment">##   Lengths:    2</span></div><div class="line"><span class="comment">##   Values : chrZ</span></div><div class="line"><span class="comment">## Levels(1): chrZ</span></div><div class="line">seqlengths(gr)</div><div class="line"><span class="comment">## chrZ </span></div><div class="line"><span class="comment">##  100</span></div></pre></td></tr></table></figure>
<p>我们可以使用 <code>shift</code> 函数（就像在IRanges中使用的一样）来处理GRanges。但是，当我们尝试超过染色体长度之外的范围时，就会出现警告信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">shift(gr, <span class="number">10</span>)</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [15, 45]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [20, 55]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div><div class="line">shift(gr, <span class="number">80</span>)</div><div class="line"><span class="comment">## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequence</span></div><div class="line"><span class="comment">##   chrZ. Note that only ranges located on a non-circular sequence</span></div><div class="line"><span class="comment">##   whose length is not NA can be considered out-of-bound (use</span></div><div class="line"><span class="comment">##   seqlengths() and isCircular() to get the lengths and circularity</span></div><div class="line"><span class="comment">##   flags of the underlying sequences). You can use trim() to trim</span></div><div class="line"><span class="comment">##   these ranges. See ?`trim,GenomicRanges-method` for more</span></div><div class="line"><span class="comment">##   information.</span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ [85, 115]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ [90, 125]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>如果我们使用 <code>trim</code> 处理这个范围，我们就能获取剩下的范围，而不考虑超出染色体长的部分：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">trim(shift(gr, <span class="number">80</span>))</div><div class="line"><span class="comment">## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequence</span></div><div class="line"><span class="comment">##   chrZ. Note that only ranges located on a non-circular sequence</span></div><div class="line"><span class="comment">##   whose length is not NA can be considered out-of-bound (use</span></div><div class="line"><span class="comment">##   seqlengths() and isCircular() to get the lengths and circularity</span></div><div class="line"><span class="comment">##   flags of the underlying sequences). You can use trim() to trim</span></div><div class="line"><span class="comment">##   these ranges. See ?`trim,GenomicRanges-method` for more</span></div><div class="line"><span class="comment">##   information.</span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ [85, 100]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ [90, 100]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们可以使用 <code>mcols</code> 函数（链表示元数据列）向每个范围添加列信息。需要注意的是，这个操作对IRanges也是可行的。我们还可以通过指定 <code>NULL</code> 来移除列信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">mcols(gr)</div><div class="line"><span class="comment">## DataFrame with 2 rows and 0 columns</span></div><div class="line">mcols(gr)$value &lt;- c(-<span class="number">1</span>,<span class="number">4</span>)</div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 1 metadata column:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand |     value</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      + |        -1</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      + |         4</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div><div class="line">mcols(gr)$value &lt;- <span class="literal">NULL</span></div><div class="line">&gt; gr</div><div class="line"><span class="comment"># GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">#      seqnames    ranges strand</span></div><div class="line"><span class="comment">#         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">#  [1]     chrZ      5-35      +</span></div><div class="line"><span class="comment">#  [2]     chrZ     10-45      +</span></div><div class="line"><span class="comment">#  -------</span></div><div class="line"><span class="comment">#  seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<h3 id="grangeslist">GRangesList</h3>
<p>当我们涉及基因时，我们通过创建一个GRanges列表是很有用的，例如用于表示分组信息（比如每个基因的外显子）。该列表的元素是基因，并且在每个元素中，外显子的范围被定义为GRanges。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">11</span>:<span class="number">13</span>,<span class="number">51</span>:<span class="number">53</span>))</div><div class="line">grl &lt;- GRangesList(gr, gr2)</div><div class="line">grl</div><div class="line"><span class="comment">## GRangesList object of length 2:</span></div><div class="line"><span class="comment">## [[1]] </span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## [[2]] </span></div><div class="line"><span class="comment">## GRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames   ranges strand</span></div><div class="line"><span class="comment">##   [1]     chrZ [11, 51]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ [12, 52]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ [13, 53]      *</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## -------</span></div><div class="line"><span class="comment">## seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p><code>GRangesList</code> 对象中的长度就是其中的<code>GRanges</code> 对象个数。为了获取每个GRanges的长度，我们可以调用 <code>elementNROWS</code>。我们可以通过两个方括号的典型列表索引方法来建立索引，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">length(grl)</div><div class="line"><span class="comment">## [1] 2</span></div><div class="line">elementNROWS(grl)</div><div class="line"><span class="comment">## [1] 2 3</span></div><div class="line">grl[[<span class="number">1</span>]]</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p><code>width</code>函数会返回一个 <code>integerList</code>。如果使用<code>sum</code>，我们会得到列表中每个GRanges对象的宽度向量，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">width(grl)</div><div class="line"><span class="comment">## IntegerList of length 2</span></div><div class="line"><span class="comment">## [[1]] 31 36</span></div><div class="line"><span class="comment">## [[2]] 41 41 41</span></div><div class="line">sum(width(grl))</div><div class="line"><span class="comment">## [1]  67 123</span></div></pre></td></tr></table></figure>
<p>我们可以像以前那样添加元数据(metadata)，现在每个GRange对象都有一行元数据。当我们输出GRangesList时它们并不业显示出来，但我们可以通过 <code>mcols</code> 进行访问，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line">mcols(grl)$value &lt;- c(<span class="number">5</span>,<span class="number">7</span>)</div><div class="line">grl</div><div class="line"><span class="comment">## GRangesList object of length 2:</span></div><div class="line"><span class="comment">## [[1]] </span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## [[2]] </span></div><div class="line"><span class="comment">## GRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames   ranges strand</span></div><div class="line"><span class="comment">##   [1]     chrZ [11, 51]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ [12, 52]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ [13, 53]      *</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## -------</span></div><div class="line"><span class="comment">## seqinfo: 1 sequence from hg19 genome</span></div><div class="line">mcols(grl)</div><div class="line"><span class="comment">## DataFrame with 2 rows and 1 column</span></div><div class="line"><span class="comment">##       value</span></div><div class="line"><span class="comment">##   &lt;numeric&gt;</span></div><div class="line"><span class="comment">## 1         5</span></div><div class="line"><span class="comment">## 2         7</span></div></pre></td></tr></table></figure>
<h3 id="findoverlaps与over">findOverlaps与%over%</h3>
<p>现在我们来看两个比较<code>GRanges</code>对象的常用方法，首先我们要创建两组范围如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">(gr1 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(c(<span class="number">1</span>,<span class="number">11</span>,<span class="number">21</span>,<span class="number">31</span>,<span class="number">41</span>),width=<span class="number">5</span>),strand=<span class="string">"*"</span>))</div><div class="line"><span class="comment">## GRanges object with 5 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 1,  5]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [11, 15]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ  [21, 25]      *</span></div><div class="line"><span class="comment">##   [4]     chrZ  [31, 35]      *</span></div><div class="line"><span class="comment">##   [5]     chrZ  [41, 45]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div><div class="line">(gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(c(<span class="number">19</span>,<span class="number">33</span>),c(<span class="number">38</span>,<span class="number">35</span>)),strand=<span class="string">"*"</span>))</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [19, 38]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [33, 35]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p><code>findOverlaps</code> 会返回一个 <code>Hits</code> 对象，这个对象含有关于检索范围的信息（第一个参数），这个范围与主题（第二个参数）有哪些地方重叠。还有许多参数用于指定计算哪种类型的重叠，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">fo &lt;- findOverlaps(gr1, gr2)</div><div class="line">fo</div><div class="line"><span class="comment">## Hits object with 3 hits and 0 metadata columns:</span></div><div class="line"><span class="comment">##       queryHits subjectHits</span></div><div class="line"><span class="comment">##       &lt;integer&gt;   &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3           1</span></div><div class="line"><span class="comment">##   [2]         4           1</span></div><div class="line"><span class="comment">##   [3]         4           2</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   queryLength: 5 / subjectLength: 2</span></div><div class="line">queryHits(fo)</div><div class="line"><span class="comment">## [1] 3 4 4</span></div><div class="line">subjectHits(fo)</div><div class="line"><span class="comment">## [1] 1 1 2</span></div></pre></td></tr></table></figure>
<p>得到重叠信息的另外一种方式就是使用 <code>%over%</code> ，它返回一个逻辑向量，这个向量表示了第一个参数中的范围与第二个参数中任何范围重叠的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">gr1 %over% gr2</div><div class="line"><span class="comment">## [1] FALSE FALSE  TRUE  TRUE FALSE</span></div><div class="line">gr1[gr1 %over% gr2]</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [21, 25]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [31, 35]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p>需要注意的是，这两种操作方式都是链特异性的，虽然 <code>findOverlaps</code> 有一个 <code>ignore.strand</code> 选择参数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">gr1 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">1</span>,<span class="number">10</span>),strand=<span class="string">"+"</span>)</div><div class="line">gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">1</span>,<span class="number">10</span>),strand=<span class="string">"-"</span>)</div><div class="line">gr1 %over% gr2</div><div class="line"><span class="comment">## [1] FALSE</span></div></pre></td></tr></table></figure>
<h3 id="rle-和views">Rle 和Views</h3>
<p>最后，我们来简单了解一下由IRanges定义的两个相关类，即Rle类和Views类。Rle全称为run-length encoding，它表示的是重复数据的压缩工，用于代替类似于[1, 1, 1, 1]这样的储存。</p>
<p>我们只用储存数据1，以及重复次数4即可。数据越重复，使用Rle对象的压缩效果就越明显。</p>
<p>我们使用 <code>str</code> 来查看 Rle的内部结构，用于显示它的储存数据和重复次数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">(r &lt;- Rle(c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,-<span class="number">2</span>,-<span class="number">2</span>,-<span class="number">2</span>,rep(-<span class="number">1</span>,<span class="number">20</span>))))</div><div class="line"><span class="comment">## numeric-Rle of length 28 with 4 runs</span></div><div class="line"><span class="comment">##   Lengths:  3  2  3 20</span></div><div class="line"><span class="comment">##   Values :  1  0 -2 -1</span></div><div class="line">str(r)</div><div class="line"><span class="comment">## Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></div><div class="line"><span class="comment">##   ..@ values         : num [1:4] 1 0 -2 -1</span></div><div class="line"><span class="comment">##   ..@ lengths        : int [1:4] 3 2 3 20</span></div><div class="line"><span class="comment">##   ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   ..@ metadata       : list()</span></div><div class="line">as.numeric(r)</div><div class="line"><span class="comment">##  [1]  1  1  1  0  0 -2 -2 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1</span></div><div class="line"><span class="comment">## [24] -1 -1 -1 -1 -1</span></div></pre></td></tr></table></figure>
<p>Views对象可以被为一个“窗口”用于查看一个序列信息，如下所示：</p>
<p>A <em>Views</em> object can be thought of as “windows” looking into a sequence.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">(v &lt;- Views(r, start=c(<span class="number">4</span>,<span class="number">2</span>), end=c(<span class="number">7</span>,<span class="number">6</span>)))</div><div class="line"><span class="comment">## Views on a 28-length Rle subject</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## views:</span></div><div class="line"><span class="comment">##     start end width</span></div><div class="line"><span class="comment">## [1]     4   7     4 [ 0  0 -2 -2]</span></div><div class="line"><span class="comment">## [2]     2   6     5 [ 1  1  0  0 -2]</span></div></pre></td></tr></table></figure>
<p>需要注意的是，Veiws对象的内部结构只是原始对象，以及指定窗口的IRangs。使用Views的最好好处就是，当原始对象没有储存在内存中时，在这种情况下，View对象就是一个轻量级的类，它能帮助我们引用子序列，而不必将整个序列都加载到内存中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">str(v)</div><div class="line"><span class="comment">## Formal class 'RleViews' [package "IRanges"] with 5 slots</span></div><div class="line"><span class="comment">##   ..@ subject        :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></div><div class="line"><span class="comment">##   .. .. ..@ values         : num [1:4] 1 0 -2 -1</span></div><div class="line"><span class="comment">##   .. .. ..@ lengths        : int [1:4] 3 2 3 20</span></div><div class="line"><span class="comment">##   .. .. ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ metadata       : list()</span></div><div class="line"><span class="comment">##   ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></div><div class="line"><span class="comment">##   .. .. ..@ start          : int [1:2] 4 2</span></div><div class="line"><span class="comment">##   .. .. ..@ width          : int [1:2] 4 5</span></div><div class="line"><span class="comment">##   .. .. ..@ NAMES          : NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ elementType    : chr "integer"</span></div><div class="line"><span class="comment">##   .. .. ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ metadata       : list()</span></div><div class="line"><span class="comment">##   ..@ elementType    : chr "ANY"</span></div><div class="line"><span class="comment">##   ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   ..@ metadata       : list()</span></div></pre></td></tr></table></figure>
<h2 id="基因组元件的应用链特异性操作">基因组元件的应用：链特异性操作</h2>
<p>在这一部分，我们使用一个小型的范围来说明一下基因的范围内(intra-range)操作，包括reduce, disjoin与gaps。我们添加链和seqname信息，并显示如想调整大小和侧翼(flank)用于识别TSS和启动子区域，如下所示：</p>
<h3 id="一组简单的范围">一组简单的范围</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">ir &lt;- IRanges(c(3, 8, 14, 15, 19, 34, 40),</div><div class="line">  width = c(12, 6, 6, 15, 6, 2, 7))</div></pre></td></tr></table></figure>
<p>现在我们使用图片形式查看一下<code>ir</code>，以及内部范围(intra-range)操作，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(4,1), mar=c(4,2,2,2))</div><div class="line">plotRanges(ir, xlim=c(0,60))</div><div class="line">plotRanges(reduce(ir), xlim=c(0,60))</div><div class="line">plotRanges(disjoin(ir), xlim=c(0,60))</div><div class="line">plotRanges(gaps(ir), xlim=c(0,60))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-lkir-1.png" alt="plot of chunk lkir">
<p class="caption">plot of chunk lkir</p>
</div>
<p>注：<code>plotRange()</code>函数似乎不在这个包中，网上找到了它的原代码，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">plotRanges &lt;- <span class="keyword">function</span>(x, xlim = x, main = deparse(substitute(x)),</div><div class="line">                       col = <span class="string">"black"</span>, sep = <span class="number">0.5</span>, <span class="keyword">...</span>)</div><div class="line">&#123;</div><div class="line">  height &lt;- <span class="number">1</span></div><div class="line">  <span class="keyword">if</span> (is(xlim, <span class="string">"Ranges"</span>))</div><div class="line">    xlim &lt;- c(min(start(xlim)), max(end(xlim)))</div><div class="line">  bins &lt;- disjointBins(IRanges(start(x), end(x) + <span class="number">1</span>))</div><div class="line">  plot.new()</div><div class="line">  plot.window(xlim, c(<span class="number">0</span>, max(bins)*(height + sep)))</div><div class="line">  ybottom &lt;- bins * (sep + height) - height</div><div class="line">  rect(start(x)-<span class="number">0.5</span>, ybottom, end(x)+<span class="number">0.5</span>, ybottom + height, col = col, <span class="keyword">...</span>)</div><div class="line">  title(main)</div><div class="line">  axis(<span class="number">1</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p><code>reduce(x)</code>函数生成一组不重叠的范围，它覆盖了x中的所有位点。这种操作用于生了你许多转录本的基因模式的复杂性，其中我们可能只想要转录出已知区间的地址，不考虑transcript of residence。</p>
<p><code>disjoin(x)</code>生成一组覆盖x的所有位置的范围，使得分离输出中的任何范围都不会与x中的任何间隔的端点重叠。这为我们提供了连续间隔最大可能的集合，这些连续间隔在原始间隔集中具有端点的地方被分割开。</p>
<p><code>gaps(x)</code>产生一组覆盖[start(X)，end(X)]中未被x中任何范围覆盖的位置的范围。它能给出编码序列位置和外显子间隔，这可用于枚举内含子。</p>
<h3 id="granges的扩展">GRanges的扩展</h3>
<p>现在我们添加上染色体与链信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(GenomicRanges)</div><div class="line">gir = GRanges(seqnames=&quot;chr1&quot;, ir, strand=c(rep(&quot;+&quot;, 4), rep(&quot;-&quot;,3)))</div></pre></td></tr></table></figure>
<p>让我们假设间隔代表基因。下图说明了转录起始位点(绿色)、上游启动子区域(紫色)、下游启动子区域(棕色)的位置，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(4,1), mar=c(4,2,2,2))</div><div class="line">plotGRanges(gir, xlim=c(0,60))</div><div class="line"># </div><div class="line">plotGRanges(resize(gir,1), xlim=c(0,60),col=&quot;green&quot;)</div><div class="line">plotGRanges(flank(gir,3), xlim=c(0,60), col=&quot;purple&quot;)</div><div class="line">plotGRanges(flank(gir,2,start=FALSE), xlim=c(0,60), col=&quot;brown&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-dopr-1.png" alt="plot of chunk dopr">
<p class="caption">plot of chunk dopr</p>
</div>
<p>注：<code>plotGRanges</code>的源代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">plotGRanges = <span class="keyword">function</span> (x, xlim = x, col = <span class="string">"black"</span>, sep = <span class="number">0.5</span>, xlimits = c(<span class="number">0</span>, </div><div class="line">    <span class="number">60</span>), <span class="keyword">...</span>) </div><div class="line">&#123;</div><div class="line">    main = deparse(substitute(x))</div><div class="line">    ch = as.character(seqnames(x)[<span class="number">1</span>])</div><div class="line">    x = ranges(x)</div><div class="line">    height &lt;- <span class="number">1</span></div><div class="line">    <span class="keyword">if</span> (is(xlim, <span class="string">"Ranges"</span>)) </div><div class="line">        xlim &lt;- c(min(start(xlim)), max(end(xlim)))</div><div class="line">    bins &lt;- disjointBins(IRanges(start(x), end(x) + <span class="number">1</span>))</div><div class="line">    plot.new()</div><div class="line">    plot.window(xlim = xlimits, c(<span class="number">0</span>, max(bins) * (height + sep)))</div><div class="line">    ybottom &lt;- bins * (sep + height) - height</div><div class="line">    rect(start(x) - <span class="number">0.5</span>, ybottom, end(x) + <span class="number">0.5</span>, ybottom + height, </div><div class="line">        col = col, <span class="keyword">...</span>)</div><div class="line">    title(main, xlab = ch)</div><div class="line">    axis(<span class="number">1</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>请注意，我们不需要采取特殊步骤来处理链中的差异</p>
<h2 id="甲基化微阵列数据的可视化">甲基化微阵列数据的可视化</h2>
<p>在我们讨论 <a href="http://genomicsclass.github.io/book/pages/dataman2017.html#methy" target="_blank" rel="external">SummarizedExperiment applications</a>时，我们输入了由Illumina 450k甲基化微阵列生成的数据。</p>
<p>在这一部分中，我们将介绍如何使用GenomicRanges和Gviz来研究一下在基因水平注释下的甲基化模型。这个思路非常简单：只需从所有样本中提取M值（甲基化与总DNA 比值的基因座特定的估计的log转换），并在选定基因的基因模型中绘制出来，如下所示：</p>
<p>我们回顾一下数据是如何获取和导入的：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ArrayExpress)</div><div class="line"><span class="keyword">if</span> (!file.exists(<span class="string">"E-MTAB-5797.sdrf.txt"</span>)) nano = getAE(<span class="string">"E-MTAB-5797"</span>)</div><div class="line"><span class="keyword">library</span>(minfi)</div><div class="line">pref = unique(substr(dir(patt=<span class="string">"idat"</span>),<span class="number">1</span>,<span class="number">17</span>)) <span class="comment"># find the prefix strings</span></div><div class="line">raw = read.metharray(pref)</div><div class="line">glioMeth = preprocessQuantile(raw) <span class="comment"># generate SummarizedExperiment</span></div><div class="line"><span class="comment">## [preprocessQuantile] Mapping to genome.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Fixing outliers.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Quantile normalizing.</span></div></pre></td></tr></table></figure>
<p>这些步骤需要网联，并且需要花几分钟。</p>
<p>一旦我们的会话中有 <code>glioMeth</code> ，需要将以下代码添加到会话中。我们将会介绍它是如何工作的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">MbyGene = <span class="keyword">function</span>(mset, symbol=<span class="string">"TP53"</span>, rad=<span class="number">5000</span>) &#123;</div><div class="line"><span class="comment"># phase 1: annotated GRanges for the gene</span></div><div class="line"> <span class="keyword">require</span>(erma)</div><div class="line"> <span class="keyword">require</span>(Gviz)</div><div class="line"> gmod = suppressMessages(genemodel(symbol))     <span class="comment"># erma utility</span></div><div class="line"> gseq = as.character(seqnames(gmod)[<span class="number">1</span>])</div><div class="line"> gmod$transcript = symbol</div><div class="line"><span class="comment"># phase 2: filter down to the region of interest</span></div><div class="line"> mlim = mset[which(seqnames(mset)==gseq),] <span class="comment"># restrict to chromosome</span></div><div class="line"> <span class="comment"># now focus the methylation data to vicinity of gene</span></div><div class="line"> d1 = subsetByOverlaps(GRanges(rowRanges(mlim),,, getM(mlim)), </div><div class="line">         range(gmod)+rad)</div><div class="line"><span class="comment"># phase 3: use Gviz</span></div><div class="line"> plotTracks(list(DataTrack(d1), </div><div class="line">              GeneRegionTrack(gmod, </div><div class="line">               transcriptAnnotation=<span class="string">"transcript"</span>, name=gseq), </div><div class="line">              GenomeAxisTrack(name=gseq, showTitle=<span class="literal">TRUE</span>)))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>代码表明了三个阶段：获取基因区域，并添加转录本(<code>transcript</code>)注释用于绘制所有外显子的结合；将<code>GenomicRatioSet</code>（继承自</p>
<p>对代码的注释表明了三个阶段：获取基因区并添加用于所有外显子联合的信息性绘图的转录注释；将GenomicRatioSet（继承自RangedSummarizedExperiment）缩小到感兴趣的间隔，这个间隔是由基因模式和rad参数确定；使用Gviz来构建可用于绘图的对象，并绘制出来。</p>
<p>Gviz的详细信息在该包的文档中。现在我们绘制出以基因为中心的图形，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">MbyGene(glioMeth, symbol=<span class="string">"TERT"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104142451.png">

</div>
<p>从上图中我们可以发现这些信息：</p>
<ul>
<li>基因组背景（以兆(megabase)为单位的染色体和区域）。</li>
<li>感兴趣的基因在单行中的结构，它们表示了表达间隔的联合。</li>
<li>450k针近的位置（蓝色数据点的x坐标）。</li>
<li>甲基化估计中的样本间变异。</li>
<li>所选基因组区域的甲基化变异。</li>
</ul>
<p>在6x模块中，我们将学习如何使用额外的软件包来创建这种类型的交互式展示，允许我们选择基因并随意放大感兴趣的区域。</p>
<h2 id="脚注">脚注</h2>
<p>更多的信息可以查看 <code>GenomicRanges</code> 包，以及查阅PLOS Comp Bio上的文献，它是GenomicRanges的作者写的：</p>
<p>http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118</p>
<p>使用vignettes也能查看这个包的众多功能文档，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">browseVignettes(&quot;GenomicRanges&quot;)</div></pre></td></tr></table></figure>
<p>或者是使用help函数：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">help(package=&quot;GenomicRanges&quot;, help_type=&quot;html&quot;)</div></pre></td></tr></table></figure>
<p>对于Bed Tools的用户来说，<code>HelloRanges</code>包对于在BEd和GRanges概念框架之间转换概念非常有用。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://github.com/genomicsclass/labs/blob/master/biocintro_5x/bioc1_grangeOps.Rmd" target="_blank" rel="external">bioc1_grangeOps.Rmd</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_igranges.html" target="_blank" rel="external">IRanges and GRanges</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/07/Experiment/HIPC介绍/" rel="next" title="HIPC介绍">
                <i class="fa fa-chevron-left"></i> HIPC介绍
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/" rel="prev" title="GDAS009-Bioconductor中的基因组注释">
                GDAS009-Bioconductor中的基因组注释 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">246</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">117</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#iranges"><span class="nav-number">2.</span> <span class="nav-text">IRanges</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#内部范围intra-range操作"><span class="nav-number">2.1.</span> <span class="nav-text">内部范围(Intra-range)操作</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#内部范围方法"><span class="nav-number">2.2.</span> <span class="nav-text">内部范围方法</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#granges"><span class="nav-number">3.</span> <span class="nav-text">GRanges</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#grangeslist"><span class="nav-number">3.1.</span> <span class="nav-text">GRangesList</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#findoverlaps与over"><span class="nav-number">3.2.</span> <span class="nav-text">findOverlaps与%over%</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#rle-和views"><span class="nav-number">3.3.</span> <span class="nav-text">Rle 和Views</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#基因组元件的应用链特异性操作"><span class="nav-number">4.</span> <span class="nav-text">基因组元件的应用：链特异性操作</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#一组简单的范围"><span class="nav-number">4.1.</span> <span class="nav-text">一组简单的范围</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#granges的扩展"><span class="nav-number">4.2.</span> <span class="nav-text">GRanges的扩展</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#甲基化微阵列数据的可视化"><span class="nav-number">5.</span> <span class="nav-text">甲基化微阵列数据的可视化</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#脚注"><span class="nav-number">6.</span> <span class="nav-text">脚注</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">7.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共980.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
