<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言 本篇笔记主要是介绍了Bioconductor与基因组级数据分析的关系。 R语言，R包与仓库(repositories) 学习这一系列课程的前提是你已经有了一定的R语言基因。并且我们推荐使用RStudio作为R语言的使用环境。如果你跳过前面的1x-4x部分，直接进入5x，可以先参考一下其它的R语言教程，例如 Try R 。 Why R? Bioconductor的基础就是R。使用R有">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS003-Bioconductor与基因组级数据分析简介">
<meta property="og:url" content="http://rvdsd.top/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 本篇笔记主要是介绍了Bioconductor与基因组级数据分析的关系。 R语言，R包与仓库(repositories) 学习这一系列课程的前提是你已经有了一定的R语言基因。并且我们推荐使用RStudio作为R语言的使用环境。如果你跳过前面的1x-4x部分，直接进入5x，可以先参考一下其它的R语言教程，例如 Try R 。 Why R? Bioconductor的基础就是R。使用R有">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191105142824.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/biointro-lkycc2-1.png">
<meta property="og:updated_time" content="2019-11-13T08:49:51.027Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS003-Bioconductor与基因组级数据分析简介">
<meta name="twitter:description" content="前言 本篇笔记主要是介绍了Bioconductor与基因组级数据分析的关系。 R语言，R包与仓库(repositories) 学习这一系列课程的前提是你已经有了一定的R语言基因。并且我们推荐使用RStudio作为R语言的使用环境。如果你跳过前面的1x-4x部分，直接进入5x，可以先参考一下其它的R语言教程，例如 Try R 。 Why R? Bioconductor的基础就是R。使用R有">
<meta name="twitter:image" content="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191105142824.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/"/>





  <title>GDAS003-Bioconductor与基因组级数据分析简介 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS003-Bioconductor与基因组级数据分析简介</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-03T12:00:00+08:00">
                2019-09-03
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  3,445
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  13
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>本篇笔记主要是介绍了Bioconductor与基因组级数据分析的关系。</p>
<h2 id="r语言r包与仓库repositories">R语言，R包与仓库(repositories)</h2>
<p>学习这一系列课程的前提是你已经有了一定的R语言基因。并且我们推荐使用RStudio作为R语言的使用环境。如果你跳过前面的1x-4x部分，直接进入5x，可以先参考一下其它的R语言教程，例如 <a href="http://tryr.codeschool.com/" target="_blank" rel="external">Try R</a> 。</p>
<h2 id="why-r">Why R?</h2>
<p>Bioconductor的基础就是R。使用R有3个原因：</p>
<ul>
<li>许多统计学家与生物统计学家使用R语言，并且他们使用R语言设计了许多算法可以帮助我们理解复杂的实验数据。</li>
<li>R语言有着高度的互操作性(interoperable)，促进了使用其它语言编写软件组件的重用。</li>
<li>R可移植到在各种计算机上的操作系统里(Linux、MacOSX、Windows)，初学者可以使用这些平台中的任一个来学习。</li>
</ul>
<p>总之，R的易用性和在统计学和“数据科学”中扮演着核心角色，因此当生物学家和统计学家面对基因组级的实验数据时，选择R作为处理工具就是自然而然的事情了。Bioconductor项目开始于2001年，自那以后，它就与基因组级生物学中不断增长的数据量和复杂性保持着同步。</p>
<h2 id="函数式面向对象程序设计">函数式面向对象程序设计</h2>
<p>R语言是一种混合了函数式编程与面向对象的混合型语言。</p>
<p>在R中进行函数式编程时，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">square = <span class="keyword">function</span>(x) x^<span class="number">2</span></div></pre></td></tr></table></figure>
<p>上面就是一个典型的函数式编程。我们定义了一个函数<code>square()</code>，它会计算输入的数据字的平方，即<code>^2</code>，例如当我们输入3时，结果就是在R中，所有的计算都是通过函数进行的。</p>
<p>在进行面向对象的编程时，我们所关注的生就是构建数据结构，定义能够操作这些结构化数据的方法。这是一种接近于人正常思维的编程思想。最早出现在20世纪90年代。现在我们以Bioconductor中的一个案例说明一下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">class(Homo.sapiens)</div><div class="line">methods(class=class(Homo.sapiens))</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(Homo.sapiens)</div><div class="line">&gt; class(Homo.sapiens)</div><div class="line">[<span class="number">1</span>] <span class="string">"OrganismDb"</span></div><div class="line">attr(,<span class="string">"package"</span>)</div><div class="line">[<span class="number">1</span>] <span class="string">"OrganismDbi"</span></div><div class="line">&gt; methods(class=class(Homo.sapiens))</div><div class="line"> [<span class="number">1</span>] asBED                 asGFF                 cds                   cdsBy                </div><div class="line"> [<span class="number">5</span>] coerce&lt;-              columns               dbconn                dbfile               </div><div class="line"> [<span class="number">9</span>] disjointExons         distance              exons                 exonsBy              </div><div class="line">[<span class="number">13</span>] extractUpstreamSeqs   fiveUTRsByTranscript  genes                 getTxDbIfAvailable   </div><div class="line">[<span class="number">17</span>] intronsByTranscript   isActiveSeq           isActiveSeq&lt;-         keys                 </div><div class="line">[<span class="number">21</span>] keytypes              mapIds                mapToTranscripts      metadata             </div><div class="line">[<span class="number">25</span>] microRNAs             promoters             resources             select               </div><div class="line">[<span class="number">29</span>] selectByRanges        selectRangesById      seqinfo               show                 </div><div class="line">[<span class="number">33</span>] taxonomyId            threeUTRsByTranscript transcripts           transcriptsBy        </div><div class="line">[<span class="number">37</span>] tRNAs                 TxDb                  TxDb&lt;-               </div><div class="line">see <span class="string">'?methods'</span> <span class="keyword">for</span> accessing help and <span class="keyword">source</span> code</div></pre></td></tr></table></figure>
<p>在这个案例中，我们可以看到：</p>
<ol style="list-style-type: decimal">
<li><code>OrganismDb</code>是一个类(<code>class</code>)；</li>
<li><code>Homo.sapiens</code>是<code>OrganismDb</code>这个类的一个实例(instance)；</li>
<li>某个类的所有方法都可以应用到这个类的所有实例中去（<code>methods(class=class(Homo.sapiens))</code>就是查看这个类的方法）；</li>
<li>每个方法的实现是通过R中的函数进行的（我的理解就是，方法其实本质上就是函数）。并且函数的运行依赖于它自身的参数。</li>
</ol>
<p>其中需要特别注意的是juncture(交界区)的方法，即<code>genes</code>，<code>exons</code>，<code>transcripts</code>，这三个方法输出基因组基本组成成分的一些信息。上面我们列出的所有方法都可以应用于<code>OrganismDb</code>类定义的一些实例，也就是一些模式生物，例如小鼠(Mus musculus)，酒酿酵母(S. cerevisiae)和秀丽隐杆线虫(C. elegans)。</p>
<h2 id="r包模块化与持续集成continuous-integration">R包，模块化与持续集成(Continuous* *Integration)</h2>
<p>这一部分在阅读的时候可以跳过。</p>
<h3 id="r包的结构">R包的结构</h3>
<p>我们可以通过编写R代码来用R执行面向对象的函数式编程。一种基本的方法就是创建“脚本”，用于定义数据的导入和分析的基本流程。当脚本以仅定义函数和数据结构的方式编写时，我们就可以将这些脚本打包，从而方便地发布，供其他有类似数据处理问题的用户使用。</p>
<p>R软件的<a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html" target="_blank" rel="external">打包协议</a>规定了如何将R和其他语言的源代码与元数据以及文档一起组织起来，以促进简便的测试和发布的流程。例如，它义些文档的包的早期版本具有如下的目录和文件布局：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">├── DESCRIPTION  (文本文件，用于说明提供的元数据，授权信息)</div><div class="line">├── NAMESPACE    (定义输入与输出)</div><div class="line">├── R            (文件夹，包含了一些R代码)</div><div class="line">├── README.md    (可选，Github上的说明内容)</div><div class="line">├── data         (文件夹，含有示例数据)</div><div class="line">├── man          (文件夹，包的详细文档)</div><div class="line">├── tests        (文件夹，正式软件测试代码)</div><div class="line">└── vignettes    (文件夹，含有高质量的文档)</div><div class="line">    ├── biocOv1.Rmd</div><div class="line">    ├── biocOv1.html</div></pre></td></tr></table></figure>
<p>打包协议文档“Writing R Extensions”提供了完整的详细信息。使用R命令 <code>R CMD build [foldername]</code> 将对程度包文件夹的内容进行打包，从而创建可使用<code>R CMD install[archivename]</code> 添加到R安装的文件。不过现在都使用R Studio来完成这些任务。</p>
<h3 id="模块化与包的相互依赖">模块化与包的相互依赖</h3>
<p>打包协议可以帮助我们隔离执行有限操作集的软件，并识别随时间固有变化的程序集合的版本。没有客观的方法来确定一组操作是否适合打包。一些非常有用的包仅执行少量任务，而其他包则具有非常广泛的应用范围。重要的是打包的概念允许软件的模块化。这在两个方面很重要：范围和时间。范围的模块化对于允许并行独立开发解决不同问题的软件工具非常重要。时间上的模块化对于识别行为稳定的软件版本是很重要的。</p>
<h3 id="持续集成测试包的正确性和互操作性">持续集成：测试包的正确性和互操作性</h3>
<p>下列的图片就是开发Bioconductor分支的一个构建报告的<a href="http://bioconductor.org/checkResults/3.6/bioc-LATEST/" target="_blank" rel="external">快照</a>：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191105142824.png">

</div>
<p>在上面的表格中，一共有6列，其中最后1列标明了“Installed Pkgs”，其中2000表示用于Linux平台，这个数据在不同平台之间有所不同，并且通常会伴随着开发分支的时间而增加。</p>
<h2 id="汇总">汇总</h2>
<p>Bioconductor的核心开发者小组致力于开发数据结构，使用户能够方便地处理基因组和基因组规模的数据。用于设计的结构是为了支持基因组规模生物学实验的主要目标：</p>
<ul>
<li>解析由微阵列或测序仪产生的大规模数据集。</li>
<li>对（相对）原始数据进行预处理，以支持可靠的统计解释。</li>
<li>将分析量化与样本信息数据进行结合，用于检验分子过程和生物体水平特征（例如生长、疾病状态）之间关系的假设。</li>
<li>在本课程里，我们将回顾你可以在自己的研究中用于执行这些相关任务的无明和功能。</li>
</ul>
<h1 id="x的基因前提和概述">5x的基因前提和概述</h1>
<p>你知道如何使用R来操作和分析数据后，并且还对统计建模有不错的理解后。Bioconductor项目就会证明，在进行基因组级规模的计算生物操作时（但非所有），R是一个有效的工具。</p>
<p>将Bioconductor与处理基因组数据的其它软件系统的不同之处在于：</p>
<ul>
<li>使用面向对象的设计理解统一了基因组实验中出现的不同数据类型；</li>
<li>致力于基因组注释的可互操作结构，从核苷酸到人口规模；</li>
<li>发布和开发周期的持续集成原则，在多个广泛使用的计算平台上进行日常测试。我们这个为期四周的模块化学习目标旨在针对基因组规模数据分析的方方面面建立起对该系统使用的专业能力。</li>
</ul>
<p>525.5x课程主要为了四个主要部分，每部分花费一周时间来学习，内容如下：</p>
<ul>
<li>动机和技术：我们为什么要检测，以及如何检测数据，以及我们如何使用R语言来管理数据。</li>
<li>基因组注释：尤其是要注意基因组坐标中的区间（ranges，有的译为“范围”）在识别基因组结构方面的作用。</li>
<li>基因组级数据的预处理概念，重要如何通过Bioconductor来实现。</li>
<li>使用Bioconductor来做基因组级数据的假设检验。</li>
</ul>
<p>Some of the fundamental concepts that distinguish Bioconductor from other software systems addressing genome-scale data are</p>
<p>本章的不同小节将简要描述一下这些概念，以及说明如何进行计算。</p>
<h2 id="动机与技术">动机与技术</h2>
<p>“What we measure and why”这个视频（这个视频位于Youtube上）给出了基本生物过程的示意图，我们可以通过计算来研究这些过程。我们注意到，对有机体的生命过程至关重要的所有蛋白质的编码序列都位于有机体的基因组DNA中。研究生物体之间的差异， 以及生物体内的某些变化（例如肿瘤的发展），通常涉及基因组DNA序列的计算。</p>
<p>Bioconductor提供了处理许多生物体基因组DNA序列的工具。一种处理序列的基本方法计算就是给定一个“参考序列”，然后我们计算另外一条序列与参考序列之间的差异。</p>
<h3 id="参考序列的获取">参考序列的获取</h3>
<p>使用Bioconductor很容易处理人类(Homo sapiens)的参考序列。现在我们来看一下17号染色体，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">Hsapiens$chr17</div><div class="line"><span class="comment">##   81195210-letter "DNAString" instance</span></div><div class="line"><span class="comment">## seq: AAGCTTCTCACCCTGTTCCTGCATAGATAATTGC...GGTGTGGGTGTGGTGTGTGGGTGTGGGTGTGGT</span></div></pre></td></tr></table></figure>
<p>我们注意到：</p>
<ul>
<li>我们通过了一个R包来提供了序列。</li>
<li>包的名称表明了这个是来源于UCSC的参考基因组hg19。</li>
<li>我们使用美元符号 <code>$</code> 提取了染色体的序列。</li>
</ul>
<h3 id="表示dna突变体">表示DNA突变体</h3>
<p>与参考序列有单个偏离的标准表示方法就是使用VCF格式的文件（ <a href="http://samtools.github.io/hts-specs/VCFv4.1.pdf" target="_blank" rel="external">Variant Call Format</a>）。<code>VariantAnnotation</code>包就含有这样一个案例。我们有两个某些DNA突变体的两个高质量表示方法(high-level representations)，即VCF内容中的摘要，以及突变体本身在基因组上的地址，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div></pre></td><td class="code"><pre><div class="line">fl &lt;- system.file(<span class="string">"extdata"</span>, <span class="string">"ex2.vcf"</span>, package=<span class="string">"VariantAnnotation"</span>) </div><div class="line">vcf &lt;- readVcf(fl, <span class="string">"hg19"</span>)</div><div class="line">vcf</div><div class="line"><span class="comment">## class: CollapsedVCF </span></div><div class="line"><span class="comment">## dim: 5 3 </span></div><div class="line"><span class="comment">## rowRanges(vcf):</span></div><div class="line"><span class="comment">##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></div><div class="line"><span class="comment">## info(vcf):</span></div><div class="line"><span class="comment">##   DataFrame with 6 columns: NS, DP, AF, AA, DB, H2</span></div><div class="line"><span class="comment">## info(header(vcf)):</span></div><div class="line"><span class="comment">##       Number Type    Description                </span></div><div class="line"><span class="comment">##    NS 1      Integer Number of Samples With Data</span></div><div class="line"><span class="comment">##    DP 1      Integer Total Depth                </span></div><div class="line"><span class="comment">##    AF A      Float   Allele Frequency           </span></div><div class="line"><span class="comment">##    AA 1      String  Ancestral Allele           </span></div><div class="line"><span class="comment">##    DB 0      Flag    dbSNP membership, build 129</span></div><div class="line"><span class="comment">##    H2 0      Flag    HapMap2 membership         </span></div><div class="line"><span class="comment">## geno(vcf):</span></div><div class="line"><span class="comment">##   SimpleList of length 4: GT, GQ, DP, HQ</span></div><div class="line"><span class="comment">## geno(header(vcf)):</span></div><div class="line"><span class="comment">##       Number Type    Description      </span></div><div class="line"><span class="comment">##    GT 1      String  Genotype         </span></div><div class="line"><span class="comment">##    GQ 1      Integer Genotype Quality </span></div><div class="line"><span class="comment">##    DP 1      Integer Read Depth       </span></div><div class="line"><span class="comment">##    HQ 2      Integer Haplotype Quality</span></div><div class="line">rowRanges(vcf)</div><div class="line"><span class="comment">## GRanges object with 5 ranges and 5 metadata columns:</span></div><div class="line"><span class="comment">##                  seqnames             ranges strand | paramRangeID</span></div><div class="line"><span class="comment">##                     &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;factor&gt;</span></div><div class="line"><span class="comment">##        rs6054257       20 [  14370,   14370]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##     20:17330_T/A       20 [  17330,   17330]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##        rs6040355       20 [1110696, 1110696]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##   20:1230237_T/.       20 [1230237, 1230237]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##        microsat1       20 [1234567, 1234569]      * |         &lt;NA&gt;</span></div><div class="line"><span class="comment">##                             REF                ALT      QUAL      FILTER</span></div><div class="line"><span class="comment">##                  &lt;DNAStringSet&gt; &lt;DNAStringSetList&gt; &lt;numeric&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##        rs6054257              G                  A        29        PASS</span></div><div class="line"><span class="comment">##     20:17330_T/A              T                  A         3         q10</span></div><div class="line"><span class="comment">##        rs6040355              A                G,T        67        PASS</span></div><div class="line"><span class="comment">##   20:1230237_T/.              T                           47        PASS</span></div><div class="line"><span class="comment">##        microsat1            GTC             G,GTCT        50        PASS</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们需要注意：</p>
<ul>
<li>演示中的数据已经包含在了包中，它仅用于说明和测试</li>
<li>变量 <code>vcf</code> 向用户简洁地展示了信息</li>
<li>使用<code>rowRanges</code>函数提取了突变体在hg19基因组中的位置，并与标签一同显示</li>
</ul>
<h3 id="检测基因表达">检测基因表达</h3>
<p>我们将以观察模式生物酿酒酵母(Sacchomyces Cerevisiae)中基因表达的检测数值来结束这一部分的讨论。现在我们来看一个实验，这个实验检测了酿酒酵母生殖周期中一系列时间点上的全基因组mRNA丰度。我们会使用R包来管理这些数据，并且我们使用特殊的Bioconductor定义 的数据结构来访问有关的实验和结果信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(yeastCC)</div><div class="line">data(spYCCES)</div><div class="line">spYCCES</div><div class="line"><span class="comment">## ExpressionSet (storageMode: lockedEnvironment)</span></div><div class="line"><span class="comment">## assayData: 6178 features, 77 samples </span></div><div class="line"><span class="comment">##   element names: exprs </span></div><div class="line"><span class="comment">## protocolData: none</span></div><div class="line"><span class="comment">## phenoData</span></div><div class="line"><span class="comment">##   sampleNames: cln3_40 cln3_30 ... elu_390 (77 total)</span></div><div class="line"><span class="comment">##   varLabels: syncmeth time</span></div><div class="line"><span class="comment">##   varMetadata: labelDescription</span></div><div class="line"><span class="comment">## featureData: none</span></div><div class="line"><span class="comment">## experimentData: use 'experimentData(object)'</span></div><div class="line"><span class="comment">##   pubMedIds: 9843569 </span></div><div class="line"><span class="comment">## Annotation:</span></div><div class="line">experimentData(spYCCES)</div><div class="line"><span class="comment">## Experiment data</span></div><div class="line"><span class="comment">##   Experimenter name: Spellman PT </span></div><div class="line"><span class="comment">##   Laboratory: Department of Genetics, Stanford University Medical Center, Stanford, California 94306-5120, USA. </span></div><div class="line"><span class="comment">##   Contact information:  </span></div><div class="line"><span class="comment">##   Title: Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. </span></div><div class="line"><span class="comment">##   URL:  </span></div><div class="line"><span class="comment">##   PMIDs: 9843569 </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##   Abstract: A 150 word abstract is available. Use 'abstract' method.</span></div></pre></td></tr></table></figure>
<p>经过一段时间的学习，你就会在接下来的几周内成为这方面的高手，你可以将整个细胞周期中参与调控的基因随时间变化的过程绘制出来。</p>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/biointro-lkycc2-1.png" alt="plot of chunk lkycc2">
<p class="caption">plot of chunk lkycc2</p>
</div>
<p>需要注意：</p>
<ul>
<li>关于实验的元数据已经绑定到R包的数据中（通过 <code>experimentData</code>可以可看PubmedID和摘要）；</li>
<li>我们使用简单的语法就能选择这些复杂实验设计中的信息；在这个案例中，我们使用<code>spYCCES[, spYCCES$syncmeth==&quot;alpha&quot;]</code> 就能提取出控制alpha pheromone的基因；</li>
<li>R的绘图工具支持通用的绘图注释与增强功能 (plot annotation and enhancemetn)；</li>
<li>统计模型工具能够直接帮我们区分周期与非周期的基因。</li>
</ul>
<h2 id="结束语">结束语</h2>
<p>你即将学习一些关于基因组结构和检测它们的分子生物学技术的高水平讲座。当你遇到这些概念时，请你记住那些与理解结构以及数据处理过程相关的计算思路与方法。在Bioconductor中找到相应的计算工具，并熟练掌握这些工具。如果你找不到它们，请通知我们，或者是如果满足某些需求的工具不存在，你也可以选择构建它们。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://rafalab.github.io/pages/harvardx.html" target="_blank" rel="external">HarvardX Biomedical Data Science Open Online Training</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/biointro.html" target="_blank" rel="external">Bioconductor for genome-scale data – quick intro</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/02/Genomics Data Analysis Series/GDAS002-Bioconductor备忘录/" rel="next" title="GDAS002-Bioconductor备忘录">
                <i class="fa fa-chevron-left"></i> GDAS002-Bioconductor备忘录
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/04/Genomics Data Analysis Series/GDAS004-Bioconductor与基因组级数据分析简介/" rel="prev" title="GDAS004-Bioconductor与基因组级数据分析简介">
                GDAS004-Bioconductor与基因组级数据分析简介 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">249</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">118</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#r语言r包与仓库repositories"><span class="nav-number">2.</span> <span class="nav-text">R语言，R包与仓库(repositories)</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#why-r"><span class="nav-number">3.</span> <span class="nav-text">Why R?</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#函数式面向对象程序设计"><span class="nav-number">4.</span> <span class="nav-text">函数式面向对象程序设计</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#r包模块化与持续集成continuous-integration"><span class="nav-number">5.</span> <span class="nav-text">R包，模块化与持续集成(Continuous* *Integration)</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#r包的结构"><span class="nav-number">5.1.</span> <span class="nav-text">R包的结构</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#模块化与包的相互依赖"><span class="nav-number">5.2.</span> <span class="nav-text">模块化与包的相互依赖</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#持续集成测试包的正确性和互操作性"><span class="nav-number">5.3.</span> <span class="nav-text">持续集成：测试包的正确性和互操作性</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#汇总"><span class="nav-number">6.</span> <span class="nav-text">汇总</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#x的基因前提和概述"><span class="nav-number"></span> <span class="nav-text">5x的基因前提和概述</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#动机与技术"><span class="nav-number">1.</span> <span class="nav-text">动机与技术</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#参考序列的获取"><span class="nav-number">1.1.</span> <span class="nav-text">参考序列的获取</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#表示dna突变体"><span class="nav-number">1.2.</span> <span class="nav-text">表示DNA突变体</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#检测基因表达"><span class="nav-number">1.3.</span> <span class="nav-text">检测基因表达</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#结束语"><span class="nav-number">2.</span> <span class="nav-text">结束语</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">3.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共988.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
