<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="生物信息学,Linux,Ubuntu," />










<meta name="description" content="前言这个教程源于曾健明，题库地址为其博客。所使用的Linux系统是曾健明的服务器，服务器信息如下：  /usr/local/bin/miniconda3/bin路径下面安装了生物信息学软件，可以使用全路径调用它们，或者添加该目录到环境变量。 两个练手数据：2.1 转录组数据：/public/study/mRNAseq/tair/的转录组的测试数据，具体教程可以看其博客：http://www.bio">
<meta name="keywords" content="生物信息学,Linux,Ubuntu">
<meta property="og:type" content="article">
<meta property="og:title" content="Linux学习笔记">
<meta property="og:url" content="http://rvdsd.top/2018/03/28/Linux/Linux学习笔记/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言这个教程源于曾健明，题库地址为其博客。所使用的Linux系统是曾健明的服务器，服务器信息如下：  /usr/local/bin/miniconda3/bin路径下面安装了生物信息学软件，可以使用全路径调用它们，或者添加该目录到环境变量。 两个练手数据：2.1 转录组数据：/public/study/mRNAseq/tair/的转录组的测试数据，具体教程可以看其博客：http://www.bio">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/AKEdmBFIIH.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/7fA66bCmEk.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/8ki0H89cfC.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/AAeBcaD17A.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/b9EHafGmmE.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/jggF50km1I.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/6J951D225C.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/CmBJ842fB1.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/d87Hd2LjLm.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/2EhCCfda4e.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/h0kI4HLFg8.jpg?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/I2gIbJihKE.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/cf45ebk16a.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/HGHLBKhlEK.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/cbj6hmgGGB.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/KDd1bmlk0F.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/096904lAFg.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/2hChiFB4FK.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/hE93c3HKlH.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/f5K40hH9mh.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/Ih5GL0fIAb.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/20jK2DK3Kl.png?imageslim">
<meta property="og:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180329/gHHAJ0II9A.png?imageslim">
<meta property="og:updated_time" content="2018-05-11T00:21:36.542Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Linux学习笔记">
<meta name="twitter:description" content="前言这个教程源于曾健明，题库地址为其博客。所使用的Linux系统是曾健明的服务器，服务器信息如下：  /usr/local/bin/miniconda3/bin路径下面安装了生物信息学软件，可以使用全路径调用它们，或者添加该目录到环境变量。 两个练手数据：2.1 转录组数据：/public/study/mRNAseq/tair/的转录组的测试数据，具体教程可以看其博客：http://www.bio">
<meta name="twitter:image" content="http://ortxpw68f.bkt.clouddn.com/professional/180328/AKEdmBFIIH.png?imageslim">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2018/03/28/Linux/Linux学习笔记/"/>





  <title>Linux学习笔记 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">学习过程中的输出</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2018/03/28/Linux/Linux学习笔记/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Linux学习笔记</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2018-03-28T12:00:00+08:00">
                2018-03-28
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Linux/" itemprop="url" rel="index">
                    <span itemprop="name">Linux</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,640
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h2><p>这个教程源于曾健明，<a href="http://www.bio-info-trainee.com/2900.html" target="_blank" rel="external">题库地址为其博客</a>。所使用的Linux系统是曾健明的服务器，服务器信息如下：</p>
<ol>
<li><code>/usr/local/bin/miniconda3/bin</code>路径下面安装了生物信息学软件，可以使用全路径调用它们，或者添加该目录到环境变量。</li>
<li>两个练手数据：<br>2.1 转录组数据：<code>/public/study/mRNAseq/tair/</code>的转录组的测试数据，具体教程可以看其博客：<a href="http://www.bio-info-trainee.com/2809.html" target="_blank" rel="external">http://www.bio-info-trainee.com/2809.html</a>，可以用<code>ls -lh  --color=auto /public/study/mRNAseq/tair/</code> 命令查看，然后所有的处理，在自己的目录处理，不需要拷贝那些数据.<br>2.2 找变异：<code>/public/study/variant-calling/tair</code>。<br>可以参见教程：<a href="http://www.biotrainee.com/thread-696-1-1.html" target="_blank" rel="external">http://www.biotrainee.com/thread-696-1-1.html</a></li>
</ol>
<p>其他的NGS流程，参见Github：<a href="https://github.com/jmzeng1314/NGS-pipeline" target="_blank" rel="external">https://github.com/jmzeng1314/NGS-pipeline</a>。</p>
<h2 id="基础知识"><a href="#基础知识" class="headerlink" title="基础知识"></a>基础知识</h2><h3 id="基础知识-1"><a href="#基础知识-1" class="headerlink" title="基础知识"></a>基础知识</h3><h3 id="文件目录操作"><a href="#文件目录操作" class="headerlink" title="文件目录操作"></a>文件目录操作</h3><ol>
<li>查看当前目录<code>pwd</code><br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/AKEdmBFIIH.png?imageslim" alt="mark"></li>
</ol>
<h3 id="系统管理"><a href="#系统管理" class="headerlink" title="系统管理"></a>系统管理</h3><p>1.查看当前登录用户<code>w</code><br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/7fA66bCmEk.png?imageslim" alt="mark"></p>
<p>2.查看命令历史<code>l~/.bash_history</code><br>Bash shell在“~/.bash_history”（“~/”表示用户目录）文件中保存了500条使用过的命令，这样可以使你输入使用过的长命令变得容易。每个在系统中拥有账号的用户在他的目录下都有一个“.bash_history”文件。</p>
<h3 id="用户权限"><a href="#用户权限" class="headerlink" title="用户权限"></a>用户权限</h3><h3 id="文本操作"><a href="#文本操作" class="headerlink" title="文本操作"></a>文本操作</h3><h2 id="题库测试"><a href="#题库测试" class="headerlink" title="题库测试"></a>题库测试</h2><p><strong>1. 在任意文件夹下面创建形如 1/2/3/4/5/6/7/8/9 格式的文件夹系列。</strong></p>
<p><code>mkdir -p 1/2/3/4/5/6/7/8/9</code><br>mkdir是建立文件夹命名，添加<code>-p</code>参数可以直接创建多级文件夹，如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/8ki0H89cfC.png?imageslim" alt="mark"></p>
<p><strong>2.在创建好的文件夹下面，比如我的是 /Users/jimmy/</strong>tmp/1/2/3/4/5/6/7/8/9 ，里面创建文本文件 me.txt</p>
<p>先切换到/1/2/3/4/5/6/7/8/9，直接输入<code>cd</code>，然后不停地按<code>Tab</code>键，就会进入下一层文件夹。<br><code>touch me.txt</code>创建txt文件。<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/AAeBcaD17A.png?imageslim" alt="mark"></p>
<p><strong>3.在文本文件 me.txt 里面输入内容:</strong><br>Go to: <a href="http://www.biotrainee.com/" target="_blank" rel="external">http://www.biotrainee.com/</a><br>I love bioinfomatics.<br>And you ?</p>
<p>输入<code>vim me.txt</code>，输入上述文字，然后按<code>Esc</code>键，输入<code>wp</code>退出。</p>
<p><strong>4. 删除上面创建的文件夹 1/2/3/4/5/6/7/8/9 及文本文件 me.txt</strong></p>
<p><code>rm -rf *</code></p>
<p><strong>5. 在任意文件夹下面创建 folder1~5这5个文件夹，然后每个文件夹下面继续创建 folder1~5这5个文件夹，效果如下：</strong></p>
<p><code>mkdir -p folder_{1..5}/folder_{1..5}</code><br>如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/b9EHafGmmE.png?imageslim" alt="mark"></p>
<p><strong>6. 在第五题创建的每一个文件夹下面都 创建第二题文本文件 me.txt ，内容也要一样。</strong></p>
<p>先创建一个tmp3目录，其目录下有5个文件夹，5个文件夹下面又有5个文件夹，然后将me.txt拷贝到这25个文件夹中：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">mkdir -p tmp3/folder_&#123;1..5&#125;/folder_&#123;1..5&#125;</div><div class="line">echo tmp3/folder_&#123;1..5&#125;/folder_&#123;1..5&#125; | xargs -n 1 cp -v me.txt</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/jggF50km1I.png?imageslim" alt="mark"></p>
<p><strong>7. 再次删除掉前面几个步骤建立的文件夹及文件。</strong><br><code>rm -rf *</code></p>
<p><strong>8. 下载 <a href="http://www.biotrainee.com/jmzeng/igv/test.bed" target="_blank" rel="external">http://www.biotrainee.com/jmzeng/igv/test.bed</a> 文件，后在里面选择含有 H3K4me3 的那一行是第几行，该文件总共有几行。</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">wget http://www.biotrainee.com/jmzeng/igv/test.bed</div><div class="line">grep -n &quot;H3K4me3&quot; test.bed</div><div class="line">wc test.bed</div></pre></td></tr></table></figure>
<p>结果如下：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/6J951D225C.png?imageslim" alt="mark"><br>可知，H3K4me3在第8行，用Notepad++打开的话，核对一下，确实如此。<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/CmBJ842fB1.png?imageslim" alt="mark"><br>用<code>wc</code>命令查看有多少行，一共10行。</p>
<p><strong>9. 下载 <a href="http://www.biotrainee.com/jmzeng/rmDuplicate.zip" target="_blank" rel="external">http://www.biotrainee.com/jmzeng/rmDuplicate.zip</a> 文件，并且解压，查看里面的文件夹结构</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">wget http://www.biotrainee.com/jmzeng/rmDuplicate.zip</div><div class="line">unzip rmDuplicate.zip</div><div class="line">cd rmDuplicate.zip</div></pre></td></tr></table></figure>
<p>查看文件：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/d87Hd2LjLm.png?imageslim" alt="mark"></p>
<p><strong>10. 打开第九题解压的文件，进入 rmDuplicate/samtools/single 文件夹里面，查看后缀为 .sam 的文件，搞清楚 生物信息学里面的SAM/BAM 定义是什么。</strong><br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/2EhCCfda4e.png?imageslim" alt="mark"></p>
<p><strong>SAM文件与BAM文件</strong></p>
<blockquote>
<p>SAM是一种序列比对格式标准， 由sanger制定，是以TAB为分割符的文本格式。主要应用于测序序列mapping到基因组上的结果表示，当然也可以表示任意的多重比对结果。</p>
<p>SAM要处理好的问题：</p>
<p>非常多序列（read)，mapping到多个参考基因组（reference）上；<br>同一条序列，分多段（segment）比对到参考基因组上；<br>无限量的，结构化信息表示，包括错配、删除、插入等比对信息；<br>SAM分为两部分，注释信息（header section）和比对结果部分（alignment section），注释信息可有可无，都是以@开头，用不同的tag表示不同的信息，主要有@HD，说明符合标准的版本、对比序列的排列顺序；@SQ，参考序列说明；@RG，比对上的序列（read）说明；@PG，使用的程序说明；@CO，任意的说明信息。</p>
</blockquote>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/h0kI4HLFg8.jpg?imageslim" alt="mark"></p>
<p>&gt;<br>比对结果部分（alignment section），每一行表示一个片段（segment）的比对信息，包括11个必须的字段（mandatory fields）和一个可选的字段，字段之间用tag分割。必须的字段有11个，顺序固定，不可用时，根据字段定义，可以为’0‘或者’*‘，这是11个字段包括：</p>
<blockquote>
<p>QNAME，比对片段的（template）的编号；<br>FLAG，位标识，template mapping情况的数字表示，每一个数字代表一种比对情况，这里的值是符合情况的数字相加总和；<br>RNAME，参考序列的编号，如果注释中对SQ-SN进行了定义，这里必须和其保持一致，另外对于没有mapping上的序列，这里是’<em>‘；<br>POS，比对上的位置，注意是从1开始计数，没有比对上，此处为0；<br>MAPQ，mappint的质量；<br>CIGAR，简要比对信息表达式（Compact Idiosyncratic Gapped Alignment Report），其以参考序列为基础，使用数字加字母表示比对结果，比如3S6M1P1I4M，前三个碱基被剪切去除了，然后6个比对上了，然后打开了一个缺口，有一个碱基插入，最后是4个比对上了，是按照顺序的；<br>RNEXT，下一个片段比对上的参考序列的编号，没有另外的片段，这里是’</em>‘，同一个片段，用’=‘；<br>PNEXT，下一个片段比对上的位置，如果不可用，此处为0；<br>TLEN，Template的长度，最左边得为正，最右边的为负，中间的不用定义正负，不分区段（single-segment)的比对上，或者不可用时，此处为0；<br>SEQ，序列片段的序列信息，如果不存储此类信息，此处为’*‘，注意CIGAR中M/I/S/=/X对应数字的和要等于序列长度；<br>QUAL，序列的质量信息，格式同FASTQ一样。<br>可选字段（optional fields)，格式如：TAG:TYPE:VALUE，其中TAG有两个大写字母组成，每个TAG代表一类信息，每一行一个TAG只能出现一次，TYPE表示TAG对应值的类型，可以是字符串、整数、字节、数组等。</p>
</blockquote>
<p><strong>11. 安装 samtools 软件</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">mkdir soft</div><div class="line">cd soft</div><div class="line">wget https://github.com/samtools/samtools/releases/download/1.7/samtools-1.7.tar.bz2</div><div class="line">tar jxvf samtools-1.7.tar.bz2</div><div class="line">cd samtools-1.7/</div><div class="line">./configure --prefix=/trainee/home/jlzhang/biosoft</div><div class="line">make</div><div class="line">make install</div><div class="line">export PATH=$PATH:/trainee/home/jlzhang/biosoft/bin:$PATH &gt;&gt;~/.bashrc</div><div class="line">source ~/.bashrc</div></pre></td></tr></table></figure>
<p>可以通过<code>echo $PATH</code>来查看用户变量。</p>
<p><strong>12. 打开 后缀为BAM 的文件，找到产生该文件的命令。</strong></p>
<p>用samtools view打开BAM文件，过程如下：</p>
<p>进入相应文件夹，即<code>samtools view tmp.rmdup.bam | head</code><br>输入命令：<code>samtools view tmp.rmdup.bam | head</code><br>打开后的BAM文件如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180328/I2gIbJihKE.png?imageslim" alt="mark"></p>
<p><strong>13. 根据上面的命令，找到我使用的参考基因组 /home/jianmingzeng/reference/index/bowtie/hg38 具体有多少条染色体。</strong><br>从网上找到的信息，hg38含有24条染色体，统计代码如下所示：</p>
<p>先找到hg38.fa文件，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">locate hg38.fa</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/cf45ebk16a.png?imageslim" alt="mark"><br>随便挑一个进行统计，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">grep &quot;chr[1-9,X,Y]&quot; /umac/reference/genome/hg38/hg38.fa| cut -d_ -f1 &gt; statistic.chrom</div><div class="line">sort statistic.chrom | uniq | wc</div></pre></td></tr></table></figure>
<p>可以发现，hg38有24条染色体，如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/HGHLBKhlEK.png?imageslim" alt="mark"></p>
<p>其中,sort是对文本文件中的字符进行排序，uniq是去掉重复的行，但是uniq只能去掉相邻的重复行，因此之前要进行sort。</p>
<p><strong>14. 上面的后缀为BAM 的文件的第二列，只有 0 和 16 两个数字，用 cut/sort/uniq等命令统计它们的个数。</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">samtools view tmp.rmdup.bam &gt; bamfile</div><div class="line">awk &apos;&#123;print $2&#125;&apos; bamfile &gt; zero_one</div><div class="line">grep -o &quot;0&quot; zero_one |wc -l</div><div class="line">grep -o &quot;16&quot; zero_one |wc -l</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/cbj6hmgGGB.png?imageslim" alt="mark"></p>
<p><strong>15. 重新打开 rmDuplicate/samtools/paired 文件夹下面的后缀为BAM 的文件，再次查看第二列，并且统计</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">samtools view tmp.rmdup.bam &gt; pairbam</div><div class="line">awk &apos;&#123;print $2&#125;&apos; pairbam &gt; number </div><div class="line">awk &apos;&#123;print $2&#125;&apos; pairbam |sort|uniq -c</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/KDd1bmlk0F.png?imageslim" alt="mark"></p>
<p><strong>16. 下载 <a href="http://www.biotrainee.com/jmzeng/sickle/sickle-results.zip" target="_blank" rel="external">http://www.biotrainee.com/jmzeng/sickle/sickle-results.zip</a> 文件，并且解压，查看里面的文件夹结构， 这个文件有2.3M，注意留心下载时间及下载速度。</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">wget http://www.biotrainee.com/jmzeng/sickle/sickle-results.zip</div><div class="line">ls -lh</div><div class="line">unzip sickle-results.zip</div><div class="line">cd sickle-results</div><div class="line">ls -lh</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/096904lAFg.png?imageslim" alt="mark"></p>
<p><strong>17. 解压 sickle-results/single_tmp_fastqc.zip 文件，并且进入解压后的文件夹，找到 fastqc_data.txt 文件，并且搜索该文本文件以 &gt;&gt;开头的有多少行？</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">cd sickle-results</div><div class="line">unzip single_tmp_fastqc.zip</div><div class="line">cd single_tmp_fastqc</div><div class="line">grep &quot;&gt;&gt;&quot;  fastqc_data.txt |wc -l</div></pre></td></tr></table></figure>
<p>一共是24行<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/2hChiFB4FK.png?imageslim" alt="mark"></p>
<p><strong>18. 下载 <a href="http://www.biotrainee.com/jmzeng/tmp/hg38.tss" target="_blank" rel="external">http://www.biotrainee.com/jmzeng/tmp/hg38.tss</a> 文件，去NCBI找到TP53/BRCA1等自己感兴趣的基因对应的 refseq数据库 ID，然后找到它们的hg38.tss 文件的哪一行。<br><a href="https://www.ncbi.nlm.nih.gov/gene/7157" target="_blank" rel="external">https://www.ncbi.nlm.nih.gov/gene/7157</a></strong></p>
<ol>
<li>打开PUBMED，选GENE，输入“TP53”，点击Search，如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/hE93c3HKlH.png?imageslim" alt="mark"></li>
</ol>
<p>Pubmed中有很多TP53的信息，但是hg38是人类基因组，因此选择第一行的TP53就行，进入即可，找到Go to reference sequence details，点进去即可，如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/f5K40hH9mh.png?imageslim" alt="mark"></p>
<p>找到Go to reference sequence details，进去，其中有NM开头的编号，NM表示mRNA的RefSeq，TP53的就是NM_00054635（小数点部分是版本号，不用管），在hg38.tss中查询即可，BRCA1按照同样操作，查询结果如下所示：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">wget http://www.biotrainee.com/jmzeng/tmp/hg38.tss</div><div class="line">grep &quot;NM_000546&quot; hg38.tss </div><div class="line">grep &quot;NM_005905&quot; hg38.tss</div></pre></td></tr></table></figure></p>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/Ih5GL0fIAb.png?imageslim" alt="mark"></p>
<p><strong>19. 解析hg38.tss 文件，统计每条染色体的基因个数。</strong></p>
<p>hg38.tss文件是基因的坐标，查看了一下文件，文件内容如下所示：<br><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/20jK2DK3Kl.png?imageslim" alt="mark"></p>
<p>第1列是基因名称，第2列是染色体名称，第3列到第4列是转录起止坐标，第5列不清楚。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">awk &apos;$5 &gt; 0&apos; hg38.tss &gt; gene2</div><div class="line">awk &apos;&#123;print $2&#125;&apos; gene2 &gt; gene</div><div class="line">cut -d_ -f1 gene |sort |uniq -c</div></pre></td></tr></table></figure>
<p><img src="http://ortxpw68f.bkt.clouddn.com/professional/180329/gHHAJ0II9A.png?imageslim" alt="mark"></p>
<p><strong>20. 解析hg38.tss 文件，统计NM和NR开头的序列，了解NM和NR开头的含义。</strong></p>
<p>没在hg38.tss文件里找到相应的序列，只找到了名称，提取NM和NR开头的名称命令如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">grep &quot;NM&quot; hg38.tss</div><div class="line">grep &quot;NR&quot; hg38.tss</div></pre></td></tr></table></figure>
<p>NM，mRNA mixed，转录组产物序列；成熟mRNA转录本序列。<br>NR，RNA mixed，非编码的转录子序列，包括结构RNAs，假基因转子等。</p>
<p>参考资料：</p>
<ol>
<li><a href="http://boyun.sh.cn/bio/?p=1890" target="_blank" rel="external">博耘生物</a></li>
<li><a href="http://blog.sina.com.cn/s/blog_72512a1d0102xp6f.html" target="_blank" rel="external">NCBI中编号/编码说明</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/生物信息学/" <i class="fa fa-tag"></i> 生物信息学</a>
          
            <a href="/tags/Linux/" <i class="fa fa-tag"></i> Linux</a>
          
            <a href="/tags/Ubuntu/" <i class="fa fa-tag"></i> Ubuntu</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/03/20/BioInformation/生信-实战-生信技能树转录组数据分析入门实战/" rel="next" title="转录组数据分析入门实战">
                <i class="fa fa-chevron-left"></i> 转录组数据分析入门实战
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/03/30/BioInformation/生信-有关测序的数据储存问题/" rel="prev" title="有关测序的数据储存问题">
                有关测序的数据储存问题 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">108</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">10</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">103</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#基础知识"><span class="nav-number">2.</span> <span class="nav-text">基础知识</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#基础知识-1"><span class="nav-number">2.1.</span> <span class="nav-text">基础知识</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#文件目录操作"><span class="nav-number">2.2.</span> <span class="nav-text">文件目录操作</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#系统管理"><span class="nav-number">2.3.</span> <span class="nav-text">系统管理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#用户权限"><span class="nav-number">2.4.</span> <span class="nav-text">用户权限</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#文本操作"><span class="nav-number">2.5.</span> <span class="nav-text">文本操作</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#题库测试"><span class="nav-number">3.</span> <span class="nav-text">题库测试</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_uv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共381.4k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  










  
  <script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/valine@1.1.4/dist/Valine.min.js"></script>
  <script type="text/javascript">
    new Valine({
        av: AV,
        el: '#vcomments' ,
        verify: true,
        notify: true,
        app_id: '4ORLDMxXCFDLSHPJ4wSSDkkw-gzGzoHsz',
        app_key: 'HSWpCY7XlN0pGV5DCrKoC0ic',
        placeholder: 'Comment input placeholder'
    });
  </script>



  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
