<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言 以前在网上看到了哈佛大学的PH525x系列课程，这个系列教程主要是生物信息学相关的内容，一共是17章。这17章内容可以大致分为三部分。 第一部分：第1章到10章主要是讲的与生物信息学有关的统计学基础，参考书是《Data Analysis for the life sciences》，第1章到第10章的笔记已经做完。 第二部分：第11章到第17章，涉及生物信息学的应用部分，例如Bioco">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="PH525x01-Bioconductor备忘录">
<meta property="og:url" content="http://rvdsd.top/2019/09/01/PH525x/PH525x01-Bioconductor备忘录/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 以前在网上看到了哈佛大学的PH525x系列课程，这个系列教程主要是生物信息学相关的内容，一共是17章。这17章内容可以大致分为三部分。 第一部分：第1章到10章主要是讲的与生物信息学有关的统计学基础，参考书是《Data Analysis for the life sciences》，第1章到第10章的笔记已经做完。 第二部分：第11章到第17章，涉及生物信息学的应用部分，例如Bioco">
<meta property="og:locale" content="zh-Hans">
<meta property="og:updated_time" content="2019-10-28T05:24:08.243Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="PH525x01-Bioconductor备忘录">
<meta name="twitter:description" content="前言 以前在网上看到了哈佛大学的PH525x系列课程，这个系列教程主要是生物信息学相关的内容，一共是17章。这17章内容可以大致分为三部分。 第一部分：第1章到10章主要是讲的与生物信息学有关的统计学基础，参考书是《Data Analysis for the life sciences》，第1章到第10章的笔记已经做完。 第二部分：第11章到第17章，涉及生物信息学的应用部分，例如Bioco">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/01/PH525x/PH525x01-Bioconductor备忘录/"/>





  <title>PH525x01-Bioconductor备忘录 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/01/PH525x/PH525x01-Bioconductor备忘录/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">PH525x01-Bioconductor备忘录</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-01T12:00:00+08:00">
                2019-09-01
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/PH525x/" itemprop="url" rel="index">
                    <span itemprop="name">PH525x</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,609
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  13
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="前言">前言</h2>
<p>以前在网上看到了哈佛大学的<a href="http://genomicsclass.github.io/book/" target="_blank" rel="external">PH525x系列课程</a>，这个系列教程主要是生物信息学相关的内容，一共是17章。这17章内容可以大致分为三部分。</p>
<p>第一部分：第1章到10章主要是讲的与生物信息学有关的统计学基础，参考书是《Data Analysis for the life sciences》，<a href="http://rvdsd.top/categories/Data-Analysis-for-the-life-sciences/">第1章到第10章</a>的笔记已经做完。</p>
<p>第二部分：第11章到第17章，涉及生物信息学的应用部分，例如Bioconductor上的几个包使用等。</p>
<p>第三部分：一些使用案例。</p>
<p>这篇笔记是这是第二部分的第一篇笔记，主要内容是关于Bioconductor的一些备忘录(cheat sheet)，作者总结的不错，我就按原文翻译了下来，由于一些知识点并没有学到，因此未翻译，后面学到后，再回头过来翻译。</p>
<h2 id="安装">安装</h2>
<p>安装细节参考http://bioconductor.org/install/，具体代码如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">if (!requireNamespace(&quot;BiocManager&quot;))</div><div class="line">    install.packages(&quot;BiocManager&quot;)</div><div class="line">BiocManager::install()</div><div class="line">BiocManager::install(c(&quot;package1&quot;,&quot;package2&quot;)</div><div class="line">BiocManager::valid() # are packages up to date?</div><div class="line"></div><div class="line"># what Bioc version is release right now?</div><div class="line">http://bioconductor.org/bioc-version</div><div class="line"># what Bioc versions are release/devel?</div><div class="line">http://bioconductor.org/js/versions.js</div></pre></td></tr></table></figure>
<h2 id="r语言帮助">R语言帮助</h2>
<p>简单的帮忙如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">?函数名</div><div class="line">?&quot;eSet-class&quot; # 如果要查看类的帮助，需要后缀&apos;-class&apos;</div><div class="line">help(package=&quot;foo&quot;,help_type=&quot;html&quot;) # 使用web浏览器来查看帮助文档</div><div class="line">vignette(&quot;topic&quot;)</div><div class="line">browseVignettes(package=&quot;package&quot;) # 某包的简单案例</div></pre></td></tr></table></figure>
<p>高级用户的帮助功能</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">functionName # 直接输入函数名（不带括号），则显示函数代码</div><div class="line">getMethod(method,&quot;class&quot;)  # 显示某类的方法代码</div><div class="line">selectMethod(method, &quot;class&quot;) # 查看继承来的方法</div><div class="line">showMethods(classes=&quot;class&quot;) # 显示某类的所有方法show all methods for class</div><div class="line">methods(class=&quot;GRanges&quot;) # 此函数限于R &gt;=3.2</div><div class="line">?&quot;functionName,class-method&quot; # 用于查看S4对象的方法文档</div><div class="line">?&quot;plotMA,data.frame-method&quot; # 需要加载geneplotter包，from library(geneplotter)</div><div class="line">?&quot;method.class&quot; # 查看S3对象的方法 </div><div class="line">?&quot;plot.lm&quot;</div><div class="line">sessionInfo() # 获取R语言版本的相关帮助信息 </div><div class="line">packageVersion(&quot;foo&quot;) # 查看某个包的版本</div></pre></td></tr></table></figure>
<p>Bioconductor相关的论坛： https://support.bioconductor.org</p>
<p>如果你使用的Rstudio，则可以使用<code>?</code>或<code>help</code>来查看帮忙文档，如果使用是命令行模式（例如Linux），则可以使用下面的代码通过浏览器来查看帮助文档（我使用的Rstudio，因此未尝试以下代码）：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">alias rhelp=&quot;Rscript -e &apos;args &lt;- commandArgs(TRUE); help(args[2], package=args[3], help_type=\&quot;html\&quot;); Sys.sleep(5)&apos; --args&quot;</div></pre></td></tr></table></figure>
<h2 id="debugging-r">debugging R</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">traceback() # 查看哪一步导致了R错误 </div><div class="line"># debug a function</div><div class="line">debug(myFunction) # 逐行遍历函数中的代码 </div><div class="line">undebug(myFunction) # 停止debugging</div><div class="line">debugonce(myFunction) # 功能如上，但是不需要undebug()</div><div class="line"># 在写代码的过程中，可以将函数browser()放到一个关系的节点上</div><div class="line"># this plus devtools::load_all() can be useful for programming</div><div class="line"># to jump in function on error:</div><div class="line">options(error=recover)</div><div class="line"># turn that behavior off:</div><div class="line">options(error=NULL)</div><div class="line"># debug, e.g. estimateSizeFactors from DESeq2...</div><div class="line"># debugging an S4 method is more difficult; this gives you a peek inside:</div><div class="line">trace(estimateSizeFactors, browser, exit=browser, signature=&quot;DESeqDataSet&quot;)</div></pre></td></tr></table></figure>
<h2 id="显示某个类的包特异方法">显示某个类的包特异方法</h2>
<p>以下两个段代码实现的功能大致相同：查看特定的某个包中，某个特定类对象的操作方法。给定类的对象的操作方法。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">intersect(sapply(strsplit(as.character(methods(class=&quot;DESeqDataSet&quot;)), &quot;,&quot;), `[`, 1), ls(&quot;package:DESeq2&quot;))</div><div class="line">sub(&quot;Function: (.*) \\(package .*\\)&quot;,&quot;\\1&quot;,grep(&quot;Function&quot;,showMethods(classes=&quot;DESeqDataSet&quot;, </div><div class="line">    where=getNamespace(&quot;DESeq2&quot;), printTo=FALSE), value=TRUE))</div></pre></td></tr></table></figure>
<h2 id="注释annotations">注释Annotations</h2>
<p>这里需要先更新一下AnnotationHub，现在我们查看一下AnnotationHub的某个案例：</p>
<p>https://master.bioconductor.org/packages/release/bioc/html/AnnotationHub.html</p>
<p>以下代码是如何使用数据库包，以及<code>biomart</code>，先看<a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html" target="_blank" rel="external">AnnotationDbi</a>：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"># 某用包个注释包</div><div class="line">library(AnnotationDbi)</div><div class="line">library(org.Hs.eg.db) # 这个是人类注释包，即Homo.sapiens</div><div class="line">columns(org.Hs.eg.db)</div><div class="line">keytypes(org.Hs.eg.db) # columns与keytypes返回的结果一样，即有什么样的注释内容</div><div class="line">k &lt;- head(keys(org.Hs.eg.db, keytype=&quot;ENTREZID&quot;))</div><div class="line"></div><div class="line"># 返回一个字符向量（1对1的关系），查看mapIds则可以查看多个选项</div><div class="line">res &lt;- mapIds(org.Hs.eg.db, keys=k, column=&quot;ENSEMBL&quot;, keytype=&quot;ENTREZID&quot;)</div><div class="line"></div><div class="line"># 产生1对多的映射关系 </div><div class="line">res &lt;- select(org.Hs.eg.db, keys=k,</div><div class="line">  columns=c(&quot;ENTREZID&quot;,&quot;ENSEMBL&quot;,&quot;SYMBOL&quot;),</div><div class="line">  keytype=&quot;ENTREZID&quot;)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html" target="_blank" rel="external">biomaRt</a></p>
<p>这个包主要用于各种基因ID的转换。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line"># 使用biomart包，将一个注释与另外一个注释进行映射</div><div class="line">library(biomaRt)</div><div class="line">m &lt;- useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)</div><div class="line">map &lt;- getBM(mart = m,</div><div class="line">  attributes = c(&quot;ensembl_gene_id&quot;, &quot;entrezgene&quot;),</div><div class="line">  filters = &quot;ensembl_gene_id&quot;, </div><div class="line">  values = some.ensembl.genes)</div></pre></td></tr></table></figure>
<h2 id="genomicranges包">GenomicRanges包</h2>
<p><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" target="_blank" rel="external">GenomicRanges</a>这个包用于分析高通量测序数据，它能有效地表壳与操作基因组注释信息(genomic annotations)与比对结果(alignments)。GenomicRanges包定义了一个通用容器，此容器用于储存与操作基因间隔(genomic intervals)与变量(variables)。在GenomicAlignments包和SummarizedExperiment包中定义了一个更加特殊的容器，GenomicAlignments包中的这个更加特殊的容器可以根据一个参考基因组来表示和操作一些短比对结果(short alignments)，SummarizedExperiment包中容器可以构建一个实验的矩阵样汇总信息。</p>
<p>此包的使用如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">library(GenomicRanges)</div><div class="line">z &lt;- GRanges(&quot;chr1&quot;,IRanges(1000001,1001000),strand=&quot;+&quot;)</div><div class="line">start(z)</div><div class="line">end(z)</div><div class="line">width(z)</div><div class="line">strand(z)</div><div class="line">mcols(z) # the &apos;metadata columns&apos;, any information stored alongside each range</div><div class="line">ranges(z) # gives the IRanges</div><div class="line">seqnames(z) # the chromosomes for each ranges</div><div class="line">seqlevels(z) # the possible chromosomes</div><div class="line">seqlengths(z) # the lengths for each chromosome</div></pre></td></tr></table></figure>
<h3 id="intra-range-methods">Intra-range methods</h3>
<p>不清楚功能，后面补充。</p>
<p>Affects ranges independently</p>
<table>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>shift</td>
<td>moves left/right</td>
</tr>
<tr class="even">
<td>narrow</td>
<td>narrows by relative position within range</td>
</tr>
<tr class="odd">
<td>resize</td>
<td>resizes to width, fixing start for +, end for -</td>
</tr>
<tr class="even">
<td>flank</td>
<td>returns flanking ranges to the left +, or right -</td>
</tr>
<tr class="odd">
<td>promoters</td>
<td>similar to flank</td>
</tr>
<tr class="even">
<td>restrict</td>
<td>restricts ranges to a start and end position</td>
</tr>
<tr class="odd">
<td>trim</td>
<td>trims out of bound ranges</td>
</tr>
<tr class="even">
<td>+/-</td>
<td>expands/contracts by adding/subtracting fixed amount</td>
</tr>
<tr class="odd">
<td>*</td>
<td>zooms in (positive) or out (negative) by multiples</td>
</tr>
</tbody>
</table>
<h3 id="inter-range-methods">Inter-range methods</h3>
<p>Affects ranges as a group</p>
<table>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>range</td>
<td>one range, leftmost start to rightmost end</td>
</tr>
<tr class="even">
<td>reduce</td>
<td>cover all positions with only one range</td>
</tr>
<tr class="odd">
<td>gaps</td>
<td>uncovered positions within range</td>
</tr>
<tr class="even">
<td>disjoin</td>
<td>breaks into discrete ranges based on original starts/ends</td>
</tr>
</tbody>
</table>
<h3 id="nearest-methods">Nearest methods</h3>
<p>Given two sets of ranges, <code>x</code> and <code>subject</code>, for each range in <code>x</code>, returns…</p>
<table style="width:100%;">
<colgroup>
<col width="22%">
<col width="77%">
</colgroup>
<thead>
<tr class="header">
<th>function</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nearest</td>
<td>index of the nearest neighbor range in subject</td>
</tr>
<tr class="even">
<td>precede</td>
<td>index of the range in subject that is directly preceded by the range in x</td>
</tr>
<tr class="odd">
<td>follow</td>
<td>index of the range in subject that is directly followed by the range in x</td>
</tr>
<tr class="even">
<td>distanceToNearest</td>
<td>distances to its nearest neighbor in subject (Hits object)</td>
</tr>
<tr class="odd">
<td>distance</td>
<td>distances to nearest neighbor (integer vector)</td>
</tr>
</tbody>
</table>
<p>A Hits object can be accessed with <code>queryHits</code>, <code>subjectHits</code> and <code>mcols</code> if a distance is associated.</p>
<h3 id="set-methods">set methods</h3>
<p>If <code>y</code> is a GRangesList, then use <code>punion</code>, etc. All functions have default <code>ignore.strand=FALSE</code>, so are strand specific.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">union(x,y) </div><div class="line">intersect(x,y)</div><div class="line">setdiff(x,y)</div></pre></td></tr></table></figure>
<h3 id="overlaps">Overlaps</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">x %over% y  # logical vector of which x overlaps any in y</div><div class="line">fo &lt;- findOverlaps(x,y) # returns a Hits object</div><div class="line">queryHits(fo)   # which in x</div><div class="line">subjectHits(fo) # which in y</div></pre></td></tr></table></figure>
<h3 id="seqnames-and-seqlevels">Seqnames and seqlevels</h3>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html" target="_blank" rel="external">GenomicRanges</a> and <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html" target="_blank" rel="external">GenomeInfoDb</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">gr.sub &lt;- gr[seqlevels(gr) == &quot;chr1&quot;]</div><div class="line">seqlevelsStyle(x) &lt;- &quot;UCSC&quot; # convert to &apos;chr1&apos; style from &quot;NCBI&quot; style &apos;1&apos;</div></pre></td></tr></table></figure>
<h2 id="sequences">Sequences</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Biostrings.html" target="_blank" rel="external">Biostrings</a>是一个操作序列或序列集的包，简要操作可以参考<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/Biostrings/inst/doc/BiostringsQuickOverview.pdf" target="_blank" rel="external">Biostrings Quick Overview PDF</a>，关于一些物种的，缩写的信息可以参考<a href="http://genomicsclass.github.io/book/pages/annoCheat.html" target="_blank" rel="external">cheat sheet for annotation</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">dnastringset &lt;- getSeq(Hsapiens, granges) # returns a DNAStringSet</div><div class="line"># also Views() for Bioconductor &gt;= 3.1</div></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(Biostrings)</div><div class="line">dnastringset &lt;- readDNAStringSet(&quot;transcripts.fa&quot;)</div></pre></td></tr></table></figure>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">substr(dnastringset, 1, 10) # to character string</div><div class="line">subseq(dnastringset, 1, 10) # returns DNAStringSet</div><div class="line">Views(dnastringset, 1, 10) # lightweight views into object</div><div class="line">complement(dnastringset)</div><div class="line">reverseComplement(dnastringset)</div><div class="line">matchPattern(&quot;ACGTT&quot;, dnastring) # also countPattern, also works on Hsapiens/genome</div><div class="line">vmatchPattern(&quot;ACGTT&quot;, dnastringset) # also vcountPattern</div><div class="line">letterFrequecy(dnastringset, &quot;CG&quot;) # how many C&apos;s or G&apos;s</div><div class="line"># also letterFrequencyInSlidingView</div><div class="line">alphabetFrequency(dnastringset, as.prob=TRUE)</div><div class="line"># also oligonucleotideFrequency, dinucleotideFrequency, trinucleotideFrequency</div><div class="line"># transcribe/translate for imitating biological processes</div></pre></td></tr></table></figure>
<h2 id="测序数据">测序数据</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsamtools.html" target="_blank" rel="external">Rsamtools</a>中的 <code>scanBam</code> 返回BAM文件中的原始数值：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(Rsamtools)</div><div class="line">which &lt;- GRanges(&quot;chr1&quot;,IRanges(1000001,1001000))</div><div class="line">what &lt;- c(&quot;rname&quot;,&quot;strand&quot;,&quot;pos&quot;,&quot;qwidth&quot;,&quot;seq&quot;)</div><div class="line">param &lt;- ScanBamParam(which=which, what=what)</div><div class="line"># for more BamFile functions/details see ?BamFile</div><div class="line"># yieldSize for chunk-wise access</div><div class="line">bamfile &lt;- BamFile(&quot;/path/to/file.bam&quot;)</div><div class="line">reads &lt;- scanBam(bamfile, param=param)</div><div class="line">res &lt;- countBam(bamfile, param=param) </div><div class="line"># for more sophisticated counting modes</div><div class="line"># see summarizeOverlaps() below</div><div class="line"></div><div class="line"># quickly check chromosome names</div><div class="line">seqinfo(BamFile(&quot;/path/to/file.bam&quot;))</div><div class="line"></div><div class="line"># DNAStringSet is defined in the Biostrings package</div><div class="line"># see the Biostrings Quick Overview PDF</div><div class="line">dnastringset &lt;- scanFa(fastaFile, param=granges)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html" target="_blank" rel="external">GenomicAlignments</a> 返回一个Bioconductor对象(GRanges-based)</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line">ga &lt;- readGAlignments(bamfile) # single-end</div><div class="line">ga &lt;- readGAlignmentPairs(bamfile) # paired-end</div></pre></td></tr></table></figure>
<h2 id="转录本数据库">转录本数据库</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html" target="_blank" rel="external">GenomicFeatures</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div></pre></td><td class="code"><pre><div class="line"># get a transcript database, which stores exon, trancript, and gene information</div><div class="line">library(GenomicFeatures)</div><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line"></div><div class="line"># or build a txdb from GTF file (e.g. downloadable from Ensembl FTP site)</div><div class="line">txdb &lt;- makeTranscriptDbFromGFF(&quot;file.GTF&quot;, format=&quot;gtf&quot;)</div><div class="line"></div><div class="line"># or build a txdb from Biomart (however, not as easy to reproduce later)</div><div class="line">txdb &lt;- makeTranscriptDbFromBiomart(biomart = &quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)</div><div class="line"></div><div class="line"># in Bioconductor &gt;= 3.1, also makeTxDbFromGRanges</div><div class="line"></div><div class="line"># saving and loading</div><div class="line">saveDb(txdb, file=&quot;txdb.sqlite&quot;)</div><div class="line">loadDb(&quot;txdb.sqlite&quot;)</div><div class="line"></div><div class="line"># extracting information from txdb</div><div class="line">g &lt;- genes(txdb) # GRanges, just start to end, no exon/intron information</div><div class="line">tx &lt;- transcripts(txdb) # GRanges, similar to genes()</div><div class="line">e &lt;- exons(txdb) # GRanges for each exon</div><div class="line">ebg &lt;- exonsBy(txdb, by=&quot;gene&quot;) # exons grouped in a GRangesList by gene</div><div class="line">ebt &lt;- exonsBy(txdb, by=&quot;tx&quot;) # similar but by transcript</div><div class="line"></div><div class="line"># then get the transcript sequence</div><div class="line">txSeq &lt;- extractTranscriptSeqs(Hsapiens, ebt)</div></pre></td></tr></table></figure>
<h2 id="根据范围ranges和实验汇总信息">根据范围(ranges)和实验汇总信息</h2>
<p><code>SummarizedExperiment</code>是高维数据的储存类， 它与<code>GRanges</code>或<code>GRangesList</code>配合使用（使用每个基因外显子的read计数）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line">fls &lt;- list.files(pattern=&quot;*.bam$&quot;)</div><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line">ebg &lt;- exonsBy(txdb, by=&quot;gene&quot;)</div><div class="line"># see yieldSize argument for restricting memory</div><div class="line">bf &lt;- BamFileList(fls)</div><div class="line">library(BiocParallel)</div><div class="line">register(MulticoreParam(4))</div><div class="line"># lots of options in the man page</div><div class="line"># singleEnd, ignore.strand, inter.features, fragments, etc.</div><div class="line">se &lt;- summarizeOverlaps(ebg, bf)</div><div class="line"></div><div class="line"># operations on SummarizedExperiment</div><div class="line">assay(se) # the counts from summarizeOverlaps</div><div class="line">colData(se)</div><div class="line">rowRanges(se)</div></pre></td></tr></table></figure>
<p><a href="https://combine-lab.github.io/salmon/" target="_blank" rel="external">Salmon</a>是一个定量工具，如果数据占有GC依赖的数据，那么就添加上<code>--gcBias</code>参数，接着使用 <a href="http://bioconductor.org/pacakges/tximport" target="_blank" rel="external">tximport</a>，以下是使用案例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">coldata &lt;- read.table(&quot;samples.txt&quot;)</div><div class="line">rownames(coldata) &lt;- coldata$id</div><div class="line">files &lt;- coldata$files; names(files) &lt;- coldata$id</div><div class="line">txi &lt;- tximport(files, type=&quot;salmon&quot;, tx2gene=tx2gene)</div><div class="line">dds &lt;- DESeqDataSetFromTximport(txi, coldata, ~condition)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsubread.html" target="_blank" rel="external">Rsubread</a>包中的<code>featureCounts</code>函数也可以用于read计算，速度可能更快。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(Rsubread)</div><div class="line">res &lt;- featureCounts(files, annot.ext=&quot;annotation.gtf&quot;,</div><div class="line">  isGTFAnnotationFile=TRUE,</div><div class="line">  GTF.featureType=&quot;exon&quot;,</div><div class="line">  GTF.attrType=&quot;gene_id&quot;)</div><div class="line">res$counts</div></pre></td></tr></table></figure>
<h2 id="rna-seq分析方法">RNA-seq分析方法</h2>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html" target="_blank" rel="external">DESeq2</a>：My preferred pipeline for DESeq2 users is to start with a lightweight transcript abundance quantifier such as <a href="https://combine-lab.github.io/salmon/" target="_blank" rel="external">Salmon</a> and to use <a href="http://bioconductor.org/packages/tximport" target="_blank" rel="external">tximport</a>, followed by <code>DESeqDataSetFromTximport</code>.</p>
<p>Here, <code>coldata</code> is a <em>data.frame</em> with <code>group</code> as a column.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">library(DESeq2)</div><div class="line"># from tximport</div><div class="line">dds &lt;- DESeqDataSetFromTximport(txi, coldata, ~ group)</div><div class="line"># from SummarizedExperiment</div><div class="line">dds &lt;- DESeqDataSet(se, ~ group)</div><div class="line"># from count matrix</div><div class="line">dds &lt;- DESeqDataSetFromMatrix(counts, coldata, ~ group)</div><div class="line"># minimal filtering helps keep things fast </div><div class="line"># one can set &apos;n&apos; to e.g. min(5, smallest group sample size)</div><div class="line">keep &lt;- rowSums(counts(dds) &gt;= 10) &gt;= n </div><div class="line">dds &lt;- dds[keep,]</div><div class="line">dds &lt;- DESeq(dds)</div><div class="line">res &lt;- results(dds) # no shrinkage of LFC, or:</div><div class="line">res &lt;- lfcShrink(dds, coef = 2, type=&quot;apeglm&quot;) # shrink LFCs</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank" rel="external">edgeR</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"># this chunk from the Quick start in the edgeR User Guide</div><div class="line">library(edgeR) </div><div class="line">y &lt;- DGEList(counts=counts,group=group)</div><div class="line">keep &lt;- filterByExpr(y)</div><div class="line">y &lt;- y[keep,]</div><div class="line">y &lt;- calcNormFactors(y)</div><div class="line">design &lt;- model.matrix(~group)</div><div class="line">y &lt;- estimateDisp(y,design)</div><div class="line">fit &lt;- glmFit(y,design)</div><div class="line">lrt &lt;- glmLRT(fit)</div><div class="line">topTags(lrt)</div><div class="line"># or use the QL methods:</div><div class="line">qlfit &lt;- glmQLFit(y,design)</div><div class="line">qlft &lt;- glmQLFTest(qlfit)</div><div class="line">topTags(qlft)</div></pre></td></tr></table></figure>
<p><a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html" target="_blank" rel="external">limma-voom</a></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">library(limma)</div><div class="line">design &lt;- model.matrix(~ group)</div><div class="line">y &lt;- DGEList(counts)</div><div class="line">keep &lt;- filterByExpr(y)</div><div class="line">y &lt;- y[keep,]</div><div class="line">y &lt;- calcNormFactors(y)</div><div class="line">v &lt;- voom(y,design)</div><div class="line">fit &lt;- lmFit(v,design)</div><div class="line">fit &lt;- eBayes(fit)</div><div class="line">topTable(fit)</div></pre></td></tr></table></figure>
<p>更多有关RNA-seq操作的包可以参考：<a href="http://www.bioconductor.org/packages/release/BiocViews.html#___RNASeq" target="_blank" rel="external">Many more RNA-seq packages</a></p>
<h2 id="表达数据集expression-set">表达数据集Expression set</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(Biobase)</div><div class="line">data(sample.ExpressionSet)</div><div class="line">e &lt;- sample.ExpressionSet</div><div class="line">exprs(e)</div><div class="line">pData(e)</div><div class="line">fData(e)</div></pre></td></tr></table></figure>
<h2 id="下载geo数据库">下载GEO数据库</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(GEOquery)</div><div class="line">e &lt;- getGEO(&quot;GSE9514&quot;)</div></pre></td></tr></table></figure>
<h2 id="芯片分析">芯片分析</h2>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">library(affy)</div><div class="line">library(limma)</div><div class="line">phenoData &lt;- read.AnnotatedDataFrame(&quot;sample-description.csv&quot;)</div><div class="line">eset &lt;- justRMA(&quot;/celfile-directory&quot;, phenoData=phenoData)</div><div class="line">design &lt;- model.matrix(~ Disease, pData(eset))</div><div class="line">fit &lt;- lmFit(eset, design)</div><div class="line">efit &lt;- eBayes(fit)</div><div class="line">topTable(efit, coef=2)</div></pre></td></tr></table></figure>
<h2 id="icobra包">iCOBRA包</h2>
<p>iCOBRA包用于计算和可视化排序数据以及二分类匹配后的数据。例如，iCOBRA包的一个典型用途就是用于评估基因表达实验中不同组之间的比较方法，我们可以视之为一个排序问题（评估正确的效应量(effect size)以及按照显著性进行排列的基因），还是一个二分类比较问题（binary assignment problem）（例如将基因分为差异表达和非差异表达）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">library(iCOBRA)</div><div class="line">cd &lt;- COBRAData(pval=pval.df, padj=padj.df, score=score.df, truth=truth.df)</div><div class="line">cp &lt;- calculate_performance(cd, binary_truth = &quot;status&quot;, cont_truth = &quot;logFC&quot;)</div><div class="line">cobraplot &lt;- prepare_data_for_plot(cp)</div><div class="line">plot_fdrtprcurve(cobraplot)</div><div class="line"># interactive shiny app:</div><div class="line">COBRAapp(cd)</div></pre></td></tr></table></figure>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/annoCheat.html" target="_blank" rel="external">Genomic annotation in Bioconductor: Cheat sheet</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/08/28/DAL/DALS028_Batch_Effect04_FactorAnalysis/" rel="next" title="DALS028-批次效应04-因子分析">
                <i class="fa fa-chevron-left"></i> DALS028-批次效应04-因子分析
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/07/Experiment/HIPC介绍/" rel="prev" title="HIPC介绍">
                HIPC介绍 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">229</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">113</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#安装"><span class="nav-number">2.</span> <span class="nav-text">安装</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#r语言帮助"><span class="nav-number">3.</span> <span class="nav-text">R语言帮助</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#debugging-r"><span class="nav-number">4.</span> <span class="nav-text">debugging R</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#显示某个类的包特异方法"><span class="nav-number">5.</span> <span class="nav-text">显示某个类的包特异方法</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#注释annotations"><span class="nav-number">6.</span> <span class="nav-text">注释Annotations</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#genomicranges包"><span class="nav-number">7.</span> <span class="nav-text">GenomicRanges包</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#intra-range-methods"><span class="nav-number">7.1.</span> <span class="nav-text">Intra-range methods</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#inter-range-methods"><span class="nav-number">7.2.</span> <span class="nav-text">Inter-range methods</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#nearest-methods"><span class="nav-number">7.3.</span> <span class="nav-text">Nearest methods</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#set-methods"><span class="nav-number">7.4.</span> <span class="nav-text">set methods</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#overlaps"><span class="nav-number">7.5.</span> <span class="nav-text">Overlaps</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#seqnames-and-seqlevels"><span class="nav-number">7.6.</span> <span class="nav-text">Seqnames and seqlevels</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#sequences"><span class="nav-number">8.</span> <span class="nav-text">Sequences</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#测序数据"><span class="nav-number">9.</span> <span class="nav-text">测序数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#转录本数据库"><span class="nav-number">10.</span> <span class="nav-text">转录本数据库</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#根据范围ranges和实验汇总信息"><span class="nav-number">11.</span> <span class="nav-text">根据范围(ranges)和实验汇总信息</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#rna-seq分析方法"><span class="nav-number">12.</span> <span class="nav-text">RNA-seq分析方法</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#表达数据集expression-set"><span class="nav-number">13.</span> <span class="nav-text">表达数据集Expression set</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#下载geo数据库"><span class="nav-number">14.</span> <span class="nav-text">下载GEO数据库</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#芯片分析"><span class="nav-number">15.</span> <span class="nav-text">芯片分析</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#icobra包"><span class="nav-number">16.</span> <span class="nav-text">iCOBRA包</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">17.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共920.5k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
