<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="PH525x series - Biomedical Data Science Management of genome-scale dataIntroduction (revised October 2017)It is very common for life scientists to use spreadsheets to record and manage experimental da">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS004-管理基因组级数据">
<meta property="og:url" content="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS005-管理基因组级数据/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="PH525x series - Biomedical Data Science Management of genome-scale dataIntroduction (revised October 2017)It is very common for life scientists to use spreadsheets to record and manage experimental da">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-dochks-1.png">
<meta property="og:updated_time" content="2019-11-06T01:04:26.874Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS004-管理基因组级数据">
<meta name="twitter:description" content="PH525x series - Biomedical Data Science Management of genome-scale dataIntroduction (revised October 2017)It is very common for life scientists to use spreadsheets to record and manage experimental da">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS005-管理基因组级数据/"/>





  <title>GDAS004-管理基因组级数据 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/04/Genomics Data Analysis Series/GDAS005-管理基因组级数据/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS004-管理基因组级数据</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-04T12:00:00+08:00">
                2019-09-04
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,318
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  14
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p><a href="http://genomicsclass.github.io/book" target="_blank" rel="external">PH525x series - Biomedical Data Science</a></p>
<h2 id="Management-of-genome-scale-data"><a href="#Management-of-genome-scale-data" class="headerlink" title="Management of genome-scale data"></a>Management of genome-scale data</h2><h1 id="Introduction-revised-October-2017"><a href="#Introduction-revised-October-2017" class="headerlink" title="Introduction (revised October 2017)"></a>Introduction (revised October 2017)</h1><p>It is very common for life scientists to use spreadsheets to record and manage experimental data. In these problems we will build up a discipline of data management for genome-scale experiments using R and Bioconductor.</p>
<p>It may seem unusual to recommend R-based software for managing data. Ordinarily R is used for programming and computing statistical analyses. However, Bioconductor has taken very seriously the need to carefully annotate, in a unified way, many aspects of complex biological experiments. This approach allows us to import assay outputs (such as multiple microarray scans, or multiple sequencing alignments), combining them with relevant metadata about the experiment and the samples assayed, saving all the information in a single R variable. By carefully designing the structure of this variable, programs can operate reliably on various aspects of the data. Ultimately a more efficient management and analysis environment is obtained.</p>
<p>We will come to understand this enhancement of efficiency by working from some very basic representations of experimental data, proceeding stepwise to more advanced integrative representations that are the hallmark of Bioconductor’s approach to genome-scale data.</p>
<h2 id="Illustrative-exercises-worked-through-in-video"><a href="#Illustrative-exercises-worked-through-in-video" class="headerlink" title="Illustrative exercises (worked through in video)"></a>Illustrative exercises (worked through in video)</h2><p>1.0 Distinct components from an experiment.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">library(GSE5859Subset)</div><div class="line">library(Biobase)</div><div class="line">.oldls = ls()</div><div class="line">data(GSE5859Subset)</div><div class="line">.newls = ls()</div></pre></td></tr></table></figure>
<p>The objects obtained from the data() command are</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">newstuff = setdiff(.newls, .oldls)</div><div class="line">newstuff</div><div class="line">## [1] &quot;geneAnnotation&quot; &quot;geneExpression&quot; &quot;sampleInfo&quot;</div></pre></td></tr></table></figure>
<p>and they have classes</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">cls = lapply(newstuff, function(x) class(get(x)))</div><div class="line">names(cls) = newstuff</div><div class="line">cls</div><div class="line">## $$geneAnnotation</div><div class="line">## [1] &quot;data.frame&quot;</div><div class="line">## </div><div class="line">## $$geneExpression</div><div class="line">## [1] &quot;matrix&quot;</div><div class="line">## </div><div class="line">## $$sampleInfo</div><div class="line">## [1] &quot;data.frame&quot;</div></pre></td></tr></table></figure>
<p>It is easy to make an informative plot of hybridization date by ethnicity of sample source using standard R:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(date~factor(ethnicity), data=sampleInfo, ylab=&quot;chip date&quot;)</div></pre></td></tr></table></figure>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-lket-1.png" alt="plot of chunk lket"></p>
<p>What if we want to look at expression of the gene BRCA2 by ethnicity? We have to do several things:</p>
<ul>
<li>We need to determine which row of <code>geneExpression</code> corresponds to BRCA2, and to do this we need to know which column of <code>geneAnnotation</code> provides the information;</li>
<li>We need to ensure that the gene expression data in the matrix <code>geneExpression</code> are ordered consistently with the ordering of data in <code>sampleInfo</code>, and to do this we need to know which column of <code>sampleInfo</code>identifies the samples using the same tags as <code>geneExpression</code>.</li>
</ul>
<p>Therefore, to relate gene expression to a sample characteristic we need to manipulate three different entities (expression, sample annotation, and gene annotation), ensuring that they are properly coordinated.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">all.equal(colnames(geneExpression), sampleInfo$filename)</div><div class="line">## [1] TRUE</div><div class="line">all.equal(rownames(geneExpression), geneAnnotation$PROBEID)</div><div class="line">## [1] TRUE</div><div class="line">ind = which(geneAnnotation$SYMBOL==&quot;BRCA2&quot;)</div><div class="line">boxplot(geneExpression[ind,]~sampleInfo$ethnicity, ylab=&quot;BRCA2 by hgfocus&quot;)</div></pre></td></tr></table></figure>
<p><img src="http://genomicsclass.github.io/book/pages/figure/bioc1_mgt_gsd-dochks-1.png" alt="plot of chunk dochks"></p>
<p>It works, but it is complicated. To reduce complexity, Bioconductor defines <code>ExpressionSet</code>.</p>
<p>1.1 <em>The ExpressionSet container</em>.</p>
<p>There are high-level commands that simplify ExpressionSet construction.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">es1 = ExpressionSet(geneExpression)</div><div class="line">pData(es1) = sampleInfo</div><div class="line">fData(es1) = geneAnnotation</div><div class="line">es1</div><div class="line">## ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">## assayData: 8793 features, 24 samples </div><div class="line">##   element names: exprs </div><div class="line">## protocolData: none</div><div class="line">## phenoData</div><div class="line">##   sampleNames: 107 122 ... 159 (24 total)</div><div class="line">##   varLabels: ethnicity date filename group</div><div class="line">##   varMetadata: labelDescription</div><div class="line">## featureData</div><div class="line">##   featureNames: 1 30 ... 12919 (8793 total)</div><div class="line">##   fvarLabels: PROBEID CHR CHRLOC SYMBOL</div><div class="line">##   fvarMetadata: labelDescription</div><div class="line">## experimentData: use &apos;experimentData(object)&apos;</div><div class="line">## Annotation:</div></pre></td></tr></table></figure>
<p>When we mention <code>es1</code> we get a compact representation of lots of information about the experiment. Use <code>es1</code>to visualize the confounding of hybridization date with ethnicity:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">boxplot(es1$date~es1$ethnicity)</div></pre></td></tr></table></figure>
<p>Now you’ve been introduced to the ExpressionSet container design and you’ve even constructed an ExpressionSet instance. We’ll use a more comprehensive representation of the data in the underlying experiment in the next task. We’ll use</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(GSE5859)</div><div class="line">data(GSE5859)</div><div class="line">e</div><div class="line">## ExpressionSet (storageMode: lockedEnvironment)</div><div class="line">## assayData: 8793 features, 208 samples </div><div class="line">##   element names: exprs, se.exprs </div><div class="line">## protocolData</div><div class="line">##   sampleNames: GSM25349.CEL.gz GSM25350.CEL.gz ...</div><div class="line">##     GSM136729.CEL.gz (208 total)</div><div class="line">##   varLabels: ScanDate</div><div class="line">##   varMetadata: labelDescription</div><div class="line">## phenoData</div><div class="line">##   sampleNames: 1 2 ... 208 (208 total)</div><div class="line">##   varLabels: ethnicity date filename</div><div class="line">##   varMetadata: labelDescription</div><div class="line">## featureData: none</div><div class="line">## experimentData: use &apos;experimentData(object)&apos;</div><div class="line">## Annotation: hgfocus</div></pre></td></tr></table></figure>
<p>1.2 <em>Metadata management</em>. It is useful to have information about an experiment’s intentions and methods ready-to-hand. Some of the arrays used for GSE5859 were produced for a 2004 paper with Pubmed ID 15269782. We can, if connected to the internet, import information about that paper into R.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">library(annotate)</div><div class="line">p = pmid2MIAME(&quot;15269782&quot;)</div><div class="line">p</div><div class="line">## Experiment data</div><div class="line">##   Experimenter name: Morley M </div><div class="line">##   Laboratory: NA </div><div class="line">##   Contact information:  </div><div class="line">##   Title: Genetic analysis of genome-wide variation in human gene expression. </div><div class="line">##   URL:  </div><div class="line">##   PMIDs: 15269782 </div><div class="line">## </div><div class="line">##   Abstract: A 167 word abstract is available. Use &apos;abstract&apos; method.</div></pre></td></tr></table></figure>
<p>Check the display of <code>e</code> in your session. Then Use the command <code>experimentData(e) &lt;- p</code>. How has the display of <code>e</code> changed? What is the result of <code>experimentData(e)</code>?</p>
<h1 id="ExpressionSet-Why-and-How"><a href="#ExpressionSet-Why-and-How" class="headerlink" title="ExpressionSet: Why? and How?"></a>ExpressionSet: Why? and How?</h1><p>We have moved fairly quickly through the creation and use of an <code>ExpressionSet</code> instance to illustrate some of its virtues. We will now provide some details on why and how this structure is designed and used.</p>
<p>Motivations underlying the ExpressionSet design are many. A few of the most important are</p>
<ul>
<li>to reduce complexity of programming with complex assay, annotation, and sample-level information,</li>
<li>to support simple and idiomatic filtering of features and samples,</li>
<li>to control memory consumption as assay data are passed to analytical functions.</li>
</ul>
<p>So much for “why”, to this point. “How” takes us for an excursion in software design strategy: object-oriented programming in R with the class system known as S4.</p>
<h2 id="Object-oriented-programming-and-S4-classes"><a href="#Object-oriented-programming-and-S4-classes" class="headerlink" title="Object-oriented programming and S4 classes"></a>Object-oriented programming and S4 classes</h2><p>Object-oriented programming (OOP) is a discipline of computer programming that helps to reduce program complexity and maintenance requirements, and to enhance program interoperability. There are various approaches to OOP, even within R, and we focus on the one known as S4. The basic idea is that an object</p>
<ul>
<li>brings together data of different types in a unified structure,</li>
<li>is identified as a member of a formally specified class, and may be formally situated in a system of related objects,</li>
<li>“responds to” generic method calls in a class-specific way.</li>
</ul>
<p>We can see the class structure for an ExpressionSet with <code>getClass</code>:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(&quot;ExpressionSet&quot;)</div><div class="line">## Class &quot;ExpressionSet&quot; [package &quot;Biobase&quot;]</div><div class="line">## </div><div class="line">## Slots:</div><div class="line">##                                                                </div><div class="line">## Name:      experimentData          assayData          phenoData</div><div class="line">## Class:              MIAME          AssayData AnnotatedDataFrame</div><div class="line">##                                                                </div><div class="line">## Name:         featureData         annotation       protocolData</div><div class="line">## Class: AnnotatedDataFrame          character AnnotatedDataFrame</div><div class="line">##                          </div><div class="line">## Name:   .__classVersion__</div><div class="line">## Class:           Versions</div><div class="line">## </div><div class="line">## Extends: </div><div class="line">## Class &quot;eSet&quot;, directly</div><div class="line">## Class &quot;VersionedBiobase&quot;, by class &quot;eSet&quot;, distance 2</div><div class="line">## Class &quot;Versioned&quot;, by class &quot;eSet&quot;, distance 3</div></pre></td></tr></table></figure>
<p>This shows that there are 7 ‘slots’ comprising the class definition. Only the <code>annotation</code> slot is occupied by a simple R data type. All the other slots are to be occupied by instances of other S4 classes (or more basic R data types).</p>
<p>As an example, the <code>MIAME</code> class has the following definition:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">getClass(&quot;MIAME&quot;)</div><div class="line">## Class &quot;MIAME&quot; [package &quot;Biobase&quot;]</div><div class="line">## </div><div class="line">## Slots:</div><div class="line">##                                                             </div><div class="line">## Name:               name               lab           contact</div><div class="line">## Class:         character         character         character</div><div class="line">##                                                             </div><div class="line">## Name:              title          abstract               url</div><div class="line">## Class:         character         character         character</div><div class="line">##                                                             </div><div class="line">## Name:          pubMedIds           samples    hybridizations</div><div class="line">## Class:         character              list              list</div><div class="line">##                                                             </div><div class="line">## Name:       normControls     preprocessing             other</div><div class="line">## Class:              list              list              list</div><div class="line">##                         </div><div class="line">## Name:  .__classVersion__</div><div class="line">## Class:          Versions</div><div class="line">## </div><div class="line">## Extends: </div><div class="line">## Class &quot;MIAxE&quot;, directly</div><div class="line">## Class &quot;characterORMIAME&quot;, directly</div><div class="line">## Class &quot;Versioned&quot;, by class &quot;MIAxE&quot;, distance 2</div></pre></td></tr></table></figure>
<p>So we see that S4 allows us to define new data structures that encapsulate diverse types of information. We could do this with the very general “list” structure provided by R. But S4 allows us to guarantee properties of our new structures so that programs defined for these structures do not have to “check” on what they are operating on. The checks are built in to the class system.</p>
<h2 id="DataFrame-an-enhanced-structure-for-tabular-data"><a href="#DataFrame-an-enhanced-structure-for-tabular-data" class="headerlink" title="DataFrame: an enhanced structure for tabular data"></a>DataFrame: an enhanced structure for tabular data</h2><p>We are familiar with the <code>data.frame</code> structure from base R. Bioconductor introduced the <code>DataFrame</code> class to provide some additional capabilities that will be heavily exploited later.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">library(S4Vectors)</div><div class="line">sa2 = DataFrame(sampleInfo)</div><div class="line">sa2</div><div class="line">## DataFrame with 24 rows and 4 columns</div><div class="line">##     ethnicity       date         filename     group</div><div class="line">##      &lt;factor&gt;     &lt;Date&gt;      &lt;character&gt; &lt;numeric&gt;</div><div class="line">## 1         ASN 2005-06-23 GSM136508.CEL.gz         1</div><div class="line">## 2         ASN 2005-06-27 GSM136530.CEL.gz         1</div><div class="line">## 3         ASN 2005-06-27 GSM136517.CEL.gz         1</div><div class="line">## 4         ASN 2005-10-28 GSM136576.CEL.gz         1</div><div class="line">## 5         ASN 2005-10-07 GSM136566.CEL.gz         1</div><div class="line">## ...       ...        ...              ...       ...</div><div class="line">## 20        ASN 2005-06-27 GSM136524.CEL.gz         0</div><div class="line">## 21        ASN 2005-06-23 GSM136514.CEL.gz         0</div><div class="line">## 22        ASN 2005-10-07 GSM136563.CEL.gz         0</div><div class="line">## 23        ASN 2005-10-07 GSM136564.CEL.gz         0</div><div class="line">## 24        ASN 2005-10-07 GSM136572.CEL.gz         0</div></pre></td></tr></table></figure>
<p>The <code>show</code> method depicts top and bottom table row data, provides information on column datatype, and has the capacity to manage complex data types in columns more effectively than <code>data.frame</code>. Here is the class definition:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">getClass(&quot;DataFrame&quot;)</div><div class="line">## Class &quot;DataFrame&quot; [package &quot;S4Vectors&quot;]</div><div class="line">## </div><div class="line">## Slots:</div><div class="line">##                                                             </div><div class="line">## Name:           rownames             nrows          listData</div><div class="line">## Class: character_OR_NULL           integer              list</div><div class="line">##                                                             </div><div class="line">## Name:        elementType   elementMetadata          metadata</div><div class="line">## Class:         character DataTable_OR_NULL              list</div><div class="line">## </div><div class="line">## Extends: </div><div class="line">## Class &quot;DataTable&quot;, directly</div><div class="line">## Class &quot;SimpleList&quot;, directly</div><div class="line">## Class &quot;DataTable_OR_NULL&quot;, by class &quot;DataTable&quot;, distance 2</div><div class="line">## Class &quot;List&quot;, by class &quot;SimpleList&quot;, distance 2</div><div class="line">## Class &quot;Vector&quot;, by class &quot;SimpleList&quot;, distance 3</div><div class="line">## Class &quot;Annotated&quot;, by class &quot;SimpleList&quot;, distance 4</div></pre></td></tr></table></figure>
<h2 id="Representations-for-mature-archives-from-multisample-NGS-experiments"><a href="#Representations-for-mature-archives-from-multisample-NGS-experiments" class="headerlink" title="Representations for mature archives from multisample NGS experiments"></a>Representations for mature archives from multisample NGS experiments</h2><p>Short read sequencing leads to data collections that are more voluminous than gene-oriented microarray outputs.<br>Sequencing experiments may also be analyzed in more varied ways, so that it is valuable to provide efficient access to assay outputs at various levels of granularity.</p>
<p>A typical data flow for RNA-seq studies has phases 1) chromatographic display of base calls per cycle (seldom provided after, say, 2008), 2) base calls over the read length, with quality values, in FASTQ format, 3) feature quantification through statistical processing of genomic or transcriptomic alignments, 4) genome-scale archiving of multiple samples for inference on biologic questions of interest. FASTQ files are often developed in pairs for each sample.</p>
<h3 id="A-set-of-BAM-files-in-Bioconductor"><a href="#A-set-of-BAM-files-in-Bioconductor" class="headerlink" title="A set of BAM files in Bioconductor"></a>A set of BAM files in Bioconductor</h3><p>The <em>RNAseqData.HNRNPC.bam.chr14</em> package includes BAM files derived from paired-end sequencing of 8 samples of HeLa cell lines, 4 of which were subjected to RNA interference to “knock down” the expression of the heterogeneous nuclear ribonucleoproteins C1 and C2 (denoted hnRNPC). This family of nuclear proteins is involved in pre-mRNA processing. We will demonstrate some approaches to managing the data from BAM to archived counts related to HNRNPC.</p>
<p>The BAM file pathnames are available in a character vector stored in the RNAseqData… package.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">library(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">bfp = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line">length(bfp)</div><div class="line">## [1] 8</div><div class="line">bfp[1]</div><div class="line">##                                                                                                                ERR127306 </div><div class="line">## &quot;/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RNAseqData.HNRNPC.bam.chr14/extdata/ERR127306_chr14.bam&quot;</div></pre></td></tr></table></figure>
<p>We use the <em>Rsamtools</em> package to manage these using a single variable.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">library(Rsamtools)</div><div class="line">bfl = BamFileList(file=bfp)</div><div class="line">bfl</div><div class="line">## BamFileList of length 8</div><div class="line">## names(8): ERR127306 ERR127307 ERR127308 ... ERR127303 ERR127304 ERR127305</div></pre></td></tr></table></figure>
<p>The BAM files include a limited amount of metadata about the genome to which the reads were aligned. This is propagated to the BamFileList instance.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">seqinfo(bfl)</div><div class="line">## Seqinfo object with 93 sequences from an unspecified genome:</div><div class="line">##   seqnames              seqlengths isCircular genome</div><div class="line">##   chr1                   249250621       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chr10                  135534747       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chr11                  135006516       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chr11_gl000202_random      40103       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chr12                  133851895       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   ...                          ...        ...    ...</div><div class="line">##   chrUn_gl000247             36422       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chrUn_gl000248             39786       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chrUn_gl000249             38502       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chrX                   155270560       &lt;NA&gt;   &lt;NA&gt;</div><div class="line">##   chrY                    59373566       &lt;NA&gt;   &lt;NA&gt;</div></pre></td></tr></table></figure>
<p>We see that the genome is “unspecified”, but we happen to know that the reference build ‘hg19’ was used to define the genomic sequence for alignment. [genome(bfl) &lt;- “hg19” should rectify this but at present there is a bug. This should be fixed so that attempts to compare alignments using different references can be flagged or throw errors.]</p>
<p>In order to verify the success of the knockdown we will check the alignments in the vicinity of the genomic sequence coding for HNRNPC. Later we will show how to obtain the gene address using Bioconductor annotation tools; for now we just assert that for hg19</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">hnrnpcLoc = GRanges(&quot;chr14&quot;, IRanges(21677296, 21737638))</div><div class="line">hnrnpcLoc</div><div class="line">## GRanges object with 1 range and 0 metadata columns:</div><div class="line">##       seqnames               ranges strand</div><div class="line">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</div><div class="line">##   [1]    chr14 [21677296, 21737638]      *</div><div class="line">##   -------</div><div class="line">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div></pre></td></tr></table></figure>
<p>This identifies a region of chromosome 14 to which the HNRNPC coding gene is annotated for hg19. A simple way of assessing the expression quantifications uses <code>summarizeOverlaps</code> from the <em>GenomicAlignments</em>package.  This counting method will implicitly use multicore computation.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line">library(BiocParallel)</div><div class="line">register(SerialParam())</div><div class="line">hnse = summarizeOverlaps(hnrnpcLoc,bfl)</div></pre></td></tr></table></figure>
<h3 id="Exercise"><a href="#Exercise" class="headerlink" title="Exercise"></a>Exercise</h3><p>5.1 Examine the contents of the <code>count</code> component of <code>hnse</code>. What are the column identifiers for quantifications corresponding to the HNRNPC knockdown?</p>
<p>5.2 Add a variable <code>condition</code> with values <code>knockdown</code> and <code>wildtype</code> to the <code>colData</code> component of <code>hnse</code>.</p>
<p>5.3 Samples with identifier suffixes 306 and 307 are technical replicates of a single control sample; likewise for 308 and 309. Consequently there are six different biological samples in use. Add a variable to <code>colData</code>distinguishing these six samples.</p>
<h2 id="Some-intricacies-of-the-ExpressionSet-design-advanced"><a href="#Some-intricacies-of-the-ExpressionSet-design-advanced" class="headerlink" title="Some intricacies of the ExpressionSet design (advanced)"></a>Some intricacies of the ExpressionSet design (advanced)</h2><p>Bioinformaticians work in an environment that is constantly changing. Computer capacity and throughput has increased dramatically since Bioconductor began (around 2000CE). The R language has also changed considerably. In the early days, it was important to be very economical about memory consumption, and R’s semantics at that time led to some inefficiencies through copying of objects in memory as they are passed to functions. Copying does not occur for R objects of class ‘environment’, and this led to the adoption of environments for storage of voluminous assay data.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">class(slot(es1, &quot;assayData&quot;))</div><div class="line">## [1] &quot;environment&quot;</div><div class="line">ae = slot(es1, &quot;assayData&quot;)</div><div class="line">ae</div><div class="line">## &lt;environment: 0x7f813b4dc698&gt;</div></pre></td></tr></table></figure>
<p>This is a very special type of object for R. It can be thought of as a container for a reference to the matrix of expression values.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">ls(ae)</div><div class="line">## [1] &quot;exprs&quot;</div><div class="line">dim(ae$exprs)</div><div class="line">## [1] 8793   24</div><div class="line">dim(exprs(es1))</div><div class="line">## [1] 8793   24</div><div class="line">object.size(ae)</div><div class="line">## 56 bytes</div><div class="line">object.size(ae$exprs)</div><div class="line">## 2253824 bytes</div></pre></td></tr></table></figure>
<h3 id="Exercise-optional"><a href="#Exercise-optional" class="headerlink" title="Exercise (optional)"></a>Exercise (optional)</h3><p>4.1. Explain the following findings.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">e1 = new.env()</div><div class="line">l1 = list()</div><div class="line">e1$x = 5</div><div class="line">l1$x = 5</div><div class="line">all.equal(e1$x, l1$x)</div><div class="line">## [1] TRUE</div><div class="line">f = function(z) &#123;z$x = 4; z$x&#125;</div><div class="line">all.equal(f(e1), f(l1))</div><div class="line">## [1] TRUE</div><div class="line">all.equal(e1$x, l1$x)  # different from before</div><div class="line">## [1] &quot;Mean relative difference: 0.25&quot;</div></pre></td></tr></table></figure>
<hr>
<p><a href="http://genomicsclass.github.io/book/" target="_blank" rel="external">PH525x</a>, Rafael Irizarry and Michael Love, <a href="https://github.com/genomicsclass/labs/blob/master/LICENSE" target="_blank" rel="external">MIT License</a></p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/03/Genomics Data Analysis Series/GDAS003-Bioconductor与R语言/" rel="next" title="GDAS003-Bioconductor与基因组级数据分析简介">
                <i class="fa fa-chevron-left"></i> GDAS003-Bioconductor与基因组级数据分析简介
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/05/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/" rel="prev" title="GDAS006-深入理解SummarizedExperiment类">
                GDAS006-深入理解SummarizedExperiment类 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">236</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">113</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#Management-of-genome-scale-data"><span class="nav-number">1.</span> <span class="nav-text">Management of genome-scale data</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Introduction-revised-October-2017"><span class="nav-number"></span> <span class="nav-text">Introduction (revised October 2017)</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Illustrative-exercises-worked-through-in-video"><span class="nav-number">1.</span> <span class="nav-text">Illustrative exercises (worked through in video)</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#ExpressionSet-Why-and-How"><span class="nav-number"></span> <span class="nav-text">ExpressionSet: Why? and How?</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Object-oriented-programming-and-S4-classes"><span class="nav-number">1.</span> <span class="nav-text">Object-oriented programming and S4 classes</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#DataFrame-an-enhanced-structure-for-tabular-data"><span class="nav-number">2.</span> <span class="nav-text">DataFrame: an enhanced structure for tabular data</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Representations-for-mature-archives-from-multisample-NGS-experiments"><span class="nav-number">3.</span> <span class="nav-text">Representations for mature archives from multisample NGS experiments</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#A-set-of-BAM-files-in-Bioconductor"><span class="nav-number">3.1.</span> <span class="nav-text">A set of BAM files in Bioconductor</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Exercise"><span class="nav-number">3.2.</span> <span class="nav-text">Exercise</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Some-intricacies-of-the-ExpressionSet-design-advanced"><span class="nav-number">4.</span> <span class="nav-text">Some intricacies of the ExpressionSet design (advanced)</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Exercise-optional"><span class="nav-number">4.1.</span> <span class="nav-text">Exercise (optional)</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共938.3k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
