<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Hexo, NexT" />










<meta property="og:type" content="website">
<meta property="og:title" content="RVDSD的个人笔记本">
<meta property="og:url" content="http://rvdsd.top/page/2/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RVDSD的个人笔记本">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/page/2/"/>





  <title>RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/17/Genomics Data Analysis Series/GDAS018-The parallel package lapply and mclapply/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/17/Genomics Data Analysis Series/GDAS018-The parallel package lapply and mclapply/" itemprop="url">GDAS017-使用Bioconductor进行高性能计算</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-17T12:00:00+08:00">
                2019-09-17
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,519
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="性能增强概览">性能增强概览</h2>
<p>软件系统和数据分析的可扩展性(scalability)的概念是复杂的，精确的度量(precise metrics)和标准现在还不好处理。可以参考一下来自计算机界的 <a href="http://repository.cmu.edu/sei/399/" target="_blank" rel="external">Weinstock and Goodenough</a> 的一个概述。基于Bioconductor的分析工作流的可扩展性(scalability)可以沿着不同的路径进行，Bioconductor的两个核<a href="http://arxiv.org/abs/1409.2864" target="_blank" rel="external">Mike Lawrence和Martin Morgan</a>在工作就提到了许多有意的想法。</p>
<p>本文档仅解决Bioconductor直接支持的并行计算，主要通过两种方法实现：</p>
<ul>
<li>共享内存并行性。使用内存映像的完整副本对R进程进行任意次数的分叉，并且对每个映像独立地进行计算。选定的结果将返回到主进程。可以有效地在多环境中实现。</li>
<li>分布式并行性。可以运行不同操作系统的独立处理器可以运行R的兼容实例。R或集群调度程序可以执行作业控制(job control)。</li>
</ul>
<h3 id="简要说明">简要说明</h3>
<p>现在我们来说明一下共享内存方法：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">system.time( lapply(<span class="number">1</span>:<span class="number">8</span>, <span class="keyword">function</span>(x)Sys.sleep(<span class="number">1</span>) ) )</div><div class="line"><span class="comment">##    user  system elapsed </span></div><div class="line"><span class="comment">##   0.004   0.001   8.020</span></div><div class="line"><span class="keyword">library</span>(parallel)</div><div class="line">detectCores()</div><div class="line"><span class="comment">## [1] 4</span></div><div class="line">options(mc.cores=<span class="number">4</span>)</div><div class="line">system.time( mclapply(<span class="number">1</span>:<span class="number">8</span>, <span class="keyword">function</span>(x)Sys.sleep(<span class="number">1</span>) ) )</div><div class="line"><span class="comment">##    user  system elapsed </span></div><div class="line"><span class="comment">##   0.007   0.022   2.031</span></div></pre></td></tr></table></figure>
<p>上面只是一个无意义的计算，我们就是为了说明并行计算的原理，通过并行计算，我们加快了计算速度，计算过程的时间缩短为原来的四分之一。</p>
<h3 id="rna-seq实验中的bam文件">RNA-seq实验中的BAM文件</h3>
<p>我们将使用 <a href="http://www.pubmedcentral.nih.gov/articlerender.fcgi?artid=3629564&amp;tool=pmcentrez&amp;rendertype=abstract" target="_blank" rel="external">Zarnack et al. 2013</a> 等的人研究中的一些BAM文件。2013年的这项假设是，有一类蛋白质，即异质核糖蛋白C1和C2，也就是指的hnRNP C(heterogeneous nuclear ribonucleoproteins C1 and C2)，这个蛋白会阻止Alu因子整合到成熟到转录本上。这样的整合会导致蛋白功能障碍，进而导致疾病；如果想了解更深入的信息，可以去查看文献原文。</p>
<p>我们下面所使用的包含有8个BAM文件，它们已经与chr14染色体进行了比对。其中4个文件是正常的HeLa细胞的读长文件(reads)（阴性对照），另外4个文件是敲低了hnRNP C后的HeLa的读长文件，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">dir(system.file(<span class="string">"extdata"</span>, package=<span class="string">"RNAseqData.HNRNPC.bam.chr14"</span>))</div><div class="line"><span class="comment">##  [1] "ERR127302_chr14.bam"     "ERR127302_chr14.bam.bai"</span></div><div class="line"><span class="comment">##  [3] "ERR127303_chr14.bam"     "ERR127303_chr14.bam.bai"</span></div><div class="line"><span class="comment">##  [5] "ERR127304_chr14.bam"     "ERR127304_chr14.bam.bai"</span></div><div class="line"><span class="comment">##  [7] "ERR127305_chr14.bam"     "ERR127305_chr14.bam.bai"</span></div><div class="line"><span class="comment">##  [9] "ERR127306_chr14.bam"     "ERR127306_chr14.bam.bai"</span></div><div class="line"><span class="comment">## [11] "ERR127307_chr14.bam"     "ERR127307_chr14.bam.bai"</span></div><div class="line"><span class="comment">## [13] "ERR127308_chr14.bam"     "ERR127308_chr14.bam.bai"</span></div><div class="line"><span class="comment">## [15] "ERR127309_chr14.bam"     "ERR127309_chr14.bam.bai"</span></div><div class="line"><span class="comment">## [17] "tophat2_out"</span></div></pre></td></tr></table></figure>
<p>这些都是由Tabix建立索引的BAM文件。其中<code>*.bam</code> 是原始的BAM文件，而<code>*.bam.bai</code>则是添加了索引后的BAM文件，从上面我们可以看到有8个bam.bai文件，前面四个是敲低组，后面四个是对照组。</p>
<h3 id="使用biocparallel实现隐式并行">使用BiocParallel实现隐式并行</h3>
<p>为了促进可靠的执行程序的和谐发展，Bioconductor开发了<code>BiocParallel</code>包。在下面的案例中，我们将读取HNRNP C案例数据集中的reads读取到由外显子地址定义的bins中（这里我不清楚bin是什么，列出了原文）：</p>
<blockquote>
<p>Consider the following example: we will count reads into bins defined by exon addresses in the HNRNPC example dataset.</p>
</blockquote>
<p><code>RNAseqData.HNRNPC.bam.chr14</code> 这个包含有BAM文件的绝对路径的一个向量，我将其分配给fns，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">fns = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line">length(fns)</div><div class="line"><span class="comment">## [1] 8</span></div></pre></td></tr></table></figure>
<p>Here we establish the exon bins into which we will count reads.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line">seqlevels(txdb) = &quot;chr14&quot;</div><div class="line">ebg = exonsBy(txdb, by=&quot;gene&quot;)</div></pre></td></tr></table></figure>
<p>Now we use the <code>summarizeOverlaps</code> function from the GenomicAlignments package to count reads into the exon bins. We’ll time the counting from a single file, and then time the counting from all files at once.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">library(GenomicAlignments)</div><div class="line"># summarizeOverlaps uses bplapply to iterate over files</div><div class="line">s1 = system.time(i1 &lt;- summarizeOverlaps( ebg, fns[3] ))</div><div class="line">s1</div><div class="line">##    user  system elapsed </div><div class="line">##   2.975   0.178   3.178</div><div class="line"># show implicit config</div><div class="line">BiocParallel::bpparam()</div><div class="line">## class: MulticoreParam</div><div class="line">##   bpisup: FALSE; bpnworkers: 4; bptasks: 0; bpjobname: BPJOB</div><div class="line">##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</div><div class="line">##   bptimeout: 2592000; bpprogressbar: FALSE</div><div class="line">##   bpRNGseed: </div><div class="line">##   bplogdir: NA</div><div class="line">##   bpresultdir: NA</div><div class="line">##   cluster type: FORK</div><div class="line">s2 = system.time(i2 &lt;- summarizeOverlaps( ebg, fns ))</div><div class="line">s2</div><div class="line">##    user  system elapsed </div><div class="line">##   0.271   0.162  10.480</div></pre></td></tr></table></figure>
<p>This is not a thorough way of measuring speedup but it shows reasonable enhancement. In the second computation, we did approximately 8 times as much computation, but the clock time elapsed increased only by a factor of (3.3). The default behavior of BiocParallel on the MacBook Air used to produce this document is to pick up the value of the option <code>mc.cores</code> and use this as the number of workers in a <code>MulticoreParam</code> configuration; if <code>options()$mc.cores</code> is NULL, 2 workers are specified.</p>
<p>When BiocParallel is attached, the <code>summarizeOverlaps</code> function will iterate over the files using <code>bplapply</code> from the BiocParallel package. That function will check the R session for specific parallelization configuration information. If it finds none, it will check for multiple cores and make arrangements to use them if present. The “check” occurs via the function <code>bpparam</code>.</p>
<p>The default situation on a MacBook Air running MacOSX 10.9.5 with an Intel core i7 processor, (two physical cores with two logical cores each, allowing for four concurrent threads) as follows.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">options(mc.cores=NULL)</div><div class="line">library(BiocParallel)</div><div class="line">register(MulticoreParam())</div><div class="line">bpparam()</div><div class="line">## class: MulticoreParam</div><div class="line">##   bpisup: FALSE; bpnworkers: 2; bptasks: 0; bpjobname: BPJOB</div><div class="line">##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</div><div class="line">##   bptimeout: 2592000; bpprogressbar: FALSE</div><div class="line">##   bpRNGseed: </div><div class="line">##   bplogdir: NA</div><div class="line">##   bpresultdir: NA</div><div class="line">##   cluster type: FORK</div></pre></td></tr></table></figure>
<p>This identifies an object called <code>MulticoreParam</code> which is used to configure the behavior of <code>bplapply</code> and other utilities of the BiocParallel package. There are various configuration object classes that can be used.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">‘SnowParam’: distributed memory computing</div><div class="line"></div><div class="line">‘MulticoreParam’: shared memory computing</div><div class="line"></div><div class="line">‘BatchJobsParam’: scheduled cluster computing</div><div class="line"></div><div class="line">‘DoparParam’: foreach computing</div><div class="line"></div><div class="line">‘SerialParam’: non-parallel execution</div></pre></td></tr></table></figure>
<p>We need to use <code>register</code> to determine the type of concurrent computation that will be performed.</p>
<p>If process size is large, we may want to leave some cores idle. We can accomplish that by using <code>register</code>. Here we’ll check for any advantage of using all logical cores.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">library(BiocParallel)</div><div class="line">register(MulticoreParam(workers=1))</div><div class="line">system.time(i3 &lt;- summarizeOverlaps( ebg, fns ))</div><div class="line">##    user  system elapsed </div><div class="line">##  15.653   1.286  17.070</div><div class="line">register(MulticoreParam(workers=2))</div><div class="line">system.time(i3 &lt;- summarizeOverlaps( ebg, fns ))</div><div class="line">##    user  system elapsed </div><div class="line">##   0.103   0.029   8.818</div><div class="line">register(MulticoreParam(workers=3))</div><div class="line">system.time(i3 &lt;- summarizeOverlaps( ebg, fns ))</div><div class="line">##    user  system elapsed </div><div class="line">##   0.051   0.023   9.543</div><div class="line">register(MulticoreParam(workers=4))</div><div class="line">system.time(i3 &lt;- summarizeOverlaps( ebg, fns ))</div><div class="line">##    user  system elapsed </div><div class="line">##   0.056   0.026   8.774</div><div class="line">all.equal(i3,i2)  # check that the results do not change</div><div class="line">## [1] TRUE</div></pre></td></tr></table></figure>
<p>Note that in this environment, despite increasing the number of CPUs linearly do not appreciably reduce run time after moving to two cores. This due mainly to communication costs that typically increase with the number of CPUs, and this phenomenon will be environment-specific.</p>
<p>In summary, it is very easy to perform embarrassingly parallel tasks with R, and this carries over to genomic data analysis thanks to BiocParallel. There are some strategic considerations concerning control of memory consumption and communication costs, and full mastery of the topic involves attention to profiling and benchmarking methods that can be addressed in an advanced software development course.</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc2_parallel.html" target="_blank" rel="external">Architecture: considerations on high performance computing with Bioconductor</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/15/Genomics Data Analysis Series/GDAS015-NGS的可视化/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/15/Genomics Data Analysis Series/GDAS015-NGS的可视化/" itemprop="url">GDAS015-NGS的可视化</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-15T12:00:00+08:00">
                2019-09-15
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,728
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  9
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="ngs的可视化">NGS的可视化</h2>
<p>我们将用以下的包来说明一下如何对NGS数据进行可视化。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("pasillaBamSubset")</span></div><div class="line"><span class="comment">#biocLite("TxDb.Dmelanogaster.UCSC.dm3.ensGene")</span></div><div class="line"><span class="keyword">library</span>(pasillaBamSubset)</div><div class="line"><span class="keyword">library</span>(TxDb.Dmelanogaster.UCSC.dm3.ensGene)</div><div class="line">fl1 &lt;- untreated1_chr4()</div><div class="line">fl2 &lt;- untreated3_chr4()</div></pre></td></tr></table></figure>
<p>我们将会采用四种方式来查看NGS的数据，分别是IGV，<code>plot</code>,<code>Gviz</code>和<code>ggbio</code>。</p>
<h2 id="igv">IGV</h2>
<p>将这些文献从R library的目录中复制到当前的工作目标。首先将工作目录设置为源文件位置。我们需要使用<code>Rsamtools</code>包来对BAM文件加上索引，然后使用IGV来查看，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">file.copy(from=fl1,to=basename(fl1))</div><div class="line"><span class="comment">## [1] FALSE</span></div><div class="line">file.copy(from=fl2,to=basename(fl2))</div><div class="line"><span class="comment">## [1] FALSE</span></div><div class="line"><span class="keyword">library</span>(Rsamtools)</div><div class="line"><span class="comment">## Loading required package: Biostrings</span></div><div class="line"><span class="comment">## Loading required package: XVector</span></div><div class="line">indexBam(basename(fl1))</div><div class="line"><span class="comment">##       untreated1_chr4.bam </span></div><div class="line"><span class="comment">## "untreated1_chr4.bam.bai"</span></div><div class="line">indexBam(basename(fl2))</div><div class="line"><span class="comment">##       untreated3_chr4.bam </span></div><div class="line"><span class="comment">## "untreated3_chr4.bam.bai"</span></div></pre></td></tr></table></figure>
<p>IGV可以在这个网站下载: https://www.broadinstitute.org/igv/home</p>
<p>下载的时候需要提供你的电子邮箱，现在我们使用IGV来查看一个基因，例如<code>igs</code>。</p>
<p>如果无法下载IGV，那么可以使用UCSC的 Genome Browser：http://genome.ucsc.edu/cgi-bin/hgTracks</p>
<p>UCSC Genome Browser是一个很好的资源，有许多涉及基因注释，多个物种的保护，以及ENCODE表观遗传信号。但是，UCSC Genome Browser要求你将基因组文件上传到他们的服务器，或将数据放在公共可用服务器上。如果你处理的数据涉及保密，那就另当别论了。</p>
<h2 id="simple-plot">Simple plot</h2>
<p>现在我们来看一使用<code>plot()</code>函数来绘制基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicRanges)</div></pre></td></tr></table></figure>
<p>注意：如果你使用的是Bioconductor 14，它与R 3.1配合使用，需要加载这个包。如想是Bioconductor 13（它与R 3.0.x配合使用），则不需要加载这个包。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line"><span class="comment">## Loading required package: SummarizedExperiment</span></div></pre></td></tr></table></figure>
<p>我们从文件<code>fl1</code>中读取比对结果(alignments)。然后我们使用<code>coverage</code>函数来计算碱基对的覆盖。然后我们提取与我们感兴趣的基因重叠的覆盖区子集，并将此覆盖率从<code>RleList</code>转换为<code>numeric</code>向量。我们前面提到，<code>Rle</code>对象是一种压缩对象，例如重复数字会被压缩为2个数字，其中一个是重复的数字，另外一个是表示这个数字重复的次数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">x &lt;- readGAlignments(fl1)</div><div class="line">xcov &lt;- coverage(x)</div><div class="line">z &lt;- GRanges(<span class="string">"chr4"</span>,IRanges(<span class="number">456500</span>,<span class="number">466000</span>))</div><div class="line"><span class="comment"># Bioconductor 2.14</span></div><div class="line">xcov[z]</div><div class="line"><span class="comment">## RleList of length 1</span></div><div class="line"><span class="comment">## $$chr4</span></div><div class="line"><span class="comment">## integer-Rle of length 9501 with 1775 runs</span></div><div class="line"><span class="comment">##   Lengths: 1252   10   52    4    7    2 ...   10    7   12  392   75 1041</span></div><div class="line"><span class="comment">##   Values :    0    2    3    4    5    6 ...    3    2    1    0    1    0</span></div><div class="line"><span class="comment"># Bioconductor 2.13</span></div><div class="line">xcov$chr4[ranges(z)]</div><div class="line"><span class="comment">## integer-Rle of length 9501 with 1775 runs</span></div><div class="line"><span class="comment">##   Lengths: 1252   10   52    4    7    2 ...   10    7   12  392   75 1041</span></div><div class="line"><span class="comment">##   Values :    0    2    3    4    5    6 ...    3    2    1    0    1    0</span></div><div class="line">xnum &lt;- as.numeric(xcov$chr4[ranges(z)])</div><div class="line">plot(xnum)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-5-1.png" alt="plot of chunk unnamed-chunk-5">
<p class="caption">plot of chunk unnamed-chunk-5</p>
</div>
<p>现在我们读取另外一个文件<code>fl2</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">y &lt;- readGAlignmentPairs(fl2)</div><div class="line"><span class="comment">## Warning in .make_GAlignmentPairs_from_GAlignments(gal, strandMode = strandMode, : 63 pairs (0 proper, 63 not proper) were dropped because the seqname</span></div><div class="line"><span class="comment">##   or strand of the alignments in the pair were not concordant.</span></div><div class="line"><span class="comment">##   Note that a GAlignmentPairs object can only hold concordant pairs at the</span></div><div class="line"><span class="comment">##   moment, that is, pairs where the 2 alignments are on the opposite strands</span></div><div class="line"><span class="comment">##   of the same reference sequence.</span></div><div class="line">ycov &lt;- coverage(y)</div><div class="line">ynum &lt;- as.numeric(ycov$chr4[ranges(z)])</div><div class="line">plot(xnum, type=<span class="string">"l"</span>, col=<span class="string">"blue"</span>, lwd=<span class="number">2</span>)</div><div class="line">lines(ynum, col=<span class="string">"red"</span>, lwd=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<p>我们可以单独放大一个外显子，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">plot(xnum, type=<span class="string">"l"</span>, col=<span class="string">"blue"</span>, lwd=<span class="number">2</span>, xlim=c(<span class="number">6200</span>,<span class="number">6600</span>))</div><div class="line">lines(ynum, col=<span class="string">"red"</span>, lwd=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-7-1.png" alt="plot of chunk unnamed-chunk-7">
<p class="caption">plot of chunk unnamed-chunk-7</p>
</div>
<h2 id="使用转录本数据库提取感兴趣的基因">使用转录本数据库提取感兴趣的基因</h2>
<p>假设我们想可视化基因igs。我们可以从转录本数据库<code>TxDb.Dmelanogaster.UCSC.dm3.ensGene</code> 中提取它，但我们首先要知道这个基因的Ensembl基因名，现在我们使用前面学到的函数来实现这一功能，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># biocLite("biomaRt")</span></div><div class="line"><span class="keyword">library</span>(biomaRt)</div><div class="line">m &lt;- useMart(<span class="string">"ensembl"</span>, dataset = <span class="string">"dmelanogaster_gene_ensembl"</span>)</div><div class="line">lf &lt;- listFilters(m)</div><div class="line">lf[grep(<span class="string">"name"</span>, lf$description, ignore.case=<span class="literal">TRUE</span>),]</div><div class="line"><span class="comment">##                             name</span></div><div class="line"><span class="comment">## 1                chromosome_name</span></div><div class="line"><span class="comment">## 17         with_flybasename_gene</span></div><div class="line"><span class="comment">## 19   with_flybasename_transcript</span></div><div class="line"><span class="comment">## 23  with_flybasename_translation</span></div><div class="line"><span class="comment">## 52            external_gene_name</span></div><div class="line"><span class="comment">## 66              flybasename_gene</span></div><div class="line"><span class="comment">## 67        flybasename_transcript</span></div><div class="line"><span class="comment">## 68       flybasename_translation</span></div><div class="line"><span class="comment">## 93                 wikigene_name</span></div><div class="line"><span class="comment">## 107               go_parent_name</span></div><div class="line"><span class="comment">## 208               so_parent_name</span></div><div class="line"><span class="comment">##                                      description</span></div><div class="line"><span class="comment">## 1                                Chromosome name</span></div><div class="line"><span class="comment">## 17                   with FlyBaseName gene ID(s)</span></div><div class="line"><span class="comment">## 19             with FlyBaseName transcript ID(s)</span></div><div class="line"><span class="comment">## 23                with FlyBaseName protein ID(s)</span></div><div class="line"><span class="comment">## 52          Associated Gene Name(s) [e.g. BRCA2]</span></div><div class="line"><span class="comment">## 66           FlyBaseName Gene ID(s) [e.g. cul-2]</span></div><div class="line"><span class="comment">## 67  FlyBaseName Transcript ID(s) [e.g. cul-2-RB]</span></div><div class="line"><span class="comment">## 68     FlyBaseName Protein ID(s) [e.g. cul-2-PB]</span></div><div class="line"><span class="comment">## 93                 WikiGene Name(s) [e.g. Ir21a]</span></div><div class="line"><span class="comment">## 107                             Parent term name</span></div><div class="line"><span class="comment">## 208                             Parent term name</span></div><div class="line">map &lt;- getBM(mart = m,</div><div class="line">  attributes = c(<span class="string">"ensembl_gene_id"</span>, <span class="string">"flybasename_gene"</span>),</div><div class="line">  filters = <span class="string">"flybasename_gene"</span>, </div><div class="line">  values = <span class="string">"lgs"</span>)</div><div class="line">map</div><div class="line"><span class="comment">##   ensembl_gene_id flybasename_gene</span></div><div class="line"><span class="comment">## 1     FBgn0039907              lgs</span></div></pre></td></tr></table></figure>
<p>现在我们提取每个基因的外显子，然后提取igs基因的外显子，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicFeatures)</div><div class="line">grl &lt;- exonsBy(TxDb.Dmelanogaster.UCSC.dm3.ensGene, by=<span class="string">"gene"</span>)</div><div class="line">gene &lt;- grl[[map$ensembl_gene_id[<span class="number">1</span>]]]</div><div class="line">gene</div><div class="line"><span class="comment">## GRanges object with 6 ranges and 2 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames           ranges strand |   exon_id   exon_name</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##   [1]     chr4 [457583, 459544]      - |     63350        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [2]     chr4 [459601, 459791]      - |     63351        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [3]     chr4 [460074, 462077]      - |     63352        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [4]     chr4 [462806, 463015]      - |     63353        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [5]     chr4 [463490, 463780]      - |     63354        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [6]     chr4 [463839, 464533]      - |     63355        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 15 sequences (1 circular) from dm3 genome</span></div></pre></td></tr></table></figure>
<p>最后，我们绘制出这些外显子的范围，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">rg &lt;- range(gene)</div><div class="line">plot(c(start(rg), end(rg)), c(0,0), type=&quot;n&quot;, xlab=seqnames(gene)[1], ylab=&quot;&quot;)</div><div class="line">arrows(start(gene),rep(0,length(gene)),</div><div class="line">       end(gene),rep(0,length(gene)),</div><div class="line">       lwd=3, length=.1)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>不过实际上这个基因位于负链，因此我们需要改一下箭头的方向，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">plot(c(start(rg), end(rg)), c(<span class="number">0</span>,<span class="number">0</span>), type=<span class="string">"n"</span>, xlab=seqnames(gene)[<span class="number">1</span>], ylab=<span class="string">""</span>)</div><div class="line">arrows(start(gene),rep(<span class="number">0</span>,length(gene)),</div><div class="line">       end(gene),rep(<span class="number">0</span>,length(gene)),</div><div class="line">       lwd=<span class="number">3</span>, length=<span class="number">.1</span>, </div><div class="line">       code=ifelse(as.character(strand(gene)[<span class="number">1</span>]) == <span class="string">"+"</span>, <span class="number">2</span>, <span class="number">1</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11">
<p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<h2 id="gviz">Gviz</h2>
<p>我们将简要展示两个在Bioconductor中可视化基因组数据的软件包。请注意，这两个包中的任何一个都可以用于绘制多种数据的图形。我们将在这里展示如何像以前一样制作覆盖图(converage plot)：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("Gviz")</span></div><div class="line"><span class="keyword">library</span>(Gviz)</div><div class="line"><span class="comment">## Loading required package: grid</span></div><div class="line">gtrack &lt;- GenomeAxisTrack()</div><div class="line">atrack &lt;- AnnotationTrack(gene, name = <span class="string">"Gene Model"</span>)</div><div class="line">plotTracks(list(gtrack, atrack))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-12-1.png" alt="plot of chunk unnamed-chunk-12">
<p class="caption">plot of chunk unnamed-chunk-12</p>
</div>
<p>提取覆盖。<code>Gviz</code>中的数据最好是是<code>GRanges</code>对象，因此我们需要将<code>RleList</code>转换为<code>GRanges</code>对象，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">xgr &lt;- as(xcov, <span class="string">"GRanges"</span>)</div><div class="line">ygr &lt;- as(ycov, <span class="string">"GRanges"</span>)</div><div class="line">dtrack1 &lt;- DataTrack(xgr[xgr %over% z], name = <span class="string">"sample 1"</span>)</div><div class="line">dtrack2 &lt;- DataTrack(ygr[ygr %over% z], name = <span class="string">"sample 2"</span>)</div><div class="line">plotTracks(list(gtrack, atrack, dtrack1, dtrack2))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-1.png" alt="plot of chunk unnamed-chunk-13">
<p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">plotTracks(list(gtrack, atrack, dtrack1, dtrack2), type=<span class="string">"polygon"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-2.png" alt="plot of chunk unnamed-chunk-13">
<p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<h2 id="ggbio">ggbio</h2>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("ggbio")</span></div><div class="line"><span class="keyword">library</span>(ggbio)</div><div class="line"><span class="comment">## Loading required package: ggplot2</span></div><div class="line"><span class="comment">## Warning: replacing previous import by 'BiocGenerics::Position' when loading</span></div><div class="line"><span class="comment">## 'ggbio'</span></div><div class="line"><span class="comment">## Need specific help about ggbio? try mailing </span></div><div class="line"><span class="comment">##  the maintainer or visit http://tengfei.github.com/ggbio/</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Attaching package: 'ggbio'</span></div><div class="line"><span class="comment">## The following objects are masked from 'package:ggplot2':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     geom_bar, geom_rect, geom_segment, ggsave, stat_bin,</span></div><div class="line"><span class="comment">##     stat_identity, xlim</span></div><div class="line">autoplot(gene)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-1.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">autoplot(fl1, which=z)</div><div class="line"><span class="comment">## reading in as Bamfile</span></div><div class="line"><span class="comment">## Parsing raw coverage...</span></div><div class="line"><span class="comment">## Read GAlignments from BamFile...</span></div><div class="line"><span class="comment">## extracting information...</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-2.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">autoplot(fl2, which=z)</div><div class="line"><span class="comment">## reading in as Bamfile</span></div><div class="line"><span class="comment">## Parsing raw coverage...</span></div><div class="line"><span class="comment">## Read GAlignments from BamFile...</span></div><div class="line"><span class="comment">## extracting information...</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-3.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc2_vizNGS.html" target="_blank" rel="external">Visualizing NGS data</a></li>
<li>IGV: https://www.broadinstitute.org/igv/home</li>
<li>Gviz: http://www.bioconductor.org/packages/release/bioc/html/Gviz.html</li>
<li>ggbio:http://www.bioconductor.org/packages/release/bioc/html/ggbio.html</li>
<li>UCSC Genome Browser: zooms and scrolls over chromosomes, showing the work of annotators worldwide: http://genome.ucsc.edu/</li>
<li>Ensembl genome browser: genome databases for vertebrates and other eukaryotic species:http://ensembl.org</li>
<li>Roadmap Epigenome browser: public resource of human epigenomic data: http://www.epigenomebrowser.org http://genomebrowser.wustl.edu/ http://epigenomegateway.wustl.edu/</li>
<li>“Sashimi plots” for RNA-Seq: http://genes.mit.edu/burgelab/miso/docs/sashimi.html</li>
<li>Circos: designed for visualizing genomic data in a cirlce: http://circos.ca/</li>
<li>SeqMonk: a tool to visualise and analyse high throughput mapped sequence data: http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/14/Genomics Data Analysis Series/GDAS014-使用autoplot和Gviz绘图/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/14/Genomics Data Analysis Series/GDAS014-使用autoplot和Gviz绘图/" itemprop="url">GDAS014-使用autoplot和Gviz绘图</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-14T12:00:00+08:00">
                2019-09-14
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  902
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  4
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="使用ggbio包绘制染色体图">使用<code>ggbio</code>包绘制染色体图</h2>
<p>我们可以通过各种方法来绘制基因组级规模的数据。Bioiconductor中有两个这样的包，即<code>Gviz</code> 和 <code>ggbio</code>。这两个包旨在绘制基因组级数据结构的图形示意图。</p>
<p><code>ggbio</code> 包中的 <code>autoplot</code> 方法用途广泛。对于GRanges实例来说，存在于每个范围中的数据都可以绘制成染色体上条带。核布图(karyogram)布局提供了全基因组的视图，不过这种图在处理染色体外序列水平方面有着重要作用。</p>
<p>原文如下：</p>
<blockquote>
<p>The karyogram layout gives a genome-wide view, but it can be important to control the handling of extra-chromosomal sequence levels.</p>
</blockquote>
<p>以下是肝细胞系的布局：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ERBS)</div><div class="line">data(HepG2)</div><div class="line"><span class="keyword">library</span>(GenomeInfoDb)  <span class="comment"># trim all but autosomal chroms</span></div><div class="line">HepG2 = keepStandardChromosomes(HepG2)</div><div class="line">data(GM12878)</div><div class="line">GM12878 = keepStandardChromosomes(GM12878)</div><div class="line"><span class="keyword">library</span>(ggbio)</div><div class="line">autoplot(HepG2, layout=<span class="string">"karyogram"</span>, main=<span class="string">"ESRRA binding on HepG2"</span>)</div><div class="line"><span class="comment">## Scale for 'x' is already present. Adding another scale for 'x', which</span></div><div class="line"><span class="comment">## will replace the existing scale.</span></div><div class="line"><span class="comment">## Scale for 'x' is already present. Adding another scale for 'x', which</span></div><div class="line"><span class="comment">## will replace the existing scale.</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_ggbio-lkd-1.png" alt="plot of chunk lkd">
<p class="caption">plot of chunk lkd</p>
</div>
<p>以下是B细胞系的布局：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">autoplot(GM12878, layout=&quot;karyogram&quot;, main=&quot;ESRRA binding on GM12878&quot;)</div><div class="line">## Scale for &apos;x&apos; is already present. Adding another scale for &apos;x&apos;, which</div><div class="line">## will replace the existing scale.</div><div class="line">## Scale for &apos;x&apos; is already present. Adding another scale for &apos;x&apos;, which</div><div class="line">## will replace the existing scale.</div></pre></td></tr></table></figure>
<h2 id="使用gviz包绘制esrra结合位点与基因注释">使用<code>Gviz</code>包绘制ESRRA结合位点与基因注释</h2>
<p>我们通常会将观察到的数据放在基因组的参考信息中来绘制这些数据的图形。现在我们演示一下如何ESRRA结合的图形。</p>
<p>第一步我们先加载相关的数据、注释包与<code>Gviz</code>包，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ERBS)</div><div class="line"><span class="keyword">library</span>(Gviz)</div><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line"><span class="keyword">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene</div></pre></td></tr></table></figure>
<h3 id="esrra附近的基因">ESRRA附近的基因</h3>
<p>我们如何确定含有ESRRA以该基因附近的人类基因组片段呢？有几种方法，我们先从ENTREZ ID开始，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">eid = select(Homo.sapiens, keys=<span class="string">"ESRRA"</span>, keytype=<span class="string">"SYMBOL"</span>, columns=<span class="string">"ENTREZID"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line">eid</div><div class="line"><span class="comment">##   SYMBOL ENTREZID</span></div><div class="line"><span class="comment">## 1  ESRRA     2101</span></div></pre></td></tr></table></figure>
<p>现在我们获取ESRRA基因的地址，并收拾它相邻基因的地址，并绑定这些基因的符号，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">allg = genes(txdb)</div><div class="line">esrraAddr = genes(txdb, filter=list(gene_id=<span class="number">2101</span>)) <span class="comment"># redundant...</span></div><div class="line">esrraNeigh = subsetByOverlaps(allg, esrraAddr+<span class="number">500000</span>)</div><div class="line">esrraNeigh$symbol = mapIds(Homo.sapiens, keys=esrraNeigh$gene_id, keytype=<span class="string">"ENTREZID"</span>,</div><div class="line">  column=<span class="string">"SYMBOL"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div></pre></td></tr></table></figure>
<p>使用<code>Gviz</code> 包快速绘制图形，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">plotTracks(GeneRegionTrack(esrraNeigh, showId=TRUE))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_gvfeat-lknei-1.png" alt="plot of chunk lknei">
<p class="caption">plot of chunk lknei</p>
</div>
<h3 id="此区域内的esrra结合峰">此区域内的ESRRA结合峰</h3>
<p>我们从某个样本（GM12878 EBV-transformed B-cell）获取ESRRA结合数据，以及我们目标基因附近的子集，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">data(GM12878)</div><div class="line">sc = subsetByOverlaps(GM12878, range(esrraNeigh))</div><div class="line">sc</div><div class="line"><span class="comment">## GRanges object with 6 ranges and 7 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames               ranges strand |      name     score       col</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;logical&gt;</span></div><div class="line"><span class="comment">##   [1]    chr11 [64071338, 64073242]      * |         3         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [2]    chr11 [63913452, 63914077]      * |        61         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [3]    chr11 [63741506, 63742754]      * |       210         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [4]    chr11 [63992515, 63994725]      * |       408         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [5]    chr11 [64098744, 64100083]      * |      1778         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [6]    chr11 [63639263, 63640021]      * |      2248         0      &lt;NA&gt;</span></div><div class="line"><span class="comment">##       signalValue    pValue    qValue      peak</span></div><div class="line"><span class="comment">##         &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]       69.87       310        32      1040</span></div><div class="line"><span class="comment">##   [2]       52.48    123.72  1.785156       259</span></div><div class="line"><span class="comment">##   [3]       18.24    59.652  2.022276       622</span></div><div class="line"><span class="comment">##   [4]        9.04    41.001  1.910095       990</span></div><div class="line"><span class="comment">##   [5]        9.03    19.758  2.075721       653</span></div><div class="line"><span class="comment">##   [6]        8.65    17.481  2.031517       436</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 93 sequences (1 circular) from hg19 genome</span></div></pre></td></tr></table></figure>
<h3 id="计算表义文字ideogram从而给出染色体体的信息">计算表义文字(ideogram)从而给出染色体体的信息</h3>
<p>计算过程有点慢。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">idxTrack = IdeogramTrack(genome=<span class="string">"hg19"</span>, chr=<span class="string">"chr11"</span>)</div></pre></td></tr></table></figure>
<h3 id="数据合并">数据合并</h3>
<p>我们从顶部开始，用表意文字来确定染色体和染色体上的区域，并且通过观察到的数据和结构信息放大这个区域，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">plotTracks(list(idxTrack, GenomeAxisTrack(), </div><div class="line">   DataTrack(sc[,<span class="number">7</span>], name=<span class="string">"ESRRA peak values"</span>), </div><div class="line">   GeneRegionTrack(esrraNeigh, showId=<span class="literal">TRUE</span>,</div><div class="line">         name=<span class="string">"genes near ESRRA"</span>), GenomeAxisTrack()))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_gvfeat-dofull-1.png" alt="plot of chunk dofull">
<p class="caption">plot of chunk dofull</p>
</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc2_ggbio.html" target="_blank" rel="external">Sketching the binding landscape over chromosomes with ggbio’s karyogram layout</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/bioc2_gvfeat.html" target="_blank" rel="external">Gviz for plotting data with genomic features</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/13/Genomics Data Analysis Series/GDAS013- 基因集分析/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/13/Genomics Data Analysis Series/GDAS013- 基因集分析/" itemprop="url">GDAS013-基因集分析</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-13T12:00:00+08:00">
                2019-09-13
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  4,393
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  19
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>在这一节中我们将学习基因集分析(gene set analysis)背后的统计学原理。在这一部分里，我们要记住三个重要的步骤，这三个步骤都非常重要，但它们又相互独立，分别如下所示：</p>
<ul>
<li>参与统计的基因集</li>
<li>汇总每个基因集的统计学指标</li>
<li>评估不确定性的方法</li>
</ul>
<h2 id="准备数据与基因集数据库">准备数据与基因集数据库</h2>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">source</span>(<span class="string">"http://www.bioconductor.org/biocLite.R"</span>)</div><div class="line">biocLite(<span class="string">"GSEABase"</span>)</div></pre></td></tr></table></figure>
<p>加载以下包：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(rafalib)</div><div class="line">library(GSEABase)</div><div class="line">library(GSE5859Subset)</div><div class="line">library(sva)</div><div class="line">library(limma)</div><div class="line">library(matrixStats)</div></pre></td></tr></table></figure>
<p>我们使用下面包中的数据来说明一下，这个包可以从Github中下载，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(devtools)</div><div class="line">install_github(<span class="string">"genomicsclass/GSE5859Subset"</span>)</div></pre></td></tr></table></figure>
<p>我们构建的这个数据含有批次效应，因此我们需要使用SVA对其进行校正，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSE5859Subset)</div><div class="line">data(GSE5859Subset)</div><div class="line"><span class="keyword">library</span>(sva)</div><div class="line"><span class="keyword">library</span>(limma)</div><div class="line"><span class="keyword">library</span>(matrixStats)</div><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">X = sampleInfo$group</div><div class="line">mod&lt;-model.matrix(~X)</div><div class="line">svafit &lt;- sva(geneExpression,mod)</div><div class="line"><span class="comment">## Number of significant surrogate variables is:  5 </span></div><div class="line"><span class="comment">## Iteration (out of 5 ):1  2  3  4  5</span></div><div class="line">svaX&lt;-model.matrix(~X+svafit$sv)</div><div class="line">lmfit &lt;- lmFit(geneExpression,svaX)</div><div class="line">tt&lt;-lmfit$coef[,<span class="number">2</span>]*sqrt(lmfit$df.residual)/(<span class="number">2</span>*lmfit$sigma)</div><div class="line">pval&lt;-<span class="number">2</span>*(<span class="number">1</span>-pt(abs(tt),lmfit$df.residual[<span class="number">1</span>]))</div><div class="line">qval &lt;- p.adjust(pval,<span class="string">"BH"</span>)</div></pre></td></tr></table></figure>
<p>我们需要注意一下，在这一节中，我们需要每个基因的评分和差异表达基因基因的候选列表。我们如何获取这个候选基因列表与所表示的基因丰度的单位无关。为了说明这一点，我们会假装我们事先不知道在不同组之间基因表达的关系。</p>
<p>我们构建一个FDR为25%的候选基因列表。这个列表中有23个基因。我们从中能了解到什么生物学信息呢？既然我们知道基因存在着相互作用，那么研究一组基因，而不是单个基因有着重要意义。通常来说，这就是基因集分析的原始想法。</p>
<p>网上有各种提供基因集的资源。这些资源中的基因都是经过缜密筛选的，它们都有着一些共同的特征。这些基因分组的方式是通过通路（例如KEGG）或GO来划分的。在下面的网站上就能找到许多这样的资源：</p>
<p>http://www.broadinstitute.org/gsea/msigdb/index.jsp</p>
<p>在以下的案例中，我闪将使用一个根据染色体进行分组的基因集来分析数据。第一步就是下载数据库，不过你需要在<a href="http://software.broadinstitute.org/gsea/login.jsp;jsessionid=C7D472C24D6D4FF96526F3868A8CCCC9" target="_blank" rel="external">网站注册</a>。</p>
<p>当你下载好文件后，就可以通过以下的命令读入数据库（文件扩展名为gmt，其实就是GSEA分析中的基因集文件），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">gsets &lt;- getGmt(<span class="string">"c1.all.v6.1.entrez.gmt"</span>) <span class="comment">##must provide correct path to file</span></div></pre></td></tr></table></figure>
<p>这个对象中含有326个基因集。每个基因集使用ENTREZ ID标明了基因，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">length(gsets)</div><div class="line">## [1] 326</div><div class="line">head(names(gsets))</div><div class="line">## [1] &quot;chr5q23&quot;  &quot;chr16q24&quot; &quot;chr8q24&quot;  &quot;chr13q11&quot; &quot;chr7p21&quot;  &quot;chr10q23&quot;</div><div class="line">gsets[[&quot;chryq11&quot;]]</div><div class="line">## setName: chryq11 </div><div class="line">## geneIds: 728512, 392597, ..., 360027 (total: 204)</div><div class="line">## geneIdType: Null</div><div class="line">## collectionType: Null </div><div class="line">## details: use &apos;details(object)&apos;</div></pre></td></tr></table></figure>
<p>我们可以使用以下方式获取ENTREZ ID，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">head(geneIds(gsets[[&quot;chryq11&quot;]]))</div><div class="line">## [1] &quot;728512&quot; &quot;392597&quot; &quot;9086&quot;   &quot;359796&quot; &quot;83864&quot;  &quot;425057&quot;</div></pre></td></tr></table></figure>
<p>现在，为了将其应用于我们的数据，我们必须将这些ID映射到Affymetrix ID上。现在我们编写这样一个函数，这个函数会使用Expression Set类，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">mapGMT2Affy &lt;- <span class="keyword">function</span>(object,gsets)&#123;</div><div class="line">  ann&lt;-annotation(object)</div><div class="line">  dbname&lt;-paste(ann,<span class="string">"db"</span>,sep=<span class="string">"."</span>)</div><div class="line">  <span class="keyword">require</span>(dbname,character.only=<span class="literal">TRUE</span>)</div><div class="line">  gns&lt;-featureNames(object)</div><div class="line"></div><div class="line">  <span class="comment">##This call may generate warnings</span></div><div class="line">  map&lt;-select(get(dbname), keys=gns,columns=c(<span class="string">"ENTREZID"</span>, <span class="string">"PROBEID"</span>))</div><div class="line">  map&lt;-split(map[,<span class="number">1</span>],map[,<span class="number">2</span>])</div><div class="line">  indexes&lt;-sapply(gsets,<span class="keyword">function</span>(ids)&#123;</div><div class="line">    gns2&lt;-unlist(map[geneIds(ids)])</div><div class="line">    match(gns2,gns)</div><div class="line">    &#125;)</div><div class="line">  names(indexes)&lt;-names(gsets)</div><div class="line">  <span class="keyword">return</span>(indexes)</div><div class="line">&#125;</div><div class="line"><span class="comment">##create an Expression Set</span></div><div class="line">rownames(sampleInfo)&lt;- colnames(geneExpression)</div><div class="line">e=ExpressionSet(assay=geneExpression,</div><div class="line">                phenoData=AnnotatedDataFrame(sampleInfo),</div><div class="line">                annotation=<span class="string">"hgfocus"</span>)</div><div class="line"><span class="comment">##can safely ignore the warning</span></div><div class="line">gsids &lt;- mapGMT2Affy(e,gsets) </div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div></pre></td></tr></table></figure>
<h2 id="基于相关检验的方法">基于相关检验的方法</h2>
<p>现在我们就可以计算差异表达基因是否在某个基因集中富集。那么我们如何计算呢？最简单的方法就是独立卡方检验，这种统计方法我们以前学过。现在我们对这些差异基因是否在Y染色体上的基因集中富集进行计算，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">tab &lt;- table(ingenset=<span class="number">1</span>:nrow(e) %<span class="keyword">in</span>% gsids[[<span class="string">"chryq11"</span>]],signif=qval&lt;<span class="number">0.05</span>)</div><div class="line">tab</div><div class="line"><span class="comment">##         signif</span></div><div class="line"><span class="comment">## ingenset FALSE TRUE</span></div><div class="line"><span class="comment">##    FALSE  8769    9</span></div><div class="line"><span class="comment">##    TRUE     11    4</span></div><div class="line">chisq.test(tab)$p.val</div><div class="line"><span class="comment">## Warning in chisq.test(tab): Chi-squared approximation may be incorrect</span></div><div class="line"><span class="comment">## [1] 5.264033e-121</span></div></pre></td></tr></table></figure>
<p>因为我们比较的是男性与女性，因此我们使用Y染色体上基因集来进行富集，这样比较便于富集差异基因，更能说明这种算法的原理。但是，由于只有13个基因有显著性（在Y染色体上的只有4个），这种方法对于寻找可能在某些方面具有生物学意义的基因集来说，实际意义不大。例如，X染色体上的基因集是我们感兴趣的基因集么？这里需要注意的是，一些基因在X染色体上并不活性，因此与男性相比，女性体内这样的基因数目更多。</p>
<p>这里我们面临的问题是，FDR值为0.25，这是任意设置的一个截止值。为什么不设为0.05或0.01呢？我们如何设置这个截止值？</p>
<p>如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">qs&lt;-seq(<span class="number">0</span>,<span class="number">1</span>,len=length(tt)+<span class="number">1</span>)-<span class="number">1</span>/(<span class="number">2</span>*length(tt));qs&lt;-qs[-<span class="number">1</span>]</div><div class="line">qqplot(qt(qs,lmfit$df.resid),tt,ylim=c(-<span class="number">10</span>,<span class="number">10</span>),xlab=<span class="string">"Theoretical quantiles"</span>,ylab=<span class="string">"Observed"</span>) <span class="comment">##note we leave some of the obvious ones out</span></div><div class="line">abline(<span class="number">0</span>,<span class="number">1</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<h2 id="基因集相关的统计学">基因集相关的统计学</h2>
<p>目前大多数进行基因集分析的方法都没有将基因分析为差异表达部分和非差异表达部分。相反，我们会使用一些类似于t检验的方法来进行计算。为了了解这种方法比基因列表的二分法更为强大，现在我们提出一个问题：一些基因是否在X染色体上并没有失活？我们不像前面提到的那样创建一个基因表，而是查看一下X染色体中第一组基因t检验中t值的分布，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">ind &lt;- gsids[[<span class="string">"chrxp11"</span>]]</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">plot(density(tt[-ind]),xlim=c(-<span class="number">7</span>,<span class="number">7</span>),main=<span class="string">""</span>,xlab=<span class="string">"t-stat"</span>,sub=<span class="string">""</span>,lwd=<span class="number">4</span>)</div><div class="line">lines(density(tt[ind],bw=<span class="number">.7</span>),col=<span class="number">2</span>,lty=<span class="number">2</span>,lwd=<span class="number">4</span>)</div><div class="line">rug(tt[ind],col=<span class="number">2</span>)</div><div class="line">legend(-<span class="number">6.5</span>, <span class="number">.3</span>, legend=c(<span class="string">"present on xp11"</span>, <span class="string">"not on xp11"</span>),</div><div class="line">     col=c(<span class="number">2</span>,<span class="number">1</span>), lty=c(<span class="number">2</span>,<span class="number">1</span>), lwd=<span class="number">4</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11">
<p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<p>从上图我们可以看到基因集的t统计量的分布，它的分布似乎不同于所有基因的分布。那么我们如何地这个结果进行定量呢？</p>
<p>有关基因集分析的第一篇文献是于2001年由[Virtaneva]发表的，他在前面部分中讨论了Wilcoxon Mann-Whitney检验。那篇文献的基本思路就是统计实验组和对照组之间的差异基因在某个特定的基因集中的数目是否比在其它基因集中的数目更大。Wilcoxon比较的是基因集上基因的秩是否在中位数以上。</p>
<p>在这里我们计算了每个基因集的Wilcoxon检验。然后我们对Wilcoxon统计量的结果进行标准(standardize)，也就是将它们转换为均值为0，标准差为1的数据。注意，中心极限理论告诉我们，如果基因集足够大，并且每个基因的效应大小彼此独立，那么这些统计量将遵循正态分布，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">es &lt;- lmfit$coef[,<span class="number">2</span>]</div><div class="line">wilcox &lt;- t(sapply(gsids,<span class="keyword">function</span>(i)&#123;</div><div class="line">  <span class="keyword">if</span>(length(i)&gt;<span class="number">2</span>)&#123;</div><div class="line">  tmp&lt;- wilcox.test(es[i],es[-i])  </div><div class="line">  n1&lt;-length(i);n2&lt;-length(es)-n1</div><div class="line">  z &lt;- (tmp$stat -n1*n2/<span class="number">2</span>) / sqrt(n1*n2*(n1+n2+<span class="number">1</span>)/<span class="number">12</span>)</div><div class="line">  <span class="keyword">return</span>(c(z,tmp$p.value))</div><div class="line">  &#125; <span class="keyword">else</span> <span class="keyword">return</span>(rep(<span class="literal">NA</span>,<span class="number">2</span>))  </div><div class="line">  &#125;))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">cols &lt;- rep(<span class="number">1</span>,nrow(wilcox))</div><div class="line">cols[grep(<span class="string">"chrx"</span>,rownames(wilcox))]&lt;-<span class="number">2</span></div><div class="line">cols[grep(<span class="string">"chry"</span>,rownames(wilcox))]&lt;-<span class="number">3</span></div><div class="line">qqnorm(wilcox[,<span class="number">1</span>],col=cols,cex=ifelse(cols==<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>),pch=<span class="number">16</span>)</div><div class="line">qqline(wilcox[,<span class="number">1</span>])</div><div class="line">legend(<span class="string">"topleft"</span>,c(<span class="string">"Autosome"</span>,<span class="string">"chrX"</span>,<span class="string">"chrY"</span>),pch=<span class="number">16</span>,col=<span class="number">1</span>:<span class="number">3</span>,box.lwd=<span class="number">0</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-12-1.png" alt="plot of chunk unnamed-chunk-12">
<p class="caption">plot of chunk unnamed-chunk-12</p>
</div>
<p>从上图我们可以看出来，排名前10位的基因并没像我们预期的那样在X染色体和Y染色体中高表达。使用Wilcoxon（稳健,robust)为基础的背后逻辑是，我们并不想看到一个或两个高表达的基因对基因集分析产生重要影响（Wilcoxon检验是功效比较低，但比较稳健，对异常值不敏感）。虽然这种统计方法能够降低低秩基因的假阳性，但是这牺牲了灵敏度（功效,power）。下面我们来考虑一个低稳健性(less robust)，但是通常情况下更加强大的方法。</p>
<p>如果我们对基个基因集中向左分布（通常来说是表达量少的基因）的基因或者说是向右分布（通常来说是就是表达高的基因）的基因戌兴趣，那么我们可以使用的最简单的统计就是比较两组之间的均值，具体可以参考这篇文献<a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=20048385" target="_blank" rel="external">《Gene set enrichment analysis made simple》</a>，其中G是基因在基因集中的索引，我们定义以下的统计量： <span class="math display">\[
\bar{t}=\frac{1}{N} \sum_{i \in G} t_i
\]</span></p>
<p>零假设是基因集中的基因表达没有差异，这些t检验的均值为0。如果t检验统计量之间互相是独立的（后面我们会学习避免此假设的方法），那么<span class="math inline">\(\sqrt{N} \bar{t}\)</span> 服从以下分布： <span class="math display">\[
\sqrt{N} \bar{t} \sim N(0,1)
\]</span></p>
<p>以下是这些统计量的qq图：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">avgt &lt;- sapply(gsids,function(i) sqrt(length(i))*mean(tt[i]))</div><div class="line">qqnorm(avgt,col=cols,cex=ifelse(cols==1,1,2),pch=16)</div><div class="line">qqline(avgt)</div><div class="line">legend(&quot;topleft&quot;,c(&quot;Autosome&quot;,&quot;chrX&quot;,&quot;chrY&quot;),pch=16,col=1:3,box.lwd=0)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-13-1.png" alt="plot of chunk unnamed-chunk-13">
<p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">avgt[order(-abs(avgt))[<span class="number">1</span>:<span class="number">10</span>]]</div><div class="line"><span class="comment">##    chryp11    chryq11    chrxq13    chr6p22   chr17p13    chrxp11 </span></div><div class="line"><span class="comment">## -17.749865 -15.287203   6.592806  -4.558008  -4.545851   4.420686 </span></div><div class="line"><span class="comment">##   chr20p13    chr4q35    chr7p22    chrxp22 </span></div><div class="line"><span class="comment">##  -4.241892  -4.033233  -3.642068   3.608926</span></div></pre></td></tr></table></figure>
<p>有一些算法就是基于以上原理来转统计量的均值的，例如 <a href="http://www.ncbi.nlm.nih.gov/pubmed/16174746" target="_blank" rel="external">Tian et al 2005</a> , <a href="http://www.ncbi.nlm.nih.gov/pubmed/15647293" target="_blank" rel="external">SAFE</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=20610611" target="_blank" rel="external">ROAST</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/22638577" target="_blank" rel="external">CAMERA</a>.</p>
<p>我们可以构建其它的得分来检验其它替代假设。例如我们可以检验<a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=20048385" target="_blank" rel="external">changes in variances</a> 与这个分布的常规变化 [<a href="http://www.ncbi.nlm.nih.gov/pubmed/12808457" target="_blank" rel="external">1</a>], [<a href="http://www.ncbi.nlm.nih.gov/pubmed/16199517" target="_blank" rel="external">2</a>]。</p>
<h2 id="基因集的假设检验">基因集的假设检验</h2>
<p>上面我们对<span class="math inline">\(\sqrt{N} \bar{t}\)</span> 的正态近似表明，t检验的统计量是独立的，并且分布相同，例如，现在我们看第二个公式，如下所示： <span class="math display">\[
Var(\bar{t})= \frac{1}{N^2} \mbox{var}(t_1 + \dots + t_N) = \frac{1}{N^2}(\mbox{var}(t_1)+\dots+\mbox{var}(t_N)) = \frac{1}{N}
\]</span> 这个公众就是假设t检验统计量是独立的。但是，即使在零假设下，我们也许会考虑到基因集之间是相互关联的。为了了解这一点，我们可以取每个基因集中所有成对相关的平均值，我们就会看到，它们并不都是0，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">N &lt;- sapply(gsids,length)</div><div class="line">ind1&lt;-which(X==<span class="number">0</span>)</div><div class="line">ind2&lt;-which(X==<span class="number">1</span>)</div><div class="line">corrs &lt;- t(sapply(gsids,<span class="keyword">function</span>(ind)&#123;</div><div class="line">  <span class="keyword">if</span>(length(ind)&gt;=<span class="number">2</span>)&#123;</div><div class="line">    cc1&lt;-cor(t(geneExpression[ind,ind1]))</div><div class="line">    cc2&lt;-cor(t(geneExpression[ind,ind2]))</div><div class="line">  <span class="keyword">return</span>(c(median(cc1[lower.tri(cc1)]),</div><div class="line">    median(cc2[lower.tri(cc2)])))</div><div class="line">  &#125; <span class="keyword">else</span> <span class="keyword">return</span>(c(<span class="literal">NA</span>,<span class="literal">NA</span>))</div><div class="line">&#125;))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">plot(corrs[N&gt;<span class="number">10</span>,],xlim=c(-<span class="number">1</span>,<span class="number">1</span>),ylim=c(-<span class="number">1</span>,<span class="number">1</span>),xlab=<span class="string">"Correlation within females"</span>,ylab=<span class="string">"Correlation within males"</span>)</div><div class="line">abline(h=<span class="number">0</span>,v=<span class="number">0</span>,lty=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-14-1.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<p>注意，在这个案例中，在一个基因集中，基因的检验统计量具有相关性的 <span class="math inline">\(\rho\)</span>， 平均t检验统计量的方差如下所示： <span class="math display">\[
\begin{eqnarray}
Var(\bar{t}) &amp;=&amp; \frac{1}{N^2} \mbox{var}( (1 \dots 1) (t_1 \dots 
t_N)&#39; ) \\
&amp;=&amp; \frac{1}{N^2}(1 \dots 1)
\begin{pmatrix}
1 &amp; \rho &amp; \dots &amp; \rho &amp; \rho \\
\rho &amp; 1 &amp; \rho &amp; \dots  &amp; \rho \\
\dots &amp; \dots &amp; \dots &amp; \dots &amp; \dots \\
\rho &amp;  \rho &amp; \dots &amp; \rho &amp; 1 \\
\end{pmatrix}
(1 \dots 1) &#39; \\
&amp;=&amp; \frac{1}{N^2}\{N + (N-1) N \rho \} \\
&amp;=&amp; \frac{1}{N}\{1 + (N-1) \rho \} \\
\end{eqnarray} %]]&gt;
\]</span> 在上面公式里，<span class="math inline">\(\rho\)</span> 是正值，方差与 ${1 + (N-1) } <span class="math inline">\(有关。虽然常规情况下，在一个基因集内，\)</span>$ 并不一定都相等，但是我们仍然可以根据一个基因集内的平均 <span class="math inline">\(\rho\)</span> 来计算一个校正因子(correction factor)。这个公式可以参考这篇文献<a href="https://projecteuclid.org/euclid.aoas/1206367822" target="_blank" rel="external">《A statistical framework for testing functional categories in microarray data》</a>。公式中使用的校正方法还包括Wald统计量。</p>
<p>这些校正因子可以对上述提到的统计量进行简单的汇总。事实上，现在已经有论文提出了基于简单的均值转换方法的算法，这些算法还考虑到了相关性，它们基于渐进近似的<a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=20610611" target="_blank" rel="external">ROAST</a>和<a href="https://www.ncbi.nlm.nih.gov/pubmed/22638577" target="_blank" rel="external">CAMERA</a>，用于开发统计摘要(summary statistics)与统计推断。在计算机演示案例中，我们将会演示其中一款软件。简单起见，我们这里基于 <span class="math inline">\(\sqrt{N} \bar{t}\)</span> 计算近似，并使用校正因子。这里我们比较一下初始 <span class="math inline">\(\sqrt{N} \bar{t}\)</span> 以及校正后的 <span class="math inline">\(\sqrt{N} \bar{t}\)</span> ，如下所示：</p>
<p>在下图里，我们注意到这个校正会降低分数(socres)（也就是将分数值降低到接近0的程度）。尤其注意的是，第三个最高的值现在非常接近0，它不再显著（参见箭头），如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">avgcorrs &lt;- rowMeans(corrs)</div><div class="line">cf &lt;- (<span class="number">1</span>+(N-<span class="number">1</span>)*avgcorrs)</div><div class="line">cf[cf&lt;<span class="number">1</span>] &lt;- <span class="number">1</span> <span class="comment">## we ignore negative correlations</span></div><div class="line">correctedavgt &lt;- avgt/sqrt(cf)</div><div class="line">parampvaliid &lt;- <span class="number">2</span>*pnorm(-abs(avgt))</div><div class="line">parampval&lt;- <span class="number">2</span>*pnorm(-abs(correctedavgt))</div><div class="line">plot(avgt,correctedavgt,bg=cols,pch=<span class="number">21</span>,xlab=<span class="string">"Original"</span>,ylab=<span class="string">"With correction factor"</span>,xlim=c(-<span class="number">7</span>,<span class="number">20</span>),ylim=c(-<span class="number">7</span>,<span class="number">20</span>),cex=<span class="number">1.5</span>)</div><div class="line">abline(<span class="number">0</span>,<span class="number">1</span>)</div><div class="line">abline(h=<span class="number">0</span>,v=<span class="number">0</span>,lty=<span class="number">2</span>)</div><div class="line">thirdhighest &lt;- order(-avgt)[<span class="number">3</span>]</div><div class="line">arrows(avgt[thirdhighest]+<span class="number">3</span>,correctedavgt[thirdhighest],x1=avgt[thirdhighest]+<span class="number">0.5</span>,lwd=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-15-1.png" alt="plot of chunk unnamed-chunk-15">
<p class="caption">plot of chunk unnamed-chunk-15</p>
</div>
<p>需要注意，标注出来的基因集具有相对高的相关性和大小(size)，它就使得校正因子变得很大，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">avgcorrs[thirdhighest]</div><div class="line"><span class="comment">##   chrxp22 </span></div><div class="line"><span class="comment">## 0.0287052</span></div><div class="line">length(gsids[[thirdhighest]])</div><div class="line"><span class="comment">## [1] 76</span></div><div class="line">cf[thirdhighest]</div><div class="line"><span class="comment">## chrxp22 </span></div><div class="line"><span class="comment">## 3.15289</span></div></pre></td></tr></table></figure>
<p>在这一节中，我们已经使用了平均t检验统计量作为一个简单的案例来说明基因集在零假设下的统计摘要。我们证明了，当基因集中的基因并不独立时，需要一个校正因子。这里要注意的是，我们使用平均t检验统计量只是做了一个简单的说明，如果需要更加严格处理方法来实现均值偏移方法(testing-for-mean-shifts)，则需要参考 <a href="http://www.ncbi.nlm.nih.gov/pubmed/15647293" target="_blank" rel="external">SAFE</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=20610611" target="_blank" rel="external">ROAST</a>和 <a href="http://www.ncbi.nlm.nih.gov/pubmed/22638577" target="_blank" rel="external">CAMERA</a>。</p>
<p>即使在使用了膨胀因子(inflation factor)校正了平均t检验统计量后，零分布也有可能不成立。例如，对于平均t检验统计量是正态的假设而言，原始的样本大小和基因集大小有可能太小。在一次部分里，我们会演示如何在不使用参数假设的情况下，使用置换检验(Permutations)来估计零分布。</p>
<h2 id="置换检验permutations">置换检验(Permutations)</h2>
<p>当我们感兴趣的结果允许以及样本足够的情况下，我们可以使用基因集特定的转换检验来替换参数检验。参考文献中列出了几种这样的方法[<a href="http://www.ncbi.nlm.nih.gov/pubmed/16199517" target="_blank" rel="external">1</a>], [<a href="http://www.ncbi.nlm.nih.gov/pubmed/16174746" target="_blank" rel="external">2</a>]。对于大多数的富集分布而言，我们可以使用置换来创建一个零分布。例如，在这个案例中，我们对简单的t检验摘要执行置换检验，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(matrixStats)</div><div class="line"><span class="comment">## matrixStats v0.50.1 (2015-12-14) successfully loaded. See ?matrixStats for help.</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Attaching package: 'matrixStats'</span></div><div class="line"><span class="comment">## The following object is masked from 'package:GenomicFiles':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     rowRanges</span></div><div class="line"><span class="comment">## The following object is masked from 'package:SummarizedExperiment':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     rowRanges</span></div><div class="line"><span class="comment">## The following objects are masked from 'package:genefilter':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     rowSds, rowVars</span></div><div class="line"><span class="comment">## The following objects are masked from 'package:Biobase':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     anyMissing, rowMedians</span></div><div class="line">set.seed(<span class="number">1</span>)</div><div class="line">B &lt;- <span class="number">400</span> <span class="comment">##takes a few minutes</span></div><div class="line">null &lt;- sapply(<span class="number">1</span>:B,<span class="keyword">function</span>(b)&#123;</div><div class="line"> nullX&lt;- sample(X)</div><div class="line"> nullsvaX&lt;-model.matrix(~nullX+svafit$sv) <span class="comment">##note that we are not recomupting the surrogate values. </span></div><div class="line"> nulllmfit &lt;- lmFit(geneExpression,nullsvaX)</div><div class="line"> nulltt&lt;-nulllmfit$coef[,<span class="number">2</span>]*sqrt(nulllmfit$df.residual)/(<span class="number">2</span>*nulllmfit$sigma)</div><div class="line"> nullavgt &lt;- sapply(gsids,<span class="keyword">function</span>(i) sqrt(length(i))*mean(nulltt[i]))</div><div class="line"> <span class="keyword">return</span>(nullavgt)</div><div class="line">&#125;)</div><div class="line">permavgt &lt;- avgt/rowSds(null)</div><div class="line">permpval&lt;- rowMeans(abs(avgt) &lt; abs(null))</div></pre></td></tr></table></figure>
<p>在这个案例中，使用置换得到的结果非常类似于使用了校正因子得到的参数检验结果，我们从比较的结果中就能看出这些信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">plot(correctedavgt,permavgt,bg=cols,pch=<span class="number">21</span>,xlab=<span class="string">"Parametric z-score (with correction)"</span>,ylab=<span class="string">"Permutation z-score"</span>,cex=<span class="number">1.5</span>,ylim=c(-<span class="number">5</span>,<span class="number">15</span>),xlim=c(-<span class="number">5</span>,<span class="number">15</span>))</div><div class="line">abline(<span class="number">0</span>,<span class="number">1</span>)</div><div class="line">abline(h=<span class="number">0</span>,v=<span class="number">0</span>,lty=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_geneset_1-unnamed-chunk-18-1.png" alt="plot of chunk unnamed-chunk-18">
<p class="caption">plot of chunk unnamed-chunk-18</p>
</div>
<p>我们看到这两种方法得到的q值也是比较类似：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div></pre></td><td class="code"><pre><div class="line">tab &lt;- data.frame(avgt=p.adjust(signif(<span class="number">2</span>*pnorm(<span class="number">1</span>-abs(avgt)),<span class="number">2</span>),method=<span class="string">"BH"</span>),</div><div class="line">                  correctedavgt=p.adjust(signif(<span class="number">2</span>*pnorm(<span class="number">1</span>-abs(correctedavgt)),<span class="number">2</span>),method=<span class="string">"BH"</span>),</div><div class="line">                  permutations=p.adjust(permpval,method=<span class="string">"BH"</span>))</div><div class="line"><span class="comment">##include only gene sets with 10 or more genes in comparison</span></div><div class="line">tab&lt;-tab[N&gt;=<span class="number">10</span>,]</div><div class="line">tab &lt;- cbind(signif(tab,<span class="number">2</span>),apply(tab,<span class="number">2</span>,rank))</div><div class="line">tab&lt;-tab[order(tab[,<span class="number">1</span>]),]</div><div class="line">tab &lt;- tab[tab[,<span class="number">1</span>]&lt; <span class="number">0.25</span>,]</div><div class="line">tab</div><div class="line"><span class="comment">##             avgt correctedavgt permutations avgt correctedavgt</span></div><div class="line"><span class="comment">## chryp11  1.9e-60       1.5e-37         0.00  1.0             1</span></div><div class="line"><span class="comment">## chryq11  4.2e-44       6.3e-02         0.00  2.0             3</span></div><div class="line"><span class="comment">## chrxq13  2.4e-06       5.3e-03         0.00  3.0             2</span></div><div class="line"><span class="comment">## chr17p13 2.5e-02       1.0e+00         0.00  4.5           124</span></div><div class="line"><span class="comment">## chr6p22  2.5e-02       1.0e+00         0.77  4.5           124</span></div><div class="line"><span class="comment">## chrxp11  3.4e-02       1.0e+00         0.70  6.0           124</span></div><div class="line"><span class="comment">## chr20p13 5.6e-02       9.7e-02         0.00  7.0             4</span></div><div class="line"><span class="comment">## chr4q35  9.7e-02       1.0e+00         0.39  8.0           124</span></div><div class="line"><span class="comment">##          permutations</span></div><div class="line"><span class="comment">## chryp11             4</span></div><div class="line"><span class="comment">## chryq11             4</span></div><div class="line"><span class="comment">## chrxq13             4</span></div><div class="line"><span class="comment">## chr17p13            4</span></div><div class="line"><span class="comment">## chr6p22           114</span></div><div class="line"><span class="comment">## chrxp11            78</span></div><div class="line"><span class="comment">## chr20p13            4</span></div><div class="line"><span class="comment">## chr4q35            18</span></div></pre></td></tr></table></figure>
<p>不过置换检验方法的一个局限就是，为了能够得到非常小的p值，我们需要进行多次置换。例如为了得到一个在10e-5到10e-6之间的一个p值，我们霜肆进行100万次置换。对于那些使用了保守的多重比较校正方法（例如Bonferonni校正）的人来说，很有必要进行更多次地的模拟。置换检验的另外一个局限就是，在更复杂的设计中，例如具有许多轭和连续协变量的实验中，如何对置换进行解读并不容易。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_geneset_1.html" target="_blank" rel="external">Gene set analysis</a></li>
<li>Virtaneva K, “Expression profiling reveals fundamental biological differences in acute myeloid leukemia with isolated trisomy 8 and normal cytogenetics.” PNAS 2001 http://www.ncbi.nlm.nih.gov/pubmed/?term=11158605</li>
<li>Jelle J. Goeman and Peter Bühlmann, “Analyzing gene expression data in terms of gene sets: methodological issues” Bioinformatics 2006. http://bioinformatics.oxfordjournals.org/content/23/8/980</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/12/Genomics Data Analysis Series/GDAS012- 折衷t检验之limma分析微阵列数据/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/12/Genomics Data Analysis Series/GDAS012- 折衷t检验之limma分析微阵列数据/" itemprop="url">GDAS012-折衷t检验之limma分析微阵列数据</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-12T12:00:00+08:00">
                2019-09-12
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,729
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  8
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="简单t检验">简单t检验</h2>
<p>在这一节中，我们将展示一下使用简单t检验和使用limma层次模型进行差异分析的区别。limma的相关参考文献已经列到参考资料中。</p>
<p>在这一节中，我们还民法了执行limma分析的基本步骤。需要注意的是，limma的功能非常强大，它有数百页的文档，我们不能详细地介绍这个包以及函数，有兴趣的同学可以直接查看它的文档。</p>
<h3 id="spike-in数据集">spike-in数据集</h3>
<p>我们先加载一批由Affymetrix生成的均一化(normalized)后的spike-in数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># biocLite("SpikeInSubset")</span></div><div class="line"><span class="keyword">library</span>(SpikeInSubset)</div><div class="line">data(rma95)</div><div class="line">fac &lt;- factor(rep(<span class="number">1</span>:<span class="number">2</span>,each=<span class="number">3</span>))</div></pre></td></tr></table></figure>
<p>简单来说，这里一共有16个mRNA的物种，它们都已经使用了固定的浓度，并混合起来使用了hgu95芯片进行了检验。这个子集中对于每个mRNA物种，ExpressionSet中的第一个三元组(trio)和第二个三元组有着固定的不同值。通过<code>pData</code>就能看到，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">pData(rma95)</div><div class="line"><span class="comment">##                  37777_at 684_at 1597_at 38734_at 39058_at 36311_at</span></div><div class="line"><span class="comment">## 1521a99hpp_av06      0.00   0.25     0.5        1        2        4</span></div><div class="line"><span class="comment">## 1532a99hpp_av04      0.00   0.25     0.5        1        2        4</span></div><div class="line"><span class="comment">## 2353a99hpp_av08      0.00   0.25     0.5        1        2        4</span></div><div class="line"><span class="comment">## 1521b99hpp_av06      0.25   0.50     1.0        2        4        8</span></div><div class="line"><span class="comment">## 1532b99hpp_av04      0.25   0.50     1.0        2        4        8</span></div><div class="line"><span class="comment">## 2353b99hpp_av08r     0.25   0.50     1.0        2        4        8</span></div><div class="line"><span class="comment">##                  36889_at 1024_at 36202_at 36085_at 40322_at 407_at</span></div><div class="line"><span class="comment">## 1521a99hpp_av06         8      16       32       64      128   0.00</span></div><div class="line"><span class="comment">## 1532a99hpp_av04         8      16       32       64      128   0.00</span></div><div class="line"><span class="comment">## 2353a99hpp_av08         8      16       32       64      128   0.00</span></div><div class="line"><span class="comment">## 1521b99hpp_av06        16      32       64      128      256   0.25</span></div><div class="line"><span class="comment">## 1532b99hpp_av04        16      32       64      128      256   0.25</span></div><div class="line"><span class="comment">## 2353b99hpp_av08r       16      32       64      128      256   0.25</span></div><div class="line"><span class="comment">##                  1091_at 1708_at 33818_at 546_at</span></div><div class="line"><span class="comment">## 1521a99hpp_av06      512    1024      256     32</span></div><div class="line"><span class="comment">## 1532a99hpp_av04      512    1024      256     32</span></div><div class="line"><span class="comment">## 2353a99hpp_av08      512    1024      256     32</span></div><div class="line"><span class="comment">## 1521b99hpp_av06     1024       0      512     64</span></div><div class="line"><span class="comment">## 1532b99hpp_av04     1024       0      512     64</span></div><div class="line"><span class="comment">## 2353b99hpp_av08r    1024       0      512     64</span></div></pre></td></tr></table></figure>
<p>如果要了解这批数据这样设计的原理，现在我们来看一下下面的四张spiked mRNA，原文如下：</p>
<blockquote>
<p>To get a feel for the response of the array quantifications to this design, consider the following plots for four of the spiked mRNAs.</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(2,2))</div><div class="line">for (i in 1:4) &#123;</div><div class="line">  spg = names(pData(rma95))</div><div class="line">  plot(1:6, exprs(rma95)[spg[i+6],], main=spg[i+6], ylab=&quot;RMA&quot;,</div><div class="line">    xlab=&quot;nominal&quot;, axes=FALSE)</div><div class="line">  axis(2)</div><div class="line">  axis(1, at=1:6, labels=pData(rma95)[[spg[i+6]]])</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_limma-lkpl-1.png" alt="plot of chunk lkpl">
<p class="caption">plot of chunk lkpl</p>
</div>
<p>由于RMA(robust multi-array average,鲁棒多阵列平均 )是以log2转换为定量标准的，因此在归一化的检测值中观察到的差异是比较合理的。但是，在每个设计好的浓度内部，存在着相当大的变异。</p>
<h3 id="rowttests"><code>rowttests</code></h3>
<p>我们现在使用 <code>genefilter</code> 包中的 <code>rowttests</code> 函数来做一个简单的t检验，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(genefilter)</div><div class="line">rtt &lt;- rowttests(exprs(rma95),fac)</div></pre></td></tr></table></figure>
<p>我们将根据p值，均值，以及特征值是否是spike-in值来设置颜色，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">mask &lt;- with(rtt, abs(dm) &lt; .2 &amp; p.value &lt; .01)</div><div class="line">spike &lt;- rownames(rma95) %in% colnames(pData(rma95))</div><div class="line">cols &lt;- ifelse(mask,&quot;red&quot;,ifelse(spike,&quot;dodgerblue&quot;,&quot;black&quot;))</div></pre></td></tr></table></figure>
<p>我们现在使用上面定义的颜色来绘制结果。我们将<code>dm</code>乘以-1，因为国我们感兴趣的是，第二组与第一组的差异（这是由<code>lm</code>与<code>limma</code>包默认情况下使用的差异）。spike-in基因是蓝色的，它们大多数有着较小的p值与较大的均值差异。红点表示较小p值的基因，同时也有着较小的均值差异。我们将会看到这些点使用<code>limma</code>后的变化，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">with(rtt, plot(-dm, -log10(p.value), cex=.8, pch=16,</div><div class="line">     xlim=c(-1,1), ylim=c(0,5),</div><div class="line">     xlab=&quot;difference in means&quot;,</div><div class="line">     col=cols))</div><div class="line">abline(h=2,v=c(-.2,.2), lty=2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191109170623.png">

</div>
<p>Note that the red genes have mostly low estimates of standard deviation.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">rtt$s &lt;- apply(exprs(rma95), 1, function(row) sqrt(.5 * (var(row[1:3]) + var(row[4:6]))))</div><div class="line">with(rtt, plot(s, -log10(p.value), cex=.8, pch=16,</div><div class="line">              log=&quot;x&quot;,xlab=&quot;estimate of standard deviation&quot;,</div><div class="line">              col=cols))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_limma-unnamed-chunk-5-1.png" alt="plot of chunk unnamed-chunk-5">
<p class="caption">plot of chunk unnamed-chunk-5</p>
</div>
<h2 id="limma步骤">limma步骤</h2>
<p>以下是执行基本<code>limma</code>分析的三个步骤。我们指定<code>coef=2</code>，因为我感兴趣的是组与组之间的差异，而不是截距(intercept)，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">library(limma)</div><div class="line">options(digits=3)</div><div class="line">fit &lt;- lmFit(rma95, design=model.matrix(~ fac))  # step 1 least squares estimates</div><div class="line">colnames(coef(fit))     </div><div class="line">## [1] &quot;(Intercept)&quot; &quot;fac2&quot;</div><div class="line">fit &lt;- eBayes(fit)           # step 2 moderate the t statistics</div><div class="line">tt &lt;- topTable(fit, coef=2)  # step 3 report</div><div class="line">tt</div><div class="line">##           logFC AveExpr      t  P.Value adj.P.Val    B</div><div class="line">## 1708_at  -7.061    7.95 -73.53 7.82e-17  9.87e-13 8.65</div><div class="line">## 36202_at  0.853    9.37   9.98 4.94e-07  3.12e-03 4.59</div><div class="line">## 36311_at  0.832    8.56   8.36 3.02e-06  1.27e-02 3.57</div><div class="line">## 33264_at  0.712    4.92   7.43 9.67e-06  2.71e-02 2.84</div><div class="line">## 32660_at  0.655    8.68   7.36 1.07e-05  2.71e-02 2.77</div><div class="line">## 38734_at  0.747    6.26   7.19 1.35e-05  2.83e-02 2.62</div><div class="line">## 1024_at   0.843    9.70   6.73 2.50e-05  4.40e-02 2.20</div><div class="line">## 36085_at  0.645   12.19   6.65 2.79e-05  4.40e-02 2.12</div><div class="line">## 33818_at  0.532   12.29   6.45 3.70e-05  5.19e-02 1.92</div><div class="line">## 39058_at  0.609    7.53   6.28 4.77e-05  5.69e-02 1.74</div></pre></td></tr></table></figure>
<p><code>topTable</code> 会返回你定义的值排序的前几个基因。你还可以使用<code>topTable</code>通过指定<code>none</code>来要求返回所有的值。需要注意的是，有一列会给出校正后的每个基因的p值。默认情况下，通过<code>p.adjust</code>函数中的<code>Benjamini-Hochberg</code>方法用于校正p值，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"># ?topTable</div><div class="line">dim(topTable(fit, coef=2, number=Inf, sort.by=&quot;none&quot;))</div><div class="line">## [1] 12626     6</div><div class="line"># ?p.adjust</div></pre></td></tr></table></figure>
<p>在这里，我们比较一下使用<code>limma</code>分析的结果与以前得到的火山图结果。注意，红点现在都位于<code>-log10(p.value)=2</code>之下。此外，表示实际差异的蓝点的p值比以前更高，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">limmares &lt;- data.frame(dm=coef(fit)[,&quot;fac2&quot;], p.value=fit$p.value[,&quot;fac2&quot;])</div><div class="line">with(limmares, plot(dm, -log10(p.value),cex=.8, pch=16,</div><div class="line">     col=cols,xlab=&quot;difference in means&quot;,</div><div class="line">     xlim=c(-1,1), ylim=c(0,5)))</div><div class="line">abline(h=2,v=c(-.2,.2), lty=2)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191109172143.png">

</div>
<p>最后我们来构建一张图，用于展示<code>limma</code>是如何将方差估计缩小到一个共同值(common value)的，这能消除由于方差估计过低而导致的假阳性。</p>
<p>在这里，我们为不同的方差估计值设计40个小区间(bin)，每个区间中落入一个基因。我们会删除那些不含有任何基因的区间，原文如下：</p>
<blockquote>
<p>Here we pick, for each of 40 bins of different variance estimates, a single gene which falls in that bin. We remove bins which do not have any such genes.</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">n &lt;- 40</div><div class="line">qs &lt;- seq(from=0,to=.2,length=n)</div><div class="line">idx &lt;- sapply(seq_len(n),function(i) which(as.integer(cut(rtt$s^2,qs)) == i)[1])</div><div class="line">idx &lt;- idx[!is.na(idx)]</div></pre></td></tr></table></figure>
<p>现在绘制一条直线，这个直线展示了那些基因最初的估计方差到运行了<code>limma</code>之后的估计方差，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">par(mar=c(5,5,2,2))</div><div class="line">plot(1,1,xlim=c(0,.21),ylim=c(0,1),type=&quot;n&quot;,</div><div class="line">     xlab=&quot;variance estimates&quot;,ylab=&quot;&quot;,yaxt=&quot;n&quot;)</div><div class="line">axis(2,at=c(.1,.9),c(&quot;before&quot;,&quot;after&quot;),las=2)</div><div class="line">segments((rtt$s^2)[idx],rep(.1,n),</div><div class="line">         fit$s2.post[idx],rep(.9,n))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_limma-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>关于limma校正的一些层级模型资料，可以参考以前的笔记<a href="http://rvdsd.top/2019/08/19/DAL/DALS019_StatisticalModels2_Beyes_Hierarchical/">《DALS019-统计模型2-贝叶斯分布与层次分析》</a>。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_limma.html" target="_blank" rel="external">Using limma for microarray analysis</a></li>
<li>Smyth GK, “Linear models and empirical bayes methods for assessing differential expression in microarray experiments”. Stat Appl Genet Mol Biol. 2004 http://www.ncbi.nlm.nih.gov/pubmed/16646809</li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/11/Genomics Data Analysis Series/GDAS011-基因组级数据中的t检验/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/11/Genomics Data Analysis Series/GDAS011-基因组级数据中的t检验/" itemprop="url">GDAS011-基因组级数据中的t检验与多重比较</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-11T12:00:00+08:00">
                2019-09-11
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,000
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  4
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>在上一节中，我们集中研究了一对基因来说明两个方面的变异。其中一个基因似乎在不同的小鼠之间存在着很高的变异，但是这种变异隐藏在了合并数据背后。当我们使用合并的数据对不同的品系小鼠进行统计时，获得的p值在表面上看来是非常低的（p&lt;10e-6），但是当我们使用含有几个代表个体的数据进行比较时，我们发现p值又变得比较大（p=0.089）。在这一节中我们会认识到，找到最令人信服的问题就是要研究生物学变异，只有通过良好的实验设计才能直接检测到生物学变异。准确解释生物学变异的来源(origin)和大小(size)需要合适的统计分析。</p>
<p>在这一节上，我们将会介绍基因组实验中的统计推断，在此，我们需要了解以下几个方面的注意事项：</p>
<ul>
<li>如果需要进行多次检验，那么对于数以千计的特征值（通常是一个基因）来说，每个特征值至少要经历一次检验；</li>
<li>每个特征值都会表现出自身的技术变异与生物学变异；</li>
<li>有可能存在未检测到或未报道的生物学变异（例如以天为单位时间时，不同时间点的检测）（注：原文是there may be unmeasured or unreported sources of biological variation (such as time of day)）；</li>
<li>许多特征值有着内在的联系，这些检验并不是独立的。</li>
</ul>
<p>我们将会使用到前面章节中提到的一些概念，包括t检验和多重比较；在后面部分里，我们将会根据分层模型(hierarchical models)计算标准偏差估计。</p>
<p>我们先加载合并的数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line"><span class="keyword">library</span>(maPooling)</div><div class="line">data(maPooling)</div><div class="line">pd=pData(maPooling)</div><div class="line">individuals=which(rowSums(pd)==<span class="number">1</span>)</div></pre></td></tr></table></figure>
<p>现在我们提取单个小鼠及品系，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">individuals=which(rowSums(pd)==<span class="number">1</span>)</div><div class="line">individuals=individuals[-grep(<span class="string">"tr"</span>,names(individuals))]</div><div class="line">y=exprs(maPooling)[,individuals]</div><div class="line">g=factor(as.numeric(grepl(<span class="string">"b"</span>,names(individuals))))</div></pre></td></tr></table></figure>
<h2 id="t检验">T检验</h2>
<p>我们可以使用<code>genefilter</code>包中的<code>rowttest</code>函数来进行t检验，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(genefilter)</div><div class="line">tt=rowttests(y,g)</div></pre></td></tr></table></figure>
<p>现在我们来找一些哪些基因有统计学意义。由于某些历史原因，在科学研究中，我们判断是否有统计学意义的截止值是p值为0.01或0.05，现在我们找到那些p值小于0.01的基因数目，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">NsigAt01 = sum(tt$p.value&lt;0.01)</div><div class="line">NsigAt01</div><div class="line">## [1] 1578</div><div class="line">NsigAt05 = sum(tt$p.value&lt;0.05)</div><div class="line">NsigAt05</div><div class="line">## [1] 2908</div></pre></td></tr></table></figure>
<h2 id="多重检验">多重检验</h2>
<p>我们在前面的章节中提到了多重检验（参考笔记<a href="http://rvdsd.top/2019/08/17/DAL/DALS017_InferenceForHighDimensionalData2_Principle/">DALS017-高维数据推断2-统计原理</a>），现在我们来进行一个快速总结。</p>
<p>我们是否要报道前面提到的名义上的所有的具有显著性的基因？现在让我们来研究一下，如果将第一组分为两组，强制零假设为真，会发生什么情况。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">set.seed(<span class="number">0</span>)</div><div class="line">shuffledIndex &lt;- factor(sample(c(<span class="number">0</span>,<span class="number">1</span>),sum(g==<span class="number">0</span>),replace=<span class="literal">TRUE</span> ))</div><div class="line">nulltt &lt;- rowttests(y[,g==<span class="number">0</span>],shuffledIndex)</div><div class="line">NfalselySigAt01 = sum(nulltt$p.value&lt;<span class="number">0.01</span>)</div><div class="line">NfalselySigAt01 </div><div class="line"><span class="comment">## [1] 79</span></div><div class="line">NfalselySigAt05 = sum(nulltt$p.value&lt;<span class="number">0.05</span>)</div><div class="line">NfalselySigAt05</div><div class="line"><span class="comment">## [1] 840</span></div></pre></td></tr></table></figure>
<p>如果我们使用0.05作为截止值，则将出现840个假阳性。我们在前面部分中已经提到了几种校正方法，其中就包括<code>qvalue</code>包中的<code>qvalue</code>函数。经过这样的调整，我们就能降低假阳性，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(qvalue)</div><div class="line">qvals = qvalue(tt$p.value)$qvalue</div><div class="line">sum(qvals&lt;<span class="number">0.05</span>)</div><div class="line"><span class="comment">## [1] 1179</span></div><div class="line">sum(qvals&lt;<span class="number">0.01</span>)</div><div class="line"><span class="comment">## [1] 538</span></div></pre></td></tr></table></figure>
<p>并且在零亿是高的情况下，也不会产生误报，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(qvalue)</div><div class="line">nullqvals = qvalue(nulltt$p.value)$qvalue</div><div class="line">sum(nullqvals&lt;<span class="number">0.05</span>)</div><div class="line"><span class="comment">## [1] 0</span></div><div class="line">sum(nullqvals&lt;<span class="number">0.01</span>)</div><div class="line"><span class="comment">## [1] 0</span></div></pre></td></tr></table></figure>
<p>这以相当常规的方式解决了在固定名义显著性水平上执行多次假设检验时造成的假阳性问题。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_t_mult.html" target="_blank" rel="external">Inference: t-tests, multiple comparisons</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/10/BioTools/fgsea/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/10/BioTools/fgsea/" itemprop="url">fgsea包做GSEA分析</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-10T12:00:00+08:00">
                2019-09-10
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生信工具/" itemprop="url" rel="index">
                    <span itemprop="name">生信工具</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,259
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="gsea图">GSEA图</h2>
<p>这篇教程翻译自生信博客DAVE TANGE’S BLOG(https://davetang.org/muse/)上的一篇教程，参考资料中已经列出了地址。</p>
<p><code>fgsea</code>这个包用于做GSEA分析，先来看一下使用这个包做的图，如下所示：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911114456.png">

</div>
<p>现在简单解释一下这个图形：</p>
<p>x轴——排序后的基因列表<code>L</code>位置对应的坐标，也就是我们自己通过RNA-seq，芯片，qPCR等手段获得的基因表达值倍数变化，或p值排序，总之，这是一个有序列表。</p>
<p>垂直的黑色细胞——上图中类似条形码的图形，这是指的是某一个基因集<code>S</code>中基因对应于<code>L</code>基因中，的位置，在上图中，这个基因集是细胞周期(Cell Cycle)，明天看到<code>S</code>中的成员在<code>L</code>的左侧比较密集。</p>
<p>y轴——富集分布，从上面我们可以看到，细胞周期(Cell Cycle)这个基因集在左侧富集，也就是绿色曲线表示的位置。</p>
<h2 id="fgsea使用">fgsea使用</h2>
<h3 id="安装">安装</h3>
<p>先安装<code>fgsea</code>包，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">install.packages(<span class="string">"BiocManager"</span>)</div><div class="line">BiocManager::install(<span class="string">"fgsea"</span>)</div><div class="line"><span class="keyword">library</span>(fgsea)</div></pre></td></tr></table></figure>
<h3 id="数据">数据</h3>
<p><code>fgsea</code>包中内置的有数据集<code>examplePathways</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(fgsea)</div><div class="line"></div><div class="line">data(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># examplePathways是从'reactome.db'包中</span></div><div class="line"><span class="comment"># 提取的信息，并以列表的形式存了这些信息，</span></div><div class="line"><span class="comment"># 这些信息主要是小鼠的基因</span></div><div class="line">help(<span class="string">"examplePathways"</span>)</div><div class="line"></div><div class="line"><span class="comment"># 这个数据集是一个列表</span></div><div class="line">class(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># 一共有1,457 "通路"</span></div><div class="line">length(examplePathways)</div><div class="line"></div><div class="line"><span class="comment"># 第1列含有Meiotic_Synapsis pathway的EntrezID</span></div><div class="line">examplePathways[<span class="number">1</span>]</div><div class="line"></div><div class="line"><span class="comment"># exampleRanks储存的是排序信息</span></div><div class="line">data(exampleRanks)</div><div class="line"></div><div class="line"><span class="comment"># exampleRnak是一个数字型的向量</span></div><div class="line">class(exampleRanks)</div><div class="line"></div><div class="line"><span class="comment"># exampleRanks中的向量名称则是对称的Entrez ID</span></div><div class="line">head(exampleRanks)</div><div class="line">tail(unname(exampleRanks))</div></pre></td></tr></table></figure>
<p>运行结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(fgsea)</div><div class="line">&gt; </div><div class="line">&gt; data(examplePathways)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># examplePathways是从'reactome.db'包中</span></div><div class="line"><span class="comment"># 提取的信息，并以列表的形式</span></div><div class="line">&gt; <span class="comment"># 储存了这些信息，这些信息主要是小鼠的基因</span></div><div class="line">&gt; help(<span class="string">"examplePathways"</span>)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 这个数据集是一个列表</span></div><div class="line">&gt; class(examplePathways)</div><div class="line">[<span class="number">1</span>] <span class="string">"list"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 一共有1,457 "通路"</span></div><div class="line">&gt; length(examplePathways)</div><div class="line">[<span class="number">1</span>] <span class="number">1457</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 第1列含有Meiotic_Synapsis pathway的EntrezID</span></div><div class="line">&gt; examplePathways[<span class="number">1</span>]</div><div class="line">$`1221633_Meiotic_Synapsis`</div><div class="line"> [<span class="number">1</span>] <span class="string">"12189"</span>     <span class="string">"13006"</span>     <span class="string">"15077"</span>     <span class="string">"15078"</span>     <span class="string">"15270"</span>     <span class="string">"15512"</span>    </div><div class="line"> [<span class="number">7</span>] <span class="string">"16905"</span>     <span class="string">"16906"</span>     <span class="string">"19357"</span>     <span class="string">"20842"</span>     <span class="string">"20843"</span>     <span class="string">"20957"</span>    </div><div class="line">[<span class="number">13</span>] <span class="string">"20962"</span>     <span class="string">"21749"</span>     <span class="string">"21750"</span>     <span class="string">"22196"</span>     <span class="string">"23856"</span>     <span class="string">"24061"</span>    </div><div class="line">[<span class="number">19</span>] <span class="string">"28113"</span>     <span class="string">"50878"</span>     <span class="string">"56739"</span>     <span class="string">"57321"</span>     <span class="string">"64009"</span>     <span class="string">"66654"</span>    </div><div class="line">[<span class="number">25</span>] <span class="string">"69386"</span>     <span class="string">"71846"</span>     <span class="string">"74075"</span>     <span class="string">"77053"</span>     <span class="string">"94244"</span>     <span class="string">"97114"</span>    </div><div class="line">[<span class="number">31</span>] <span class="string">"97122"</span>     <span class="string">"97908"</span>     <span class="string">"101185"</span>    <span class="string">"140557"</span>    <span class="string">"223697"</span>    <span class="string">"260423"</span>   </div><div class="line">[<span class="number">37</span>] <span class="string">"319148"</span>    <span class="string">"319149"</span>    <span class="string">"319150"</span>    <span class="string">"319151"</span>    <span class="string">"319152"</span>    <span class="string">"319153"</span>   </div><div class="line">[<span class="number">43</span>] <span class="string">"319154"</span>    <span class="string">"319155"</span>    <span class="string">"319156"</span>    <span class="string">"319157"</span>    <span class="string">"319158"</span>    <span class="string">"319159"</span>   </div><div class="line">[<span class="number">49</span>] <span class="string">"319160"</span>    <span class="string">"319161"</span>    <span class="string">"319565"</span>    <span class="string">"320332"</span>    <span class="string">"320558"</span>    <span class="string">"326619"</span>   </div><div class="line">[<span class="number">55</span>] <span class="string">"326620"</span>    <span class="string">"360198"</span>    <span class="string">"497652"</span>    <span class="string">"544973"</span>    <span class="string">"625328"</span>    <span class="string">"667250"</span>   </div><div class="line">[<span class="number">61</span>] <span class="string">"100041230"</span> <span class="string">"102641229"</span> <span class="string">"102641751"</span> <span class="string">"102642045"</span></div><div class="line"></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRanks储存的是排序信息</span></div><div class="line">&gt; data(exampleRanks)</div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRnak是一个数字型的向量</span></div><div class="line">&gt; class(exampleRanks)</div><div class="line">[<span class="number">1</span>] <span class="string">"numeric"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># exampleRanks中的向量名称则是对称的Entrez ID</span></div><div class="line">&gt; head(exampleRanks)</div><div class="line">   <span class="number">170942</span>    <span class="number">109711</span>     <span class="number">18124</span>     <span class="number">12775</span>     <span class="number">72148</span>     <span class="number">16010</span> </div><div class="line">-<span class="number">63.33703</span> -<span class="number">49.74779</span> -<span class="number">43.63878</span> -<span class="number">41.51889</span> -<span class="number">33.26039</span> -<span class="number">32.77626</span> </div><div class="line">&gt; tail(unname(exampleRanks))</div><div class="line">[<span class="number">1</span>] <span class="number">47.58235</span> <span class="number">49.87543</span> <span class="number">50.25179</span> <span class="number">50.86532</span> <span class="number">51.16110</span> <span class="number">53.28400</span></div></pre></td></tr></table></figure>
<h3 id="分析">分析</h3>
<p>现在我们使用上面的数据进行GSEA分析，进行GSEA分析时，我们的通路文件（也就是GSEA官网中的GMT文件，对应的就是基因集<code>S</code>）储存在<code>pathways</code>参数中，用户自己的数据（排过序的数据）放在<code>stats</code>数据集中，剩下的参数不用管，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">fgseaRes &lt;- fgsea(pathways = examplePathways, </div><div class="line">                  stats = exampleRanks,</div><div class="line">                  minSize=<span class="number">15</span>,</div><div class="line">                  maxSize=<span class="number">500</span>,</div><div class="line">                  nperm=<span class="number">100000</span>)</div></pre></td></tr></table></figure>
<p>分析的结果fgseaRes是一个<code>data.table</code>格式的文件，使用<code>plotEnrichment</code>函数可以绘制出GSEA分析的结果，如下所示： <figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">library(ggplot2)</div><div class="line"># 参考 https://github.com/Rdatatable/data.table/wiki</div><div class="line"># 来查看data.table相关的信息</div><div class="line">class(fgseaRes)</div><div class="line"></div><div class="line"># 查看p值小于0.01的通路的数目</div><div class="line">sum(fgseaRes[, padj &lt; 0.01])</div><div class="line"></div><div class="line"># 绘制出最显著的富集通路</div><div class="line">plotEnrichment(examplePathways[[head(fgseaRes[order(pval), ], 1)$pathway]],</div><div class="line">               exampleRanks) + </div><div class="line">               labs(title=head(fgseaRes[order(pval), ], 1)$pathway)</div></pre></td></tr></table></figure></p>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">&gt; class(fgseaRes)</div><div class="line">[<span class="number">1</span>] <span class="string">"data.table"</span> <span class="string">"data.frame"</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># 查看p值小于0.01的通路的数目</span></div><div class="line">&gt; sum(fgseaRes[, padj &lt; <span class="number">0.01</span>])</div><div class="line">[<span class="number">1</span>] <span class="number">78</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911122517.jpeg">

</div>
<p>可以画出具体的某条通路，如下所示；</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">plotEnrichment(examplePathways[[<span class="string">"5991130_Programmed_Cell_Death"</span>]],</div><div class="line">               exampleRanks) + </div><div class="line">  labs(title=<span class="string">"Programmed Cell Death"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911125331.jpeg">

</div>
<p>还可以在一张图中绘制出前10个富集通路，以及后10个富集通路，一共20个，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">topPathwaysUp &lt;- fgseaRes[ES &gt; 0][head(order(pval), n=10), pathway]</div><div class="line">topPathwaysDown &lt;- fgseaRes[ES &lt; 0][head(order(pval), n=10), pathway]</div><div class="line">topPathways &lt;- c(topPathwaysUp, rev(topPathwaysDown))</div><div class="line">plotGseaTable(examplePathways[topPathways], exampleRanks, fgseaRes, </div><div class="line">              gseaParam = 0.5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911122626.jpeg">

</div>
<h3 id="reactome">Reactome</h3>
<p>也可以使用<a href="https://reactome.org/" target="_blank" rel="external">Reactome</a>通路来进行GSEA分析，此时需要安装<code>reactome.db</code>包，这个包很大，600多M，下载的时间很长，安装过程如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># BiocManager::install("reactome.db")</span></div><div class="line"><span class="keyword">library</span>(reactome.db)</div><div class="line"></div><div class="line">my_pathways &lt;- reactomePathways(names(exampleRanks))</div><div class="line"> </div><div class="line"><span class="comment"># Reactome pathways have a median of 11 genes</span></div><div class="line">summary(sapply(my_pathways, length))</div><div class="line">   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </div><div class="line">    <span class="number">1.0</span>     <span class="number">4.0</span>    <span class="number">11.0</span>    <span class="number">30.3</span>    <span class="number">30.0</span>  <span class="number">1140.0</span></div><div class="line"> </div><div class="line">fgsea_reactome &lt;- fgsea(pathways = my_pathways, </div><div class="line">                        stats = exampleRanks,</div><div class="line">                        minSize=<span class="number">15</span>,</div><div class="line">                        maxSize=<span class="number">500</span>,</div><div class="line">                        nperm=<span class="number">100000</span>)</div><div class="line"> </div><div class="line">head(fgsea_reactome[order(pval), ])</div><div class="line">sum(fgsea_reactome[, padj &lt; <span class="number">0.01</span>])</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">&gt; head(fgsea_reactome[order(pval), ])</div><div class="line">                    pathway         pval         padj</div><div class="line"><span class="number">1</span>:               Cell Cycle <span class="number">1.212856e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">2</span>:      Cell Cycle, Mitotic <span class="number">1.236552e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">3</span>: Neutrophil degranulation <span class="number">1.274974e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">4</span>: Signaling by Rho GTPases <span class="number">1.291339e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">5</span>:                  M Phase <span class="number">1.303509e-05</span> <span class="number">0.0002571089</span></div><div class="line"><span class="number">6</span>:   Cell Cycle Checkpoints <span class="number">1.350749e-05</span> <span class="number">0.0002571089</span></div><div class="line">          ES      NES nMoreExtreme size</div><div class="line"><span class="number">1</span>: <span class="number">0.5324037</span> <span class="number">2.674889</span>            <span class="number">0</span>  <span class="number">414</span></div><div class="line"><span class="number">2</span>: <span class="number">0.5475346</span> <span class="number">2.719737</span>            <span class="number">0</span>  <span class="number">363</span></div><div class="line"><span class="number">3</span>: <span class="number">0.4258088</span> <span class="number">2.074081</span>            <span class="number">0</span>  <span class="number">296</span></div><div class="line"><span class="number">4</span>: <span class="number">0.4073481</span> <span class="number">1.966039</span>            <span class="number">0</span>  <span class="number">271</span></div><div class="line"><span class="number">5</span>: <span class="number">0.5022189</span> <span class="number">2.408762</span>            <span class="number">0</span>  <span class="number">255</span></div><div class="line"><span class="number">6</span>: <span class="number">0.6085986</span> <span class="number">2.836881</span>            <span class="number">0</span>  <span class="number">200</span></div><div class="line">                                  leadingEdge</div><div class="line"><span class="number">1</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line"><span class="number">2</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line"><span class="number">3</span>:    <span class="number">11676</span>,<span class="number">14190</span>,<span class="number">53381</span>,<span class="number">12306</span>,<span class="number">20430</span>,<span class="number">12505</span>,<span class="keyword">...</span></div><div class="line"><span class="number">4</span>: <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">20430</span>,<span class="number">104215</span>,<span class="number">233406</span>,<span class="number">107995</span>,<span class="keyword">...</span></div><div class="line"><span class="number">5</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="number">52276</span>,<span class="keyword">...</span></div><div class="line"><span class="number">6</span>:   <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="number">12428</span>,<span class="keyword">...</span></div><div class="line">&gt; sum(fgsea_reactome[, padj &lt; <span class="number">0.01</span>])</div><div class="line">[<span class="number">1</span>] <span class="number">103</span></div></pre></td></tr></table></figure>
<h3 id="leading-edge">Leading edge</h3>
<p>在GSEA分析中，我们通常会提取那些构成高得分的核心基因。我们对高得分的核心基因的定义就是，基因集<code>S</code>中位于排序基因列表<code>L</code>位置中的得分最大处之前或之后的那些基因集（也就是GSEA结果中绿色曲线最高点的前面或后面）。</p>
<p>前面我们注意到，GSEA分析的结果中有1列被命名为<code>leadingEdge</code>。这一列包含了高得分的基因。我们使用Reactome通路的富集结果来提取这些基因，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># the most significant pathway</span></div><div class="line">fgsea_reactome[order(pval),][<span class="number">1</span>,]</div><div class="line"></div><div class="line"><span class="comment"># list of Entrez gene IDs that contributed to the enrichment score</span></div><div class="line">fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge</div><div class="line"></div><div class="line"><span class="comment"># how many genes are in the leading edge?</span></div><div class="line">length(fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge[[<span class="number">1</span>]])</div><div class="line"></div><div class="line"><span class="comment"># how many genes are in the Cell Cycle pathway?</span></div><div class="line">length(my_pathways[[<span class="string">'Cell Cycle'</span>]])</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="comment"># the most significant pathway</span></div><div class="line">&gt; fgsea_reactome[order(pval),][<span class="number">1</span>,]</div><div class="line">      pathway         pval         padj        ES</div><div class="line"><span class="number">1</span>: Cell Cycle <span class="number">1.212856e-05</span> <span class="number">0.0002571089</span> <span class="number">0.5324037</span></div><div class="line">        NES nMoreExtreme size</div><div class="line"><span class="number">1</span>: <span class="number">2.674889</span>            <span class="number">0</span>  <span class="number">414</span></div><div class="line">                                leadingEdge</div><div class="line"><span class="number">1</span>: <span class="number">66336</span>,<span class="number">66977</span>,<span class="number">15366</span>,<span class="number">12442</span>,<span class="number">107995</span>,<span class="number">66442</span>,<span class="keyword">...</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># list of Entrez gene IDs that contributed to the enrichment score</span></div><div class="line">&gt; fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge</div><div class="line">[[<span class="number">1</span>]]</div><div class="line">  [<span class="number">1</span>] <span class="string">"66336"</span>  <span class="string">"66977"</span>  <span class="string">"15366"</span>  <span class="string">"12442"</span>  <span class="string">"107995"</span></div><div class="line">  [<span class="number">6</span>] <span class="string">"66442"</span>  <span class="string">"12571"</span>  <span class="string">"12428"</span>  <span class="string">"52276"</span>  <span class="string">"54392"</span> </div><div class="line"> [<span class="number">11</span>] <span class="string">"66311"</span>  <span class="string">"215387"</span> <span class="string">"67629"</span>  <span class="string">"12649"</span>  <span class="string">"72415"</span> </div><div class="line"> [<span class="number">16</span>] <span class="string">"56150"</span>  <span class="string">"57441"</span>  <span class="string">"20877"</span>  <span class="string">"67121"</span>  <span class="string">"12615"</span> </div><div class="line"> [<span class="number">21</span>] <span class="string">"11799"</span>  <span class="string">"66468"</span>  <span class="string">"67849"</span>  <span class="string">"19053"</span>  <span class="string">"73804"</span> </div><div class="line"> [<span class="number">26</span>] <span class="string">"76044"</span>  <span class="string">"20878"</span>  <span class="string">"15270"</span>  <span class="string">"13555"</span>  <span class="string">"60411"</span> </div><div class="line"> [<span class="number">31</span>] <span class="string">"12580"</span>  <span class="string">"17219"</span>  <span class="string">"69270"</span>  <span class="string">"12575"</span>  <span class="string">"69263"</span> </div><div class="line"> [<span class="number">36</span>] <span class="string">"12448"</span>  <span class="string">"14211"</span>  <span class="string">"20873"</span>  <span class="string">"18005"</span>  <span class="string">"72119"</span> </div><div class="line"> [<span class="number">41</span>] <span class="string">"71988"</span>  <span class="string">"12189"</span>  <span class="string">"17215"</span>  <span class="string">"12534"</span>  <span class="string">"66156"</span> </div><div class="line"> [<span class="number">46</span>] <span class="string">"208628"</span> <span class="string">"237911"</span> <span class="string">"22390"</span>  <span class="string">"68240"</span>  <span class="string">"228421"</span></div><div class="line"> [<span class="number">51</span>] <span class="string">"68014"</span>  <span class="string">"269582"</span> <span class="string">"19348"</span>  <span class="string">"12236"</span>  <span class="string">"72151"</span> </div><div class="line"> [<span class="number">56</span>] <span class="string">"18817"</span>  <span class="string">"21781"</span>  <span class="string">"18968"</span>  <span class="string">"217653"</span> <span class="string">"66934"</span> </div><div class="line"> [<span class="number">61</span>] <span class="string">"272551"</span> <span class="string">"227613"</span> <span class="string">"67141"</span>  <span class="string">"67951"</span>  <span class="string">"68612"</span> </div><div class="line"> [<span class="number">66</span>] <span class="string">"68298"</span>  <span class="string">"108000"</span> <span class="string">"23834"</span>  <span class="string">"106344"</span> <span class="string">"56452"</span> </div><div class="line"> [<span class="number">71</span>] <span class="string">"242705"</span> <span class="string">"18141"</span>  <span class="string">"223921"</span> <span class="string">"26886"</span>  <span class="string">"13557"</span> </div><div class="line"> [<span class="number">76</span>] <span class="string">"26909"</span>  <span class="string">"72787"</span>  <span class="string">"268697"</span> <span class="string">"72155"</span>  <span class="string">"56371"</span> </div><div class="line"> [<span class="number">81</span>] <span class="string">"17535"</span>  <span class="string">"107823"</span> <span class="string">"12531"</span>  <span class="string">"327762"</span> <span class="string">"12567"</span> </div><div class="line"> [<span class="number">86</span>] <span class="string">"229841"</span> <span class="string">"67052"</span>  <span class="string">"16319"</span>  <span class="string">"66634"</span>  <span class="string">"171567"</span></div><div class="line"> [<span class="number">91</span>] <span class="string">"26931"</span>  <span class="string">"67203"</span>  <span class="string">"12235"</span>  <span class="string">"19891"</span>  <span class="string">"74470"</span> </div><div class="line"> [<span class="number">96</span>] <span class="string">"72083"</span>  <span class="string">"381318"</span> <span class="string">"66570"</span>  <span class="string">"17216"</span>  <span class="string">"76308"</span> </div><div class="line">[<span class="number">101</span>] <span class="string">"19687"</span>  <span class="string">"17218"</span>  <span class="string">"102920"</span> <span class="string">"29870"</span>  <span class="string">"18973"</span> </div><div class="line">[<span class="number">106</span>] <span class="string">"16881"</span>  <span class="string">"17463"</span>  <span class="string">"75786"</span>  <span class="string">"19645"</span>  <span class="string">"19075"</span> </div><div class="line">[<span class="number">111</span>] <span class="string">"26417"</span>  <span class="string">"69736"</span>  <span class="string">"19357"</span>  <span class="string">"76816"</span>  <span class="string">"70385"</span> </div><div class="line">[<span class="number">116</span>] <span class="string">"70645"</span>  <span class="string">"22628"</span>  <span class="string">"225182"</span> <span class="string">"22627"</span>  <span class="string">"52683"</span> </div><div class="line">[<span class="number">121</span>] <span class="string">"19076"</span>  <span class="string">"18972"</span>  <span class="string">"231863"</span> <span class="string">"26932"</span>  <span class="string">"12544"</span> </div><div class="line">[<span class="number">126</span>] <span class="string">"17997"</span>  <span class="string">"51788"</span>  <span class="string">"26440"</span>  <span class="string">"68549"</span>  <span class="string">"12445"</span> </div><div class="line">[<span class="number">131</span>] <span class="string">"19088"</span>  <span class="string">"269113"</span> <span class="string">"26444"</span>  <span class="string">"19324"</span>  <span class="string">"103733"</span></div><div class="line">[<span class="number">136</span>] <span class="string">"59001"</span>  <span class="string">"107976"</span> <span class="string">"19179"</span>  <span class="string">"12579"</span>  <span class="string">"232987"</span></div><div class="line">[<span class="number">141</span>] <span class="string">"17420"</span>  <span class="string">"228769"</span> <span class="string">"219072"</span> <span class="string">"26445"</span>  <span class="string">"105988"</span></div><div class="line">[<span class="number">146</span>] <span class="string">"69745"</span>  <span class="string">"18538"</span>  <span class="string">"69928"</span>  <span class="string">"11651"</span>  <span class="string">"235559"</span></div><div class="line">[<span class="number">151</span>] <span class="string">"68097"</span>  <span class="string">"57296"</span>  <span class="string">"63955"</span>  <span class="string">"14235"</span>  <span class="string">"19170"</span> </div><div class="line">[<span class="number">156</span>] <span class="string">"17246"</span>  <span class="string">"17220"</span>  <span class="string">"12144"</span>  <span class="string">"50793"</span>  <span class="string">"77605"</span> </div><div class="line">[<span class="number">161</span>] <span class="string">"18392"</span>  <span class="string">"236930"</span> <span class="string">"67151"</span>  <span class="string">"70024"</span>  <span class="string">"59126"</span> </div><div class="line">[<span class="number">166</span>] <span class="string">"66296"</span>  <span class="string">"16906"</span>  <span class="string">"109145"</span> <span class="string">"71819"</span>  <span class="string">"67733"</span> </div><div class="line">[<span class="number">171</span>] <span class="string">"50883"</span>  <span class="string">"12447"</span>  <span class="string">"12532"</span>  <span class="string">"14156"</span>  <span class="string">"26442"</span> </div><div class="line">[<span class="number">176</span>] <span class="string">"19177"</span>  <span class="string">"230376"</span> <span class="string">"245688"</span></div><div class="line"></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># how many genes are in the leading edge?</span></div><div class="line">&gt; length(fgsea_reactome[order(pval),][<span class="number">1</span>,]$leadingEdge[[<span class="number">1</span>]])</div><div class="line">[<span class="number">1</span>] <span class="number">178</span></div><div class="line">&gt; </div><div class="line">&gt; <span class="comment"># how many genes are in the Cell Cycle pathway?</span></div><div class="line">&gt; length(my_pathways[[<span class="string">'Cell Cycle'</span>]])</div><div class="line">[<span class="number">1</span>] <span class="number">414</span></div></pre></td></tr></table></figure>
<h2 id="总结">总结</h2>
<p>GSEA是于2005年首次提出来的，现在已经成了基因表达分析中的常规分析手段，它不同于GO分析，GO分析只关注差异基因，而GSEA分析则关注所有的基因。<code>fgsea</code>包可以使用预先排列好的基因一R中进行GSEA分析。p值的计算结果是基于置换检验(permutation test)，这种方法并不是十分精我，因为它忽略了基因之间的相关性，有可能会导致假阳性。但是，在这种方法在研究上调与上调基因方面还是很有用的。即使你计算出的GSEA结果中，p值大于0.05，但是是也可以参考leading edge基因集，为你的实验进行指导。</p>
<h2 id="案例分析">案例分析</h2>
<p>作者提供了用于生成类似于<strong>exampleRanks</strong>文件的R脚本，不过使用的GEO的数据，平时自己利用<code>fgsea</code>包进行GSEA分析时，生成就好，现在看一下如何将GEO的数据生成类似于****exampleRanks****文件的排序信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">source</span>(<span class="string">"https://raw.githubusercontent.com/assaron/r-utils/master/R/exprs.R"</span>)</div><div class="line"><span class="keyword">library</span>(GEOquery)</div><div class="line"><span class="keyword">library</span>(limma)</div><div class="line">gse14308 &lt;- getGEO(<span class="string">"GSE14308"</span>)[[<span class="number">1</span>]]</div><div class="line">pData(gse14308)$condition &lt;- sub(<span class="string">"-.*$"</span>, <span class="string">""</span>, gse14308$title)</div><div class="line">es &lt;- collapseBy(gse14308, fData(gse14308)$ENTREZ_GENE_ID, FUN=median)</div><div class="line">es &lt;- es[!grepl(<span class="string">"///"</span>, rownames(es)), ]</div><div class="line">es &lt;- es[rownames(es) != <span class="string">""</span>, ]</div><div class="line">exprs(es) &lt;- normalizeBetweenArrays(log2(exprs(es)+<span class="number">1</span>), method=<span class="string">"quantile"</span>)</div><div class="line">es.design &lt;- model.matrix(~<span class="number">0</span>+condition, data=pData(es))</div><div class="line">fit &lt;- lmFit(es, es.design)</div><div class="line">fit2 &lt;- contrasts.fit(fit, makeContrasts(conditionTh1-conditionNaive,</div><div class="line">                                         levels=es.design))</div><div class="line">fit2 &lt;- eBayes(fit2)</div><div class="line">de &lt;- data.table(topTable(fit2, adjust.method=<span class="string">"BH"</span>, number=<span class="number">12000</span>, sort.by = <span class="string">"B"</span>), keep.rownames = <span class="literal">T</span>)</div><div class="line">ranks &lt;- de[order(t), list(rn, t)]</div><div class="line">head(ranks)</div><div class="line">tail(ranks)</div></pre></td></tr></table></figure>
<p>结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">&gt; head(ranks)</div><div class="line">       rn         t</div><div class="line"><span class="number">1</span>: <span class="number">109711</span> -<span class="number">50.99316</span></div><div class="line"><span class="number">2</span>:  <span class="number">18124</span> -<span class="number">44.40559</span></div><div class="line"><span class="number">3</span>:  <span class="number">12775</span> -<span class="number">43.22109</span></div><div class="line"><span class="number">4</span>:  <span class="number">72148</span> -<span class="number">33.74441</span></div><div class="line"><span class="number">5</span>:  <span class="number">16010</span> -<span class="number">33.34034</span></div><div class="line"><span class="number">6</span>:  <span class="number">16206</span> -<span class="number">30.67962</span></div><div class="line">&gt; tail(ranks)</div><div class="line">      rn        t</div><div class="line"><span class="number">1</span>: <span class="number">80901</span> <span class="number">47.55618</span></div><div class="line"><span class="number">2</span>: <span class="number">58801</span> <span class="number">48.53756</span></div><div class="line"><span class="number">3</span>: <span class="number">15937</span> <span class="number">49.88068</span></div><div class="line"><span class="number">4</span>: <span class="number">13730</span> <span class="number">50.17693</span></div><div class="line"><span class="number">5</span>: <span class="number">12772</span> <span class="number">50.32302</span></div><div class="line"><span class="number">6</span>: <span class="number">80876</span> <span class="number">51.72987</span></div></pre></td></tr></table></figure>
<p>现在绘制一下上述数据中6个上升与6个下降的基因热图：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(pheatmap)</div><div class="line"> </div><div class="line">my_group &lt;- data.frame(group = pData(es)$condition)</div><div class="line">row.names(my_group) &lt;- colnames(exprs(es))</div><div class="line"> </div><div class="line">pheatmap(</div><div class="line">  mat = es[c(head(de[order(t), <span class="number">1</span>])$rn, tail(de[order(t), <span class="number">1</span>])$rn),],</div><div class="line">  annotation_col = my_group,</div><div class="line">  cluster_rows = <span class="literal">FALSE</span>,</div><div class="line">  cellwidth=<span class="number">25</span>,</div><div class="line">  cellheight=<span class="number">15</span></div><div class="line">)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20190911133906.png">

</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html" target="_blank" rel="external">Using fgsea package</a></li>
<li><a href="https://www.r-bloggers.com/comparison-of-clusterprofiler-and-gsea-p/" target="_blank" rel="external">Comparison of clusterProfiler and GSEA-P</a></li>
<li><a href="https://davetang.org/muse/2018/01/10/using-fast-preranked-gene-set-enrichment-analysis-fgsea-package/" target="_blank" rel="external">Using the fast preranked gene set enrichment analysis (fgsea) package</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/10/Genomics Data Analysis Series/GDAS010- 生物学变异与技术变异/" itemprop="url">GDAS010-生物学变异与技术变异</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-10T12:00:00+08:00">
                2019-09-10
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  2,677
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  11
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="前言">前言</h1>
<p>在后面的章节中，我们将介绍基因组学实验中的统计推断。我们会使用前面几节中介绍的一些概念，其中包括t检验和多重比较。在这一节中，我们先介绍一个在基因组学数据分析中一个极为重要的概念：生物学变异与技术变异(biological and technical variability)以及它们之间的区别。</p>
<p>通常而言，我们在一个种群里所观察到的生物学单位，例如某个个体的变异称为生物学变异(biological variability)。我们将那些对同一个生物学单位进行测量时所观察到的变异称为技术性变异(technical variability，例如同一个样本测3次，这3次就是技术重复，3次数据的变异称为技术变异)。由于在基因组学研究中经常会用到新技术，因此技术重复(technical replicates)在很多时候都会用于评估实验数据。理想情况下，同一样本的检测结果应该是相同的，但在实际检测中同一样本的多次检测会出现一定的偏差，因此使用技术重复我们就能我们就能够评估新技术的技术变异。我还经常使用生物学重复(biological replicates)与技术重复(technical replicates)来分别代指生物学变异和技术变异。</p>
<p>我们在进行统计推断来解释有差异时，需要注意不要混淆生物学变异和技术变异，因为它们所代表的意义是不同的。例如，当我们分析技术重复时，样本仅仅只有一个，而不是代表某个总体，例如人类或小鼠。这一节中，我们通过设计一个实验来说明一下技术重复和生物重复。</p>
<h1 id="合并实验数据">合并实验数据</h1>
<p>我们将要研究的数据集来源于基因表达微阵列。在这个实验中，我们从两个品系的小鼠中分别随机选择了12只小鼠，并提取了它们的RNA。我们将得到的24个样本使用微阵列的手段进行检测，引外，我们还构成了两个库，这两个库中的样本则是分别来源于这两个品系的12只小鼠的混合样本。</p>
<p>现在我们安装以下包，这个包中含有这些数据，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(devtools)</div><div class="line">install_github(&quot;genomicsclass/maPooling&quot;)</div></pre></td></tr></table></figure>
<p>使用<code>pData</code>函数我们就能查看实验设计。每行表示一个样本，每列表示1只小鼠。例如<code>i,j</code>单元格，中如果是1，则表示RNA数据来源于<code>j</code>只小鼠的样本<code>i</code>。通过行名我们可以知道品系信息（不推荐这种方法来进行分组，你可以向phenoData中添加一个变量，用于指定小鼠品系信息），数据如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Biobase)</div><div class="line"><span class="keyword">library</span>(maPooling)</div><div class="line">data(maPooling)</div><div class="line">head(pData(maPooling))</div><div class="line"><span class="comment">##          a2 a3 a4 a5 a6 a7 a8 a9 a10 a11 a12 a14 b2 b3 b5 b6 b8 b9 b10 b11</span></div><div class="line"><span class="comment">## a10       0  0  0  0  0  0  0  0   1   0   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a10a11    0  0  0  0  0  0  0  0   1   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a10a11a4  0  0  1  0  0  0  0  0   1   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a11       0  0  0  0  0  0  0  0   0   1   0   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a12       0  0  0  0  0  0  0  0   0   0   1   0  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">## a12a14    0  0  0  0  0  0  0  0   0   0   1   1  0  0  0  0  0  0   0   0</span></div><div class="line"><span class="comment">##          b12 b13 b14 b15</span></div><div class="line"><span class="comment">## a10        0   0   0   0</span></div><div class="line"><span class="comment">## a10a11     0   0   0   0</span></div><div class="line"><span class="comment">## a10a11a4   0   0   0   0</span></div><div class="line"><span class="comment">## a11        0   0   0   0</span></div><div class="line"><span class="comment">## a12        0   0   0   0</span></div><div class="line"><span class="comment">## a12a14     0   0   0   0</span></div></pre></td></tr></table></figure>
<p>现在我们创建一个函数，用于说明小鼠的样本信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rafalib)</div><div class="line">mypar()</div><div class="line">flipt &lt;- <span class="keyword">function</span>(m) t(m[nrow(m):<span class="number">1</span>,])</div><div class="line">myimage &lt;- <span class="keyword">function</span>(m,<span class="keyword">...</span>) &#123;</div><div class="line">  image(flipt(m),xaxt=<span class="string">"n"</span>,yaxt=<span class="string">"n"</span>,<span class="keyword">...</span>)</div><div class="line">  &#125;</div><div class="line"></div><div class="line">myimage(as.matrix(pData(maPooling)),col=c(<span class="string">"white"</span>,<span class="string">"black"</span>),</div><div class="line">        xlab=<span class="string">"experiments"</span>,</div><div class="line">        ylab=<span class="string">"individuals"</span>,</div><div class="line">        main=<span class="string">"phenoData"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-3-1.png" alt="plot of chunk unnamed-chunk-3">
<p class="caption">plot of chunk unnamed-chunk-3</p>
</div>
<p>需要注意的是，最终我们只对两个品系小鼠的差异基因感兴趣，也就是数据中标为0的品系和1的品系。我们可以对合并样本的技术重复进行检验或来源于12只小鼠的数据进行检验。我们可以提取出这些合并的样本，因为来自于每个品系的所有小鼠都都可以用0或1来表示，因此那些来源于某品系（例如用1表示的这个品系）代表了这些样本，因此我们提取出矩阵行之和为12的基因，这段话不太理解，原文如下：</p>
<blockquote>
<p>Note that ultimately we are interested in detecting genes that are differentially expressed between the two strains of mice which we will refer to as strain 0 and 1. We can apply tests to the technical replicates of pooled samples or the data from 12 individual mice. We can identify these pooled samples because all mice from each strain were represented in these samples and thus the sum of the rows of experimental design matrix add up to 12:</p>
</blockquote>
<p>如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">data(maPooling)</div><div class="line">pd=pData(maPooling)</div><div class="line">pooled=which(rowSums(pd)==12)</div></pre></td></tr></table></figure>
<p>我们可以从列名确定品系，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">factor(as.numeric(grepl(<span class="string">"b"</span>,names(pooled))))</div><div class="line"><span class="comment">## [1] 0 0 0 0 1 1 1 1</span></div><div class="line"><span class="comment">## Levels: 0 1</span></div></pre></td></tr></table></figure>
<p>如果我们要比较每个基因在不同品系之间的差异，我们就会发现有几个基因出现了比较一致的差异，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="comment">###look at 2 pre-selected genes for illustration</span></div><div class="line">i=<span class="number">11425</span>;j=<span class="number">11878</span></div><div class="line">pooled_y=exprs(maPooling[,pooled])</div><div class="line">pooled_g=factor(as.numeric(grepl(<span class="string">"b"</span>,names(pooled))))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">2</span>)</div><div class="line">stripchart(split(pooled_y[i,],pooled_g),vertical=<span class="literal">TRUE</span>,method=<span class="string">"jitter"</span>,col=c(<span class="number">1</span>,<span class="number">2</span>),</div><div class="line">           main=<span class="string">"Gene 1"</span>,xlab=<span class="string">"Group"</span>,pch=<span class="number">15</span>)</div><div class="line">stripchart(split(pooled_y[j,],pooled_g),vertical=<span class="literal">TRUE</span>,method=<span class="string">"jitter"</span>,col=c(<span class="number">1</span>,<span class="number">2</span>),</div><div class="line">           main=<span class="string">"Gene 2"</span>,xlab=<span class="string">"Group"</span>,pch=<span class="number">15</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<p>这里要注意一下，如果我们根据这些值来进行t检验，那么我们就会得到非常显著的结果，也就是说p值极低，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(genefilter)</div><div class="line">pooled_tt=rowttests(pooled_y,pooled_g)</div><div class="line">pooled_tt$p.value[i]</div><div class="line"><span class="comment">## [1] 2.075617e-07</span></div><div class="line">pooled_tt$p.value[j]</div><div class="line"><span class="comment">## [1] 3.400476e-07</span></div></pre></td></tr></table></figure>
<p>但是，如果我们再次从两个品系的小鼠中各选择12个小鼠的话，这样的结论会成立吗？请注意，t检验的定义包括要比较的总体的标准差。我们这里做的计算结果是否能够反映出总体差异？（注：这里要回顾一下t检验的定义，样本与方差的差异等概念）。</p>
<p>需要注意的是，我们现在重复计算的过程就是在重复原来的实验流程。我们为每个合并的样本创建了4个技术重复。或许Gene 1在小鼠品系内存在着很高的变异，而Gene 2的变异则比较低，但是我们不能看到这一点，因为小鼠之间的变异被合并后的数据所掩盖了（注：这里作者只是假设了Gene 1可能存在着比较高的组内变异）。</p>
<p>我们还有每只小鼠的微阵列数据。对于每个品系的小鼠来说，我们有12个生物学重复。我们可以通过查询1来找到这个品系，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">individuals=which(rowSums(pd)==1)</div></pre></td></tr></table></figure>
<p>事实证明，某些单独的小鼠本身就含有技术重复，因此现在我们将这种情况剔除掉，从而用于说明仅用生物学重复的分析结果，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="comment">##remove replicates</span></div><div class="line">individuals=individuals[-grep(<span class="string">"tr"</span>,names(individuals))]</div><div class="line">y=exprs(maPooling)[,individuals]</div><div class="line">g=factor(as.numeric(grepl(<span class="string">"b"</span>,names(individuals))))</div></pre></td></tr></table></figure>
<p>我们可以计算每个基因的在某个品系中的方差，并与通过技术重复获得的标准差进行比较，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">technicalsd &lt;- rowSds(pooled_y[,pooled_g==<span class="number">0</span>])</div><div class="line">biologicalsd &lt;- rowSds(y[,g==<span class="number">0</span>])</div><div class="line">LIM=range(c(technicalsd,biologicalsd))</div><div class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</div><div class="line">boxplot(technicalsd,biologicalsd,names=c(<span class="string">"technical"</span>,<span class="string">"biological"</span>),ylab=<span class="string">"standard deviation"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>从上图上我们可以看到，生物学方差(variance)远大于技术方差。并且对于方差的变异而言，生物学方差更大。这只是我们前面提到的2个基因的情况，现在我们展示每只小鼠上测得的这2个基因的表达值，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">mypar(1,2)</div><div class="line">stripchart(split(y[i,],g),vertical=TRUE,method=&quot;jitter&quot;,col=c(1,2),xlab=&quot;Gene 1&quot;,pch=15)</div><div class="line">points(c(1,2),tapply(y[i,],g,mean),pch=4,cex=1.5)</div><div class="line">stripchart(split(y[j,],g),vertical=TRUE,method=&quot;jitter&quot;,col=c(1,2),xlab=&quot;Gene 2&quot;,pch=15)</div><div class="line">points(c(1,2),tapply(y[j,],g,mean),pch=4,cex=1.5)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_btvari-unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11">
<p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<p>现在p值就发生了一点变化，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(genefilter)</div><div class="line">tt=rowttests(y,g)</div><div class="line">tt$p.value[i]</div><div class="line">## [1] 0.0898726</div><div class="line">tt$p.value[j]</div><div class="line">## [1] 1.979172e-07</div></pre></td></tr></table></figure>
<p>当我们在不同品系的小鼠之间比较这2个基因的差异时，我们对哪个基因的差异更有信心呢（Gene 1还是Gene 2）？如果另一位研究人员另外一批随机样本来做实验，你认识他会发现哪个基因有差异（Gene 1还是Gene 2）？如果我们希望我们的结论是关于小鼠品系的通用结论（通用结论这里我的理解就是，无论放哪个实验室做，都能做出这个结果，得到这个结论），而不仅仅是我们这次实验的结论（也就是说，有可能就是只有自己实验室能做出这个结果，那么这个结果就不太可信，有可能是假阳性），那么检测生物学的变异则非常重要。</p>
<p>在这一节中，我们使用了两种分析方法。第一种分析方法是将这两个品系的生物学重复做为总体总体来进行分析析（也就是说把某个品系的12只小鼠数据汇总起来分析）。第二种则是将合并后的数据来作为总体进行分析（我的理解就是把12只小鼠的RNA混合在一起，作为一个样本，然后使用4个技术重复来进行分析，这种分析方法其实检测的就是技术变异）。在科学研究中，我们通常关注的是总体（也就是第一种分析方法）。作为一个非常实际的例子，这里需要注意的是，如果另外一个实验室重复该实验，他们会有另外一组12个小鼠的样本，因此关于我们实验中对总体的推断也会在另外的实验室中得到重复。</p>
<blockquote>
<p>An analysis with biological replicates has as a population these two strains of mice. An analysis with technical replicates has as a population the twelve mice we selected and the variability is related to the measurement technology. In science we typically are concerned with populations. As a very practical example, note that if another lab performs this experiment they will have another set of twelve mice and thus inferences about <em>populations</em> are more likely to be reproducible.</p>
</blockquote>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_btvari.html" target="_blank" rel="external">Biological versus technical variability</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/09/Genomics Data Analysis Series/GDAS009-Bioconductor中的基因组注释/" itemprop="url">GDAS009-Bioconductor中的基因组注释</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-09T12:00:00+08:00">
                2019-09-09
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  6,586
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  31
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>这一部分内容涉及R中使用人类基因且，内含子，外显子，转录本，AnnotationHub，基因组的注释包，GO分析，KEGG分析等，笔记末尾的参考文献是原文。</p>
<h2 id="基础注释资源与发现">基础注释资源与发现</h2>
<p>在这一部分里，我们将回顾Bioconductor中用于处理和注释基因组序列的一些工具。我们将研究参考基因组序列，转录本和基因，并以基因通路(gene pathway)作为结束。我们学习这一部分的最终目标就是使用注释信息来帮助我们对基因组实验进行可靠的解释。Bioconductor的基本目标就是更加方便地有关基因组结构和功能的信息统计统计分析程序。</p>
<h3 id="注释概念的层次结构">注释概念的层次结构</h3>
<p>Bioconductor包括许多不同类型的基因组注释。我们可以在层次结构中来理解这些注释资源。</p>
<ul>
<li>最基因的注释就是某个物种的参考基因组序列。它总是按照核苷酸的线性方式排列成染色体（例如参考基因组。</li>
<li>在此之上就是将染色体序列排列到感兴趣的区域中。最感兴趣的区域就是基因，但是注释中也含有其它的信息，例如SNP或CpG位点。基因具有内部结构，即被转录的部分和未被转录的部分。“基因模式”定义了在基因组坐标中的标记和布置这些结构的方式。</li>
<li>在感兴趣的区域(regions of interest)的理念下，我们还定义了面向平台的注释(platform-oriented annotation)。这处类型的注释通常首先是由厂家提供的，但随着研究的进行，对这些平台中最初有歧义信息进行了确认和更新，从而完善了这些注释内容。密歇根大学的<a href="http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp" target="_blank" rel="external">brainarray project</a> 说明了affymetrix阵列注释的过程。我们将在本节最后讨论面向平台注释的问题。</li>
<li>在此之是是将区域（通常是基因或基因的产物）组成成具有共同结构或功能特性的组。例如在细胞中共同被发现的，或者是被鉴定为在生物学过程中协同作用的基因组（我的理解就是GO分析，KEGG分析这一类）。</li>
</ul>
<h3 id="发现可用的参考基因组">发现可用的参考基因组</h3>
<p>Bioconductor已经包含了注释包的合成，将它这一层次结构上的所有元素都带了可编程环境中。参考基因组序列是使用Biostrings和BSgenome包中的工具进行管理的，<code>available.genomes</code> 函数能够列出构建好的人和现在各种模式生物的参考基因组，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div></pre></td><td class="code"><pre><div class="line">library(Biostrings)</div><div class="line">ag = available.genomes()</div><div class="line">length(ag)</div><div class="line">## [1] 87</div><div class="line">head(ag)</div><div class="line">## [1] &quot;BSgenome.Alyrata.JGI.v1&quot;                </div><div class="line">## [2] &quot;BSgenome.Amellifera.BeeBase.assembly4&quot;  </div><div class="line">## [3] &quot;BSgenome.Amellifera.UCSC.apiMel2&quot;       </div><div class="line">## [4] &quot;BSgenome.Amellifera.UCSC.apiMel2.masked&quot;</div><div class="line">## [5] &quot;BSgenome.Athaliana.TAIR.04232008&quot;       </div><div class="line">## [6] &quot;BSgenome.Athaliana.TAIR.TAIR9&quot;</div></pre></td></tr></table></figure>
<h3 id="参考基因组的版本很重要">参考基因组的版本很重要</h3>
<p>不同物种的参考基因组是从头构建的，然后随着算法和测序数据的不断改进而进一步完善。对人类而言，基因组研究联盟（Genome Research Consortium）于2009年构建了37号版本，并于2013年构建了38号版本。</p>
<p>一旦参考基因组构建完成，就哦可以很轻松地对某个物种进行信息丰富的基因组序列分析，因为人们可以专注于那引起已知含有等位基因多样性的区域。</p>
<p>The reference build for an organism is created de novo and then refined as algorithms and sequenced data improve. For humans, the Genome Research Consortium signed off on build 37 in 2009, and on build 38 in 2013.</p>
<p>需要注意的是，基因组序列包含有很长的名称，这个名称里包括版本信息。这样命名的方式就是为了避免与不同版本的参考基因组混淆。在LiftOver这节视频里，我们就展示了如何使用UCSC的<code>liftOver</code>工具与<code>rtracklayer</code>包中的接口对接，从而实现不同版本的基因组坐标转化的过程。</p>
<p>为了帮助用户避免混淆从不同参考基因组坐标上收集分析来的数据，我们提供了一个”基因组“标签，这个标签填充了大多关于序列的信息。在随后的部分里，我们会看到一些案例。用于序列比对的软件可以检查被比对上的序列的兼容标签，从而有助于确保有意义的结果。</p>
<h2 id="h.-sapiens的参考基因序列">H. sapiens的参考基因序列</h2>
<p>通过安装并添加一个单独的R包就能获取智人(Homo sapiens)的参考序列。这个程序包定义了一个<code>Hsapiens</code>对象，试剂公司对象是染色体序列的来源，但是当对其进行单独显示时，它会提供相关序列数据来源的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(BSgenome.Hsapiens.UCSC.hg19)</div><div class="line">Hsapiens</div><div class="line"><span class="comment">## Human genome:</span></div><div class="line"><span class="comment">## # organism: Homo sapiens (Human)</span></div><div class="line"><span class="comment">## # provider: UCSC</span></div><div class="line"><span class="comment">## # provider version: hg19</span></div><div class="line"><span class="comment">## # release date: Feb. 2009</span></div><div class="line"><span class="comment">## # release name: Genome Reference Consortium GRCh37</span></div><div class="line"><span class="comment">## # 93 sequences:</span></div><div class="line"><span class="comment">## #   chr1                  chr2                  chr3                 </span></div><div class="line"><span class="comment">## #   chr4                  chr5                  chr6                 </span></div><div class="line"><span class="comment">## #   chr7                  chr8                  chr9                 </span></div><div class="line"><span class="comment">## #   chr10                 chr11                 chr12                </span></div><div class="line"><span class="comment">## #   chr13                 chr14                 chr15                </span></div><div class="line"><span class="comment">## #   ...                   ...                   ...                  </span></div><div class="line"><span class="comment">## #   chrUn_gl000235        chrUn_gl000236        chrUn_gl000237       </span></div><div class="line"><span class="comment">## #   chrUn_gl000238        chrUn_gl000239        chrUn_gl000240       </span></div><div class="line"><span class="comment">## #   chrUn_gl000241        chrUn_gl000242        chrUn_gl000243       </span></div><div class="line"><span class="comment">## #   chrUn_gl000244        chrUn_gl000245        chrUn_gl000246       </span></div><div class="line"><span class="comment">## #   chrUn_gl000247        chrUn_gl000248        chrUn_gl000249       </span></div><div class="line"><span class="comment">## # (use 'seqnames()' to see all the sequence names, use the '$' or '[['</span></div><div class="line"><span class="comment">## # operator to access a given sequence)</span></div><div class="line">head(genome(Hsapiens))  <span class="comment"># see the tag</span></div><div class="line"><span class="comment">##   chr1   chr2   chr3   chr4   chr5   chr6 </span></div><div class="line"><span class="comment">## "hg19" "hg19" "hg19" "hg19" "hg19" "hg19"</span></div></pre></td></tr></table></figure>
<p>我们使用 <code>$</code> 符号来获取17号染色体的序列，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div></pre></td><td class="code"><pre><div class="line">Hsapiens$chr17</div><div class="line">##   81195210-letter &quot;DNAString&quot; instance</div><div class="line">## seq: AAGCTTCTCACCCTGTTCCTGCATAGATAATTGC...GGTGTGGGTGTGGTGTGTGGGTGTGGGTGTGGT</div></pre></td></tr></table></figure>
<h2 id="参考序列的转录本和基因">参考序列的转录本和基因</h2>
<h3 id="ucsc注释">UCSC注释</h3>
<p><code>TxDb</code>包家族和数据对象管理了转录本和基因模式信息。我们可以认为这些信息来源于UCSC基因组浏览器的注释表，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">library(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene # abbreviate</div><div class="line">txdb</div><div class="line">## TxDb object:</div><div class="line">## # Db type: TxDb</div><div class="line">## # Supporting package: GenomicFeatures</div><div class="line">## # Data source: UCSC</div><div class="line">## # Genome: hg19</div><div class="line">## # Organism: Homo sapiens</div><div class="line">## # Taxonomy ID: 9606</div><div class="line">## # UCSC Table: knownGene</div><div class="line">## # Resource URL: http://genome.ucsc.edu/</div><div class="line">## # Type of Gene ID: Entrez Gene ID</div><div class="line">## # Full dataset: yes</div><div class="line">## # miRBase build ID: GRCh37</div><div class="line">## # transcript_nrow: 82960</div><div class="line">## # exon_nrow: 289969</div><div class="line">## # cds_nrow: 237533</div><div class="line">## # Db created by: GenomicFeatures package from Bioconductor</div><div class="line">## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)</div><div class="line">## # GenomicFeatures version at creation time: 1.21.30</div><div class="line">## # RSQLite version at creation time: 1.0.0</div><div class="line">## # DBSCHEMAVERSION: 1.1</div></pre></td></tr></table></figure>
<p>我们使用 <code>genes()</code> 来获取Entrez Gene ID的地址，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">ghs = genes(txdb)</div><div class="line">ghs</div><div class="line">## GRanges object with 23056 ranges and 1 metadata column:</div><div class="line">##         seqnames                 ranges strand |     gene_id</div><div class="line">##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</div><div class="line">##       1    chr19 [ 58858172,  58874214]      - |           1</div><div class="line">##      10     chr8 [ 18248755,  18258723]      + |          10</div><div class="line">##     100    chr20 [ 43248163,  43280376]      - |         100</div><div class="line">##    1000    chr18 [ 25530930,  25757445]      - |        1000</div><div class="line">##   10000     chr1 [243651535, 244006886]      - |       10000</div><div class="line">##     ...      ...                    ...    ... .         ...</div><div class="line">##    9991     chr9 [114979995, 115095944]      - |        9991</div><div class="line">##    9992    chr21 [ 35736323,  35743440]      + |        9992</div><div class="line">##    9993    chr22 [ 19023795,  19109967]      - |        9993</div><div class="line">##    9994     chr6 [ 90539619,  90584155]      + |        9994</div><div class="line">##    9997    chr22 [ 50961997,  50964905]      - |        9997</div><div class="line">##   -------</div><div class="line">##   seqinfo: 93 sequences (1 circular) from hg19 genome</div></pre></td></tr></table></figure>
<p>我们也可以使用合适的标识符进行信息过滤。现在我们提取两个不同基因的外显子，这些外显子由其Entrez基因ID标明，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div></pre></td><td class="code"><pre><div class="line">exons(txdb, columns=c(<span class="string">"EXONID"</span>, <span class="string">"TXNAME"</span>, <span class="string">"GENEID"</span>),</div><div class="line">                  filter=list(gene_id=c(<span class="number">100</span>, <span class="number">101</span>)))</div><div class="line"><span class="comment">## GRanges object with 39 ranges and 3 metadata columns:</span></div><div class="line"><span class="comment">##        seqnames                 ranges strand |    EXONID</span></div><div class="line"><span class="comment">##           &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span></div><div class="line"><span class="comment">##    [1]    chr10 [135075920, 135076737]      - |    144421</span></div><div class="line"><span class="comment">##    [2]    chr10 [135077192, 135077269]      - |    144422</span></div><div class="line"><span class="comment">##    [3]    chr10 [135080856, 135080921]      - |    144423</span></div><div class="line"><span class="comment">##    [4]    chr10 [135081433, 135081570]      - |    144424</span></div><div class="line"><span class="comment">##    [5]    chr10 [135081433, 135081622]      - |    144425</span></div><div class="line"><span class="comment">##    ...      ...                    ...    ... .       ...</span></div><div class="line"><span class="comment">##   [35]    chr20   [43254210, 43254325]      - |    256371</span></div><div class="line"><span class="comment">##   [36]    chr20   [43255097, 43255240]      - |    256372</span></div><div class="line"><span class="comment">##   [37]    chr20   [43257688, 43257810]      - |    256373</span></div><div class="line"><span class="comment">##   [38]    chr20   [43264868, 43264929]      - |    256374</span></div><div class="line"><span class="comment">##   [39]    chr20   [43280216, 43280376]      - |    256375</span></div><div class="line"><span class="comment">##                                  TXNAME          GENEID</span></div><div class="line"><span class="comment">##                         &lt;CharacterList&gt; &lt;CharacterList&gt;</span></div><div class="line"><span class="comment">##    [1] uc009ybi.3,uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [2]            uc009ybi.3,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [3] uc009ybi.3,uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    [4]                       uc009ybi.3             101</span></div><div class="line"><span class="comment">##    [5]            uc010qva.2,uc021qbe.1             101</span></div><div class="line"><span class="comment">##    ...                              ...             ...</span></div><div class="line"><span class="comment">##   [35]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [36]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [37]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [38]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   [39]                       uc002xmj.3             100</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 93 sequences (1 circular) from hg19 genome</span></div></pre></td></tr></table></figure>
<h3 id="ensembl注释">ENSEMBL注释</h3>
<p><a href="http://www.ensembl.org/index.html" target="_blank" rel="external">Ensembl home主页</a>上写道：Ensembl创建，整合和发布研究基因组的参考数据库和工具。该项目位于 <a href="https://www.ebi.ac.uk/" target="_blank" rel="external">欧洲分子生物学实验室</a>，该实验室的数据库支持其注释资源可以与Bioconductor兼容。</p>
<p><code>ensembldb</code> 包含有一个简要说明，其内容如下所示：</p>
<p>ensembldb包提供了一些函数，这些函数用于创建和使用以转录本为中心的注释数据库/包。使用注释数据库的Perl API可以从Ensembl 1中直接获取这些数据。<code>TxDb</code> 包的功能和数据类似于<code>GenomicFeatures</code>包，另外，除了从数据库检索所有的基因/转录本模型和注释外，ensembldb包还提供了一个过滤框架，用于检索特定条目的注释，例如位于染色体区域上的某编码基因或某LincRNA转录模式的特定条目。从1.7版本开始，由ensembldb创建的EnsDb数据库还包含蛋白质注释数据库（参考<a href="http://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html#org35014ed" target="_blank" rel="external">第11节</a>：数据库而已和可用属性/列的概述）。有关蛋白质注释的信息请参考蛋白质的vignette，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ensembldb)</div><div class="line"><span class="keyword">library</span>(EnsDb.Hsapiens.v75)</div><div class="line">names(listTables(EnsDb.Hsapiens.v75))</div><div class="line"><span class="comment">##  [1] "gene"           "tx"             "tx2exon"        "exon"          </span></div><div class="line"><span class="comment">##  [5] "chromosome"     "protein"        "uniprot"        "protein_domain"</span></div><div class="line"><span class="comment">##  [9] "entrezgene"     "metadata"</span></div></pre></td></tr></table></figure>
<p>举例说明如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div></pre></td><td class="code"><pre><div class="line">edb = EnsDb.Hsapiens.v75  <span class="comment"># abbreviate</span></div><div class="line">txs &lt;- transcripts(edb, filter = GenenameFilter(<span class="string">"ZBTB16"</span>),</div><div class="line">                   columns = c(<span class="string">"protein_id"</span>, <span class="string">"uniprot_id"</span>, <span class="string">"tx_biotype"</span>))</div><div class="line">txs</div><div class="line"><span class="comment">## GRanges object with 20 ranges and 5 metadata columns:</span></div><div class="line"><span class="comment">##                   seqnames                 ranges strand |      protein_id</span></div><div class="line"><span class="comment">##                      &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##   ENST00000335953       11 [113930315, 114121398]      + | ENSP00000338157</span></div><div class="line"><span class="comment">##               ...      ...                    ...    ... .             ...</span></div><div class="line"><span class="comment">##   ENST00000392996       11 [113931229, 114121374]      + | ENSP00000376721</span></div><div class="line"><span class="comment">##   ENST00000539918       11 [113935134, 114118066]      + | ENSP00000445047</span></div><div class="line"><span class="comment">##   ENST00000545851       11 [114051488, 114118018]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ENST00000535379       11 [114107929, 114121279]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##   ENST00000535509       11 [114117512, 114121198]      + |            &lt;NA&gt;</span></div><div class="line"><span class="comment">##                     uniprot_id              tx_biotype           tx_id</span></div><div class="line"><span class="comment">##                    &lt;character&gt;             &lt;character&gt;     &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953  ZBT16_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL7_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL6_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 Q71UL5_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##   ENST00000335953 F5H6C3_HUMAN          protein_coding ENST00000335953</span></div><div class="line"><span class="comment">##               ...          ...                     ...             ...</span></div><div class="line"><span class="comment">##   ENST00000392996 F5H5Y7_HUMAN          protein_coding ENST00000392996</span></div><div class="line"><span class="comment">##   ENST00000539918         &lt;NA&gt; nonsense_mediated_decay ENST00000539918</span></div><div class="line"><span class="comment">##   ENST00000545851         &lt;NA&gt;    processed_transcript ENST00000545851</span></div><div class="line"><span class="comment">##   ENST00000535379         &lt;NA&gt;    processed_transcript ENST00000535379</span></div><div class="line"><span class="comment">##   ENST00000535509         &lt;NA&gt;         retained_intron ENST00000535509</span></div><div class="line"><span class="comment">##                     gene_name</span></div><div class="line"><span class="comment">##                   &lt;character&gt;</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000335953      ZBTB16</span></div><div class="line"><span class="comment">##               ...         ...</span></div><div class="line"><span class="comment">##   ENST00000392996      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000539918      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000545851      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000535379      ZBTB16</span></div><div class="line"><span class="comment">##   ENST00000535509      ZBTB16</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from GRCh37 genome</span></div></pre></td></tr></table></figure>
<h2 id="你的数据将会成他人的注释导入导出">你的数据将会成他人的注释：导入/导出</h2>
<p>ENCODE项目很地说明了今天的实验是明天的注释。你应该以同样的方式考虑自己的实验（当然，要使实验成为可靠且持久的注释，它必须解决有关基因组结构或功能的重要问题，并且必须使用适当的，能正确执行的实验流程。需要注意，ENCODE能够非常明确地将实验流程链接到数据）。</p>
<p>例如，我们来看一个雌激素受体结合数据，它是由ENCODE发布的一个narrowPeak 数据。它的碱基是用ascii文本表示的，因此可以很容易地导入为一组文本数据。如果记录的字段有一定的规律性，则可以将文件作为表格导入。</p>
<p>但是，我们不仅是想导入数据，还想将导入的数据作为可计算的对象。我们认识到arrowePeak和bedGraph格式之间的联系后，我们就可以立即将其导入GRanges中。</p>
<p>为了说明这一点，我们在ERBS包中找到narrowPeak原始数据文件的路径，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">f1 = dir(system.file(<span class="string">"extdata"</span>,package=<span class="string">"ERBS"</span>), full=<span class="literal">TRUE</span>)[<span class="number">1</span>]</div><div class="line">readLines(f1, <span class="number">4</span>) <span class="comment"># look at a few lines</span></div><div class="line"><span class="comment">## [1] "chrX\t1509354\t1512462\t5\t0\t.\t157.92\t310\t32.000000\t1991"    </span></div><div class="line"><span class="comment">## [2] "chrX\t26801421\t26802448\t6\t0\t.\t147.38\t310\t32.000000\t387"   </span></div><div class="line"><span class="comment">## [3] "chr19\t11694101\t11695359\t1\t0\t.\t99.71\t311.66\t32.000000\t861"</span></div><div class="line"><span class="comment">## [4] "chr19\t4076892\t4079276\t4\t0\t.\t84.74\t310\t32.000000\t1508"</span></div></pre></td></tr></table></figure>
<p>使用<code>import</code>命令非常简单，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rtracklayer)</div><div class="line">imp = import(f1, format=<span class="string">"bedGraph"</span>)</div><div class="line">imp</div><div class="line"><span class="comment">## GRanges object with 1873 ranges and 7 metadata columns:</span></div><div class="line"><span class="comment">##          seqnames               ranges strand |     score       NA.</span></div><div class="line"><span class="comment">##             &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##      [1]     chrX [ 1509355,  1512462]      * |         5         0</span></div><div class="line"><span class="comment">##      [2]     chrX [26801422, 26802448]      * |         6         0</span></div><div class="line"><span class="comment">##      [3]    chr19 [11694102, 11695359]      * |         1         0</span></div><div class="line"><span class="comment">##      [4]    chr19 [ 4076893,  4079276]      * |         4         0</span></div><div class="line"><span class="comment">##      [5]     chr3 [53288568, 53290767]      * |         9         0</span></div><div class="line"><span class="comment">##      ...      ...                  ...    ... .       ...       ...</span></div><div class="line"><span class="comment">##   [1869]    chr19 [11201120, 11203985]      * |      8701         0</span></div><div class="line"><span class="comment">##   [1870]    chr19 [ 2234920,  2237370]      * |       990         0</span></div><div class="line"><span class="comment">##   [1871]     chr1 [94311336, 94313543]      * |      4035         0</span></div><div class="line"><span class="comment">##   [1872]    chr19 [45690614, 45691210]      * |     10688         0</span></div><div class="line"><span class="comment">##   [1873]    chr19 [ 6110100,  6111252]      * |      2274         0</span></div><div class="line"><span class="comment">##               NA.1      NA.2      NA.3      NA.4      NA.5</span></div><div class="line"><span class="comment">##          &lt;logical&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##      [1]      &lt;NA&gt;    157.92       310        32      1991</span></div><div class="line"><span class="comment">##      [2]      &lt;NA&gt;    147.38       310        32       387</span></div><div class="line"><span class="comment">##      [3]      &lt;NA&gt;     99.71    311.66        32       861</span></div><div class="line"><span class="comment">##      [4]      &lt;NA&gt;     84.74       310        32      1508</span></div><div class="line"><span class="comment">##      [5]      &lt;NA&gt;      78.2   299.505        32      1772</span></div><div class="line"><span class="comment">##      ...       ...       ...       ...       ...       ...</span></div><div class="line"><span class="comment">##   [1869]      &lt;NA&gt;      8.65     7.281   0.26576      2496</span></div><div class="line"><span class="comment">##   [1870]      &lt;NA&gt;      8.65    26.258  1.995679      1478</span></div><div class="line"><span class="comment">##   [1871]      &lt;NA&gt;      8.65    12.511   1.47237      1848</span></div><div class="line"><span class="comment">##   [1872]      &lt;NA&gt;      8.65     6.205         0       298</span></div><div class="line"><span class="comment">##   [1873]      &lt;NA&gt;      8.65    17.356  2.013228       496</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></div><div class="line">genome(imp)  <span class="comment"># genome identifier tag not set, but you should set it</span></div><div class="line"><span class="comment">##  chrX chr19  chr3 chr17  chr8 chr11 chr16  chr1  chr2  chr6  chr9  chr7 </span></div><div class="line"><span class="comment">##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA </span></div><div class="line"><span class="comment">##  chr5 chr12 chr20 chr21 chr22 chr18 chr10 chr14 chr15  chr4 chr13 </span></div><div class="line"><span class="comment">##    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA</span></div></pre></td></tr></table></figure>
<p>我们可以通过一次获取GRanges。元数据列中还有一些其他字段用于指定名称，但是如果我们只对范围感兴趣，除了添加基因组元数据以防止与不兼容的坐标中记录的数据非法组合外，我们就完成了这个任务（这一段不太理解，原文如下）：</p>
<blockquote>
<p>We obtain a GRanges in one stroke. There are some additional fields in the metadata columns whose names should be specified, but if we are interested only in the ranges, we are done, with the exception of adding the genome metadata to protect against illegitimate combination with data recorded in an incompatible coordinate system.</p>
</blockquote>
<p>为了与其他得养家或系统进行交流，我们有两个主要选择。我们可以将GRanges保存为<code>RData</code>对象，轻松地传递给另外一个R用户使用。或者，我们歌词采用其他标准格式进行导出。例如，如果我们仅对间隔地址和绑定的得分感兴趣，则仅保存为“bed”格式就足够了，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">export(imp, <span class="string">"demoex.bed"</span>)  <span class="comment"># implicit format choice</span></div><div class="line">cat(readLines(<span class="string">"demoex.bed"</span>, n=<span class="number">5</span>), sep=<span class="string">"\n"</span>)</div><div class="line"><span class="comment">## chrX	1509354	1512462	.	5	.</span></div><div class="line"><span class="comment">## chrX	26801421	26802448	.	6	.</span></div><div class="line"><span class="comment">## chr19	11694101	11695359	.	1	.</span></div><div class="line"><span class="comment">## chr19	4076892	4079276	.	4	.</span></div><div class="line"><span class="comment">## chr3	53288567	53290767	.	9	.</span></div></pre></td></tr></table></figure>
<p>我们已经进行了导入，建模和导入实验数据之间的“往返”，该实验数据可以与其他数据集成在一起，从而增进生物学的理解。</p>
<p>我们需要注意的是，注释在某种程度上是永久正确的，它与在知识边界上的研究进展乏味地隔离开来。我们已经看到了，甚至人类染色体的参考序列也受到了修订。在使用ERBS包时，我们将未知的实验结果视为定义ER结合位点从而进入潜在的生物学解释。不确定性，峰鉴定的可变质量，尚未得到明确估计，但应该是这个样子。</p>
<p>Bioconductor已经尽力致力于这种情况的多个方面。我们维护软件先前版本和注释的存档，以便可以检查或修改过去的工作。我们每年会两次更新中疏注释资源，以确保正在进行的工作以及获得新知识的稳定性。而且，我们已经简化了导入和创建实验数据和注释数据的表示形式。</p>
<h2 id="annotationhub">AnnotationHub</h2>
<p><code>AnnotationHub</code>包用于获取GRanges或其它的合适设计的容器，用于机构设计的容器，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(AnnotationHub)</div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Attaching package: 'AnnotationHub'</span></div><div class="line"><span class="comment">## The following object is masked from 'package:Biobase':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     cache</span></div><div class="line">ah = AnnotationHub()</div><div class="line"><span class="comment">## snapshotDate(): 2017-10-27</span></div><div class="line">ah</div><div class="line"><span class="comment">## AnnotationHub with 42282 records</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # $$dataprovider: BroadInstitute, Ensembl, UCSC, ftp://ftp.ncbi.nlm.nih....</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos ta...</span></div><div class="line"><span class="comment">## # $$rdataclass: GRanges, BigWigFile, FaFile, TwoBitFile, Rle, ChainFile,...</span></div><div class="line"><span class="comment">## # additional mcols(): taxonomyid, genome, description,</span></div><div class="line"><span class="comment">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,</span></div><div class="line"><span class="comment">## #   tags, rdatapath, sourceurl, sourcetype </span></div><div class="line"><span class="comment">## # retrieve records with, e.g., 'object[["AH2"]]' </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##             title                                                         </span></div><div class="line"><span class="comment">##   AH2     | Ailuropoda_melanoleuca.ailMel1.69.dna.toplevel.fa             </span></div><div class="line"><span class="comment">##   AH3     | Ailuropoda_melanoleuca.ailMel1.69.dna_rm.toplevel.fa          </span></div><div class="line"><span class="comment">##   AH4     | Ailuropoda_melanoleuca.ailMel1.69.dna_sm.toplevel.fa          </span></div><div class="line"><span class="comment">##   AH5     | Ailuropoda_melanoleuca.ailMel1.69.ncrna.fa                    </span></div><div class="line"><span class="comment">##   AH6     | Ailuropoda_melanoleuca.ailMel1.69.pep.all.fa                  </span></div><div class="line"><span class="comment">##   ...       ...                                                           </span></div><div class="line"><span class="comment">##   AH58988 | org.Flavobacterium_piscicida.eg.sqlite                        </span></div><div class="line"><span class="comment">##   AH58989 | org.Bacteroides_fragilis_YCH46.eg.sqlite                      </span></div><div class="line"><span class="comment">##   AH58990 | org.Pseudomonas_mendocina_ymp.eg.sqlite                       </span></div><div class="line"><span class="comment">##   AH58991 | org.Salmonella_enterica_subsp._enterica_serovar_Typhimurium...</span></div><div class="line"><span class="comment">##   AH58992 | org.Acinetobacter_baumannii.eg.sqlite</span></div></pre></td></tr></table></figure>
<p>我们可以通过AnnotationHub获得许多与HepG2细胞系相关的实验数据对象，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">query(ah, <span class="string">"HepG2"</span>)</div><div class="line"><span class="comment">## AnnotationHub with 440 records</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # $$dataprovider: UCSC, BroadInstitute, Pazar</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens, NA</span></div><div class="line"><span class="comment">## # $$rdataclass: GRanges, BigWigFile</span></div><div class="line"><span class="comment">## # additional mcols(): taxonomyid, genome, description,</span></div><div class="line"><span class="comment">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,</span></div><div class="line"><span class="comment">## #   tags, rdatapath, sourceurl, sourcetype </span></div><div class="line"><span class="comment">## # retrieve records with, e.g., 'object[["AH22246"]]' </span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##             title                                                         </span></div><div class="line"><span class="comment">##   AH22246 | pazar_CEBPA_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22249 | pazar_CTCF_HEPG2_Schmidt_20120522.csv                         </span></div><div class="line"><span class="comment">##   AH22273 | pazar_HNF4A_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22309 | pazar_STAG1_HEPG2_Schmidt_20120522.csv                        </span></div><div class="line"><span class="comment">##   AH22348 | wgEncodeAffyRnaChipFiltTransfragsHepg2CytosolLongnonpolya.b...</span></div><div class="line"><span class="comment">##   ...       ...                                                           </span></div><div class="line"><span class="comment">##   AH41564 | E118-H4K5ac.imputed.pval.signal.bigwig                        </span></div><div class="line"><span class="comment">##   AH41691 | E118-H4K8ac.imputed.pval.signal.bigwig                        </span></div><div class="line"><span class="comment">##   AH41818 | E118-H4K91ac.imputed.pval.signal.bigwig                       </span></div><div class="line"><span class="comment">##   AH46971 | E118_15_coreMarks_mnemonics.bed.gz                            </span></div><div class="line"><span class="comment">##   AH49484 | E118_RRBS_FractionalMethylation.bigwig</span></div></pre></td></tr></table></figure>
<p><code>query</code> 方法可以使用过滤字符串的向量。要限制对寻址组蛋白H4K5的注释资源的响应，只需要添加该标签，如下所示(To limit response to annotation resources addressing the histone H4K5, simply add that tag)：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">query(ah, c(<span class="string">"HepG2"</span>, <span class="string">"H4K5"</span>))</div><div class="line"><span class="comment">## AnnotationHub with 1 record</span></div><div class="line"><span class="comment">## # snapshotDate(): 2017-10-27 </span></div><div class="line"><span class="comment">## # names(): AH41564</span></div><div class="line"><span class="comment">## # $$dataprovider: BroadInstitute</span></div><div class="line"><span class="comment">## # $$species: Homo sapiens</span></div><div class="line"><span class="comment">## # $$rdataclass: BigWigFile</span></div><div class="line"><span class="comment">## # $$rdatadateadded: 2015-05-08</span></div><div class="line"><span class="comment">## # $$title: E118-H4K5ac.imputed.pval.signal.bigwig</span></div><div class="line"><span class="comment">## # $$description: Bigwig File containing -log10(p-value) signal tracks fr...</span></div><div class="line"><span class="comment">## # $$taxonomyid: 9606</span></div><div class="line"><span class="comment">## # $$genome: hg19</span></div><div class="line"><span class="comment">## # $$sourcetype: BigWig</span></div><div class="line"><span class="comment">## # $$sourceurl: http://egg2.wustl.edu/roadmap/data/byFileType/signal/cons...</span></div><div class="line"><span class="comment">## # $$sourcesize: 226630905</span></div><div class="line"><span class="comment">## # $$tags: c("EpigenomeRoadMap", "signal", "consolidatedImputed",</span></div><div class="line"><span class="comment">## #   "H4K5ac", "E118", "ENCODE2012", "LIV.HEPG2.CNCR", "HepG2</span></div><div class="line"><span class="comment">## #   Hepatocellular Carcinoma Cell Line") </span></div><div class="line"><span class="comment">## # retrieve record with 'object[["AH41564"]]'</span></div></pre></td></tr></table></figure>
<h2 id="the-orgdb基因注释图">The OrgDb基因注释图</h2>
<p>那些命名为<code>org.*.ge.db</code> 的包含在基因水平上链接到位置，蛋白产物标识符，KEGG途径和GO term，PMIDs以及其它注释资源的标识符的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(org.Hs.eg.db)</div><div class="line">keytypes(org.Hs.eg.db) <span class="comment"># columns() gives same answer</span></div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" </span></div><div class="line"><span class="comment">##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    </span></div><div class="line"><span class="comment">##  [9] "EVIDENCEALL"  "GENENAME"     "GO"           "GOALL"       </span></div><div class="line"><span class="comment">## [13] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    </span></div><div class="line"><span class="comment">## [17] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        </span></div><div class="line"><span class="comment">## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      </span></div><div class="line"><span class="comment">## [25] "UNIGENE"      "UNIPROT"</span></div><div class="line">head(select(org.Hs.eg.db, keys=<span class="string">"ORMDL3"</span>, keytype=<span class="string">"SYMBOL"</span>, </div><div class="line">   columns=<span class="string">"PMID"</span>))</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">##   SYMBOL     PMID</span></div><div class="line"><span class="comment">## 1 ORMDL3 11042152</span></div><div class="line"><span class="comment">## 2 ORMDL3 12093374</span></div><div class="line"><span class="comment">## 3 ORMDL3 12477932</span></div><div class="line"><span class="comment">## 4 ORMDL3 14702039</span></div><div class="line"><span class="comment">## 5 ORMDL3 15489334</span></div><div class="line"><span class="comment">## 6 ORMDL3 16169070</span></div></pre></td></tr></table></figure>
<h2 id="基因集和通路资源">基因集和通路资源</h2>
<h3 id="基因本体论">基因本体论</h3>
<p><a href="http://www.geneontology.org" target="_blank" rel="external">Gene Ontology</a> (GO)是一种广泛使用的结构化词汇，它组织了基因和基因产物在以下方面的内容：</p>
<ul>
<li>生物过程</li>
<li>分子功能</li>
<li>细胞组分。</li>
</ul>
<p>这套词汇本身旨在与所有生物有关。它采用有向无环图的形式，其中term作为节点，使用<code>is-a</code>和<code>part-of</code>关系作构成了大多数链接。</p>
<p>将生物体特定基因链接到基因本体中的术语的注释与词汇表本身是分开的，并且涉及不同类型的证据。这些记录都在Bioconductor的注释包中。</p>
<p>我们可以使用<code>GO.db</code>包来快速地访问GO词汇，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GO.db)</div><div class="line">GO.db <span class="comment"># metadata</span></div><div class="line"><span class="comment">## GODb object:</span></div><div class="line"><span class="comment">## | GOSOURCENAME: Gene Ontology</span></div><div class="line"><span class="comment">## | GOSOURCEURL: ftp://ftp.geneontology.org/pub/go/godatabase/archive/latest-lite/</span></div><div class="line"><span class="comment">## | GOSOURCEDATE: 2017-Nov01</span></div><div class="line"><span class="comment">## | Db type: GODb</span></div><div class="line"><span class="comment">## | package: AnnotationDbi</span></div><div class="line"><span class="comment">## | DBSCHEMA: GO_DB</span></div><div class="line"><span class="comment">## | GOEGSOURCEDATE: 2017-Nov6</span></div><div class="line"><span class="comment">## | GOEGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | DBSCHEMAVERSION: 2.1</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Please see: help('select') for usage information</span></div></pre></td></tr></table></figure>
<p>使用<code>AnnotationDbi</code>包中的<code>keys</code>，<code>columns</code>和<code>select</code>函数也很容易在地id与不同terms之间进行映射，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div></pre></td><td class="code"><pre><div class="line">k5 = keys(GO.db)[<span class="number">1</span>:<span class="number">5</span>]</div><div class="line">cgo = columns(GO.db)</div><div class="line">select(GO.db, keys=k5, columns=cgo[<span class="number">1</span>:<span class="number">3</span>])</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line"><span class="comment">##         GOID</span></div><div class="line"><span class="comment">## 1 GO:0000001</span></div><div class="line"><span class="comment">## 2 GO:0000002</span></div><div class="line"><span class="comment">## 3 GO:0000003</span></div><div class="line"><span class="comment">## 4 GO:0000006</span></div><div class="line"><span class="comment">## 5 GO:0000007</span></div><div class="line"><span class="comment">##                                                                                                                                                                                                                                                                                      DEFINITION</span></div><div class="line"><span class="comment">## 1                                                                                                       The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.</span></div><div class="line"><span class="comment">## 2                                                                                                                                             The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.</span></div><div class="line"><span class="comment">## 3                                                                                                                                                                  The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.</span></div><div class="line"><span class="comment">## 4                                      Enables the transfer of zinc ions (Zn2+) from one side of a membrane to the other, probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations.</span></div><div class="line"><span class="comment">## 5 Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.</span></div><div class="line"><span class="comment">##   ONTOLOGY</span></div><div class="line"><span class="comment">## 1       BP</span></div><div class="line"><span class="comment">## 2       BP</span></div><div class="line"><span class="comment">## 3       BP</span></div><div class="line"><span class="comment">## 4       MF</span></div><div class="line"><span class="comment">## 5       MF</span></div></pre></td></tr></table></figure>
<p>词汇表的图形结构被编码在<code>SQLite</code>数据库的表中。我们可以使用<code>RSQLite</code>接口对此进行查询，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">con = GO_dbconn()</div><div class="line">dbListTables(con)</div><div class="line"><span class="comment">##  [1] "go_bp_offspring" "go_bp_parents"   "go_cc_offspring"</span></div><div class="line"><span class="comment">##  [4] "go_cc_parents"   "go_mf_offspring" "go_mf_parents"  </span></div><div class="line"><span class="comment">##  [7] "go_obsolete"     "go_ontology"     "go_synonym"     </span></div><div class="line"><span class="comment">## [10] "go_term"         "map_counts"      "map_metadata"   </span></div><div class="line"><span class="comment">## [13] "metadata"        "sqlite_stat1"</span></div></pre></td></tr></table></figure>
<p>以下查询提示了一些内部标识符：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term limit 5"</span>)</div><div class="line"><span class="comment">##   _id      go_id                                        term</span></div><div class="line"><span class="comment">## 1  30 GO:0000001                   mitochondrion inheritance</span></div><div class="line"><span class="comment">## 2  32 GO:0000002            mitochondrial genome maintenance</span></div><div class="line"><span class="comment">## 3  33 GO:0000003                                reproduction</span></div><div class="line"><span class="comment">## 4  37 GO:0042254                         ribosome biogenesis</span></div><div class="line"><span class="comment">## 5  38 GO:0044183 protein binding involved in protein folding</span></div></pre></td></tr></table></figure>
<p>我们可以将 <code>mitochondrion inheritance</code> term追溯到父项和祖父母项，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">dbGetQuery(con, <span class="string">"select * from go_bp_parents where _id=30"</span>)</div><div class="line"><span class="comment">##   _id _parent_id relationship_type</span></div><div class="line"><span class="comment">## 1  30      26537              is_a</span></div><div class="line"><span class="comment">## 2  30      26540              is_a</span></div><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term where _id=26616"</span>)</div><div class="line"><span class="comment">##     _id      go_id</span></div><div class="line"><span class="comment">## 1 26616 GO:0048387</span></div><div class="line"><span class="comment">##                                                              term</span></div><div class="line"><span class="comment">## 1 negative regulation of retinoic acid receptor signaling pathway</span></div><div class="line">dbGetQuery(con, <span class="string">"select * from go_bp_parents where _id=26616"</span>)</div><div class="line"><span class="comment">##     _id _parent_id    relationship_type</span></div><div class="line"><span class="comment">## 1 26616       8389                 is_a</span></div><div class="line"><span class="comment">## 2 26616      26614                 is_a</span></div><div class="line"><span class="comment">## 3 26616      26613 negatively_regulates</span></div><div class="line">dbGetQuery(con, <span class="string">"select _id, go_id, term from go_term where _id=5932"</span>)</div><div class="line"><span class="comment">##    _id      go_id                    term</span></div><div class="line"><span class="comment">## 1 5932 GO:0019237 centromeric DNA binding</span></div></pre></td></tr></table></figure>
<p>将 “mitochondrion inheritance” 视为过程“mitochondrion distribution”和 “organelle inheritance”在概念上的精练是有意义的，这两个term在数据库中被为父项。</p>
<p>可以使用 <code>GO_dbschema()</code>来查看整个数据库模式。</p>
<h3 id="kegg">KEGG</h3>
<p>自Bioconductor诞生以来，KEGG的注释就能在Bioconductor中人使用了，但KEGG的数据库使用权限已经进行了更改。当我们使用<code>KEGG.db</code>加载后会出现以下信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">&gt; <span class="keyword">library</span>(KEGG.db)</div><div class="line">KEGG.db contains mappings based on older data because the original</div><div class="line">  resource was removed from the the public domain before the most</div><div class="line">  recent update was produced. This package should now be considered</div><div class="line">  deprecated and future versions of Bioconductor may not have it</div><div class="line">  available.  Users who want more current data are encouraged to look</div><div class="line">  at the KEGGREST or reactome.db packages</div></pre></td></tr></table></figure>
<p>因此我们可以关注KEGGREST这个包，它需要联网。这是一个非常有用的，基于Entrez标识符的工具。现在我们查询一下BRCA2的信息（它的EntrezID为675），如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">library(KEGGREST)</div><div class="line">brca2K = keggGet(&quot;hsa:675&quot;)</div><div class="line">names(brca2K[[1]])</div><div class="line">##  [1] &quot;ENTRY&quot;      &quot;NAME&quot;       &quot;DEFINITION&quot; &quot;ORTHOLOGY&quot;  &quot;ORGANISM&quot;  </div><div class="line">##  [6] &quot;PATHWAY&quot;    &quot;DISEASE&quot;    &quot;BRITE&quot;      &quot;POSITION&quot;   &quot;MOTIF&quot;     </div><div class="line">## [11] &quot;DBLINKS&quot;    &quot;STRUCTURE&quot;  &quot;AASEQ&quot;      &quot;NTSEQ&quot;</div></pre></td></tr></table></figure>
<p>我们也可以通过<code>keggGet</code>函数来获取构成通路模式的基因列表，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">brpat = keggGet(<span class="string">"path:hsa05212"</span>)</div><div class="line">names(brpat[[<span class="number">1</span>]])</div><div class="line"><span class="comment">##  [1] "ENTRY"       "NAME"        "DESCRIPTION" "CLASS"       "PATHWAY_MAP"</span></div><div class="line"><span class="comment">##  [6] "DISEASE"     "DRUG"        "ORGANISM"    "GENE"        "COMPOUND"   </span></div><div class="line"><span class="comment">## [11] "KO_PATHWAY"  "REFERENCE"</span></div><div class="line">brpat[[<span class="number">1</span>]]$GENE[seq(<span class="number">1</span>,<span class="number">132</span>,<span class="number">2</span>)] <span class="comment"># entrez gene ids</span></div><div class="line"><span class="comment">##  [1] "3845"  "5290"  "5293"  "5291"  "5295"  "5296"  "8503"  "9459" </span></div><div class="line"><span class="comment">##  [9] "5879"  "5880"  "5881"  "4790"  "5970"  "207"   "208"   "10000"</span></div><div class="line"><span class="comment">## [17] "1147"  "3551"  "8517"  "572"   "598"   "842"   "369"   "673"  </span></div><div class="line"><span class="comment">## [25] "5894"  "5604"  "5594"  "5595"  "5599"  "5602"  "5601"  "5900" </span></div><div class="line"><span class="comment">## [33] "5898"  "5899"  "10928" "998"   "7039"  "1950"  "1956"  "2064" </span></div><div class="line"><span class="comment">## [41] "2475"  "6198"  "6199"  "3716"  "6774"  "6772"  "7422"  "1029" </span></div><div class="line"><span class="comment">## [49] "1019"  "1021"  "595"   "5925"  "1869"  "1870"  "1871"  "7157" </span></div><div class="line"><span class="comment">## [57] "1026"  "1647"  "4616"  "10912" "581"   "578"   "1643"  "51426"</span></div><div class="line"><span class="comment">## [65] "7040"  "7042"</span></div></pre></td></tr></table></figure>
<p>KEGGREST还有许多值得研究的地方，例如还可以查询BRCA2（人类）关于胰腺癌途径的静态图像，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(png)</div><div class="line"><span class="keyword">library</span>(grid)</div><div class="line">brpng = keggGet(<span class="string">"hsa05212"</span>, <span class="string">"image"</span>)</div><div class="line">grid.raster(brpng)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_annoOverview-getp-1.png" alt="plot of chunk getp">
<p class="caption">plot of chunk getp</p>
</div>
<h3 id="其它本体">其它本体</h3>
<p><code>rols</code>包含有与EMBL-EBI连接的接口 <a href="https://www.ebi.ac.uk/ols/index" target="_blank" rel="external">Ontology Lookup Service</a>.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(rols)</div><div class="line">oo = Ontologies()</div><div class="line">oo</div><div class="line"><span class="comment">## Object of class 'Ontologies' with 198 entries</span></div><div class="line"><span class="comment">##    GENEPIO, MP ... SEPIO, SIBO</span></div><div class="line">oo[[<span class="number">1</span>]]</div><div class="line"><span class="comment">## Ontology: Genomic Epidemiology Ontology (genepio)  </span></div><div class="line"><span class="comment">##   The Genomic Epidemiology Ontology (GenEpiO) covers vocabulary</span></div><div class="line"><span class="comment">##   necessary to identify, document and research foodborne pathogens</span></div><div class="line"><span class="comment">##   and associated outbreaks.</span></div><div class="line"><span class="comment">##    Loaded: 2017-04-10 Updated: 2017-10-20 Version: 2017-04-09 </span></div><div class="line"><span class="comment">##    4351 terms  137 properties  38 individuals</span></div></pre></td></tr></table></figure>
<p>为了控制查询检索中涉及的网络流量，搜索分为几个阶段，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">glis = OlsSearch(<span class="string">"glioblastoma"</span>)</div><div class="line">glis</div><div class="line"><span class="comment">## Object of class 'OlsSearch':</span></div><div class="line"><span class="comment">##   query: glioblastoma </span></div><div class="line"><span class="comment">##   requested: 20 (out of 502)</span></div><div class="line"><span class="comment">##   response(s): 0</span></div><div class="line">res = olsSearch(glis)</div><div class="line">dim(res)</div><div class="line"><span class="comment">## NULL</span></div><div class="line">resdf = as(res, <span class="string">"data.frame"</span>) <span class="comment"># get content</span></div><div class="line">resdf[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]</div><div class="line"><span class="comment">##                                                     id</span></div><div class="line"><span class="comment">## 1 ncit:class:http://purl.obolibrary.org/obo/NCIT_C3058</span></div><div class="line"><span class="comment">## 2     omit:http://purl.obolibrary.org/obo/OMIT_0007102</span></div><div class="line"><span class="comment">## 3    ordo:class:http://www.orpha.net/ORDO/Orphanet_360</span></div><div class="line"><span class="comment">## 4   hp:class:http://purl.obolibrary.org/obo/HP_0100843</span></div><div class="line"><span class="comment">##                                           iri   short_form        label</span></div><div class="line"><span class="comment">## 1   http://purl.obolibrary.org/obo/NCIT_C3058   NCIT_C3058 Glioblastoma</span></div><div class="line"><span class="comment">## 2 http://purl.obolibrary.org/obo/OMIT_0007102 OMIT_0007102 Glioblastoma</span></div><div class="line"><span class="comment">## 3      http://www.orpha.net/ORDO/Orphanet_360 Orphanet_360 Glioblastoma</span></div><div class="line"><span class="comment">## 4   http://purl.obolibrary.org/obo/HP_0100843   HP_0100843 Glioblastoma</span></div><div class="line">resdf[<span class="number">1</span>,<span class="number">5</span>]  <span class="comment"># full description for one instance</span></div><div class="line"><span class="comment">## [[1]]</span></div><div class="line"><span class="comment">## [1] "The most malignant astrocytic tumor (WHO grade IV).  It is composed of poorly differentiated neoplastic astrocytes and it is characterized by the presence of cellular polymorphism, nuclear atypia, brisk mitotic activity, vascular thrombosis, microvascular proliferation and necrosis. It typically affects adults and is preferentially located in the cerebral hemispheres. It may develop from diffuse astrocytoma WHO grade II or anaplastic astrocytoma (secondary glioblastoma, IDH-mutant), but more frequently, it manifests after a short clinical history de novo, without evidence of a less malignant precursor lesion (primary glioblastoma, IDH- wildtype). (Adapted from WHO)"</span></div></pre></td></tr></table></figure>
<p><code>ontologyIndex</code> 包支持导入开放生物本体(OBO, Open Biological Ontologies)格式的数据，并含有用于查询和可视化本体系统高效的工具。</p>
<h3 id="通用基因集管理">通用基因集管理</h3>
<p><code>GSEABase</code> 包有一个用于管理基因集和集合的优秀工具。我们可以从<a href="http://software.broadinstitute.org/gsea/msigdb/search.jsp" target="_blank" rel="external">MSigDb</a>中导入胶质母细胞瘤相关的基因集来说明一下，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GSEABase)</div><div class="line">glioG = getGmt(system.file(<span class="string">"gmt/glioSets.gmt"</span>, package=<span class="string">"ph525x"</span>))</div><div class="line"><span class="comment">## Warning in readLines(con, ...): incomplete final line found on '/</span></div><div class="line"><span class="comment">## Library/Frameworks/R.framework/Versions/3.4/Resources/library/ph525x/gmt/</span></div><div class="line"><span class="comment">## glioSets.gmt'</span></div><div class="line">glioG</div><div class="line"><span class="comment">## GeneSetCollection</span></div><div class="line"><span class="comment">##   names: BALDWIN_PRKCI_TARGETS_UP, BEIER_GLIOMA_STEM_CELL_DN, ..., ZHENG_GLIOBLASTOMA_PLASTICITY_UP (47 total)</span></div><div class="line"><span class="comment">##   unique identifiers: ADA, AQP9, ..., ZFP28 (3671 total)</span></div><div class="line"><span class="comment">##   types in collection:</span></div><div class="line"><span class="comment">##     geneIdType: NullIdentifier (1 total)</span></div><div class="line"><span class="comment">##     collectionType: NullCollection (1 total)</span></div><div class="line">head(geneIds(glioG[[<span class="number">1</span>]]))</div><div class="line"><span class="comment">## [1] "ADA"      "AQP9"     "ATP2B4"   "ATP6V1G1" "CBX6"     "CCDC165"</span></div></pre></td></tr></table></figure>
<h2 id="模式生物的统一自我描述方法">模式生物的统一，自我描述方法</h2>
<p><code>OrganismDb</code>包简化了对注释的访问。还可以针对<code>TxDb</code>和<code>org.[Nn].eg.db</code>进行直接查询，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(Homo.sapiens)</div><div class="line">class(Homo.sapiens)</div><div class="line"><span class="comment">## [1] "OrganismDb"</span></div><div class="line"><span class="comment">## attr(,"package")</span></div><div class="line"><span class="comment">## [1] "OrganismDbi"</span></div><div class="line">Homo.sapiens</div><div class="line"><span class="comment">## OrganismDb Object:</span></div><div class="line"><span class="comment">## # Includes GODb Object:  GO.db </span></div><div class="line"><span class="comment">## # With data about:  Gene Ontology </span></div><div class="line"><span class="comment">## # Includes OrgDb Object:  org.Hs.eg.db </span></div><div class="line"><span class="comment">## # Gene data about:  Homo sapiens </span></div><div class="line"><span class="comment">## # Taxonomy Id:  9606 </span></div><div class="line"><span class="comment">## # Includes TxDb Object:  TxDb.Hsapiens.UCSC.hg19.knownGene </span></div><div class="line"><span class="comment">## # Transcriptome data about:  Homo sapiens </span></div><div class="line"><span class="comment">## # Based on genome:  hg19 </span></div><div class="line"><span class="comment">## # The OrgDb gene id ENTREZID is mapped to the TxDb gene id GENEID .</span></div><div class="line">tx = transcripts(Homo.sapiens)</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line">keytypes(Homo.sapiens)</div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "CDSID"        "CDSNAME"     </span></div><div class="line"><span class="comment">##  [5] "DEFINITION"   "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"</span></div><div class="line"><span class="comment">##  [9] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></div><div class="line"><span class="comment">## [13] "EXONID"       "EXONNAME"     "GENEID"       "GENENAME"    </span></div><div class="line"><span class="comment">## [17] "GO"           "GOALL"        "GOID"         "IPI"         </span></div><div class="line"><span class="comment">## [21] "MAP"          "OMIM"         "ONTOLOGY"     "ONTOLOGYALL" </span></div><div class="line"><span class="comment">## [25] "PATH"         "PFAM"         "PMID"         "PROSITE"     </span></div><div class="line"><span class="comment">## [29] "REFSEQ"       "SYMBOL"       "TERM"         "TXID"        </span></div><div class="line"><span class="comment">## [33] "TXNAME"       "UCSCKG"       "UNIGENE"      "UNIPROT"</span></div><div class="line">columns(Homo.sapiens)</div><div class="line"><span class="comment">##  [1] "ACCNUM"       "ALIAS"        "CDSCHROM"     "CDSEND"      </span></div><div class="line"><span class="comment">##  [5] "CDSID"        "CDSNAME"      "CDSSTART"     "CDSSTRAND"   </span></div><div class="line"><span class="comment">##  [9] "DEFINITION"   "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"</span></div><div class="line"><span class="comment">## [13] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL" </span></div><div class="line"><span class="comment">## [17] "EXONCHROM"    "EXONEND"      "EXONID"       "EXONNAME"    </span></div><div class="line"><span class="comment">## [21] "EXONRANK"     "EXONSTART"    "EXONSTRAND"   "GENEID"      </span></div><div class="line"><span class="comment">## [25] "GENENAME"     "GO"           "GOALL"        "GOID"        </span></div><div class="line"><span class="comment">## [29] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    </span></div><div class="line"><span class="comment">## [33] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        </span></div><div class="line"><span class="comment">## [37] "PROSITE"      "REFSEQ"       "SYMBOL"       "TERM"        </span></div><div class="line"><span class="comment">## [41] "TXCHROM"      "TXEND"        "TXID"         "TXNAME"      </span></div><div class="line"><span class="comment">## [45] "TXSTART"      "TXSTRAND"     "TXTYPE"       "UCSCKG"      </span></div><div class="line"><span class="comment">## [49] "UNIGENE"      "UNIPROT"</span></div></pre></td></tr></table></figure>
<h2 id="面向平台的注释">面向平台的注释</h2>
<p>通过在NCBI GEO的<a href="https://www.ncbi.nlm.nih.gov/geo/browse/?view=platforms&amp;sort=series&amp;display=20&amp;page=892" target="_blank" rel="external">GPL信息页面</a> 上对信息进行排序，我们就可以看到最常用的寡核苷阵列平台（数据库中有4760个系列）就是Affy Human Genome U133 plus 2.0 array (GPL 570)。我们可以使用<code>hgu133plus2.db</code>对这些数据进行注释，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(hgu133plus2.db)</div><div class="line"><span class="comment">## </span></div><div class="line">hgu133plus2.db</div><div class="line"><span class="comment">## ChipDb object:</span></div><div class="line"><span class="comment">## | DBSCHEMAVERSION: 2.1</span></div><div class="line"><span class="comment">## | Db type: ChipDb</span></div><div class="line"><span class="comment">## | Supporting package: AnnotationDbi</span></div><div class="line"><span class="comment">## | DBSCHEMA: HUMANCHIP_DB</span></div><div class="line"><span class="comment">## | ORGANISM: Homo sapiens</span></div><div class="line"><span class="comment">## | SPECIES: Human</span></div><div class="line"><span class="comment">## | MANUFACTURER: Affymetrix</span></div><div class="line"><span class="comment">## | CHIPNAME: Human Genome U133 Plus 2.0 Array</span></div><div class="line"><span class="comment">## | MANUFACTURERURL: http://www.affymetrix.com/support/technical/byproduct.affx?product=hg-u133-plus</span></div><div class="line"><span class="comment">## | EGSOURCEDATE: 2015-Sep27</span></div><div class="line"><span class="comment">## | EGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | CENTRALID: ENTREZID</span></div><div class="line"><span class="comment">## | TAXID: 9606</span></div><div class="line"><span class="comment">## | GOSOURCENAME: Gene Ontology</span></div><div class="line"><span class="comment">## | GOSOURCEURL: ftp://ftp.geneontology.org/pub/go/godatabase/archive/latest-lite/</span></div><div class="line"><span class="comment">## | GOSOURCEDATE: 20150919</span></div><div class="line"><span class="comment">## | GOEGSOURCEDATE: 2015-Sep27</span></div><div class="line"><span class="comment">## | GOEGSOURCENAME: Entrez Gene</span></div><div class="line"><span class="comment">## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></div><div class="line"><span class="comment">## | KEGGSOURCENAME: KEGG GENOME</span></div><div class="line"><span class="comment">## | KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes</span></div><div class="line"><span class="comment">## | KEGGSOURCEDATE: 2011-Mar15</span></div><div class="line"><span class="comment">## | GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)</span></div><div class="line"><span class="comment">## | GPSOURCEURL: ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19</span></div><div class="line"><span class="comment">## | GPSOURCEDATE: 2010-Mar22</span></div><div class="line"><span class="comment">## | ENSOURCEDATE: 2015-Jul16</span></div><div class="line"><span class="comment">## | ENSOURCENAME: Ensembl</span></div><div class="line"><span class="comment">## | ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta</span></div><div class="line"><span class="comment">## | UPSOURCENAME: Uniprot</span></div><div class="line"><span class="comment">## | UPSOURCEURL: http://www.uniprot.org/</span></div><div class="line"><span class="comment">## | UPSOURCEDATE: Thu Oct  1 23:31:58 2015</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Please see: help('select') for usage information</span></div></pre></td></tr></table></figure>
<p>这个资源（以及ChipDb类的所有实例）的基本目的是在探针集(probeset)识别符和更高层次的基因组注释之间进行映射。</p>
<p>有关探针的详细信息（探针集的组成部分）已经由那些后缀为<code>probe</code>的文件包提供，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">library(hgu133plus2probe)</div><div class="line">head(hgu133plus2probe)</div><div class="line">##                    sequence    x   y Probe.Set.Name</div><div class="line">## 1 CACCCAGCTGGTCCTGTGGATGGGA  718 317      1007_s_at</div><div class="line">## 2 GCCCCACTGGACAACACTGATTCCT 1105 483      1007_s_at</div><div class="line">## 3 TGGACCCCACTGGCTGAGAATCTGG  584 901      1007_s_at</div><div class="line">## 4 AAATGTTTCCTTGTGCCTGCTCCTG  192 205      1007_s_at</div><div class="line">## 5 TCCTTGTGCCTGCTCCTGTACTTGT  844 979      1007_s_at</div><div class="line">## 6 TGCCTGCTCCTGTACTTGTCCTCAG  537 971      1007_s_at</div><div class="line">##   Probe.Interrogation.Position Target.Strandedness</div><div class="line">## 1                         3330           Antisense</div><div class="line">## 2                         3443           Antisense</div><div class="line">## 3                         3512           Antisense</div><div class="line">## 4                         3563           Antisense</div><div class="line">## 5                         3570           Antisense</div><div class="line">## 6                         3576           Antisense</div><div class="line">dim(hgu133plus2probe)</div><div class="line">## [1] 604258      6</div></pre></td></tr></table></figure>
<p>将探针集标识符映射到基因水平的信息可以提示一些有意思的歧视，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">select(hgu133plus2.db, keytype=<span class="string">"PROBEID"</span>, </div><div class="line">  columns=c(<span class="string">"SYMBOL"</span>, <span class="string">"GENENAME"</span>, <span class="string">"PATH"</span>, <span class="string">"MAP"</span>), keys=<span class="string">"1007_s_at"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:many mapping between keys and columns</span></div><div class="line"><span class="comment">##     PROBEID  SYMBOL                                    GENENAME PATH</span></div><div class="line"><span class="comment">## 1 1007_s_at    DDR1 discoidin domain receptor tyrosine kinase 1 &lt;NA&gt;</span></div><div class="line"><span class="comment">## 2 1007_s_at MIR4640                               microRNA 4640 &lt;NA&gt;</span></div><div class="line"><span class="comment">##       MAP</span></div><div class="line"><span class="comment">## 1 6p21.33</span></div><div class="line"><span class="comment">## 2 6p21.33</span></div></pre></td></tr></table></figure>
<p>显然，该探针集合可以用于mRNA和miRNA丰度的定量。作为稳定的检查，我们可以看到，不同的符号映射到了相同的细胞带（最后一句不懂，原文为： As a sanity check we see that the distinct symbols map to the same cytoband）。</p>
<h2 id="总结">总结</h2>
<p>我们现在已经拥有了含有从核酸到通路水平的许多数据。通过Bioconductor.org上的View就可以查看现有的一些资源。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_annoOverview.html" target="_blank" rel="external">Genomic annotation in Bioconductor: The general situation</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/09/08/Genomics Data Analysis Series/GDAS008-IRanges和GRanges/" itemprop="url">GDAS008-IRanges和GRanges</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-08T12:00:00+08:00">
                2019-09-08
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  5,178
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  24
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h2 id="前言">前言</h2>
<p>原始的Rmd文档可以参考<a href="https://github.com/genomicsclass/labs/blob/master/biocintro_5x/bioc1_grangeOps.Rmd" target="_blank" rel="external">Github</a>上。</p>
<p><code>IRanges</code> 和 <code>GRanges</code> 对象是Bioconductor基础数据类型的核心成员，其中 <code>IRanges</code> 定义了 <code>integer ranges</code> ，它用于解决基因组上的位置问题，而 <code>GRanges</code> 则包含了染色体和DNA链的信息。这里我们简介绍一下这两个对象，以及如何操作 <code>IRanges</code> 和 <code>GRanges</code> 。</p>
<h2 id="iranges">IRanges</h2>
<p>第一步是加载<code>IRanges</code>包，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">library(IRanges)</div></pre></td></tr></table></figure>
<p><code>IRanges</code> 函数定义了区间范围(interval ranges)。如果你输入2个数据，那么就表示一个区间，例如 [5,10]={5,6,7,8,9,10}，它的宽度是6。另外，如果我们要查看一个 <code>IRanges</code> 定义的区间宽度，我们使用的<code>width()</code>函数，而非 <code>length()</code>函数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div></pre></td><td class="code"><pre><div class="line">ir &lt;- IRanges(<span class="number">5</span>,<span class="number">10</span>)</div><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">start(ir)</div><div class="line"><span class="comment">## [1] 5</span></div><div class="line">end(ir)</div><div class="line"><span class="comment">## [1] 10</span></div><div class="line">width(ir)</div><div class="line"><span class="comment">## [1] 6</span></div><div class="line"><span class="comment"># for detailed information on the IRanges class:</span></div><div class="line"><span class="comment"># ?IRanges</span></div><div class="line">length(ir)</div><div class="line"><span class="comment"># [1] 1</span></div></pre></td></tr></table></figure>
<p>一个单独的<code>IRanges</code>对象可以含有多个范围。现在我们可以在一个向量中指定<code>start</code>，在另外一个向量中指定<code>end</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">IRanges(start=c(<span class="number">3</span>,<span class="number">5</span>,<span class="number">17</span>), end=c(<span class="number">10</span>,<span class="number">8</span>,<span class="number">20</span>))</div><div class="line"><span class="comment">## IRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]        17        20         4</span></div></pre></td></tr></table></figure>
<h3 id="内部范围intra-range操作">内部范围(Intra-range)操作</h3>
<p>现在我们继续使用单个范围[5,10]，我们查看内部范围(intra-range)<code>{…,−2,−1,0,1,2,…}</code>的一个数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># full details on the intra-range methods:</span></div><div class="line"><span class="comment"># ?"intra-range-methods"</span></div><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">shift(ir, -<span class="number">2</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3         8         6</span></div></pre></td></tr></table></figure>
<p>这里我们展示了应用于<code>ir</code>的多个不同操作的结果，结果如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div></pre></td><td class="code"><pre><div class="line">ir</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5        10         6</span></div><div class="line">narrow(ir, start=<span class="number">2</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         6        10         5</span></div><div class="line">narrow(ir, end=<span class="number">5</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5         9         5</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">TRUE</span>, both=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2         4         3</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">FALSE</span>, both=<span class="literal">FALSE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]        11        13         3</span></div><div class="line">flank(ir, width=<span class="number">3</span>, start=<span class="literal">TRUE</span>, both=<span class="literal">TRUE</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2         7         6</span></div><div class="line">ir * <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         6         8         3</span></div><div class="line">ir * -<span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         2        13        12</span></div><div class="line">ir + <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        12        10</span></div><div class="line">ir - <span class="number">2</span></div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         7         8         2</span></div><div class="line">resize(ir, <span class="number">1</span>)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         5         5         1</span></div></pre></td></tr></table></figure>
<p>我们使用下图展示了上面操作的结果。红色柱状部分显示了最初的<code>ir</code>范围。掌握这些操作的最好方法是在自己定义的范围内在控制台中亲自操作一下。</p>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<h3 id="内部范围方法">内部范围方法</h3>
<p>针对<code>intra-range</code>有着一系列的方法。还有一些方法是针对一系列范围的操作，它的输出取决于所有的范围，因此这些方法与<code>intra-range</code>方法不同，后者不会改变输出。现在我们一些案例来说明一下。<code>range()</code>函数提供了从最左侧到最右侧的整数输出，如下所示：</p>
<p>注：原文我不太理解，翻译的不精准，原文如下：</p>
<blockquote>
<p>There are also a set of <em>inter-range</em> methods. These are operations which work on a set of ranges, and the output depends on all the ranges, thus distinguishes these methods from the <em>intra-range</em> methods, for which the other ranges in the set do not change the output. This is best explained with some examples. The <code>range</code> function gives the integer range from the start of the leftmost range to the end of the rightmost range:</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># full details on the inter-range methods:</span></div><div class="line"><span class="comment"># ?"inter-range-methods"</span></div><div class="line">(ir &lt;- IRanges(start=c(<span class="number">3</span>,<span class="number">5</span>,<span class="number">17</span>), end=c(<span class="number">10</span>,<span class="number">8</span>,<span class="number">20</span>)))</div><div class="line"><span class="comment">## IRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]        17        20         4</span></div><div class="line">range(ir)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        20        18</span></div></pre></td></tr></table></figure>
<p><code>reduce</code> 函数能折叠这些范围，以便于整个可以只被输出范围中的一个范围覆盖，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">reduce(ir)</div><div class="line"><span class="comment">## IRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3        10         8</span></div><div class="line"><span class="comment">##   [2]        17        20         4</span></div></pre></td></tr></table></figure>
<p>注：ir原来有3个范围，其中[5,8]范围是在[3,10]之间，因此前者被折叠到了后者之中。</p>
<p><code>gaps</code> 函数能够给出上面两个范围的区间，但是它不覆盖任何<code>ir</code>中的范围，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">gaps(ir)</div><div class="line"><span class="comment">## IRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]        11        16         6</span></div></pre></td></tr></table></figure>
<p><code>disjoin</code> 会打断 <code>ir</code>的范围，并它们分成不同的范围，如下所示：</p>
<blockquote>
<p>disjoin返回一个disjoint对象，它会查看参数中端点的并集，返回不相交的对象。换话句讲，结果中由区间的最大长度范围组成，在该范围上，对象x中的重叠范围集合是相同的，并且至少为1。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">disjoin(ir)</div><div class="line"><span class="comment">## IRanges object with 4 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##           start       end     width</span></div><div class="line"><span class="comment">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3         4         2</span></div><div class="line"><span class="comment">##   [2]         5         8         4</span></div><div class="line"><span class="comment">##   [3]         9        10         2</span></div><div class="line"><span class="comment">##   [4]        17        20         4</span></div></pre></td></tr></table></figure>
<p>现在我们使用示意图来说明一下：</p>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104114940.png">

</div>
<p>第一二行是原始的范围，第四行是经过disjoin处理后的范围。</p>
<h2 id="granges">GRanges</h2>
<p>GRanges对象包含了Iranges，它含有2个方面的重要信息：</p>
<ul>
<li>染色体信息（Bioconductor称为 <code>seqnames</code>）。</li>
<li>DNA链。</li>
</ul>
<p>链的信息可以指定为 <code>+</code> 号或 <code>-</code> ，或者是保留未指定的 <code>*</code> 。正义链的特征则是在编号线上从左到右的生物学方面，而负链特征则是从右到左。就IRanges而言，正链特征是从<code>start</code>到 <code>end</code> ,负链则是从 <code>end</code> 到 <code>start</code> 。这初看起来比较混乱，但这是必需的，因此 <code>width</code> 被定义为 <code>end - start + 1</code>，并且不允许有负的宽度范围。因为DNA有两条链，这两条链具有相反的方向性，因此当我们说DNA时，可以只指明其中的一条链。</p>
<p>通过IRnage、染色体名称和一条链，我们可以确定我们与另外一个研究人员所指的DNA分子具有相同的区域和链，因为我们使用的相同的基因组构建的。GRanges对象中还可以包括其它部分的信息，不过上述的两个信息则是最重要的 。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicRanges)</div></pre></td></tr></table></figure>
<p>现在我们创建一个虚构染色体<code>chrZ</code> 上的两个范围。我们会说这两个范围指的基因组<code>hg19</code>上的范围。因为我们没有将我们的基因组链接到数据库，因此我们可以指定一个在hg19中并不存在的染色体，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">gr &lt;- GRanges(<span class="string">"chrZ"</span>, IRanges(start=c(<span class="number">5</span>,<span class="number">10</span>),end=c(<span class="number">35</span>,<span class="number">45</span>)),</div><div class="line">              strand=<span class="string">"+"</span>, seqlengths=c(chrZ=<span class="number">100L</span>))</div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome</span></div><div class="line">genome(gr) &lt;- <span class="string">"hg19"</span></div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们可以调用 <code>seqnames</code> 和 <code>seqlengths</code> 来查看上面的信息：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div></pre></td><td class="code"><pre><div class="line">seqnames(gr)</div><div class="line"><span class="comment">## factor-Rle of length 2 with 1 run</span></div><div class="line"><span class="comment">##   Lengths:    2</span></div><div class="line"><span class="comment">##   Values : chrZ</span></div><div class="line"><span class="comment">## Levels(1): chrZ</span></div><div class="line">seqlengths(gr)</div><div class="line"><span class="comment">## chrZ </span></div><div class="line"><span class="comment">##  100</span></div></pre></td></tr></table></figure>
<p>我们可以使用 <code>shift</code> 函数（就像在IRanges中使用的一样）来处理GRanges。但是，当我们尝试超过染色体长度之外的范围时，就会出现警告信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">shift(gr, <span class="number">10</span>)</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [15, 45]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [20, 55]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div><div class="line">shift(gr, <span class="number">80</span>)</div><div class="line"><span class="comment">## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequence</span></div><div class="line"><span class="comment">##   chrZ. Note that only ranges located on a non-circular sequence</span></div><div class="line"><span class="comment">##   whose length is not NA can be considered out-of-bound (use</span></div><div class="line"><span class="comment">##   seqlengths() and isCircular() to get the lengths and circularity</span></div><div class="line"><span class="comment">##   flags of the underlying sequences). You can use trim() to trim</span></div><div class="line"><span class="comment">##   these ranges. See ?`trim,GenomicRanges-method` for more</span></div><div class="line"><span class="comment">##   information.</span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ [85, 115]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ [90, 125]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>如果我们使用 <code>trim</code> 处理这个范围，我们就能获取剩下的范围，而不考虑超出染色体长的部分：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line">trim(shift(gr, <span class="number">80</span>))</div><div class="line"><span class="comment">## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 2 out-of-bound ranges located on sequence</span></div><div class="line"><span class="comment">##   chrZ. Note that only ranges located on a non-circular sequence</span></div><div class="line"><span class="comment">##   whose length is not NA can be considered out-of-bound (use</span></div><div class="line"><span class="comment">##   seqlengths() and isCircular() to get the lengths and circularity</span></div><div class="line"><span class="comment">##   flags of the underlying sequences). You can use trim() to trim</span></div><div class="line"><span class="comment">##   these ranges. See ?`trim,GenomicRanges-method` for more</span></div><div class="line"><span class="comment">##   information.</span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ [85, 100]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ [90, 100]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p>我们可以使用 <code>mcols</code> 函数（链表示元数据列）向每个范围添加列信息。需要注意的是，这个操作对IRanges也是可行的。我们还可以通过指定 <code>NULL</code> 来移除列信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">mcols(gr)</div><div class="line"><span class="comment">## DataFrame with 2 rows and 0 columns</span></div><div class="line">mcols(gr)$value &lt;- c(-<span class="number">1</span>,<span class="number">4</span>)</div><div class="line">gr</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 1 metadata column:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand |     value</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      + |        -1</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      + |         4</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div><div class="line">mcols(gr)$value &lt;- <span class="literal">NULL</span></div><div class="line">&gt; gr</div><div class="line"><span class="comment"># GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">#      seqnames    ranges strand</span></div><div class="line"><span class="comment">#         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">#  [1]     chrZ      5-35      +</span></div><div class="line"><span class="comment">#  [2]     chrZ     10-45      +</span></div><div class="line"><span class="comment">#  -------</span></div><div class="line"><span class="comment">#  seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<h3 id="grangeslist">GRangesList</h3>
<p>当我们涉及基因时，我们通过创建一个GRanges列表是很有用的，例如用于表示分组信息（比如每个基因的外显子）。该列表的元素是基因，并且在每个元素中，外显子的范围被定义为GRanges。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div></pre></td><td class="code"><pre><div class="line">gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">11</span>:<span class="number">13</span>,<span class="number">51</span>:<span class="number">53</span>))</div><div class="line">grl &lt;- GRangesList(gr, gr2)</div><div class="line">grl</div><div class="line"><span class="comment">## GRangesList object of length 2:</span></div><div class="line"><span class="comment">## [[1]] </span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## [[2]] </span></div><div class="line"><span class="comment">## GRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames   ranges strand</span></div><div class="line"><span class="comment">##   [1]     chrZ [11, 51]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ [12, 52]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ [13, 53]      *</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## -------</span></div><div class="line"><span class="comment">## seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p><code>GRangesList</code> 对象中的长度就是其中的<code>GRanges</code> 对象个数。为了获取每个GRanges的长度，我们可以调用 <code>elementNROWS</code>。我们可以通过两个方括号的典型列表索引方法来建立索引，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div></pre></td><td class="code"><pre><div class="line">length(grl)</div><div class="line"><span class="comment">## [1] 2</span></div><div class="line">elementNROWS(grl)</div><div class="line"><span class="comment">## [1] 2 3</span></div><div class="line">grl[[<span class="number">1</span>]]</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from hg19 genome</span></div></pre></td></tr></table></figure>
<p><code>width</code>函数会返回一个 <code>integerList</code>。如果使用<code>sum</code>，我们会得到列表中每个GRanges对象的宽度向量，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">width(grl)</div><div class="line"><span class="comment">## IntegerList of length 2</span></div><div class="line"><span class="comment">## [[1]] 31 36</span></div><div class="line"><span class="comment">## [[2]] 41 41 41</span></div><div class="line">sum(width(grl))</div><div class="line"><span class="comment">## [1]  67 123</span></div></pre></td></tr></table></figure>
<p>我们可以像以前那样添加元数据(metadata)，现在每个GRange对象都有一行元数据。当我们输出GRangesList时它们并不业显示出来，但我们可以通过 <code>mcols</code> 进行访问，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div></pre></td><td class="code"><pre><div class="line">mcols(grl)$value &lt;- c(<span class="number">5</span>,<span class="number">7</span>)</div><div class="line">grl</div><div class="line"><span class="comment">## GRangesList object of length 2:</span></div><div class="line"><span class="comment">## [[1]] </span></div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 5, 35]      +</span></div><div class="line"><span class="comment">##   [2]     chrZ  [10, 45]      +</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## [[2]] </span></div><div class="line"><span class="comment">## GRanges object with 3 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames   ranges strand</span></div><div class="line"><span class="comment">##   [1]     chrZ [11, 51]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ [12, 52]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ [13, 53]      *</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## -------</span></div><div class="line"><span class="comment">## seqinfo: 1 sequence from hg19 genome</span></div><div class="line">mcols(grl)</div><div class="line"><span class="comment">## DataFrame with 2 rows and 1 column</span></div><div class="line"><span class="comment">##       value</span></div><div class="line"><span class="comment">##   &lt;numeric&gt;</span></div><div class="line"><span class="comment">## 1         5</span></div><div class="line"><span class="comment">## 2         7</span></div></pre></td></tr></table></figure>
<h3 id="findoverlaps与over">findOverlaps与%over%</h3>
<p>现在我们来看两个比较<code>GRanges</code>对象的常用方法，首先我们要创建两组范围如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div></pre></td><td class="code"><pre><div class="line">(gr1 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(c(<span class="number">1</span>,<span class="number">11</span>,<span class="number">21</span>,<span class="number">31</span>,<span class="number">41</span>),width=<span class="number">5</span>),strand=<span class="string">"*"</span>))</div><div class="line"><span class="comment">## GRanges object with 5 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [ 1,  5]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [11, 15]      *</span></div><div class="line"><span class="comment">##   [3]     chrZ  [21, 25]      *</span></div><div class="line"><span class="comment">##   [4]     chrZ  [31, 35]      *</span></div><div class="line"><span class="comment">##   [5]     chrZ  [41, 45]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div><div class="line">(gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(c(<span class="number">19</span>,<span class="number">33</span>),c(<span class="number">38</span>,<span class="number">35</span>)),strand=<span class="string">"*"</span>))</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [19, 38]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [33, 35]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p><code>findOverlaps</code> 会返回一个 <code>Hits</code> 对象，这个对象含有关于检索范围的信息（第一个参数），这个范围与主题（第二个参数）有哪些地方重叠。还有许多参数用于指定计算哪种类型的重叠，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">fo &lt;- findOverlaps(gr1, gr2)</div><div class="line">fo</div><div class="line"><span class="comment">## Hits object with 3 hits and 0 metadata columns:</span></div><div class="line"><span class="comment">##       queryHits subjectHits</span></div><div class="line"><span class="comment">##       &lt;integer&gt;   &lt;integer&gt;</span></div><div class="line"><span class="comment">##   [1]         3           1</span></div><div class="line"><span class="comment">##   [2]         4           1</span></div><div class="line"><span class="comment">##   [3]         4           2</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   queryLength: 5 / subjectLength: 2</span></div><div class="line">queryHits(fo)</div><div class="line"><span class="comment">## [1] 3 4 4</span></div><div class="line">subjectHits(fo)</div><div class="line"><span class="comment">## [1] 1 1 2</span></div></pre></td></tr></table></figure>
<p>得到重叠信息的另外一种方式就是使用 <code>%over%</code> ，它返回一个逻辑向量，这个向量表示了第一个参数中的范围与第二个参数中任何范围重叠的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">gr1 %over% gr2</div><div class="line"><span class="comment">## [1] FALSE FALSE  TRUE  TRUE FALSE</span></div><div class="line">gr1[gr1 %over% gr2]</div><div class="line"><span class="comment">## GRanges object with 2 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames    ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]     chrZ  [21, 25]      *</span></div><div class="line"><span class="comment">##   [2]     chrZ  [31, 35]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div></pre></td></tr></table></figure>
<p>需要注意的是，这两种操作方式都是链特异性的，虽然 <code>findOverlaps</code> 有一个 <code>ignore.strand</code> 选择参数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">gr1 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">1</span>,<span class="number">10</span>),strand=<span class="string">"+"</span>)</div><div class="line">gr2 &lt;- GRanges(<span class="string">"chrZ"</span>,IRanges(<span class="number">1</span>,<span class="number">10</span>),strand=<span class="string">"-"</span>)</div><div class="line">gr1 %over% gr2</div><div class="line"><span class="comment">## [1] FALSE</span></div></pre></td></tr></table></figure>
<h3 id="rle-和views">Rle 和Views</h3>
<p>最后，我们来简单了解一下由IRanges定义的两个相关类，即Rle类和Views类。Rle全称为run-length encoding，它表示的是重复数据的压缩工，用于代替类似于[1, 1, 1, 1]这样的储存。</p>
<p>我们只用储存数据1，以及重复次数4即可。数据越重复，使用Rle对象的压缩效果就越明显。</p>
<p>我们使用 <code>str</code> 来查看 Rle的内部结构，用于显示它的储存数据和重复次数，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">(r &lt;- Rle(c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">0</span>,<span class="number">0</span>,-<span class="number">2</span>,-<span class="number">2</span>,-<span class="number">2</span>,rep(-<span class="number">1</span>,<span class="number">20</span>))))</div><div class="line"><span class="comment">## numeric-Rle of length 28 with 4 runs</span></div><div class="line"><span class="comment">##   Lengths:  3  2  3 20</span></div><div class="line"><span class="comment">##   Values :  1  0 -2 -1</span></div><div class="line">str(r)</div><div class="line"><span class="comment">## Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></div><div class="line"><span class="comment">##   ..@ values         : num [1:4] 1 0 -2 -1</span></div><div class="line"><span class="comment">##   ..@ lengths        : int [1:4] 3 2 3 20</span></div><div class="line"><span class="comment">##   ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   ..@ metadata       : list()</span></div><div class="line">as.numeric(r)</div><div class="line"><span class="comment">##  [1]  1  1  1  0  0 -2 -2 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1</span></div><div class="line"><span class="comment">## [24] -1 -1 -1 -1 -1</span></div></pre></td></tr></table></figure>
<p>Views对象可以被为一个“窗口”用于查看一个序列信息，如下所示：</p>
<p>A <em>Views</em> object can be thought of as “windows” looking into a sequence.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div></pre></td><td class="code"><pre><div class="line">(v &lt;- Views(r, start=c(<span class="number">4</span>,<span class="number">2</span>), end=c(<span class="number">7</span>,<span class="number">6</span>)))</div><div class="line"><span class="comment">## Views on a 28-length Rle subject</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## views:</span></div><div class="line"><span class="comment">##     start end width</span></div><div class="line"><span class="comment">## [1]     4   7     4 [ 0  0 -2 -2]</span></div><div class="line"><span class="comment">## [2]     2   6     5 [ 1  1  0  0 -2]</span></div></pre></td></tr></table></figure>
<p>需要注意的是，Veiws对象的内部结构只是原始对象，以及指定窗口的IRangs。使用Views的最好好处就是，当原始对象没有储存在内存中时，在这种情况下，View对象就是一个轻量级的类，它能帮助我们引用子序列，而不必将整个序列都加载到内存中，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">str(v)</div><div class="line"><span class="comment">## Formal class 'RleViews' [package "IRanges"] with 5 slots</span></div><div class="line"><span class="comment">##   ..@ subject        :Formal class 'Rle' [package "S4Vectors"] with 4 slots</span></div><div class="line"><span class="comment">##   .. .. ..@ values         : num [1:4] 1 0 -2 -1</span></div><div class="line"><span class="comment">##   .. .. ..@ lengths        : int [1:4] 3 2 3 20</span></div><div class="line"><span class="comment">##   .. .. ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ metadata       : list()</span></div><div class="line"><span class="comment">##   ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots</span></div><div class="line"><span class="comment">##   .. .. ..@ start          : int [1:2] 4 2</span></div><div class="line"><span class="comment">##   .. .. ..@ width          : int [1:2] 4 5</span></div><div class="line"><span class="comment">##   .. .. ..@ NAMES          : NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ elementType    : chr "integer"</span></div><div class="line"><span class="comment">##   .. .. ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   .. .. ..@ metadata       : list()</span></div><div class="line"><span class="comment">##   ..@ elementType    : chr "ANY"</span></div><div class="line"><span class="comment">##   ..@ elementMetadata: NULL</span></div><div class="line"><span class="comment">##   ..@ metadata       : list()</span></div></pre></td></tr></table></figure>
<h2 id="基因组元件的应用链特异性操作">基因组元件的应用：链特异性操作</h2>
<p>在这一部分，我们使用一个小型的范围来说明一下基因的范围内(intra-range)操作，包括reduce, disjoin与gaps。我们添加链和seqname信息，并显示如想调整大小和侧翼(flank)用于识别TSS和启动子区域，如下所示：</p>
<h3 id="一组简单的范围">一组简单的范围</h3>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">ir &lt;- IRanges(c(3, 8, 14, 15, 19, 34, 40),</div><div class="line">  width = c(12, 6, 6, 15, 6, 2, 7))</div></pre></td></tr></table></figure>
<p>现在我们使用图片形式查看一下<code>ir</code>，以及内部范围(intra-range)操作，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(4,1), mar=c(4,2,2,2))</div><div class="line">plotRanges(ir, xlim=c(0,60))</div><div class="line">plotRanges(reduce(ir), xlim=c(0,60))</div><div class="line">plotRanges(disjoin(ir), xlim=c(0,60))</div><div class="line">plotRanges(gaps(ir), xlim=c(0,60))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-lkir-1.png" alt="plot of chunk lkir">
<p class="caption">plot of chunk lkir</p>
</div>
<p>注：<code>plotRange()</code>函数似乎不在这个包中，网上找到了它的原代码，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line">plotRanges &lt;- <span class="keyword">function</span>(x, xlim = x, main = deparse(substitute(x)),</div><div class="line">                       col = <span class="string">"black"</span>, sep = <span class="number">0.5</span>, <span class="keyword">...</span>)</div><div class="line">&#123;</div><div class="line">  height &lt;- <span class="number">1</span></div><div class="line">  <span class="keyword">if</span> (is(xlim, <span class="string">"Ranges"</span>))</div><div class="line">    xlim &lt;- c(min(start(xlim)), max(end(xlim)))</div><div class="line">  bins &lt;- disjointBins(IRanges(start(x), end(x) + <span class="number">1</span>))</div><div class="line">  plot.new()</div><div class="line">  plot.window(xlim, c(<span class="number">0</span>, max(bins)*(height + sep)))</div><div class="line">  ybottom &lt;- bins * (sep + height) - height</div><div class="line">  rect(start(x)-<span class="number">0.5</span>, ybottom, end(x)+<span class="number">0.5</span>, ybottom + height, col = col, <span class="keyword">...</span>)</div><div class="line">  title(main)</div><div class="line">  axis(<span class="number">1</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p><code>reduce(x)</code>函数生成一组不重叠的范围，它覆盖了x中的所有位点。这种操作用于生了你许多转录本的基因模式的复杂性，其中我们可能只想要转录出已知区间的地址，不考虑transcript of residence。</p>
<p><code>disjoin(x)</code>生成一组覆盖x的所有位置的范围，使得分离输出中的任何范围都不会与x中的任何间隔的端点重叠。这为我们提供了连续间隔最大可能的集合，这些连续间隔在原始间隔集中具有端点的地方被分割开。</p>
<p><code>gaps(x)</code>产生一组覆盖[start(X)，end(X)]中未被x中任何范围覆盖的位置的范围。它能给出编码序列位置和外显子间隔，这可用于枚举内含子。</p>
<h3 id="granges的扩展">GRanges的扩展</h3>
<p>现在我们添加上染色体与链信息，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">library(GenomicRanges)</div><div class="line">gir = GRanges(seqnames=&quot;chr1&quot;, ir, strand=c(rep(&quot;+&quot;, 4), rep(&quot;-&quot;,3)))</div></pre></td></tr></table></figure>
<p>让我们假设间隔代表基因。下图说明了转录起始位点(绿色)、上游启动子区域(紫色)、下游启动子区域(棕色)的位置，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(4,1), mar=c(4,2,2,2))</div><div class="line">plotGRanges(gir, xlim=c(0,60))</div><div class="line"># </div><div class="line">plotGRanges(resize(gir,1), xlim=c(0,60),col=&quot;green&quot;)</div><div class="line">plotGRanges(flank(gir,3), xlim=c(0,60), col=&quot;purple&quot;)</div><div class="line">plotGRanges(flank(gir,2,start=FALSE), xlim=c(0,60), col=&quot;brown&quot;)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_igranges-dopr-1.png" alt="plot of chunk dopr">
<p class="caption">plot of chunk dopr</p>
</div>
<p>注：<code>plotGRanges</code>的源代码如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">plotGRanges = <span class="keyword">function</span> (x, xlim = x, col = <span class="string">"black"</span>, sep = <span class="number">0.5</span>, xlimits = c(<span class="number">0</span>, </div><div class="line">    <span class="number">60</span>), <span class="keyword">...</span>) </div><div class="line">&#123;</div><div class="line">    main = deparse(substitute(x))</div><div class="line">    ch = as.character(seqnames(x)[<span class="number">1</span>])</div><div class="line">    x = ranges(x)</div><div class="line">    height &lt;- <span class="number">1</span></div><div class="line">    <span class="keyword">if</span> (is(xlim, <span class="string">"Ranges"</span>)) </div><div class="line">        xlim &lt;- c(min(start(xlim)), max(end(xlim)))</div><div class="line">    bins &lt;- disjointBins(IRanges(start(x), end(x) + <span class="number">1</span>))</div><div class="line">    plot.new()</div><div class="line">    plot.window(xlim = xlimits, c(<span class="number">0</span>, max(bins) * (height + sep)))</div><div class="line">    ybottom &lt;- bins * (sep + height) - height</div><div class="line">    rect(start(x) - <span class="number">0.5</span>, ybottom, end(x) + <span class="number">0.5</span>, ybottom + height, </div><div class="line">        col = col, <span class="keyword">...</span>)</div><div class="line">    title(main, xlab = ch)</div><div class="line">    axis(<span class="number">1</span>)</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>请注意，我们不需要采取特殊步骤来处理链中的差异</p>
<h2 id="甲基化微阵列数据的可视化">甲基化微阵列数据的可视化</h2>
<p>在我们讨论 <a href="http://genomicsclass.github.io/book/pages/dataman2017.html#methy" target="_blank" rel="external">SummarizedExperiment applications</a>时，我们输入了由Illumina 450k甲基化微阵列生成的数据。</p>
<p>在这一部分中，我们将介绍如何使用GenomicRanges和Gviz来研究一下在基因水平注释下的甲基化模型。这个思路非常简单：只需从所有样本中提取M值（甲基化与总DNA 比值的基因座特定的估计的log转换），并在选定基因的基因模型中绘制出来，如下所示：</p>
<p>我们回顾一下数据是如何获取和导入的：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ArrayExpress)</div><div class="line"><span class="keyword">if</span> (!file.exists(<span class="string">"E-MTAB-5797.sdrf.txt"</span>)) nano = getAE(<span class="string">"E-MTAB-5797"</span>)</div><div class="line"><span class="keyword">library</span>(minfi)</div><div class="line">pref = unique(substr(dir(patt=<span class="string">"idat"</span>),<span class="number">1</span>,<span class="number">17</span>)) <span class="comment"># find the prefix strings</span></div><div class="line">raw = read.metharray(pref)</div><div class="line">glioMeth = preprocessQuantile(raw) <span class="comment"># generate SummarizedExperiment</span></div><div class="line"><span class="comment">## [preprocessQuantile] Mapping to genome.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Fixing outliers.</span></div><div class="line"><span class="comment">## [preprocessQuantile] Quantile normalizing.</span></div></pre></td></tr></table></figure>
<p>这些步骤需要网联，并且需要花几分钟。</p>
<p>一旦我们的会话中有 <code>glioMeth</code> ，需要将以下代码添加到会话中。我们将会介绍它是如何工作的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line">MbyGene = <span class="keyword">function</span>(mset, symbol=<span class="string">"TP53"</span>, rad=<span class="number">5000</span>) &#123;</div><div class="line"><span class="comment"># phase 1: annotated GRanges for the gene</span></div><div class="line"> <span class="keyword">require</span>(erma)</div><div class="line"> <span class="keyword">require</span>(Gviz)</div><div class="line"> gmod = suppressMessages(genemodel(symbol))     <span class="comment"># erma utility</span></div><div class="line"> gseq = as.character(seqnames(gmod)[<span class="number">1</span>])</div><div class="line"> gmod$transcript = symbol</div><div class="line"><span class="comment"># phase 2: filter down to the region of interest</span></div><div class="line"> mlim = mset[which(seqnames(mset)==gseq),] <span class="comment"># restrict to chromosome</span></div><div class="line"> <span class="comment"># now focus the methylation data to vicinity of gene</span></div><div class="line"> d1 = subsetByOverlaps(GRanges(rowRanges(mlim),,, getM(mlim)), </div><div class="line">         range(gmod)+rad)</div><div class="line"><span class="comment"># phase 3: use Gviz</span></div><div class="line"> plotTracks(list(DataTrack(d1), </div><div class="line">              GeneRegionTrack(gmod, </div><div class="line">               transcriptAnnotation=<span class="string">"transcript"</span>, name=gseq), </div><div class="line">              GenomeAxisTrack(name=gseq, showTitle=<span class="literal">TRUE</span>)))</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>代码表明了三个阶段：获取基因区域，并添加转录本(<code>transcript</code>)注释用于绘制所有外显子的结合；将<code>GenomicRatioSet</code>（继承自</p>
<p>对代码的注释表明了三个阶段：获取基因区并添加用于所有外显子联合的信息性绘图的转录注释；将GenomicRatioSet（继承自RangedSummarizedExperiment）缩小到感兴趣的间隔，这个间隔是由基因模式和rad参数确定；使用Gviz来构建可用于绘图的对象，并绘制出来。</p>
<p>Gviz的详细信息在该包的文档中。现在我们绘制出以基因为中心的图形，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">MbyGene(glioMeth, symbol=<span class="string">"TERT"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="https://pic-1256416512.cos.ap-chengdu.myqcloud.com/img/20191104142451.png">

</div>
<p>从上图中我们可以发现这些信息：</p>
<ul>
<li>基因组背景（以兆(megabase)为单位的染色体和区域）。</li>
<li>感兴趣的基因在单行中的结构，它们表示了表达间隔的联合。</li>
<li>450k针近的位置（蓝色数据点的x坐标）。</li>
<li>甲基化估计中的样本间变异。</li>
<li>所选基因组区域的甲基化变异。</li>
</ul>
<p>在6x模块中，我们将学习如何使用额外的软件包来创建这种类型的交互式展示，允许我们选择基因并随意放大感兴趣的区域。</p>
<h2 id="脚注">脚注</h2>
<p>更多的信息可以查看 <code>GenomicRanges</code> 包，以及查阅PLOS Comp Bio上的文献，它是GenomicRanges的作者写的：</p>
<p>http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118</p>
<p>使用vignettes也能查看这个包的众多功能文档，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">browseVignettes(&quot;GenomicRanges&quot;)</div></pre></td></tr></table></figure>
<p>或者是使用help函数：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">help(package=&quot;GenomicRanges&quot;, help_type=&quot;html&quot;)</div></pre></td></tr></table></figure>
<p>对于Bed Tools的用户来说，<code>HelloRanges</code>包对于在BEd和GRanges概念框架之间转换概念非常有用。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="https://github.com/genomicsclass/labs/blob/master/biocintro_5x/bioc1_grangeOps.Rmd" target="_blank" rel="external">bioc1_grangeOps.Rmd</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_igranges.html" target="_blank" rel="external">IRanges and GRanges</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <a class="extend prev" rel="prev" href="/"><i class="fa fa-angle-left"></i></a><a class="page-number" href="/">1</a><span class="page-number current">2</span><a class="page-number" href="/page/3/">3</a><span class="space">&hellip;</span><a class="page-number" href="/page/25/">25</a><a class="extend next" rel="next" href="/page/3/"><i class="fa fa-angle-right"></i></a>
  </nav>



          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">249</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">118</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共988.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
