<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="Bioconductor," />










<meta name="description" content="前言 在数据管理部分，我们应用summarizeOverlaps 来管理RNAseqData.HNRNPC.bam.chr14中的BAM文件。现在我们再次来操作一下。 使用SummarizedExperiment来管理BAM文件 123456789101112131415161718library(RNAseqData.HNRNPC.bam.chr14)bfp = RNAseqData.HN">
<meta name="keywords" content="Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS006-深入理解SummarizedExperiment类">
<meta property="og:url" content="http://rvdsd.top/2019/09/06/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="前言 在数据管理部分，我们应用summarizeOverlaps 来管理RNAseqData.HNRNPC.bam.chr14中的BAM文件。现在我们再次来操作一下。 使用SummarizedExperiment来管理BAM文件 123456789101112131415161718library(RNAseqData.HNRNPC.bam.chr14)bfp = RNAseqData.HN">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkbo-1.png">
<meta property="og:updated_time" content="2019-11-13T08:50:40.066Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS006-深入理解SummarizedExperiment类">
<meta name="twitter:description" content="前言 在数据管理部分，我们应用summarizeOverlaps 来管理RNAseqData.HNRNPC.bam.chr14中的BAM文件。现在我们再次来操作一下。 使用SummarizedExperiment来管理BAM文件 123456789101112131415161718library(RNAseqData.HNRNPC.bam.chr14)bfp = RNAseqData.HN">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/06/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/"/>





  <title>GDAS006-深入理解SummarizedExperiment类 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/06/Genomics Data Analysis Series/GDAS006-深入理解SummarizedExperiment类/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS006-深入理解SummarizedExperiment类</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-06T12:00:00+08:00">
                2019-09-06
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,289
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  6
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="前言">前言</h1>
<p>在数据管理部分，我们应用<code>summarizeOverlaps</code> 来管理<code>RNAseqData.HNRNPC.bam.chr14</code>中的BAM文件。现在我们再次来操作一下。</p>
<h2 id="使用summarizedexperiment来管理bam文件">使用SummarizedExperiment来管理BAM文件</h2>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(RNAseqData.HNRNPC.bam.chr14)</div><div class="line">bfp = RNAseqData.HNRNPC.bam.chr14_BAMFILES</div><div class="line"><span class="keyword">library</span>(Rsamtools)</div><div class="line">bfl = BamFileList(file=bfp)</div><div class="line">hnrnpcLoc = GRanges(<span class="string">"chr14"</span>, IRanges(<span class="number">21677296</span>, <span class="number">21737638</span>))</div><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line"><span class="keyword">library</span>(BiocParallel)</div><div class="line">register(SerialParam())</div><div class="line">hnse = summarizeOverlaps(hnrnpcLoc,bfl)</div><div class="line">hnse</div><div class="line"><span class="comment">## class: RangedSummarizedExperiment </span></div><div class="line"><span class="comment">## dim: 1 8 </span></div><div class="line"><span class="comment">## metadata(0):</span></div><div class="line"><span class="comment">## assays(1): counts</span></div><div class="line"><span class="comment">## rownames: NULL</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(8): ERR127306 ERR127307 ... ERR127304 ERR127305</span></div><div class="line"><span class="comment">## colData names(0):</span></div></pre></td></tr></table></figure>
<p><code>hnse</code> 是<code>RangedSummarizedExperiment</code> 类的一实例。这个类就类似于 <code>ExpressionSet</code> ，不过有着更多的内容来管理元数据，它的流程如下所示：</p>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkseee-1.png" alt="plot of chunk lkseee">
<p class="caption">plot of chunk lkseee</p>
</div>
<p>有效地使用<code>SummarizedExperiment</code>实例涉及学习它的一些方法。为了获取HNRNPC基因的读长/区域(read/region)，我们可以使用<code>assay</code>方法，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">assay(hnse)</div><div class="line"><span class="comment">##      ERR127306 ERR127307 ERR127308 ERR127309 ERR127302 ERR127303 ERR127304</span></div><div class="line"><span class="comment">## [1,]      5422      6320      5896      5558       172       196       316</span></div><div class="line"><span class="comment">##      ERR127305</span></div><div class="line"><span class="comment">## [1,]       282</span></div></pre></td></tr></table></figure>
<p>以上就是最基本的结果表示方法。列名则是样本标识符，但是有关区域检查的一些信息则已经丢失。</p>
<h2 id="summarizedexperiment中的元数据">SummarizedExperiment中的元数据</h2>
<p><code>hnse</code> 对象还有一些其它的信息，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">rowRanges(hnse)</div><div class="line"><span class="comment">## GRanges object with 1 range and 0 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames               ranges strand</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">##   [1]    chr14 [21677296, 21737638]      *</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></div><div class="line">seqinfo(hnse)</div><div class="line"><span class="comment">## Seqinfo object with 1 sequence from an unspecified genome; no seqlengths:</span></div><div class="line"><span class="comment">##   seqnames seqlengths isCircular genome</span></div><div class="line"><span class="comment">##   chr14            NA         NA   &lt;NA&gt;</span></div><div class="line">metadata(hnse)</div><div class="line"><span class="comment">## list()</span></div></pre></td></tr></table></figure>
<p>我们还可以进一步分析差异信息，从而输出更多，更广泛的信息。</p>
<h1 id="通过添加输入信息来更有效地生成summarizedexperiment">通过添加输入信息来更有效地生成SummarizedExperiment</h1>
<h2 id="使用元数据定义感兴趣的区域">使用元数据定义感兴趣的区域</h2>
<p>We have seen that it is sufficient to define a single GRanges to drive <code>summarizeOverlaps</code> over a set of BAM files. We’d like to preserve more metadata about the regions examined. We’ll use the TxDb infrastructure, to be described in more detail later, to get a structure defining gene regions on chr14. We’ll also use the Homo.sapiens annotation package to add gene symbols.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</div><div class="line">txdb = TxDb.Hsapiens.UCSC.hg19.knownGene</div><div class="line">gr14 = genes(txdb, vals=list(tx_chrom=<span class="string">"chr14"</span>))</div><div class="line"><span class="comment">## Warning: 'elementLengths' is deprecated.</span></div><div class="line"><span class="comment">## Use 'elementNROWS' instead.</span></div><div class="line"><span class="comment">## See help("Deprecated")</span></div><div class="line"></div><div class="line"><span class="comment">## Warning: 'elementLengths' is deprecated.</span></div><div class="line"><span class="comment">## Use 'elementNROWS' instead.</span></div><div class="line"><span class="comment">## See help("Deprecated")</span></div><div class="line">gr14$symbol = mapIds(Homo.sapiens, keys=gr14$gene_id, keytype=<span class="string">"ENTREZID"</span>,</div><div class="line">   column=<span class="string">"SYMBOL"</span>)</div><div class="line"><span class="comment">## 'select()' returned 1:1 mapping between keys and columns</span></div><div class="line">gr14</div><div class="line"><span class="comment">## GRanges object with 781 ranges and 2 metadata columns:</span></div><div class="line"><span class="comment">##             seqnames                 ranges strand |     gene_id</span></div><div class="line"><span class="comment">##                &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></div><div class="line"><span class="comment">##       10001    chr14 [ 71050957,  71067384]      - |       10001</span></div><div class="line"><span class="comment">##   100113389    chr14 [ 45580078,  45580176]      + |   100113389</span></div><div class="line"><span class="comment">##   100113391    chr14 [ 20794600,  20794698]      - |   100113391</span></div><div class="line"><span class="comment">##   100124539    chr14 [ 91592770,  91592896]      + |   100124539</span></div><div class="line"><span class="comment">##   100126297    chr14 [101507700, 101507781]      + |   100126297</span></div><div class="line"><span class="comment">##         ...      ...                    ...    ... .         ...</span></div><div class="line"><span class="comment">##        9870    chr14 [ 75127955,  75179807]      - |        9870</span></div><div class="line"><span class="comment">##        9878    chr14 [ 21945335,  21967319]      + |        9878</span></div><div class="line"><span class="comment">##        9895    chr14 [102829300, 102968818]      + |        9895</span></div><div class="line"><span class="comment">##        9950    chr14 [ 93260650,  93306304]      + |        9950</span></div><div class="line"><span class="comment">##        9985    chr14 [ 24641234,  24649463]      + |        9985</span></div><div class="line"><span class="comment">##                  symbol</span></div><div class="line"><span class="comment">##             &lt;character&gt;</span></div><div class="line"><span class="comment">##       10001        MED6</span></div><div class="line"><span class="comment">##   100113389    SNORD127</span></div><div class="line"><span class="comment">##   100113391    SNORD126</span></div><div class="line"><span class="comment">##   100124539    SNORA11B</span></div><div class="line"><span class="comment">##   100126297      MIR300</span></div><div class="line"><span class="comment">##         ...         ...</span></div><div class="line"><span class="comment">##        9870       AREL1</span></div><div class="line"><span class="comment">##        9878        TOX4</span></div><div class="line"><span class="comment">##        9895      TECPR2</span></div><div class="line"><span class="comment">##        9950      GOLGA5</span></div><div class="line"><span class="comment">##        9985        REC8</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 93 sequences (1 circular) from hg19 genome</span></div></pre></td></tr></table></figure>
<h2 id="定义bam文件的样本信息">定义BAM文件的样本信息</h2>
<p>现在我们有三个样本，一个用于控制，两个用于敲低。我们使用<code>GenomicFiles</code>来绑定样本信息的元数据，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">char = rep(c(<span class="string">"hela_wt"</span>, <span class="string">"hela_hkd"</span>), each=<span class="number">4</span>)</div><div class="line">bff = GenomicFiles(files=path(bfl))</div><div class="line">colData(bff)$condition = char</div><div class="line">sid = c(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>,<span class="number">3</span>,<span class="number">3</span>)</div><div class="line">bff$sample = sid</div><div class="line">bff</div><div class="line"><span class="comment">## GenomicFiles object with 0 ranges and 8 files: </span></div><div class="line"><span class="comment">## files: ERR127306_chr14.bam, ERR127307_chr14.bam, ..., ERR127304_chr14.bam, ERR127305_chr14.bam </span></div><div class="line"><span class="comment">## detail: use files(), rowRanges(), colData(), ...</span></div></pre></td></tr></table></figure>
<h2 id="比较读长覆盖区保留元数据">比较读长覆盖区，保留元数据</h2>
<p>我们来查看5个基因，其中就包括HNRNPC。当计算后，我们会将样本信息再绑定回结果，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div></pre></td><td class="code"><pre><div class="line">hnse = summarizeOverlaps(gr14[c(<span class="number">1</span>:<span class="number">4</span>,<span class="number">305</span>)],files(bff))</div><div class="line">colData(hnse) = cbind(colData(hnse), colData(bff))</div><div class="line">hnse</div><div class="line"><span class="comment">## class: RangedSummarizedExperiment </span></div><div class="line"><span class="comment">## dim: 5 8 </span></div><div class="line"><span class="comment">## metadata(0):</span></div><div class="line"><span class="comment">## assays(1): counts</span></div><div class="line"><span class="comment">## rownames(5): 10001 100113389 100113391 100124539 3183</span></div><div class="line"><span class="comment">## rowData names(2): gene_id symbol</span></div><div class="line"><span class="comment">## colnames(8): ERR127306 ERR127307 ... ERR127304 ERR127305</span></div><div class="line"><span class="comment">## colData names(2): condition sample</span></div><div class="line">assay(hnse)</div><div class="line"><span class="comment">##           ERR127306 ERR127307 ERR127308 ERR127309 ERR127302 ERR127303</span></div><div class="line"><span class="comment">## 10001           114       156       129       144       175       213</span></div><div class="line"><span class="comment">## 100113389         0         0         0         0         0         0</span></div><div class="line"><span class="comment">## 100113391         0         0         0         0         0         0</span></div><div class="line"><span class="comment">## 100124539         0         0         0         0         0         0</span></div><div class="line"><span class="comment">## 3183           2711      3160      2948      2779        86        98</span></div><div class="line"><span class="comment">##           ERR127304 ERR127305</span></div><div class="line"><span class="comment">## 10001           210       165</span></div><div class="line"><span class="comment">## 100113389         0         0</span></div><div class="line"><span class="comment">## 100113391         0         0</span></div><div class="line"><span class="comment">## 100124539         0         0</span></div><div class="line"><span class="comment">## 3183            158       141</span></div></pre></td></tr></table></figure>
<p>从上面我们可以看出，行标识符是随计数矩阵一起出现的，现在我们绘制出这几个基因的箱线图，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line">par(mfrow=c(<span class="number">2</span>,<span class="number">2</span>))</div><div class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">2</span>:<span class="number">5</span>) &#123;</div><div class="line">  boxplot(assay(hnse)[i,]~hnse$condition, ylab=rowRanges(hnse)$symbol[i])</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc1_summex-lkbo-1.png" alt="plot of chunk lkbo">
<p class="caption">plot of chunk lkbo</p>
</div>
<h1 id="从expressionset提取数据">从ExpressionSet提取数据</h1>
<p>从ExpressionSet中提取数据很容易，但是还需要基于基因组范围查询的阵列探针的子集，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(ALL)</div><div class="line">data(ALL)</div><div class="line">allse = makeSummarizedExperimentFromExpressionSet(ALL)</div><div class="line">allse</div><div class="line"><span class="comment">## class: RangedSummarizedExperiment </span></div><div class="line"><span class="comment">## dim: 12625 128 </span></div><div class="line"><span class="comment">## metadata(3): experimentData annotation protocolData</span></div><div class="line"><span class="comment">## assays(1): exprs</span></div><div class="line"><span class="comment">## rownames(12625): 1000_at 1001_at ... AFFX-YEL021w/URA3_at</span></div><div class="line"><span class="comment">##   AFFX-YEL024w/RIP1_at</span></div><div class="line"><span class="comment">## rowData names(0):</span></div><div class="line"><span class="comment">## colnames(128): 01005 01010 ... 83001 LAL4</span></div><div class="line"><span class="comment">## colData names(21): cod diagnosis ... f.u date.last.seen</span></div><div class="line">rowRanges(allse)</div><div class="line"><span class="comment">## Warning: 'elementLengths' is deprecated.</span></div><div class="line"><span class="comment">## Use 'elementNROWS' instead.</span></div><div class="line"><span class="comment">## See help("Deprecated")</span></div><div class="line"><span class="comment">## GRangesList object of length 12625:</span></div><div class="line"><span class="comment">## $$1000_at </span></div><div class="line"><span class="comment">## GRanges object with 0 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##    seqnames    ranges strand</span></div><div class="line"><span class="comment">##       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$1001_at </span></div><div class="line"><span class="comment">## GRanges object with 0 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##      seqnames ranges strand</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## $$1002_f_at </span></div><div class="line"><span class="comment">## GRanges object with 0 ranges and 0 metadata columns:</span></div><div class="line"><span class="comment">##      seqnames ranges strand</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## ...</span></div><div class="line"><span class="comment">## &lt;12622 more elements&gt;</span></div><div class="line"><span class="comment">## -------</span></div><div class="line"><span class="comment">## seqinfo: no sequences</span></div></pre></td></tr></table></figure>
<h1 id="总结">总结</h1>
<p><code>RangedSummarizedExperiment</code>类实例化了一些Bioconductor数据结构设计的一些关键原则：</p>
<ul>
<li>基于样本特征的数据分析和元数据分析，并将它们以某种方式绑定在一起。</li>
<li>类似于矩阵的子集直接用于分析和样本数据分析。</li>
<li>基于范围的子集设置适用于可通过基因组坐标寻址的分析。</li>
<li>可以在<code>mcol</code>中提供关键分析特征的任意元数据(<code>rowRanges(se)</code>)。</li>
<li>可以通过<code>metadata(se)&lt;-</code> 来添加任何元数据。</li>
</ul>
<p>在后面的内容里我们还将会了解更多的关于<code>SummarizedExperiment</code>改造的一些数据结构，从而用于专门用于RNA-seq的多个阶段处理和数据分析。</p>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc1_summex.html" target="_blank" rel="external">SummarizedExperiment class in depth</a></li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/05/Genomics Data Analysis Series/GDAS005-基因组实验中的数据结构与管理/" rel="next" title="GDAS005-基因组实验中的数据结构与管理">
                <i class="fa fa-chevron-left"></i> GDAS005-基因组实验中的数据结构与管理
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/06/Genomics Data Analysis Series/GDAS016-shiny的交互式操作/" rel="prev" title="GDAS010-Bioconductor中的基因组注释">
                GDAS010-Bioconductor中的基因组注释 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">249</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">118</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#前言"><span class="nav-number">1.</span> <span class="nav-text">前言</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#使用summarizedexperiment来管理bam文件"><span class="nav-number">1.1.</span> <span class="nav-text">使用SummarizedExperiment来管理BAM文件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#summarizedexperiment中的元数据"><span class="nav-number">1.2.</span> <span class="nav-text">SummarizedExperiment中的元数据</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#通过添加输入信息来更有效地生成summarizedexperiment"><span class="nav-number">2.</span> <span class="nav-text">通过添加输入信息来更有效地生成SummarizedExperiment</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#使用元数据定义感兴趣的区域"><span class="nav-number">2.1.</span> <span class="nav-text">使用元数据定义感兴趣的区域</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#定义bam文件的样本信息"><span class="nav-number">2.2.</span> <span class="nav-text">定义BAM文件的样本信息</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#比较读长覆盖区保留元数据"><span class="nav-number">2.3.</span> <span class="nav-text">比较读长覆盖区，保留元数据</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#从expressionset提取数据"><span class="nav-number">3.</span> <span class="nav-text">从ExpressionSet提取数据</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#总结"><span class="nav-number">4.</span> <span class="nav-text">总结</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">4.1.</span> <span class="nav-text">参考资料</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共988.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
