<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang="zh-Hans">
<head>
  <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.3" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.3">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.3">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.3" color="#222">





  <meta name="keywords" content="RNA-seq,Bioconductor," />










<meta name="description" content="NGS的可视化 我们将用以下的包来说明一下如何对NGS数据进行可视化。 123456#biocLite(&quot;pasillaBamSubset&quot;)#biocLite(&quot;TxDb.Dmelanogaster.UCSC.dm3.ensGene&quot;)library(pasillaBamSubset)library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)fl1 &amp;lt;-">
<meta name="keywords" content="RNA-seq,Bioconductor">
<meta property="og:type" content="article">
<meta property="og:title" content="GDAS015-NGS的可视化">
<meta property="og:url" content="http://rvdsd.top/2019/09/15/Genomics Data Analysis Series/GDAS015-NGS的可视化/index.html">
<meta property="og:site_name" content="RVDSD的个人笔记本">
<meta property="og:description" content="NGS的可视化 我们将用以下的包来说明一下如何对NGS数据进行可视化。 123456#biocLite(&quot;pasillaBamSubset&quot;)#biocLite(&quot;TxDb.Dmelanogaster.UCSC.dm3.ensGene&quot;)library(pasillaBamSubset)library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)fl1 &amp;lt;-">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-5-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-6-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-7-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-10-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-11-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-12-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-2.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-1.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-2.png">
<meta property="og:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-3.png">
<meta property="og:updated_time" content="2019-11-18T10:11:27.881Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="GDAS015-NGS的可视化">
<meta name="twitter:description" content="NGS的可视化 我们将用以下的包来说明一下如何对NGS数据进行可视化。 123456#biocLite(&quot;pasillaBamSubset&quot;)#biocLite(&quot;TxDb.Dmelanogaster.UCSC.dm3.ensGene&quot;)library(pasillaBamSubset)library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)fl1 &amp;lt;-">
<meta name="twitter:image" content="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-5-1.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.3',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://rvdsd.top/2019/09/15/Genomics Data Analysis Series/GDAS015-NGS的可视化/"/>





  <title>GDAS015-NGS的可视化 | RVDSD的个人笔记本</title><!-- hexo-inject:begin --><!-- hexo-inject:end -->
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">RVDSD的个人笔记本</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-tags" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://rvdsd.top/2019/09/15/Genomics Data Analysis Series/GDAS015-NGS的可视化/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="RVDSD">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="RVDSD的个人笔记本">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">GDAS015-NGS的可视化</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-09-15T12:00:00+08:00">
                2019-09-15
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Genomics-Data-Analysis-Series/" itemprop="url" rel="index">
                    <span itemprop="name">Genomics Data Analysis Series</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                  <span class="post-meta-divider">|</span>
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1,728
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  9
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="ngs的可视化">NGS的可视化</h2>
<p>我们将用以下的包来说明一下如何对NGS数据进行可视化。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("pasillaBamSubset")</span></div><div class="line"><span class="comment">#biocLite("TxDb.Dmelanogaster.UCSC.dm3.ensGene")</span></div><div class="line"><span class="keyword">library</span>(pasillaBamSubset)</div><div class="line"><span class="keyword">library</span>(TxDb.Dmelanogaster.UCSC.dm3.ensGene)</div><div class="line">fl1 &lt;- untreated1_chr4()</div><div class="line">fl2 &lt;- untreated3_chr4()</div></pre></td></tr></table></figure>
<p>我们将会采用四种方式来查看NGS的数据，分别是IGV，<code>plot</code>,<code>Gviz</code>和<code>ggbio</code>。</p>
<h2 id="igv">IGV</h2>
<p>将这些文献从R library的目录中复制到当前的工作目标。首先将工作目录设置为源文件位置。我们需要使用<code>Rsamtools</code>包来对BAM文件加上索引，然后使用IGV来查看，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div></pre></td><td class="code"><pre><div class="line">file.copy(from=fl1,to=basename(fl1))</div><div class="line"><span class="comment">## [1] FALSE</span></div><div class="line">file.copy(from=fl2,to=basename(fl2))</div><div class="line"><span class="comment">## [1] FALSE</span></div><div class="line"><span class="keyword">library</span>(Rsamtools)</div><div class="line"><span class="comment">## Loading required package: Biostrings</span></div><div class="line"><span class="comment">## Loading required package: XVector</span></div><div class="line">indexBam(basename(fl1))</div><div class="line"><span class="comment">##       untreated1_chr4.bam </span></div><div class="line"><span class="comment">## "untreated1_chr4.bam.bai"</span></div><div class="line">indexBam(basename(fl2))</div><div class="line"><span class="comment">##       untreated3_chr4.bam </span></div><div class="line"><span class="comment">## "untreated3_chr4.bam.bai"</span></div></pre></td></tr></table></figure>
<p>IGV可以在这个网站下载: https://www.broadinstitute.org/igv/home</p>
<p>下载的时候需要提供你的电子邮箱，现在我们使用IGV来查看一个基因，例如<code>igs</code>。</p>
<p>如果无法下载IGV，那么可以使用UCSC的 Genome Browser：http://genome.ucsc.edu/cgi-bin/hgTracks</p>
<p>UCSC Genome Browser是一个很好的资源，有许多涉及基因注释，多个物种的保护，以及ENCODE表观遗传信号。但是，UCSC Genome Browser要求你将基因组文件上传到他们的服务器，或将数据放在公共可用服务器上。如果你处理的数据涉及保密，那就另当别论了。</p>
<h2 id="simple-plot">Simple plot</h2>
<p>现在我们来看一使用<code>plot()</code>函数来绘制基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicRanges)</div></pre></td></tr></table></figure>
<p>注意：如果你使用的是Bioconductor 14，它与R 3.1配合使用，需要加载这个包。如想是Bioconductor 13（它与R 3.0.x配合使用），则不需要加载这个包。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicAlignments)</div><div class="line"><span class="comment">## Loading required package: SummarizedExperiment</span></div></pre></td></tr></table></figure>
<p>我们从文件<code>fl1</code>中读取比对结果(alignments)。然后我们使用<code>coverage</code>函数来计算碱基对的覆盖。然后我们提取与我们感兴趣的基因重叠的覆盖区子集，并将此覆盖率从<code>RleList</code>转换为<code>numeric</code>向量。我们前面提到，<code>Rle</code>对象是一种压缩对象，例如重复数字会被压缩为2个数字，其中一个是重复的数字，另外一个是表示这个数字重复的次数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div></pre></td><td class="code"><pre><div class="line">x &lt;- readGAlignments(fl1)</div><div class="line">xcov &lt;- coverage(x)</div><div class="line">z &lt;- GRanges(<span class="string">"chr4"</span>,IRanges(<span class="number">456500</span>,<span class="number">466000</span>))</div><div class="line"><span class="comment"># Bioconductor 2.14</span></div><div class="line">xcov[z]</div><div class="line"><span class="comment">## RleList of length 1</span></div><div class="line"><span class="comment">## $$chr4</span></div><div class="line"><span class="comment">## integer-Rle of length 9501 with 1775 runs</span></div><div class="line"><span class="comment">##   Lengths: 1252   10   52    4    7    2 ...   10    7   12  392   75 1041</span></div><div class="line"><span class="comment">##   Values :    0    2    3    4    5    6 ...    3    2    1    0    1    0</span></div><div class="line"><span class="comment"># Bioconductor 2.13</span></div><div class="line">xcov$chr4[ranges(z)]</div><div class="line"><span class="comment">## integer-Rle of length 9501 with 1775 runs</span></div><div class="line"><span class="comment">##   Lengths: 1252   10   52    4    7    2 ...   10    7   12  392   75 1041</span></div><div class="line"><span class="comment">##   Values :    0    2    3    4    5    6 ...    3    2    1    0    1    0</span></div><div class="line">xnum &lt;- as.numeric(xcov$chr4[ranges(z)])</div><div class="line">plot(xnum)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-5-1.png" alt="plot of chunk unnamed-chunk-5">
<p class="caption">plot of chunk unnamed-chunk-5</p>
</div>
<p>现在我们读取另外一个文件<code>fl2</code>，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div></pre></td><td class="code"><pre><div class="line">y &lt;- readGAlignmentPairs(fl2)</div><div class="line"><span class="comment">## Warning in .make_GAlignmentPairs_from_GAlignments(gal, strandMode = strandMode, : 63 pairs (0 proper, 63 not proper) were dropped because the seqname</span></div><div class="line"><span class="comment">##   or strand of the alignments in the pair were not concordant.</span></div><div class="line"><span class="comment">##   Note that a GAlignmentPairs object can only hold concordant pairs at the</span></div><div class="line"><span class="comment">##   moment, that is, pairs where the 2 alignments are on the opposite strands</span></div><div class="line"><span class="comment">##   of the same reference sequence.</span></div><div class="line">ycov &lt;- coverage(y)</div><div class="line">ynum &lt;- as.numeric(ycov$chr4[ranges(z)])</div><div class="line">plot(xnum, type=<span class="string">"l"</span>, col=<span class="string">"blue"</span>, lwd=<span class="number">2</span>)</div><div class="line">lines(ynum, col=<span class="string">"red"</span>, lwd=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6">
<p class="caption">plot of chunk unnamed-chunk-6</p>
</div>
<p>我们可以单独放大一个外显子，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div></pre></td><td class="code"><pre><div class="line">plot(xnum, type=<span class="string">"l"</span>, col=<span class="string">"blue"</span>, lwd=<span class="number">2</span>, xlim=c(<span class="number">6200</span>,<span class="number">6600</span>))</div><div class="line">lines(ynum, col=<span class="string">"red"</span>, lwd=<span class="number">2</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-7-1.png" alt="plot of chunk unnamed-chunk-7">
<p class="caption">plot of chunk unnamed-chunk-7</p>
</div>
<h2 id="使用转录本数据库提取感兴趣的基因">使用转录本数据库提取感兴趣的基因</h2>
<p>假设我们想可视化基因igs。我们可以从转录本数据库<code>TxDb.Dmelanogaster.UCSC.dm3.ensGene</code> 中提取它，但我们首先要知道这个基因的Ensembl基因名，现在我们使用前面学到的函数来实现这一功能，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div></pre></td><td class="code"><pre><div class="line"><span class="comment"># biocLite("biomaRt")</span></div><div class="line"><span class="keyword">library</span>(biomaRt)</div><div class="line">m &lt;- useMart(<span class="string">"ensembl"</span>, dataset = <span class="string">"dmelanogaster_gene_ensembl"</span>)</div><div class="line">lf &lt;- listFilters(m)</div><div class="line">lf[grep(<span class="string">"name"</span>, lf$description, ignore.case=<span class="literal">TRUE</span>),]</div><div class="line"><span class="comment">##                             name</span></div><div class="line"><span class="comment">## 1                chromosome_name</span></div><div class="line"><span class="comment">## 17         with_flybasename_gene</span></div><div class="line"><span class="comment">## 19   with_flybasename_transcript</span></div><div class="line"><span class="comment">## 23  with_flybasename_translation</span></div><div class="line"><span class="comment">## 52            external_gene_name</span></div><div class="line"><span class="comment">## 66              flybasename_gene</span></div><div class="line"><span class="comment">## 67        flybasename_transcript</span></div><div class="line"><span class="comment">## 68       flybasename_translation</span></div><div class="line"><span class="comment">## 93                 wikigene_name</span></div><div class="line"><span class="comment">## 107               go_parent_name</span></div><div class="line"><span class="comment">## 208               so_parent_name</span></div><div class="line"><span class="comment">##                                      description</span></div><div class="line"><span class="comment">## 1                                Chromosome name</span></div><div class="line"><span class="comment">## 17                   with FlyBaseName gene ID(s)</span></div><div class="line"><span class="comment">## 19             with FlyBaseName transcript ID(s)</span></div><div class="line"><span class="comment">## 23                with FlyBaseName protein ID(s)</span></div><div class="line"><span class="comment">## 52          Associated Gene Name(s) [e.g. BRCA2]</span></div><div class="line"><span class="comment">## 66           FlyBaseName Gene ID(s) [e.g. cul-2]</span></div><div class="line"><span class="comment">## 67  FlyBaseName Transcript ID(s) [e.g. cul-2-RB]</span></div><div class="line"><span class="comment">## 68     FlyBaseName Protein ID(s) [e.g. cul-2-PB]</span></div><div class="line"><span class="comment">## 93                 WikiGene Name(s) [e.g. Ir21a]</span></div><div class="line"><span class="comment">## 107                             Parent term name</span></div><div class="line"><span class="comment">## 208                             Parent term name</span></div><div class="line">map &lt;- getBM(mart = m,</div><div class="line">  attributes = c(<span class="string">"ensembl_gene_id"</span>, <span class="string">"flybasename_gene"</span>),</div><div class="line">  filters = <span class="string">"flybasename_gene"</span>, </div><div class="line">  values = <span class="string">"lgs"</span>)</div><div class="line">map</div><div class="line"><span class="comment">##   ensembl_gene_id flybasename_gene</span></div><div class="line"><span class="comment">## 1     FBgn0039907              lgs</span></div></pre></td></tr></table></figure>
<p>现在我们提取每个基因的外显子，然后提取igs基因的外显子，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">library</span>(GenomicFeatures)</div><div class="line">grl &lt;- exonsBy(TxDb.Dmelanogaster.UCSC.dm3.ensGene, by=<span class="string">"gene"</span>)</div><div class="line">gene &lt;- grl[[map$ensembl_gene_id[<span class="number">1</span>]]]</div><div class="line">gene</div><div class="line"><span class="comment">## GRanges object with 6 ranges and 2 metadata columns:</span></div><div class="line"><span class="comment">##       seqnames           ranges strand |   exon_id   exon_name</span></div><div class="line"><span class="comment">##          &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;</span></div><div class="line"><span class="comment">##   [1]     chr4 [457583, 459544]      - |     63350        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [2]     chr4 [459601, 459791]      - |     63351        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [3]     chr4 [460074, 462077]      - |     63352        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [4]     chr4 [462806, 463015]      - |     63353        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [5]     chr4 [463490, 463780]      - |     63354        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   [6]     chr4 [463839, 464533]      - |     63355        &lt;NA&gt;</span></div><div class="line"><span class="comment">##   -------</span></div><div class="line"><span class="comment">##   seqinfo: 15 sequences (1 circular) from dm3 genome</span></div></pre></td></tr></table></figure>
<p>最后，我们绘制出这些外显子的范围，如下所示：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">rg &lt;- range(gene)</div><div class="line">plot(c(start(rg), end(rg)), c(0,0), type=&quot;n&quot;, xlab=seqnames(gene)[1], ylab=&quot;&quot;)</div><div class="line">arrows(start(gene),rep(0,length(gene)),</div><div class="line">       end(gene),rep(0,length(gene)),</div><div class="line">       lwd=3, length=.1)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10">
<p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>不过实际上这个基因位于负链，因此我们需要改一下箭头的方向，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">plot(c(start(rg), end(rg)), c(<span class="number">0</span>,<span class="number">0</span>), type=<span class="string">"n"</span>, xlab=seqnames(gene)[<span class="number">1</span>], ylab=<span class="string">""</span>)</div><div class="line">arrows(start(gene),rep(<span class="number">0</span>,length(gene)),</div><div class="line">       end(gene),rep(<span class="number">0</span>,length(gene)),</div><div class="line">       lwd=<span class="number">3</span>, length=<span class="number">.1</span>, </div><div class="line">       code=ifelse(as.character(strand(gene)[<span class="number">1</span>]) == <span class="string">"+"</span>, <span class="number">2</span>, <span class="number">1</span>))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11">
<p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<h2 id="gviz">Gviz</h2>
<p>我们将简要展示两个在Bioconductor中可视化基因组数据的软件包。请注意，这两个包中的任何一个都可以用于绘制多种数据的图形。我们将在这里展示如何像以前一样制作覆盖图(converage plot)：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("Gviz")</span></div><div class="line"><span class="keyword">library</span>(Gviz)</div><div class="line"><span class="comment">## Loading required package: grid</span></div><div class="line">gtrack &lt;- GenomeAxisTrack()</div><div class="line">atrack &lt;- AnnotationTrack(gene, name = <span class="string">"Gene Model"</span>)</div><div class="line">plotTracks(list(gtrack, atrack))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-12-1.png" alt="plot of chunk unnamed-chunk-12">
<p class="caption">plot of chunk unnamed-chunk-12</p>
</div>
<p>提取覆盖。<code>Gviz</code>中的数据最好是是<code>GRanges</code>对象，因此我们需要将<code>RleList</code>转换为<code>GRanges</code>对象，如下所示：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">xgr &lt;- as(xcov, <span class="string">"GRanges"</span>)</div><div class="line">ygr &lt;- as(ycov, <span class="string">"GRanges"</span>)</div><div class="line">dtrack1 &lt;- DataTrack(xgr[xgr %over% z], name = <span class="string">"sample 1"</span>)</div><div class="line">dtrack2 &lt;- DataTrack(ygr[ygr %over% z], name = <span class="string">"sample 2"</span>)</div><div class="line">plotTracks(list(gtrack, atrack, dtrack1, dtrack2))</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-1.png" alt="plot of chunk unnamed-chunk-13">
<p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div></pre></td><td class="code"><pre><div class="line">plotTracks(list(gtrack, atrack, dtrack1, dtrack2), type=<span class="string">"polygon"</span>)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-13-2.png" alt="plot of chunk unnamed-chunk-13">
<p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<h2 id="ggbio">ggbio</h2>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div></pre></td><td class="code"><pre><div class="line"><span class="comment">#biocLite("ggbio")</span></div><div class="line"><span class="keyword">library</span>(ggbio)</div><div class="line"><span class="comment">## Loading required package: ggplot2</span></div><div class="line"><span class="comment">## Warning: replacing previous import by 'BiocGenerics::Position' when loading</span></div><div class="line"><span class="comment">## 'ggbio'</span></div><div class="line"><span class="comment">## Need specific help about ggbio? try mailing </span></div><div class="line"><span class="comment">##  the maintainer or visit http://tengfei.github.com/ggbio/</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">## Attaching package: 'ggbio'</span></div><div class="line"><span class="comment">## The following objects are masked from 'package:ggplot2':</span></div><div class="line"><span class="comment">## </span></div><div class="line"><span class="comment">##     geom_bar, geom_rect, geom_segment, ggsave, stat_bin,</span></div><div class="line"><span class="comment">##     stat_identity, xlim</span></div><div class="line">autoplot(gene)</div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-1.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">autoplot(fl1, which=z)</div><div class="line"><span class="comment">## reading in as Bamfile</span></div><div class="line"><span class="comment">## Parsing raw coverage...</span></div><div class="line"><span class="comment">## Read GAlignments from BamFile...</span></div><div class="line"><span class="comment">## extracting information...</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-2.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<figure class="highlight r"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div></pre></td><td class="code"><pre><div class="line">autoplot(fl2, which=z)</div><div class="line"><span class="comment">## reading in as Bamfile</span></div><div class="line"><span class="comment">## Parsing raw coverage...</span></div><div class="line"><span class="comment">## Read GAlignments from BamFile...</span></div><div class="line"><span class="comment">## extracting information...</span></div></pre></td></tr></table></figure>
<div class="figure">
<img src="http://genomicsclass.github.io/book/pages/figure/bioc2_vizNGS-unnamed-chunk-14-3.png" alt="plot of chunk unnamed-chunk-14">
<p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<h2 id="参考资料">参考资料</h2>
<ol style="list-style-type: decimal">
<li><a href="http://genomicsclass.github.io/book/pages/bioc2_vizNGS.html" target="_blank" rel="external">Visualizing NGS data</a></li>
<li>IGV: https://www.broadinstitute.org/igv/home</li>
<li>Gviz: http://www.bioconductor.org/packages/release/bioc/html/Gviz.html</li>
<li>ggbio:http://www.bioconductor.org/packages/release/bioc/html/ggbio.html</li>
<li>UCSC Genome Browser: zooms and scrolls over chromosomes, showing the work of annotators worldwide: http://genome.ucsc.edu/</li>
<li>Ensembl genome browser: genome databases for vertebrates and other eukaryotic species:http://ensembl.org</li>
<li>Roadmap Epigenome browser: public resource of human epigenomic data: http://www.epigenomebrowser.org http://genomebrowser.wustl.edu/ http://epigenomegateway.wustl.edu/</li>
<li>“Sashimi plots” for RNA-Seq: http://genes.mit.edu/burgelab/miso/docs/sashimi.html</li>
<li>Circos: designed for visualizing genomic data in a cirlce: http://circos.ca/</li>
<li>SeqMonk: a tool to visualise and analyse high throughput mapped sequence data: http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/</li>
</ol>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/RNA-seq/" <i class="fa fa-tag"></i> RNA-seq</a>
          
            <a href="/tags/Bioconductor/" <i class="fa fa-tag"></i> Bioconductor</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/09/14/Genomics Data Analysis Series/GDAS014-使用autoplot和Gviz绘图/" rel="next" title="GDAS014-使用autoplot和Gviz绘图">
                <i class="fa fa-chevron-left"></i> GDAS014-使用autoplot和Gviz绘图
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/09/24/ggplot2/R语言拼图/" rel="prev" title="R语言拼图">
                R语言拼图 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="vcomments"></div>
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">RVDSD</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">246</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">117</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          

          <div class="links-of-author motion-element">
            
          </div>

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#ngs的可视化"><span class="nav-number">1.</span> <span class="nav-text">NGS的可视化</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#igv"><span class="nav-number">2.</span> <span class="nav-text">IGV</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#simple-plot"><span class="nav-number">3.</span> <span class="nav-text">Simple plot</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#使用转录本数据库提取感兴趣的基因"><span class="nav-number">4.</span> <span class="nav-text">使用转录本数据库提取感兴趣的基因</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#gviz"><span class="nav-number">5.</span> <span class="nav-text">Gviz</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#ggbio"><span class="nav-number">6.</span> <span class="nav-text">ggbio</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考资料"><span class="nav-number">7.</span> <span class="nav-text">参考资料</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">RVDSD</span>

  
</div>



<div class="powered-by">
<i class="fa fa-user-md"></i><span id="busuanzi_container_site_pv">
  本站访客数:<span id="busuanzi_value_site_uv"></span>
</span>
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>


<div class="BbeiAn-info">
	<a target="_blank" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=41018102000118" style="color:#909090;text-decoration:none;padding-left:0px;no-repeat left center" rel="nofollow">豫公网安备 41018102000118</a>	  <!--这里将图标作为了背景，以使得能和后面的文字在同一行-->
</div>

  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.3</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共980.9k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.3"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.3"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.3"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.3"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.3"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.3"></script>



  


  




	





  





  












  





  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script><!-- hexo-inject:begin --><!-- Begin: Injected MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config("");
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
      all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });
</script>

<script type="text/javascript" src="custom_mathjax_source">
</script>
<!-- End: Injected MathJax -->
<!-- hexo-inject:end -->
  


  

  

</body>
</html>
